// @wp-g2/styles/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  compiler: true,
  cache: true,
  css: true,
  cx: true,
  injectGlobal: true,
  keyframes: true,
  config: true,
  createTheme: true,
  getIsReducedMotion: true,
  SUPPORTED_COLORS: true
};
Object.defineProperty(exports, "compiler", {
  enumerable: true,
  get: function get() {
    return _system.compiler;
  }
});
Object.defineProperty(exports, "cache", {
  enumerable: true,
  get: function get() {
    return _styleSystem.cache;
  }
});
Object.defineProperty(exports, "css", {
  enumerable: true,
  get: function get() {
    return _styleSystem.css;
  }
});
Object.defineProperty(exports, "cx", {
  enumerable: true,
  get: function get() {
    return _styleSystem.cx;
  }
});
Object.defineProperty(exports, "injectGlobal", {
  enumerable: true,
  get: function get() {
    return _styleSystem.injectGlobal;
  }
});
Object.defineProperty(exports, "keyframes", {
  enumerable: true,
  get: function get() {
    return _styleSystem.keyframes;
  }
});
Object.defineProperty(exports, "config", {
  enumerable: true,
  get: function get() {
    return _theme.config;
  }
});
Object.defineProperty(exports, "createTheme", {
  enumerable: true,
  get: function get() {
    return _theme.createTheme;
  }
});
Object.defineProperty(exports, "getIsReducedMotion", {
  enumerable: true,
  get: function get() {
    return _theme.getIsReducedMotion;
  }
});
Object.defineProperty(exports, "SUPPORTED_COLORS", {
  enumerable: true,
  get: function get() {
    return _theme.SUPPORTED_COLORS;
  }
});

var _core = require("./core");

Object.keys(_core).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _core[key];
    }
  });
});

var _system = require("./system");

var _styleSystem = require("./style-system");

var _styled = require("./styled");

Object.keys(_styled).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _styled[key];
    }
  });
});

var _hooks = require("./hooks");

Object.keys(_hooks).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _hooks[key];
    }
  });
});

var _mixins = require("./mixins");

Object.keys(_mixins).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _mixins[key];
    }
  });
});

var _namespaces = require("./namespaces");

Object.keys(_namespaces).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _namespaces[key];
    }
  });
});

var _theme = require("./theme");

var _ui = require("./presets/ui");

Object.keys(_ui).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _ui[key];
    }
  });
});

var _components = require("./components");

Object.keys(_components).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _components[key];
    }
  });
});
