// @wp-g2/styles/dist/cjs/style-system.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "css", {
  enumerable: true,
  get: function get() {
    return _css.css;
  }
});
exports.sheet = exports.merge = exports.keyframes = exports.injectGlobal = exports.hydrate = exports.getRegisteredStyles = exports.flush = exports.cx = exports.cache = exports.breakpoints = void 0;

var _system = require("./system");

var _css = require("./css");

var breakpoints = _system.compiler.breakpoints,
    cache = _system.compiler.cache,
    cx = _system.compiler.cx,
    flush = _system.compiler.flush,
    getRegisteredStyles = _system.compiler.getRegisteredStyles,
    hydrate = _system.compiler.hydrate,
    injectGlobal = _system.compiler.injectGlobal,
    keyframes = _system.compiler.keyframes,
    merge = _system.compiler.merge,
    sheet = _system.compiler.sheet;
exports.sheet = sheet;
exports.merge = merge;
exports.keyframes = keyframes;
exports.injectGlobal = injectGlobal;
exports.hydrate = hydrate;
exports.getRegisteredStyles = getRegisteredStyles;
exports.flush = flush;
exports.cx = cx;
exports.cache = cache;
exports.breakpoints = breakpoints;
