// @wp-g2/components/dist/es/ButtonGroup/__stories__/ButtonGroup.stories.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { formatBold, formatItalic, formatUnderline } from '@wordpress/icons';
import React from 'react';
import { Button, FormGroup, ListGroup, ListGroupHeader, ListGroups } from '../../index';
import { ButtonGroup } from '../index';
export default {
  component: ButtonGroup,
  title: 'Components/ButtonGroup'
};
export var _default = function () {
  var _React$useState = React.useState('bold'),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      value = _React$useState2[0],
      setValue = _React$useState2[1];

  return /*#__PURE__*/React.createElement(ListGroups, null, /*#__PURE__*/React.createElement(ListGroup, null, /*#__PURE__*/React.createElement(ListGroupHeader, null, "Default"), /*#__PURE__*/React.createElement(FormGroup, null, /*#__PURE__*/React.createElement(ButtonGroup, {
    onChange: setValue,
    value: value
  }, /*#__PURE__*/React.createElement(Button, {
    icon: formatBold,
    value: "bold"
  }), /*#__PURE__*/React.createElement(Button, {
    icon: formatItalic,
    value: "italic"
  }), /*#__PURE__*/React.createElement(Button, {
    icon: formatUnderline,
    value: "underline"
  })))), /*#__PURE__*/React.createElement(ListGroup, null, /*#__PURE__*/React.createElement(ListGroupHeader, null, "Segmented"), /*#__PURE__*/React.createElement(FormGroup, null, /*#__PURE__*/React.createElement(ButtonGroup, {
    onChange: setValue,
    segmented: true,
    value: value
  }, /*#__PURE__*/React.createElement(Button, {
    icon: formatBold,
    value: "bold"
  }), /*#__PURE__*/React.createElement(Button, {
    icon: formatItalic,
    value: "italic"
  }), /*#__PURE__*/React.createElement(Button, {
    icon: formatUnderline,
    value: "underline"
  })))), /*#__PURE__*/React.createElement(ListGroup, null, /*#__PURE__*/React.createElement(ListGroupHeader, null, "Expanded"), /*#__PURE__*/React.createElement(FormGroup, null, /*#__PURE__*/React.createElement(ButtonGroup, {
    expanded: true,
    onChange: setValue,
    value: value
  }, /*#__PURE__*/React.createElement(Button, {
    icon: formatBold,
    value: "bold"
  }), /*#__PURE__*/React.createElement(Button, {
    icon: formatItalic,
    value: "italic"
  }), /*#__PURE__*/React.createElement(Button, {
    icon: formatUnderline,
    value: "underline"
  })))), /*#__PURE__*/React.createElement(ListGroup, null, /*#__PURE__*/React.createElement(ListGroupHeader, null, "Segmented + Expanded"), /*#__PURE__*/React.createElement(FormGroup, null, /*#__PURE__*/React.createElement(ButtonGroup, {
    expanded: true,
    onChange: setValue,
    segmented: true,
    value: value
  }, /*#__PURE__*/React.createElement(Button, {
    icon: formatBold,
    value: "bold"
  }), /*#__PURE__*/React.createElement(Button, {
    icon: formatItalic,
    value: "italic"
  }), /*#__PURE__*/React.createElement(Button, {
    icon: formatUnderline,
    value: "underline"
  })))));
};
export var _expanded = function () {
  var _React$useState3 = React.useState('bold'),
      _React$useState4 = _slicedToArray(_React$useState3, 2),
      value = _React$useState4[0],
      setValue = _React$useState4[1];

  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(ButtonGroup, {
    expanded: true,
    onChange: setValue,
    value: value
  }, /*#__PURE__*/React.createElement(Button, {
    icon: formatBold,
    value: "bold"
  }), /*#__PURE__*/React.createElement(Button, {
    icon: formatItalic,
    value: "italic"
  }), /*#__PURE__*/React.createElement(Button, {
    icon: formatUnderline,
    value: "underline"
  })));
};
export var _segmented = function () {
  var _React$useState5 = React.useState('bold'),
      _React$useState6 = _slicedToArray(_React$useState5, 2),
      value = _React$useState6[0],
      setValue = _React$useState6[1];

  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(ButtonGroup, {
    onChange: setValue,
    segmented: true,
    value: value
  }, /*#__PURE__*/React.createElement(Button, {
    icon: formatBold,
    value: "bold"
  }), /*#__PURE__*/React.createElement(Button, {
    icon: formatItalic,
    value: "italic"
  }), /*#__PURE__*/React.createElement(Button, {
    icon: formatUnderline,
    value: "underline"
  })));
};
