// @wordpress/block-editor/build-module/components/font-appearance-control/index.js
import { createElement } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { CustomSelectControl } from '@wordpress/components';
import { useMemo } from '@wordpress/element';
import { __, sprintf } from '@wordpress/i18n';
var FONT_STYLES = [{
  name: __('Regular'),
  value: 'normal'
}, {
  name: __('Italic'),
  value: 'italic'
}];
var FONT_WEIGHTS = [{
  name: __('Thin'),
  value: '100'
}, {
  name: __('Extra Light'),
  value: '200'
}, {
  name: __('Light'),
  value: '300'
}, {
  name: __('Regular'),
  value: '400'
}, {
  name: __('Medium'),
  value: '500'
}, {
  name: __('Semi Bold'),
  value: '600'
}, {
  name: __('Bold'),
  value: '700'
}, {
  name: __('Extra Bold'),
  value: '800'
}, {
  name: __('Black'),
  value: '900'
}];
/**
 * Control to display unified font style and weight options.
 *
 * @param  {Object}   props Component props.
 * @return {WPElement}      Font appearance control.
 */

export default function FontAppearanceControl(props) {
  var _onChange = props.onChange,
      _props$hasFontStyles = props.hasFontStyles,
      hasFontStyles = _props$hasFontStyles === void 0 ? true : _props$hasFontStyles,
      _props$hasFontWeights = props.hasFontWeights,
      hasFontWeights = _props$hasFontWeights === void 0 ? true : _props$hasFontWeights,
      _props$value = props.value,
      fontStyle = _props$value.fontStyle,
      fontWeight = _props$value.fontWeight;
  var hasStylesOrWeights = hasFontStyles || hasFontWeights;
  var defaultOption = {
    key: 'default',
    name: __('Default'),
    style: {
      fontStyle: undefined,
      fontWeight: undefined
    }
  }; // Combines both font style and weight options into a single dropdown.

  var combineOptions = function combineOptions() {
    var combinedOptions = [defaultOption];
    FONT_STYLES.forEach(function (_ref) {
      var styleName = _ref.name,
          styleValue = _ref.value;
      FONT_WEIGHTS.forEach(function (_ref2) {
        var weightName = _ref2.name,
            weightValue = _ref2.value;
        var optionName = styleValue === 'normal' ? weightName : sprintf(
        /* translators: 1: Font weight name. 2: Font style name. */
        __('%1$s %2$s'), weightName, styleName);
        combinedOptions.push({
          key: "".concat(styleValue, "-").concat(weightValue),
          name: optionName,
          style: {
            fontStyle: styleValue,
            fontWeight: weightValue
          }
        });
      });
    });
    return combinedOptions;
  }; // Generates select options for font styles only.


  var styleOptions = function styleOptions() {
    var combinedOptions = [defaultOption];
    FONT_STYLES.forEach(function (_ref3) {
      var name = _ref3.name,
          value = _ref3.value;
      combinedOptions.push({
        key: value,
        name: name,
        style: {
          fontStyle: value,
          fontWeight: undefined
        }
      });
    });
    return combinedOptions;
  }; // Generates select options for font weights only.


  var weightOptions = function weightOptions() {
    var combinedOptions = [defaultOption];
    FONT_WEIGHTS.forEach(function (_ref4) {
      var name = _ref4.name,
          value = _ref4.value;
      combinedOptions.push({
        key: value,
        name: name,
        style: {
          fontStyle: undefined,
          fontWeight: value
        }
      });
    });
    return combinedOptions;
  }; // Map font styles and weights to select options.


  var selectOptions = useMemo(function () {
    if (hasFontStyles && hasFontWeights) {
      return combineOptions();
    }

    return hasFontStyles ? styleOptions() : weightOptions();
  }, [props.options]); // Find current selection by comparing font style & weight against options.

  var currentSelection = selectOptions.find(function (option) {
    return option.style.fontStyle === fontStyle && option.style.fontWeight === fontWeight;
  }); // Adjusts field label in case either styles or weights are disabled.

  var getLabel = function getLabel() {
    if (!hasFontStyles) {
      return __('Font weight');
    }

    if (!hasFontWeights) {
      return __('Font style');
    }

    return __('Appearance');
  };

  return createElement("fieldset", {
    className: "components-font-appearance-control"
  }, hasStylesOrWeights && createElement(CustomSelectControl, {
    className: "components-font-appearance-control__select",
    label: getLabel(),
    options: selectOptions,
    value: currentSelection,
    onChange: function onChange(_ref5) {
      var selectedItem = _ref5.selectedItem;
      return _onChange(selectedItem.style);
    }
  }));
}
