// @wp-g2/components/dist/cjs/TextInput/TextInput.styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "scrollableScrollbar", {
  enumerable: true,
  get: function get() {
    return _Scrollable.scrollableScrollbar;
  }
});
exports.globalDraggableY = exports.globalDraggableX = exports.SteppersDown = exports.SteppersUp = exports.StepperButton = exports.Steppers = exports.SpinnerArrowDown = exports.SpinnerArrowUp = exports.SpinnerArrow = exports.Spinner = exports.SpinnerWrapper = exports.resizable = exports.small = exports.large = exports.Input = exports.inputFontSize = exports.inputMultiline = exports.multiline = exports.focus = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _styles = require("@wp-g2/styles");

var _Scrollable = require("../Scrollable/Scrollable.styles");

function _templateObject19() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t* {\n\t\tcursor: ns-resize;\n\t}\n"]);

  _templateObject19 = function () {
    return data;
  };

  return data;
}

function _templateObject18() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t* {\n\t\tcursor: ew-resize;\n\t}\n"]);

  _templateObject18 = function () {
    return data;
  };

  return data;
}

function _templateObject17() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n"]);

  _templateObject17 = function () {
    return data;
  };

  return data;
}

function _templateObject16() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n"]);

  _templateObject16 = function () {
    return data;
  };

  return data;
}

function _templateObject15() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\tborder: 1px solid transparent;\n\tcolor: ", ";\n\tcursor: pointer;\n\tdisplay: block;\n\toutline: none;\n\tpadding: 3px;\n\n\t&:hover {\n\t\tbackground-color: ", ";\n\t}\n\n\t&:focus {\n\t\tbackground-color: ", ";\n\t\tborder-color: ", ";\n\t}\n\n\t&[disabled] {\n\t\tpointer-events: none;\n\t}\n"]);

  _templateObject15 = function () {
    return data;
  };

  return data;
}

function _templateObject14() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\tmargin: 0 -2px 0 0 !important;\n\topacity: 1;\n"]);

  _templateObject14 = function () {
    return data;
  };

  return data;
}

function _templateObject13() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\tmargin-top: -1px;\n"]);

  _templateObject13 = function () {
    return data;
  };

  return data;
}

function _templateObject12() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\tmargin-bottom: -1px;\n"]);

  _templateObject12 = function () {
    return data;
  };

  return data;
}

function _templateObject11() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbackground-color: transparent;\n\tborder-radius: ", ";\n\tcolor: ", ";\n\tcursor: pointer;\n\tpadding: 0 2px;\n\n\t&:hover:active {\n\t\tbackground-color: ", ";\n\t}\n"]);

  _templateObject11 = function () {
    return data;
  };

  return data;
}

function _templateObject10() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\theight: 24px;\n\tmargin: 0 -6px 0 0 !important;\n\topacity: 0.6;\n\tuser-select: none;\n"]);

  _templateObject10 = function () {
    return data;
  };

  return data;
}

function _templateObject9() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\talign-items: center;\n\tdisplay: flex;\n\theight: 100%;\n\tjustify-content: center;\n\tmargin: 0 !important;\n\tmin-height: 0;\n"]);

  _templateObject9 = function () {
    return data;
  };

  return data;
}

function _templateObject8() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tresize: vertical;\n"]);

  _templateObject8 = function () {
    return data;
  };

  return data;
}

function _templateObject7() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tmin-height: calc(", " - 2px);\n\tpadding-bottom: calc((", " - 2px - 18px) / 2);\n\tpadding-top: calc((", " - 2px - 18px) / 2);\n"]);

  _templateObject7 = function () {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tmin-height: calc(", " - 2px);\n\tpadding-bottom: calc((", " - 2px - 18px) / 2);\n\tpadding-top: calc((", " - 2px - 18px) / 2);\n"]);

  _templateObject6 = function () {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tappearance: none;\n\tbackground: transparent;\n\tborder: none;\n\tborder-radius: ", ";\n\tbox-shadow: none;\n\tbox-sizing: border-box;\n\tcolor: ", ";\n\tdisplay: block;\n\tflex: 1;\n\tline-height: 18px;\n\tmin-height: calc(", " - 2px);\n\toutline: none;\n\tpadding: 0;\n\tpadding-bottom: calc((", " - 2px - 18px) / 2);\n\tpadding-top: calc((", " - 2px - 18px) / 2);\n\tresize: none;\n\ttext-align: left;\n\twidth: 100%;\n\n\t", ";\n\n\t&::-webkit-outer-spin-button,\n\t&::-webkit-inner-spin-button {\n\t\t-webkit-appearance: none;\n\t\tmargin: 0;\n\t}\n\n\t/* Firefox */\n\t&[type='number'] {\n\t\t-moz-appearance: textfield;\n\t}\n"]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tfont-size: ", ";\n\n\t@media (min-width: 36em) {\n\t\tfont-size: ", ";\n\t}\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tpadding-left: 8px;\n\tpadding-right: 8px;\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tpadding-left: 0;\n\tpadding-right: 0;\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tborder-color: ", ";\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var focus = (0, _styles.css)(_templateObject(), _styles.ui.color.admin);
exports.focus = focus;
var multiline = (0, _styles.css)(_templateObject2());
exports.multiline = multiline;
var inputMultiline = (0, _styles.css)(_templateObject3());
exports.inputMultiline = inputMultiline;
var inputFontSize = (0, _styles.css)(_templateObject4(), _styles.ui.get('fontSizeInputMobile'), _styles.ui.get('fontSize'));
exports.inputFontSize = inputFontSize;
var Input = (0, _styles.css)(_templateObject5(), _styles.ui.get('controlBorderRadius'), _styles.ui.color.text, _styles.ui.get('controlHeight'), _styles.ui.get('controlHeight'), _styles.ui.get('controlHeight'), inputFontSize);
exports.Input = Input;
var large = (0, _styles.css)(_templateObject6(), _styles.ui.get('controlHeightLarge'), _styles.ui.get('controlHeightLarge'), _styles.ui.get('controlHeightLarge'));
exports.large = large;
var small = (0, _styles.css)(_templateObject7(), _styles.ui.get('controlHeightSmall'), _styles.ui.get('controlHeightSmall'), _styles.ui.get('controlHeightSmall'));
exports.small = small;
var resizable = (0, _styles.css)(_templateObject8());
exports.resizable = resizable;
var SpinnerWrapper = (0, _styles.css)(_templateObject9());
exports.SpinnerWrapper = SpinnerWrapper;
var Spinner = (0, _styles.css)(_templateObject10());
exports.Spinner = Spinner;
var SpinnerArrow = (0, _styles.css)(_templateObject11(), _styles.ui.get('controlBorderRadius'), _styles.ui.get('colorText'), _styles.ui.get('controlBackgroundColorHover'));
exports.SpinnerArrow = SpinnerArrow;
var SpinnerArrowUp = (0, _styles.css)(_templateObject12(), SpinnerArrow);
exports.SpinnerArrowUp = SpinnerArrowUp;
var SpinnerArrowDown = (0, _styles.css)(_templateObject13(), SpinnerArrow);
exports.SpinnerArrowDown = SpinnerArrowDown;
var Steppers = (0, _styles.css)(_templateObject14(), Spinner);
exports.Steppers = Steppers;
var StepperButton = (0, _styles.css)(_templateObject15(), SpinnerArrow, _styles.ui.get('controlInnerControltextColor'), _styles.ui.get('controlBackgroundColorHover'), _styles.ui.get('controlBackgroundColorHover'), _styles.ui.get('colorAdmin'));
exports.StepperButton = StepperButton;
var SteppersUp = (0, _styles.css)(_templateObject16(), StepperButton);
exports.SteppersUp = SteppersUp;
var SteppersDown = (0, _styles.css)(_templateObject17(), StepperButton);
exports.SteppersDown = SteppersDown;
var globalDraggableX = (0, _styles.css)(_templateObject18());
exports.globalDraggableX = globalDraggableX;
var globalDraggableY = (0, _styles.css)(_templateObject19());
exports.globalDraggableY = globalDraggableY;
