// @wp-g2/components/dist/cjs/TextInput/TextInputArrows.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _icons = require("@wordpress/icons");

var _styles = require("@wp-g2/styles");

var _lodash = require("lodash");

var _react = _interopRequireWildcard(require("react"));

var _Icon = require("../Icon");

var _View = require("../View");

var _VStack = require("../VStack");

var styles = _interopRequireWildcard(require("./TextInput.styles"));

/**
 * @typedef Props
 * @property {import('./useTextInputState').TextInputState} __store
 * @property {(int: number) => void} decrement
 * @property {(int: number) => void} increment
 */

/**
 *
 * @param {Props} props
 * @param {import('react').Ref<any>} forwardedRef
 */
function TextInputArrows(props, forwardedRef) {
  var decrement = props.decrement,
      dragHandlersRef = props.dragHandlersRef,
      increment = props.increment;
  var dragHandlers = dragHandlersRef.current;
  return /*#__PURE__*/_react.default.createElement(_View.View, {
    className: styles.SpinnerWrapper
  }, /*#__PURE__*/_react.default.createElement(_VStack.VStack, Object.assign({}, dragHandlers, {
    className: styles.Spinner,
    expanded: true,
    spacing: 0
  }, _styles.ui.$('TextInputArrows'), {
    ref: forwardedRef
  }), /*#__PURE__*/_react.default.createElement(UpDownArrows, {
    onDecrement: decrement,
    onIncrement: increment
  })));
}

var _UpDownArrows = function (_ref) {
  var _ref$onIncrement = _ref.onIncrement,
      onIncrement = _ref$onIncrement === void 0 ? _lodash.noop : _ref$onIncrement,
      _ref$onDecrement = _ref.onDecrement,
      onDecrement = _ref$onDecrement === void 0 ? _lodash.noop : _ref$onDecrement;

  /** @type {import('react').MutableRefObject<number | undefined>} */
  var timeoutRef = (0, _react.useRef)();
  var timeoutDurationStart = 500;
  var timeoutDurationEnd = 20;
  var timeoutDurationRef = (0, _react.useRef)(timeoutDurationStart);
  var handleOnClearTimers = (0, _react.useCallback)(function () {
    if (timeoutRef.current) {
      clearTimeout(timeoutRef.current);
    }

    timeoutDurationRef.current = timeoutDurationStart;
  }, []);
  var handleOnMouseDownIncrement = (0, _react.useCallback)(function (event) {
    if (event) {
      event.preventDefault();
    }

    timeoutRef.current = setTimeout(function () {
      onIncrement();
      timeoutDurationRef.current = (0, _lodash.clamp)(timeoutDurationRef.current / 2, timeoutDurationEnd, timeoutDurationStart);
      handleOnMouseDownIncrement();
    }, timeoutDurationRef.current);
  }, [onIncrement]);
  var handleOnMouseDownDecrement = (0, _react.useCallback)(function (event) {
    if (event) {
      event.preventDefault();
    }

    timeoutRef.current = setTimeout(function () {
      onDecrement();
      timeoutDurationRef.current = (0, _lodash.clamp)(timeoutDurationRef.current / 2, timeoutDurationEnd, timeoutDurationStart);
      handleOnMouseDownDecrement();
    }, timeoutDurationRef.current);
  }, [onDecrement]);
  (0, _react.useEffect)(function () {
    return function () {
      return handleOnClearTimers();
    };
  }, [handleOnClearTimers]);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_Icon.Icon, Object.assign({
    className: styles.SpinnerArrowUp,
    onClick: onIncrement,
    onMouseDown: handleOnMouseDownIncrement,
    onMouseLeave: handleOnClearTimers,
    onMouseUp: handleOnClearTimers,
    tabIndex: -1
  }, _styles.ui.$('TextInputArrowUp'), {
    icon: _icons.chevronUp,
    size: 12,
    width: 16
  })), /*#__PURE__*/_react.default.createElement(_Icon.Icon, Object.assign({
    className: styles.SpinnerArrowDown,
    onClick: onDecrement,
    onMouseDown: handleOnMouseDownDecrement,
    onMouseLeave: handleOnClearTimers,
    onMouseUp: handleOnClearTimers,
    tabIndex: -1
  }, _styles.ui.$('TextInputArrowDown'), {
    icon: _icons.chevronDown,
    size: 12,
    width: 16
  })));
};

var UpDownArrows = /*#__PURE__*/_react.default.memo(_UpDownArrows);

var _default = /*#__PURE__*/_react.default.memo( /*#__PURE__*/_react.default.forwardRef(TextInputArrows));

exports.default = _default;
