// @wp-g2/components/dist/cjs/VStack/__stories__/VStack.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _HStack = require("../../HStack");

var _index = require("../../index");

var _index2 = require("../index");

var _default2 = {
  component: _index2.VStack,
  title: 'Components/VStack'
};
exports.default = _default2;

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(_index.View, {
    css: {
      margin: 'auto',
      maxWidth: 960
    }
  }, /*#__PURE__*/_react.default.createElement(_index2.VStack, null, /*#__PURE__*/_react.default.createElement(_HStack.HStack, {
    justify: "space-between"
  }, /*#__PURE__*/_react.default.createElement(_index.View, null, /*#__PURE__*/_react.default.createElement(_index.Heading, {
    size: 3
  }, "Components: HStack, VStack, ZStack", ' ', /*#__PURE__*/_react.default.createElement(_index.View, {
    as: "span",
    css: {
      fontSize: 'inherit',
      opacity: 0.5
    }
  }, "#22"))), /*#__PURE__*/_react.default.createElement(_index.View, null, /*#__PURE__*/_react.default.createElement(_HStack.HStack, null, /*#__PURE__*/_react.default.createElement(_index.Button, null, "Edit"), /*#__PURE__*/_react.default.createElement(_index.Button, {
    variant: "primary"
  }, "New Issue")))), /*#__PURE__*/_react.default.createElement(_HStack.HStack, {
    justify: "flex-start"
  }, /*#__PURE__*/_react.default.createElement(_index.Badge, {
    color: "green"
  }, "Open"), /*#__PURE__*/_react.default.createElement(_index.Text, {
    weight: "bold"
  }, "ItsJonQ"), /*#__PURE__*/_react.default.createElement(_index.Text, null, "opened this issue 2 days ago \xB7 1 comment"))));
};

exports._default = _default;
