// @wp-g2/styles/dist/cjs/theme/tokens/colors.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DARK_MODE_RGBA_COLORS = exports.DARK_MODE_COLORS = exports.BACKGROUND_COLOR_PROPS = exports.G2_COLORS = exports.WORDPRESS_COLORS = exports.BLUE_COLORS = exports.PURPLE_COLORS = exports.GREEN_COLORS = exports.YELLOW_COLORS = exports.ORANGE_COLORS = exports.RED_COLORS = exports.LIGHT_GRAY_COLORS = exports.DARK_GRAY_COLORS = exports.CORE_BLUE_COLORS = exports.CORE_GREEN_COLORS = exports.CORE_YELLOW_COLORS = exports.CORE_ORANGE_COLORS = exports.CORE_RED_COLORS = exports.CORE_LIGHT_GRAY_COLORS = exports.CORE_DARK_GRAY_COLORS = exports.CORE_PURPLE_COLORS = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _createStyles = require("@wp-g2/create-styles");

var _utils = require("../utils");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var CORE_PURPLE_COLORS = {
  purple100: '#f3f1f8',
  purple300: '#B4A8D2',
  purple500: '#826eb4',
  purple700: '#4e426c',
  purple900: '#342c48'
};
exports.CORE_PURPLE_COLORS = CORE_PURPLE_COLORS;
var CORE_DARK_GRAY_COLORS = {
  darkGray100: '#8F98A1',
  darkGray300: '#6C7781',
  darkGray500: '#555D66',
  darkGray700: '#32373C',
  darkGray900: '#191E23'
};
exports.CORE_DARK_GRAY_COLORS = CORE_DARK_GRAY_COLORS;
var CORE_LIGHT_GRAY_COLORS = {
  lightGray100: '#fbfbfc',
  lightGray300: '#edeff0',
  lightGray500: '#e2e4e7',
  lightGray700: '#ccd0d4',
  lightGray900: '#a2aab2'
};
exports.CORE_LIGHT_GRAY_COLORS = CORE_LIGHT_GRAY_COLORS;
var CORE_RED_COLORS = {
  red100: '#fcebeb',
  red300: '#ea8484',
  red500: '#dc3232',
  red700: '#b02828',
  red900: '#841e1e'
};
exports.CORE_RED_COLORS = CORE_RED_COLORS;
var CORE_ORANGE_COLORS = {
  orange100: '#fef1ea',
  orange300: '#f9a87e',
  orange500: '#F56E28',
  orange700: '#ca4a1f',
  orange900: '#aa3e1a'
};
exports.CORE_ORANGE_COLORS = CORE_ORANGE_COLORS;
var CORE_YELLOW_COLORS = {
  yellow100: '#fff8e6',
  yellow300: '#ffd566',
  yellow500: '#ffb900',
  yellow700: '#ee8e0d',
  yellow900: '#dd631a'
};
exports.CORE_YELLOW_COLORS = CORE_YELLOW_COLORS;
var CORE_GREEN_COLORS = {
  green100: '#edf8ee',
  green300: '#90d296',
  green500: '#46b450',
  green700: '#328540',
  green900: '#25612f'
};
exports.CORE_GREEN_COLORS = CORE_GREEN_COLORS;
var CORE_BLUE_COLORS = {
  blue100: '#e6f6fb',
  blue300: '#66c6e4',
  blue500: '#00a0d2',
  blue700: '#0085ba',
  blue900: '#0072A8'
};
exports.CORE_BLUE_COLORS = CORE_BLUE_COLORS;

var DARK_GRAY_COLORS = _objectSpread(_objectSpread(_objectSpread({}, CORE_DARK_GRAY_COLORS), (0, _utils.createTextColors)(CORE_DARK_GRAY_COLORS)), (0, _utils.createRgbaColors)(CORE_DARK_GRAY_COLORS));

exports.DARK_GRAY_COLORS = DARK_GRAY_COLORS;

var LIGHT_GRAY_COLORS = _objectSpread(_objectSpread(_objectSpread({}, CORE_LIGHT_GRAY_COLORS), (0, _utils.createTextColors)(CORE_LIGHT_GRAY_COLORS)), (0, _utils.createRgbaColors)(CORE_LIGHT_GRAY_COLORS));

exports.LIGHT_GRAY_COLORS = LIGHT_GRAY_COLORS;

var RED_COLORS = _objectSpread(_objectSpread(_objectSpread({}, CORE_RED_COLORS), (0, _utils.createTextColors)(CORE_RED_COLORS)), (0, _utils.createRgbaColors)(CORE_RED_COLORS));

exports.RED_COLORS = RED_COLORS;

var ORANGE_COLORS = _objectSpread(_objectSpread(_objectSpread({}, CORE_ORANGE_COLORS), (0, _utils.createTextColors)(CORE_ORANGE_COLORS)), (0, _utils.createRgbaColors)(CORE_ORANGE_COLORS));

exports.ORANGE_COLORS = ORANGE_COLORS;

var YELLOW_COLORS = _objectSpread(_objectSpread(_objectSpread({}, CORE_YELLOW_COLORS), (0, _utils.createTextColors)(CORE_YELLOW_COLORS)), (0, _utils.createRgbaColors)(CORE_YELLOW_COLORS));

exports.YELLOW_COLORS = YELLOW_COLORS;

var GREEN_COLORS = _objectSpread(_objectSpread(_objectSpread({}, CORE_GREEN_COLORS), (0, _utils.createTextColors)(CORE_GREEN_COLORS)), (0, _utils.createRgbaColors)(CORE_GREEN_COLORS));

exports.GREEN_COLORS = GREEN_COLORS;

var PURPLE_COLORS = _objectSpread(_objectSpread(_objectSpread({}, CORE_PURPLE_COLORS), (0, _utils.createTextColors)(CORE_PURPLE_COLORS)), (0, _utils.createRgbaColors)(CORE_PURPLE_COLORS));

exports.PURPLE_COLORS = PURPLE_COLORS;

var BLUE_COLORS = _objectSpread(_objectSpread(_objectSpread({}, CORE_BLUE_COLORS), (0, _utils.createTextColors)(CORE_BLUE_COLORS)), (0, _utils.createRgbaColors)(CORE_BLUE_COLORS));

exports.BLUE_COLORS = BLUE_COLORS;

var WORDPRESS_COLORS = _objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread({}, DARK_GRAY_COLORS), LIGHT_GRAY_COLORS), RED_COLORS), ORANGE_COLORS), YELLOW_COLORS), GREEN_COLORS), PURPLE_COLORS), BLUE_COLORS);

exports.WORDPRESS_COLORS = WORDPRESS_COLORS;
var G2_COLORS = {
  black: '#000000',
  blueberry: '#3858E9',
  blueberryDark: '#1D35B4',
  greens: '#33F078',
  grey: '#40464D',
  greyBlack: '#1E1E1E',
  lightBlue: '#33F078',
  lightGrey: '#40464D',
  lighterGrey: '#dddddd',
  pomegrade: '#E26F56',
  wordpressBlue: '#007cba',
  white: '#ffffff'
};
exports.G2_COLORS = G2_COLORS;
var BACKGROUND_COLOR_PROPS = {
  colorBackgroundBlue: (0, _createStyles.get)('blueRgba10'),
  colorBackgroundBlueText: (0, _createStyles.get)('blue900'),
  colorBackgroundDarkGray: (0, _createStyles.get)('darkGrayRgba10'),
  colorBackgroundDarkGrayText: (0, _createStyles.get)('darkGray900'),
  colorBackgroundGreen: (0, _createStyles.get)('greenRgba10'),
  colorBackgroundGreenText: (0, _createStyles.get)('green900'),
  colorBackgroundLightGray: (0, _createStyles.get)('lightGrayRgba10'),
  colorBackgroundLightGrayText: (0, _createStyles.get)('lightGray900'),
  colorBackgroundOrange: (0, _createStyles.get)('orangeRgba10'),
  colorBackgroundOrangeText: (0, _createStyles.get)('orange900'),
  colorBackgroundPurple: (0, _createStyles.get)('purpleRgba10'),
  colorBackgroundPurpleText: (0, _createStyles.get)('purple900'),
  colorBackgroundRed: (0, _createStyles.get)('redRgba10'),
  colorBackgroundRedText: (0, _createStyles.get)('red900'),
  colorBackgroundYellow: (0, _createStyles.get)('yellowRgba10'),
  colorBackgroundYellowText: (0, _createStyles.get)('yellow900')
};
exports.BACKGROUND_COLOR_PROPS = BACKGROUND_COLOR_PROPS;
var DARK_MODE_COLORS = {
  colorBackgroundBlue: (0, _createStyles.get)('blueRgba20'),
  colorBackgroundBlueText: (0, _createStyles.get)('blue300'),
  colorBackgroundDarkGray: (0, _createStyles.get)('darkGrayRgba20'),
  colorBackgroundDarkGrayText: (0, _createStyles.get)('white'),
  colorBackgroundGreen: (0, _createStyles.get)('greenRgba20'),
  colorBackgroundGreenText: (0, _createStyles.get)('green300'),
  colorBackgroundLightGray: (0, _createStyles.get)('lightGrayRgba20'),
  colorBackgroundLightGrayText: (0, _createStyles.get)('white'),
  colorBackgroundOrange: (0, _createStyles.get)('orangeRgba20'),
  colorBackgroundOrangeText: (0, _createStyles.get)('orange300'),
  colorBackgroundPurple: (0, _createStyles.get)('purpleRgba20'),
  colorBackgroundPurpleText: (0, _createStyles.get)('purple300'),
  colorBackgroundRed: (0, _createStyles.get)('redRgba20'),
  colorBackgroundRedText: (0, _createStyles.get)('red300'),
  colorBackgroundYellow: (0, _createStyles.get)('yellowRgba20'),
  colorBackgroundYellowText: (0, _createStyles.get)('yellow300')
};
exports.DARK_MODE_COLORS = DARK_MODE_COLORS;

var DARK_MODE_RGBA_COLORS = _objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread({}, (0, _utils.createRgbaColors)(CORE_BLUE_COLORS,
/* isDark */
true)), (0, _utils.createRgbaColors)(CORE_GREEN_COLORS,
/* isDark */
true)), (0, _utils.createRgbaColors)(CORE_ORANGE_COLORS,
/* isDark */
true)), (0, _utils.createRgbaColors)(CORE_PURPLE_COLORS,
/* isDark */
true)), (0, _utils.createRgbaColors)(CORE_RED_COLORS,
/* isDark */
true)), (0, _utils.createRgbaColors)(CORE_YELLOW_COLORS,
/* isDark */
true)), (0, _utils.createRgbaColors)(CORE_DARK_GRAY_COLORS,
/* isDark */
true)), (0, _utils.createRgbaColors)(CORE_LIGHT_GRAY_COLORS,
/* isDark */
true));

exports.DARK_MODE_RGBA_COLORS = DARK_MODE_RGBA_COLORS;
