// @wp-g2/components/dist/es/TextInput/__stories__/TextInput.stories.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import React from 'react';
import NumberFormat from 'react-number-format';
import { Container } from '../../Container';
import { Text, VStack } from '../../index';
import { TextInput } from '../index';
export default {
  component: TextInput,
  title: 'Components/TextInput'
};
export var _default = function () {
  return /*#__PURE__*/React.createElement(TextInput, {
    placeholder: "Into The Unknown"
  });
};
export var Number = function () {
  var _React$useState = React.useState(0),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      value = _React$useState2[0],
      setValue = _React$useState2[1];

  return /*#__PURE__*/React.createElement(Container, {
    css: "\n\t\t\t\tmargin-top: 20vh;\n\t\t\t",
    width: 480
  }, /*#__PURE__*/React.createElement(TextInput, {
    max: 100,
    min: 0,
    onChange: setValue,
    type: "number",
    value: value
  }));
};
export var NumberWithExternalChange = function () {
  var _React$useState3 = React.useState(),
      _React$useState4 = _slicedToArray(_React$useState3, 2),
      value = _React$useState4[0],
      setValue = _React$useState4[1];

  return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("div", null, "Value: ", value), /*#__PURE__*/React.createElement("button", {
    onClick: function onClick() {
      return setValue(10);
    }
  }, "Make ten"), /*#__PURE__*/React.createElement(TextInput, {
    onChange: setValue,
    type: "number",
    value: value
  }));
};
export var numberStepper = function () {
  return /*#__PURE__*/React.createElement(Container, {
    css: "\n\t\t\t\tmargin-top: 20vh;\n\t\t\t",
    width: 480
  }, /*#__PURE__*/React.createElement(TextInput, {
    arrows: "stepper",
    type: "number",
    value: "1"
  }));
};
export var multiline = function () {
  return /*#__PURE__*/React.createElement(TextInput, {
    maxRows: 6,
    minRows: 3,
    multiline: true
  });
};
export var custom = function () {
  return /*#__PURE__*/React.createElement(VStack, null, /*#__PURE__*/React.createElement(Text, null, "Rendered using ", /*#__PURE__*/React.createElement("code", null, "react-number-format")), /*#__PURE__*/React.createElement(TextInput, {
    as: NumberFormat,
    prefix: '$',
    thousandSeparator: true
  }));
};
export var inlineRendering = function () {
  return /*#__PURE__*/React.createElement(VStack, {
    css: "\n\t\t\t\tmargin: auto;\n\t\t\t\twidth: 320px;\n\t\t\t"
  }, /*#__PURE__*/React.createElement(Text, {
    adjustLineHeightForInnerControls: true
  }, "My site name is ", /*#__PURE__*/React.createElement(TextInput, {
    isInline: true
  })));
};

var MadLib = function (_ref) {
  var type = _ref.type;
  return /*#__PURE__*/React.createElement(TextInput, {
    isInline: true,
    placeholder: type
  });
};

var Adjective = function () {
  return /*#__PURE__*/React.createElement(MadLib, {
    type: "Adjective"
  });
};

var Noun = function () {
  return /*#__PURE__*/React.createElement(MadLib, {
    type: "Noun"
  });
};

var PluralNoun = function () {
  return /*#__PURE__*/React.createElement(MadLib, {
    isInline: true,
    type: "Plural noun"
  });
};

var Gerund = function () {
  return /*#__PURE__*/React.createElement(MadLib, {
    isInline: true,
    type: "Gerund"
  });
};
/**
 * Adapted from https://www.madlibs.com/printables/
 */


export var madLibs = function () {
  return /*#__PURE__*/React.createElement(VStack, {
    css: "\n\t\t\tmargin: auto;\n\t\t\ttext-align: justify;\n\t\t"
  }, /*#__PURE__*/React.createElement(Text, {
    adjustLineHeightForInnerControls: true
  }, "A vacation is when you take a trip to some ", /*#__PURE__*/React.createElement(Adjective, null), " place with your ", /*#__PURE__*/React.createElement(Adjective, null), " family. Usually you go to some place that is near a/an ", /*#__PURE__*/React.createElement(Noun, null), " or up on a/an ", /*#__PURE__*/React.createElement(Noun, null), ". A good vacation place is one where you can ride ", /*#__PURE__*/React.createElement(PluralNoun, null), " or play ", /*#__PURE__*/React.createElement(MadLib, {
    type: "Game"
  }), " or go birding for ", /*#__PURE__*/React.createElement(MadLib, {
    type: "Bird species"
  }), ". I like to spend my time ", /*#__PURE__*/React.createElement(Gerund, null), " or ", /*#__PURE__*/React.createElement(Gerund, null), ". When parents go on a vacation, they spend their time eating three ", /*#__PURE__*/React.createElement(PluralNoun, null), " a day. Last summer, my little sibling fell in a/an ", /*#__PURE__*/React.createElement(Noun, null), " and got poison", ' ', /*#__PURE__*/React.createElement(MadLib, {
    type: "Plant"
  }), " all over their", ' ', /*#__PURE__*/React.createElement(MadLib, {
    type: "Part of the body"
  }), ". My family is going to go to (the) ", /*#__PURE__*/React.createElement(MadLib, {
    type: "Place"
  }), ", and I will practice ", /*#__PURE__*/React.createElement(Gerund, null), ". Parents need vacations more than kids because parents are always very ", /*#__PURE__*/React.createElement(Adjective, null), " and because they have to work", ' ', /*#__PURE__*/React.createElement(MadLib, {
    type: "Number"
  }), " hours every day all year making enough", ' ', /*#__PURE__*/React.createElement(PluralNoun, null), " to pay for the vacation."));
};
