// @wp-g2/components/dist/es/Navigator/__stories__/Navigator.Sidebar.stories.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _taggedTemplateLiteralLoose from "@babel/runtime/helpers/esm/taggedTemplateLiteralLoose";

function _templateObject() {
  var data = _taggedTemplateLiteralLoose(["\n\tbottom: 0;\n\tleft: 0;\n\tright: 0;\n\ttop: 0;\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

import { wordpress } from '@wordpress/icons';
import { faker, Schema } from '@wp-g2/protokit';
import { styled, ThemeProvider, ui } from '@wp-g2/styles';
import { useLocalState } from '@wp-g2/utils';
import { chunk } from 'lodash';
import React, { useEffect, useState } from 'react';
import { MemoryRouter as Router, NavLink, Route, Switch } from 'react-router-dom';
import { Avatar, Background, Badge, Flex, Grid, Heading, HStack, Icon, Menu, MenuItem, Scrollable, SearchInput, Spacer, Spinner, Subheading, Surface, Text, TextInput, Truncate, View, VStack } from '../../index';
import { Navigator, NavigatorLink, NavigatorScreen, NavigatorScreens } from '../index';
export default {
  component: Navigator,
  title: 'Components/Navigator'
};
var ScreenView = styled(Surface)(_templateObject());

var Screen = function (_ref) {
  var children = _ref.children;
  return /*#__PURE__*/React.createElement(ScreenView, null, children);
};

var MenuLink = function (_ref2) {
  var children = _ref2.children,
      props = _objectWithoutProperties(_ref2, ["children"]);

  return /*#__PURE__*/React.createElement(NavigatorLink, props, /*#__PURE__*/React.createElement(MenuItem, null, children));
};

var ExternalMenuLink = function (props) {
  return /*#__PURE__*/React.createElement(MenuItem, Object.assign({
    activeClassName: "is-active",
    as: NavLink,
    exact: true
  }, props));
};

var DashboardNav = function () {
  var _useState = useState(null),
      _useState2 = _slicedToArray(_useState, 2),
      commentCount = _useState2[0],
      setCommentCount = _useState2[1];

  useEffect(function () {
    setTimeout(function () {
      setCommentCount(3);
    }, 1000);
  }, []);
  return /*#__PURE__*/React.createElement(Screen, null, /*#__PURE__*/React.createElement(Spacer, {
    py: 4
  }), /*#__PURE__*/React.createElement(Spacer, {
    mb: 8
  }, /*#__PURE__*/React.createElement(MenuLink, null, "My sites")), /*#__PURE__*/React.createElement(Spacer, {
    mb: 4,
    pl: 3
  }, /*#__PURE__*/React.createElement(Heading, {
    size: 4
  }, "My site name")), /*#__PURE__*/React.createElement(Spacer, {
    mb: 4
  }, /*#__PURE__*/React.createElement(Menu, null, /*#__PURE__*/React.createElement(ExternalMenuLink, {
    to: "/"
  }, "Dashboard"), /*#__PURE__*/React.createElement(MenuLink, {
    to: "Pages"
  }, "Pages"), /*#__PURE__*/React.createElement(MenuLink, {
    to: "Posts"
  }, "Posts"), /*#__PURE__*/React.createElement(ExternalMenuLink, {
    to: "/comments"
  }, /*#__PURE__*/React.createElement(Flex, null, /*#__PURE__*/React.createElement(View, null, /*#__PURE__*/React.createElement(Text, null, "Comments")), /*#__PURE__*/React.createElement(View, {
    css: {
      margin: '-3px 0'
    }
  }, commentCount === null ? /*#__PURE__*/React.createElement(Spinner, null) : /*#__PURE__*/React.createElement(Badge, null, commentCount)))), /*#__PURE__*/React.createElement(ExternalMenuLink, {
    to: "/media"
  }, "Media"), /*#__PURE__*/React.createElement(ExternalMenuLink, {
    to: "/users"
  }, "Users"))), /*#__PURE__*/React.createElement(Spacer, {
    mb: 4,
    mt: 12,
    pl: 3
  }, /*#__PURE__*/React.createElement(Heading, {
    size: 4
  }, "Plugins")), /*#__PURE__*/React.createElement(Spacer, {
    mb: 4
  }, /*#__PURE__*/React.createElement(Menu, null, /*#__PURE__*/React.createElement(ExternalMenuLink, {
    to: "/plugins/hello-dolly"
  }, "Hello Dolly"), /*#__PURE__*/React.createElement(MenuLink, {
    to: "PluginsWooCommerce"
  }, "WooCommerce"), /*#__PURE__*/React.createElement(ExternalMenuLink, {
    to: "/plugins"
  }, "All plugins"))));
};

var pageSchema = new Schema(function () {
  return {
    title: faker.commerce.productName()
  };
});
var pages = pageSchema.make(100);

var PagesNav = function () {
  var _chunk = chunk(pages, 5),
      _chunk2 = _slicedToArray(_chunk, 2),
      recentPages = _chunk2[0],
      draftPages = _chunk2[1];

  var _chunk3 = chunk(draftPages, 2),
      _chunk4 = _slicedToArray(_chunk3, 1),
      drafts = _chunk4[0];

  var _useState3 = useState(''),
      _useState4 = _slicedToArray(_useState3, 2),
      query = _useState4[0],
      setQuery = _useState4[1];

  var isSearching = !!query;
  var searchResults = pages.filter(function (page) {
    return page.title.toLowerCase().includes(query.toLowerCase());
  });

  var _chunk5 = chunk(searchResults, 10),
      _chunk6 = _slicedToArray(_chunk5, 1),
      _results = _chunk6[0];

  var results = _results || [];
  return /*#__PURE__*/React.createElement(Screen, null, /*#__PURE__*/React.createElement(Spacer, {
    py: 4
  }), /*#__PURE__*/React.createElement(Spacer, {
    mb: 8
  }, /*#__PURE__*/React.createElement(MenuLink, {
    isBack: true,
    to: "Dashboard"
  }, "Dashboard")), /*#__PURE__*/React.createElement(Spacer, {
    mb: 4,
    pl: 3
  }, /*#__PURE__*/React.createElement(Heading, {
    size: 4
  }, "Pages")), /*#__PURE__*/React.createElement(View, {
    css: [ui.padding.x(3), ui.margin.bottom(4)]
  }, /*#__PURE__*/React.createElement(SearchInput, {
    onChange: function onChange(next) {
      return setQuery(next);
    },
    placeholder: "Search pages",
    value: query
  })), /*#__PURE__*/React.createElement(VStack, {
    spacing: 10
  }, isSearching && results.length && /*#__PURE__*/React.createElement(View, null, results.map(function (page) {
    return /*#__PURE__*/React.createElement(ExternalMenuLink, {
      key: page.id,
      to: "/pages/" + page.id
    }, /*#__PURE__*/React.createElement(Truncate, null, page.title));
  })), isSearching && !results.length && /*#__PURE__*/React.createElement(View, {
    css: [ui.padding.x(3)]
  }, /*#__PURE__*/React.createElement(Text, null, "No Results")), !isSearching && /*#__PURE__*/React.createElement(VStack, {
    spacing: 10
  }, /*#__PURE__*/React.createElement(VStack, null, /*#__PURE__*/React.createElement(View, {
    css: [ui.padding.x(3)]
  }, /*#__PURE__*/React.createElement(Subheading, null, "Recent Pages")), /*#__PURE__*/React.createElement(View, null, recentPages.map(function (page) {
    return /*#__PURE__*/React.createElement(ExternalMenuLink, {
      key: page.id,
      to: "/pages/" + page.id
    }, /*#__PURE__*/React.createElement(Truncate, null, page.title));
  }))), /*#__PURE__*/React.createElement(VStack, null, /*#__PURE__*/React.createElement(View, {
    css: [ui.padding.x(3)]
  }, /*#__PURE__*/React.createElement(Subheading, null, "Drafts")), /*#__PURE__*/React.createElement(View, null, drafts.map(function (page) {
    return /*#__PURE__*/React.createElement(ExternalMenuLink, {
      key: page.id,
      to: "/pages/" + page.id
    }, /*#__PURE__*/React.createElement(Truncate, null, page.title));
  }))))));
};

var PostsNav = function () {
  return /*#__PURE__*/React.createElement(Screen, null, /*#__PURE__*/React.createElement(Spacer, {
    py: 4
  }), /*#__PURE__*/React.createElement(Spacer, {
    mb: 8
  }, /*#__PURE__*/React.createElement(MenuLink, {
    isBack: true,
    to: "Dashboard"
  }, "Dashboard")), /*#__PURE__*/React.createElement(Spacer, {
    mb: 4,
    pl: 3
  }, /*#__PURE__*/React.createElement(Heading, {
    size: 4
  }, "Posts")), /*#__PURE__*/React.createElement(Menu, null, /*#__PURE__*/React.createElement(ExternalMenuLink, {
    to: "/posts"
  }, "All posts"), /*#__PURE__*/React.createElement(ExternalMenuLink, {
    to: "/posts/categories"
  }, "Categories"), /*#__PURE__*/React.createElement(ExternalMenuLink, {
    to: "/posts/tags"
  }, "Tags"), /*#__PURE__*/React.createElement(ExternalMenuLink, {
    to: "/posts/new"
  }, "New page")));
};

var PluginsWooCommerceNav = function () {
  var _useLocalState = useLocalState('@wp-g2/components/Navigator/Sidebar/PluginsWooCommerceNav', false),
      _useLocalState2 = _slicedToArray(_useLocalState, 2),
      isFavorite = _useLocalState2[0],
      setIsFavorite = _useLocalState2[1];

  return /*#__PURE__*/React.createElement(Screen, null, /*#__PURE__*/React.createElement(Spacer, {
    py: 4
  }), /*#__PURE__*/React.createElement(Spacer, {
    mb: 8
  }, /*#__PURE__*/React.createElement(MenuLink, {
    isBack: true,
    to: "Dashboard"
  }, "Dashboard")), /*#__PURE__*/React.createElement(Spacer, {
    mb: 4,
    pl: 3
  }, /*#__PURE__*/React.createElement(HStack, null, /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(Heading, {
    size: 4
  }, "WooCommerce")), /*#__PURE__*/React.createElement(Text, {
    css: [ui.padding.y(3), ui.padding.x(2), {
      cursor: 'pointer',
      color: isFavorite ? 'yellow' : null
    }, ui.animation.default],
    isBlock: true,
    onClick: function onClick() {
      return setIsFavorite(!isFavorite);
    },
    variant: !isFavorite && 'muted'
  }, "\u2605"))), /*#__PURE__*/React.createElement(Menu, null, /*#__PURE__*/React.createElement(ExternalMenuLink, {
    to: "/plugins/woo-commerce/store"
  }, "Store home"), /*#__PURE__*/React.createElement(MenuLink, {
    showArrow: true
  }, "Analytics"), /*#__PURE__*/React.createElement(MenuLink, {
    showArrow: true
  }, "Orders"), /*#__PURE__*/React.createElement(MenuLink, {
    showArrow: true
  }, "Marketing"), /*#__PURE__*/React.createElement(MenuLink, {
    showArrow: true
  }, "Products"), /*#__PURE__*/React.createElement(MenuLink, {
    showArrow: true
  }, "Customers"), /*#__PURE__*/React.createElement(MenuLink, {
    showArrow: true
  }, "Settings"), /*#__PURE__*/React.createElement(MenuLink, {
    showArrow: true
  }, "Extensions"), /*#__PURE__*/React.createElement(MenuLink, {
    showArrow: true
  }, "Tools")));
};

var screens = [{
  component: DashboardNav,
  path: 'Dashboard'
}, {
  component: PagesNav,
  path: 'Pages'
}, {
  component: PostsNav,
  path: 'Posts'
}, {
  component: PluginsWooCommerceNav,
  path: 'PluginsWooCommerce'
}];

var NavigationSidebar = function () {
  return /*#__PURE__*/React.createElement(ThemeProvider, {
    isDark: true,
    isGlobal: false
  }, /*#__PURE__*/React.createElement(Surface, {
    borderRight: "1px solid",
    css: {
      height: '100%'
    }
  }, /*#__PURE__*/React.createElement(VStack, null, /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(Navigator, {
    initialPath: 'Dashboard'
  }, /*#__PURE__*/React.createElement(Scrollable, {
    css: {
      height: "calc(100vh - 100px - 40px)",
      padding: 8
    }
  }, /*#__PURE__*/React.createElement(HStack, {
    css: [ui.padding(3)],
    spacing: 3
  }, /*#__PURE__*/React.createElement(Icon, {
    icon: wordpress,
    size: 30
  }), /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(Heading, {
    size: 5,
    truncate: true
  }, "My Site Name"))), /*#__PURE__*/React.createElement(NavigatorScreens, null, screens.map(function (screen) {
    return /*#__PURE__*/React.createElement(NavigatorScreen, Object.assign({}, screen, {
      key: screen.path
    }));
  }))))), /*#__PURE__*/React.createElement(Surface, {
    css: [ui.padding.y(4), ui.padding.x(3)]
  }, /*#__PURE__*/React.createElement(HStack, {
    spacing: 4
  }, /*#__PURE__*/React.createElement(Avatar, {
    name: "Elsa Oldenburg",
    size: "medium",
    src: "https://picsum.photos/id/1041/300/300"
  }), /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(Text, {
    weight: "bold"
  }, "Elsa")))))));
};

var MockPage = function (_ref3) {
  var match = _ref3.match;
  var params = match.params,
      path = match.path;
  var title;

  if (path === '/pages/:id' && (params == null ? void 0 : params.id)) {
    var page = pages.find(function (page) {
      return page.id === (params == null ? void 0 : params.id);
    });
    title = page.title;
  } else {
    var route = routes.find(function (item) {
      return item.path === path;
    });
    title = route.title;
  }

  return /*#__PURE__*/React.createElement(Surface, {
    borderBottom: true
  }, /*#__PURE__*/React.createElement(View, {
    css: {
      padding: '20px 40px'
    }
  }, /*#__PURE__*/React.createElement(Heading, null, title)));
};

var routes = [{
  path: '/',
  title: 'Dashboard'
}, {
  path: '/pages',
  title: 'All pages'
}, {
  path: '/pages/new',
  title: 'New page'
}, {
  path: '/pages/:id',
  title: 'Single Page'
}, {
  path: '/posts',
  title: 'All posts'
}, {
  path: '/posts/categories',
  title: 'Categories'
}, {
  path: '/posts/tags',
  title: 'Tags'
}, {
  path: '/posts/new',
  title: 'New post'
}, {
  path: '/media',
  title: 'Media'
}, {
  path: '/comments',
  title: 'Comments'
}, {
  path: '/users',
  title: 'Users'
}, {
  path: '/plugins',
  title: 'All plugins'
}, {
  path: '/plugins/hello-dolly',
  title: 'Hello Dolly'
}, {
  path: '/plugins/woo-commerce/store',
  title: 'Store home'
}];

var BrowserBar = function (_ref4) {
  var _history$location;

  var history = _ref4.history;
  var pathname = (history == null ? void 0 : (_history$location = history.location) == null ? void 0 : _history$location.pathname) || '/';
  return /*#__PURE__*/React.createElement(Background, {
    borderBottom: true,
    css: {
      padding: 4
    }
  }, /*#__PURE__*/React.createElement(TextInput, {
    css: {
      textAlign: 'center'
    },
    readOnly: true,
    value: "http://mysite.com/wp-admin" + pathname
  }));
};

var App = function () {
  return /*#__PURE__*/React.createElement(View, {
    css: {
      border: "2px solid " + ui.get('surfaceBorderColor'),
      bottom: 100,
      left: 0,
      position: 'fixed',
      right: 0,
      top: 0
    }
  }, /*#__PURE__*/React.createElement(Router, {
    initialEntries: ['/']
  }, /*#__PURE__*/React.createElement(VStack, {
    spacing: 0
  }, /*#__PURE__*/React.createElement(Route, {
    component: BrowserBar
  }), /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(Grid, {
    css: {
      height: "100%"
    },
    gap: 0,
    templateColumns: "220px 1fr"
  }, /*#__PURE__*/React.createElement(NavigationSidebar, null), /*#__PURE__*/React.createElement(Background, null, /*#__PURE__*/React.createElement(Switch, null, routes.map(function (route) {
    return /*#__PURE__*/React.createElement(Route, Object.assign({
      exact: true
    }, route, {
      component: MockPage,
      key: route.path
    }));
  }))))))));
};

export var Sidebar = function () {
  return /*#__PURE__*/React.createElement(App, null);
};
