// @wordpress/components/build/ui/spinner/component.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _styles2 = require("./styles");

var _utils = require("./utils");

/**
 * External dependencies
 */

/**
 * Internal dependencies
 */

/* eslint-disable jsdoc/valid-types */

/**
 * @typedef Props
 * @property {import('react').CSSProperties['color']} [color] Color of `Spinner`.
 * @property {number} [size=16] Size of `Spinner`.
 */

/* eslint-enable jsdoc/valid-types */

/**
 *
 * @param {import('@wp-g2/create-styles').ViewOwnProps<Props, 'div'>} props
 * @param {import('react').Ref<any>} forwardedRef
 */
function Spinner(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'Spinner'),
      _useContextSystem$col = _useContextSystem.color,
      color = _useContextSystem$col === void 0 ? (0, _styles.get)('colorText') : _useContextSystem$col,
      _useContextSystem$siz = _useContextSystem.size,
      size = _useContextSystem$siz === void 0 ? _utils.BASE_SIZE : _useContextSystem$siz,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["color", "size"]);

  var ratio = size / _utils.BASE_SIZE;
  var scale = ratio * _utils.BASE_SIZE / _utils.WRAPPER_SIZE;
  var transform = "scale(".concat(scale, ")");
  var styles = {
    transform: transform
  };
  return (0, _element.createElement)(_styles2.ContainerView, (0, _extends2.default)({}, otherProps, {
    "aria-busy": true,
    ref: forwardedRef,
    style: {
      height: size,
      width: size
    }
  }), (0, _element.createElement)(_styles2.BarsWrapperView, {
    "aria-hidden": true,
    style: styles
  }, (0, _element.createElement)(_styles2.BarsView, {
    style: {
      color: color
    }
  }, (0, _element.createElement)("div", {
    className: "InnerBar1"
  }), (0, _element.createElement)("div", {
    className: "InnerBar2"
  }), (0, _element.createElement)("div", {
    className: "InnerBar3"
  }), (0, _element.createElement)("div", {
    className: "InnerBar4"
  }), (0, _element.createElement)("div", {
    className: "InnerBar5"
  }), (0, _element.createElement)("div", {
    className: "InnerBar6"
  }), (0, _element.createElement)("div", {
    className: "InnerBar7"
  }), (0, _element.createElement)("div", {
    className: "InnerBar8"
  }), (0, _element.createElement)("div", {
    className: "InnerBar9"
  }), (0, _element.createElement)("div", {
    className: "InnerBar10"
  }), (0, _element.createElement)("div", {
    className: "InnerBar11"
  }), (0, _element.createElement)("div", {
    className: "InnerBar12"
  }))));
}
/**
 * `Spinner` is a component that notify users that their action is being processed.
 *
 * @example
 * ```jsx
 * import { Spinner } from `@wp-g2/components/ui`;
 *
 * function Example() {
 * 	return (
 * 		<Spinner />
 * 	);
 * }
 * ```
 */


var _default = (0, _context.contextConnect)(Spinner, 'Spinner');

exports.default = _default;
