// @wordpress/components/build-module/mobile/gradient/index.native.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { View, Platform } from 'react-native';
import RNLinearGradient from 'react-native-linear-gradient';
import gradientParser from 'gradient-parser';
/**
 * WordPress dependencies
 */

import { colorsUtils } from '@wordpress/components';
import { RadialGradient, Stop, SVG, Defs, Rect } from '@wordpress/primitives';
import { useResizeObserver } from '@wordpress/compose';
import { useMemo } from '@wordpress/element';
/**
 * Internal dependencies
 */

import styles from './style.scss';
export function getGradientAngle(gradientValue) {
  var angleBase = 45;
  var matchAngle = /\(((\d+deg)|(to\s[^,]+))/;
  var angle = matchAngle.exec(gradientValue) ? matchAngle.exec(gradientValue)[1] : '180deg';
  var angleType = angle.includes('deg') ? 'angle' : 'sideOrCorner';

  if (angleType === 'sideOrCorner') {
    switch (angle) {
      case 'to top':
        return 0;

      case 'to top right':
      case 'to right top':
        return angleBase;

      case 'to right':
        return 2 * angleBase;

      case 'to bottom right':
      case 'to right bottom':
        return 3 * angleBase;

      case 'to bottom':
        return 4 * angleBase;

      case 'to bottom left':
      case 'to left bottom':
        return 5 * angleBase;

      case 'to left':
        return 6 * angleBase;

      case 'to top left':
      case 'to left top':
        return 7 * angleBase;
    }
  } else if (angleType === 'angle') {
    return parseFloat(angle);
  } else return 4 * angleBase;
}
export function getGradientColorGroup(gradientValue) {
  var _ref, _gradientParser$parse;

  var colorNeedParenthesis = ['rgb', 'rgba'];
  var excludeSideOrCorner = /linear-gradient\(to\s+([a-z\s]+,)/; // Parser has some difficulties with angle defined as a side or corner (e.g. `to left`)
  // so it's going to be excluded in order to matching color groups

  var modifiedGradientValue = gradientValue.replace(excludeSideOrCorner, 'linear-gradient(');
  return (_ref = []).concat.apply(_ref, _toConsumableArray((_gradientParser$parse = gradientParser.parse(modifiedGradientValue)) === null || _gradientParser$parse === void 0 ? void 0 : _gradientParser$parse.map(function (gradient) {
    var _gradient$colorStops;

    return (_gradient$colorStops = gradient.colorStops) === null || _gradient$colorStops === void 0 ? void 0 : _gradient$colorStops.map(function (color, index) {
      var type = color.type,
          value = color.value,
          length = color.length;
      var fallbackLength = "".concat(100 * (index / (gradient.colorStops.length - 1)), "%");
      var colorLength = length ? "".concat(length.value).concat(length.type) : fallbackLength;

      if (colorNeedParenthesis.includes(type)) {
        return ["".concat(type, "(").concat(value.join(','), ")"), colorLength];
      } else if (type === 'literal') {
        return [value, colorLength];
      }

      return ["#".concat(value), colorLength];
    });
  })));
}
export function getGradientBaseColors(colorGroup) {
  return colorGroup.map(function (color) {
    return color[0];
  });
}
export function getColorLocations(colorGroup) {
  return colorGroup.map(function (location) {
    return Number(location[1].replace('%', '')) / 100;
  });
}

function Gradient(_ref2) {
  var gradientValue = _ref2.gradientValue,
      style = _ref2.style,
      _ref2$angleCenter = _ref2.angleCenter,
      angleCenter = _ref2$angleCenter === void 0 ? {
    x: 0.5,
    y: 0.5
  } : _ref2$angleCenter,
      children = _ref2.children,
      otherProps = _objectWithoutProperties(_ref2, ["gradientValue", "style", "angleCenter", "children"]);

  var _useResizeObserver = useResizeObserver(),
      _useResizeObserver2 = _slicedToArray(_useResizeObserver, 2),
      resizeObserver = _useResizeObserver2[0],
      sizes = _useResizeObserver2[1];

  var _ref3 = sizes || {},
      _ref3$width = _ref3.width,
      width = _ref3$width === void 0 ? 0 : _ref3$width,
      _ref3$height = _ref3.height,
      height = _ref3$height === void 0 ? 0 : _ref3$height;

  var isGradient = colorsUtils.isGradient,
      getGradientType = colorsUtils.getGradientType,
      gradients = colorsUtils.gradients;
  var colorGroup = useMemo(function () {
    return getGradientColorGroup(gradientValue);
  }, [gradientValue]);
  var locations = useMemo(function () {
    return getColorLocations(colorGroup);
  }, [colorGroup]);
  var colors = useMemo(function () {
    return getGradientBaseColors(colorGroup);
  }, [colorGroup]);

  if (!gradientValue || !isGradient(gradientValue)) {
    return null;
  }

  var isLinearGradient = getGradientType(gradientValue) === gradients.linear;

  if (isLinearGradient) {
    return createElement(RNLinearGradient, _extends({
      colors: colors,
      useAngle: true,
      angle: getGradientAngle(gradientValue),
      locations: locations,
      angleCenter: angleCenter,
      style: style
    }, otherProps), children);
  }

  return createElement(View, {
    style: [style, styles.overflow]
  }, createElement(View, {
    style: styles.radialGradientContent
  }, children), resizeObserver, createElement(SVG, null, createElement(Defs, null, createElement(RadialGradient //eslint-disable-next-line no-restricted-syntax
  , {
    id: "radialGradient",
    gradientUnits: "userSpaceOnUse",
    rx: "70%",
    ry: "70%",
    cy: Platform.OS === 'android' ? width / 2 : '50%'
  }, colorGroup.map(function (group) {
    return createElement(Stop, {
      offset: group[1],
      stopColor: group[0],
      stopOpacity: "1",
      key: "".concat(group[1], "-").concat(group[0])
    });
  }))), createElement(Rect, {
    height: height,
    width: width,
    fill: "url(#radialGradient)"
  })));
}

export default Gradient;
