// @wp-g2/components/dist/cjs/ZStack/ZStack.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _utils = require("@wp-g2/utils");

var _react = _interopRequireDefault(require("react"));

var _View = require("../View");

var styles = _interopRequireWildcard(require("./ZStack.styles"));

var ZStackView = styles.ZStackView;

function ZStack(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'ZStack'),
      children = _useContextSystem.children,
      className = _useContextSystem.className,
      _useContextSystem$isL = _useContextSystem.isLayered,
      isLayered = _useContextSystem$isL === void 0 ? true : _useContextSystem$isL,
      _useContextSystem$isR = _useContextSystem.isReversed,
      isReversed = _useContextSystem$isR === void 0 ? false : _useContextSystem$isR,
      _useContextSystem$off = _useContextSystem.offset,
      offset = _useContextSystem$off === void 0 ? 0 : _useContextSystem$off,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["children", "className", "isLayered", "isReversed", "offset"]);

  var validChildren = (0, _utils.getValidChildren)(children);
  var childrenCount = validChildren.length - 1;
  var clonedChildren = validChildren.map(function (child, index) {
    var zIndex = isReversed ? childrenCount - index : index;
    var classes = (0, _styles.cx)(isLayered ? styles.positionAbsolute : styles.positionRelative, (0, _styles.css)({
      marginLeft: !isLayered && offset * -1
    }));
    return /*#__PURE__*/_react.default.createElement(_View.View, Object.assign({}, _styles.ui.$('ZStackItem'), {
      className: classes,
      key: child.key,
      style: {
        zIndex: zIndex
      }
    }), child);
  });
  var classes = (0, _styles.cx)((0, _styles.css)({
    paddingLeft: !isLayered ? offset : null
  }), className);
  return /*#__PURE__*/_react.default.createElement(ZStackView, Object.assign({}, otherProps, {
    className: classes,
    ref: forwardedRef
  }), clonedChildren);
}

var _default = (0, _context.contextConnect)(ZStack, 'ZStack');

exports.default = _default;
