// @wp-g2/components/dist/cjs/Animated/useAnimated.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useAnimated = useAnimated;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _framerMotion = require("framer-motion");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function useAnimated(props) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'Animated'),
      _useContextSystem$as = _useContextSystem.as,
      as = _useContextSystem$as === void 0 ? 'div' : _useContextSystem$as,
      _useContextSystem$aut = _useContextSystem.auto,
      auto = _useContextSystem$aut === void 0 ? false : _useContextSystem$aut,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["as", "auto"]);

  var _useReducedMotion = (0, _styles.useReducedMotion)(),
      _useReducedMotion2 = (0, _slicedToArray2.default)(_useReducedMotion, 1),
      isProviderReducedMotion = _useReducedMotion2[0];

  var isSystemReducedMotion = (0, _framerMotion.useReducedMotion)();
  var baseProps = {
    initial: false
  };

  if (auto) {
    baseProps = {
      animate: {
        opacity: 1,
        scale: 1,
        transition: props.transition || {
          duration: 0.2,
          ease: 'easeInOut'
        }
      },
      exit: {
        opacity: 0,
        scale: 0.9,
        transition: {
          duration: 0.2,
          ease: 'easeInOut'
        }
      },
      initial: {
        opacity: 0,
        scale: 0.9
      }
    };
  }

  var finalProps = _objectSpread(_objectSpread({}, baseProps), otherProps);

  if (isSystemReducedMotion || isProviderReducedMotion) {
    finalProps = _objectSpread(_objectSpread({}, finalProps), {}, {
      layout: undefined,
      transition: {
        duration: 0
      }
    });
  }

  return _objectSpread(_objectSpread({}, finalProps), {}, {
    as: as
  });
}
