// @wp-g2/styles/dist/cjs/theme/dark-mode-config.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.darkModeConfig = exports.DARK_THEME = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _createStyles = require("@wp-g2/create-styles");

var _tokens = require("./tokens");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var DARK_MODE_PROPS = _objectSpread(_objectSpread(_objectSpread({}, _tokens.DARK_MODE_COLORS), _tokens.DARK_MODE_RGBA_COLORS), {}, {
  buttonPrimaryTextColorActive: (0, _createStyles.get)('controlPrimaryTextColorActive'),
  buttonControlActiveStateTextColor: (0, _createStyles.get)('colorTextInverted'),
  colorBodyBackground: '#18191A',
  colorDivider: 'rgba(255, 255, 255, 0.1)',
  colorScrollbarThumb: 'rgba(255, 255, 255, 0.2)',
  colorScrollbarThumbHover: 'rgba(255, 255, 255, 0.5)',
  colorScrollbarTrack: 'rgba(0, 0, 0, 0.04)',
  colorText: '#E4E6EB',
  colorTextMuted: '#7a7a7a',
  colorTextInverted: '#050505',
  colorTextHeading: '#ffffff',
  controlBackgroundColor: (0, _createStyles.get)('colorBodyBackground'),
  controlBackgroundColorHover: 'rgba(255, 255, 255, 0.3)',
  controlBackgroundBrightColor: 'rgba(255, 255, 255, 0.08)',
  controlBackgroundDimColor: 'rgba(255, 255, 255, 0.2)',
  controlBorderSubtleColor: 'rgba(255, 255, 255, 0.5)',
  controlPrimaryTextColorActive: (0, _createStyles.get)('black'),
  controlPrimaryTextColor: (0, _createStyles.get)('white'),
  controlSurfaceColor: 'rgba(255, 255, 255, 0.3)',
  controlTextActiveColor: (0, _createStyles.get)('white'),
  surfaceBackgroundColor: (0, _createStyles.get)('colorBodyBackground'),
  surfaceBackgroundSubtleColor: '#151515',
  surfaceBackgroundTintColor: '#252525',
  surfaceBackgroundTertiaryColor: '#000',
  surfaceBorderColor: 'rgba(255, 255, 255, 0.1)',
  surfaceBorderBoldColor: 'rgba(255, 255, 255, 0.15)',
  surfaceBorderSubtleColor: 'rgba(255, 255, 255, 0.05)',
  surfaceColor: '#292929'
});

var DARK_THEME = _objectSpread({}, DARK_MODE_PROPS);

exports.DARK_THEME = DARK_THEME;
var darkModeConfig = DARK_THEME;
exports.darkModeConfig = darkModeConfig;
