// @wp-g2/components/dist/es/Select/__stories__/Select.stories.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import React from 'react';
import { Select } from '../index';
export default {
  component: Select,
  title: 'Components/Select'
};
export var _default = function () {
  return /*#__PURE__*/React.createElement(Select, {
    options: [{
      label: 'Today',
      value: 'today'
    }, {
      label: 'Yesterday',
      value: 'yesterday'
    }, {
      label: 'Last 7 days',
      value: 'lastWeek'
    }],
    required: true
  });
};
export var _groups = function () {
  var _React$useState = React.useState(''),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      value = _React$useState2[0],
      setValue = _React$useState2[1];

  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Select, {
    onChange: setValue,
    options: [{
      label: 'Frozen',
      options: [{
        label: 'Frozen Heart',
        value: 'frozen-heart'
      }, {
        label: 'Do You Want To Build A Snowman?',
        value: 'do-you-want-to-build-a-snowman'
      }, {
        label: 'First Time In Forever',
        value: 'first-time-in-forever'
      }]
    }, {
      label: 'Frozen 2',
      options: [{
        label: 'All Is Found',
        value: 'all-is-found'
      }, {
        label: 'Some Things Never Change',
        value: 'some-things-never-change'
      }, {
        label: 'Into The Unknown',
        value: 'into-the-unknown'
      }]
    }],
    placeholder: "Select a song",
    value: value
  }));
};
export var _multiple = function () {
  var _React$useState3 = React.useState([]),
      _React$useState4 = _slicedToArray(_React$useState3, 2),
      value = _React$useState4[0],
      setValue = _React$useState4[1];

  return /*#__PURE__*/React.createElement(Select, {
    multiple: true,
    onChange: setValue,
    options: [{
      label: 'Today',
      value: 'today'
    }, {
      label: 'Yesterday',
      value: 'yesterday'
    }, {
      label: 'Last 7 days',
      value: 'lastWeek'
    }],
    value: value
  });
};
