// @wordpress/components/build/ui/control-group/styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.itemGrid = exports.itemFocus = exports.lastRow = exports.firstRow = exports.last = exports.middle = exports.first = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _styles = require("@wp-g2/styles");

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7;

var first = (0, _styles.css)(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2.default)(["\n\tborder-bottom-right-radius: 0;\n\tborder-top-right-radius: 0;\n"])));
exports.first = first;
var middle = (0, _styles.css)(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2.default)(["\n\tborder-radius: 0;\n"])));
exports.middle = middle;
var last = (0, _styles.css)(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2.default)(["\n\tborder-bottom-left-radius: 0;\n\tborder-top-left-radius: 0;\n"])));
exports.last = last;
var firstRow = (0, _styles.css)(_templateObject4 || (_templateObject4 = (0, _taggedTemplateLiteral2.default)(["\n\tborder-bottom-left-radius: 0;\n\tborder-bottom-right-radius: 0;\n"])));
exports.firstRow = firstRow;
var lastRow = (0, _styles.css)(_templateObject5 || (_templateObject5 = (0, _taggedTemplateLiteral2.default)(["\n\tborder-top-left-radius: 0;\n\tborder-top-right-radius: 0;\n"])));
exports.lastRow = lastRow;
var itemFocus = (0, _styles.css)(_templateObject6 || (_templateObject6 = (0, _taggedTemplateLiteral2.default)(["\n\t> * {\n\t\t&:focus-within {\n\t\t\tz-index: 1;\n\t\t}\n\t}\n"])));
exports.itemFocus = itemFocus;
var itemGrid = (0, _styles.css)(_templateObject7 || (_templateObject7 = (0, _taggedTemplateLiteral2.default)(["\n\t> * + *:not( marquee ) {\n\t\tmargin-left: -1px;\n\t\twidth: calc( 100% + 1px );\n\t}\n"])));
exports.itemGrid = itemGrid;
