// @wp-g2/components/dist/cjs/ColorControl/ColorControl.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _Button = require("../Button");

var _ColorCircle = require("../ColorCircle");

var styles = _interopRequireWildcard(require("./ColorControl.styles"));

function ColorControl(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'ColorControl'),
      children = _useContextSystem.children,
      className = _useContextSystem.className,
      color = _useContextSystem.color,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["children", "className", "color"]);

  var classes = (0, _styles.cx)(styles.ColorControl, className);
  return /*#__PURE__*/_react.default.createElement(_Button.Button, Object.assign({
    className: classes,
    isBlock: true,
    isControl: true,
    isSubtle: true,
    ref: forwardedRef,
    textAlign: "left",
    variant: "tertiary"
  }, otherProps, {
    prefix: /*#__PURE__*/_react.default.createElement(_ColorCircle.ColorCircle, {
      color: color,
      size: "small"
    })
  }), children);
}

var _default = (0, _context.contextConnect)(ColorControl, 'ColorControl');

exports.default = _default;
