// @wp-g2/components/dist/cjs/AlignmentMatrixControl/AlignmentMatrixControl.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _i18n = require("@wordpress/i18n");

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _utils = require("@wp-g2/utils");

var _lodash = require("lodash");

var _react = _interopRequireWildcard(require("react"));

var _reakit = require("reakit");

var _AlignmentMatrixControl = require("./AlignmentMatrixControl.styles");

var _AlignmentMatrixControl2 = require("./AlignmentMatrixControl.utils");

var _AlignmentMatrixControlCell = _interopRequireDefault(require("./AlignmentMatrixControlCell"));

var _constants = require("./constants");

function AlignmentMatrixControl(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'AlignmentMatrixControl'),
      _useContextSystem$bas = _useContextSystem.baseId,
      baseId = _useContextSystem$bas === void 0 ? 'alignment-matrix-control' : _useContextSystem$bas,
      defaultValue = _useContextSystem.defaultValue,
      _useContextSystem$onC = _useContextSystem.onChange,
      onChange = _useContextSystem$onC === void 0 ? _lodash.noop : _useContextSystem$onC,
      _useContextSystem$lab = _useContextSystem.label,
      label = _useContextSystem$lab === void 0 ? (0, _i18n.__)('Alignment Matrix Control') : _useContextSystem$lab,
      value = _useContextSystem.value,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["baseId", "defaultValue", "onChange", "label", "value"]);

  var _useState = (0, _react.useState)(value != null ? value : defaultValue),
      _useState2 = (0, _slicedToArray2.default)(_useState, 1),
      immutableDefaultValue = _useState2[0];

  var isRTL = (0, _styles.useRTL)();
  var initialCurrentId = (0, _AlignmentMatrixControl2.getItemId)(baseId, immutableDefaultValue);
  var composite = (0, _reakit.useCompositeState)({
    baseId: baseId,
    currentId: initialCurrentId,
    rtl: isRTL
  });
  var handleOnChange = (0, _react.useCallback)(function (nextValue) {
    return function () {
      onChange(nextValue);
      composite.setCurrentId((0, _AlignmentMatrixControl2.getItemId)(baseId, nextValue));
    };
  }, [baseId, composite, onChange]);
  (0, _utils.useUpdateEffect)(function () {
    composite.setCurrentId((0, _AlignmentMatrixControl2.getItemId)(baseId, value));
  }, [value]);
  return /*#__PURE__*/_react.default.createElement(_reakit.Composite, Object.assign({
    "aria-label": label,
    as: _AlignmentMatrixControl.AlignmentMatrixControlView,
    ref: forwardedRef
  }, otherProps, composite, {
    role: "grid"
  }), _constants.GRID.map(function (cells, index) {
    return /*#__PURE__*/_react.default.createElement(_reakit.CompositeGroup, Object.assign({
      as: _AlignmentMatrixControl.RowView,
      key: index,
      role: "row"
    }, composite), cells.map(function (cell) {
      var cellId = (0, _AlignmentMatrixControl2.getItemId)(baseId, cell);
      var isActive = composite.currentId === cellId;
      return /*#__PURE__*/_react.default.createElement(_AlignmentMatrixControlCell.default, Object.assign({}, composite, {
        id: cellId,
        isActive: isActive,
        key: cell,
        onFocus: handleOnChange(cell),
        value: cell
      }));
    }));
  }));
}

var _default = (0, _context.contextConnect)(AlignmentMatrixControl, 'AlignmentMatrixControl');

exports.default = _default;
