// @wp-g2/components/dist/cjs/AlignmentMatrixControl/AlignmentMatrixControlCell.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _reakit = require("reakit");

var _Tooltip = require("../Tooltip");

var _VisuallyHidden = require("../VisuallyHidden");

var styles = _interopRequireWildcard(require("./AlignmentMatrixControl.styles"));

var _constants = require("./constants");

var CellView = styles.CellView,
    PointView = styles.PointView;

function AlignmentMatrixControlCell(_ref) {
  var _cx;

  var id = _ref.id,
      _ref$isActive = _ref.isActive,
      isActive = _ref$isActive === void 0 ? false : _ref$isActive,
      value = _ref.value,
      props = (0, _objectWithoutProperties2.default)(_ref, ["id", "isActive", "value"]);
  var tooltipText = _constants.ALIGNMENT_LABEL[value];
  var pointClasses = (0, _styles.cx)((_cx = {}, _cx[styles.activePoint] = isActive, _cx));
  return /*#__PURE__*/_react.default.createElement(_reakit.CompositeItem, Object.assign({
    as: CellView,
    role: "gridcell"
  }, props, {
    id: id
  }), /*#__PURE__*/_react.default.createElement(_VisuallyHidden.VisuallyHidden, null, value), /*#__PURE__*/_react.default.createElement(_Tooltip.Tooltip, {
    baseId: id,
    content: tooltipText,
    focusable: false
  }, /*#__PURE__*/_react.default.createElement(PointView, {
    className: pointClasses,
    role: "presentation"
  })));
}

var _default = /*#__PURE__*/_react.default.memo(AlignmentMatrixControlCell);

exports.default = _default;
