// @wordpress/block-editor/build-module/components/inserter/hooks/use-patterns-state.js
/**
 * External dependencies
 */
import { map } from 'lodash';
/**
 * WordPress dependencies
 */

import { useCallback } from '@wordpress/element';
import { cloneBlock } from '@wordpress/blocks';
import { useDispatch, useSelect } from '@wordpress/data';
import { __, sprintf } from '@wordpress/i18n';
import { store as noticesStore } from '@wordpress/notices';
/**
 * Internal dependencies
 */

import { store as blockEditorStore } from '../../../store';
/**
 * Retrieves the block patterns inserter state.
 *
 * @param {Function} onInsert     function called when inserter a list of blocks.
 * @param {string=}  rootClientId Insertion's root client ID.
 *
 * @return {Array} Returns the patterns state. (patterns, categories, onSelect handler)
 */

var usePatternsState = function usePatternsState(onInsert, rootClientId) {
  var _useSelect = useSelect(function (select) {
    var _select = select(blockEditorStore),
        __experimentalGetAllowedPatterns = _select.__experimentalGetAllowedPatterns,
        getSettings = _select.getSettings;

    var inserterPatterns = __experimentalGetAllowedPatterns(rootClientId).filter(function (pattern) {
      return !pattern.scope || pattern.scope.inserter;
    });

    return {
      patterns: inserterPatterns,
      patternCategories: getSettings().__experimentalBlockPatternCategories
    };
  }, [rootClientId]),
      patternCategories = _useSelect.patternCategories,
      patterns = _useSelect.patterns;

  var _useDispatch = useDispatch(noticesStore),
      createSuccessNotice = _useDispatch.createSuccessNotice;

  var onClickPattern = useCallback(function (pattern, blocks) {
    onInsert(map(blocks, function (block) {
      return cloneBlock(block);
    }), pattern.name);
    createSuccessNotice(sprintf(
    /* translators: %s: block pattern title. */
    __('Block pattern "%s" inserted.'), pattern.title), {
      type: 'snackbar'
    });
  }, []);
  return [patterns, patternCategories, onClickPattern];
};

export default usePatternsState;
