// @wp-g2/components/dist/cjs/Scrollable/Scrollable.styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.scrollAuto = exports.scrollY = exports.scrollX = exports.smoothScroll = exports.Content = exports.Scrollable = exports.scrollableScrollbar = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _styles = require("@wp-g2/styles");

function _templateObject7() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\toverflow-y: auto;\n"]);

  _templateObject7 = function () {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\toverflow-x: hidden;\n\toverflow-y: auto;\n"]);

  _templateObject6 = function () {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\toverflow-x: auto;\n\toverflow-y: hidden;\n"]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tscroll-behavior: smooth;\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tposition: relative;\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\theight: 100%;\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t@media only screen and (min-device-width: 40em) {\n\t\t&::-webkit-scrollbar {\n\t\t\theight: 12px;\n\t\t\twidth: 12px;\n\t\t}\n\n\t\t&::-webkit-scrollbar-track {\n\t\t\tbackground-color: transparent;\n\t\t}\n\n\t\t&::-webkit-scrollbar-track {\n\t\t\tbackground: ", ";\n\t\t\tborder-radius: 8px;\n\t\t}\n\n\t\t&::-webkit-scrollbar-thumb {\n\t\t\tbackground-clip: padding-box;\n\t\t\tbackground-color: ", ";\n\t\t\tborder: 2px solid rgba(0, 0, 0, 0);\n\t\t\tborder-radius: 7px;\n\t\t}\n\n\t\t&:hover::-webkit-scrollbar-thumb {\n\t\t\tbackground-color: ", ";\n\t\t}\n\t}\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var scrollableScrollbar = (0, _styles.css)(_templateObject(), _styles.ui.get('colorScrollbarTrack'), _styles.ui.get('colorScrollbarThumb'), _styles.ui.get('colorScrollbarThumbHover'));
exports.scrollableScrollbar = scrollableScrollbar;
var Scrollable = (0, _styles.css)(_templateObject2());
exports.Scrollable = Scrollable;
var Content = (0, _styles.css)(_templateObject3());
exports.Content = Content;
var smoothScroll = (0, _styles.css)(_templateObject4());
exports.smoothScroll = smoothScroll;
var scrollX = (0, _styles.css)(_templateObject5());
exports.scrollX = scrollX;
var scrollY = (0, _styles.css)(_templateObject6());
exports.scrollY = scrollY;
var scrollAuto = (0, _styles.css)(_templateObject7());
exports.scrollAuto = scrollAuto;
