// @wp-g2/components/dist/cjs/ColorPicker/__stories__/ColorPicker.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HTMLPickerTest = exports.BasicTest = exports._default = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _tinycolor = _interopRequireDefault(require("tinycolor2"));

var _index = require("../../index");

var _index2 = require("../index");

var _default2 = {
  component: _index2.ColorPicker,
  title: 'Components/ColorPicker'
};
exports.default = _default2;

var Example = function () {
  var _React$useState = _react.default.useState('red'),
      _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
      color = _React$useState2[0],
      setColor = _React$useState2[1];

  return /*#__PURE__*/_react.default.createElement(_index.Container, null, /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement("strong", null, color)), /*#__PURE__*/_react.default.createElement(_index.VStack, {
    spacing: 10
  }, /*#__PURE__*/_react.default.createElement(_index.Grid, {
    gap: 8
  }, /*#__PURE__*/_react.default.createElement(_index2.ColorPicker, {
    color: color,
    disableAlpha: false,
    inputType: "rgb",
    onChange: function handleOnChange(next) {
      setColor(next);
    },
    width: 300
  }), /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("button", {
    onClick: function onClick() {
      return setColor(_tinycolor.default.random().toRgbString());
    }
  }, "Set Random Color")))));
};

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(Example, null);
};

exports._default = _default;

var BasicTest = function () {
  var _React$useState3 = _react.default.useState('red'),
      _React$useState4 = (0, _slicedToArray2.default)(_React$useState3, 2),
      color = _React$useState4[0],
      setColor = _React$useState4[1];

  return /*#__PURE__*/_react.default.createElement(_index.Container, null, /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement("strong", null, color)), /*#__PURE__*/_react.default.createElement(_index.VStack, {
    spacing: 10
  }, /*#__PURE__*/_react.default.createElement(_index.Grid, {
    gap: 8
  }, /*#__PURE__*/_react.default.createElement(_index2.ColorPicker, {
    color: color,
    onChange: function handleOnChange(next) {
      setColor(next);
    },
    width: 300
  }), /*#__PURE__*/_react.default.createElement("input", {
    onChange: function onChange(e) {
      return setColor(e.target.value);
    },
    type: "color",
    value: _styles.ui.color(color).toHexString()
  }))));
};

exports.BasicTest = BasicTest;

var HTMLPickerTest = function () {
  var _React$useState5 = _react.default.useState('#ff0000'),
      _React$useState6 = (0, _slicedToArray2.default)(_React$useState5, 2),
      color = _React$useState6[0],
      setColor = _React$useState6[1];

  return /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.Heading, {
    size: 1
  }, /*#__PURE__*/_react.default.createElement("span", {
    "aria-label": "wave",
    role: "img"
  }, "\uD83D\uDC4B"), ' ', color), /*#__PURE__*/_react.default.createElement("input", {
    onChange: function onChange(e) {
      return setColor(e.target.value);
    },
    type: "color"
  }))))));
};

exports.HTMLPickerTest = HTMLPickerTest;
