// @wp-g2/components/dist/cjs/Menu/Menu.styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.showBackArrow = exports.showArrow = exports.offset = exports.MenuHeader = exports.xxSmall = exports.xSmall = exports.small = exports.large = exports.MenuItem = exports.Menu = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _styles = require("@wp-g2/styles");

function _templateObject11() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n"]);

  _templateObject11 = function () {
    return data;
  };

  return data;
}

function _templateObject10() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n"]);

  _templateObject10 = function () {
    return data;
  };

  return data;
}

function _templateObject9() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\t", ";\n\t", ";\n\tmin-height: calc(", " + ", ");\n\twidth: calc(100% + ", ");\n"]);

  _templateObject9 = function () {
    return data;
  };

  return data;
}

function _templateObject8() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\talign-items: center;\n\tdisplay: flex;\n\tmin-height: ", ";\n\tpadding-bottom: ", ";\n\tpadding-left: ", ";\n\tpadding-right: ", ";\n\tpadding-top: ", ";\n"]);

  _templateObject8 = function () {
    return data;
  };

  return data;
}

function _templateObject7() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tmin-height: ", ";\n"]);

  _templateObject7 = function () {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tmin-height: ", ";\n"]);

  _templateObject6 = function () {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tmin-height: ", ";\n"]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tmin-height: ", ";\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t\t&:hover,\n\t\t&:focus {\n\t\t\tborder-color: ", ";\n\t\t}\n\t"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\t", ";\n\tborder-color: transparent;\n\tborder-width: ", ";\n\tbox-sizing: border-box;\n\tcursor: pointer;\n\tmin-height: ", ";\n\toutline: none;\n\tposition: relative;\n\ttext-decoration: none;\n\ttransition: background ", " linear,\n\t\tborder-color ", " linear;\n\twidth: 100%;\n\n\ta:hover > &,\n\t&:hover {\n\t\tbackground: ", ";\n\t}\n\n\ta:focus > &,\n\t&:focus {\n\t\t", ";\n\t\tbackground-color: ", ";\n\t\tborder-color: ", ";\n\t\tbox-shadow: ", ";\n\t\tcolor: ", ";\n\t}\n\n\ta:active > &,\n\t&:active {\n\t\tbackground: ", ";\n\t\tborder-color: ", ";\n\t\tbox-shadow: ", ";\n\t\tcolor: ", ";\n\t}\n\n\t&.is-active,\n\t&[aria-current='page'],\n\t&[aria-selected='true'] {\n\t\tbackground-color: ", ";\n\t\tcolor: ", ";\n\n\t\t&:active {\n\t\t\tcolor: ", ";\n\t\t}\n\n\t\t&:hover,\n\t\t&:focus {\n\t\t\tbackground-color: ", ";\n\t\t}\n\n\t\t&:focus {\n\t\t\tborder-color: ", ";\n\t\t}\n\n\t\t&:active {\n\t\t\tbackground-color: ", ";\n\t\t}\n\t}\n\n\t", "\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\toutline: none;\n\tpadding: 0;\n\tposition: relative;\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var Menu = (0, _styles.css)(_templateObject());
exports.Menu = Menu;
var MenuItem = (0, _styles.css)(_templateObject2(), _styles.ui.font.color.text, _styles.ui.font.size(), _styles.ui.get('menuItemBorderWidth'), _styles.ui.get('menuItemHeight'), _styles.ui.get('transitionDurationFastest'), _styles.ui.get('transitionDurationFastest'), _styles.ui.get('controlBackgroundBrightColor'), _styles.ui.zIndex('ControlFocus'), _styles.ui.get('menuItemFocusBackgroundColor'), _styles.ui.get('menuItemFocusBorderColor'), _styles.ui.get('menuItemFocusBoxShadow'), _styles.ui.get('menuItemFocusTextColor'), _styles.ui.get('menuItemActiveBackgroundColor'), _styles.ui.get('menuItemActiveBorderColor'), _styles.ui.get('menuItemActiveBoxShadow'), _styles.ui.get('menuItemActiveTextColor'), _styles.ui.get('surfaceBackgroundSubtleColor'), _styles.ui.color.text, _styles.ui.color.textInverted, _styles.ui.get('surfaceBackgroundSubtleColor'), _styles.ui.get('surfaceBackgroundSubtleColor'), _styles.ui.color.text, (0, _styles.highContrastMode)(_templateObject3(), _styles.ui.get('controlBorderColor')));
exports.MenuItem = MenuItem;
var large = (0, _styles.css)(_templateObject4(), _styles.ui.get('menuItemHeightLarge'));
exports.large = large;
var small = (0, _styles.css)(_templateObject5(), _styles.ui.get('menuItemHeightSmall'));
exports.small = small;
var xSmall = (0, _styles.css)(_templateObject6(), _styles.ui.get('menuItemHeightXSmall'));
exports.xSmall = xSmall;
var xxSmall = (0, _styles.css)(_templateObject7(), _styles.ui.get('menuItemHeightXXSmall'));
exports.xxSmall = xxSmall;
var MenuHeader = (0, _styles.css)(_templateObject8(), _styles.ui.get('menuItemHeight'), _styles.ui.space(1), _styles.ui.get('controlPaddingX'), _styles.ui.get('controlPaddingX'), _styles.ui.space(1));
exports.MenuHeader = MenuHeader;
var offset = (0, _styles.css)(_templateObject9(), _styles.ui.margin.x(-2), _styles.ui.margin.y(-1), _styles.ui.padding.y(2), _styles.ui.get('menuItemHeight'), _styles.ui.space(1), _styles.ui.space(4));
exports.offset = offset;
var showArrow = (0, _styles.css)(_templateObject10(), _styles.ui.padding.right(1));
exports.showArrow = showArrow;
var showBackArrow = (0, _styles.css)(_templateObject11(), _styles.ui.padding.left(1));
exports.showBackArrow = showBackArrow;
