// @wp-g2/components/dist/es/Flex/__stories__/Flex.stories.js
import React from 'react';
import { Placeholder, Spacer } from '../../index';
import { Flex, FlexItem } from '../index';
export default {
  component: Flex,
  title: 'Components/Flex'
};

var ItemView = function (props) {
  return /*#__PURE__*/React.createElement(Placeholder, props);
};

export var _default = function () {
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Spacer, {
    mb: 4
  }, /*#__PURE__*/React.createElement(Flex, {
    gap: 3
  }, /*#__PURE__*/React.createElement(ItemView, null, "Item"), /*#__PURE__*/React.createElement(ItemView, null, "Item"))), /*#__PURE__*/React.createElement(Flex, {
    direction: ['column', 'row'],
    gap: 3
  }, /*#__PURE__*/React.createElement(ItemView, {
    css: {
      width: '180px'
    }
  }, "Item"), /*#__PURE__*/React.createElement(FlexItem, {
    isBlock: true
  }, /*#__PURE__*/React.createElement(ItemView, null, "Item")), /*#__PURE__*/React.createElement(ItemView, null, "Item"), /*#__PURE__*/React.createElement(ItemView, null, "Item")));
};
