// @wp-g2/components/dist/cjs/Slider/Slider.styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.focusedError = exports.error = exports.small = exports.large = exports.focused = exports.Slider = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _styles = require("@wp-g2/styles");

function _templateObject7() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t\t&::-webkit-slider-thumb {\n\t\t\tbox-shadow: ", ";\n\t\t}\n\t\t&::-moz-range-thumb {\n\t\t\tbox-shadow: ", ";\n\t\t}\n\t"]);

  _templateObject7 = function () {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n"]);

  _templateObject6 = function () {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t&::-webkit-slider-runnable-track {\n\t\tbackground: linear-gradient(\n\t\t\tto right,\n\t\t\t", " calc(var(--progress)),\n\t\t\t", " calc(var(--progress))\n\t\t);\n\t}\n\t&::-moz-range-track {\n\t\tbackground: linear-gradient(\n\t\t\tto right,\n\t\t\t", " calc(var(--progress)),\n\t\t\t", " calc(var(--progress))\n\t\t);\n\t}\n\n\t&::-webkit-slider-thumb {\n\t\tbackground-color: ", ";\n\t\tborder: 1px solid ", ";\n\t}\n\t&::-moz-range-thumb {\n\t\tbackground-color: ", ";\n\t\tborder: 1px solid ", ";\n\t}\n\n\t&:focus {\n\t\t", ";\n\t}\n"]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\theight: ", ";\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\theight: ", ";\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", "\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tappearance: none;\n\tbackground-color: transparent;\n\tborder: 1px solid transparent;\n\tborder-radius: ", ";\n\tcursor: pointer;\n\tdisplay: block;\n\theight: ", ";\n\tmax-width: 100%;\n\tmin-width: 0;\n\tpadding: ", ";\n\twidth: 100%;\n\n\t&:focus {\n\t\toutline: none;\n\t}\n\n\t&::-moz-focus-outer {\n\t\tborder: 0;\n\t}\n\n\t&::-webkit-slider-runnable-track {\n\t\tbackground: linear-gradient(\n\t\t\tto right,\n\t\t\t", " calc(var(--progress)),\n\t\t\t", " calc(var(--progress))\n\t\t);\n\t\tborder-radius: 2px;\n\t\theight: 2px;\n\n\t\t*:disabled& {\n\t\t\tbackground: ", ";\n\t\t}\n\t}\n\t&::-moz-range-track {\n\t\tbackground: linear-gradient(\n\t\t\tto right,\n\t\t\t", " calc(var(--progress)),\n\t\t\t", " calc(var(--progress))\n\t\t);\n\t\tborder-radius: 2px;\n\t\theight: 2px;\n\t\twill-change: transform;\n\n\t\t*:disabled& {\n\t\t\tbackground: ", ";\n\t\t}\n\t}\n\n\t&::-webkit-slider-thumb {\n\t\tappearance: none;\n\t\tbackground-color: ", ";\n\t\tborder: 1px solid ", ";\n\t\tborder-radius: 50%;\n\t\tbox-shadow: ", ";\n\t\tcursor: pointer;\n\t\theight: 12px;\n\t\tmargin-top: -5px;\n\t\topacity: 1;\n\t\twidth: 12px;\n\t\ttransition: box-shadow ease ", ";\n\n\t\t*:disabled& {\n\t\t\tbackground: ", ";\n\t\t\tborder-color: ", ";\n\t\t}\n\t}\n\t&::-moz-range-thumb {\n\t\tappearance: none;\n\t\tbackground-color: ", ";\n\t\tborder: 1px solid ", ";\n\t\tborder-radius: 50%;\n\t\tbox-shadow: ", ";\n\t\tcursor: pointer;\n\t\theight: 12px;\n\t\tmargin-top: -5px;\n\t\topacity: 1;\n\t\twidth: 12px;\n\t\ttransition: box-shadow ease ", ";\n\t\twill-change: transform;\n\n\t\t*:disabled& {\n\t\t\tbackground: ", ";\n\t\t\tborder-color: ", ";\n\t\t}\n\t}\n\n\t&:focus {\n\t\t", "\n\t}\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var Slider = (0, _styles.css)(_templateObject(), _styles.ui.get('controlBorderRadius'), _styles.ui.get('controlHeight'), _styles.ui.space(1), _styles.ui.color.admin, _styles.ui.get('controlBackgroundDimColor'), _styles.ui.get('controlBackgroundDimColor'), _styles.ui.color.admin, _styles.ui.get('controlBackgroundDimColor'), _styles.ui.get('controlBackgroundDimColor'), _styles.ui.get('sliderThumbBackgroundColor'), _styles.ui.get('sliderThumbBorderColor'), _styles.ui.get('sliderThumbBoxShadow'), _styles.ui.get('transitionDurationFast'), _styles.ui.get('colorTextMuted'), _styles.ui.get('colorTextMuted'), _styles.ui.get('sliderThumbBackgroundColor'), _styles.ui.get('sliderThumbBorderColor'), _styles.ui.get('sliderThumbBoxShadow'), _styles.ui.get('transitionDurationFast'), _styles.ui.get('colorTextMuted'), _styles.ui.get('colorTextMuted'), getFocusBoxShadow());
exports.Slider = Slider;
var focused = (0, _styles.css)(_templateObject2(), getFocusBoxShadow());
exports.focused = focused;
var large = (0, _styles.css)(_templateObject3(), _styles.ui.get('controlHeightLarge'));
exports.large = large;
var small = (0, _styles.css)(_templateObject4(), _styles.ui.get('controlHeightSmall'));
exports.small = small;
var error = (0, _styles.css)(_templateObject5(), _styles.ui.get('controlDestructiveBorderColor'), _styles.ui.get('controlBackgroundDimColor'), _styles.ui.get('controlDestructiveBorderColor'), _styles.ui.get('controlBackgroundDimColor'), _styles.ui.get('controlDestructiveBorderColor'), _styles.ui.get('controlDestructiveBorderColor'), _styles.ui.get('controlDestructiveBorderColor'), _styles.ui.get('controlDestructiveBorderColor'), getFocusBoxShadow(_styles.ui.get('controlDestructivePseudoBoxShadowFocusSmall')));
exports.error = error;
var focusedError = (0, _styles.css)(_templateObject6(), getFocusBoxShadow(_styles.ui.get('controlDestructivePseudoBoxShadowFocusSmall')));
exports.focusedError = focusedError;

function getFocusBoxShadow(color) {
  if (color === void 0) {
    color = _styles.ui.get('controlPseudoBoxShadowFocusSmall');
  }

  return (0, _styles.css)(_templateObject7(), color, color);
}
