// @wp-g2/components/dist/es/__fixtures__/components/Sidebar.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { ContextSystemProvider } from '@wp-g2/context';
import { ThemeProvider, ui } from '@wp-g2/styles';
import { useLocalState } from '@wp-g2/utils';
import React from 'react';
import { Card, CardBody, ControlLabel, FormGroup, Panel, PanelBody, PanelHeader, Scrollable, Surface, Switch, Tab, TabList, TabPanel, Tabs, TextInput, View } from '../../index';
export var Sidebar = function (_ref) {
  var children = _ref.children;

  var _useLocalState = useLocalState('G2/Components/Sidebar/DarkenedSidebar', false),
      _useLocalState2 = _slicedToArray(_useLocalState, 2),
      darkenedSidebar = _useLocalState2[0],
      setDarkenedSidebar = _useLocalState2[1];

  var _useLocalState3 = useLocalState('G2/Components/Sidebar/outlinedControls', false),
      _useLocalState4 = _slicedToArray(_useLocalState3, 2),
      outlinedControls = _useLocalState4[0],
      setOutlinedControls = _useLocalState4[1];

  var _useLocalState5 = useLocalState('G2/Components/Sidebar/darkMode', false),
      _useLocalState6 = _slicedToArray(_useLocalState5, 2),
      darkMode = _useLocalState6[0],
      setDarkMode = _useLocalState6[1];

  var theme = {
    controlBackgroundColor: outlinedControls && ui.get('surfaceColor'),
    controlBorderColor: outlinedControls && ui.get('surfaceBorderColor'),
    controlSurfaceBoxShadow: outlinedControls ? 'none' : null
  };
  return /*#__PURE__*/React.createElement(ContextSystemProvider, {
    value: {
      Grid: {
        gap: 2
      },
      Icon: {
        size: 16
      }
    }
  }, /*#__PURE__*/React.createElement(View, {
    css: [ui.position.fixed, {
      top: 8,
      left: 8
    }, ui.frame.width(240)]
  }, /*#__PURE__*/React.createElement(Card, null, /*#__PURE__*/React.createElement(CardBody, null, /*#__PURE__*/React.createElement(FormGroup, {
    label: "Darkend Sidebar",
    templateColumns: "1fr 1fr"
  }, /*#__PURE__*/React.createElement(Switch, {
    checked: darkenedSidebar,
    onChange: setDarkenedSidebar
  })), /*#__PURE__*/React.createElement(FormGroup, {
    label: "Outlined Controls",
    templateColumns: "1fr 1fr"
  }, /*#__PURE__*/React.createElement(Switch, {
    checked: outlinedControls,
    onChange: setOutlinedControls
  })), /*#__PURE__*/React.createElement(FormGroup, {
    label: "Dark Mode",
    templateColumns: "1fr 1fr"
  }, /*#__PURE__*/React.createElement(Switch, {
    checked: darkMode,
    onChange: setDarkMode
  }))))), /*#__PURE__*/React.createElement(ThemeProvider, {
    isDark: darkMode,
    theme: theme
  }, /*#__PURE__*/React.createElement(Surface, {
    css: "\n\t\t\t\t\t\twidth: 280px;\n\t\t\t\t\t\tposition: absolute;\n\t\t\t\t\t\theight: 100vh;\n\t\t\t\t\t\ttop: 0;\n\t\t\t\t\t\tright: 0;\n\t\t\t\t\t\tborder-left: 1px solid " + ui.get('colorDivider') + ";\n\t\t\t\t\t",
    variant: darkenedSidebar ? 'secondary' : 'tertiary'
  }, /*#__PURE__*/React.createElement(Tabs, {
    selectedId: "block"
  }, /*#__PURE__*/React.createElement(TabList, null, /*#__PURE__*/React.createElement(Tab, {
    id: "document"
  }, "Document"), /*#__PURE__*/React.createElement(Tab, {
    id: "block"
  }, "Block")), /*#__PURE__*/React.createElement(TabPanel, {
    css: {
      height: "calc(100vh - (" + ui.get('controlHeightXLarge') + " + 1px))"
    }
  }, /*#__PURE__*/React.createElement(Scrollable, {
    css: {
      paddingBottom: '20vh'
    }
  }, /*#__PURE__*/React.createElement(Panel, {
    visible: true
  }, /*#__PURE__*/React.createElement(PanelHeader, {
    title: "Status & Visibility"
  }), /*#__PURE__*/React.createElement(PanelBody, null, /*#__PURE__*/React.createElement(FormGroup, {
    label: "Visibility"
  }, /*#__PURE__*/React.createElement(ControlLabel, {
    as: "div"
  }, "Public")), /*#__PURE__*/React.createElement(FormGroup, {
    label: "Slug"
  }, /*#__PURE__*/React.createElement(TextInput, {
    value: "/my-blog-post"
  })))), /*#__PURE__*/React.createElement(Panel, null, /*#__PURE__*/React.createElement(PanelHeader, {
    title: "Excerpt"
  }), /*#__PURE__*/React.createElement(PanelBody, null, /*#__PURE__*/React.createElement(ControlLabel, null, "Write an excerpt (optional)"), /*#__PURE__*/React.createElement(TextInput, {
    maxRows: 6,
    minRows: 3,
    multiline: true
  }))), /*#__PURE__*/React.createElement(Panel, null, /*#__PURE__*/React.createElement(PanelHeader, {
    title: "Discussion"
  }), /*#__PURE__*/React.createElement(PanelBody, null, "TBD")))), /*#__PURE__*/React.createElement(TabPanel, {
    css: {
      height: "calc(100vh - (" + ui.get('controlHeightXLarge') + " + 1px))"
    }
  }, /*#__PURE__*/React.createElement(Scrollable, {
    css: {
      paddingBottom: '20vh'
    }
  }, children))))));
};
