// @wp-g2/components/dist/cjs/Panel/__stories__/Panel.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _index = require("../../index");

var _index2 = require("../index");

var _default2 = {
  component: _index2.Panel,
  title: 'Components/Panel'
};
exports.default = _default2;

var InputSuffix = function (props) {
  return /*#__PURE__*/_react.default.createElement(_index.Text, Object.assign({
    css: {
      userSelect: 'none'
    },
    isBlock: true,
    lineHeight: 1,
    size: 11,
    variant: "muted"
  }, props));
};

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(_index2.Panel, {
    css: {
      width: 300
    },
    visible: true
  }, /*#__PURE__*/_react.default.createElement(_index2.PanelHeader, {
    title: "Title"
  }), /*#__PURE__*/_react.default.createElement(_index2.PanelBody, null, /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: 2
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    suffix: /*#__PURE__*/_react.default.createElement(InputSuffix, null, "W"),
    value: 300
  }), /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    suffix: /*#__PURE__*/_react.default.createElement(InputSuffix, null, "H"),
    value: 200
  }))), /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: 3
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    suffix: /*#__PURE__*/_react.default.createElement(InputSuffix, null, "X")
  }), /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    suffix: /*#__PURE__*/_react.default.createElement(InputSuffix, null, "Y")
  }), /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    suffix: /*#__PURE__*/_react.default.createElement(InputSuffix, null, "Z")
  }))), /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.Grid, {
    templateColumns: "2fr 1fr"
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, null), /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    suffix: /*#__PURE__*/_react.default.createElement(InputSuffix, null, "%")
  })))));
};

exports._default = _default;
