// @wp-g2/components/dist/cjs/Alert/__stories__/Alert.stories.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _index = require("../../index");

var _index2 = require("../index");

var _default2 = {
  component: _index2.Alert,
  title: 'Components/Alert'
};
exports.default = _default2;
var statuses = ['default', 'success', 'warning', 'info', 'critical'];

var Example = function () {
  var _useState = (0, _react.useState)(statuses),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      alerts = _useState2[0],
      setAlerts = _useState2[1];

  var remove = function (id) {
    return setAlerts(function (prev) {
      return prev.filter(function (i) {
        return i !== id;
      });
    });
  };

  return /*#__PURE__*/_react.default.createElement(_index.View, {
    css: "\n\t\t\t\tmax-width: 480px;\n\t\t\t\tmargin: auto;\n\t\t\t"
  }, /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.Button, {
    disabled: alerts.length === 5,
    onClick: function reset() {
      return setAlerts(statuses);
    }
  }, "Reset")), /*#__PURE__*/_react.default.createElement(_index2.Alerts, null, alerts.map(function (status) {
    return /*#__PURE__*/_react.default.createElement(_index2.Alert, {
      isDismissable: true,
      key: status,
      onDismiss: function onDismiss() {
        remove(status);
      },
      status: status,
      title: "Title"
    }, /*#__PURE__*/_react.default.createElement(_index.Text, null, "Alert Content"));
  })));
};

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(Example, null);
};

exports._default = _default;
