// @wp-g2/context/dist/cjs/__stories__/ComponentsProvider.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Default = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _components = require("@wp-g2/components");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _index = require("../index");

var SomeContext = /*#__PURE__*/_react.default.createContext();

var useSomeContext = function () {
  return _react.default.useContext(SomeContext);
};

var _default = {
  component: _index.ContextSystemProvider,
  title: 'Context/ContextSystemProvider'
};
exports.default = _default;
var innerContext = {
  Card: {
    css: {
      background: 'white'
    }
  }
};

var InnerContent = /*#__PURE__*/_react.default.memo(function () {
  var state = useSomeContext();
  var style = state % 2 === 0 ? {
    background: 'red'
  } : null;
  return /*#__PURE__*/_react.default.createElement(_components.View, {
    css: style
  }, state);
});

var InnerCard = /*#__PURE__*/_react.default.memo(function () {
  return /*#__PURE__*/_react.default.createElement(_components.View, {
    css: {
      padding: 40
    }
  }, /*#__PURE__*/_react.default.createElement(_components.Card, null, /*#__PURE__*/_react.default.createElement(_components.CardBody, {
    css: [{
      border: '3px solid green'
    }]
  }, /*#__PURE__*/_react.default.createElement(InnerContent, null))));
});

var Default = function () {
  var _React$useState = _react.default.useState(0),
      _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
      state = _React$useState2[0],
      update = _React$useState2[1];

  var value = {
    Card: {
      css: {
        background: state % 2 === 0 ? 'blue' : 'yellow'
      }
    },
    CardBody: {
      css: ["\n\t\t\t\tpadding: 20px;\n\t\t\t\t*:hover > & {\n\t\t\t\t\tpadding: 40px;\n\t\t\t\t}\n\t\t\t", _styles.ui.animation.default]
    }
  };
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(SomeContext.Provider, {
    value: state
  }, /*#__PURE__*/_react.default.createElement("button", {
    onClick: function forceUpdate() {
      return update(function (prev) {
        return prev + 1;
      });
    }
  }, "Force Update"), /*#__PURE__*/_react.default.createElement(_index.ContextSystemProvider, {
    value: value
  }, /*#__PURE__*/_react.default.createElement(_components.Card, null, /*#__PURE__*/_react.default.createElement(_components.CardBody, {
    css: [{
      border: '3px solid green'
    }]
  }, /*#__PURE__*/_react.default.createElement(_components.Text, {
    optimizeReadabilityFor: "blue"
  }, "Card"), /*#__PURE__*/_react.default.createElement(_index.ContextSystemProvider, {
    value: innerContext
  }, /*#__PURE__*/_react.default.createElement(InnerCard, null))))), /*#__PURE__*/_react.default.createElement(_components.Card, null, /*#__PURE__*/_react.default.createElement(_components.CardBody, null, /*#__PURE__*/_react.default.createElement(_components.Text, null, "Card")))));
};

exports.Default = Default;
