// @wp-g2/context/dist/cjs/__stories__/ContextSystemProviderLayers.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _components = require("@wp-g2/components");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _index = require("../index");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var _default2 = {
  component: _index.ContextSystemProvider,
  title: 'Context/ContextSystemProvider/Layers'
};
exports.default = _default2;

var Test = function () {
  var _useContextSystem = (0, _index.useContextSystem)({}, 'WPComponentsButton'),
      version = _useContextSystem.version;

  return /*#__PURE__*/_react.default.createElement(_components.Subheading, null, version);
};

var Layer = function (_ref) {
  var children = _ref.children,
      _ref$isActive = _ref.isActive,
      isActive = _ref$isActive === void 0 ? false : _ref$isActive,
      _ref$isEnabled = _ref.isEnabled,
      isEnabled = _ref$isEnabled === void 0 ? false : _ref$isEnabled,
      _ref$label = _ref.label,
      label = _ref$label === void 0 ? 'Layer 1' : _ref$label;
  var nextValue = isActive ? {
    WPComponentsButton: {
      version: 'next'
    }
  } : {
    WPComponentsButton: {
      version: 'current'
    }
  };

  if (!isEnabled) {
    nextValue = {};
  }

  if (!isEnabled && isActive) {
    nextValue = {
      WPComponentsButton: {
        version: 'current'
      }
    };
  }

  return /*#__PURE__*/_react.default.createElement(_index.ContextSystemProvider, {
    value: nextValue
  }, /*#__PURE__*/_react.default.createElement(_components.View, {
    css: [{
      border: '1px solid #ddd'
    }]
  }, /*#__PURE__*/_react.default.createElement(_components.VStack, null, /*#__PURE__*/_react.default.createElement(_components.Text, null, label), /*#__PURE__*/_react.default.createElement(Test, null), /*#__PURE__*/_react.default.createElement(_components.View, null, children))));
};

function App() {
  var _React$useState = _react.default.useState({
    layer1: false,
    layer2: false,
    layer3: false,
    layer4: false,
    layer5: false
  }),
      _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
      versions = _React$useState2[0],
      setVersions = _React$useState2[1];

  var _React$useState3 = _react.default.useState({
    layer1: false,
    layer2: false,
    layer3: false,
    layer4: false,
    layer5: false
  }),
      _React$useState4 = (0, _slicedToArray2.default)(_React$useState3, 2),
      enabled = _React$useState4[0],
      setEnabled = _React$useState4[1];

  var handleOnChange = function (key) {
    return function (value) {
      return setVersions(function (prev) {
        var _objectSpread2;

        return _objectSpread(_objectSpread({}, prev), {}, (_objectSpread2 = {}, _objectSpread2[key] = value, _objectSpread2));
      });
    };
  };

  var handleOnChangeEnabled = function (key) {
    return function (value) {
      setEnabled(function (prev) {
        var _objectSpread3;

        return _objectSpread(_objectSpread({}, prev), {}, (_objectSpread3 = {}, _objectSpread3[key] = value, _objectSpread3));
      });
    };
  };

  return /*#__PURE__*/_react.default.createElement("div", {
    className: "App"
  }, /*#__PURE__*/_react.default.createElement(_components.Container, null, /*#__PURE__*/_react.default.createElement(_components.Spacer, {
    mb: 8
  }, /*#__PURE__*/_react.default.createElement(_components.Grid, {
    columns: 5,
    css: {
      width: '70%'
    },
    gap: 8
  }, /*#__PURE__*/_react.default.createElement(_components.FormGroup, {
    label: "Sidebar Section"
  }, /*#__PURE__*/_react.default.createElement(_components.Checkbox, {
    checked: enabled.layer5,
    label: "Enable",
    onChange: handleOnChangeEnabled('layer5')
  }), enabled.layer5 && /*#__PURE__*/_react.default.createElement(_components.FormGroup, {
    horizontal: true,
    label: "Activate",
    templateColumns: "1fr 60px"
  }, /*#__PURE__*/_react.default.createElement(_components.Switch, {
    checked: versions.layer5,
    disabled: !enabled.layer5,
    onChange: handleOnChange('layer5')
  }))), /*#__PURE__*/_react.default.createElement(_components.FormGroup, {
    label: "Sidebar"
  }, /*#__PURE__*/_react.default.createElement(_components.Checkbox, {
    checked: enabled.layer4,
    label: "Enable",
    onChange: handleOnChangeEnabled('layer4')
  }), enabled.layer4 && /*#__PURE__*/_react.default.createElement(_components.FormGroup, {
    horizontal: true,
    label: "Activate",
    templateColumns: "1fr 60px"
  }, /*#__PURE__*/_react.default.createElement(_components.Switch, {
    checked: versions.layer4,
    disabled: !enabled.layer4,
    onChange: handleOnChange('layer4')
  }))), /*#__PURE__*/_react.default.createElement(_components.FormGroup, {
    label: "Content"
  }, /*#__PURE__*/_react.default.createElement(_components.Checkbox, {
    checked: enabled.layer3,
    label: "Enable",
    onChange: handleOnChangeEnabled('layer3')
  }), enabled.layer3 && /*#__PURE__*/_react.default.createElement(_components.FormGroup, {
    horizontal: true,
    label: "Activate",
    templateColumns: "1fr 60px"
  }, /*#__PURE__*/_react.default.createElement(_components.Switch, {
    checked: versions.layer3,
    disabled: !enabled.layer3,
    onChange: handleOnChange('layer3')
  }))), /*#__PURE__*/_react.default.createElement(_components.FormGroup, {
    label: "Nav menu"
  }, /*#__PURE__*/_react.default.createElement(_components.Checkbox, {
    checked: enabled.layer2,
    label: "Enable",
    onChange: handleOnChangeEnabled('layer2')
  }), enabled.layer2 && /*#__PURE__*/_react.default.createElement(_components.FormGroup, {
    horizontal: true,
    label: "Activate",
    templateColumns: "1fr 60px"
  }, /*#__PURE__*/_react.default.createElement(_components.Switch, {
    checked: versions.layer2,
    disabled: !enabled.layer2,
    onChange: handleOnChange('layer2')
  }))), /*#__PURE__*/_react.default.createElement(_components.FormGroup, {
    label: "App"
  }, /*#__PURE__*/_react.default.createElement(_components.Checkbox, {
    checked: enabled.layer1,
    label: "Enable",
    onChange: handleOnChangeEnabled('layer1')
  }), enabled.layer1 && /*#__PURE__*/_react.default.createElement(_components.FormGroup, {
    horizontal: true,
    label: "Activate",
    templateColumns: "1fr 60px"
  }, /*#__PURE__*/_react.default.createElement(_components.Switch, {
    checked: versions.layer1,
    disabled: !enabled.layer1,
    onChange: handleOnChange('layer1')
  }))))), /*#__PURE__*/_react.default.createElement(Layer, {
    isActive: versions.layer1,
    isEnabled: enabled.layer1,
    label: "App"
  }, /*#__PURE__*/_react.default.createElement(_components.HStack, null, /*#__PURE__*/_react.default.createElement(_components.Button, null, "Button"), /*#__PURE__*/_react.default.createElement(_components.Button, null, "Button"), /*#__PURE__*/_react.default.createElement(_components.Button, null, "Button")), /*#__PURE__*/_react.default.createElement(_components.Grid, {
    templateColumns: "1fr 3fr 1fr"
  }, /*#__PURE__*/_react.default.createElement(Layer, {
    isActive: versions.layer2,
    isEnabled: enabled.layer2,
    label: "Nav Menu"
  }, /*#__PURE__*/_react.default.createElement(_components.VStack, null, /*#__PURE__*/_react.default.createElement(_components.Button, null, "Button"), /*#__PURE__*/_react.default.createElement(_components.Button, null, "Button"), /*#__PURE__*/_react.default.createElement(_components.Button, null, "Button"))), /*#__PURE__*/_react.default.createElement(Layer, {
    isActive: versions.layer3,
    isEnabled: enabled.layer3,
    label: "Content Area"
  }, /*#__PURE__*/_react.default.createElement(_components.HStack, null, /*#__PURE__*/_react.default.createElement(_components.Button, null, "Button"), /*#__PURE__*/_react.default.createElement(_components.Button, null, "Button"), /*#__PURE__*/_react.default.createElement(_components.Button, null, "Button"))), /*#__PURE__*/_react.default.createElement(Layer, {
    isActive: versions.layer4,
    isEnabled: enabled.layer4,
    label: "Sidebar"
  }, /*#__PURE__*/_react.default.createElement(_components.VStack, null, /*#__PURE__*/_react.default.createElement(_components.Button, null, "Button"), /*#__PURE__*/_react.default.createElement(_components.Button, null, "Button"), /*#__PURE__*/_react.default.createElement(_components.Button, null, "Button")), /*#__PURE__*/_react.default.createElement(Layer, {
    isActive: versions.layer5,
    isEnabled: enabled.layer5,
    label: "Sidebar Section"
  }, /*#__PURE__*/_react.default.createElement(_components.VStack, null, /*#__PURE__*/_react.default.createElement(_components.Button, null, "Button"), /*#__PURE__*/_react.default.createElement(_components.Button, null, "Button"), /*#__PURE__*/_react.default.createElement(_components.Button, null, "Button"))))))));
}

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(App, null);
};

exports._default = _default;
