// @wordpress/components/build-module/mobile/bottom-sheet/range-cell.native.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import { createElement } from "@wordpress/element";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

/**
 * External dependencies
 */
import { Platform, AccessibilityInfo, View } from 'react-native';
import Slider from '@react-native-community/slider';
/**
 * WordPress dependencies
 */

import { __, sprintf } from '@wordpress/i18n';
import { Component } from '@wordpress/element';
import { withPreferredColorScheme } from '@wordpress/compose';
/**
 * Internal dependencies
 */

import Cell from './cell';
import styles from './range-cell.scss';
import RangeTextInput from './range-text-input';
import { toFixed } from '../utils';
var isIOS = Platform.OS === 'ios';

var BottomSheetRangeCell = /*#__PURE__*/function (_Component) {
  _inherits(BottomSheetRangeCell, _Component);

  var _super = _createSuper(BottomSheetRangeCell);

  function BottomSheetRangeCell(props) {
    var _this;

    _classCallCheck(this, BottomSheetRangeCell);

    _this = _super.call(this, props);
    _this.onSliderChange = _this.onSliderChange.bind(_assertThisInitialized(_this));
    _this.onTextInputChange = _this.onTextInputChange.bind(_assertThisInitialized(_this));
    _this.a11yIncrementValue = _this.a11yIncrementValue.bind(_assertThisInitialized(_this));
    _this.a11yDecrementValue = _this.a11yDecrementValue.bind(_assertThisInitialized(_this));
    _this.a11yUpdateValue = _this.a11yUpdateValue.bind(_assertThisInitialized(_this));
    var value = props.value,
        defaultValue = props.defaultValue,
        minimumValue = props.minimumValue;
    var initialValue = Number(value || defaultValue || minimumValue);
    _this.state = {
      inputValue: initialValue,
      sliderValue: initialValue
    };
    return _this;
  }

  _createClass(BottomSheetRangeCell, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      clearTimeout(this.timeoutAnnounceValue);
    }
  }, {
    key: "onSliderChange",
    value: function onSliderChange(initialValue) {
      var _this$props = this.props,
          decimalNum = _this$props.decimalNum,
          onChange = _this$props.onChange;
      initialValue = toFixed(initialValue, decimalNum);
      this.setState({
        inputValue: initialValue
      });
      onChange(initialValue);
    }
  }, {
    key: "onTextInputChange",
    value: function onTextInputChange(nextValue) {
      var _this$props2 = this.props,
          onChange = _this$props2.onChange,
          onComplete = _this$props2.onComplete;
      this.setState({
        sliderValue: nextValue
      });
      onChange(nextValue);

      if (onComplete) {
        onComplete(nextValue);
      }
    }
    /*
     * Only used with screenreaders like VoiceOver and TalkBack. Increments the
     * value of this setting programmatically.
     */

  }, {
    key: "a11yIncrementValue",
    value: function a11yIncrementValue() {
      var _this$props3 = this.props,
          _this$props3$step = _this$props3.step,
          step = _this$props3$step === void 0 ? 5 : _this$props3$step,
          maximumValue = _this$props3.maximumValue,
          decimalNum = _this$props3.decimalNum;
      var inputValue = this.state.inputValue;
      var newValue = toFixed(inputValue + step, decimalNum);

      if (newValue <= maximumValue || maximumValue === undefined) {
        this.a11yUpdateValue(newValue);
      }
    }
    /*
     * Only used with screenreaders like VoiceOver and TalkBack. Decrements the
     * value of this setting programmatically.
     */

  }, {
    key: "a11yDecrementValue",
    value: function a11yDecrementValue() {
      var _this$props4 = this.props,
          _this$props4$step = _this$props4.step,
          step = _this$props4$step === void 0 ? 5 : _this$props4$step,
          minimumValue = _this$props4.minimumValue,
          decimalNum = _this$props4.decimalNum;
      var sliderValue = this.state.sliderValue;
      var newValue = toFixed(sliderValue - step, decimalNum);

      if (newValue >= minimumValue) {
        this.a11yUpdateValue(newValue);
      }
    }
  }, {
    key: "a11yUpdateValue",
    value: function a11yUpdateValue(newValue) {
      var _this$props5 = this.props,
          onChange = _this$props5.onChange,
          onComplete = _this$props5.onComplete;
      this.setState({
        sliderValue: newValue,
        inputValue: newValue
      });
      onChange(newValue);

      if (onComplete) {
        onComplete(newValue);
      }

      this.announceValue(newValue);
    }
    /*
     * Only used with screenreaders like VoiceOver and TalkBack.
     */

  }, {
    key: "announceValue",
    value: function announceValue(value) {
      var _this$props6 = this.props,
          label = _this$props6.label,
          _this$props6$unitLabe = _this$props6.unitLabel,
          unitLabel = _this$props6$unitLabe === void 0 ? '' : _this$props6$unitLabe;

      if (isIOS) {
        // On Android it triggers the accessibilityLabel with the value change, but
        // on iOS we need to do this manually.
        clearTimeout(this.timeoutAnnounceValue);
        this.timeoutAnnounceValue = setTimeout(function () {
          AccessibilityInfo.announceForAccessibility("".concat(value, " ").concat(unitLabel, ",  ").concat(label));
        }, 300);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props7 = this.props,
          value = _this$props7.value,
          defaultValue = _this$props7.defaultValue,
          _this$props7$minimumV = _this$props7.minimumValue,
          minimumValue = _this$props7$minimumV === void 0 ? 0 : _this$props7$minimumV,
          _this$props7$maximumV = _this$props7.maximumValue,
          maximumValue = _this$props7$maximumV === void 0 ? 10 : _this$props7$maximumV,
          disabled = _this$props7.disabled,
          _this$props7$step = _this$props7.step,
          step = _this$props7$step === void 0 ? 1 : _this$props7$step,
          preferredColorScheme = _this$props7.preferredColorScheme,
          _this$props7$minimumT = _this$props7.minimumTrackTintColor,
          minimumTrackTintColor = _this$props7$minimumT === void 0 ? preferredColorScheme === 'light' ? '#00669b' : '#5198d9' : _this$props7$minimumT,
          _this$props7$maximumT = _this$props7.maximumTrackTintColor,
          maximumTrackTintColor = _this$props7$maximumT === void 0 ? isIOS ? '#e9eff3' : '#909090' : _this$props7$maximumT,
          _this$props7$thumbTin = _this$props7.thumbTintColor,
          thumbTintColor = _this$props7$thumbTin === void 0 ? !isIOS && '#00669b' : _this$props7$thumbTin,
          preview = _this$props7.preview,
          cellContainerStyle = _this$props7.cellContainerStyle,
          onComplete = _this$props7.onComplete,
          _this$props7$shouldDi = _this$props7.shouldDisplayTextInput,
          shouldDisplayTextInput = _this$props7$shouldDi === void 0 ? true : _this$props7$shouldDi,
          _this$props7$unitLabe = _this$props7.unitLabel,
          unitLabel = _this$props7$unitLabe === void 0 ? '' : _this$props7$unitLabe,
          _this$props7$settingL = _this$props7.settingLabel,
          settingLabel = _this$props7$settingL === void 0 ? 'Value' : _this$props7$settingL,
          openUnitPicker = _this$props7.openUnitPicker,
          children = _this$props7.children,
          decimalNum = _this$props7.decimalNum,
          cellProps = _objectWithoutProperties(_this$props7, ["value", "defaultValue", "minimumValue", "maximumValue", "disabled", "step", "preferredColorScheme", "minimumTrackTintColor", "maximumTrackTintColor", "thumbTintColor", "preview", "cellContainerStyle", "onComplete", "shouldDisplayTextInput", "unitLabel", "settingLabel", "openUnitPicker", "children", "decimalNum"]);

      var _this$state = this.state,
          inputValue = _this$state.inputValue,
          sliderValue = _this$state.sliderValue;

      var getAccessibilityHint = function getAccessibilityHint() {
        return openUnitPicker ? __('double-tap to change unit') : '';
      };

      var getAccessibilityLabel = function getAccessibilityLabel() {
        return sprintf(
        /* translators: accessibility text. Inform about current value. %1$s: Control label %2$s: setting label (example: width), %3$s: Current value. %4$s: value measurement unit (example: pixels) */
        __('%1$s. %2$s is %3$s %4$s.'), cellProps.label, settingLabel, value, unitLabel);
      };

      var containerStyle = [styles.container, isIOS ? styles.containerIOS : styles.containerAndroid];
      return createElement(View, {
        accessible: true,
        accessibilityRole: "adjustable",
        accessibilityActions: [{
          name: 'increment'
        }, {
          name: 'decrement'
        }, {
          name: 'activate'
        }],
        onAccessibilityAction: function onAccessibilityAction(event) {
          switch (event.nativeEvent.actionName) {
            case 'increment':
              _this2.a11yIncrementValue();

              break;

            case 'decrement':
              _this2.a11yDecrementValue();

              break;

            case 'activate':
              openUnitPicker();
              break;
          }
        },
        accessibilityLabel: getAccessibilityLabel(),
        accessibilityHint: getAccessibilityHint()
      }, createElement(View, {
        importantForAccessibility: "no-hide-descendants"
      }, createElement(Cell, _extends({}, cellProps, {
        cellContainerStyle: [styles.cellContainerStyles, cellContainerStyle],
        cellRowContainerStyle: containerStyle,
        leftAlign: true,
        editable: false,
        activeOpacity: 1,
        accessible: false,
        valueStyle: styles.valueStyle
      }), createElement(View, {
        style: containerStyle
      }, preview, createElement(Slider, {
        value: sliderValue,
        defaultValue: defaultValue,
        disabled: disabled,
        step: step,
        minimumValue: minimumValue,
        maximumValue: maximumValue,
        minimumTrackTintColor: minimumTrackTintColor,
        maximumTrackTintColor: maximumTrackTintColor,
        thumbTintColor: thumbTintColor,
        onValueChange: this.onSliderChange,
        onSlidingComplete: onComplete,
        ref: function ref(slider) {
          _this2.sliderRef = slider;
        },
        style: isIOS ? styles.sliderIOS : styles.sliderAndroid
      }), shouldDisplayTextInput && createElement(RangeTextInput, {
        label: cellProps.label,
        onChange: this.onTextInputChange,
        defaultValue: "".concat(inputValue),
        value: inputValue,
        min: minimumValue,
        max: maximumValue,
        step: step,
        decimalNum: decimalNum
      }, children)))));
    }
  }]);

  return BottomSheetRangeCell;
}(Component);

export default withPreferredColorScheme(BottomSheetRangeCell);
