// @wordpress/components/build-module/mobile/bottom-sheet/range-text-input.native.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import { createElement } from "@wordpress/element";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

/**
 * External dependencies
 */
import { AccessibilityInfo, View, TextInput, PixelRatio, AppState, Platform, Text, TouchableWithoutFeedback } from 'react-native';
/**
 * WordPress dependencies
 */

import { Component } from '@wordpress/element';
import { withPreferredColorScheme } from '@wordpress/compose';
import { __, sprintf } from '@wordpress/i18n';
/**
 * Internal dependencies
 */

import { toFixed, removeNonDigit } from '../utils';
import styles from './styles.scss';
import borderStyles from './borderStyles.scss';
var isIOS = Platform.OS === 'ios';

var RangeTextInput = /*#__PURE__*/function (_Component) {
  _inherits(RangeTextInput, _Component);

  var _super = _createSuper(RangeTextInput);

  function RangeTextInput(props) {
    var _this;

    _classCallCheck(this, RangeTextInput);

    _this = _super.call(this, props);
    _this.announceCurrentValue = _this.announceCurrentValue.bind(_assertThisInitialized(_this));
    _this.onInputFocus = _this.onInputFocus.bind(_assertThisInitialized(_this));
    _this.onInputBlur = _this.onInputBlur.bind(_assertThisInitialized(_this));
    _this.handleChangePixelRatio = _this.handleChangePixelRatio.bind(_assertThisInitialized(_this));
    _this.onSubmitEditing = _this.onSubmitEditing.bind(_assertThisInitialized(_this));
    _this.onChangeText = _this.onChangeText.bind(_assertThisInitialized(_this));
    var value = props.value,
        defaultValue = props.defaultValue,
        min = props.min,
        decimalNum = props.decimalNum;
    var initialValue = toFixed(value || defaultValue || min, decimalNum);

    var fontScale = _this.getFontScale();

    _this.state = {
      fontScale: fontScale,
      inputValue: initialValue,
      controlValue: initialValue,
      hasFocus: false
    };
    return _this;
  }

  _createClass(RangeTextInput, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      AppState.addEventListener('change', this.handleChangePixelRatio);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      AppState.removeEventListener('change', this.handleChangePixelRatio);
      clearTimeout(this.timeoutAnnounceValue);
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps, prevState) {
      var value = this.props.value;
      var _this$state = this.state,
          hasFocus = _this$state.hasFocus,
          inputValue = _this$state.inputValue;

      if (prevProps.value !== value) {
        this.setState({
          inputValue: value
        });
      }

      if (prevState.hasFocus !== hasFocus) {
        var validValue = this.validateInput(inputValue);
        this.setState({
          inputValue: validValue
        });
      }

      if (!prevState.hasFocus && hasFocus) {
        this._valueTextInput.focus();
      }
    }
  }, {
    key: "getFontScale",
    value: function getFontScale() {
      return PixelRatio.getFontScale() < 1 ? 1 : PixelRatio.getFontScale();
    }
  }, {
    key: "handleChangePixelRatio",
    value: function handleChangePixelRatio(nextAppState) {
      if (nextAppState === 'active') {
        this.setState({
          fontScale: this.getFontScale()
        });
      }
    }
  }, {
    key: "onInputFocus",
    value: function onInputFocus() {
      this.setState({
        hasFocus: true
      });
    }
  }, {
    key: "onInputBlur",
    value: function onInputBlur() {
      var inputValue = this.state.inputValue;
      this.onChangeText("".concat(inputValue));
      this.setState({
        hasFocus: false
      });
    }
  }, {
    key: "validateInput",
    value: function validateInput(text) {
      var _this$props = this.props,
          min = _this$props.min,
          max = _this$props.max,
          decimalNum = _this$props.decimalNum;
      var result = min;

      if (!text) {
        return min;
      }

      if (typeof text === 'number') {
        result = Math.max(text, min);
        return max ? Math.min(result, max) : result;
      }

      result = Math.max(removeNonDigit(text, decimalNum), min);
      return max ? Math.min(result, max) : result;
    }
  }, {
    key: "updateValue",
    value: function updateValue(value) {
      var onChange = this.props.onChange;
      var validValue = this.validateInput(value);
      this.announceCurrentValue("".concat(validValue));
      onChange(validValue);
    }
  }, {
    key: "onChangeText",
    value: function onChangeText(textValue) {
      var decimalNum = this.props.decimalNum;
      var inputValue = removeNonDigit(textValue, decimalNum);
      textValue = inputValue.replace(',', '.');
      textValue = toFixed(textValue, decimalNum);
      var value = this.validateInput(textValue);
      this.setState({
        inputValue: inputValue,
        controlValue: value
      });
      this.updateValue(value);
    }
  }, {
    key: "onSubmitEditing",
    value: function onSubmitEditing(_ref) {
      var text = _ref.nativeEvent.text;
      var decimalNum = this.props.decimalNum;
      var inputValue = this.state.inputValue;

      if (!isNaN(Number(text))) {
        text = toFixed(text.replace(',', '.'), decimalNum);
        var validValue = this.validateInput(text);

        if (inputValue !== validValue) {
          this.setState({
            inputValue: validValue
          });
          this.announceCurrentValue("".concat(validValue));
          this.props.onChange(validValue);
        }
      }
    }
  }, {
    key: "announceCurrentValue",
    value: function announceCurrentValue(value) {
      /* translators: %s: current cell value. */
      var announcement = sprintf(__('Current value is %s'), value);
      AccessibilityInfo.announceForAccessibility(announcement);
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props2 = this.props,
          getStylesFromColorScheme = _this$props2.getStylesFromColorScheme,
          children = _this$props2.children;
      var _this$state2 = this.state,
          fontScale = _this$state2.fontScale,
          inputValue = _this$state2.inputValue,
          hasFocus = _this$state2.hasFocus;
      var textInputStyle = getStylesFromColorScheme(styles.textInput, styles.textInputDark);
      var verticalBorderStyle = getStylesFromColorScheme(styles.verticalBorder, styles.verticalBorderDark);
      var inputBorderStyles = [textInputStyle, borderStyles.borderStyle, hasFocus && borderStyles.isSelected];
      var valueFinalStyle = [!isIOS ? inputBorderStyles : verticalBorderStyle, {
        width: 50 * fontScale
      }];
      return createElement(TouchableWithoutFeedback, {
        onPress: this.onInputFocus,
        accessible: false
      }, createElement(View, {
        style: [styles.textInputContainer, isIOS && inputBorderStyles],
        accessible: false
      }, isIOS || hasFocus ? createElement(TextInput, {
        ref: function ref(c) {
          return _this2._valueTextInput = c;
        },
        style: valueFinalStyle,
        onChangeText: this.onChangeText,
        onSubmitEditing: this.onSubmitEditing,
        onFocus: this.onInputFocus,
        onBlur: this.onInputBlur,
        keyboardType: "numeric",
        returnKeyType: "done",
        numberOfLines: 1,
        defaultValue: "".concat(inputValue),
        value: inputValue,
        pointerEvents: hasFocus ? 'auto' : 'none'
      }) : createElement(Text, {
        style: valueFinalStyle,
        numberOfLines: 1,
        ellipsizeMode: "clip"
      }, inputValue), children));
    }
  }]);

  return RangeTextInput;
}(Component);

export default withPreferredColorScheme(RangeTextInput);
