// @wordpress/block-editor/build-module/components/block-mover/index.native.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement, Fragment } from "@wordpress/element";

/**
 * External dependencies
 */
import { first, last, partial, castArray } from 'lodash';
import { Platform } from 'react-native';
/**
 * WordPress dependencies
 */

import { __ } from '@wordpress/i18n';
import { Picker, ToolbarButton } from '@wordpress/components';
import { withInstanceId, compose } from '@wordpress/compose';
import { withSelect, withDispatch } from '@wordpress/data';
import { useRef, useState } from '@wordpress/element';
/**
 * Internal dependencies
 */

import { getMoversSetup } from './mover-description';
import { store as blockEditorStore } from '../../store';
export var BLOCK_MOVER_DIRECTION_TOP = 'blockPageMoverOptions-moveToTop';
export var BLOCK_MOVER_DIRECTION_BOTTOM = 'blockPageMoverOptions-moveToBottom';
export var BlockMover = function BlockMover(_ref) {
  var isFirst = _ref.isFirst,
      isLast = _ref.isLast,
      isLocked = _ref.isLocked,
      onMoveDown = _ref.onMoveDown,
      onMoveUp = _ref.onMoveUp,
      onLongMove = _ref.onLongMove,
      firstIndex = _ref.firstIndex,
      numberOfBlocks = _ref.numberOfBlocks,
      rootClientId = _ref.rootClientId,
      isStackedHorizontally = _ref.isStackedHorizontally;
  var pickerRef = useRef();

  var _useState = useState(undefined),
      _useState2 = _slicedToArray(_useState, 2),
      blockPageMoverState = _useState2[0],
      setBlockPageMoverState = _useState2[1];

  var showBlockPageMover = function showBlockPageMover(direction) {
    return function () {
      if (!pickerRef.current) {
        setBlockPageMoverState(undefined);
        return;
      }

      setBlockPageMoverState(direction);
      pickerRef.current.presentPicker();
    };
  };

  var _getMoversSetup = getMoversSetup(isStackedHorizontally, {
    firstIndex: firstIndex
  }),
      _getMoversSetup$descr = _getMoversSetup.description,
      backwardButtonHint = _getMoversSetup$descr.backwardButtonHint,
      forwardButtonHint = _getMoversSetup$descr.forwardButtonHint,
      firstBlockTitle = _getMoversSetup$descr.firstBlockTitle,
      lastBlockTitle = _getMoversSetup$descr.lastBlockTitle,
      _getMoversSetup$icon = _getMoversSetup.icon,
      backwardButtonIcon = _getMoversSetup$icon.backward,
      forwardButtonIcon = _getMoversSetup$icon.forward,
      _getMoversSetup$title = _getMoversSetup.title,
      backwardButtonTitle = _getMoversSetup$title.backward,
      forwardButtonTitle = _getMoversSetup$title.forward;

  var blockPageMoverOptions = [{
    icon: backwardButtonIcon,
    label: __('Move to top'),
    value: BLOCK_MOVER_DIRECTION_TOP,
    onSelect: function onSelect() {
      onLongMove()(0);
    }
  }, {
    icon: forwardButtonIcon,
    label: __('Move to bottom'),
    value: BLOCK_MOVER_DIRECTION_BOTTOM,
    onSelect: function onSelect() {
      onLongMove()(numberOfBlocks);
    }
  }].filter(function (el) {
    return el.value === blockPageMoverState;
  });

  var onPickerSelect = function onPickerSelect(value) {
    var option = blockPageMoverOptions.find(function (el) {
      return el.value === value;
    });
    if (option && option.onSelect) option.onSelect();
  };

  if (isLocked || isFirst && isLast && !rootClientId) {
    return null;
  }

  return createElement(Fragment, null, createElement(ToolbarButton, {
    title: !isFirst ? backwardButtonTitle : firstBlockTitle,
    isDisabled: isFirst,
    onClick: onMoveUp,
    onLongPress: showBlockPageMover(BLOCK_MOVER_DIRECTION_TOP),
    icon: backwardButtonIcon,
    extraProps: {
      hint: backwardButtonHint
    }
  }), createElement(ToolbarButton, {
    title: !isLast ? forwardButtonTitle : lastBlockTitle,
    isDisabled: isLast,
    onClick: onMoveDown,
    onLongPress: showBlockPageMover(BLOCK_MOVER_DIRECTION_BOTTOM),
    icon: forwardButtonIcon,
    extraProps: {
      hint: forwardButtonHint
    }
  }), createElement(Picker, {
    ref: pickerRef,
    options: blockPageMoverOptions,
    onChange: onPickerSelect,
    title: __('Change block position'),
    leftAlign: true,
    hideCancelButton: Platform.OS !== 'ios'
  }));
};
export default compose(withSelect(function (select, _ref2) {
  var clientIds = _ref2.clientIds;

  var _select = select(blockEditorStore),
      getBlockIndex = _select.getBlockIndex,
      getTemplateLock = _select.getTemplateLock,
      getBlockRootClientId = _select.getBlockRootClientId,
      getBlockOrder = _select.getBlockOrder;

  var normalizedClientIds = castArray(clientIds);
  var firstClientId = first(normalizedClientIds);
  var rootClientId = getBlockRootClientId(firstClientId);
  var blockOrder = getBlockOrder(rootClientId);
  var firstIndex = getBlockIndex(firstClientId, rootClientId);
  var lastIndex = getBlockIndex(last(normalizedClientIds), rootClientId);
  return {
    firstIndex: firstIndex,
    numberOfBlocks: blockOrder.length - 1,
    isFirst: firstIndex === 0,
    isLast: lastIndex === blockOrder.length - 1,
    isLocked: getTemplateLock(rootClientId) === 'all',
    rootClientId: rootClientId
  };
}), withDispatch(function (dispatch, _ref3) {
  var clientIds = _ref3.clientIds,
      rootClientId = _ref3.rootClientId;

  var _dispatch = dispatch(blockEditorStore),
      moveBlocksDown = _dispatch.moveBlocksDown,
      moveBlocksUp = _dispatch.moveBlocksUp,
      moveBlocksToPosition = _dispatch.moveBlocksToPosition;

  return {
    onMoveDown: partial(moveBlocksDown, clientIds, rootClientId),
    onMoveUp: partial(moveBlocksUp, clientIds, rootClientId),
    onLongMove: function onLongMove(targetIndex) {
      return partial(moveBlocksToPosition, clientIds, rootClientId, targetIndex);
    }
  };
}), withInstanceId)(BlockMover);
