// @wp-g2/components/dist/cjs/SegmentedControl/__stories__/SegmentedControl.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _icons = require("@wordpress/icons");

var _react = _interopRequireDefault(require("react"));

var _index = require("../../index");

var _index2 = require("../index");

var _default2 = {
  component: _index2.SegmentedControl,
  title: 'Components/SegmentedControl'
};
exports.default = _default2;
var align = [{
  label: /*#__PURE__*/_react.default.createElement(_index.Icon, {
    icon: _icons.alignLeft,
    size: 14
  }),
  value: 'left'
}, {
  label: /*#__PURE__*/_react.default.createElement(_index.Icon, {
    icon: _icons.alignCenter,
    size: 14
  }),
  value: 'center'
}, {
  label: /*#__PURE__*/_react.default.createElement(_index.Icon, {
    icon: _icons.alignRight,
    size: 14
  }),
  value: 'right'
}, {
  label: /*#__PURE__*/_react.default.createElement(_index.Icon, {
    icon: _icons.alignJustify,
    size: 14
  }),
  value: 'justify'
}];

var _default = function () {
  var _React$useState = _react.default.useState(align[0].value),
      _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
      alignState = _React$useState2[0],
      setAlignState = _React$useState2[1];

  return /*#__PURE__*/_react.default.createElement(_index.View, null, /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    css: {
      width: [null, 300]
    }
  }, /*#__PURE__*/_react.default.createElement(_index2.SegmentedControl, {
    isBlock: true,
    onChange: setAlignState,
    options: align,
    value: alignState
  })), /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    css: {
      width: [null, 300]
    }
  }, /*#__PURE__*/_react.default.createElement(_index2.SegmentedControl, {
    isBlock: true,
    onChange: setAlignState,
    options: align,
    value: alignState
  })), /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index2.SegmentedControl, {
    options: [{
      label: 'Horizontal',
      value: 'horizontal'
    }, {
      label: 'Vertical',
      value: 'vertical'
    }]
  })), /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index2.SegmentedControl, {
    isAdaptiveWidth: true,
    options: [{
      label: 'Short',
      value: 'short'
    }, {
      label: 'Looooooooooooong',
      value: 'long'
    }]
  })));
};

exports._default = _default;
