// @wordpress/block-editor/build/components/block-list/block-selection-button.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _components = require("@wordpress/components");

var _data = require("@wordpress/data");

var _compose = require("@wordpress/compose");

var _blocks = require("@wordpress/blocks");

var _reactNative = require("react-native");

var _blockTitle = _interopRequireDefault(require("../block-title"));

var _useBlockDisplayInformation = _interopRequireDefault(require("../use-block-display-information"));

var _subdirectoryIcon = _interopRequireDefault(require("./subdirectory-icon"));

var _store = require("../../store");

var _blockSelectionButton = _interopRequireDefault(require("./block-selection-button.scss"));

/**
 * WordPress dependencies
 */

/**
 * External dependencies
 */

/**
 * Internal dependencies
 */
var BlockSelectionButton = function BlockSelectionButton(_ref) {
  var _blockInformation$ico;

  var clientId = _ref.clientId,
      rootClientId = _ref.rootClientId,
      rootBlockIcon = _ref.rootBlockIcon,
      isRTL = _ref.isRTL;
  var blockInformation = (0, _useBlockDisplayInformation.default)(clientId);
  return (0, _element.createElement)(_reactNative.View, {
    style: [_blockSelectionButton.default.selectionButtonContainer, rootClientId && _blockSelectionButton.default.densedPaddingLeft]
  }, (0, _element.createElement)(_reactNative.TouchableOpacity, {
    style: _blockSelectionButton.default.button,
    onPress: function onPress() {
      /* Open BottomSheet with markup */
    },
    disabled: true
    /* Disable temporarily since onPress function is empty */

  }, rootClientId && rootBlockIcon && [(0, _element.createElement)(_components.Icon, {
    key: "parent-icon",
    size: 24,
    icon: rootBlockIcon.src,
    fill: _blockSelectionButton.default.icon.color
  }), (0, _element.createElement)(_reactNative.View, {
    key: "subdirectory-icon",
    style: _blockSelectionButton.default.arrow
  }, (0, _element.createElement)(_subdirectoryIcon.default, {
    fill: _blockSelectionButton.default.arrow.color,
    isRTL: isRTL
  }))], (0, _element.createElement)(_components.Icon, {
    size: 24,
    icon: blockInformation === null || blockInformation === void 0 ? void 0 : (_blockInformation$ico = blockInformation.icon) === null || _blockInformation$ico === void 0 ? void 0 : _blockInformation$ico.src,
    fill: _blockSelectionButton.default.icon.color
  }), (0, _element.createElement)(_reactNative.Text, {
    maxFontSizeMultiplier: 1.25,
    ellipsizeMode: "tail",
    numberOfLines: 1,
    style: _blockSelectionButton.default.selectionButtonTitle
  }, (0, _element.createElement)(_blockTitle.default, {
    clientId: clientId
  }))));
};

var _default = (0, _compose.compose)([(0, _data.withSelect)(function (select, _ref2) {
  var clientId = _ref2.clientId;

  var _select = select(_store.store),
      getBlockRootClientId = _select.getBlockRootClientId,
      getBlockName = _select.getBlockName,
      getSettings = _select.getSettings;

  var rootClientId = getBlockRootClientId(clientId);

  if (!rootClientId) {
    return {
      clientId: clientId
    };
  }

  var rootBlockName = getBlockName(rootClientId);
  var rootBlockType = (0, _blocks.getBlockType)(rootBlockName);
  var rootBlockIcon = rootBlockType ? rootBlockType.icon : {};
  return {
    clientId: clientId,
    rootClientId: rootClientId,
    rootBlockIcon: rootBlockIcon,
    isRTL: getSettings().isRTL
  };
})])(BlockSelectionButton);

exports.default = _default;
