// @wp-g2/components/dist/cjs/__stories__/__References/Backgrounds.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _index = require("../../index");

function _templateObject4() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t\t\t", "\n\t\t"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t\t\t", "\n\t\t"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t\t\t", "\n\t\t"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t\t\t", "\n\t\t"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var _default2 = {
  title: 'References/Background'
};
exports.default = _default2;

var ColorSet = function (_ref) {
  var color = _ref.color;
  var classes = (0, _styles.cx)((0, _styles.css)(_templateObject(), (0, _styles.getBackgroundColor)(color)), (0, _styles.css)(_templateObject2(), (0, _styles.getBackgroundColorText)(color)));
  return /*#__PURE__*/_react.default.createElement(_index.Card, null, /*#__PURE__*/_react.default.createElement(_index.CardBody, {
    className: classes
  }, color));
};

var ColorSetBold = function (_ref2) {
  var color = _ref2.color;
  var classes = (0, _styles.cx)((0, _styles.css)(_templateObject3(), (0, _styles.getBackgroundColor)(color, {
    isBold: true
  })), (0, _styles.css)(_templateObject4(), (0, _styles.getBackgroundColorText)(color, {
    isBold: true
  })));
  return /*#__PURE__*/_react.default.createElement(_index.Card, null, /*#__PURE__*/_react.default.createElement(_index.CardBody, {
    className: classes
  }, color));
};

var ColorCollection = function (_ref3) {
  var title = _ref3.title;
  return /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    css: {
      position: 'relative'
    },
    mb: 8,
    p: 5
  }, /*#__PURE__*/_react.default.createElement(_index.View, {
    css: {
      margin: 'auto',
      maxWidth: 1200,
      position: 'relative',
      zIndex: 1
    }
  }, /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.Heading, {
    size: 2
  }, title)), /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    mb: 8
  }, /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: [1, 2, 4]
  }, _styles.SUPPORTED_COLORS.map(function (color) {
    return /*#__PURE__*/_react.default.createElement(ColorSet, {
      color: color,
      key: color
    });
  }))), /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.Heading, {
    size: 4
  }, "Bold")), /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    mb: 8
  }, /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: [1, 2, 4]
  }, _styles.SUPPORTED_COLORS.map(function (color) {
    return /*#__PURE__*/_react.default.createElement(ColorSetBold, {
      color: color,
      key: color
    });
  })))), /*#__PURE__*/_react.default.createElement(_index.Surface, {
    css: {
      bottom: 0,
      left: 0,
      position: 'absolute',
      right: 0,
      top: 0,
      zIndex: 0
    }
  }));
};

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(ColorCollection, {
    title: "Backgrounds"
  }), /*#__PURE__*/_react.default.createElement(_styles.ThemeProvider, {
    isDark: true,
    isGlobal: false
  }, /*#__PURE__*/_react.default.createElement(ColorCollection, {
    title: "Backgrounds + Dark"
  })));
};

exports._default = _default;
