// @wp-g2/components/dist/cjs/__stories__/__References/Colors.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _index = require("../../index");

var _default2 = {
  title: 'References/Colors'
};
exports.default = _default2;
var grades = [100, 300, 500, 700, 900];
var alphas = [10, 20, 30, 40, 50, 60, 70, 80, 90];

var ColorSet = function (_ref) {
  var color = _ref.color;
  return /*#__PURE__*/_react.default.createElement(_index.Card, null, grades.map(function (grade) {
    return /*#__PURE__*/_react.default.createElement(_index.CardBody, {
      key: grade,
      style: {
        background: (0, _styles.get)("" + color + grade),
        color: (0, _styles.get)("" + color + grade + "Text")
      }
    }, /*#__PURE__*/_react.default.createElement(_index.Flex, null, /*#__PURE__*/_react.default.createElement(_index.FlexItem, null, color, grade)));
  }));
};

var ColorRGBSet = function (_ref2) {
  var color = _ref2.color;
  return /*#__PURE__*/_react.default.createElement(_index.Card, null, alphas.map(function (alpha) {
    return /*#__PURE__*/_react.default.createElement(_index.CardBody, {
      key: alpha,
      style: {
        background: (0, _styles.get)(color + "Rgba" + alpha),
        color: (0, _styles.get)("" + color + alpha + "Text")
      }
    }, /*#__PURE__*/_react.default.createElement(_index.Flex, null, /*#__PURE__*/_react.default.createElement(_index.FlexItem, null, color + "Rgba" + alpha)));
  }));
};

var ColorCollection = function (_ref3) {
  var title = _ref3.title;
  return /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    css: {
      position: 'relative'
    },
    mb: 8,
    p: 5
  }, /*#__PURE__*/_react.default.createElement(_index.View, {
    css: {
      margin: 'auto',
      maxWidth: 1200,
      position: 'relative',
      zIndex: 1
    }
  }, /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.Heading, {
    size: 2
  }, title)), /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    mb: 8
  }, /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: [1, 2, 4]
  }, _styles.SUPPORTED_COLORS.map(function (color) {
    return /*#__PURE__*/_react.default.createElement(ColorSet, {
      color: color,
      key: color
    });
  }))), /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.Heading, {
    size: 4
  }, "RGBA")), /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: [1, 2, 4]
  }, _styles.SUPPORTED_COLORS.map(function (color) {
    return /*#__PURE__*/_react.default.createElement(ColorRGBSet, {
      color: color,
      key: color
    });
  })))), /*#__PURE__*/_react.default.createElement(_index.Surface, {
    css: {
      bottom: 0,
      left: 0,
      position: 'absolute',
      right: 0,
      top: 0,
      zIndex: 0
    }
  }));
};

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(ColorCollection, {
    title: "Colors"
  }), /*#__PURE__*/_react.default.createElement(_styles.ThemeProvider, {
    isDark: true,
    isGlobal: false
  }, /*#__PURE__*/_react.default.createElement(ColorCollection, {
    title: "Colors + Dark"
  })));
};

exports._default = _default;
