// @wordpress/block-editor/build-module/components/block-toolbar/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import { createElement, Fragment } from "@wordpress/element";

/**
 * External dependencies
 */
import classnames from 'classnames';
/**
 * WordPress dependencies
 */

import { useSelect, useDispatch } from '@wordpress/data';
import { useRef } from '@wordpress/element';
import { useViewportMatch } from '@wordpress/compose';
import { getBlockType, hasBlockSupport } from '@wordpress/blocks';
import { ToolbarGroup } from '@wordpress/components';
/**
 * Internal dependencies
 */

import BlockMover from '../block-mover';
import BlockParentSelector from '../block-parent-selector';
import BlockSwitcher from '../block-switcher';
import BlockControls from '../block-controls';
import BlockSettingsMenu from '../block-settings-menu';
import { useShowMoversGestures } from './utils';
import { store as blockEditorStore } from '../../store';
export default function BlockToolbar(_ref) {
  var hideDragHandle = _ref.hideDragHandle;

  var _useSelect = useSelect(function (select) {
    var _select = select(blockEditorStore),
        getBlockName = _select.getBlockName,
        getBlockMode = _select.getBlockMode,
        getSelectedBlockClientIds = _select.getSelectedBlockClientIds,
        isBlockValid = _select.isBlockValid,
        getBlockRootClientId = _select.getBlockRootClientId,
        getSettings = _select.getSettings;

    var selectedBlockClientIds = getSelectedBlockClientIds();
    var selectedBlockClientId = selectedBlockClientIds[0];
    var blockRootClientId = getBlockRootClientId(selectedBlockClientId);
    var settings = getSettings();
    return {
      blockClientIds: selectedBlockClientIds,
      blockClientId: selectedBlockClientId,
      blockType: selectedBlockClientId && getBlockType(getBlockName(selectedBlockClientId)),
      hasFixedToolbar: settings.hasFixedToolbar,
      hasReducedUI: settings.hasReducedUI,
      rootClientId: blockRootClientId,
      isValid: selectedBlockClientIds.every(function (id) {
        return isBlockValid(id);
      }),
      isVisual: selectedBlockClientIds.every(function (id) {
        return getBlockMode(id) === 'visual';
      })
    };
  }, []),
      blockClientIds = _useSelect.blockClientIds,
      blockClientId = _useSelect.blockClientId,
      blockType = _useSelect.blockType,
      hasFixedToolbar = _useSelect.hasFixedToolbar,
      hasReducedUI = _useSelect.hasReducedUI,
      isValid = _useSelect.isValid,
      isVisual = _useSelect.isVisual; // Handles highlighting the current block outline on hover or focus of the
  // block type toolbar area.


  var _useDispatch = useDispatch(blockEditorStore),
      toggleBlockHighlight = _useDispatch.toggleBlockHighlight;

  var nodeRef = useRef();

  var _useShowMoversGesture = useShowMoversGestures({
    ref: nodeRef,
    onChange: function onChange(isFocused) {
      if (isFocused && hasReducedUI) {
        return;
      }

      toggleBlockHighlight(blockClientId, isFocused);
    }
  }),
      showMovers = _useShowMoversGesture.showMovers,
      showMoversGestures = _useShowMoversGesture.gestures; // Account for the cases where the block toolbar is rendered within the
  // header area and not contextually to the block.


  var displayHeaderToolbar = useViewportMatch('medium', '<') || hasFixedToolbar;

  if (blockType) {
    if (!hasBlockSupport(blockType, '__experimentalToolbar', true)) {
      return null;
    }
  }

  var shouldShowMovers = displayHeaderToolbar || showMovers;

  if (blockClientIds.length === 0) {
    return null;
  }

  var shouldShowVisualToolbar = isValid && isVisual;
  var isMultiToolbar = blockClientIds.length > 1;
  var classes = classnames('block-editor-block-toolbar', shouldShowMovers && 'is-showing-movers');
  return createElement("div", {
    className: classes
  }, !isMultiToolbar && !displayHeaderToolbar && createElement(BlockParentSelector, {
    clientIds: blockClientIds
  }), createElement("div", _extends({
    ref: nodeRef
  }, showMoversGestures), (shouldShowVisualToolbar || isMultiToolbar) && createElement(ToolbarGroup, {
    className: "block-editor-block-toolbar__block-controls"
  }, createElement(BlockSwitcher, {
    clientIds: blockClientIds
  }), createElement(BlockMover, {
    clientIds: blockClientIds,
    hideDragHandle: hideDragHandle || hasReducedUI
  }))), shouldShowVisualToolbar && createElement(Fragment, null, createElement(BlockControls.Slot, {
    group: "block",
    className: "block-editor-block-toolbar__slot"
  }), createElement(BlockControls.Slot, {
    className: "block-editor-block-toolbar__slot"
  }), createElement(BlockControls.Slot, {
    group: "inline",
    className: "block-editor-block-toolbar__slot"
  }), createElement(BlockControls.Slot, {
    group: "other",
    className: "block-editor-block-toolbar__slot"
  })), createElement(BlockSettingsMenu, {
    clientIds: blockClientIds
  }));
}
