// @wordpress/components/build/ui/form-group/form-group-content.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _styles = require("@wp-g2/styles");

var _vStack = require("../v-stack");

var _formGroupContext = require("./form-group-context");

var _formGroupHelp = _interopRequireDefault(require("./form-group-help"));

var _formGroupLabel = _interopRequireDefault(require("./form-group-label"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

/**
 * @param {import('@wp-g2/create-styles').ViewOwnProps<import('./types').FormGroupContentProps, 'label'>} props
 */
function FormGroupContent(_ref) {
  var alignLabel = _ref.alignLabel,
      children = _ref.children,
      help = _ref.help,
      _ref$horizontal = _ref.horizontal,
      horizontal = _ref$horizontal === void 0 ? false : _ref$horizontal,
      id = _ref.id,
      label = _ref.label,
      labelHidden = _ref.labelHidden,
      _ref$spacing = _ref.spacing,
      spacing = _ref$spacing === void 0 ? 2 : _ref$spacing,
      truncate = _ref.truncate,
      props = (0, _objectWithoutProperties2.default)(_ref, ["alignLabel", "children", "help", "horizontal", "id", "label", "labelHidden", "spacing", "truncate"]);
  var contextProps = (0, _element.useMemo)(function () {
    return {
      id: id,
      horizontal: horizontal
    };
  }, [id, horizontal]);
  var content = help ? (0, _element.createElement)(_vStack.VStack, (0, _extends2.default)({
    expanded: false
  }, _styles.ui.$('FormGroupContentContainer'), {
    spacing: spacing
  }), children, (0, _element.createElement)(_formGroupHelp.default, null, help)) : children;
  return (0, _element.createElement)(_formGroupContext.FormGroupContext.Provider, {
    value: contextProps
  }, (0, _element.createElement)(_formGroupLabel.default, (0, _extends2.default)({
    align: alignLabel,
    id: id,
    labelHidden: labelHidden,
    truncate: truncate
  }, props), label), content);
}

var _default = (0, _element.memo)(FormGroupContent);

exports.default = _default;
