// @wordpress/components/build/ui/form-group/use-form-group.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useFormGroup = useFormGroup;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var styles = _interopRequireWildcard(require("./form-group-styles"));

var _utils = require("../utils");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * @param {import('@wp-g2/create-styles').ViewOwnProps<import('./types').FormGroupProps, 'div'>} props
 */
function useFormGroup(props) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'FormGroup'),
      _useContextSystem$ali = _useContextSystem.alignLabel,
      alignLabel = _useContextSystem$ali === void 0 ? 'left' : _useContextSystem$ali,
      children = _useContextSystem.children,
      className = _useContextSystem.className,
      help = _useContextSystem.help,
      _useContextSystem$hor = _useContextSystem.horizontal,
      horizontal = _useContextSystem$hor === void 0 ? false : _useContextSystem$hor,
      idProp = _useContextSystem.id,
      label = _useContextSystem.label,
      _useContextSystem$lab = _useContextSystem.labelHidden,
      labelHidden = _useContextSystem$lab === void 0 ? false : _useContextSystem$lab,
      _useContextSystem$tru = _useContextSystem.truncate,
      truncate = _useContextSystem$tru === void 0 ? false : _useContextSystem$tru,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["alignLabel", "children", "className", "help", "horizontal", "id", "label", "labelHidden", "truncate"]);

  var id = (0, _utils.useInstanceId)(useFormGroup, 'form-group', idProp);
  var classes = (0, _styles.cx)(styles.FormGroup, className);
  var contentProps = {
    alignLabel: alignLabel,
    children: children,
    help: help,
    id: id,
    horizontal: horizontal,
    label: label,
    labelHidden: labelHidden,
    truncate: truncate
  };
  return _objectSpread(_objectSpread({}, otherProps), {}, {
    className: classes,
    contentProps: contentProps,
    horizontal: horizontal
  });
}
