// @wp-g2/components/dist/cjs/FormGroup/FormGroupContent.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _VStack = require("../VStack");

var _FormGroup = require("./FormGroup.Context");

var _FormGroupHelp = _interopRequireDefault(require("./FormGroupHelp"));

var _FormGroupLabel = _interopRequireDefault(require("./FormGroupLabel"));

function FormGroupContent(_ref) {
  var alignLabel = _ref.alignLabel,
      children = _ref.children,
      help = _ref.help,
      horizontal = _ref.horizontal,
      id = _ref.id,
      _ref$isLabelBlock = _ref.isLabelBlock,
      isLabelBlock = _ref$isLabelBlock === void 0 ? true : _ref$isLabelBlock,
      label = _ref.label,
      labelHidden = _ref.labelHidden,
      _ref$spacing = _ref.spacing,
      spacing = _ref$spacing === void 0 ? 2 : _ref$spacing,
      truncate = _ref.truncate;

  var contextProps = _react.default.useMemo(function () {
    return {
      id: id,
      horizontal: horizontal
    };
  }, [id, horizontal]);

  var content = help ? /*#__PURE__*/_react.default.createElement(_VStack.VStack, Object.assign({
    expanded: false
  }, _styles.ui.$('FormGroupContentContainer'), {
    spacing: spacing
  }), children, /*#__PURE__*/_react.default.createElement(_FormGroupHelp.default, null, help)) : children;
  return /*#__PURE__*/_react.default.createElement(_FormGroup.FormGroupContext.Provider, {
    value: contextProps
  }, /*#__PURE__*/_react.default.createElement(_FormGroupLabel.default, {
    align: alignLabel,
    id: id,
    isLabelBlock: isLabelBlock,
    labelHidden: labelHidden,
    truncate: truncate
  }, label), content);
}

var _default = /*#__PURE__*/_react.default.memo(FormGroupContent);

exports.default = _default;
