// @wp-g2/components/dist/cjs/ArrowIndicator/ArrowIndicator.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _Icon = require("../Icon");

var _View = require("../View");

var styles = _interopRequireWildcard(require("./ArrowIndicator.styles"));

var ArrowIndicatorView = styles.ArrowIndicatorView;

function ArrowIndicator(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'ArrowIndicator'),
      className = _useContextSystem.className,
      _useContextSystem$dir = _useContextSystem.direction,
      direction = _useContextSystem$dir === void 0 ? 'right' : _useContextSystem$dir,
      height = _useContextSystem.height,
      _useContextSystem$ico = _useContextSystem.iconSize,
      iconSize = _useContextSystem$ico === void 0 ? 5 : _useContextSystem$ico,
      _useContextSystem$siz = _useContextSystem.size,
      size = _useContextSystem$siz === void 0 ? 24 : _useContextSystem$siz,
      width = _useContextSystem.width,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["className", "direction", "height", "iconSize", "size", "width"]);

  var rotations = {
    down: 90,
    left: 180,
    right: 0,
    up: 270
  };
  var rotate = rotations[direction] || rotations.right;
  var classes = (0, _styles.cx)((0, _styles.css)({
    height: height || size,
    width: width || size
  }), _styles.ui.rotate(rotate), _styles.ui.animation.default, className);
  return /*#__PURE__*/_react.default.createElement(ArrowIndicatorView, Object.assign({}, otherProps, {
    className: classes,
    ref: forwardedRef
  }), /*#__PURE__*/_react.default.createElement(_View.View, {
    animate: {
      rotate: rotate
    }
  }, /*#__PURE__*/_react.default.createElement(_Icon.Icon, {
    icon: /*#__PURE__*/_react.default.createElement("svg", {
      fill: "currentColor",
      height: "24",
      viewBox: "0 0 24 24",
      width: "24",
      xmlns: "http://www.w3.org/2000/svg"
    }, /*#__PURE__*/_react.default.createElement("path", {
      d: "M21 12l-18 12v-24z"
    })),
    size: iconSize
  })));
}

var _default = (0, _context.contextConnect)(ArrowIndicator, 'ArrowIndicator');

exports.default = _default;
