// @wp-g2/components/dist/cjs/__fixtures__/components/ContextPopover.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ContextPopover = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _index = require("../../index");

var ContextPopover = function (_ref) {
  var children = _ref.children,
      title = _ref.title,
      trigger = _ref.trigger;

  var _useState = (0, _react.useState)(0),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      offset = _useState2[0],
      setOffset = _useState2[1];

  var triggerRef = (0, _react.useRef)();
  (0, _react.useLayoutEffect)(function () {
    requestAnimationFrame(function () {
      var node = triggerRef.current;

      if (node) {
        setOffset(node.offsetLeft);
      }
    });
  }, []);
  return /*#__PURE__*/_react.default.createElement(_index.Popover, {
    maxWidth: 280,
    placement: "right-start",
    style: {
      right: offset
    },
    trigger: /*#__PURE__*/_react.default.createElement(_index.View, {
      css: {
        outline: 'none'
      },
      ref: triggerRef
    }, trigger),
    unstable_offset: [0, 12]
  }, /*#__PURE__*/_react.default.createElement(_index.CardHeader, null, /*#__PURE__*/_react.default.createElement(_index.Text, {
    weight: 600
  }, title)), /*#__PURE__*/_react.default.createElement(_index.CardBody, null, children));
};

exports.ContextPopover = ContextPopover;
