// @wp-g2/components/dist/cjs/__fixtures__/components/Sidebar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Sidebar = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _utils = require("@wp-g2/utils");

var _react = _interopRequireDefault(require("react"));

var _index = require("../../index");

var Sidebar = function (_ref) {
  var children = _ref.children;

  var _useLocalState = (0, _utils.useLocalState)('G2/Components/Sidebar/DarkenedSidebar', false),
      _useLocalState2 = (0, _slicedToArray2.default)(_useLocalState, 2),
      darkenedSidebar = _useLocalState2[0],
      setDarkenedSidebar = _useLocalState2[1];

  var _useLocalState3 = (0, _utils.useLocalState)('G2/Components/Sidebar/outlinedControls', false),
      _useLocalState4 = (0, _slicedToArray2.default)(_useLocalState3, 2),
      outlinedControls = _useLocalState4[0],
      setOutlinedControls = _useLocalState4[1];

  var _useLocalState5 = (0, _utils.useLocalState)('G2/Components/Sidebar/darkMode', false),
      _useLocalState6 = (0, _slicedToArray2.default)(_useLocalState5, 2),
      darkMode = _useLocalState6[0],
      setDarkMode = _useLocalState6[1];

  var theme = {
    controlBackgroundColor: outlinedControls && _styles.ui.get('surfaceColor'),
    controlBorderColor: outlinedControls && _styles.ui.get('surfaceBorderColor'),
    controlSurfaceBoxShadow: outlinedControls ? 'none' : null
  };
  return /*#__PURE__*/_react.default.createElement(_context.ContextSystemProvider, {
    value: {
      Grid: {
        gap: 2
      },
      Icon: {
        size: 16
      }
    }
  }, /*#__PURE__*/_react.default.createElement(_index.View, {
    css: [_styles.ui.position.fixed, {
      top: 8,
      left: 8
    }, _styles.ui.frame.width(240)]
  }, /*#__PURE__*/_react.default.createElement(_index.Card, null, /*#__PURE__*/_react.default.createElement(_index.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Darkend Sidebar",
    templateColumns: "1fr 1fr"
  }, /*#__PURE__*/_react.default.createElement(_index.Switch, {
    checked: darkenedSidebar,
    onChange: setDarkenedSidebar
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Outlined Controls",
    templateColumns: "1fr 1fr"
  }, /*#__PURE__*/_react.default.createElement(_index.Switch, {
    checked: outlinedControls,
    onChange: setOutlinedControls
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Dark Mode",
    templateColumns: "1fr 1fr"
  }, /*#__PURE__*/_react.default.createElement(_index.Switch, {
    checked: darkMode,
    onChange: setDarkMode
  }))))), /*#__PURE__*/_react.default.createElement(_styles.ThemeProvider, {
    isDark: darkMode,
    theme: theme
  }, /*#__PURE__*/_react.default.createElement(_index.Surface, {
    css: "\n\t\t\t\t\t\twidth: 280px;\n\t\t\t\t\t\tposition: absolute;\n\t\t\t\t\t\theight: 100vh;\n\t\t\t\t\t\ttop: 0;\n\t\t\t\t\t\tright: 0;\n\t\t\t\t\t\tborder-left: 1px solid " + _styles.ui.get('colorDivider') + ";\n\t\t\t\t\t",
    variant: darkenedSidebar ? 'secondary' : 'tertiary'
  }, /*#__PURE__*/_react.default.createElement(_index.Tabs, {
    selectedId: "block"
  }, /*#__PURE__*/_react.default.createElement(_index.TabList, null, /*#__PURE__*/_react.default.createElement(_index.Tab, {
    id: "document"
  }, "Document"), /*#__PURE__*/_react.default.createElement(_index.Tab, {
    id: "block"
  }, "Block")), /*#__PURE__*/_react.default.createElement(_index.TabPanel, {
    css: {
      height: "calc(100vh - (" + _styles.ui.get('controlHeightXLarge') + " + 1px))"
    }
  }, /*#__PURE__*/_react.default.createElement(_index.Scrollable, {
    css: {
      paddingBottom: '20vh'
    }
  }, /*#__PURE__*/_react.default.createElement(_index.Panel, {
    visible: true
  }, /*#__PURE__*/_react.default.createElement(_index.PanelHeader, {
    title: "Status & Visibility"
  }), /*#__PURE__*/_react.default.createElement(_index.PanelBody, null, /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Visibility"
  }, /*#__PURE__*/_react.default.createElement(_index.ControlLabel, {
    as: "div"
  }, "Public")), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Slug"
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    value: "/my-blog-post"
  })))), /*#__PURE__*/_react.default.createElement(_index.Panel, null, /*#__PURE__*/_react.default.createElement(_index.PanelHeader, {
    title: "Excerpt"
  }), /*#__PURE__*/_react.default.createElement(_index.PanelBody, null, /*#__PURE__*/_react.default.createElement(_index.ControlLabel, null, "Write an excerpt (optional)"), /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    maxRows: 6,
    minRows: 3,
    multiline: true
  }))), /*#__PURE__*/_react.default.createElement(_index.Panel, null, /*#__PURE__*/_react.default.createElement(_index.PanelHeader, {
    title: "Discussion"
  }), /*#__PURE__*/_react.default.createElement(_index.PanelBody, null, "TBD")))), /*#__PURE__*/_react.default.createElement(_index.TabPanel, {
    css: {
      height: "calc(100vh - (" + _styles.ui.get('controlHeightXLarge') + " + 1px))"
    }
  }, /*#__PURE__*/_react.default.createElement(_index.Scrollable, {
    css: {
      paddingBottom: '20vh'
    }
  }, children))))));
};

exports.Sidebar = Sidebar;
