// @wp-g2/components/dist/cjs/font-size-control/__stories__/index.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._periodicRerender = exports._default = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireDefault(require("react"));

var _Grid = require("../../Grid");

var _index = require("../index");

var _default2 = {
  component: _index.FontSizeControl,
  title: 'Components/FontSizeControl'
};
exports.default = _default2;
var fontSizes = [{
  name: 'Small',
  slug: 'small',
  size: 12
}, {
  name: 'Normallllllllllllllllllllllll Normalllllllllllllllllll',
  slug: 'normal',
  size: 16
}, {
  name: 'Big',
  slug: 'big',
  size: 26
}];

var _default = function () {
  var _React$useState = _react.default.useState(undefined),
      _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
      value = _React$useState2[0],
      setValue = _React$useState2[1];

  var handleOnChange = function (next) {
    setValue(next);
  };

  return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_Grid.Grid, {
    templateColumns: "260px 260px 1fr"
  }, /*#__PURE__*/_react.default.createElement(_index.FontSizeControl, {
    fontSizes: fontSizes,
    isPreviewable: true,
    onChange: handleOnChange,
    placeholder: "Element",
    value: value,
    withSlider: true
  }), /*#__PURE__*/_react.default.createElement(_index.FontSizeControl, {
    fontSizes: fontSizes,
    onChange: handleOnChange,
    placeholder: "Element",
    value: value
  })));
};

exports._default = _default;

var _periodicRerender = function () {
  var _React$useState3 = _react.default.useState(0),
      _React$useState4 = (0, _slicedToArray2.default)(_React$useState3, 2),
      timer = _React$useState4[0],
      setTimer = _React$useState4[1];

  var _React$useState5 = _react.default.useState(undefined),
      _React$useState6 = (0, _slicedToArray2.default)(_React$useState5, 2),
      value = _React$useState6[0],
      setValue = _React$useState6[1];

  _react.default.useEffect(function () {
    var timeout = setTimeout(function () {
      setTimer(timer + 1);
    }, 1000);
    return function () {
      return clearTimeout(timeout);
    };
  }, [timer]);

  var renderItem = _react.default.useCallback(function (_ref) {
    var name = _ref.name,
        size = _ref.size;
    return /*#__PURE__*/_react.default.createElement("div", {
      style: {
        fontSize: size
      }
    }, name);
  }, []);

  return /*#__PURE__*/_react.default.createElement("div", null, timer, /*#__PURE__*/_react.default.createElement(_Grid.Grid, {
    templateColumns: "260px 260px 1fr"
  }, /*#__PURE__*/_react.default.createElement(_index.FontSizeControl, {
    fontSizes: fontSizes,
    onChange: function handleOnChange(next) {
      console.log('handleOnChange', next, value);
      setValue(next);
    },
    renderItem: renderItem,
    value: value
  })));
};

exports._periodicRerender = _periodicRerender;
