// @wp-g2/create-styles/dist/es/__stories__/styledComponentInterpolation.stories.js
import _taggedTemplateLiteralLoose from "@babel/runtime/helpers/esm/taggedTemplateLiteralLoose";

function _templateObject() {
  var data = _taggedTemplateLiteralLoose(["\n\tcolor: #aaa;\n\t", ";\n\t", ";\n\tfont-size: 40px;\n\tfont-weight: bold;\n\tpadding: 20px;\n\t", "\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

import React from 'react';
import { createStyleSystem } from '../index';
export default {
  title: 'create-styles/StyledComponent/StringInterpolation'
};

var _createStyleSystem = createStyleSystem({
  baseStyles: {
    boxSizing: 'border-box',
    fontFamily: 'Arial'
  }
}),
    View = _createStyleSystem.View,
    styled = _createStyleSystem.styled;

var happyMixin = function (props) {
  return props.happy ? "color: green" : "color: blue";
};

var niceMixin = function (props) {
  return props.nice ? "background-color: yellow" : "background-color: #ddd";
};

var FriendlyExample = styled.div(_templateObject(), happyMixin, niceMixin, function (props) {
  return props.title && "font-size: 90px";
});

var App = function () {
  return /*#__PURE__*/React.createElement(View, null, /*#__PURE__*/React.createElement(FriendlyExample, {
    happy: false,
    nice: false,
    title: "Weee"
  }, "Hello"));
};

export var _default = function () {
  return /*#__PURE__*/React.createElement(App, null);
};
