// @wp-g2/components/dist/cjs/ClipboardButton/__stories__/ClipboardButton.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireDefault(require("react"));

var _index = require("../index");

var _default2 = {
  component: _index.ClipboardButton,
  title: 'Components/ClipboardButton'
};
exports.default = _default2;

var Example = function () {
  var _React$useState = _react.default.useState(false),
      _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
      hasCopied = _React$useState2[0],
      setHasCopied = _React$useState2[1];

  var _React$useState3 = _react.default.useState(false),
      _React$useState4 = (0, _slicedToArray2.default)(_React$useState3, 2),
      hasFinishedCopying = _React$useState4[0],
      setHasFinishedCopying = _React$useState4[1];

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_index.ClipboardButton, {
    onCopy: function onCopy() {
      return setHasCopied(true);
    },
    onFinishCopy: function onFinishCopy() {
      return setHasFinishedCopying(true);
    },
    text: "https://wordpress.org",
    variant: "primary"
  }, "Copy \"https://wordpress.org\""), /*#__PURE__*/_react.default.createElement("div", null, hasCopied && !hasFinishedCopying && 'Copying...', hasFinishedCopying && 'Success!!!!'));
};

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(Example, null);
};

exports._default = _default;
