// @wordpress/components/build/mobile/picker/index.ios.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _reactNative = require("react-native");

var _i18n = require("@wordpress/i18n");

var _data = require("@wordpress/data");

var _components = require("@wordpress/components");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var Picker = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Picker, _Component);

  var _super = _createSuper(Picker);

  function Picker() {
    (0, _classCallCheck2.default)(this, Picker);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(Picker, [{
    key: "presentPicker",
    value: function presentPicker() {
      var _this$props = this.props,
          options = _this$props.options,
          onChange = _this$props.onChange,
          title = _this$props.title,
          destructiveButtonIndex = _this$props.destructiveButtonIndex,
          disabledButtonIndices = _this$props.disabledButtonIndices,
          getAnchor = _this$props.getAnchor,
          isBottomSheetOpened = _this$props.isBottomSheetOpened,
          closeBottomSheet = _this$props.closeBottomSheet,
          onHandleClosingBottomSheet = _this$props.onHandleClosingBottomSheet;
      var labels = options.map(function (_ref) {
        var label = _ref.label;
        return label;
      });
      var fullOptions = [(0, _i18n.__)('Cancel')].concat(labels);

      _reactNative.ActionSheetIOS.showActionSheetWithOptions({
        title: title,
        options: fullOptions,
        cancelButtonIndex: 0,
        destructiveButtonIndex: destructiveButtonIndex,
        disabledButtonIndices: disabledButtonIndices,
        anchor: getAnchor && getAnchor()
      }, function (buttonIndex) {
        if (buttonIndex === 0) {
          return;
        }

        var selected = options[buttonIndex - 1];

        if (selected.requiresModal && isBottomSheetOpened) {
          onHandleClosingBottomSheet(function () {
            onChange(selected.value);
          });
          closeBottomSheet();
        } else {
          onChange(selected.value);
        }
      });
    }
  }, {
    key: "render",
    value: function render() {
      return null;
    }
  }]);
  return Picker;
}(_element.Component);

var PickerComponent = (0, _element.forwardRef)(function (props, ref) {
  var isBottomSheetOpened = (0, _data.useSelect)(function (select) {
    return select('core/edit-post').isEditorSidebarOpened();
  });

  var _useDispatch = (0, _data.useDispatch)('core/edit-post'),
      closeGeneralSidebar = _useDispatch.closeGeneralSidebar;

  var _useContext = (0, _element.useContext)(_components.BottomSheetContext),
      onHandleClosingBottomSheet = _useContext.onHandleClosingBottomSheet;

  return (0, _element.createElement)(Picker, (0, _extends2.default)({
    ref: ref
  }, props, {
    isBottomSheetOpened: isBottomSheetOpened,
    closeBottomSheet: closeGeneralSidebar,
    onHandleClosingBottomSheet: onHandleClosingBottomSheet
  }));
});
var _default = PickerComponent;
exports.default = _default;
