// @wordpress/components/build/ui/popover/component.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _utils = require("@wp-g2/utils");

var _reakit = require("reakit");

var _portal = require("../portal");

var _context2 = require("./context");

var _utils2 = require("./utils");

var _content = _interopRequireDefault(require("./content"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 *
 * @param {import('@wp-g2/create-styles').ViewOwnProps<import('./types').Props, 'div'>} props
 * @param {import('react').Ref<any>} forwardedRef
 */
function Popover(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'Popover'),
      _useContextSystem$ani = _useContextSystem.animated,
      animated = _useContextSystem$ani === void 0 ? true : _useContextSystem$ani,
      _useContextSystem$ani2 = _useContextSystem.animationDuration,
      animationDuration = _useContextSystem$ani2 === void 0 ? 160 : _useContextSystem$ani2,
      baseId = _useContextSystem.baseId,
      children = _useContextSystem.children,
      _useContextSystem$ele = _useContextSystem.elevation,
      elevation = _useContextSystem$ele === void 0 ? 5 : _useContextSystem$ele,
      id = _useContextSystem.id,
      label = _useContextSystem.label,
      _useContextSystem$max = _useContextSystem.maxWidth,
      maxWidth = _useContextSystem$max === void 0 ? 360 : _useContextSystem$max,
      _useContextSystem$onV = _useContextSystem.onVisibleChange,
      onVisibleChange = _useContextSystem$onV === void 0 ? _utils.noop : _useContextSystem$onV,
      placement = _useContextSystem.placement,
      state = _useContextSystem.state,
      trigger = _useContextSystem.trigger,
      visible = _useContextSystem.visible,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["animated", "animationDuration", "baseId", "children", "elevation", "id", "label", "maxWidth", "onVisibleChange", "placement", "state", "trigger", "visible"]);

  var _popover = (0, _reakit.usePopoverState)(_objectSpread({
    animated: animated ? animationDuration : undefined,
    baseId: baseId || id,
    placement: placement,
    visible: visible
  }, otherProps));

  var popover = state || _popover;
  var resizeListener = (0, _utils2.usePopoverResizeUpdater)({
    onResize: popover.unstable_update
  });
  var uniqueId = "popover-".concat(popover.baseId);
  var labelId = label || uniqueId;
  var contextProps = (0, _element.useMemo)(function () {
    return {
      label: labelId,
      popover: popover
    };
  }, [labelId, popover]);
  var triggerContent = (0, _element.useCallback)(function (triggerProps) {
    return (0, _element.cloneElement)(trigger, triggerProps);
  }, [trigger]);
  (0, _utils.useUpdateEffect)(function () {
    onVisibleChange(popover.visible);
  }, [popover.visible]);
  return (0, _element.createElement)(_context2.PopoverContext.Provider, {
    value: contextProps
  }, trigger && (0, _element.createElement)(_reakit.PopoverDisclosure, (0, _extends2.default)({}, popover, {
    ref: trigger.ref
  }, trigger.props), triggerContent), (0, _element.createElement)(_portal.Portal, null, (0, _element.createElement)(_content.default, (0, _extends2.default)({
    ref: forwardedRef
  }, otherProps, {
    elevation: elevation,
    maxWidth: maxWidth
  }), resizeListener, children)));
}
/**
 * `Popover` is a component to render a floating content modal.
 * It is similar in purpose to a tooltip, but renders content of any sort,
 * not only simple text.
 *
 * @example
 * ```jsx
 * import { Button, Popover, View, Text } from `@wordpress/components/ui`;
 
 * function Example() {
 * 	return (
 * 		<Popover trigger={ <Button>Popover</Button> }>
 * 			<View>
 * 				<Text>Code is Poetry</Text>
 * 			</View>
 * 		</Popover>
 * 	);
 * }
 * ```
 */


var _default = (0, _context.contextConnect)(Popover, 'Popover');

exports.default = _default;
