// @wordpress/components/build/ui/popover/content.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _reakit = require("reakit");

var _card = require("../card");

var _view = require("../view");

var _context2 = require("./context");

var styles = _interopRequireWildcard(require("./styles"));

/**
 * External dependencies
 */
// eslint-disable-next-line no-restricted-imports

/**
 * Internal dependencies
 */

/**
 *
 * @param {import('@wp-g2/create-styles').ViewOwnProps<import('./types').ContentProps, 'div'>} props
 * @param {import('react').Ref<any>} forwardedRef
 */
function PopoverContent(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'PopoverContent'),
      children = _useContextSystem.children,
      className = _useContextSystem.className,
      _useContextSystem$ele = _useContextSystem.elevation,
      elevation = _useContextSystem$ele === void 0 ? 5 : _useContextSystem$ele,
      _useContextSystem$max = _useContextSystem.maxWidth,
      maxWidth = _useContextSystem$max === void 0 ? 360 : _useContextSystem$max,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["children", "className", "elevation", "maxWidth"]);

  var _usePopoverContext = (0, _context2.usePopoverContext)(),
      label = _usePopoverContext.label,
      popover = _usePopoverContext.popover;

  var classes = (0, _styles.cx)(styles.PopoverContent, (0, _styles.css)({
    maxWidth: maxWidth
  }), className);

  if (!popover) {
    throw new Error('`PopoverContent` must only be used inside a `Popover`.');
  }

  var showContent = popover.visible || popover.animating;
  return (0, _element.createElement)(_reakit.Popover, (0, _extends2.default)({
    "aria-label": label,
    as: _view.View,
    className: classes
  }, otherProps, popover), showContent && (0, _element.createElement)(_card.Card, {
    className: styles.cardStyle,
    elevation: elevation,
    ref: forwardedRef
  }, children));
}

var _default = (0, _context.contextConnect)(PopoverContent, 'PopoverContent');

exports.default = _default;
