// @wordpress/components/build/utils/style-mixins.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "color", {
  enumerable: true,
  get: function get() {
    return _colors.color;
  }
});
Object.defineProperty(exports, "rgba", {
  enumerable: true,
  get: function get() {
    return _colors.rgba;
  }
});
Object.defineProperty(exports, "reduceMotion", {
  enumerable: true,
  get: function get() {
    return _reduceMotion.reduceMotion;
  }
});
Object.defineProperty(exports, "rtl", {
  enumerable: true,
  get: function get() {
    return _rtl.rtl;
  }
});
Object.defineProperty(exports, "space", {
  enumerable: true,
  get: function get() {
    return _space.space;
  }
});
Object.defineProperty(exports, "font", {
  enumerable: true,
  get: function get() {
    return _font.font;
  }
});
Object.defineProperty(exports, "breakpoint", {
  enumerable: true,
  get: function get() {
    return _breakpoint.breakpoint;
  }
});
Object.defineProperty(exports, "config", {
  enumerable: true,
  get: function get() {
    return _config.config;
  }
});

var _colors = require("./colors");

var _reduceMotion = require("./reduce-motion");

var _rtl = require("./rtl");

var _space = require("./space");

var _font = require("./font");

var _breakpoint = require("./breakpoint");

var _config = require("./config");
