// @wp-g2/components/dist/es/ControlGroup/__stories__/ControlGroup.stories.js
import React from 'react';
import { Button, Container, ListGroup, ListGroupHeader, ListGroups, Select, TextInput } from '../../index';
import { ControlGroup } from '../index';
export default {
  component: ControlGroup,
  title: 'Components/ControlGroup'
};
export var _default = function () {
  return /*#__PURE__*/React.createElement(Container, null, /*#__PURE__*/React.createElement(ListGroups, {
    spacing: 20
  }, /*#__PURE__*/React.createElement(ListGroup, null, /*#__PURE__*/React.createElement(ListGroupHeader, null, "Grid"), /*#__PURE__*/React.createElement(ControlGroup, {
    templateColumns: "auto 1fr 1fr auto"
  }, /*#__PURE__*/React.createElement(Select, {
    options: [{
      label: 'Ms.',
      value: 'ms'
    }]
  }), /*#__PURE__*/React.createElement(TextInput, {
    placeholder: "First name"
  }), /*#__PURE__*/React.createElement(TextInput, {
    placeholder: "Last name"
  }), /*#__PURE__*/React.createElement(Button, {
    variant: "primary"
  }, "Submit"))), /*#__PURE__*/React.createElement(ListGroup, null, /*#__PURE__*/React.createElement(ListGroupHeader, null, "Flex"), /*#__PURE__*/React.createElement(ControlGroup, null, /*#__PURE__*/React.createElement(Select, {
    options: [{
      label: 'Ms.',
      value: 'ms'
    }]
  }), /*#__PURE__*/React.createElement(TextInput, {
    placeholder: "First name"
  }), /*#__PURE__*/React.createElement(TextInput, {
    placeholder: "Last name"
  }), /*#__PURE__*/React.createElement(Button, {
    variant: "primary"
  }, "Submit")))));
};
