// @wp-g2/components/dist/cjs/Avatar/__stories__/Avatar.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _protokit = require("@wp-g2/protokit");

var _react = _interopRequireDefault(require("react"));

var _index = require("../../index");

var _Avatar = require("../Avatar.utils");

var _index2 = require("../index");

var _default2 = {
  component: _index2.Avatar,
  title: 'Components/Avatar'
}; // function StatusDot({ color = 'green' }) {
// 	const styles = {
// 		border: '2px solid white',
// 		zIndex: 1,
// 	};
// 	return (
// 		<View
// 			css={[
// 				{ background: color },
// 				ui.offset({ x: 4, y: -4 }),
// 				ui.borderRadius.circle,
// 				ui.frame({ height: 10, width: 10 }),
// 				ui.position.topRight,
// 				styles,
// 			]}
// 		/>
// 	);
// }

exports.default = _default2;

function AvatarGroup(_ref) {
  var children = _ref.children;
  return /*#__PURE__*/_react.default.createElement(_index.ZStack, {
    isLayered: false,
    isReversed: true,
    offset: 8
  }, children);
}

var _default = function () {
  var avatars = Object.keys(_Avatar.AVATAR_SIZES);
  return /*#__PURE__*/_react.default.createElement(_index.VStack, {
    spacing: 8
  }, /*#__PURE__*/_react.default.createElement(_index.HStack, {
    alignment: "left"
  }, avatars.map(function (size, index) {
    var randomName = _protokit.faker.name.firstName() + " " + _protokit.faker.name.lastName();

    var avatarImage = _protokit.faker.image.avatar();

    return /*#__PURE__*/_react.default.createElement(_index.View, {
      key: index
    }, /*#__PURE__*/_react.default.createElement(_index2.Avatar, {
      color: _protokit.faker.random.arrayElement(['var(--wp-g2-green-500)', 'var(--wp-g2-red-500)', 'var(--wp-g2-orange-500)', undefined]),
      key: size,
      name: randomName,
      size: size,
      src: _protokit.faker.random.arrayElement([avatarImage, undefined]),
      status: _protokit.faker.random.arrayElement(['green', 'red', 'orange', undefined])
    }));
  })), /*#__PURE__*/_react.default.createElement(_index.HStack, {
    alignment: "left"
  }, avatars.map(function (size, index) {
    var randomName = "" + _protokit.faker.name.firstName();

    var avatarImage = _protokit.faker.image.avatar();

    return /*#__PURE__*/_react.default.createElement(_index.View, {
      key: index
    }, /*#__PURE__*/_react.default.createElement(_index2.Avatar, {
      color: _protokit.faker.random.arrayElement(['green', 'red', 'orange', undefined]),
      key: size,
      name: randomName,
      size: size,
      src: _protokit.faker.random.arrayElement([avatarImage, undefined]),
      status: _protokit.faker.random.arrayElement(['green', 'red', 'orange', undefined]),
      variant: "square"
    }));
  })), /*#__PURE__*/_react.default.createElement(AvatarGroup, null, avatars.map(function (size, index) {
    var randomName = "" + _protokit.faker.name.firstName();

    var avatarImage = _protokit.faker.image.avatar();

    return /*#__PURE__*/_react.default.createElement(_index.View, {
      key: index
    }, /*#__PURE__*/_react.default.createElement(_index2.Avatar, {
      key: size,
      name: randomName,
      src: avatarImage,
      variant: "circle"
    }));
  })));
};

exports._default = _default;
