// @wp-g2/components/dist/cjs/ColorPicker/ColorPicker.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _VStack = require("../VStack");

var _ColorPicker = require("./ColorPicker.Context");

var styles = _interopRequireWildcard(require("./ColorPicker.styles"));

var _ColorPickerElement = require("./ColorPickerElement");

var _ColorPickerInputs = require("./ColorPickerInputs");

var _ColorPickerSelect = require("./ColorPickerSelect");

var _useColorPicker2 = require("./useColorPicker");

var ColorPickerView = styles.ColorPickerView;
/**
 * @typedef {import('./useColorPicker').ColorPickerProps} Props
 */

/**
 * @param {import('@wp-g2/create-styles').ViewOwnProps<Props, 'div'>} props
 * @param {import('react').Ref<any>} forwardedRef
 */

function ColorPicker(props, forwardedRef) {
  var _useColorPicker = (0, _useColorPicker2.useColorPicker)(props),
      className = _useColorPicker.className,
      store = _useColorPicker.store,
      width = _useColorPicker.width,
      otherProps = (0, _objectWithoutProperties2.default)(_useColorPicker, ["className", "store", "width"]);

  var disableAlpha = store.disableAlpha;
  var classes = (0, _styles.cx)(disableAlpha && styles.disableAlpha, className);
  var vStackClasses = (0, _styles.cx)((0, _styles.css)({
    width: width
  }));
  return /*#__PURE__*/_react.default.createElement(_ColorPicker.ColorPickerContext.Provider, {
    value: store
  }, /*#__PURE__*/_react.default.createElement(_VStack.VStack, {
    alignment: "top",
    className: vStackClasses,
    isExpanded: false
  }, /*#__PURE__*/_react.default.createElement(ColorPickerView, Object.assign({}, otherProps, {
    className: classes,
    ref: forwardedRef
  }), /*#__PURE__*/_react.default.createElement(_ColorPickerElement.ColorPickerElement, {
    width: width
  })), /*#__PURE__*/_react.default.createElement(_VStack.VStack, {
    alignment: "top",
    isExpanded: false
  }, /*#__PURE__*/_react.default.createElement(_ColorPickerSelect.ColorPickerSelect, null), /*#__PURE__*/_react.default.createElement(_ColorPickerInputs.ColorPickerInputs, null))));
}

var _default = (0, _context.contextConnect)(ColorPicker, 'ColorPicker');

exports.default = _default;
