// @wp-g2/components/dist/cjs/ColorPicker/ColorPickerPreview.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ColorPickerPreview = void 0;

var _i18n = require("@wordpress/i18n");

var _utils = require("@wp-g2/utils");

var _react = _interopRequireDefault(require("react"));

var _ColorCircle = require("../ColorCircle");

var _Tooltip = require("../Tooltip");

var _VStack = require("../VStack");

var _ColorPicker = require("./ColorPicker.Context");

var styles = _interopRequireWildcard(require("./ColorPicker.styles"));

var ColorPickerPreview = /*#__PURE__*/_react.default.memo(function () {
  var _useColorPickerContex = (0, _ColorPicker.useColorPickerContext)(),
      color = _useColorPickerContex.colorValue,
      showPreview = _useColorPickerContex.showPreview;

  var _useClipboard = (0, _utils.useClipboard)(color),
      hasCopied = _useClipboard.hasCopied,
      onCopy = _useClipboard.onCopy;

  if (!showPreview) return null;
  var tooltipContent = hasCopied ? (0, _i18n.__)('Copied') : (0, _i18n.__)('Copy');
  return /*#__PURE__*/_react.default.createElement(_VStack.VStack, {
    alignment: "center"
  }, /*#__PURE__*/_react.default.createElement(_Tooltip.Tooltip, {
    content: tooltipContent
  }, /*#__PURE__*/_react.default.createElement(_ColorCircle.ColorCircle, {
    className: styles.ColorPreview,
    color: color,
    onClick: onCopy
  })));
});

exports.ColorPickerPreview = ColorPickerPreview;
