// @wp-g2/components/dist/cjs/ColorPicker/ColorPickerElement.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ColorPickerElement = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactColorful = require("react-colorful");

var _ColorPicker = require("./ColorPicker.Context");

var ColorPickerElement = /*#__PURE__*/_react.default.memo(function (_ref) {
  var width = _ref.width;

  var _useColorPickerContex = (0, _ColorPicker.useColorPickerContext)(),
      colorRgb = _useColorPickerContex.colorRgb,
      onChange = _useColorPickerContex.onChange;
  /**
   * Focuses the interactive control from the ColorPicker library.
   * This allows for a smoother mouse -> keyboard update flow.
   */


  return /*#__PURE__*/_react.default.createElement("div", {
    onMouseDown: function handleOnFocusInteractiveControl(event) {
      return event.target.focus();
    }
  }, /*#__PURE__*/_react.default.createElement(_reactColorful.RgbaColorPicker, {
    color: colorRgb,
    onChange: onChange,
    width: width
  }));
});

exports.ColorPickerElement = ColorPickerElement;
