// @wp-g2/components/dist/cjs/ComponentDesignTool/ComponentDesignTool.styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Resizer = exports.SizeLabel = exports.ParentSizeLabel = exports.LabelR = exports.LabelL = exports.LabelB = exports.LabelT = exports.Label = exports.BaseLabel = exports.RulerB = exports.RulerT = exports.RulerR = exports.RulerL = exports.RulerY = exports.RulerX = exports.ParentElementOutline = exports.ElementOutline = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _styles = require("@wp-g2/styles");

function _templateObject17() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tcursor: pointer;\n\toutline: 1px solid #7b62ff;\n\tpointer-events: none;\n\tposition: fixed;\n\tz-index: 9999999;\n"]);

  _templateObject17 = function () {
    return data;
  };

  return data;
}

function _templateObject16() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\tbackground: orangered;\n\tleft: 0;\n\ttop: 0;\n\ttransform: translate(0, -120%);\n\twhite-space: nowrap;\n\tz-index: 2;\n"]);

  _templateObject16 = function () {
    return data;
  };

  return data;
}

function _templateObject15() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\tbackground: #7b62ff;\n\tbottom: 0;\n\tright: 0;\n\ttransform: translateY(120%);\n\twhite-space: nowrap;\n"]);

  _templateObject15 = function () {
    return data;
  };

  return data;
}

function _templateObject14() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\tright: 0;\n\ttop: 50%;\n\ttransform: translate(120%, -50%);\n"]);

  _templateObject14 = function () {
    return data;
  };

  return data;
}

function _templateObject13() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\tleft: 0;\n\ttop: 50%;\n\ttransform: translate(-120%, -50%);\n"]);

  _templateObject13 = function () {
    return data;
  };

  return data;
}

function _templateObject12() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\tbottom: 0;\n\tleft: 50%;\n\ttransform: translate(-50%, 120%);\n"]);

  _templateObject12 = function () {
    return data;
  };

  return data;
}

function _templateObject11() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\tleft: 50%;\n\ttop: 0;\n\ttransform: translate(-50%, -120%);\n"]);

  _templateObject11 = function () {
    return data;
  };

  return data;
}

function _templateObject10() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\tbackground: orangered;\n"]);

  _templateObject10 = function () {
    return data;
  };

  return data;
}

function _templateObject9() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbackground: orangered;\n\tborder-radius: 2px;\n\tcolor: white;\n\tdisplay: block;\n\tfont-size: 7px;\n\tfont-weight: bold;\n\tline-height: 1;\n\tmargin: 0;\n\tpadding: 1px 2px;\n\tposition: absolute;\n"]);

  _templateObject9 = function () {
    return data;
  };

  return data;
}

function _templateObject8() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n"]);

  _templateObject8 = function () {
    return data;
  };

  return data;
}

function _templateObject7() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n"]);

  _templateObject7 = function () {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n"]);

  _templateObject6 = function () {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n"]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tborder-left: 1px solid orangered;\n\tleft: 50%;\n\topacity: 0.5;\n\tposition: absolute;\n\twidth: 0;\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tborder-top: 1px solid orangered;\n\theight: 0;\n\topacity: 0.5;\n\tposition: absolute;\n\ttop: 50%;\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", "\n\toutline: 1px solid #7b62ff;\n\tz-index: 999999;\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tdisplay: none;\n\toutline: 1px solid orangered;\n\tpointer-events: none;\n\tposition: fixed;\n\tz-index: 9999999;\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var ElementOutline = (0, _styles.css)(_templateObject());
exports.ElementOutline = ElementOutline;
var ParentElementOutline = (0, _styles.css)(_templateObject2(), ElementOutline);
exports.ParentElementOutline = ParentElementOutline;
var RulerX = (0, _styles.css)(_templateObject3());
exports.RulerX = RulerX;
var RulerY = (0, _styles.css)(_templateObject4());
exports.RulerY = RulerY;
var RulerL = (0, _styles.css)(_templateObject5(), RulerX);
exports.RulerL = RulerL;
var RulerR = (0, _styles.css)(_templateObject6(), RulerX);
exports.RulerR = RulerR;
var RulerT = (0, _styles.css)(_templateObject7(), RulerY);
exports.RulerT = RulerT;
var RulerB = (0, _styles.css)(_templateObject8(), RulerY);
exports.RulerB = RulerB;
var BaseLabel = (0, _styles.css)(_templateObject9());
exports.BaseLabel = BaseLabel;
var Label = (0, _styles.css)(_templateObject10(), BaseLabel);
exports.Label = Label;
var LabelT = (0, _styles.css)(_templateObject11(), Label);
exports.LabelT = LabelT;
var LabelB = (0, _styles.css)(_templateObject12(), Label);
exports.LabelB = LabelB;
var LabelL = (0, _styles.css)(_templateObject13(), Label);
exports.LabelL = LabelL;
var LabelR = (0, _styles.css)(_templateObject14(), Label);
exports.LabelR = LabelR;
var ParentSizeLabel = (0, _styles.css)(_templateObject15(), BaseLabel);
exports.ParentSizeLabel = ParentSizeLabel;
var SizeLabel = (0, _styles.css)(_templateObject16(), BaseLabel);
exports.SizeLabel = SizeLabel;
var Resizer = (0, _styles.css)(_templateObject17());
exports.Resizer = Resizer;
