// @wp-g2/components/dist/cjs/ListGroup/ListGroupContent.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _react = _interopRequireDefault(require("react"));

var _Card = require("../Card");

var _VStack = require("../VStack");

var _ListGroup = require("./ListGroup.Context");

function ListGroupContent(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'ListGroupContent'),
      children = _useContextSystem.children,
      spacing = _useContextSystem.spacing,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["children", "spacing"]);

  var _useListGroupContext = (0, _ListGroup.useListGroupContext)(),
      inset = _useListGroupContext.inset;

  if (!inset) return children;
  return /*#__PURE__*/_react.default.createElement(_Card.Card, Object.assign({
    elevation: 0,
    isBorderless: true
  }, otherProps, {
    ref: forwardedRef
  }), /*#__PURE__*/_react.default.createElement(_Card.CardBody, {
    scrollable: false
  }, /*#__PURE__*/_react.default.createElement(_VStack.VStack, {
    spacing: spacing
  }, children)));
}

var _default = (0, _context.contextConnect)(ListGroupContent, 'ListGroupContent');

exports.default = _default;
