// @wp-g2/components/dist/cjs/ListGroup/ListGroupHeader.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _Flex = require("../Flex");

var _HStack = require("../HStack");

var _Subheading = require("../Subheading");

var _ListGroup = require("./ListGroup.Context");

var styles = _interopRequireWildcard(require("./ListGroup.styles"));

var listGroupTextContextProps = {
  Text: {
    size: 'subheading'
  }
};

function ListGroupHeader(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'ListGroupHeader'),
      children = _useContextSystem.children,
      className = _useContextSystem.className,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["children", "className"]);

  var _useListGroupContext = (0, _ListGroup.useListGroupContext)(),
      inset = _useListGroupContext.inset;

  var validChildren = _react.default.Children.toArray(children);

  var clonedChildren = validChildren.map(function (child, index) {
    var _key = child.key || index;

    return typeof child === 'string' ? /*#__PURE__*/_react.default.createElement(_Flex.FlexItem, {
      isBlock: true,
      key: _key
    }, /*#__PURE__*/_react.default.createElement(_Subheading.Subheading, null, child)) : child;
  });
  var classes = (0, _styles.cx)(inset && styles.headerFooterInset, className);
  return /*#__PURE__*/_react.default.createElement(_HStack.HStack, Object.assign({}, otherProps, {
    className: classes,
    ref: forwardedRef
  }), /*#__PURE__*/_react.default.createElement(_context.ContextSystemProvider, {
    value: listGroupTextContextProps
  }, clonedChildren));
}

var _default = (0, _context.contextConnect)(ListGroupHeader, 'ListGroupHeader');

exports.default = _default;
