// @wordpress/components/build-module/higher-order/with-notices/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement } from "@wordpress/element";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * External dependencies
 */
import { v4 as uuid } from 'uuid';
/**
 * WordPress dependencies
 */

import { forwardRef, useState, useMemo } from '@wordpress/element';
import { createHigherOrderComponent } from '@wordpress/compose';
/**
 * Internal dependencies
 */

import NoticeList from '../../notice/list';
/**
 * Override the default edit UI to include notices if supported.
 *
 * @param  {WPComponent} OriginalComponent Original component.
 *
 * @return {WPComponent} Wrapped component.
 */

export default createHigherOrderComponent(function (OriginalComponent) {
  function Component(props, ref) {
    var _useState = useState([]),
        _useState2 = _slicedToArray(_useState, 2),
        noticeList = _useState2[0],
        setNoticeList = _useState2[1];

    var noticeOperations = useMemo(function () {
      /**
       * Function passed down as a prop that adds a new notice.
       *
       * @param {Object} notice  Notice to add.
       */
      var createNotice = function createNotice(notice) {
        var noticeToAdd = notice.id ? notice : _objectSpread(_objectSpread({}, notice), {}, {
          id: uuid()
        });
        setNoticeList(function (current) {
          return [].concat(_toConsumableArray(current), [noticeToAdd]);
        });
      };

      return {
        createNotice: createNotice,

        /**
         * Function passed as a prop that adds a new error notice.
         *
         * @param {string} msg  Error message of the notice.
         */
        createErrorNotice: function createErrorNotice(msg) {
          createNotice({
            status: 'error',
            content: msg
          });
        },

        /**
         * Removes a notice by id.
         *
         * @param {string} id  Id of the notice to remove.
         */
        removeNotice: function removeNotice(id) {
          setNoticeList(function (current) {
            return current.filter(function (notice) {
              return notice.id !== id;
            });
          });
        },

        /**
         * Removes all notices
         */
        removeAllNotices: function removeAllNotices() {
          setNoticeList([]);
        }
      };
    }, []);

    var propsOut = _objectSpread(_objectSpread({}, props), {}, {
      noticeList: noticeList,
      noticeOperations: noticeOperations,
      noticeUI: noticeList.length > 0 && createElement(NoticeList, {
        className: "components-with-notices-ui",
        notices: noticeList,
        onRemove: noticeOperations.removeNotice
      })
    });

    return isForwardRef ? createElement(OriginalComponent, _extends({}, propsOut, {
      ref: ref
    })) : createElement(OriginalComponent, propsOut);
  }

  var isForwardRef;
  var render = OriginalComponent.render; // Returns a forwardRef if OriginalComponent appears to be a forwardRef

  if (typeof render === 'function') {
    isForwardRef = true;
    return forwardRef(Component);
  }

  return Component;
});
