// @wordpress/components/build-module/ui/text/styles.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/esm/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7;

/**
 * External dependencies
 */
import { css, ui } from '@wp-g2/styles';
export var Text = css(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n\tcolor: ", ";\n\tline-height: ", ";\n"])), ui.color.text, ui.get('fontLineHeightBase'));
export var block = css(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n\tdisplay: block;\n"])));
export var positive = css(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n\tcolor: ", ";\n"])), ui.color.positive);
export var destructive = css(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n\tcolor: ", ";\n"])), ui.color.destructive);
export var muted = css(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["\n\tcolor: ", ";\n"])), ui.get('colorTextMuted'));
export var highlighterText = css(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["\n\tmark {\n\t\tbackground: ", ";\n\t\tborder-radius: 2px;\n\t\tbox-shadow: 0 0 0 1px rgba( 0, 0, 0, 0.05 ) inset,\n\t\t\t0 -1px 0 rgba( 0, 0, 0, 0.1 ) inset;\n\t}\n"])), ui.get('yellowRgba70'));
export var upperCase = css(_templateObject7 || (_templateObject7 = _taggedTemplateLiteral(["\n\ttext-transform: uppercase;\n"])));
