// @wordpress/components/build-module/ui/surface/styles.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/esm/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7;

/**
 * External dependencies
 */
import { css, ui } from '@wp-g2/styles';
export var Surface = css(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n\tbackground-color: ", ";\n\tcolor: ", ";\n\tposition: relative;\n"])), ui.get('surfaceColor'), ui.color.text);
export var background = css(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n\tbackground-color: ", ";\n"])), ui.get('surfaceBackgroundColor'));
/**
 * @param {Object} props
 * @param {boolean} [props.border]
 * @param {boolean} [props.borderBottom]
 * @param {boolean} [props.borderLeft]
 * @param {boolean} [props.borderRight]
 * @param {boolean} [props.borderTop]
 */

export function getBorders(_ref) {
  var border = _ref.border,
      borderBottom = _ref.borderBottom,
      borderLeft = _ref.borderLeft,
      borderRight = _ref.borderRight,
      borderTop = _ref.borderTop;
  var borderStyle = "1px solid ".concat(ui.get('surfaceBorderColor'));
  return css({
    border: border ? borderStyle : undefined,
    borderBottom: borderBottom ? borderStyle : undefined,
    borderLeft: borderLeft ? borderStyle : undefined,
    borderRight: borderRight ? borderStyle : undefined,
    borderTop: borderTop ? borderStyle : undefined
  });
}
export var primary = css(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral([""])));
export var secondary = css(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n\tbackground: ", ";\n"])), ui.get('surfaceBackgroundTintColor'));
export var tertiary = css(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["\n\tbackground: ", ";\n"])), ui.get('surfaceBackgroundTertiaryColor'));
var customBackgroundSize = [ui.get('surfaceBackgroundSize'), ui.get('surfaceBackgroundSize')].join(' ');
var dottedBackground1 = ['90deg', [ui.get('surfaceBackgroundColor'), ui.get('surfaceBackgroundSizeDotted')].join(' '), 'transparent 1%'].join(',');
var dottedBackground2 = [[ui.get('surfaceBackgroundColor'), ui.get('surfaceBackgroundSizeDotted')].join(' '), 'transparent 1%'].join(',');
var dottedBackgroundCombined = ["linear-gradient( ".concat(dottedBackground1, " ) center"), "linear-gradient( ".concat(dottedBackground2, " ) center"), ui.get('surfaceBorderBoldColor')].join(',');
export var dotted = css(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["\n\tbackground: ", ";\n\tbackground-size: ", ";\n"])), dottedBackgroundCombined, customBackgroundSize);
var gridBackground1 = ["".concat(ui.get('surfaceBorderSubtleColor'), " 1px"), 'transparent 1px'].join(',');
var gridBackground2 = ['90deg', "".concat(ui.get('surfaceBorderSubtleColor'), " 1px"), 'transparent 1px'].join(',');
var gridBackgroundCombined = ["linear-gradient( ".concat(gridBackground1, " )"), "linear-gradient( ".concat(gridBackground2, " )")].join(',');
export var grid = css(_templateObject7 || (_templateObject7 = _taggedTemplateLiteral(["\n\tbackground: ", ";\n\tbackground-image: ", ";\n\tbackground-size: ", ";\n"])), ui.get('surfaceBackgroundColor'), gridBackgroundCombined, customBackgroundSize);
