// @wp-g2/components/dist/cjs/Panel/Panel.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _utils = require("@wp-g2/utils");

var _react = _interopRequireWildcard(require("react"));

var _Accordion = require("../Accordion");

var _Collapsible = require("../Collapsible");

var _Panel = require("./Panel.Context");

var styles = _interopRequireWildcard(require("./Panel.styles"));

function Panel(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'Panel'),
      baseId = _useContextSystem.baseId,
      className = _useContextSystem.className,
      children = _useContextSystem.children,
      _useContextSystem$col = _useContextSystem.collapsible,
      collapsible = _useContextSystem$col === void 0 ? true : _useContextSystem$col,
      idProp = _useContextSystem.id,
      _useContextSystem$isB = _useContextSystem.isBorderless,
      isBorderless = _useContextSystem$isB === void 0 ? false : _useContextSystem$isB,
      _useContextSystem$sea = _useContextSystem.seamless,
      seamless = _useContextSystem$sea === void 0 ? false : _useContextSystem$sea,
      _useContextSystem$onV = _useContextSystem.onVisibleChange,
      onVisibleChange = _useContextSystem$onV === void 0 ? _utils.noop : _useContextSystem$onV,
      _useContextSystem$vis = _useContextSystem.visible,
      visibleProp = _useContextSystem$vis === void 0 ? false : _useContextSystem$vis,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["baseId", "className", "children", "collapsible", "id", "isBorderless", "seamless", "onVisibleChange", "visible"]);

  var id = (0, _utils.useUniqueId)(Panel, 'Panel', baseId || idProp);
  var classes = (0, _styles.cx)(styles.Panel, isBorderless && styles.borderless, className);

  var _useAccordion = (0, _Accordion.useAccordion)({
    id: id,
    visible: visibleProp
  }),
      _useAccordion2 = (0, _slicedToArray2.default)(_useAccordion, 2),
      visible = _useAccordion2[0],
      setVisible = _useAccordion2[1];

  var handleOnVisibleChange = (0, _react.useCallback)(function (next, disclosure) {
    setVisible(next);
    onVisibleChange(next, disclosure);
  }, [onVisibleChange, setVisible]);

  var contextProps = _react.default.useMemo(function () {
    return {
      collapsible: collapsible,
      seamless: seamless
    };
  }, [collapsible, seamless]);

  return /*#__PURE__*/_react.default.createElement(_Panel.PanelContext.Provider, {
    value: contextProps
  }, /*#__PURE__*/_react.default.createElement(_Collapsible.Collapsible, Object.assign({
    baseId: id,
    visible: visible
  }, otherProps, {
    className: classes,
    onVisibleChange: handleOnVisibleChange,
    ref: forwardedRef
  }), children));
}

var _default = (0, _context.contextConnect)(Panel, 'Panel');

exports.default = _default;
