// @wp-g2/components/dist/cjs/Panel/Panel.styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.nonCollapsibleBody = exports.PanelBodyView = exports.seamless = exports.borderless = exports.nonCollapsibleHeader = exports.collapsibleHeader = exports.PanelHeader = exports.Panel = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _styles = require("@wp-g2/styles");

function _templateObject8() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tpadding-bottom: ", ";\n"]);

  _templateObject8 = function () {
    return data;
  };

  return data;
}

function _templateObject7() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tpadding: ", ";\n"]);

  _templateObject7 = function () {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tpadding-left: 0;\n\tpadding-right: 0;\n"]);

  _templateObject6 = function () {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t& + & {\n\t\tborder-top: none;\n\t}\n"]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tpadding-bottom: 0;\n\tpadding-top: ", ";\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tcursor: pointer;\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\toutline: none;\n\tpadding: ", ";\n\tposition: relative;\n\n\t&:focus {\n\t\t", ";\n\t\tbox-shadow: 0 0 0 1px ", " inset;\n\t}\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t& + & {\n\t\tborder-top: 1px solid ", ";\n\t}\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var Panel = (0, _styles.css)(_templateObject(), _styles.ui.get('colorDivider'));
exports.Panel = Panel;
var PanelHeader = (0, _styles.css)(_templateObject2(), _styles.ui.get('panelHeaderPadding'), _styles.ui.zIndex('ControlFocus'), _styles.ui.color.admin);
exports.PanelHeader = PanelHeader;
var collapsibleHeader = (0, _styles.css)(_templateObject3());
exports.collapsibleHeader = collapsibleHeader;
var nonCollapsibleHeader = (0, _styles.css)(_templateObject4(), _styles.ui.space(3));
exports.nonCollapsibleHeader = nonCollapsibleHeader;
var borderless = (0, _styles.css)(_templateObject5());
exports.borderless = borderless;
var seamless = (0, _styles.css)(_templateObject6());
exports.seamless = seamless;

var PanelBodyView = _styles.styled.div(_templateObject7(), _styles.ui.get('panelBodyPadding'));

exports.PanelBodyView = PanelBodyView;
var nonCollapsibleBody = (0, _styles.css)(_templateObject8(), _styles.ui.space(5));
exports.nonCollapsibleBody = nonCollapsibleBody;
