// @wp-g2/components/dist/cjs/Avatar/Avatar.utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getSize = getSize;
exports.getBorderRadius = getBorderRadius;
exports.getInitialsTextSize = getInitialsTextSize;
exports.AVATAR_SIZES = exports.AVATAR_SHAPES = void 0;
var AVATAR_SHAPES = {
  circle: 999999,
  square: 8
};
exports.AVATAR_SHAPES = AVATAR_SHAPES;
var AVATAR_SIZES = {
  xLarge: 48,
  large: 36,
  medium: 30,
  small: 24,
  xSmall: 16
};
exports.AVATAR_SIZES = AVATAR_SIZES;

function getSize(size) {
  var defaultSize = AVATAR_SIZES.medium;

  if (AVATAR_SIZES[size]) {
    return AVATAR_SIZES[size];
  }

  return typeof size === 'number' ? size : defaultSize;
}

function getBorderRadius(shape, size) {
  if (size === void 0) {
    size = AVATAR_SIZES.medium;
  }

  var supported = AVATAR_SHAPES[shape];

  if (shape === 'square' && size < AVATAR_SIZES.small + 1) {
    return AVATAR_SHAPES.square / 2;
  }

  return supported || shape;
}

function getInitialsTextSize(size) {
  if (size === void 0) {
    size = AVATAR_SIZES.medium;
  }

  var textSize = size < AVATAR_SIZES.medium ? 11 : undefined;

  if (size < AVATAR_SIZES.medium + 1) {
    textSize = 12;
  }

  if (size < AVATAR_SIZES.small + 1) {
    textSize = 10;
  }

  if (size < AVATAR_SIZES.xSmall + 1) {
    textSize = 8;
  }

  return textSize;
}
