// @wp-g2/components/dist/cjs/Avatar/Avatar.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireWildcard(require("react"));

var _Image = require("../Image");

var _Initials = require("../Initials");

var _View = require("../View");

var styles = _interopRequireWildcard(require("./Avatar.styles"));

var _Avatar2 = require("./Avatar.utils");

function Avatar(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'Avatar'),
      _useContextSystem$ani = _useContextSystem.animateOnImageLoad,
      animateOnImageLoad = _useContextSystem$ani === void 0 ? true : _useContextSystem$ani,
      _useContextSystem$bor = _useContextSystem.border,
      border = _useContextSystem$bor === void 0 ? false : _useContextSystem$bor,
      _useContextSystem$col = _useContextSystem.color,
      color = _useContextSystem$col === void 0 ? (0, _styles.get)('lightGray500') : _useContextSystem$col,
      name = _useContextSystem.name,
      _useContextSystem$sha = _useContextSystem.shape,
      shape = _useContextSystem$sha === void 0 ? 'circle' : _useContextSystem$sha,
      _useContextSystem$siz = _useContextSystem.size,
      sizeProp = _useContextSystem$siz === void 0 ? 'medium' : _useContextSystem$siz,
      src = _useContextSystem.src,
      className = _useContextSystem.className,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["animateOnImageLoad", "border", "color", "name", "shape", "size", "src", "className"]);

  var _useState = (0, _react.useState)(!animateOnImageLoad),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      imageLoaded = _useState2[0],
      setImageLoaded = _useState2[1];

  var size = (0, _Avatar2.getSize)(sizeProp);
  var borderRadius = (0, _Avatar2.getBorderRadius)(shape, size);
  var textSize = (0, _Avatar2.getInitialsTextSize)(size);
  var sx = {};
  sx.base = (0, _styles.css)({
    background: color,
    width: size,
    height: size
  });
  sx.borderRadius = (0, _styles.css)({
    borderRadius: borderRadius
  });
  var classes = (0, _styles.cx)(sx.base, sx.borderRadius, border && styles.border, className);
  return /*#__PURE__*/_react.default.createElement(styles.AvatarView, Object.assign({}, otherProps, {
    className: classes,
    ref: forwardedRef
  }), !!name && /*#__PURE__*/_react.default.createElement(_View.View, {
    css: (0, _styles.css)([_styles.ui.position.absolute, _styles.ui.alignment.center])
  }, /*#__PURE__*/_react.default.createElement(_Initials.Initials, Object.assign({}, _styles.ui.$('AvatarInitials'), {
    align: "center",
    as: "div",
    lineHeight: 1,
    name: name,
    optimizeReadabilityFor: color,
    size: textSize
  }))), src && /*#__PURE__*/_react.default.createElement(_Image.Image, Object.assign({}, _styles.ui.$('AvatarImage'), {
    alt: name,
    aspectRatio: 1,
    css: (0, _styles.css)([{
      borderRadius: borderRadius,
      // Prevents image clipping from revealing background colour
      transform: 'scale(1.025)'
    }, _styles.ui.opacity(imageLoaded ? 1 : 0), _styles.ui.animation.default]),
    onLoad: function handleOnImageLoad() {
      return setImageLoaded(true);
    },
    src: src
  })));
}

var _default = (0, _context.contextConnect)(Avatar, 'Avatar');

exports.default = _default;
