// @wordpress/block-editor/build/hooks/border-radius.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BorderRadiusEdit = BorderRadiusEdit;
exports.hasBorderRadiusSupport = hasBorderRadiusSupport;
exports.useIsBorderRadiusDisabled = useIsBorderRadiusDisabled;

var _element = require("@wordpress/element");

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _blocks = require("@wordpress/blocks");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _useEditorFeature = _interopRequireDefault(require("../components/use-editor-feature"));

var _border = require("./border");

var _utils = require("./utils");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var MIN_BORDER_RADIUS_VALUE = 0;
var MAX_BORDER_RADIUS_VALUE = 50;
/**
 * Inspector control panel containing the border radius related configuration.
 *
 * @param  {Object} props Block properties.
 * @return {WPElement}    Border radius edit element.
 */

function BorderRadiusEdit(props) {
  var _style$border;

  var style = props.attributes.style,
      setAttributes = props.setAttributes;

  if (useIsBorderRadiusDisabled(props)) {
    return null;
  }

  var onChange = function onChange(newRadius) {
    var newStyle = _objectSpread(_objectSpread({}, style), {}, {
      border: _objectSpread(_objectSpread({}, style === null || style === void 0 ? void 0 : style.border), {}, {
        radius: newRadius
      })
    });

    if (newRadius === undefined) {
      newStyle = (0, _utils.cleanEmptyObject)(newStyle);
    }

    setAttributes({
      style: newStyle
    });
  };

  return (0, _element.createElement)(_components.RangeControl, {
    value: style === null || style === void 0 ? void 0 : (_style$border = style.border) === null || _style$border === void 0 ? void 0 : _style$border.radius,
    label: (0, _i18n.__)('Border radius'),
    min: MIN_BORDER_RADIUS_VALUE,
    max: MAX_BORDER_RADIUS_VALUE,
    initialPosition: 0,
    allowReset: true,
    onChange: onChange
  });
}
/**
 * Determines if there is border radius support.
 *
 * @param  {string|Object} blockType Block name or Block Type object.
 * @return {boolean}                 Whether there is support.
 */


function hasBorderRadiusSupport(blockType) {
  var support = (0, _blocks.getBlockSupport)(blockType, _border.BORDER_SUPPORT_KEY);
  return !!(true === support || support !== null && support !== void 0 && support.radius);
}
/**
 * Custom hook that checks if border radius settings have been disabled.
 *
 * @param  {string} name The name of the block.
 * @return {boolean}                 Whether border radius setting is disabled.
 */


function useIsBorderRadiusDisabled() {
  var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
      blockName = _ref.name;

  var isDisabled = !(0, _useEditorFeature.default)('border.customRadius');
  return !hasBorderRadiusSupport(blockName) || isDisabled;
}
