// @wordpress/block-editor/build/hooks/font-family.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FontFamilyEdit = FontFamilyEdit;
exports.useIsFontFamilyDisabled = useIsFontFamilyDisabled;
exports.FONT_FAMILY_SUPPORT_KEY = void 0;

var _element = require("@wordpress/element");

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _lodash = require("lodash");

var _blocks = require("@wordpress/blocks");

var _utils = require("./utils");

var _useEditorFeature = _interopRequireDefault(require("../components/use-editor-feature"));

var _fontFamily = _interopRequireDefault(require("../components/font-family"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var FONT_FAMILY_SUPPORT_KEY = '__experimentalFontFamily';
exports.FONT_FAMILY_SUPPORT_KEY = FONT_FAMILY_SUPPORT_KEY;

var getFontFamilyFromAttributeValue = function getFontFamilyFromAttributeValue(fontFamilies, value) {
  var attributeParsed = /var:preset\|font-family\|(.+)/.exec(value);

  if (attributeParsed && attributeParsed[1]) {
    var fontFamilyObject = (0, _lodash.find)(fontFamilies, function (_ref) {
      var slug = _ref.slug;
      return slug === attributeParsed[1];
    });

    if (fontFamilyObject) {
      return fontFamilyObject.fontFamily;
    }
  }

  return value;
};

function FontFamilyEdit(_ref2) {
  var _style$typography;

  var name = _ref2.name,
      setAttributes = _ref2.setAttributes,
      _ref2$attributes$styl = _ref2.attributes.style,
      style = _ref2$attributes$styl === void 0 ? {} : _ref2$attributes$styl;
  var fontFamilies = (0, _useEditorFeature.default)('typography.fontFamilies');
  var isDisable = useIsFontFamilyDisabled({
    name: name
  });

  if (isDisable) {
    return null;
  }

  var value = getFontFamilyFromAttributeValue(fontFamilies, (_style$typography = style.typography) === null || _style$typography === void 0 ? void 0 : _style$typography.fontFamily);

  function onChange(newValue) {
    var predefinedFontFamily = (0, _lodash.find)(fontFamilies, function (_ref3) {
      var fontFamily = _ref3.fontFamily;
      return fontFamily === newValue;
    });
    setAttributes({
      style: (0, _utils.cleanEmptyObject)(_objectSpread(_objectSpread({}, style), {}, {
        typography: _objectSpread(_objectSpread({}, style.typography || {}), {}, {
          fontFamily: predefinedFontFamily ? "var:preset|font-family|".concat(predefinedFontFamily.slug) : newValue || undefined
        })
      }))
    });
  }

  return (0, _element.createElement)(_fontFamily.default, {
    className: "block-editor-hooks-font-family-control",
    fontFamilies: fontFamilies,
    value: value,
    onChange: onChange
  });
}
/**
 * Custom hook that checks if font-family functionality is disabled.
 *
 * @param {string} name The name of the block.
 * @return {boolean} Whether setting is disabled.
 */


function useIsFontFamilyDisabled(_ref4) {
  var name = _ref4.name;
  var fontFamilies = (0, _useEditorFeature.default)('typography.fontFamilies');
  return !fontFamilies || fontFamilies.length === 0 || !(0, _blocks.hasBlockSupport)(name, FONT_FAMILY_SUPPORT_KEY);
}
