// @wordpress/block-editor/build-module/components/inserter-draggable-blocks/index.js
import { createElement } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { Draggable } from '@wordpress/components';
/**
 * Internal dependencies
 */

import BlockDraggableChip from '../block-draggable/draggable-chip';

var InserterDraggableBlocks = function InserterDraggableBlocks(_ref) {
  var isEnabled = _ref.isEnabled,
      blocks = _ref.blocks,
      icon = _ref.icon,
      children = _ref.children;
  var transferData = {
    type: 'inserter',
    blocks: blocks
  };
  return createElement(Draggable, {
    transferData: transferData,
    __experimentalDragComponent: createElement(BlockDraggableChip, {
      count: blocks.length,
      icon: icon
    })
  }, function (_ref2) {
    var onDraggableStart = _ref2.onDraggableStart,
        onDraggableEnd = _ref2.onDraggableEnd;
    return children({
      draggable: isEnabled,
      onDragStart: isEnabled ? onDraggableStart : undefined,
      onDragEnd: isEnabled ? onDraggableEnd : undefined
    });
  });
};

export default InserterDraggableBlocks;
