// @wp-g2/components/dist/cjs/Switch/Switch.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _View = require("../View");

var _VisuallyHidden = require("../VisuallyHidden");

var styles = _interopRequireWildcard(require("./Switch.styles"));

var _SwitchBackdrop = _interopRequireDefault(require("./SwitchBackdrop"));

var _SwitchToggle = _interopRequireDefault(require("./SwitchToggle"));

var _useSwitch2 = require("./useSwitch");

function Switch(props, forwardedRef) {
  var _useSwitch = (0, _useSwitch2.useSwitch)(props),
      Component = _useSwitch.Component,
      checked = _useSwitch.checked,
      disabled = _useSwitch.disabled,
      error = _useSwitch.error,
      id = _useSwitch.id,
      isFocused = _useSwitch.isFocused,
      label = _useSwitch.label,
      onBlur = _useSwitch.onBlur,
      onChange = _useSwitch.onChange,
      onFocus = _useSwitch.onFocus,
      size = _useSwitch.size,
      type = _useSwitch.type,
      otherProps = (0, _objectWithoutProperties2.default)(_useSwitch, ["Component", "checked", "disabled", "error", "id", "isFocused", "label", "onBlur", "onChange", "onFocus", "size", "type"]);

  return /*#__PURE__*/_react.default.createElement(_View.View, Object.assign({
    as: "label"
  }, otherProps, {
    htmlFor: id
  }), /*#__PURE__*/_react.default.createElement(Component, Object.assign({}, otherProps, {
    checked: checked,
    className: styles.inputHidden,
    disabled: disabled,
    id: id,
    onBlur: onBlur,
    onChange: onChange,
    onFocus: onFocus,
    ref: forwardedRef
  }, _styles.ui.$('SwitchInput'))), /*#__PURE__*/_react.default.createElement(_SwitchBackdrop.default, {
    error: error,
    isFocused: isFocused,
    size: size
  }), /*#__PURE__*/_react.default.createElement(_SwitchToggle.default, {
    size: size
  }), /*#__PURE__*/_react.default.createElement(_VisuallyHidden.VisuallyHidden, _styles.ui.$('SwitchLabel'), label));
}

var _default = (0, _context.contextConnect)(Switch, 'Switch');

exports.default = _default;
