// @wp-g2/components/dist/cjs/Switch/Switch.styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.backdropErrorFocus = exports.backdropError = exports.formGroup = exports.toggleSmall = exports.toggleLarge = exports.Toggle = exports.focus = exports.Backdrop = exports.inputHidden = exports.small = exports.large = exports.disabled = exports.Switch = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _styles = require("@wp-g2/styles");

function _templateObject13() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tborder-color: transparent;\n\tbox-shadow: ", ";\n"]);

  _templateObject13 = function () {
    return data;
  };

  return data;
}

function _templateObject12() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tborder-color: ", ";\n\n\tinput:checked ~ & {\n\t\tbackground: ", ";\n\t\tborder-color: ", ";\n\t}\n"]);

  _templateObject12 = function () {
    return data;
  };

  return data;
}

function _templateObject11() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tmargin-left: auto;\n"]);

  _templateObject11 = function () {
    return data;
  };

  return data;
}

function _templateObject10() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tborder-radius: ", ";\n\theight: ", ";\n\twidth: ", ";\n\n\t*:active > & {\n\t\twidth: ", ";\n\t}\n"]);

  _templateObject10 = function () {
    return data;
  };

  return data;
}

function _templateObject9() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tborder-radius: ", ";\n\theight: ", ";\n\twidth: ", ";\n\n\t*:active > & {\n\t\twidth: ", ";\n\t}\n"]);

  _templateObject9 = function () {
    return data;
  };

  return data;
}

function _templateObject8() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbackground: ", ";\n\tborder: 1px solid ", ";\n\tborder-radius: ", ";\n\tbox-shadow: ", ";\n\theight: ", ";\n\tpointer-events: none;\n\tright: initial;\n\ttransform: translate(0, 0);\n\ttransition: all ", " linear;\n\twidth: ", ";\n\n\t", "\n\n\t*:active > & {\n\t\twidth: ", ";\n\t}\n\n\tinput:checked ~ & {\n\t\tbackground: ", ";\n\t\tleft: initial;\n\t\tright: ", ";\n\t}\n"]);

  _templateObject8 = function () {
    return data;
  };

  return data;
}

function _templateObject7() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbox-shadow: ", ";\n"]);

  _templateObject7 = function () {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\n\tbackground: ", ";\n\tborder: 1px solid ", ";\n\tbottom: ", ";\n\tbox-shadow: 0 0 0 0 rgba(0, 0, 0, 0);\n\tdisplay: block;\n\tleft: 0;\n\tpointer-events: none;\n\tposition: absolute;\n\ttop: ", ";\n\ttransition: all ", " linear;\n\twidth: 100%;\n\n\tinput:checked ~ & {\n\t\tbackground: ", ";\n\t\tborder-color: ", ";\n\t}\n"]);

  _templateObject6 = function () {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\topacity: 0;\n\tpadding: 0;\n\tposition: absolute;\n\tz-index: -1;\n"]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\theight: ", ";\n\twidth: ", ";\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\theight: ", ";\n\twidth: ", ";\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\topacity: 0.6;\n\tpointer-events: none;\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tcursor: pointer;\n\tdisplay: flex;\n\theight: ", ";\n\tmargin: 0;\n\toutline: none;\n\tpadding: ", " 0;\n\tposition: relative;\n\tuser-select: none;\n\twidth: ", ";\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

function getSwitchWidth(height) {
  return "calc(" + getControlHeight(height) + " * 1.85);";
}

var Switch = (0, _styles.css)(_templateObject(), _styles.ui.get('controlHeight'), _styles.ui.get('switchPaddingOffset'), getSwitchWidth('controlHeight'));
exports.Switch = Switch;
var disabled = (0, _styles.css)(_templateObject2());
exports.disabled = disabled;
var large = (0, _styles.css)(_templateObject3(), _styles.ui.get('controlHeightLarge'), getSwitchWidth('controlHeightLarge'));
exports.large = large;
var small = (0, _styles.css)(_templateObject4(), _styles.ui.get('controlHeightSmall'), getSwitchWidth('controlHeightSmall'));
exports.small = small;
var inputHidden = (0, _styles.css)(_templateObject5());
exports.inputHidden = inputHidden;
var Backdrop = (0, _styles.css)(_templateObject6(), _styles.ui.borderRadius.circle, _styles.ui.get('switchBackdropBackgroundColor'), _styles.ui.get('switchBackdropBorderColor'), _styles.ui.get('switchPaddingOffset'), _styles.ui.get('switchPaddingOffset'), _styles.ui.get('transitionDurationFast'), _styles.ui.get('switchBackdropBackgroundColorActive'), _styles.ui.get('switchBackdropBorderColorActive'));
exports.Backdrop = Backdrop;
var focus = (0, _styles.css)(_templateObject7(), _styles.ui.get('controlPseudoBoxShadowFocus'));
exports.focus = focus;
var Toggle = (0, _styles.css)(_templateObject8(), _styles.ui.get('switchToggleBackgroundColor'), _styles.ui.get('switchToggleBorderColor'), getToggleHeight('controlHeight'), _styles.ui.get('switchToggleBoxShadow'), getToggleHeight('controlHeight'), _styles.ui.get('transitionDurationFast'), getToggleHeight('controlHeight'), getTogglePosition(), getControlHeight('controlHeight'), _styles.ui.get('switchToggleBackgroundColorActive'), getTogglePositionX());
exports.Toggle = Toggle;
var toggleLarge = (0, _styles.css)(_templateObject9(), getToggleHeight('controlHeightLarge'), getToggleHeight('controlHeightLarge'), getToggleHeight('controlHeightLarge'), getControlHeight('controlHeightLarge'));
exports.toggleLarge = toggleLarge;
var toggleSmall = (0, _styles.css)(_templateObject10(), getToggleHeight('controlHeightSmall'), getToggleHeight('controlHeightSmall'), getToggleHeight('controlHeightSmall'), getControlHeight('controlHeightSmall'));
exports.toggleSmall = toggleSmall;
var formGroup = (0, _styles.css)(_templateObject11());
exports.formGroup = formGroup;
var backdropError = (0, _styles.css)(_templateObject12(), _styles.ui.get('controlDestructiveBorderColor'), _styles.ui.get('controlDestructiveBorderColor'), _styles.ui.get('controlDestructiveBorderColor'));
exports.backdropError = backdropError;
var backdropErrorFocus = (0, _styles.css)(_templateObject13(), _styles.ui.get('controlDestructivePseudoBoxShadowFocus'));
exports.backdropErrorFocus = backdropErrorFocus;

function getControlHeight(height) {
  return "calc(" + _styles.ui.get(height) + " - calc(" + _styles.ui.get('switchPaddingOffset') + " * 2))";
}

function getToggleHeight(height) {
  return "calc(" + getControlHeight(height) + " - " + _styles.ui.get('switchPaddingOffset') + ")";
}

function getTogglePosition() {
  var value = "calc(" + _styles.ui.get('switchPaddingOffset') + " * 2)";
  var left = getTogglePositionX();
  return (0, _styles.css)({
    top: "calc(" + value + " * 0.75)",
    left: left,
    position: 'absolute'
  });
}

function getTogglePositionX() {
  var value = "calc(" + _styles.ui.get('switchPaddingOffset') + " * 2)";
  return "calc(" + value + " * 0.25)";
}
