// @wordpress/rich-text/build/component/format-edit.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = FormatEdit;

var _element = require("@wordpress/element");

var _getActiveFormat = require("../get-active-format");

var _getActiveObject = require("../get-active-object");

/**
 * Internal dependencies
 */
function FormatEdit(_ref) {
  var formatTypes = _ref.formatTypes,
      onChange = _ref.onChange,
      onFocus = _ref.onFocus,
      value = _ref.value,
      forwardedRef = _ref.forwardedRef;
  return formatTypes.map(function (settings) {
    var name = settings.name,
        Edit = settings.edit;

    if (!Edit) {
      return null;
    }

    var activeFormat = (0, _getActiveFormat.getActiveFormat)(value, name);
    var isActive = activeFormat !== undefined;
    var activeObject = (0, _getActiveObject.getActiveObject)(value);
    var isObjectActive = activeObject !== undefined && activeObject.type === name;
    return (0, _element.createElement)(Edit, {
      key: name,
      isActive: isActive,
      activeAttributes: isActive ? activeFormat.attributes || {} : {},
      isObjectActive: isObjectActive,
      activeObjectAttributes: isObjectActive ? activeObject.attributes || {} : {},
      value: value,
      onChange: onChange,
      onFocus: onFocus,
      contentRef: forwardedRef
    });
  });
}
