// @wp-g2/components/dist/cjs/Navigator/Router/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Link: true,
  MemoryRouter: true,
  NavLink: true,
  Route: true,
  Router: true,
  Switch: true,
  matchPath: true,
  withRouter: true,
  HistoryContext: true,
  RouterContext: true
};
Object.defineProperty(exports, "Link", {
  enumerable: true,
  get: function get() {
    return _Link.default;
  }
});
Object.defineProperty(exports, "MemoryRouter", {
  enumerable: true,
  get: function get() {
    return _MemoryRouter.default;
  }
});
Object.defineProperty(exports, "NavLink", {
  enumerable: true,
  get: function get() {
    return _NavLink.default;
  }
});
Object.defineProperty(exports, "Route", {
  enumerable: true,
  get: function get() {
    return _Route.default;
  }
});
Object.defineProperty(exports, "Router", {
  enumerable: true,
  get: function get() {
    return _Router.default;
  }
});
Object.defineProperty(exports, "Switch", {
  enumerable: true,
  get: function get() {
    return _Switch.default;
  }
});
Object.defineProperty(exports, "matchPath", {
  enumerable: true,
  get: function get() {
    return _matchPath.default;
  }
});
Object.defineProperty(exports, "withRouter", {
  enumerable: true,
  get: function get() {
    return _withRouter.default;
  }
});
Object.defineProperty(exports, "HistoryContext", {
  enumerable: true,
  get: function get() {
    return _HistoryContext.default;
  }
});
Object.defineProperty(exports, "RouterContext", {
  enumerable: true,
  get: function get() {
    return _RouterContext.default;
  }
});

var _Link = _interopRequireDefault(require("./Link"));

var _MemoryRouter = _interopRequireDefault(require("./MemoryRouter"));

var _NavLink = _interopRequireDefault(require("./NavLink"));

var _Route = _interopRequireDefault(require("./Route"));

var _Router = _interopRequireDefault(require("./Router"));

var _Switch = _interopRequireDefault(require("./Switch"));

var _matchPath = _interopRequireDefault(require("./matchPath"));

var _withRouter = _interopRequireDefault(require("./withRouter"));

var _HistoryContext = _interopRequireDefault(require("./HistoryContext"));

var _RouterContext = _interopRequireDefault(require("./RouterContext"));

var _hooks = require("./hooks");

Object.keys(_hooks).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _hooks[key];
    }
  });
});
