// @wp-g2/components/dist/cjs/Navigator/Router/NavLink.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _Link = _interopRequireDefault(require("./Link"));

var _matchPath = _interopRequireDefault(require("./matchPath"));

var _RouterContext = _interopRequireDefault(require("./RouterContext"));

var _locationUtils = require("./utils/locationUtils");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

// React 15 compat
var forwardRefShim = function (C) {
  return C;
};

var forwardRef = _react.default.forwardRef;

if (typeof forwardRef === 'undefined') {
  forwardRef = forwardRefShim;
}

function joinClassnames() {
  for (var _len = arguments.length, classnames = new Array(_len), _key = 0; _key < _len; _key++) {
    classnames[_key] = arguments[_key];
  }

  return classnames.filter(function (i) {
    return i;
  }).join(' ');
}
/**
 * A <Link> wrapper that knows if it's "active" or not.
 */


var NavLink = forwardRef(function (_ref, forwardedRef) {
  var _ref$activeClassName = _ref.activeClassName,
      activeClassName = _ref$activeClassName === void 0 ? 'active' : _ref$activeClassName,
      activeStyle = _ref.activeStyle,
      _ref$ariaCurrent = _ref['aria-current'],
      ariaCurrent = _ref$ariaCurrent === void 0 ? 'page' : _ref$ariaCurrent,
      classNameProp = _ref.className,
      exact = _ref.exact,
      innerRef = _ref.innerRef,
      isActiveProp = _ref.isActive,
      locationProp = _ref.location,
      sensitive = _ref.sensitive,
      strict = _ref.strict,
      styleProp = _ref.style,
      to = _ref.to,
      rest = (0, _objectWithoutProperties2.default)(_ref, ["activeClassName", "activeStyle", "aria-current", "className", "exact", "innerRef", "isActive", "location", "sensitive", "strict", "style", "to"]);
  return /*#__PURE__*/_react.default.createElement(_RouterContext.default.Consumer, null, function (context) {
    var currentLocation = locationProp || context.location;
    var toLocation = (0, _locationUtils.normalizeToLocation)((0, _locationUtils.resolveToLocation)(to, currentLocation), currentLocation);
    var path = toLocation.pathname; // Regex taken from: https://github.com/pillarjs/path-to-regexp/blob/master/index.js#L202

    var escapedPath = path && path.replace(/([.+*?=^!:${}()[\]|/\\])/g, '\\$1');
    var match = escapedPath ? (0, _matchPath.default)(currentLocation.pathname, {
      exact: exact,
      path: escapedPath,
      sensitive: sensitive,
      strict: strict
    }) : null;
    var isActive = !!(isActiveProp ? isActiveProp(match, currentLocation) : match);
    var className = isActive ? joinClassnames(classNameProp, activeClassName) : classNameProp;
    var style = isActive ? _objectSpread(_objectSpread({}, styleProp), activeStyle) : styleProp;

    var props = _objectSpread({
      'aria-current': isActive && ariaCurrent || null,
      className: className,
      style: style,
      to: toLocation
    }, rest); // React 15 compat


    if (forwardRefShim !== forwardRef) {
      props.ref = forwardedRef || innerRef;
    } else {
      props.innerRef = innerRef;
    }

    return /*#__PURE__*/_react.default.createElement(_Link.default, props);
  });
});
var _default = NavLink;
exports.default = _default;
