// @wordpress/components/build/ui/text/styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.upperCase = exports.highlighterText = exports.muted = exports.destructive = exports.positive = exports.block = exports.Text = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _styles = require("@wp-g2/styles");

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7;

var Text = (0, _styles.css)(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2.default)(["\n\tcolor: ", ";\n\tline-height: ", ";\n"])), _styles.ui.color.text, _styles.ui.get('fontLineHeightBase'));
exports.Text = Text;
var block = (0, _styles.css)(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2.default)(["\n\tdisplay: block;\n"])));
exports.block = block;
var positive = (0, _styles.css)(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2.default)(["\n\tcolor: ", ";\n"])), _styles.ui.color.positive);
exports.positive = positive;
var destructive = (0, _styles.css)(_templateObject4 || (_templateObject4 = (0, _taggedTemplateLiteral2.default)(["\n\tcolor: ", ";\n"])), _styles.ui.color.destructive);
exports.destructive = destructive;
var muted = (0, _styles.css)(_templateObject5 || (_templateObject5 = (0, _taggedTemplateLiteral2.default)(["\n\tcolor: ", ";\n"])), _styles.ui.get('colorTextMuted'));
exports.muted = muted;
var highlighterText = (0, _styles.css)(_templateObject6 || (_templateObject6 = (0, _taggedTemplateLiteral2.default)(["\n\tmark {\n\t\tbackground: ", ";\n\t\tborder-radius: 2px;\n\t\tbox-shadow: 0 0 0 1px rgba( 0, 0, 0, 0.05 ) inset,\n\t\t\t0 -1px 0 rgba( 0, 0, 0, 0.1 ) inset;\n\t}\n"])), _styles.ui.get('yellowRgba70'));
exports.highlighterText = highlighterText;
var upperCase = (0, _styles.css)(_templateObject7 || (_templateObject7 = (0, _taggedTemplateLiteral2.default)(["\n\ttext-transform: uppercase;\n"])));
exports.upperCase = upperCase;
