// @wp-g2/components/dist/cjs/__stories__/Controls.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Controls = exports.default = void 0;

var _icons = require("@wordpress/icons");

var _context = require("@wp-g2/context");

var _icons2 = require("@wp-g2/icons");

var _react = _interopRequireDefault(require("react"));

var _components = require("../__fixtures__/components");

var _index = require("../index");

var _default = {
  title: 'Examples'
};
exports.default = _default;

var SuffixLabel = function (props) {
  return /*#__PURE__*/_react.default.createElement(_index.Text, Object.assign({
    isBlock: true,
    size: 10,
    variant: "muted"
  }, props, {
    css: {
      userSelect: 'none'
    }
  }));
};

var ResponsiveControls = function () {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_index.Flex, {
    justify: "center"
  }, /*#__PURE__*/_react.default.createElement(_index.SegmentedControl, {
    options: [{
      label: /*#__PURE__*/_react.default.createElement(_index.Icon, {
        icon: _icons.mobile
      }),
      value: 'phone'
    }, {
      label: /*#__PURE__*/_react.default.createElement(_index.Icon, {
        icon: _icons.tablet
      }),
      value: 'tablet'
    }, {
      label: /*#__PURE__*/_react.default.createElement(_index.Icon, {
        icon: _icons.desktop
      }),
      value: 'desktop'
    }]
  })));
};

var LayoutControls = function () {
  return /*#__PURE__*/_react.default.createElement(_index.Panel, {
    visible: true
  }, /*#__PURE__*/_react.default.createElement(_index.PanelHeader, {
    title: "Layout"
  }), /*#__PURE__*/_react.default.createElement(_index.PanelBody, null, /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Dimensions"
  }, /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: 2
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    suffix: /*#__PURE__*/_react.default.createElement(SuffixLabel, null, "W")
  }), /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    id: "d-h",
    suffix: /*#__PURE__*/_react.default.createElement(SuffixLabel, null, "H")
  }))), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Align"
  }, /*#__PURE__*/_react.default.createElement(_index.SegmentedControl, {
    isBlock: true,
    options: [{
      label: /*#__PURE__*/_react.default.createElement(_index.Icon, {
        icon: _icons.alignLeft
      }),
      value: 'left'
    }, {
      label: /*#__PURE__*/_react.default.createElement(_index.Icon, {
        icon: _icons.alignCenter
      }),
      value: 'center'
    }, {
      label: /*#__PURE__*/_react.default.createElement(_index.Icon, {
        icon: _icons.alignRight
      }),
      value: 'right'
    }]
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Position"
  }, /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: 2
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    suffix: /*#__PURE__*/_react.default.createElement(SuffixLabel, null, "X")
  }), /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    id: "pos-y",
    suffix: /*#__PURE__*/_react.default.createElement(SuffixLabel, null, "Y")
  }))), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Distribution"
  }, /*#__PURE__*/_react.default.createElement(_index.Select, {
    options: [{
      label: 'Center',
      value: 'center'
    }]
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Padding"
  }, /*#__PURE__*/_react.default.createElement(_index.View, null, /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: 2
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    suffix: /*#__PURE__*/_react.default.createElement(SuffixLabel, null, "PX")
  }), /*#__PURE__*/_react.default.createElement(_index.Stepper, {
    direction: "horizontal"
  }))), /*#__PURE__*/_react.default.createElement(_index.SegmentedControl, {
    isBlock: true,
    options: [{
      label: /*#__PURE__*/_react.default.createElement(_index.Icon, {
        icon: /*#__PURE__*/_react.default.createElement(_icons2.FiSquare, null)
      }),
      // @todo this is missing a suitable wp/icons replacement
      value: 'all'
    }, {
      label: /*#__PURE__*/_react.default.createElement(_index.Icon, {
        icon: _icons.fullscreen
      }),
      value: 'sides'
    }]
  }))), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Overflow"
  }, /*#__PURE__*/_react.default.createElement(_index.SegmentedControl, {
    isBlock: true,
    options: [{
      label: 'Show',
      value: 'show'
    }, {
      label: 'Hide',
      value: 'hide'
    }]
  })))));
};

var TypographyControls = function () {
  return /*#__PURE__*/_react.default.createElement(_index.Panel, {
    visible: true
  }, /*#__PURE__*/_react.default.createElement(_index.PanelHeader, {
    title: "Typography"
  }), /*#__PURE__*/_react.default.createElement(_index.PanelBody, null, /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Font"
  }, /*#__PURE__*/_react.default.createElement(_index.Select, {
    options: [{
      label: 'Inter',
      value: 'inter'
    }]
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Style"
  }, /*#__PURE__*/_react.default.createElement(_index.Select, {
    options: [{
      label: 'Medium',
      value: 'medium'
    }]
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Size"
  }, /*#__PURE__*/_react.default.createElement(_index.Flex, {
    gap: 2
  }, /*#__PURE__*/_react.default.createElement(_index.FlexBlock, null, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    value: "16"
  })), /*#__PURE__*/_react.default.createElement(_index.FlexBlock, null, /*#__PURE__*/_react.default.createElement(_index.Slider, null)))), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Transform"
  }, /*#__PURE__*/_react.default.createElement(_index.Select, {
    options: [{
      label: 'None',
      value: 'none'
    }]
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Dropcap"
  }, /*#__PURE__*/_react.default.createElement(_index.Flex, {
    justify: "flex-end"
  }, /*#__PURE__*/_react.default.createElement(_index.Switch, null))))));
};

var ActionsControl = function () {
  return /*#__PURE__*/_react.default.createElement(_index.Panel, {
    visible: true
  }, /*#__PURE__*/_react.default.createElement(_index.PanelHeader, {
    title: "Actions"
  }), /*#__PURE__*/_react.default.createElement(_index.PanelBody, null, /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: 2
  }, /*#__PURE__*/_react.default.createElement(_index.Button, null, "Save Draft"), /*#__PURE__*/_react.default.createElement(_index.Button, {
    variant: "primary"
  }, "Publish"))));
};

var Controls = function () {
  return /*#__PURE__*/_react.default.createElement(_context.ContextSystemProvider, {
    value: {
      FormGroup: {
        horizontal: true
      }
    }
  }, /*#__PURE__*/_react.default.createElement(_components.Sidebar, null, /*#__PURE__*/_react.default.createElement(_index.Spacer, null), /*#__PURE__*/_react.default.createElement(ResponsiveControls, null), /*#__PURE__*/_react.default.createElement(_index.Divider, {
    mb: 0,
    mt: 2
  }), /*#__PURE__*/_react.default.createElement(LayoutControls, null), /*#__PURE__*/_react.default.createElement(TypographyControls, null), /*#__PURE__*/_react.default.createElement(ActionsControl, null)));
};

exports.Controls = Controls;
