// @wordpress/components/build/ui/h-stack/hook.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useHStack = useHStack;

var _element = require("@wordpress/element");

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _utils = require("@wp-g2/utils");

var _flex = require("../flex");

var _utils2 = require("./utils");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 *
 * @param {import('@wp-g2/create-styles').ViewOwnProps<import('./types').Props, 'div'>} props
 */
function useHStack(props) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'HStack'),
      _useContextSystem$ali = _useContextSystem.alignment,
      alignment = _useContextSystem$ali === void 0 ? 'edge' : _useContextSystem$ali,
      children = _useContextSystem.children,
      direction = _useContextSystem.direction,
      _useContextSystem$spa = _useContextSystem.spacing,
      spacing = _useContextSystem$spa === void 0 ? 2 : _useContextSystem$spa,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["alignment", "children", "direction", "spacing"]);

  var align = (0, _utils2.getAlignmentProps)(alignment, direction);
  var validChildren = (0, _utils.getValidChildren)(children);
  var clonedChildren = validChildren.map( // @ts-ignore
  function (
  /** @type {import('react').ReactElement} */
  child, index) {
    var _key = child.key || "hstack-".concat(index);

    var _isSpacer = (0, _context.hasNamespace)(child, ['Spacer']);

    if (_isSpacer) {
      return (0, _element.createElement)(_flex.FlexItem, (0, _extends2.default)({
        isBlock: true,
        key: _key
      }, child.props, _styles.ui.$('Spacer')));
    }

    return child;
  });

  var propsForFlex = _objectSpread(_objectSpread(_objectSpread({
    children: clonedChildren,
    direction: direction,
    justify: 'center'
  }, align), otherProps), {}, {
    gap: spacing
  });

  var flexProps = (0, _flex.useFlex)(propsForFlex);
  return flexProps;
}
