// @wp-g2/components/dist/cjs/Surface/Surface.styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getBorders = getBorders;
exports.grid = exports.dotted = exports.tertiary = exports.secondary = exports.background = exports.Surface = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _styles = require("@wp-g2/styles");

function _templateObject6() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbackground: ", ";\n\tbackground-image: linear-gradient(\n\t\t\t", " 1px,\n\t\t\ttransparent 1px\n\t\t),\n\t\tlinear-gradient(\n\t\t\t90deg,\n\t\t\t", " 1px,\n\t\t\ttransparent 1px\n\t\t);\n\tbackground-size: ", "\n\t\t", ";\n"]);

  _templateObject6 = function () {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbackground: linear-gradient(\n\t\t\t\t90deg,\n\t\t\t\t", "\n\t\t\t\t\t", ",\n\t\t\t\ttransparent 1%\n\t\t\t)\n\t\t\tcenter,\n\t\tlinear-gradient(\n\t\t\t\t", "\n\t\t\t\t\t", ",\n\t\t\t\ttransparent 1%\n\t\t\t)\n\t\t\tcenter,\n\t\t", ";\n\tbackground-size: ", "\n\t\t", ";\n"]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbackground: ", ";\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbackground: ", ";\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbackground-color: ", ";\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbackground-color: ", ";\n\tcolor: ", ";\n\tposition: relative;\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var Surface = (0, _styles.css)(_templateObject(), _styles.ui.get('surfaceColor'), _styles.ui.color.text);
exports.Surface = Surface;
var background = (0, _styles.css)(_templateObject2(), _styles.ui.get('surfaceBackgroundColor'));
exports.background = background;

function getBorders(_ref) {
  var border = _ref.border,
      borderBottom = _ref.borderBottom,
      borderLeft = _ref.borderLeft,
      borderRight = _ref.borderRight,
      borderTop = _ref.borderTop;

  var borderStyle = "1px solid " + _styles.ui.get('surfaceBorderColor');

  return (0, _styles.css)({
    border: border && borderStyle,
    borderBottom: borderBottom && borderStyle,
    borderLeft: borderLeft && borderStyle,
    borderRight: borderRight && borderStyle,
    borderTop: borderTop && borderStyle
  });
}

var secondary = (0, _styles.css)(_templateObject3(), _styles.ui.get('surfaceBackgroundTintColor'));
exports.secondary = secondary;
var tertiary = (0, _styles.css)(_templateObject4(), _styles.ui.get('surfaceBackgroundTertiaryColor'));
exports.tertiary = tertiary;
var dotted = (0, _styles.css)(_templateObject5(), _styles.ui.get('surfaceBackgroundColor'), _styles.ui.get('surfaceBackgroundSizeDotted'), _styles.ui.get('surfaceBackgroundColor'), _styles.ui.get('surfaceBackgroundSizeDotted'), _styles.ui.get('surfaceBorderBoldColor'), _styles.ui.get('surfaceBackgroundSize'), _styles.ui.get('surfaceBackgroundSize'));
exports.dotted = dotted;
var grid = (0, _styles.css)(_templateObject6(), _styles.ui.get('surfaceBackgroundColor'), _styles.ui.get('surfaceBorderSubtleColor'), _styles.ui.get('surfaceBorderSubtleColor'), _styles.ui.get('surfaceBackgroundSize'), _styles.ui.get('surfaceBackgroundSize'));
exports.grid = grid;
