// @wp-g2/components/dist/cjs/Surface/useSurface.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useSurface = useSurface;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _react = require("react");

var styles = _interopRequireWildcard(require("./Surface.styles"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * @param {import('@wp-g2/create-styles').ViewOwnProps<import('./types').Props, 'div'>} props
 */
function useSurface(props) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'Surface'),
      _useContextSystem$bac = _useContextSystem.backgroundSize,
      backgroundSize = _useContextSystem$bac === void 0 ? 16 : _useContextSystem$bac,
      border = _useContextSystem.border,
      borderBottom = _useContextSystem.borderBottom,
      borderLeft = _useContextSystem.borderLeft,
      borderRight = _useContextSystem.borderRight,
      borderTop = _useContextSystem.borderTop,
      className = _useContextSystem.className,
      _useContextSystem$var = _useContextSystem.variant,
      variant = _useContextSystem$var === void 0 ? 'primary' : _useContextSystem$var,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["backgroundSize", "border", "borderBottom", "borderLeft", "borderRight", "borderTop", "className", "variant"]);

  var classes = (0, _react.useMemo)(function () {
    var _css;

    var sx = {};
    sx.borders = styles.getBorders({
      border: border,
      borderBottom: borderBottom,
      borderLeft: borderLeft,
      borderRight: borderRight,
      borderTop: borderTop
    });
    return (0, _styles.cx)(styles.Surface, sx.borders, styles[variant], (0, _styles.css)((_css = {}, _css[_styles.ui.createToken('surfaceBackgroundSize')] = _styles.ui.value.px(backgroundSize), _css[_styles.ui.createToken('surfaceBackgroundSizeDotted')] = _styles.ui.value.px(backgroundSize - 1), _css)), className);
  }, [backgroundSize, border, borderBottom, borderLeft, borderRight, borderTop, className, variant]);
  return _objectSpread(_objectSpread({}, otherProps), {}, {
    className: classes
  });
}
