// @wordpress/components/build-module/ui/__storybook-utils/example-grid.js
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { Container } from '@wp-g2/components';
/**
 * Internal dependencies
 */

import { Grid } from '../grid';
import { Surface } from '../surface';
import { Text } from '../text';
import { View } from '../view';
import { VStack } from '../v-stack';
export var ExampleGrid = function ExampleGrid(_ref) {
  var children = _ref.children;
  return createElement(Container, null, createElement(Grid, {
    columns: [2, 2, 4],
    gap: "1px"
  }, children));
};
export var ExampleGridItem = function ExampleGridItem(_ref2) {
  var children = _ref2.children;
  return createElement(Surface, {
    variant: "secondary",
    css: {
      padding: 20
    }
  }, createElement(VStack, {
    alignment: "center"
  }, children));
};
export var ExampleMetaContent = function ExampleMetaContent(_ref3) {
  var title = _ref3.title,
      _ref3$items = _ref3.items,
      items = _ref3$items === void 0 ? [] : _ref3$items;
  return createElement(Grid, {
    templateColumns: "60px 1fr"
  }, createElement(View, null, createElement(Text, {
    size: "caption",
    weight: "bold",
    align: "right",
    isBlock: true
  }, title)), createElement(VStack, {
    spacing: 1
  }, items.map(function (item, index) {
    return createElement(View, {
      key: index
    }, createElement(Text, {
      size: "caption",
      isBlock: true
    }, item));
  })));
};
