// @wp-g2/components/dist/cjs/BaseChoice/BaseChoice.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _react = _interopRequireDefault(require("react"));

var _FormGroup = require("../FormGroup");

var _Grid = require("../Grid");

var _View = require("../View");

function BaseChoice(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'BaseChoice'),
      children = _useContextSystem.children,
      gap = _useContextSystem.gap,
      help = _useContextSystem.help,
      label = _useContextSystem.label,
      labelHidden = _useContextSystem.labelHidden,
      _useContextSystem$spa = _useContextSystem.spacing,
      spacing = _useContextSystem$spa === void 0 ? 1 : _useContextSystem$spa,
      _useContextSystem$tem = _useContextSystem.templateColumns,
      templateColumns = _useContextSystem$tem === void 0 ? 'auto 1fr' : _useContextSystem$tem,
      _useContextSystem$tru = _useContextSystem.truncate,
      truncate = _useContextSystem$tru === void 0 ? true : _useContextSystem$tru,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["children", "gap", "help", "label", "labelHidden", "spacing", "templateColumns", "truncate"]);

  var _useFormGroup = (0, _FormGroup.useFormGroup)(otherProps),
      id = _useFormGroup.contentProps.id;

  var contextProps = _react.default.useMemo(function () {
    return {
      id: id,
      horizontal: true
    };
  }, [id]);

  if (!label) return children;
  return /*#__PURE__*/_react.default.createElement(_Grid.Grid, Object.assign({
    gap: gap,
    templateColumns: templateColumns
  }, otherProps, {
    ref: forwardedRef
  }), /*#__PURE__*/_react.default.createElement(_FormGroup.FormGroupContext.Provider, {
    value: contextProps
  }, children, /*#__PURE__*/_react.default.createElement(_View.View, null, /*#__PURE__*/_react.default.createElement(_FormGroup.FormGroupContent, {
    id: id,
    help: help,
    label: label,
    labelHidden: labelHidden,
    spacing: spacing,
    truncate: truncate
  }))));
}

var _default = (0, _context.contextConnect)(BaseChoice, 'BaseChoice');

exports.default = _default;
