// @wordpress/components/build/ui/font-size-control/font-size-control.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _i18n = require("@wordpress/i18n");

var _visuallyHidden = require("../visually-hidden");

var _vStack = require("../v-stack");

var _select = _interopRequireDefault(require("./select"));

var _slider = _interopRequireDefault(require("./slider"));

var _useFontSizeControl2 = _interopRequireDefault(require("./use-font-size-control"));

var _view = require("../view");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function FontSizeControl(props, forwardedRef) {
  var _useFontSizeControl = (0, _useFontSizeControl2.default)(props),
      options = _useFontSizeControl.options,
      inputValue = _useFontSizeControl.inputValue,
      isDefaultValue = _useFontSizeControl.isDefaultValue,
      value = _useFontSizeControl.value,
      onChange = _useFontSizeControl.onChange,
      onReset = _useFontSizeControl.onReset,
      onInputChange = _useFontSizeControl.onInputChange,
      size = _useFontSizeControl.size,
      withSlider = _useFontSizeControl.withSlider,
      withNumberInput = _useFontSizeControl.withNumberInput,
      withSelect = _useFontSizeControl.withSelect,
      otherProps = (0, _objectWithoutProperties2.default)(_useFontSizeControl, ["options", "inputValue", "isDefaultValue", "value", "onChange", "onReset", "onInputChange", "size", "withSlider", "withNumberInput", "withSelect"]);

  if (!options) return null;
  var label = (0, _i18n.__)('Font size');
  var customLabel = (0, _i18n.__)('Custom');
  var max = 100;
  var min = 1;
  return (0, _element.createElement)(_view.View, (0, _extends2.default)({
    as: "fieldset"
  }, otherProps), (0, _element.createElement)(_visuallyHidden.VisuallyHidden, {
    as: "legend"
  }, label), (0, _element.createElement)(_vStack.VStack, {
    spacing: 3
  }, (0, _element.createElement)(_select.default, {
    customLabel: customLabel,
    inputValue: inputValue,
    isDefaultValue: isDefaultValue,
    label: label,
    max: max,
    min: min,
    onChange: onChange,
    onInputChange: onInputChange,
    onReset: onReset,
    options: options,
    ref: forwardedRef,
    size: size,
    value: value,
    withNumberInput: withNumberInput,
    withSelect: withSelect
  }), (0, _element.createElement)(_slider.default, {
    max: max,
    min: min,
    onChange: onInputChange,
    size: size,
    value: inputValue,
    withSlider: withSlider
  })));
}

var _default = (0, _context.contextConnect)(FontSizeControl, 'FontSizeControl');

exports.default = _default;
