// @wordpress/components/build/ui/font-size-control/select.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _lodash = require("lodash");

var _context = require("@wp-g2/context");

var _components = require("@wp-g2/components");

var _i18n = require("@wordpress/i18n");

var _utils = require("./utils");

var _grid = require("../grid");

var _view = require("../view");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function renderItem(_ref) {
  var name = _ref.name,
      style = _ref.style;
  return (0, _element.createElement)("span", {
    style: style
  }, name);
}

function FontSizeControlSelect(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'FontSizeControlSelect'),
      customLabel = _useContextSystem.customLabel,
      disabled = _useContextSystem.disabled,
      inputValue = _useContextSystem.inputValue,
      isDefaultValue = _useContextSystem.isDefaultValue,
      label = _useContextSystem.label,
      max = _useContextSystem.max,
      min = _useContextSystem.min,
      _useContextSystem$onC = _useContextSystem.onChange,
      onChange = _useContextSystem$onC === void 0 ? _lodash.noop : _useContextSystem$onC,
      _useContextSystem$onI = _useContextSystem.onInputChange,
      onInputChange = _useContextSystem$onI === void 0 ? _lodash.noop : _useContextSystem$onI,
      _useContextSystem$onR = _useContextSystem.onReset,
      onReset = _useContextSystem$onR === void 0 ? _lodash.noop : _useContextSystem$onR,
      _useContextSystem$opt = _useContextSystem.options,
      options = _useContextSystem$opt === void 0 ? [] : _useContextSystem$opt,
      size = _useContextSystem.size,
      value = _useContextSystem.value,
      withNumberInput = _useContextSystem.withNumberInput,
      withSelect = _useContextSystem.withSelect,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["customLabel", "disabled", "inputValue", "isDefaultValue", "label", "max", "min", "onChange", "onInputChange", "onReset", "options", "size", "value", "withNumberInput", "withSelect"]);

  var templateColumns = (0, _utils.getSelectTemplateColumns)(withNumberInput);
  var subControlsTemplateColumns = withNumberInput ? '1fr 1fr' : '1fr';
  return (0, _element.createElement)(_grid.Grid, {
    alignment: "bottom",
    templateColumns: templateColumns
  }, withSelect && (0, _element.createElement)(_components.FormGroup, {
    label: label
  }, (0, _element.createElement)(_components.SelectDropdown, (0, _extends2.default)({
    disabled: disabled,
    max: 260,
    onChange: onChange,
    options: options,
    renderItem: renderItem,
    ref: forwardedRef,
    size: size,
    value: value
  }, otherProps))), (0, _element.createElement)(_grid.Grid, {
    alignment: "bottom",
    templateColumns: subControlsTemplateColumns
  }, withNumberInput && (0, _element.createElement)(_components.FormGroup, {
    label: customLabel
  }, (0, _element.createElement)(_components.TextInput, {
    disabled: disabled,
    max: max,
    min: min,
    onChange: onInputChange,
    size: size,
    type: "number",
    value: inputValue
  })), (0, _element.createElement)(_view.View, null, (0, _element.createElement)(_components.Button, {
    disabled: isDefaultValue,
    isBlock: true,
    onClick: onReset
  }, (0, _i18n.__)('Reset')))));
}

var _default = (0, _context.contextConnect)(FontSizeControlSelect, 'FontSizeControlSelect');

exports.default = _default;
