// @wordpress/components/build/ui/font-size-control/slider.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _lodash = require("lodash");

var _components = require("@wp-g2/components");

var _i18n = require("@wordpress/i18n");

var _utils = require("./utils");

var _grid = require("../grid");

var _vStack = require("../v-stack");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function FontSizeControlSlider(props) {
  var _props$label = props.label,
      label = _props$label === void 0 ? (0, _i18n.__)('Custom size') : _props$label,
      disabled = props.disabled,
      min = props.min,
      max = props.max,
      _props$onChange = props.onChange,
      onChange = _props$onChange === void 0 ? _lodash.noop : _props$onChange,
      size = props.size,
      value = props.value,
      withSlider = props.withSlider;
  if (!withSlider) return null;
  var templateColumns = (0, _utils.getSliderTemplateColumns)();
  var controlProps = {
    disabled: disabled,
    min: min,
    max: max,
    onChange: onChange,
    size: size,
    value: value
  };
  return (0, _element.createElement)(_vStack.VStack, {
    spacing: 0
  }, (0, _element.createElement)(_components.ControlLabel, null, label), (0, _element.createElement)(_grid.Grid, {
    templateColumns: templateColumns
  }, (0, _element.createElement)(_components.Slider, controlProps), (0, _element.createElement)(_components.TextInput, (0, _extends2.default)({}, controlProps, {
    type: "number"
  }))));
}

var _default = FontSizeControlSlider;
exports.default = _default;
