// @wordpress/components/build/ui/font-size-control/use-font-size-control.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useFontSizeControl;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _utils = require("@wp-g2/utils");

var _element = require("@wordpress/element");

var styles = _interopRequireWildcard(require("./styles"));

var _utils2 = require("./utils");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function useFontSizeControl(props) {
  var _fontSizes$;

  var _useContextSystem = (0, _context.useContextSystem)(props, 'FontSizeControl'),
      disableCustomFontSizes = _useContextSystem.disableCustomFontSizes,
      _useContextSystem$fon = _useContextSystem.fontSizes,
      fontSizes = _useContextSystem$fon === void 0 ? [] : _useContextSystem$fon,
      _useContextSystem$onC = _useContextSystem.onChange,
      onChange = _useContextSystem$onC === void 0 ? _utils.noop : _useContextSystem$onC,
      value = _useContextSystem.value,
      _useContextSystem$wit = _useContextSystem.withSlider,
      withSlider = _useContextSystem$wit === void 0 ? false : _useContextSystem$wit,
      className = _useContextSystem.className,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["disableCustomFontSizes", "fontSizes", "onChange", "value", "withSlider", "className"]);

  var hasUnits = (0, _utils2.hasUnit)(value || ((_fontSizes$ = fontSizes[0]) === null || _fontSizes$ === void 0 ? void 0 : _fontSizes$.size));
  var hasCustomValue = (0, _utils2.isCustomValue)(fontSizes, value);
  var options = (0, _element.useMemo)(function () {
    return (0, _utils2.getSelectOptions)({
      options: fontSizes,
      disableCustomFontSizes: disableCustomFontSizes,
      hasCustomValue: hasCustomValue
    });
  }, [fontSizes, disableCustomFontSizes, hasCustomValue]);
  var handleOnReset = (0, _element.useCallback)(function () {
    onChange(undefined);
  }, [onChange]);
  var handleOnChange = (0, _element.useCallback)(function (_ref) {
    var selectedItem = _ref.selectedItem;
    if ((0, _utils2.isCustomSelectedItem)(selectedItem)) return;

    if (hasUnits) {
      onChange(selectedItem.size);
    } else if (_utils.is.defined(selectedItem.size)) {
      onChange(Number(selectedItem.size));
    } else {
      handleOnReset();
    }
  }, [handleOnReset, hasUnits, onChange]);
  var handleOnInputChange = (0, _element.useCallback)(function (next) {
    if (!next && next !== 0) {
      handleOnReset();
      return;
    }

    if (hasUnits) {
      onChange((0, _utils.createUnitValue)(next, 'px'));
    } else {
      onChange(Number(next));
    }
  }, [handleOnReset, hasUnits, onChange]);
  var inputValue = (0, _utils2.getInputValue)(fontSizes, value);
  var selectedFontSizeSlug = (0, _utils2.getSelectValueFromFontSize)(fontSizes, value);
  var currentValue = options.find(function (option) {
    return option.key === selectedFontSizeSlug;
  });
  var isDefaultValue = !_utils.is.defined(value, className);
  var classes = (0, _styles.cx)(styles.FontSizeControl);
  var withSelect = fontSizes.length > 0;
  var withNumberInput = !withSlider && !disableCustomFontSizes;
  return _objectSpread(_objectSpread({}, otherProps), {}, {
    className: classes,
    inputValue: inputValue,
    isDefaultValue: isDefaultValue,
    onChange: handleOnChange,
    onInputChange: handleOnInputChange,
    onReset: handleOnReset,
    options: options,
    value: currentValue,
    withNumberInput: withNumberInput,
    withSelect: withSelect,
    withSlider: withSlider
  });
}
