// @wp-g2/components/dist/cjs/BaseField/BaseField.styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.inline = exports.errorFocus = exports.error = exports.subtle = exports.focus = exports.clickable = exports.BaseField = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _styles = require("@wp-g2/styles");

function _templateObject7() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tdisplay: inline-flex;\n\tvertical-align: baseline;\n\twidth: auto;\n\n\t", "\n"]);

  _templateObject7 = function () {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tborder-color: ", ";\n\tbox-shadow: ", ";\n\n\t&:hover {\n\t\tborder-color: ", ";\n\t}\n"]);

  _templateObject6 = function () {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tborder-color: ", ";\n\n\t&:hover,\n\t&:active {\n\t\tborder-color: ", ";\n\t}\n\n\t&:focus,\n\t&[data-focused='true'] {\n\t\tborder-color: ", ";\n\t\tbox-shadow: ", ";\n\t}\n"]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbackground-color: transparent;\n\n\t&:hover,\n\t&:active,\n\t&:focus,\n\t&[data-focused='true'] {\n\t\t", ";\n\t}\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tborder-color: ", ";\n\tbox-shadow: ", ";\n\n\t&:hover {\n\t\tborder-color: ", ";\n\t}\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tcursor: pointer;\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\t", ";\n\tborder: 1px solid;\n\tborder-color: ", ";\n\tbox-shadow: ", ";\n\tdisplay: flex;\n\tflex: 1;\n\tfont-family: ", ";\n\tfont-size: ", ";\n\toutline: none;\n\tpadding: 0 8px;\n\tposition: relative;\n\ttransition: border-color ", " ease;\n\twidth: 100%;\n\n\t&[disabled] {\n\t\topacity: 0.6;\n\t}\n\n\t&:hover {\n\t\tborder-color: ", ";\n\t}\n\n\t&:focus,\n\t&[data-focused='true'] {\n\t\tborder-color: ", ";\n\t\tbox-shadow: ", ";\n\t}\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var BaseField = (0, _styles.css)(_templateObject(), _styles.ui.background.control, _styles.ui.borderRadius.round, _styles.ui.get('controlBorderColor'), _styles.ui.get('controlBoxShadow'), _styles.ui.get('fontFamily'), _styles.ui.get('fontSize'), _styles.ui.get('transitionDurationFastest'), _styles.ui.get('controlBorderColorHover'), _styles.ui.color.admin, _styles.ui.get('controlBoxShadowFocus'));
exports.BaseField = BaseField;
var clickable = (0, _styles.css)(_templateObject2());
exports.clickable = clickable;
var focus = (0, _styles.css)(_templateObject3(), _styles.ui.color.admin, _styles.ui.get('controlBoxShadowFocus'), _styles.ui.color.admin);
exports.focus = focus;
var subtle = (0, _styles.css)(_templateObject4(), _styles.ui.background.control);
exports.subtle = subtle;
var error = (0, _styles.css)(_templateObject5(), _styles.ui.get('controlDestructiveBorderColor'), _styles.ui.get('controlDestructiveBorderColor'), _styles.ui.get('controlDestructiveBorderColorFocus'), _styles.ui.get('controlDestructiveBoxShadowFocus'));
exports.error = error;
var errorFocus = (0, _styles.css)(_templateObject6(), _styles.ui.get('controlDestructiveBorderColorFocus'), _styles.ui.get('controlDestructiveBoxShadowFocus'), _styles.ui.get('controlDestructiveBorderColorFocus'));
exports.errorFocus = errorFocus;
var inline = (0, _styles.css)(_templateObject7(), _styles.ui.browsers.safari("\n\t\t\tvertical-align: middle;\n\t"));
exports.inline = inline;
