// @wp-g2/components/dist/cjs/Navigator/__stories__/Navigator.Sidebar.stories.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Sidebar = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _icons = require("@wordpress/icons");

var _protokit = require("@wp-g2/protokit");

var _styles = require("@wp-g2/styles");

var _utils = require("@wp-g2/utils");

var _lodash = require("lodash");

var _react = _interopRequireWildcard(require("react"));

var _reactRouterDom = require("react-router-dom");

var _index = require("../../index");

var _index2 = require("../index");

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbottom: 0;\n\tleft: 0;\n\tright: 0;\n\ttop: 0;\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var _default = {
  component: _index2.Navigator,
  title: 'Components/Navigator'
};
exports.default = _default;
var ScreenView = (0, _styles.styled)(_index.Surface)(_templateObject());

var Screen = function (_ref) {
  var children = _ref.children;
  return /*#__PURE__*/_react.default.createElement(ScreenView, null, children);
};

var MenuLink = function (_ref2) {
  var children = _ref2.children,
      props = (0, _objectWithoutProperties2.default)(_ref2, ["children"]);
  return /*#__PURE__*/_react.default.createElement(_index2.NavigatorLink, props, /*#__PURE__*/_react.default.createElement(_index.MenuItem, null, children));
};

var ExternalMenuLink = function (props) {
  return /*#__PURE__*/_react.default.createElement(_index.MenuItem, Object.assign({
    activeClassName: "is-active",
    as: _reactRouterDom.NavLink,
    exact: true
  }, props));
};

var DashboardNav = function () {
  var _useState = (0, _react.useState)(null),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      commentCount = _useState2[0],
      setCommentCount = _useState2[1];

  (0, _react.useEffect)(function () {
    setTimeout(function () {
      setCommentCount(3);
    }, 1000);
  }, []);
  return /*#__PURE__*/_react.default.createElement(Screen, null, /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    py: 4
  }), /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    mb: 8
  }, /*#__PURE__*/_react.default.createElement(MenuLink, null, "My sites")), /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    mb: 4,
    pl: 3
  }, /*#__PURE__*/_react.default.createElement(_index.Heading, {
    size: 4
  }, "My site name")), /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    mb: 4
  }, /*#__PURE__*/_react.default.createElement(_index.Menu, null, /*#__PURE__*/_react.default.createElement(ExternalMenuLink, {
    to: "/"
  }, "Dashboard"), /*#__PURE__*/_react.default.createElement(MenuLink, {
    to: "Pages"
  }, "Pages"), /*#__PURE__*/_react.default.createElement(MenuLink, {
    to: "Posts"
  }, "Posts"), /*#__PURE__*/_react.default.createElement(ExternalMenuLink, {
    to: "/comments"
  }, /*#__PURE__*/_react.default.createElement(_index.Flex, null, /*#__PURE__*/_react.default.createElement(_index.View, null, /*#__PURE__*/_react.default.createElement(_index.Text, null, "Comments")), /*#__PURE__*/_react.default.createElement(_index.View, {
    css: {
      margin: '-3px 0'
    }
  }, commentCount === null ? /*#__PURE__*/_react.default.createElement(_index.Spinner, null) : /*#__PURE__*/_react.default.createElement(_index.Badge, null, commentCount)))), /*#__PURE__*/_react.default.createElement(ExternalMenuLink, {
    to: "/media"
  }, "Media"), /*#__PURE__*/_react.default.createElement(ExternalMenuLink, {
    to: "/users"
  }, "Users"))), /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    mb: 4,
    mt: 12,
    pl: 3
  }, /*#__PURE__*/_react.default.createElement(_index.Heading, {
    size: 4
  }, "Plugins")), /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    mb: 4
  }, /*#__PURE__*/_react.default.createElement(_index.Menu, null, /*#__PURE__*/_react.default.createElement(ExternalMenuLink, {
    to: "/plugins/hello-dolly"
  }, "Hello Dolly"), /*#__PURE__*/_react.default.createElement(MenuLink, {
    to: "PluginsWooCommerce"
  }, "WooCommerce"), /*#__PURE__*/_react.default.createElement(ExternalMenuLink, {
    to: "/plugins"
  }, "All plugins"))));
};

var pageSchema = new _protokit.Schema(function () {
  return {
    title: _protokit.faker.commerce.productName()
  };
});
var pages = pageSchema.make(100);

var PagesNav = function () {
  var _chunk = (0, _lodash.chunk)(pages, 5),
      _chunk2 = (0, _slicedToArray2.default)(_chunk, 2),
      recentPages = _chunk2[0],
      draftPages = _chunk2[1];

  var _chunk3 = (0, _lodash.chunk)(draftPages, 2),
      _chunk4 = (0, _slicedToArray2.default)(_chunk3, 1),
      drafts = _chunk4[0];

  var _useState3 = (0, _react.useState)(''),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      query = _useState4[0],
      setQuery = _useState4[1];

  var isSearching = !!query;
  var searchResults = pages.filter(function (page) {
    return page.title.toLowerCase().includes(query.toLowerCase());
  });

  var _chunk5 = (0, _lodash.chunk)(searchResults, 10),
      _chunk6 = (0, _slicedToArray2.default)(_chunk5, 1),
      _results = _chunk6[0];

  var results = _results || [];
  return /*#__PURE__*/_react.default.createElement(Screen, null, /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    py: 4
  }), /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    mb: 8
  }, /*#__PURE__*/_react.default.createElement(MenuLink, {
    isBack: true,
    to: "Dashboard"
  }, "Dashboard")), /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    mb: 4,
    pl: 3
  }, /*#__PURE__*/_react.default.createElement(_index.Heading, {
    size: 4
  }, "Pages")), /*#__PURE__*/_react.default.createElement(_index.View, {
    css: [_styles.ui.padding.x(3), _styles.ui.margin.bottom(4)]
  }, /*#__PURE__*/_react.default.createElement(_index.SearchInput, {
    onChange: function onChange(next) {
      return setQuery(next);
    },
    placeholder: "Search pages",
    value: query
  })), /*#__PURE__*/_react.default.createElement(_index.VStack, {
    spacing: 10
  }, isSearching && results.length && /*#__PURE__*/_react.default.createElement(_index.View, null, results.map(function (page) {
    return /*#__PURE__*/_react.default.createElement(ExternalMenuLink, {
      key: page.id,
      to: "/pages/" + page.id
    }, /*#__PURE__*/_react.default.createElement(_index.Truncate, null, page.title));
  })), isSearching && !results.length && /*#__PURE__*/_react.default.createElement(_index.View, {
    css: [_styles.ui.padding.x(3)]
  }, /*#__PURE__*/_react.default.createElement(_index.Text, null, "No Results")), !isSearching && /*#__PURE__*/_react.default.createElement(_index.VStack, {
    spacing: 10
  }, /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.View, {
    css: [_styles.ui.padding.x(3)]
  }, /*#__PURE__*/_react.default.createElement(_index.Subheading, null, "Recent Pages")), /*#__PURE__*/_react.default.createElement(_index.View, null, recentPages.map(function (page) {
    return /*#__PURE__*/_react.default.createElement(ExternalMenuLink, {
      key: page.id,
      to: "/pages/" + page.id
    }, /*#__PURE__*/_react.default.createElement(_index.Truncate, null, page.title));
  }))), /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.View, {
    css: [_styles.ui.padding.x(3)]
  }, /*#__PURE__*/_react.default.createElement(_index.Subheading, null, "Drafts")), /*#__PURE__*/_react.default.createElement(_index.View, null, drafts.map(function (page) {
    return /*#__PURE__*/_react.default.createElement(ExternalMenuLink, {
      key: page.id,
      to: "/pages/" + page.id
    }, /*#__PURE__*/_react.default.createElement(_index.Truncate, null, page.title));
  }))))));
};

var PostsNav = function () {
  return /*#__PURE__*/_react.default.createElement(Screen, null, /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    py: 4
  }), /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    mb: 8
  }, /*#__PURE__*/_react.default.createElement(MenuLink, {
    isBack: true,
    to: "Dashboard"
  }, "Dashboard")), /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    mb: 4,
    pl: 3
  }, /*#__PURE__*/_react.default.createElement(_index.Heading, {
    size: 4
  }, "Posts")), /*#__PURE__*/_react.default.createElement(_index.Menu, null, /*#__PURE__*/_react.default.createElement(ExternalMenuLink, {
    to: "/posts"
  }, "All posts"), /*#__PURE__*/_react.default.createElement(ExternalMenuLink, {
    to: "/posts/categories"
  }, "Categories"), /*#__PURE__*/_react.default.createElement(ExternalMenuLink, {
    to: "/posts/tags"
  }, "Tags"), /*#__PURE__*/_react.default.createElement(ExternalMenuLink, {
    to: "/posts/new"
  }, "New page")));
};

var PluginsWooCommerceNav = function () {
  var _useLocalState = (0, _utils.useLocalState)('@wp-g2/components/Navigator/Sidebar/PluginsWooCommerceNav', false),
      _useLocalState2 = (0, _slicedToArray2.default)(_useLocalState, 2),
      isFavorite = _useLocalState2[0],
      setIsFavorite = _useLocalState2[1];

  return /*#__PURE__*/_react.default.createElement(Screen, null, /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    py: 4
  }), /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    mb: 8
  }, /*#__PURE__*/_react.default.createElement(MenuLink, {
    isBack: true,
    to: "Dashboard"
  }, "Dashboard")), /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    mb: 4,
    pl: 3
  }, /*#__PURE__*/_react.default.createElement(_index.HStack, null, /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.Heading, {
    size: 4
  }, "WooCommerce")), /*#__PURE__*/_react.default.createElement(_index.Text, {
    css: [_styles.ui.padding.y(3), _styles.ui.padding.x(2), {
      cursor: 'pointer',
      color: isFavorite ? 'yellow' : null
    }, _styles.ui.animation.default],
    isBlock: true,
    onClick: function onClick() {
      return setIsFavorite(!isFavorite);
    },
    variant: !isFavorite && 'muted'
  }, "\u2605"))), /*#__PURE__*/_react.default.createElement(_index.Menu, null, /*#__PURE__*/_react.default.createElement(ExternalMenuLink, {
    to: "/plugins/woo-commerce/store"
  }, "Store home"), /*#__PURE__*/_react.default.createElement(MenuLink, {
    showArrow: true
  }, "Analytics"), /*#__PURE__*/_react.default.createElement(MenuLink, {
    showArrow: true
  }, "Orders"), /*#__PURE__*/_react.default.createElement(MenuLink, {
    showArrow: true
  }, "Marketing"), /*#__PURE__*/_react.default.createElement(MenuLink, {
    showArrow: true
  }, "Products"), /*#__PURE__*/_react.default.createElement(MenuLink, {
    showArrow: true
  }, "Customers"), /*#__PURE__*/_react.default.createElement(MenuLink, {
    showArrow: true
  }, "Settings"), /*#__PURE__*/_react.default.createElement(MenuLink, {
    showArrow: true
  }, "Extensions"), /*#__PURE__*/_react.default.createElement(MenuLink, {
    showArrow: true
  }, "Tools")));
};

var screens = [{
  component: DashboardNav,
  path: 'Dashboard'
}, {
  component: PagesNav,
  path: 'Pages'
}, {
  component: PostsNav,
  path: 'Posts'
}, {
  component: PluginsWooCommerceNav,
  path: 'PluginsWooCommerce'
}];

var NavigationSidebar = function () {
  return /*#__PURE__*/_react.default.createElement(_styles.ThemeProvider, {
    isDark: true,
    isGlobal: false
  }, /*#__PURE__*/_react.default.createElement(_index.Surface, {
    borderRight: "1px solid",
    css: {
      height: '100%'
    }
  }, /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index2.Navigator, {
    initialPath: 'Dashboard'
  }, /*#__PURE__*/_react.default.createElement(_index.Scrollable, {
    css: {
      height: "calc(100vh - 100px - 40px)",
      padding: 8
    }
  }, /*#__PURE__*/_react.default.createElement(_index.HStack, {
    css: [_styles.ui.padding(3)],
    spacing: 3
  }, /*#__PURE__*/_react.default.createElement(_index.Icon, {
    icon: _icons.wordpress,
    size: 30
  }), /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.Heading, {
    size: 5,
    truncate: true
  }, "My Site Name"))), /*#__PURE__*/_react.default.createElement(_index2.NavigatorScreens, null, screens.map(function (screen) {
    return /*#__PURE__*/_react.default.createElement(_index2.NavigatorScreen, Object.assign({}, screen, {
      key: screen.path
    }));
  }))))), /*#__PURE__*/_react.default.createElement(_index.Surface, {
    css: [_styles.ui.padding.y(4), _styles.ui.padding.x(3)]
  }, /*#__PURE__*/_react.default.createElement(_index.HStack, {
    spacing: 4
  }, /*#__PURE__*/_react.default.createElement(_index.Avatar, {
    name: "Elsa Oldenburg",
    size: "medium",
    src: "https://picsum.photos/id/1041/300/300"
  }), /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.Text, {
    weight: "bold"
  }, "Elsa")))))));
};

var MockPage = function (_ref3) {
  var match = _ref3.match;
  var params = match.params,
      path = match.path;
  var title;

  if (path === '/pages/:id' && (params == null ? void 0 : params.id)) {
    var page = pages.find(function (page) {
      return page.id === (params == null ? void 0 : params.id);
    });
    title = page.title;
  } else {
    var route = routes.find(function (item) {
      return item.path === path;
    });
    title = route.title;
  }

  return /*#__PURE__*/_react.default.createElement(_index.Surface, {
    borderBottom: true
  }, /*#__PURE__*/_react.default.createElement(_index.View, {
    css: {
      padding: '20px 40px'
    }
  }, /*#__PURE__*/_react.default.createElement(_index.Heading, null, title)));
};

var routes = [{
  path: '/',
  title: 'Dashboard'
}, {
  path: '/pages',
  title: 'All pages'
}, {
  path: '/pages/new',
  title: 'New page'
}, {
  path: '/pages/:id',
  title: 'Single Page'
}, {
  path: '/posts',
  title: 'All posts'
}, {
  path: '/posts/categories',
  title: 'Categories'
}, {
  path: '/posts/tags',
  title: 'Tags'
}, {
  path: '/posts/new',
  title: 'New post'
}, {
  path: '/media',
  title: 'Media'
}, {
  path: '/comments',
  title: 'Comments'
}, {
  path: '/users',
  title: 'Users'
}, {
  path: '/plugins',
  title: 'All plugins'
}, {
  path: '/plugins/hello-dolly',
  title: 'Hello Dolly'
}, {
  path: '/plugins/woo-commerce/store',
  title: 'Store home'
}];

var BrowserBar = function (_ref4) {
  var _history$location;

  var history = _ref4.history;
  var pathname = (history == null ? void 0 : (_history$location = history.location) == null ? void 0 : _history$location.pathname) || '/';
  return /*#__PURE__*/_react.default.createElement(_index.Background, {
    borderBottom: true,
    css: {
      padding: 4
    }
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    css: {
      textAlign: 'center'
    },
    readOnly: true,
    value: "http://mysite.com/wp-admin" + pathname
  }));
};

var App = function () {
  return /*#__PURE__*/_react.default.createElement(_index.View, {
    css: {
      border: "2px solid " + _styles.ui.get('surfaceBorderColor'),
      bottom: 100,
      left: 0,
      position: 'fixed',
      right: 0,
      top: 0
    }
  }, /*#__PURE__*/_react.default.createElement(_reactRouterDom.MemoryRouter, {
    initialEntries: ['/']
  }, /*#__PURE__*/_react.default.createElement(_index.VStack, {
    spacing: 0
  }, /*#__PURE__*/_react.default.createElement(_reactRouterDom.Route, {
    component: BrowserBar
  }), /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.Grid, {
    css: {
      height: "100%"
    },
    gap: 0,
    templateColumns: "220px 1fr"
  }, /*#__PURE__*/_react.default.createElement(NavigationSidebar, null), /*#__PURE__*/_react.default.createElement(_index.Background, null, /*#__PURE__*/_react.default.createElement(_reactRouterDom.Switch, null, routes.map(function (route) {
    return /*#__PURE__*/_react.default.createElement(_reactRouterDom.Route, Object.assign({
      exact: true
    }, route, {
      component: MockPage,
      key: route.path
    }));
  }))))))));
};

var Sidebar = function () {
  return /*#__PURE__*/_react.default.createElement(App, null);
};

exports.Sidebar = Sidebar;
