// @wp-g2/components/dist/cjs/Tag/TagRemoveButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _styles = require("@wp-g2/styles");

var _utils = require("@wp-g2/utils");

var _react = _interopRequireDefault(require("react"));

var _CloseButton = require("../CloseButton");

var styles = _interopRequireWildcard(require("./Tag.styles"));

var RemoveButtonView = styles.RemoveButtonView;

function TagRemoveButton(_ref) {
  var _ref$onClick = _ref.onClick,
      onClick = _ref$onClick === void 0 ? _utils.noop : _ref$onClick,
      removeButtonText = _ref.removeButtonText;
  if (!removeButtonText) return null;
  return /*#__PURE__*/_react.default.createElement(RemoveButtonView, _styles.ui.$('TagRemoveButtonWrapper'), /*#__PURE__*/_react.default.createElement(_CloseButton.CloseButton, Object.assign({
    "aria-label": removeButtonText,
    currentColor: true,
    iconSize: 12,
    onClick: onClick,
    size: "xSmall",
    title: removeButtonText,
    variant: "tertiary"
  }, _styles.ui.$('TagRemoveButton'))));
}

var _default = /*#__PURE__*/_react.default.memo(TagRemoveButton);

exports.default = _default;
