// @wordpress/components/build/unit-control/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _lodash = require("lodash");

var _classnames = _interopRequireDefault(require("classnames"));

var _i18n = require("@wordpress/i18n");

var _keycodes = require("@wordpress/keycodes");

var _state = require("../input-control/state");

var _unitControlStyles = require("./styles/unit-control-styles");

var _unitSelectControl = _interopRequireDefault(require("./unit-select-control"));

var _utils = require("./utils");

var _hooks = require("../utils/hooks");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function UnitControl(_ref, ref) {
  var _ref$__unstableStateR = _ref.__unstableStateReducer,
      stateReducer = _ref$__unstableStateR === void 0 ? function (state) {
    return state;
  } : _ref$__unstableStateR,
      _ref$autoComplete = _ref.autoComplete,
      autoComplete = _ref$autoComplete === void 0 ? 'off' : _ref$autoComplete,
      className = _ref.className,
      _ref$disabled = _ref.disabled,
      disabled = _ref$disabled === void 0 ? false : _ref$disabled,
      _ref$disableUnits = _ref.disableUnits,
      disableUnits = _ref$disableUnits === void 0 ? false : _ref$disableUnits,
      _ref$isPressEnterToCh = _ref.isPressEnterToChange,
      isPressEnterToChange = _ref$isPressEnterToCh === void 0 ? false : _ref$isPressEnterToCh,
      _ref$isResetValueOnUn = _ref.isResetValueOnUnitChange,
      isResetValueOnUnitChange = _ref$isResetValueOnUn === void 0 ? false : _ref$isResetValueOnUn,
      _ref$isUnitSelectTabb = _ref.isUnitSelectTabbable,
      isUnitSelectTabbable = _ref$isUnitSelectTabb === void 0 ? true : _ref$isUnitSelectTabb,
      label = _ref.label,
      _ref$onChange = _ref.onChange,
      onChange = _ref$onChange === void 0 ? _lodash.noop : _ref$onChange,
      _ref$onUnitChange = _ref.onUnitChange,
      onUnitChange = _ref$onUnitChange === void 0 ? _lodash.noop : _ref$onUnitChange,
      _ref$size = _ref.size,
      size = _ref$size === void 0 ? 'default' : _ref$size,
      style = _ref.style,
      unitProp = _ref.unit,
      _ref$units = _ref.units,
      units = _ref$units === void 0 ? _utils.CSS_UNITS : _ref$units,
      valueProp = _ref.value,
      props = (0, _objectWithoutProperties2.default)(_ref, ["__unstableStateReducer", "autoComplete", "className", "disabled", "disableUnits", "isPressEnterToChange", "isResetValueOnUnitChange", "isUnitSelectTabbable", "label", "onChange", "onUnitChange", "size", "style", "unit", "units", "value"]);

  var _getParsedValue = (0, _utils.getParsedValue)(valueProp, unitProp, units),
      _getParsedValue2 = (0, _slicedToArray2.default)(_getParsedValue, 2),
      value = _getParsedValue2[0],
      initialUnit = _getParsedValue2[1];

  var _useControlledState = (0, _hooks.useControlledState)(unitProp, {
    initial: initialUnit
  }),
      _useControlledState2 = (0, _slicedToArray2.default)(_useControlledState, 2),
      unit = _useControlledState2[0],
      setUnit = _useControlledState2[1]; // Stores parsed value for hand-off in state reducer


  var refParsedValue = (0, _element.useRef)(null);
  var classes = (0, _classnames.default)('components-unit-control', className);

  var handleOnChange = function handleOnChange(next, changeProps) {
    if (next === '') {
      onChange('', changeProps);
      return;
    }
    /*
     * Customizing the onChange callback.
     * This allows as to broadcast a combined value+unit to onChange.
     */


    next = (0, _utils.getValidParsedUnit)(next, units, value, unit).join('');
    onChange(next, changeProps);
  };

  var handleOnUnitChange = function handleOnUnitChange(next, changeProps) {
    var data = changeProps.data;
    var nextValue = "".concat(value).concat(next);

    if (isResetValueOnUnitChange && (data === null || data === void 0 ? void 0 : data.default) !== undefined) {
      nextValue = "".concat(data.default).concat(next);
    }

    onChange(nextValue, changeProps);
    onUnitChange(next, changeProps);
    setUnit(next);
  };

  var mayUpdateUnit = function mayUpdateUnit(event) {
    if (!isNaN(event.target.value)) {
      refParsedValue.current = null;
      return;
    }

    var _getValidParsedUnit = (0, _utils.getValidParsedUnit)(event.target.value, units, value, unit),
        _getValidParsedUnit2 = (0, _slicedToArray2.default)(_getValidParsedUnit, 2),
        parsedValue = _getValidParsedUnit2[0],
        parsedUnit = _getValidParsedUnit2[1];

    refParsedValue.current = parsedValue;

    if (isPressEnterToChange && parsedUnit !== unit) {
      var data = units.find(function (option) {
        return option.value === parsedUnit;
      });
      var changeProps = {
        event: event,
        data: data
      };
      onChange("".concat(parsedValue).concat(parsedUnit), changeProps);
      onUnitChange(parsedUnit, changeProps);
      setUnit(parsedUnit);
    }
  };

  var handleOnBlur = mayUpdateUnit;

  var handleOnKeyDown = function handleOnKeyDown(event) {
    var keyCode = event.keyCode;

    if (keyCode === _keycodes.ENTER) {
      mayUpdateUnit(event);
    }
  };
  /**
   * "Middleware" function that intercepts updates from InputControl.
   * This allows us to tap into actions to transform the (next) state for
   * InputControl.
   *
   * @param {Object} state State from InputControl
   * @param {Object} action Action triggering state change
   * @return {Object} The updated state to apply to InputControl
   */


  var unitControlStateReducer = function unitControlStateReducer(state, action) {
    /*
     * On commits (when pressing ENTER and on blur if
     * isPressEnterToChange is true), if a parse has been performed
     * then use that result to update the state.
     */
    if (action.type === _state.inputControlActionTypes.COMMIT) {
      if (refParsedValue.current !== null) {
        state.value = refParsedValue.current;
        refParsedValue.current = null;
      }
    }

    return state;
  };

  var inputSuffix = !disableUnits ? (0, _element.createElement)(_unitSelectControl.default, {
    "aria-label": (0, _i18n.__)('Select unit'),
    disabled: disabled,
    isTabbable: isUnitSelectTabbable,
    options: units,
    onChange: handleOnUnitChange,
    size: size,
    value: unit
  }) : null;
  return (0, _element.createElement)(_unitControlStyles.Root, {
    className: "components-unit-control-wrapper",
    style: style
  }, (0, _element.createElement)(_unitControlStyles.ValueInput, (0, _extends2.default)({
    "aria-label": label,
    type: isPressEnterToChange ? 'text' : 'number'
  }, (0, _lodash.omit)(props, ['children']), {
    autoComplete: autoComplete,
    className: classes,
    disabled: disabled,
    disableUnits: disableUnits,
    isPressEnterToChange: isPressEnterToChange,
    label: label,
    onBlur: handleOnBlur,
    onKeyDown: handleOnKeyDown,
    onChange: handleOnChange,
    ref: ref,
    size: size,
    suffix: inputSuffix,
    value: value,
    __unstableStateReducer: (0, _state.composeStateReducers)(unitControlStateReducer, stateReducer)
  })));
}

var ForwardedUnitControl = (0, _element.forwardRef)(UnitControl);
var _default = ForwardedUnitControl;
exports.default = _default;
