// @wordpress/components/build-module/ui/elevation/hook.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _taggedTemplateLiteral from "@babel/runtime/helpers/esm/taggedTemplateLiteral";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";

var _templateObject, _templateObject2, _templateObject3;

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * External dependencies
 */
import { useContextSystem } from '@wp-g2/context';
import { css, cx, getBoxShadow, ui } from '@wp-g2/styles';
import { isNil } from 'lodash';
/**
 * WordPress dependencies
 */

import { useMemo } from '@wordpress/element';
/**
 * Internal dependencies
 */

import * as styles from './styles';
/**
 * @param {import('@wp-g2/create-styles').ViewOwnProps<import('./types').Props, 'div'>} props
 */

export function useElevation(props) {
  var _useContextSystem = useContextSystem(props, 'Elevation'),
      active = _useContextSystem.active,
      _useContextSystem$bor = _useContextSystem.borderRadius,
      borderRadius = _useContextSystem$bor === void 0 ? 'inherit' : _useContextSystem$bor,
      className = _useContextSystem.className,
      focus = _useContextSystem.focus,
      hover = _useContextSystem.hover,
      _useContextSystem$isI = _useContextSystem.isInteractive,
      isInteractive = _useContextSystem$isI === void 0 ? false : _useContextSystem$isI,
      _useContextSystem$off = _useContextSystem.offset,
      offset = _useContextSystem$off === void 0 ? 0 : _useContextSystem$off,
      _useContextSystem$val = _useContextSystem.value,
      value = _useContextSystem$val === void 0 ? 0 : _useContextSystem$val,
      otherProps = _objectWithoutProperties(_useContextSystem, ["active", "borderRadius", "className", "focus", "hover", "isInteractive", "offset", "value"]);

  var classes = useMemo(function () {
    /** @type {number | undefined} */
    var hoverValue = !isNil(hover) ? hover : value * 2;
    /** @type {number | undefined} */

    var activeValue = !isNil(active) ? active : value / 2;

    if (!isInteractive) {
      hoverValue = !isNil(hover) ? hover : undefined;
      activeValue = !isNil(active) ? active : undefined;
    }

    var transition = "box-shadow ".concat(ui.get('transitionDuration'), " ").concat(ui.get('transitionTimingFunction'));
    var sx = {};
    sx.Base = css({
      borderRadius: borderRadius,
      bottom: offset,
      boxShadow: getBoxShadow(value),
      opacity: ui.get('elevationIntensity'),
      left: offset,
      right: offset,
      top: offset,
      transition: transition
    });

    if (!isNil(hoverValue)) {
      sx.hover = css(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n\t\t\t\t*:hover > & {\n\t\t\t\t\tbox-shadow: ", ";\n\t\t\t\t}\n\t\t\t"])), getBoxShadow(hoverValue));
    }

    if (!isNil(activeValue)) {
      sx.active = css(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n\t\t\t\t*:active > & {\n\t\t\t\t\tbox-shadow: ", ";\n\t\t\t\t}\n\t\t\t"])), getBoxShadow(activeValue));
    }

    if (!isNil(focus)) {
      sx.focus = css(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n\t\t\t\t*:focus > & {\n\t\t\t\t\tbox-shadow: ", ";\n\t\t\t\t}\n\t\t\t"])), getBoxShadow(focus));
    }

    return cx(styles.Elevation, sx.Base, sx.hover && sx.hover, sx.focus && sx.focus, sx.active && sx.active, className);
  }, [active, borderRadius, className, focus, hover, isInteractive, offset, value]);
  return _objectSpread(_objectSpread({}, otherProps), {}, {
    className: classes,
    'aria-hidden': true
  });
}
