// @wp-g2/components/dist/cjs/Image/Image.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _AspectRatio = require("../AspectRatio");

var _Image = require("./Image.styles");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function Image(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'Image'),
      aspectRatio = _useContextSystem.aspectRatio,
      className = _useContextSystem.className,
      fit = _useContextSystem.fit,
      height = _useContextSystem.height,
      width = _useContextSystem.width,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["aspectRatio", "className", "fit", "height", "width"]);

  var sx = {};
  var aspectFit = aspectRatio && !fit ? 'cover' : fit;
  var preferredFit = aspectRatio ? aspectFit : fit;
  sx.base = (0, _styles.css)({
    width: width,
    height: height
  });
  sx.fit = (0, _styles.css)({
    objectFit: preferredFit
  });
  sx.fitSize = (0, _styles.css)({
    height: '100%',
    width: '100%'
  });
  var classes = (0, _styles.cx)(sx.base, sx.fit, fit && sx.fitSize, className);

  var imageProps = _objectSpread(_objectSpread({}, otherProps), {}, {
    className: classes,
    height: height,
    width: width,
    ref: forwardedRef
  });

  if (aspectRatio) {
    return /*#__PURE__*/_react.default.createElement(_AspectRatio.AspectRatio, {
      ratio: aspectRatio,
      width: width
    }, /*#__PURE__*/_react.default.createElement(_Image.ImageView, imageProps));
  }

  return /*#__PURE__*/_react.default.createElement(_Image.ImageView, imageProps);
}

var _default = (0, _context.contextConnect)(Image, 'Image');

exports.default = _default;
