// @wp-g2/components/dist/cjs/ZStack/__stories__/ZStack.stories.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _index = require("../../index");

var _index2 = require("../index");

var _default2 = {
  component: _index2.ZStack,
  title: 'Components/ZStack'
};
exports.default = _default2;

var Avatar = function () {
  return /*#__PURE__*/_react.default.createElement(_index.Animated, {
    layout: true
  }, /*#__PURE__*/_react.default.createElement(_index.Placeholder, {
    css: "\n\t\t\t\t\tborder: 3px solid black;\n\t\t\t\t\tborder-radius: 9999px;\n\t\t\t\t",
    height: 48,
    width: 48
  }), /*#__PURE__*/_react.default.createElement(_index.Elevation, {
    borderRadius: 9999,
    isInteractive: false,
    value: 3
  }));
};

var AnimatedAvatars = function () {
  var _useState = (0, _react.useState)(false),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      isHover = _useState2[0],
      setIsHover = _useState2[1];

  var offset = isHover ? 0 : 20;
  return /*#__PURE__*/_react.default.createElement(_index.HStack, null, /*#__PURE__*/_react.default.createElement(_index.Animated, {
    onHoverEnd: function onHoverEnd() {
      return setIsHover(false);
    },
    onHoverStart: function onHoverStart() {
      return setIsHover(true);
    }
  }, /*#__PURE__*/_react.default.createElement(_index2.ZStack, {
    isLayered: false,
    offset: offset
  }, /*#__PURE__*/_react.default.createElement(Avatar, null), /*#__PURE__*/_react.default.createElement(Avatar, null), /*#__PURE__*/_react.default.createElement(Avatar, null), /*#__PURE__*/_react.default.createElement(Avatar, null))));
};

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(_index.View, null, /*#__PURE__*/_react.default.createElement(AnimatedAvatars, null));
};

exports._default = _default;
