// @wp-g2/components/dist/es/__stories__/__References/Backgrounds.stories.js
import _taggedTemplateLiteralLoose from "@babel/runtime/helpers/esm/taggedTemplateLiteralLoose";

function _templateObject4() {
  var data = _taggedTemplateLiteralLoose(["\n\t\t\t", "\n\t\t"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = _taggedTemplateLiteralLoose(["\n\t\t\t", "\n\t\t"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteralLoose(["\n\t\t\t", "\n\t\t"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteralLoose(["\n\t\t\t", "\n\t\t"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

import { css, cx, getBackgroundColor, getBackgroundColorText, SUPPORTED_COLORS as COLORS, ThemeProvider } from '@wp-g2/styles';
import React from 'react';
import { Card, CardBody, Grid, Heading, Spacer, Surface, View } from '../../index';
export default {
  title: 'References/Background'
};

var ColorSet = function (_ref) {
  var color = _ref.color;
  var classes = cx(css(_templateObject(), getBackgroundColor(color)), css(_templateObject2(), getBackgroundColorText(color)));
  return /*#__PURE__*/React.createElement(Card, null, /*#__PURE__*/React.createElement(CardBody, {
    className: classes
  }, color));
};

var ColorSetBold = function (_ref2) {
  var color = _ref2.color;
  var classes = cx(css(_templateObject3(), getBackgroundColor(color, {
    isBold: true
  })), css(_templateObject4(), getBackgroundColorText(color, {
    isBold: true
  })));
  return /*#__PURE__*/React.createElement(Card, null, /*#__PURE__*/React.createElement(CardBody, {
    className: classes
  }, color));
};

var ColorCollection = function (_ref3) {
  var title = _ref3.title;
  return /*#__PURE__*/React.createElement(Spacer, {
    css: {
      position: 'relative'
    },
    mb: 8,
    p: 5
  }, /*#__PURE__*/React.createElement(View, {
    css: {
      margin: 'auto',
      maxWidth: 1200,
      position: 'relative',
      zIndex: 1
    }
  }, /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(Heading, {
    size: 2
  }, title)), /*#__PURE__*/React.createElement(Spacer, {
    mb: 8
  }, /*#__PURE__*/React.createElement(Grid, {
    columns: [1, 2, 4]
  }, COLORS.map(function (color) {
    return /*#__PURE__*/React.createElement(ColorSet, {
      color: color,
      key: color
    });
  }))), /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(Heading, {
    size: 4
  }, "Bold")), /*#__PURE__*/React.createElement(Spacer, {
    mb: 8
  }, /*#__PURE__*/React.createElement(Grid, {
    columns: [1, 2, 4]
  }, COLORS.map(function (color) {
    return /*#__PURE__*/React.createElement(ColorSetBold, {
      color: color,
      key: color
    });
  })))), /*#__PURE__*/React.createElement(Surface, {
    css: {
      bottom: 0,
      left: 0,
      position: 'absolute',
      right: 0,
      top: 0,
      zIndex: 0
    }
  }));
};

export var _default = function () {
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(ColorCollection, {
    title: "Backgrounds"
  }), /*#__PURE__*/React.createElement(ThemeProvider, {
    isDark: true,
    isGlobal: false
  }, /*#__PURE__*/React.createElement(ColorCollection, {
    title: "Backgrounds + Dark"
  })));
};
