// @wp-g2/components/dist/es/__stories__/__References/Colors.stories.js
import { get, SUPPORTED_COLORS as COLORS, ThemeProvider } from '@wp-g2/styles';
import React from 'react';
import { Card, CardBody, Flex, FlexItem, Grid, Heading, Spacer, Surface, View } from '../../index';
export default {
  title: 'References/Colors'
};
var grades = [100, 300, 500, 700, 900];
var alphas = [10, 20, 30, 40, 50, 60, 70, 80, 90];

var ColorSet = function (_ref) {
  var color = _ref.color;
  return /*#__PURE__*/React.createElement(Card, null, grades.map(function (grade) {
    return /*#__PURE__*/React.createElement(CardBody, {
      key: grade,
      style: {
        background: get("" + color + grade),
        color: get("" + color + grade + "Text")
      }
    }, /*#__PURE__*/React.createElement(Flex, null, /*#__PURE__*/React.createElement(FlexItem, null, color, grade)));
  }));
};

var ColorRGBSet = function (_ref2) {
  var color = _ref2.color;
  return /*#__PURE__*/React.createElement(Card, null, alphas.map(function (alpha) {
    return /*#__PURE__*/React.createElement(CardBody, {
      key: alpha,
      style: {
        background: get(color + "Rgba" + alpha),
        color: get("" + color + alpha + "Text")
      }
    }, /*#__PURE__*/React.createElement(Flex, null, /*#__PURE__*/React.createElement(FlexItem, null, color + "Rgba" + alpha)));
  }));
};

var ColorCollection = function (_ref3) {
  var title = _ref3.title;
  return /*#__PURE__*/React.createElement(Spacer, {
    css: {
      position: 'relative'
    },
    mb: 8,
    p: 5
  }, /*#__PURE__*/React.createElement(View, {
    css: {
      margin: 'auto',
      maxWidth: 1200,
      position: 'relative',
      zIndex: 1
    }
  }, /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(Heading, {
    size: 2
  }, title)), /*#__PURE__*/React.createElement(Spacer, {
    mb: 8
  }, /*#__PURE__*/React.createElement(Grid, {
    columns: [1, 2, 4]
  }, COLORS.map(function (color) {
    return /*#__PURE__*/React.createElement(ColorSet, {
      color: color,
      key: color
    });
  }))), /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(Heading, {
    size: 4
  }, "RGBA")), /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(Grid, {
    columns: [1, 2, 4]
  }, COLORS.map(function (color) {
    return /*#__PURE__*/React.createElement(ColorRGBSet, {
      color: color,
      key: color
    });
  })))), /*#__PURE__*/React.createElement(Surface, {
    css: {
      bottom: 0,
      left: 0,
      position: 'absolute',
      right: 0,
      top: 0,
      zIndex: 0
    }
  }));
};

export var _default = function () {
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(ColorCollection, {
    title: "Colors"
  }), /*#__PURE__*/React.createElement(ThemeProvider, {
    isDark: true,
    isGlobal: false
  }, /*#__PURE__*/React.createElement(ColorCollection, {
    title: "Colors + Dark"
  })));
};
