// @wp-g2/components/dist/cjs/Draggable/__stories__/BeautifulDnd.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ReactBeautifulDnd = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _protokit = require("@wp-g2/protokit");

var _styles = require("@wp-g2/styles");

var _faker = _interopRequireDefault(require("faker"));

var _fakerSchema = require("faker-schema");

var _react = _interopRequireDefault(require("react"));

var _reactBeautifulDnd = require("react-beautiful-dnd");

var _index = require("../../index");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var _default = {
  title: 'Components/Draggable'
};
exports.default = _default;
var userSchema = new _fakerSchema.Schema(function () {
  return {
    avatar: _faker.default.image.avatar(),
    description: _faker.default.lorem.sentence(),
    id: _faker.default.random.uuid(),
    title: _faker.default.name.firstName()
  };
});

function Avatar(_ref) {
  var src = _ref.src;
  return /*#__PURE__*/_react.default.createElement(_index.View, {
    css: {
      height: 32,
      width: 32
    }
  }, /*#__PURE__*/_react.default.createElement(_index.Image, {
    css: [_styles.ui.borderRadius.circle, {
      display: 'block'
    }],
    src: src
  }));
}

var UserCard = function (_ref2) {
  var avatar = _ref2.avatar,
      description = _ref2.description,
      title = _ref2.title;
  return /*#__PURE__*/_react.default.createElement(_index.Card, null, /*#__PURE__*/_react.default.createElement(_index.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.HStack, {
    alignment: "topLeft",
    spacing: 4
  }, /*#__PURE__*/_react.default.createElement(Avatar, {
    src: avatar
  }), /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.VStack, {
    spacing: 1
  }, /*#__PURE__*/_react.default.createElement(_index.Text, {
    weight: "bold"
  }, title), /*#__PURE__*/_react.default.createElement(_index.Text, {
    variant: "muted"
  }, description))))));
};

var Example = function () {
  var _useListState = (0, _protokit.useListState)(userSchema.make(10)),
      _useListState2 = (0, _slicedToArray2.default)(_useListState, 2),
      users = _useListState2[0],
      usersData = _useListState2[1];

  var _useListState3 = (0, _protokit.useListState)(userSchema.make(10)),
      _useListState4 = (0, _slicedToArray2.default)(_useListState3, 2),
      editors = _useListState4[0],
      editorsData = _useListState4[1];

  var states = {
    editors: editors,
    users: users
  };
  var sources = {
    editors: editorsData,
    users: usersData
  };
  return /*#__PURE__*/_react.default.createElement(_reactBeautifulDnd.DragDropContext, {
    onBeforeCapture: function onBeforeCapture(_ref4) {
      var draggableId = _ref4.draggableId;
      var el = document.querySelector("[data-rbd-drag-handle-draggable-id=\"" + draggableId + "\"]");

      if (el) {
        el.style.height = '12px';
      }
    },
    onDragEnd: function onMove(_ref3) {
      var destination = _ref3.destination,
          source = _ref3.source,
          rest = (0, _objectWithoutProperties2.default)(_ref3, ["destination", "source"]);

      // Same list
      if (destination.droppableId === source.droppableId) {
        sources[destination.droppableId].move(source.index, destination.index);
        return;
      }

      var item = states[source.droppableId][source.index];
      sources[source.droppableId].remove({
        at: source.index
      });
      sources[destination.droppableId].insert({
        at: destination.index,
        item: item
      });
    }
  }, /*#__PURE__*/_react.default.createElement(_index.HStack, {
    css: [{
      maxWidth: '100%',
      width: 800
    }, _styles.ui.alignment.center],
    spacing: 0
  }, /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_reactBeautifulDnd.Droppable, {
    droppableId: "users"
  }, function (provided, snapshot) {
    return /*#__PURE__*/_react.default.createElement(_index.View, Object.assign({}, provided.droppableProps, {
      css: [_styles.ui.padding(5), snapshot.isDraggingOver && _styles.ui.background.blue],
      ref: provided.innerRef
    }), /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.Heading, null, "Users")), users.map(function (user, index) {
      return /*#__PURE__*/_react.default.createElement(_reactBeautifulDnd.Draggable, {
        draggableId: user.id,
        index: index,
        key: user.id
      }, function (provided) {
        return /*#__PURE__*/_react.default.createElement("div", Object.assign({
          ref: provided.innerRef
        }, provided.draggableProps, provided.dragHandleProps, {
          style: _objectSpread(_objectSpread({}, provided.draggableProps.style), {}, {
            marginBottom: 8
          })
        }), /*#__PURE__*/_react.default.createElement(UserCard, user));
      });
    }), provided.placeholder);
  })), /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_reactBeautifulDnd.Droppable, {
    droppableId: "editors"
  }, function (provided, snapshot) {
    return /*#__PURE__*/_react.default.createElement(_index.View, Object.assign({}, provided.droppableProps, {
      css: [_styles.ui.padding(5), snapshot.isDraggingOver && _styles.ui.background.blue],
      ref: provided.innerRef
    }), /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.Heading, null, "Editors")), editors.map(function (user, index) {
      return /*#__PURE__*/_react.default.createElement(_reactBeautifulDnd.Draggable, {
        draggableId: user.id,
        index: index,
        key: user.id
      }, function (provided) {
        return /*#__PURE__*/_react.default.createElement("div", Object.assign({
          ref: provided.innerRef
        }, provided.draggableProps, provided.dragHandleProps, {
          style: _objectSpread(_objectSpread({}, provided.draggableProps.style), {}, {
            marginBottom: 8
          })
        }), /*#__PURE__*/_react.default.createElement(UserCard, user));
      });
    }), provided.placeholder);
  }))));
};

var ReactBeautifulDnd = function () {
  return /*#__PURE__*/_react.default.createElement(Example, null);
};

exports.ReactBeautifulDnd = ReactBeautifulDnd;
