// @wp-g2/components/dist/cjs/Draggable/__stories__/DragListContext.stories.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DragListContext = exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _protokit = require("@wp-g2/protokit");

var _styles = require("@wp-g2/styles");

var _faker = _interopRequireDefault(require("faker"));

var _fakerSchema = require("faker-schema");

var _react = _interopRequireWildcard(require("react"));

var _reactBeautifulDnd = require("react-beautiful-dnd");

var _index = require("../../index");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var _default = {
  title: 'Components/Draggable'
};
exports.default = _default;
var blockSchema = new _fakerSchema.Schema(function () {
  return {
    content: _faker.default.lorem.sentences(),
    id: _faker.default.random.uuid(),
    title: 'Paragraph'
  };
});

var BlockDragIndexLine = function () {
  return /*#__PURE__*/_react.default.createElement(_index.View, {
    css: [_styles.ui.position.relative]
  }, /*#__PURE__*/_react.default.createElement(_index.View, {
    css: [_styles.ui.position.top, _styles.ui.borderRadius.round, {
      height: 3,
      width: '100%'
    }, _styles.ui.background.admin]
  }));
};

var Example = function () {
  var _useListState = (0, _protokit.useListState)(blockSchema.make(10)),
      _useListState2 = (0, _slicedToArray2.default)(_useListState, 2),
      blocks = _useListState2[0],
      setBlocks = _useListState2[1];

  var _useState = (0, _react.useState)(),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      dragIndex = _useState2[0],
      setDragIndex = _useState2[1];

  var _useState3 = (0, _react.useState)(),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      targetIndex = _useState4[0],
      setTargetIndex = _useState4[1];

  var _useState5 = (0, _react.useState)(false),
      _useState6 = (0, _slicedToArray2.default)(_useState5, 2),
      isDragging = _useState6[0],
      setIsDragging = _useState6[1];

  var setHighlightOn = function (index) {
    setTargetIndex(index);
    setIsDragging(true);
  };

  var setHighlightOff = function () {
    setTargetIndex(undefined);
    setIsDragging(false);
  };

  return /*#__PURE__*/_react.default.createElement(_reactBeautifulDnd.DragDropContext, {
    onDragEnd: function onDragEnd(_ref) {
      var destination = _ref.destination,
          source = _ref.source;
      setBlocks.move(source.index, destination.index);
      setDragIndex(undefined);
      setHighlightOff();
    },
    onDragStart: function onDragStart(_ref2) {
      var source = _ref2.source;
      setDragIndex(source == null ? void 0 : source.index);
      setTargetIndex(source == null ? void 0 : source.index);
      setHighlightOn(source == null ? void 0 : source.index);
    },
    onDragUpdate: function onDragUpdate(_ref3) {
      var destination = _ref3.destination;
      setDragIndex(destination == null ? void 0 : destination.index);
    }
  }, /*#__PURE__*/_react.default.createElement(_index.HStack, {
    css: [_styles.ui.frame.width(800), _styles.ui.alignment.center],
    spacing: 8
  }, /*#__PURE__*/_react.default.createElement(_index.View, {
    css: _styles.ui.frame.width(200)
  }, /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.Heading, {
    size: 4
  }, "Block List")), /*#__PURE__*/_react.default.createElement(_index.Scrollable, {
    css: [_styles.ui.frame.height(300), _styles.ui.padding(2)]
  }, /*#__PURE__*/_react.default.createElement(_reactBeautifulDnd.Droppable, {
    droppableId: "blockList"
  }, function (provided) {
    return /*#__PURE__*/_react.default.createElement(_index.View, Object.assign({}, provided.droppableProps, {
      ref: provided.innerRef
    }), blocks.map(function (block, index) {
      return /*#__PURE__*/_react.default.createElement(_reactBeautifulDnd.Draggable, {
        draggableId: block.id,
        index: index,
        key: block.id
      }, function (provided) {
        return /*#__PURE__*/_react.default.createElement("div", Object.assign({
          ref: provided.innerRef
        }, provided.draggableProps, provided.dragHandleProps, {
          style: _objectSpread(_objectSpread({}, provided.draggableProps.style), {}, {
            marginBottom: 8
          })
        }), /*#__PURE__*/_react.default.createElement(_index.Card, null, /*#__PURE__*/_react.default.createElement(_index.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.Text, {
          weight: "bold"
        }, block.title))));
      });
    }), provided.placeholder);
  }))), /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.Heading, {
    size: 2
  }, "Blog Post"), blocks.map(function (block, index) {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, {
      key: block.id
    }, index === dragIndex && /*#__PURE__*/_react.default.createElement(_index.Animated, {
      key: "BlockDragIndexLine",
      layout: true
    }, /*#__PURE__*/_react.default.createElement(BlockDragIndexLine, null)), /*#__PURE__*/_react.default.createElement(_index.Animated, {
      key: block.id,
      layout: true
    }, /*#__PURE__*/_react.default.createElement(_index.View, {
      css: [_styles.ui.animation.ease, isDragging && _styles.ui.opacity(0.3), targetIndex === index && _styles.ui.opacity(1)]
    }, /*#__PURE__*/_react.default.createElement(_index.Text, {
      css: {
        marginBottom: '1em',
        marginTop: '1em'
      },
      isBlock: true,
      size: 16
    }, block.content, ' '))));
  })))));
};

var DragListContext = function () {
  return /*#__PURE__*/_react.default.createElement(Example, null);
};

exports.DragListContext = DragListContext;
