// @wp-g2/components/dist/cjs/Draggable/__stories__/Draggable.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _protokit = require("@wp-g2/protokit");

var _styles = require("@wp-g2/styles");

var _faker = _interopRequireDefault(require("faker"));

var _fakerSchema = require("faker-schema");

var _react = _interopRequireDefault(require("react"));

var _reactDnd = require("react-dnd");

var _reactDndHtml5Backend = require("react-dnd-html5-backend");

var _index = require("../../index");

var _default2 = {
  title: 'Components/Draggable'
};
exports.default = _default2;
var ItemTypes = {
  DONE: 'DONE',
  TODO: 'TODO'
};

function DraggableProvider(_ref) {
  var children = _ref.children;
  return /*#__PURE__*/_react.default.createElement(_reactDnd.DndProvider, {
    backend: _reactDndHtml5Backend.HTML5Backend
  }, children);
}

function DropZone(_ref2) {
  var _ref2$accept = _ref2.accept,
      accept = _ref2$accept === void 0 ? ItemTypes.DONE : _ref2$accept,
      _ref2$dropName = _ref2.dropName,
      dropName = _ref2$dropName === void 0 ? 'Todos' : _ref2$dropName;

  var _useDrop = (0, _reactDnd.useDrop)({
    accept: accept,
    collect: function collect(monitor) {
      return {
        canDrop: monitor.canDrop(),
        isOver: monitor.isOver()
      };
    },
    drop: function () {
      return {
        name: dropName
      };
    }
  }),
      _useDrop2 = (0, _slicedToArray2.default)(_useDrop, 2),
      _useDrop2$ = _useDrop2[0],
      canDrop = _useDrop2$.canDrop,
      isOver = _useDrop2$.isOver,
      drop = _useDrop2[1];

  var isActive = canDrop && isOver;
  return /*#__PURE__*/_react.default.createElement(_index.View, {
    css: [_styles.ui.position.full, _styles.ui.alignment.content.center, _styles.ui.background.blue, isActive && _styles.ui.background.green, _styles.ui.animation.ease, _styles.ui.padding(5), isActive ? _styles.ui.background.green : _styles.ui.background.yellow, _styles.ui.borderRadius.round, {
      opacity: canDrop ? 1 : 0
    }, {
      pointerEvents: canDrop ? 'default' : 'none'
    }, {
      zIndex: 10
    }],
    ref: drop
  }, /*#__PURE__*/_react.default.createElement(_index.Heading, null, isActive ? 'Drop Here' : 'Drag Here'));
}

function Draggable(_ref3) {
  var children = _ref3.children,
      onEnd = _ref3.onEnd,
      _ref3$type = _ref3.type,
      type = _ref3$type === void 0 ? ItemTypes.TODO : _ref3$type;

  var _useDrag = (0, _reactDnd.useDrag)({
    collect: function collect(monitor) {
      return {
        isDragging: monitor.isDragging()
      };
    },
    end: function end(item, monitor) {
      onEnd && onEnd(item, monitor);
    },
    item: {
      type: type
    }
  }),
      _useDrag2 = (0, _slicedToArray2.default)(_useDrag, 2),
      isDragging = _useDrag2[0].isDragging,
      drag = _useDrag2[1];

  return /*#__PURE__*/_react.default.createElement(_index.View, {
    ref: drag
  }, children);
}

var itemSchema = new _fakerSchema.Schema(function () {
  return {
    avatar: _faker.default.image.avatar(),
    description: _faker.default.lorem.sentence(),
    id: _faker.default.random.uuid(),
    title: _faker.default.name.firstName()
  };
});

function Avatar(_ref4) {
  var src = _ref4.src;
  return /*#__PURE__*/_react.default.createElement(_index.View, {
    css: {
      height: 32,
      width: 32
    }
  }, /*#__PURE__*/_react.default.createElement(_index.Image, {
    css: _styles.ui.borderRadius.circle,
    src: src
  }));
}

function ItemCard(_ref5) {
  var avatar = _ref5.avatar,
      description = _ref5.description,
      id = _ref5.id,
      title = _ref5.title;
  return /*#__PURE__*/_react.default.createElement(_index.Animated, {
    auto: true,
    css: {
      margin: -4,
      marginBottom: 4,
      padding: 4
    },
    key: id
  }, /*#__PURE__*/_react.default.createElement(_index.Card, null, /*#__PURE__*/_react.default.createElement(_index.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.HStack, null, /*#__PURE__*/_react.default.createElement(Avatar, {
    src: avatar
  }), /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.VStack, {
    spacing: 1
  }, /*#__PURE__*/_react.default.createElement(_index.Text, {
    weight: "bold"
  }, title), /*#__PURE__*/_react.default.createElement(_index.Text, {
    numberOfLines: 2,
    truncate: true,
    variant: "muted"
  }, description)))))));
}

function Example() {
  var _useListState = (0, _protokit.useListState)(itemSchema.make(7)),
      _useListState2 = (0, _slicedToArray2.default)(_useListState, 2),
      todos = _useListState2[0],
      todosData = _useListState2[1];

  var _useListState3 = (0, _protokit.useListState)(itemSchema.make(2)),
      _useListState4 = (0, _slicedToArray2.default)(_useListState3, 2),
      dones = _useListState4[0],
      donesData = _useListState4[1];

  return /*#__PURE__*/_react.default.createElement(_index.HStack, {
    alignment: "center"
  }, /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    css: {
      maxWidth: 800
    }
  }, /*#__PURE__*/_react.default.createElement(DraggableProvider, null, /*#__PURE__*/_react.default.createElement(_index.HStack, {
    direction: ['column', 'row']
  }, /*#__PURE__*/_react.default.createElement(_index.VStack, {
    css: [{
      width: [null, 240]
    }, _styles.ui.position.relative]
  }, /*#__PURE__*/_react.default.createElement(_index.Heading, null, "Pending"), /*#__PURE__*/_react.default.createElement(DropZone, {
    accept: ItemTypes.DONE,
    dropName: "Todos"
  }), /*#__PURE__*/_react.default.createElement(_index.AnimatedContainer, null, todos.map(function (todo) {
    return /*#__PURE__*/_react.default.createElement(Draggable, {
      key: todo.id,
      onEnd: function onEnd(item, monitor) {
        var dropResult = monitor.getDropResult();

        if ((dropResult == null ? void 0 : dropResult.name) === 'Dones') {
          donesData.prepend(todo);
          todosData.remove({
            id: todo.id
          });
        }
      },
      type: ItemTypes.TODO
    }, /*#__PURE__*/_react.default.createElement(ItemCard, todo));
  }))), /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.VStack, {
    css: [_styles.ui.position.relative]
  }, /*#__PURE__*/_react.default.createElement(_index.Heading, null, "Added"), /*#__PURE__*/_react.default.createElement(DropZone, {
    accept: ItemTypes.TODO,
    dropName: "Dones"
  }), /*#__PURE__*/_react.default.createElement(_index.AnimatedContainer, null, dones.map(function (todo) {
    return /*#__PURE__*/_react.default.createElement(Draggable, {
      key: todo.id,
      onEnd: function onEnd(item, monitor) {
        var dropResult = monitor.getDropResult();

        if ((dropResult == null ? void 0 : dropResult.name) === 'Todos') {
          todosData.prepend(todo);
          donesData.remove({
            id: todo.id
          });
        }
      },
      type: ItemTypes.DONE
    }, /*#__PURE__*/_react.default.createElement(ItemCard, todo));
  }))))))));
}

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(Example, null);
};

exports._default = _default;
