// @wp-g2/components/dist/cjs/Draggable/__stories__/Sortable.stories.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SortableExample = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _protokit = require("@wp-g2/protokit");

var _faker = _interopRequireDefault(require("faker"));

var _fakerSchema = require("faker-schema");

var _react = _interopRequireWildcard(require("react"));

var _sortablejs = _interopRequireDefault(require("sortablejs"));

var _index = require("../../index");

var _default = {
  title: 'Components/Draggable'
};
exports.default = _default;

var Sortable = /*#__PURE__*/function (_React$Component) {
  (0, _inheritsLoose2.default)(Sortable, _React$Component);

  function Sortable() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;
    _this.nodeRef = /*#__PURE__*/(0, _react.createRef)();
    _this.sortable = null;
    return _this;
  }

  var _proto = Sortable.prototype;

  _proto.componentDidMount = function componentDidMount() {
    this.sortable = _sortablejs.default.create(this.nodeRef.current, {
      animation: 150,
      forceFallback: true,
      ghostClass: 'ghost'
    });
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    this.sortable.destroy();
  };

  _proto.render = function render() {
    return /*#__PURE__*/_react.default.createElement(_index.View, {
      ref: this.nodeRef
    }, this.props.children);
  };

  return Sortable;
}(_react.default.Component);

var Draggable = function (props) {
  return /*#__PURE__*/_react.default.createElement(_index.View, Object.assign({
    "data-sortable-item": "true"
  }, props, {
    css: [{
      marginBottom: 8
    }, "&.ghost {opacity: 0}"]
  }));
};

var itemSchema = new _fakerSchema.Schema(function () {
  return {
    id: _faker.default.random.uuid(),
    name: _faker.default.name.firstName()
  };
});

var Example = function () {
  var _useListState = (0, _protokit.useListState)(itemSchema.make(10)),
      _useListState2 = (0, _slicedToArray2.default)(_useListState, 1),
      items = _useListState2[0];

  return /*#__PURE__*/_react.default.createElement(_index.VStack, {
    spacing: 5
  }, /*#__PURE__*/_react.default.createElement(_index.VStack, {
    spacing: 0
  }, /*#__PURE__*/_react.default.createElement(_index.Heading, null, "SortableJS"), /*#__PURE__*/_react.default.createElement(_index.Subheading, null, "Probaby a no go.")), /*#__PURE__*/_react.default.createElement(Sortable, null, items.map(function (item) {
    return /*#__PURE__*/_react.default.createElement(Draggable, {
      key: item.id
    }, /*#__PURE__*/_react.default.createElement(_index.Card, null, /*#__PURE__*/_react.default.createElement(_index.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.Text, null, item.name))));
  })));
};

var SortableExample = function () {
  return /*#__PURE__*/_react.default.createElement(Example, null);
};

exports.SortableExample = SortableExample;
