// @wp-g2/components/dist/cjs/Collapsible/CollapsibleContent.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _reakit = require("reakit");

var _Animated = require("../Animated");

var _Collapsible = require("./Collapsible.Context");

var styles = _interopRequireWildcard(require("./Collapsible.styles"));

var animationVariants = {
  visible: {
    height: 'auto',
    opacity: 1,
    transitionEnd: {
      display: 'block'
    }
  },
  hidden: {
    height: 0,
    opacity: 0,
    transitionEnd: {
      display: 'none'
    }
  }
};

function CollapsibleContent(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'CollapsibleContent'),
      children = _useContextSystem.children,
      className = _useContextSystem.className,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["children", "className"]);

  var _useCollapsibleContex = (0, _Collapsible.useCollapsibleContext)(),
      disclosure = _useCollapsibleContex.disclosure;

  var isVisible = disclosure == null ? void 0 : disclosure.visible;
  var classes = (0, _styles.cx)(styles.CollapsibleContent, className);
  return /*#__PURE__*/_react.default.createElement(_reakit.DisclosureContent, Object.assign({
    className: classes,
    ref: forwardedRef
  }, otherProps, disclosure), /*#__PURE__*/_react.default.createElement(_Animated.Animated, Object.assign({
    animate: isVisible ? 'visible' : 'hidden',
    className: (0, _styles.cx)(styles.innerContent),
    initial: false,
    transition: {
      duration: 0.2
    },
    variants: animationVariants
  }, _styles.ui.$('CollapsibleInnerContent')), children));
}

var _default = (0, _context.contextConnect)(CollapsibleContent, 'CollapsibleContent');

exports.default = _default;
