// @wordpress/block-editor/build/components/block-media-update-progress/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.BlockMediaUpdateProgress = exports.MEDIA_SAVE_MEDIAID_CHANGED = exports.MEDIA_SAVE_FINAL_STATE_RESULT = exports.MEDIA_SAVE_STATE_RESET = exports.MEDIA_SAVE_STATE_FAILED = exports.MEDIA_SAVE_STATE_SUCCEEDED = exports.MEDIA_SAVE_STATE_SAVING = exports.MEDIA_UPLOAD_STATE_RESET = exports.MEDIA_UPLOAD_STATE_FAILED = exports.MEDIA_UPLOAD_STATE_SUCCEEDED = exports.MEDIA_UPLOAD_STATE_UPLOADING = void 0;

var _element = require("@wordpress/element");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _reactNative = require("react-native");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _reactNativeBridge = require("@wordpress/react-native-bridge");

var _styles = _interopRequireDefault(require("./styles.scss"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var MEDIA_UPLOAD_STATE_UPLOADING = 1;
exports.MEDIA_UPLOAD_STATE_UPLOADING = MEDIA_UPLOAD_STATE_UPLOADING;
var MEDIA_UPLOAD_STATE_SUCCEEDED = 2;
exports.MEDIA_UPLOAD_STATE_SUCCEEDED = MEDIA_UPLOAD_STATE_SUCCEEDED;
var MEDIA_UPLOAD_STATE_FAILED = 3;
exports.MEDIA_UPLOAD_STATE_FAILED = MEDIA_UPLOAD_STATE_FAILED;
var MEDIA_UPLOAD_STATE_RESET = 4;
exports.MEDIA_UPLOAD_STATE_RESET = MEDIA_UPLOAD_STATE_RESET;
var MEDIA_SAVE_STATE_SAVING = 5;
exports.MEDIA_SAVE_STATE_SAVING = MEDIA_SAVE_STATE_SAVING;
var MEDIA_SAVE_STATE_SUCCEEDED = 6;
exports.MEDIA_SAVE_STATE_SUCCEEDED = MEDIA_SAVE_STATE_SUCCEEDED;
var MEDIA_SAVE_STATE_FAILED = 7;
exports.MEDIA_SAVE_STATE_FAILED = MEDIA_SAVE_STATE_FAILED;
var MEDIA_SAVE_STATE_RESET = 8;
exports.MEDIA_SAVE_STATE_RESET = MEDIA_SAVE_STATE_RESET;
var MEDIA_SAVE_FINAL_STATE_RESULT = 9;
exports.MEDIA_SAVE_FINAL_STATE_RESULT = MEDIA_SAVE_FINAL_STATE_RESULT;
var MEDIA_SAVE_MEDIAID_CHANGED = 10;
exports.MEDIA_SAVE_MEDIAID_CHANGED = MEDIA_SAVE_MEDIAID_CHANGED;

var BlockMediaUpdateProgress = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(BlockMediaUpdateProgress, _Component);

  var _super = _createSuper(BlockMediaUpdateProgress);

  function BlockMediaUpdateProgress(props) {
    var _this;

    (0, _classCallCheck2.default)(this, BlockMediaUpdateProgress);
    _this = _super.call(this, props);
    _this.state = {
      progress: 0,
      isSaveInProgress: false,
      isSaveFailed: false,
      isUploadInProgress: false,
      isUploadFailed: false
    };
    _this.mediaUpload = _this.mediaUpload.bind((0, _assertThisInitialized2.default)(_this));
    _this.mediaSave = _this.mediaSave.bind((0, _assertThisInitialized2.default)(_this));
    return _this;
  }

  (0, _createClass2.default)(BlockMediaUpdateProgress, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.addMediaUploadListener();
      this.addMediaSaveListener();
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.removeMediaUploadListener();
      this.removeMediaSaveListener();
    }
  }, {
    key: "mediaIdContainedInMediaFiles",
    value: function mediaIdContainedInMediaFiles(mediaId, mediaFiles) {
      if (mediaId !== undefined && mediaFiles !== undefined) {
        return mediaFiles.some(function (element) {
          return element.id === mediaId.toString();
        });
      }

      return false;
    }
  }, {
    key: "mediaUpload",
    value: function mediaUpload(payload) {
      var mediaFiles = this.props.mediaFiles;

      if (this.mediaIdContainedInMediaFiles(payload.mediaId, mediaFiles) === false) {
        return;
      }

      switch (payload.state) {
        case MEDIA_UPLOAD_STATE_UPLOADING:
          this.updateMediaUploadProgress(payload);
          break;

        case MEDIA_UPLOAD_STATE_SUCCEEDED:
          this.finishMediaUploadWithSuccess(payload);
          break;

        case MEDIA_UPLOAD_STATE_FAILED:
          this.finishMediaUploadWithFailure(payload);
          break;

        case MEDIA_UPLOAD_STATE_RESET:
          this.mediaUploadStateReset(payload);
          break;
      }
    }
  }, {
    key: "mediaSave",
    value: function mediaSave(payload) {
      var mediaFiles = this.props.mediaFiles;

      if (this.mediaIdContainedInMediaFiles(payload.mediaId, mediaFiles) === false) {
        return;
      }

      switch (payload.state) {
        case MEDIA_SAVE_STATE_SAVING:
          this.updateMediaSaveProgress(payload);
          break;

        case MEDIA_SAVE_STATE_SUCCEEDED:
          this.finishMediaSaveWithSuccess(payload);
          break;

        case MEDIA_SAVE_STATE_FAILED:
          this.finishMediaSaveWithFailure(payload);
          break;

        case MEDIA_SAVE_STATE_RESET:
          this.mediaSaveStateReset(payload);
          break;

        case MEDIA_SAVE_FINAL_STATE_RESULT:
          this.finalSaveResult(payload);
          break;

        case MEDIA_SAVE_MEDIAID_CHANGED:
          this.mediaIdChanged(payload);
          break;
      }
    } // ---- Block media save actions

  }, {
    key: "updateMediaSaveProgress",
    value: function updateMediaSaveProgress(payload) {
      this.setState({
        progress: payload.progress,
        isUploadInProgress: false,
        isUploadFailed: false,
        isSaveInProgress: true,
        isSaveFailed: false
      });

      if (this.props.onUpdateMediaSaveProgress) {
        this.props.onUpdateMediaSaveProgress(payload);
      }
    }
  }, {
    key: "finishMediaSaveWithSuccess",
    value: function finishMediaSaveWithSuccess(payload) {
      this.setState({
        isSaveInProgress: false
      });

      if (this.props.onFinishMediaSaveWithSuccess) {
        this.props.onFinishMediaSaveWithSuccess(payload);
      }
    }
  }, {
    key: "finishMediaSaveWithFailure",
    value: function finishMediaSaveWithFailure(payload) {
      this.setState({
        isSaveInProgress: false,
        isSaveFailed: true
      });

      if (this.props.onFinishMediaSaveWithFailure) {
        this.props.onFinishMediaSaveWithFailure(payload);
      }
    }
  }, {
    key: "mediaSaveStateReset",
    value: function mediaSaveStateReset(payload) {
      this.setState({
        isSaveInProgress: false,
        isSaveFailed: false
      });

      if (this.props.onMediaSaveStateReset) {
        this.props.onMediaSaveStateReset(payload);
      }
    }
  }, {
    key: "finalSaveResult",
    value: function finalSaveResult(payload) {
      this.setState({
        progress: payload.progress,
        isUploadInProgress: false,
        isUploadFailed: false,
        isSaveInProgress: false,
        isSaveFailed: !payload.success
      });

      if (this.props.onFinalSaveResult) {
        this.props.onFinalSaveResult(payload);
      }
    }
  }, {
    key: "mediaIdChanged",
    value: function mediaIdChanged(payload) {
      this.setState({
        isUploadInProgress: false,
        isUploadFailed: false,
        isSaveInProgress: false,
        isSaveFailed: false
      });

      if (this.props.onMediaIdChanged) {
        this.props.onMediaIdChanged(payload);
      }
    } // ---- Block media upload actions

  }, {
    key: "updateMediaUploadProgress",
    value: function updateMediaUploadProgress(payload) {
      this.setState({
        progress: payload.progress,
        isUploadInProgress: true,
        isUploadFailed: false,
        isSaveInProgress: false,
        isSaveFailed: false
      });

      if (this.props.onUpdateMediaUploadProgress) {
        this.props.onUpdateMediaUploadProgress(payload);
      }
    }
  }, {
    key: "finishMediaUploadWithSuccess",
    value: function finishMediaUploadWithSuccess(payload) {
      this.setState({
        isUploadInProgress: false,
        isSaveInProgress: false
      });

      if (this.props.onFinishMediaUploadWithSuccess) {
        this.props.onFinishMediaUploadWithSuccess(payload);
      }
    }
  }, {
    key: "finishMediaUploadWithFailure",
    value: function finishMediaUploadWithFailure(payload) {
      this.setState({
        isUploadInProgress: false,
        isUploadFailed: true
      });

      if (this.props.onFinishMediaUploadWithFailure) {
        this.props.onFinishMediaUploadWithFailure(payload);
      }
    }
  }, {
    key: "mediaUploadStateReset",
    value: function mediaUploadStateReset(payload) {
      this.setState({
        isUploadInProgress: false,
        isUploadFailed: false
      });

      if (this.props.onMediaUploadStateReset) {
        this.props.onMediaUploadStateReset(payload);
      }
    }
  }, {
    key: "addMediaUploadListener",
    value: function addMediaUploadListener() {
      var _this2 = this;

      //if we already have a subscription not worth doing it again
      if (this.subscriptionParentMediaUpload) {
        return;
      }

      this.subscriptionParentMediaUpload = (0, _reactNativeBridge.subscribeMediaUpload)(function (payload) {
        _this2.mediaUpload(payload);
      });
    }
  }, {
    key: "removeMediaUploadListener",
    value: function removeMediaUploadListener() {
      if (this.subscriptionParentMediaUpload) {
        this.subscriptionParentMediaUpload.remove();
      }
    }
  }, {
    key: "addMediaSaveListener",
    value: function addMediaSaveListener() {
      var _this3 = this;

      //if we already have a subscription not worth doing it again
      if (this.subscriptionParentMediaSave) {
        return;
      }

      this.subscriptionParentMediaSave = (0, _reactNativeBridge.subscribeMediaSave)(function (payload) {
        _this3.mediaSave(payload);
      });
    }
  }, {
    key: "removeMediaSaveListener",
    value: function removeMediaSaveListener() {
      if (this.subscriptionParentMediaSave) {
        this.subscriptionParentMediaSave.remove();
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props$renderCon = this.props.renderContent,
          renderContent = _this$props$renderCon === void 0 ? function () {
        return null;
      } : _this$props$renderCon;
      var _this$state = this.state,
          isUploadInProgress = _this$state.isUploadInProgress,
          isUploadFailed = _this$state.isUploadFailed,
          isSaveInProgress = _this$state.isSaveInProgress,
          isSaveFailed = _this$state.isSaveFailed;
      var showSpinner = this.state.isUploadInProgress || this.state.isSaveInProgress;
      var progress = this.state.progress * 100; // eslint-disable-next-line @wordpress/i18n-no-collapsible-whitespace

      var retryMessageSave = (0, _i18n.__)('Failed to save files.\nPlease tap for options.'); // eslint-disable-next-line @wordpress/i18n-no-collapsible-whitespace

      var retryMessageUpload = (0, _i18n.__)('Failed to upload files.\nPlease tap for options.');
      var retryMessage = retryMessageSave;

      if (isUploadFailed) {
        retryMessage = retryMessageUpload;
      }

      return (0, _element.createElement)(_reactNative.View, {
        style: _styles.default.mediaUploadProgress,
        pointerEvents: "box-none"
      }, showSpinner && (0, _element.createElement)(_reactNative.View, {
        style: _styles.default.progressBar
      }, (0, _element.createElement)(_components.Spinner, {
        progress: progress
      })), renderContent({
        isUploadInProgress: isUploadInProgress,
        isUploadFailed: isUploadFailed,
        isSaveInProgress: isSaveInProgress,
        isSaveFailed: isSaveFailed,
        retryMessage: retryMessage
      }));
    }
  }]);
  return BlockMediaUpdateProgress;
}(_element.Component);

exports.BlockMediaUpdateProgress = BlockMediaUpdateProgress;
var _default = BlockMediaUpdateProgress;
exports.default = _default;
