// @wordpress/block-editor/build-module/components/block-styles/index.native.js
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { ScrollView } from 'react-native';
import { find } from 'lodash';
/**
 * WordPress dependencies
 */

import { store as blocksStore } from '@wordpress/blocks';
import { useSelect, useDispatch } from '@wordpress/data';
import { useMemo } from '@wordpress/element';
import { _x } from '@wordpress/i18n';
/**
 * Internal dependencies
 */

import { getActiveStyle, replaceActiveStyle } from './utils';
import StylePreview from './preview';
import containerStyles from './style.scss';
import { store as blockEditorStore } from '../../store';

function BlockStyles(_ref) {
  var clientId = _ref.clientId,
      url = _ref.url;

  var selector = function selector(select) {
    var _select = select(blockEditorStore),
        getBlock = _select.getBlock;

    var _select2 = select(blocksStore),
        getBlockStyles = _select2.getBlockStyles;

    var block = getBlock(clientId);
    return {
      styles: getBlockStyles(block.name),
      className: block.attributes.className || ''
    };
  };

  var _useSelect = useSelect(selector, [clientId]),
      styles = _useSelect.styles,
      className = _useSelect.className;

  var _useDispatch = useDispatch(blockEditorStore),
      updateBlockAttributes = _useDispatch.updateBlockAttributes;

  var renderedStyles = find(styles, 'isDefault') ? styles : [{
    name: 'default',
    label: _x('Default', 'block style'),
    isDefault: true
  }].concat(_toConsumableArray(styles));
  var mappedRenderedStyles = useMemo(function () {
    var activeStyle = getActiveStyle(renderedStyles, className);
    return renderedStyles.map(function (style) {
      var styleClassName = replaceActiveStyle(className, activeStyle, style);
      var isActive = activeStyle === style;

      var onStylePress = function onStylePress() {
        updateBlockAttributes(clientId, {
          className: styleClassName
        });
      };

      return createElement(StylePreview, {
        onPress: onStylePress,
        isActive: isActive,
        key: style.name,
        style: style,
        url: url
      });
    });
  }, [renderedStyles, className, clientId]);

  if (!styles || styles.length === 0) {
    return null;
  }

  return createElement(ScrollView, {
    horizontal: true,
    showsHorizontalScrollIndicator: false,
    contentContainerStyle: containerStyles.content
  }, mappedRenderedStyles);
}

export default BlockStyles;
