// @wp-g2/styles/dist/cjs/mixins/fonts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getFontSize = getFontSize;
exports.getHeadingFontSize = getHeadingFontSize;
exports.HEADING_FONT_SIZES = exports.PRESET_FONT_SIZES = exports.BASE_FONT_SIZE = void 0;

var _core = require("../core");

var BASE_FONT_SIZE = 13;
exports.BASE_FONT_SIZE = BASE_FONT_SIZE;
var PRESET_FONT_SIZES = {
  body: 13,
  caption: 10,
  footnote: 11,
  largeTitle: 28,
  subheadline: 12,
  title: 20
};
exports.PRESET_FONT_SIZES = PRESET_FONT_SIZES;
var HEADING_FONT_SIZES = [1, 2, 3, 4, 5, 6].flatMap(function (n) {
  return [n, n.toString()];
});
/**
 *
 * @param {import('react').CSSProperties['fontSize'] | 'body' | 'caption' | 'footnote' | 'largeTitle' | 'subheadline' | 'title'} size
 * @return {string}
 */

exports.HEADING_FONT_SIZES = HEADING_FONT_SIZES;

function getFontSize(size) {
  if (size === void 0) {
    size = BASE_FONT_SIZE;
  }

  if (PRESET_FONT_SIZES[size]) {
    return getFontSize(PRESET_FONT_SIZES[size]);
  }

  if (typeof size !== 'number') {
    var parsed = parseFloat(size);
    if (Number.isNaN(parsed)) return size;
    size = parsed;
  }

  var ratio = "(" + size + " / " + BASE_FONT_SIZE + ")";
  return "calc(" + ratio + " * " + (0, _core.get)('fontSize') + ")";
}
/**
 * @param {import('react').ReactText} size
 * @return {string}
 */


function getHeadingFontSize(size) {
  if (size === void 0) {
    size = 3;
  }

  if (!HEADING_FONT_SIZES.includes(size)) {
    return getFontSize(size);
  }

  var headingSize = "fontSizeH" + size;
  /* @ts-ignore */

  return (0, _core.get)(headingSize);
}
