// @wp-g2/components/dist/cjs/Alert/Alert.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _Animated = require("../Animated");

var _Flex = require("../Flex");

var _View = require("../View");

var styles = _interopRequireWildcard(require("./Alert.styles"));

var _AlertCloseButton = _interopRequireDefault(require("./AlertCloseButton"));

var _AlertTitle = _interopRequireDefault(require("./AlertTitle"));

var AlertView = styles.AlertView;

function Alert(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'Alert'),
      children = _useContextSystem.children,
      className = _useContextSystem.className,
      _useContextSystem$isD = _useContextSystem.isDismissable,
      isDismissableProp = _useContextSystem$isD === void 0 ? false : _useContextSystem$isD,
      onDismiss = _useContextSystem.onDismiss,
      _useContextSystem$sta = _useContextSystem.status,
      status = _useContextSystem$sta === void 0 ? 'default' : _useContextSystem$sta,
      title = _useContextSystem.title,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["children", "className", "isDismissable", "onDismiss", "status", "title"]);

  var isDismissable = isDismissableProp || !!onDismiss;
  var classes = (0, _styles.cx)(styles[status], className);
  var viewClasses = (0, _styles.cx)([isDismissable && styles.contentWithDismiss]);
  return /*#__PURE__*/_react.default.createElement(_Animated.Animated, {
    auto: true
  }, /*#__PURE__*/_react.default.createElement(AlertView, Object.assign({}, otherProps, {
    className: classes,
    ref: forwardedRef
  }), /*#__PURE__*/_react.default.createElement(_Flex.Flex, {
    align: "flex-start"
  }, /*#__PURE__*/_react.default.createElement(_Flex.FlexBlock, null, /*#__PURE__*/_react.default.createElement(_AlertTitle.default, {
    title: title
  }), /*#__PURE__*/_react.default.createElement(_View.View, Object.assign({}, _styles.ui.$('AlertContent'), {
    className: viewClasses
  }), children)), /*#__PURE__*/_react.default.createElement(_AlertCloseButton.default, {
    isDismissable: isDismissable,
    onDismiss: onDismiss,
    status: status
  }))));
}

var _default = (0, _context.contextConnect)(Alert, 'Alert');

exports.default = _default;
