// @wp-g2/components/dist/cjs/Alert/AlertCloseButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _styles = require("@wp-g2/styles");

var _utils = require("@wp-g2/utils");

var _react = _interopRequireDefault(require("react"));

var _CloseButton = require("../CloseButton");

var styles = _interopRequireWildcard(require("./Alert.styles"));

var CloseButtonWrapper = styles.CloseButtonWrapper;

function AlertCloseButton(_ref) {
  var className = _ref.className,
      isDismissable = _ref.isDismissable,
      _ref$onDismiss = _ref.onDismiss,
      onDismiss = _ref$onDismiss === void 0 ? _utils.noop : _ref$onDismiss,
      status = _ref.status,
      props = (0, _objectWithoutProperties2.default)(_ref, ["className", "isDismissable", "onDismiss", "status"]);
  if (!isDismissable) return null;
  var classes = (0, _styles.cx)((0, _styles.css)({
    color: styles.getTextColor(status)
  }), className);
  return /*#__PURE__*/_react.default.createElement(CloseButtonWrapper, Object.assign({}, props, {
    className: classes
  }), /*#__PURE__*/_react.default.createElement(_CloseButton.CloseButton, {
    currentColor: true,
    iconSize: 16,
    onClick: onDismiss,
    title: "Dismiss"
  }));
}

var _default = /*#__PURE__*/_react.default.memo(AlertCloseButton);

exports.default = _default;
