// @wp-g2/components/dist/cjs/Alert/Alert.styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getTextColor = getTextColor;
exports.info = exports.critical = exports.warning = exports.success = exports.contentWithDismiss = exports.CloseButtonWrapper = exports.AlertView = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _styles = require("@wp-g2/styles");

function _templateObject7() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n"]);

  _templateObject7 = function () {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n"]);

  _templateObject6 = function () {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n"]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tpadding-top: ", ";\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tmargin-right: ", ";\n\tmargin-top: ", ";\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbackground: ", ";\n\tborder-radius: ", ";\n\tbox-shadow: 0 0 0 1px ", " inset;\n\tpadding: ", " ", ";\n\tposition: relative;\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var AlertView = _styles.styled.div(_templateObject(), _styles.ui.get('controlBackgroundColor'), _styles.ui.get('controlBorderRadius'), _styles.ui.get('surfaceBorderColor'), _styles.ui.space(3), _styles.ui.space(4));

exports.AlertView = AlertView;

var CloseButtonWrapper = _styles.styled.div(_templateObject2(), _styles.ui.space(-1), _styles.ui.space(-1));

exports.CloseButtonWrapper = CloseButtonWrapper;
var contentWithDismiss = (0, _styles.css)(_templateObject3(), _styles.ui.space(1.25));
exports.contentWithDismiss = contentWithDismiss;
var success = (0, _styles.css)(_templateObject4(), _styles.ui.background.green);
exports.success = success;
var warning = (0, _styles.css)(_templateObject5(), _styles.ui.background.yellow);
exports.warning = warning;
var critical = (0, _styles.css)(_templateObject6(), _styles.ui.background.red);
exports.critical = critical;
var info = (0, _styles.css)(_templateObject7(), _styles.ui.background.blue);
exports.info = info;

function getTextColor(status) {
  var color = 'colorText';

  switch (status) {
    case 'success':
      color = 'colorBackgroundGreenText';
      break;

    case 'warning':
      color = 'colorBackgroundYellowText';
      break;

    case 'critical':
      color = 'colorBackgroundRedText';
      break;

    case 'info':
      color = 'colorBackgroundBlueText';
      break;

    default:
      color = 'colorText';
      break;
  }

  return _styles.ui.get(color);
}
