// @wordpress/block-editor/build/components/block-patterns-list/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _components = require("@wordpress/components");

var _compose = require("@wordpress/compose");

var _i18n = require("@wordpress/i18n");

var _data = require("@wordpress/data");

var _blockPreview = _interopRequireDefault(require("../block-preview"));

var _inserterDraggableBlocks = _interopRequireDefault(require("../inserter-draggable-blocks"));

var _store = require("../../store");

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function BlockPattern(_ref) {
  var isDraggable = _ref.isDraggable,
      pattern = _ref.pattern,
      _onClick = _ref.onClick,
      composite = _ref.composite;
  var name = pattern.name,
      viewportWidth = pattern.viewportWidth;

  var _useSelect = (0, _data.useSelect)(function (select) {
    return select(_store.store).__experimentalGetParsedPattern(name);
  }, [name]),
      blocks = _useSelect.blocks;

  var instanceId = (0, _compose.useInstanceId)(BlockPattern);
  var descriptionId = "block-editor-block-patterns-list__item-description-".concat(instanceId);
  return (0, _element.createElement)(_inserterDraggableBlocks.default, {
    isEnabled: isDraggable,
    blocks: blocks
  }, function (_ref2) {
    var draggable = _ref2.draggable,
        onDragStart = _ref2.onDragStart,
        onDragEnd = _ref2.onDragEnd;
    return (0, _element.createElement)("div", {
      className: "block-editor-block-patterns-list__list-item",
      "aria-label": pattern.title,
      "aria-describedby": pattern.description ? descriptionId : undefined,
      draggable: draggable,
      onDragStart: onDragStart,
      onDragEnd: onDragEnd
    }, (0, _element.createElement)(_components.__unstableCompositeItem, (0, _extends2.default)({
      role: "option",
      as: "div"
    }, composite, {
      className: "block-editor-block-patterns-list__item",
      onClick: function onClick() {
        return _onClick(pattern, blocks);
      }
    }), (0, _element.createElement)(_blockPreview.default, {
      blocks: blocks,
      viewportWidth: viewportWidth
    }), (0, _element.createElement)("div", {
      className: "block-editor-block-patterns-list__item-title"
    }, pattern.title), !!pattern.description && (0, _element.createElement)(_components.VisuallyHidden, {
      id: descriptionId
    }, pattern.description)));
  });
}

function BlockPatternPlaceholder() {
  return (0, _element.createElement)("div", {
    className: "block-editor-block-patterns-list__item is-placeholder"
  });
}

function BlockPatternList(_ref3) {
  var isDraggable = _ref3.isDraggable,
      blockPatterns = _ref3.blockPatterns,
      shownPatterns = _ref3.shownPatterns,
      onClickPattern = _ref3.onClickPattern,
      orientation = _ref3.orientation,
      _ref3$label = _ref3.label,
      label = _ref3$label === void 0 ? (0, _i18n.__)('Block Patterns') : _ref3$label;
  var composite = (0, _components.__unstableUseCompositeState)({
    orientation: orientation
  });
  return (0, _element.createElement)(_components.__unstableComposite, (0, _extends2.default)({}, composite, {
    role: "listbox",
    className: "block-editor-block-patterns-list",
    "aria-label": label
  }), blockPatterns.map(function (pattern) {
    var isShown = shownPatterns.includes(pattern);
    return isShown ? (0, _element.createElement)(BlockPattern, {
      key: pattern.name,
      pattern: pattern,
      onClick: onClickPattern,
      isDraggable: isDraggable,
      composite: composite
    }) : (0, _element.createElement)(BlockPatternPlaceholder, {
      key: pattern.name
    });
  }));
}

var _default = BlockPatternList;
exports.default = _default;
