// @wp-g2/components/dist/cjs/Stepper/Stepper.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _icons = require("@wordpress/icons");

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _utils = require("@wp-g2/utils");

var _react = _interopRequireDefault(require("react"));

var _Button = require("../Button");

var _ControlGroup = require("../ControlGroup");

var styles = _interopRequireWildcard(require("./Stepper.styles"));

function Stepper(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'Stepper'),
      className = _useContextSystem.className,
      _useContextSystem$dir = _useContextSystem.direction,
      direction = _useContextSystem$dir === void 0 ? 'horizontal' : _useContextSystem$dir,
      _useContextSystem$onI = _useContextSystem.onIncrement,
      onIncrement = _useContextSystem$onI === void 0 ? _utils.noop : _useContextSystem$onI,
      _useContextSystem$onD = _useContextSystem.onDecrement,
      onDecrement = _useContextSystem$onD === void 0 ? _utils.noop : _useContextSystem$onD,
      size = _useContextSystem.size,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["className", "direction", "onIncrement", "onDecrement", "size"]);

  var isVertical = direction === 'vertical';
  var controlGroupDirection = isVertical ? 'column' : 'row';
  var classes = (0, _styles.cx)(!isVertical && styles.Stepper, className);
  var AddIcon = isVertical ? _icons.chevronUp : _icons.plus;
  var MinusIcon = isVertical ? _icons.chevronDown : _icons.minus;
  var buttons = isVertical ? [/*#__PURE__*/_react.default.createElement(ControlButton, {
    icon: AddIcon,
    isVertical: isVertical,
    key: "plus",
    onClick: onIncrement,
    size: size
  }), /*#__PURE__*/_react.default.createElement(ControlButton, {
    icon: MinusIcon,
    isVertical: isVertical,
    key: "minus",
    onClick: onDecrement,
    size: size
  })] : [/*#__PURE__*/_react.default.createElement(ControlButton, {
    icon: MinusIcon,
    isVertical: isVertical,
    key: "minus",
    onClick: onDecrement,
    size: size
  }), /*#__PURE__*/_react.default.createElement(ControlButton, {
    icon: AddIcon,
    isVertical: isVertical,
    key: "plus",
    onClick: onIncrement,
    size: size
  })];
  return /*#__PURE__*/_react.default.createElement(_ControlGroup.ControlGroup, Object.assign({
    className: classes,
    direction: controlGroupDirection,
    isItemBlock: true
  }, otherProps, {
    ref: forwardedRef
  }), buttons);
}

var ControlButton = /*#__PURE__*/_react.default.memo(function (_ref) {
  var isVertical = _ref.isVertical,
      size = _ref.size,
      props = (0, _objectWithoutProperties2.default)(_ref, ["isVertical", "size"]);
  var buttonSize = isVertical ? 'half' : 'medium';

  if (size === 'large') {
    buttonSize = isVertical ? 'halfLarge' : 'large';
  }

  if (size === 'small') {
    buttonSize = isVertical ? 'halfSmall' : 'small';
  }

  return /*#__PURE__*/_react.default.createElement(_Button.Button, Object.assign({}, _styles.ui.$('StepperButton'), props, {
    iconSize: 14,
    isBlock: true,
    isControl: true,
    size: buttonSize
  }));
});

var _default = (0, _context.contextConnect)(Stepper, 'Stepper');

exports.default = _default;
