// @wp-g2/components/dist/cjs/BaseModal/BaseModal.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _i18n = require("@wordpress/i18n");

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _reakit = require("reakit");

var _Portal = require("../Portal");

var _View = require("../View");

var _BaseModal = require("./BaseModal.Context");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t\tbottom: ", ";\n\t\tleft: ", ";\n\t\tposition: absolute;\n\t\tright: ", ";\n\t\ttop: ", ";\n\t"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t\tbackground: rgba(0, 0, 0, 0.5);\n\t\tbottom: 0;\n\t\tleft: 0;\n\t\topacity: 0;\n\t\toverflow-y: auto;\n\t\tpadding: ", ";\n\t\tperspective: 800px;\n\t\tposition: fixed;\n\t\tright: 0;\n\t\ttop: 0;\n\t\ttransition: ", ";\n\t\tz-index: ", ";\n\n\t\t&[data-enter] {\n\t\t\topacity: 1;\n\t\t}\n\n\t\t", ";\n\t"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

function BaseModal(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'BaseModal'),
      _useContextSystem$bac = _useContextSystem.backdropTransitionDuration,
      backdropTransitionDuration = _useContextSystem$bac === void 0 ? 250 : _useContextSystem$bac,
      children = _useContextSystem.children,
      className = _useContextSystem.className,
      _useContextSystem$lab = _useContextSystem.label,
      label = _useContextSystem$lab === void 0 ? (0, _i18n.__)('Modal') : _useContextSystem$lab,
      _useContextSystem$ren = _useContextSystem.renderTrigger,
      renderTrigger = _useContextSystem$ren === void 0 ? null : _useContextSystem$ren,
      state = _useContextSystem.state,
      _useContextSystem$tra = _useContextSystem.transitionTimingFunction,
      transitionTimingFunction = _useContextSystem$tra === void 0 ? 'ease-in-out' : _useContextSystem$tra,
      _useContextSystem$tri = _useContextSystem.trigger,
      trigger = _useContextSystem$tri === void 0 ? null : _useContextSystem$tri,
      _useContextSystem$vis = _useContextSystem.visible,
      visible = _useContextSystem$vis === void 0 ? false : _useContextSystem$vis,
      zIndex = _useContextSystem.zIndex,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["backdropTransitionDuration", "children", "className", "label", "renderTrigger", "state", "transitionTimingFunction", "trigger", "visible", "zIndex"]);

  var _dialog = (0, _reakit.useDialogState)({
    animated: true,
    visible: visible
  });

  var dialog = state || _dialog;

  var _useModalState = (0, _BaseModal.useModalState)(dialog),
      isUnderLayer = _useModalState.isUnderLayer;

  var backdropStyles = (0, _styles.css)(_templateObject(), (0, _styles.space)(4), "opacity " + backdropTransitionDuration + "ms\n    " + transitionTimingFunction, zIndex || (0, _styles.getZIndex)('Modal'), (0, _styles.reducedMotion)("\n\t\t\t&[data-leave] {\n\t\t\t\tdisplay: none;\n\t\t\t}\n\t\t"));
  var dialogWrapperStyles = (0, _styles.css)(_templateObject2(), (0, _styles.space)(4), (0, _styles.space)(4), (0, _styles.space)(4), (0, _styles.space)(4));

  var dialogProps = _objectSpread(_objectSpread({}, otherProps), dialog);

  if (isUnderLayer) {
    dialogProps['data-underlayer'] = '';
  }

  var triggerProp = trigger || renderTrigger;
  return /*#__PURE__*/_react.default.createElement(_BaseModal.ModalContext.Provider, {
    value: {
      dialog: dialog
    }
  }, typeof triggerProp === 'function' ? triggerProp(dialog) : triggerProp ? triggerProp : null, /*#__PURE__*/_react.default.createElement(_Portal.Portal, null, /*#__PURE__*/_react.default.createElement(_reakit.DialogBackdrop, Object.assign({}, dialog, {
    className: (0, _styles.cx)(backdropStyles)
  }), /*#__PURE__*/_react.default.createElement(_View.View, {
    className: dialogWrapperStyles
  }, /*#__PURE__*/_react.default.createElement(_reakit.Dialog, Object.assign({
    "aria-label": label,
    className: className,
    ref: forwardedRef,
    tabIndex: 0
  }, dialog, dialogProps), (dialog.visible || dialog.animating) && children)))));
}

var _default = (0, _context.contextConnect)(BaseModal, 'BaseModal');

exports.default = _default;
