// @wordpress/block-editor/build/components/block-mover/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.BlockMover = exports.BLOCK_MOVER_DIRECTION_BOTTOM = exports.BLOCK_MOVER_DIRECTION_TOP = void 0;

var _element = require("@wordpress/element");

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _lodash = require("lodash");

var _reactNative = require("react-native");

var _i18n = require("@wordpress/i18n");

var _components = require("@wordpress/components");

var _compose = require("@wordpress/compose");

var _data = require("@wordpress/data");

var _moverDescription = require("./mover-description");

var _store = require("../../store");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var BLOCK_MOVER_DIRECTION_TOP = 'blockPageMoverOptions-moveToTop';
exports.BLOCK_MOVER_DIRECTION_TOP = BLOCK_MOVER_DIRECTION_TOP;
var BLOCK_MOVER_DIRECTION_BOTTOM = 'blockPageMoverOptions-moveToBottom';
exports.BLOCK_MOVER_DIRECTION_BOTTOM = BLOCK_MOVER_DIRECTION_BOTTOM;

var BlockMover = function BlockMover(_ref) {
  var isFirst = _ref.isFirst,
      isLast = _ref.isLast,
      isLocked = _ref.isLocked,
      onMoveDown = _ref.onMoveDown,
      onMoveUp = _ref.onMoveUp,
      onLongMove = _ref.onLongMove,
      firstIndex = _ref.firstIndex,
      numberOfBlocks = _ref.numberOfBlocks,
      rootClientId = _ref.rootClientId,
      isStackedHorizontally = _ref.isStackedHorizontally;
  var pickerRef = (0, _element.useRef)();

  var _useState = (0, _element.useState)(undefined),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      blockPageMoverState = _useState2[0],
      setBlockPageMoverState = _useState2[1];

  var showBlockPageMover = function showBlockPageMover(direction) {
    return function () {
      if (!pickerRef.current) {
        setBlockPageMoverState(undefined);
        return;
      }

      setBlockPageMoverState(direction);
      pickerRef.current.presentPicker();
    };
  };

  var _getMoversSetup = (0, _moverDescription.getMoversSetup)(isStackedHorizontally, {
    firstIndex: firstIndex
  }),
      _getMoversSetup$descr = _getMoversSetup.description,
      backwardButtonHint = _getMoversSetup$descr.backwardButtonHint,
      forwardButtonHint = _getMoversSetup$descr.forwardButtonHint,
      firstBlockTitle = _getMoversSetup$descr.firstBlockTitle,
      lastBlockTitle = _getMoversSetup$descr.lastBlockTitle,
      _getMoversSetup$icon = _getMoversSetup.icon,
      backwardButtonIcon = _getMoversSetup$icon.backward,
      forwardButtonIcon = _getMoversSetup$icon.forward,
      _getMoversSetup$title = _getMoversSetup.title,
      backwardButtonTitle = _getMoversSetup$title.backward,
      forwardButtonTitle = _getMoversSetup$title.forward;

  var blockPageMoverOptions = [{
    icon: backwardButtonIcon,
    label: (0, _i18n.__)('Move to top'),
    value: BLOCK_MOVER_DIRECTION_TOP,
    onSelect: function onSelect() {
      onLongMove()(0);
    }
  }, {
    icon: forwardButtonIcon,
    label: (0, _i18n.__)('Move to bottom'),
    value: BLOCK_MOVER_DIRECTION_BOTTOM,
    onSelect: function onSelect() {
      onLongMove()(numberOfBlocks);
    }
  }].filter(function (el) {
    return el.value === blockPageMoverState;
  });

  var onPickerSelect = function onPickerSelect(value) {
    var option = blockPageMoverOptions.find(function (el) {
      return el.value === value;
    });
    if (option && option.onSelect) option.onSelect();
  };

  if (isLocked || isFirst && isLast && !rootClientId) {
    return null;
  }

  return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_components.ToolbarButton, {
    title: !isFirst ? backwardButtonTitle : firstBlockTitle,
    isDisabled: isFirst,
    onClick: onMoveUp,
    onLongPress: showBlockPageMover(BLOCK_MOVER_DIRECTION_TOP),
    icon: backwardButtonIcon,
    extraProps: {
      hint: backwardButtonHint
    }
  }), (0, _element.createElement)(_components.ToolbarButton, {
    title: !isLast ? forwardButtonTitle : lastBlockTitle,
    isDisabled: isLast,
    onClick: onMoveDown,
    onLongPress: showBlockPageMover(BLOCK_MOVER_DIRECTION_BOTTOM),
    icon: forwardButtonIcon,
    extraProps: {
      hint: forwardButtonHint
    }
  }), (0, _element.createElement)(_components.Picker, {
    ref: pickerRef,
    options: blockPageMoverOptions,
    onChange: onPickerSelect,
    title: (0, _i18n.__)('Change block position'),
    leftAlign: true,
    hideCancelButton: _reactNative.Platform.OS !== 'ios'
  }));
};

exports.BlockMover = BlockMover;

var _default = (0, _compose.compose)((0, _data.withSelect)(function (select, _ref2) {
  var clientIds = _ref2.clientIds;

  var _select = select(_store.store),
      getBlockIndex = _select.getBlockIndex,
      getTemplateLock = _select.getTemplateLock,
      getBlockRootClientId = _select.getBlockRootClientId,
      getBlockOrder = _select.getBlockOrder;

  var normalizedClientIds = (0, _lodash.castArray)(clientIds);
  var firstClientId = (0, _lodash.first)(normalizedClientIds);
  var rootClientId = getBlockRootClientId(firstClientId);
  var blockOrder = getBlockOrder(rootClientId);
  var firstIndex = getBlockIndex(firstClientId, rootClientId);
  var lastIndex = getBlockIndex((0, _lodash.last)(normalizedClientIds), rootClientId);
  return {
    firstIndex: firstIndex,
    numberOfBlocks: blockOrder.length - 1,
    isFirst: firstIndex === 0,
    isLast: lastIndex === blockOrder.length - 1,
    isLocked: getTemplateLock(rootClientId) === 'all',
    rootClientId: rootClientId
  };
}), (0, _data.withDispatch)(function (dispatch, _ref3) {
  var clientIds = _ref3.clientIds,
      rootClientId = _ref3.rootClientId;

  var _dispatch = dispatch(_store.store),
      moveBlocksDown = _dispatch.moveBlocksDown,
      moveBlocksUp = _dispatch.moveBlocksUp,
      moveBlocksToPosition = _dispatch.moveBlocksToPosition;

  return {
    onMoveDown: (0, _lodash.partial)(moveBlocksDown, clientIds, rootClientId),
    onMoveUp: (0, _lodash.partial)(moveBlocksUp, clientIds, rootClientId),
    onLongMove: function onLongMove(targetIndex) {
      return (0, _lodash.partial)(moveBlocksToPosition, clientIds, rootClientId, targetIndex);
    }
  };
}), _compose.withInstanceId)(BlockMover);

exports.default = _default;
