// @wp-g2/components/dist/cjs/ColorCircle/ColorCircle.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _icons = require("@wordpress/icons");

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _utils = require("@wp-g2/utils");

var _react = _interopRequireDefault(require("react"));

var _Icon = require("../Icon");

var styles = _interopRequireWildcard(require("./ColorCircle.styles"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var CheckboxIconView = styles.CheckboxIconView,
    ColorCircleView = styles.ColorCircleView;

function ColorCircle(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'ColorCircle'),
      colorProp = _useContextSystem.color,
      _useContextSystem$siz = _useContextSystem.size,
      size = _useContextSystem$siz === void 0 ? 'medium' : _useContextSystem$siz,
      _useContextSystem$isA = _useContextSystem.isActive,
      isActive = _useContextSystem$isA === void 0 ? false : _useContextSystem$isA,
      _useContextSystem$isI = _useContextSystem.isInteractive,
      isInteractive = _useContextSystem$isI === void 0 ? false : _useContextSystem$isI,
      _useContextSystem$var = _useContextSystem.variant,
      variant = _useContextSystem$var === void 0 ? 'default' : _useContextSystem$var,
      _useContextSystem$sty = _useContextSystem.style,
      style = _useContextSystem$sty === void 0 ? {} : _useContextSystem$sty,
      className = _useContextSystem.className,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["color", "size", "isActive", "isInteractive", "variant", "style", "className"]);

  var backgroundColor = (0, _utils.colorize)(colorProp).toRgbString();
  var classes = (0, _styles.cx)(styles[size], styles[variant], styles.expand, isActive && styles.active, isInteractive && styles.interactive, className);
  var iconColor = (0, _utils.getOptimalTextColor)(backgroundColor);
  return /*#__PURE__*/_react.default.createElement(ColorCircleView, Object.assign({
    "data-active": isActive,
    style: _objectSpread(_objectSpread({}, style), {}, {
      backgroundColor: backgroundColor
    })
  }, otherProps, {
    className: classes,
    ref: forwardedRef
  }), /*#__PURE__*/_react.default.createElement(CheckboxIconView, Object.assign({}, _styles.ui.$('CheckboxIcon'), {
    "aria-hidden": true
  }), /*#__PURE__*/_react.default.createElement(_Icon.Icon, {
    fill: iconColor,
    icon: _icons.check,
    size: 12
  })));
}

var _default = (0, _context.contextConnect)(ColorCircle, 'ColorCircle');

exports.default = _default;
