// @wp-g2/components/dist/cjs/ColorCircle/ColorCircle.styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CheckboxIconView = exports.expand = exports.pill = exports.interactive = exports.active = exports.small = exports.medium = exports.large = exports.ColorCircleView = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _styles = require("@wp-g2/styles");

function _templateObject9() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\t", ";\n\tbottom: 0;\n\n\tleft: 0;\n\topacity: 0;\n\tpointer-events: none;\n\tposition: absolute;\n\tright: 0;\n\ttop: 0;\n\ttransition: opacity ", " linear;\n\n\t[data-active='true'] > & {\n\t\topacity: 1;\n\t}\n"]);

  _templateObject9 = function () {
    return data;
  };

  return data;
}

function _templateObject8() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t&[aria-expanded='true'] {\n\t\t", ";\n\t}\n"]);

  _templateObject8 = function () {
    return data;
  };

  return data;
}

function _templateObject7() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tborder-radius: 9999px;\n\twidth: 100%;\n"]);

  _templateObject7 = function () {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\tcursor: pointer;\n\n\t&::before {\n\t\t", ";\n\t\tborder: 2px solid ", ";\n\t\tcontent: '';\n\t\theight: calc(100% + 8px);\n\t\tleft: -4px;\n\t\tposition: absolute;\n\t\ttop: -4px;\n\t\twidth: calc(100% + 8px);\n\t\topacity: 0;\n\t}\n\n\t&:active {\n\t\topacity: 0.6;\n\t}\n\n\t&:focus {\n\t\t&::before {\n\t\t\topacity: 1;\n\t\t}\n\t}\n"]);

  _templateObject6 = function () {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteralLoose2.default)([""]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\theight: 20px;\n\twidth: 20px;\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\theight: 24px;\n\twidth: 24px;\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\theight: 30px;\n\twidth: 30px;\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\tbox-shadow: 0 0 0 1px rgba(0, 0, 0, 0.1) inset;\n\toutline: none;\n\tposition: relative;\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var ColorCircleView = _styles.styled.div(_templateObject(), _styles.ui.borderRadius.circle);

exports.ColorCircleView = ColorCircleView;
var large = (0, _styles.css)(_templateObject2());
exports.large = large;
var medium = (0, _styles.css)(_templateObject3());
exports.medium = medium;
var small = (0, _styles.css)(_templateObject4());
exports.small = small;
var active = (0, _styles.css)(_templateObject5());
exports.active = active;
var interactive = (0, _styles.css)(_templateObject6(), _styles.ui.animation.default, _styles.ui.borderRadius.circle, _styles.ui.color.admin);
exports.interactive = interactive;
var pill = (0, _styles.css)(_templateObject7());
exports.pill = pill;
var expand = (0, _styles.css)(_templateObject8(), active);
exports.expand = expand;

var CheckboxIconView = _styles.styled.div(_templateObject9(), _styles.ui.font.color.white, _styles.ui.alignment.content.center, _styles.ui.get('transitionDurationFastest'));

exports.CheckboxIconView = CheckboxIconView;
