// @wordpress/block-editor/build/components/inserter/block-patterns-tab.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _lodash = require("lodash");

var _i18n = require("@wordpress/i18n");

var _compose = require("@wordpress/compose");

var _patternPanel = _interopRequireDefault(require("./pattern-panel"));

var _usePatternsState3 = _interopRequireDefault(require("./hooks/use-patterns-state"));

var _blockPatternsList = _interopRequireDefault(require("../block-patterns-list"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function BlockPatternsCategory(_ref) {
  var rootClientId = _ref.rootClientId,
      onInsert = _ref.onInsert,
      selectedCategory = _ref.selectedCategory,
      onClickCategory = _ref.onClickCategory;

  var _usePatternsState = (0, _usePatternsState3.default)(onInsert, rootClientId),
      _usePatternsState2 = (0, _slicedToArray2.default)(_usePatternsState, 3),
      allPatterns = _usePatternsState2[0],
      allCategories = _usePatternsState2[1],
      onClick = _usePatternsState2[2]; // Remove any empty categories


  var populatedCategories = (0, _element.useMemo)(function () {
    return allCategories.filter(function (category) {
      return allPatterns.some(function (pattern) {
        var _pattern$categories;

        return (_pattern$categories = pattern.categories) === null || _pattern$categories === void 0 ? void 0 : _pattern$categories.includes(category.name);
      });
    });
  }, [allPatterns, allCategories]);
  var patternCategory = selectedCategory ? selectedCategory : populatedCategories[0];
  (0, _element.useEffect)(function () {
    if (allPatterns.some(function (pattern) {
      return getPatternIndex(pattern) === Infinity;
    }) && !populatedCategories.find(function (category) {
      return category.name === 'uncategorized';
    })) {
      populatedCategories.push({
        name: 'uncategorized',
        label: (0, _i18n._x)('Uncategorized')
      });
    }
  }, [populatedCategories, allPatterns]);
  var getPatternIndex = (0, _element.useCallback)(function (pattern) {
    if (!pattern.categories || !pattern.categories.length) {
      return Infinity;
    }

    var indexedCategories = (0, _lodash.fromPairs)(populatedCategories.map(function (_ref2, index) {
      var name = _ref2.name;
      return [name, index];
    }));
    return Math.min.apply(Math, (0, _toConsumableArray2.default)(pattern.categories.map(function (cat) {
      return indexedCategories[cat] !== undefined ? indexedCategories[cat] : Infinity;
    })));
  }, [populatedCategories]);
  var currentCategoryPatterns = (0, _element.useMemo)(function () {
    return allPatterns.filter(function (pattern) {
      return patternCategory.name === 'uncategorized' ? getPatternIndex(pattern) === Infinity : pattern.categories && pattern.categories.includes(patternCategory.name);
    });
  }, [allPatterns, patternCategory]); // Ordering the patterns is important for the async rendering.

  var orderedPatterns = (0, _element.useMemo)(function () {
    return currentCategoryPatterns.sort(function (a, b) {
      return getPatternIndex(a) - getPatternIndex(b);
    });
  }, [currentCategoryPatterns, getPatternIndex]);
  var currentShownPatterns = (0, _compose.useAsyncList)(orderedPatterns);
  return (0, _element.createElement)(_element.Fragment, null, !!currentCategoryPatterns.length && (0, _element.createElement)(_patternPanel.default, {
    selectedCategory: patternCategory,
    patternCategories: populatedCategories,
    onClickCategory: onClickCategory
  }, (0, _element.createElement)(_blockPatternsList.default, {
    shownPatterns: currentShownPatterns,
    blockPatterns: currentCategoryPatterns,
    onClickPattern: onClick,
    label: patternCategory.label,
    orientation: "vertical",
    isDraggable: true
  })));
}

function BlockPatternsTabs(_ref3) {
  var rootClientId = _ref3.rootClientId,
      onInsert = _ref3.onInsert,
      onClickCategory = _ref3.onClickCategory,
      selectedCategory = _ref3.selectedCategory;
  return (0, _element.createElement)(BlockPatternsCategory, {
    rootClientId: rootClientId,
    selectedCategory: selectedCategory,
    onInsert: onInsert,
    onClickCategory: onClickCategory
  });
}

var _default = BlockPatternsTabs;
exports.default = _default;
