// @wp-g2/components/dist/cjs/Button/Button.styles.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.currentColor = exports.subtleControl = exports.subtleControlActive = exports.control = exports.subtle = exports.plainLink = exports.link = exports.tertiary = exports.secondary = exports.primary = exports.icon = exports.halfSmall = exports.halfLarge = exports.half = exports.xxSmall = exports.xSmall = exports.small = exports.large = exports.Button = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _styles = require("@wp-g2/styles");

var baseButtonStyles = _interopRequireWildcard(require("../BaseButton/BaseButton.styles"));

function _templateObject16() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\t", ";\n"]);

  _templateObject16 = function () {
    return data;
  };

  return data;
}

function _templateObject15() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t&[data-active='true'] {\n\t\t", ";\n\t\t", ";\n\n\t\tbackground: ", ";\n\t\tborder-color: ", ";\n\t\tcolor: ", ";\n\n\t\t&:hover {\n\t\t\tbackground: ", ";\n\t\t}\n\t\t&:focus {\n\t\t\tbackground: ", ";\n\t\t\tbox-shadow: ", ";\n\t\t\tborder-color: ", ";\n\t\t}\n\t\t&:active {\n\t\t\tbackground: ", ";\n\t\t}\n\t}\n"]);

  _templateObject15 = function () {
    return data;
  };

  return data;
}

function _templateObject14() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\tbackground: none;\n\tborder-color: transparent;\n\tcolor: ", ";\n\n\t&:hover,\n\t&:active,\n\t&:focus,\n\t&[data-focused='true'] {\n\t\tbackground-color: transparent;\n\t\ttext-decoration: underline;\n\t}\n\n\t&:active {\n\t\tcolor: ", ";\n\t}\n\n\t&[data-destructive='true'] {\n\t\tcolor: ", ";\n\n\t\t&:active {\n\t\t\tcolor: ", ";\n\t\t}\n\t}\n"]);

  _templateObject14 = function () {
    return data;
  };

  return data;
}

function _templateObject13() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbackground: none;\n\tborder-color: transparent;\n\tcolor: ", ";\n\n\t&:active {\n\t\tcolor: ", ";\n\t}\n\n\t&[data-destructive='true'] {\n\t\tcolor: ", ";\n\n\t\t&:active {\n\t\t\tcolor: ", ";\n\t\t}\n\t}\n"]);

  _templateObject13 = function () {
    return data;
  };

  return data;
}

function _templateObject12() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbackground-color: ", ";\n\tborder-color: ", ";\n\tcolor: ", ";\n\n\t&:hover {\n\t\tbackground-color: ", ";\n\t\tborder-color: ", ";\n\t}\n\n\t&:focus,\n\t&[data-focused='true'] {\n\t\tbackground-color: ", ";\n\t\tborder-color: ", ";\n\t\tcolor: ", ";\n\t}\n\n\t&:active {\n\t\tbackground-color: ", ";\n\t\tborder-color: ", ";\n\t\tcolor: ", ";\n\t}\n\n\t&[data-destructive='true'] {\n\t\tcolor: ", ";\n\n\t\t&:hover,\n\t\t&:focus {\n\t\t\tborder-color: ", ";\n\t\t}\n\n\t\t&:active {\n\t\t\tcolor: ", ";\n\t\t}\n\t}\n"]);

  _templateObject12 = function () {
    return data;
  };

  return data;
}

function _templateObject11() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbackground-color: ", ";\n\tborder-color: ", ";\n\tcolor: ", ";\n\n\t&:hover {\n\t\tbackground-color: ", ";\n\t\tborder-color: ", ";\n\t}\n\n\t&:focus,\n\t&[data-focused='true'] {\n\t\tbackground-color: ", ";\n\t\tborder-color: ", ";\n\t\tcolor: ", ";\n\t}\n\n\t&:active {\n\t\tbackground-color: ", ";\n\t\tborder-color: ", ";\n\t\tcolor: ", ";\n\t}\n\n\t&[data-destructive='true'] {\n\t\tborder-color: ", ";\n\t\tcolor: ", ";\n\n\t\t&:hover,\n\t\t&:active,\n\t\t&:focus,\n\t\t&[data-focused='true'] {\n\t\t\tborder-color: ", ";\n\t\t\tcolor: ", ";\n\t\t}\n\n\t\t&:active {\n\t\t\tcolor: ", ";\n\t\t}\n\t}\n"]);

  _templateObject11 = function () {
    return data;
  };

  return data;
}

function _templateObject10() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbackground-color: ", ";\n\tborder-color: ", ";\n\tcolor: ", ";\n\n\t&:active {\n\t\tcolor: ", ";\n\t}\n\n\t&:hover {\n\t\tbackground-color: ", ";\n\t\tborder-color: ", ";\n\t\tcolor: ", ";\n\t}\n\n\t&:focus,\n\t&[data-focused='true'] {\n\t\tbackground-color: ", ";\n\t\tborder-color: ", ";\n\t\tcolor: ", ";\n\t}\n\n\t&:active {\n\t\tbackground-color: ", ";\n\t\tborder-color: ", ";\n\t}\n\n\t&[data-destructive='true'] {\n\t\tbackground-color: ", ";\n\t\tborder-color: ", ";\n\n\t\t&:hover,\n\t\t&:focus,\n\t\t&[data-focused='true'] {\n\t\t\tbackground-color: ", ";\n\t\t\tborder-color: ", ";\n\t\t}\n\n\t\t&:active {\n\t\t\tbackground-color: ", ";\n\t\t}\n\t}\n"]);

  _templateObject10 = function () {
    return data;
  };

  return data;
}

function _templateObject9() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n"]);

  _templateObject9 = function () {
    return data;
  };

  return data;
}

function _templateObject8() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tmin-height: calc(", " / 2);\n\n\t&[data-icon='true'] {\n\t\tmin-width: ", ";\n\t}\n"]);

  _templateObject8 = function () {
    return data;
  };

  return data;
}

function _templateObject7() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tmin-height: calc(", " / 2);\n\n\t&[data-icon='true'] {\n\t\tmin-width: ", ";\n\t}\n"]);

  _templateObject7 = function () {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tmin-height: calc(", " / 2);\n\n\t&[data-icon='true'] {\n\t\tmin-width: ", ";\n\t}\n"]);

  _templateObject6 = function () {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\tmin-height: ", ";\n\n\t&[data-icon='true'] {\n\t\tmin-width: ", ";\n\t}\n"]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\tmin-height: ", ";\n\n\t&[data-icon='true'] {\n\t\tmin-width: ", ";\n\t}\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\tmin-height: ", ";\n\n\t&[data-icon='true'] {\n\t\tmin-width: ", ";\n\t}\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tmin-height: ", ";\n\n\t&[data-icon='true'] {\n\t\tmin-width: ", ";\n\t}\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tcolor: ", ";\n\tfont-weight: 600;\n\tpadding-left: ", ";\n\tpadding-right: ", ";\n\n\t&:active {\n\t\tcolor: ", ";\n\t}\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var Button = (0, _styles.css)(_templateObject(), _styles.ui.get('buttonTextColor'), _styles.ui.get('buttonPaddingX'), _styles.ui.get('buttonPaddingX'), _styles.ui.get('buttonTextColorActive'));
exports.Button = Button;
var large = (0, _styles.css)(_templateObject2(), _styles.ui.get('controlHeightLarge'), _styles.ui.get('controlHeightLarge'));
exports.large = large;
var small = (0, _styles.css)(_templateObject3(), _styles.ui.padding.x(_styles.ui.get('controlPaddingX')), _styles.ui.get('controlHeightSmall'), _styles.ui.get('controlHeightSmall'));
exports.small = small;
var xSmall = (0, _styles.css)(_templateObject4(), _styles.ui.padding.x(_styles.ui.get('controlPaddingXSmall')), _styles.ui.get('controlHeightXSmall'), _styles.ui.get('controlHeightXSmall'));
exports.xSmall = xSmall;
var xxSmall = (0, _styles.css)(_templateObject5(), _styles.ui.padding.x(_styles.ui.get('controlHeightXXSmall')), _styles.ui.get('controlHeightXXSmall'), _styles.ui.get('controlHeightXXSmall'));
exports.xxSmall = xxSmall;
var half = (0, _styles.css)(_templateObject6(), _styles.ui.get('controlHeight'), _styles.ui.get('controlHeight'));
exports.half = half;
var halfLarge = (0, _styles.css)(_templateObject7(), _styles.ui.get('controlHeightLarge'), _styles.ui.get('controlHeightLarge'));
exports.halfLarge = halfLarge;
var halfSmall = (0, _styles.css)(_templateObject8(), _styles.ui.get('controlHeightSmall'), _styles.ui.get('controlHeightSmall'));
exports.halfSmall = halfSmall;
var icon = (0, _styles.css)(_templateObject9(), _styles.ui.padding(0));
exports.icon = icon;
var primary = (0, _styles.css)(_templateObject10(), _styles.ui.get('buttonPrimaryColor'), _styles.ui.get('buttonPrimaryBorderColor'), _styles.ui.get('buttonPrimaryTextColor'), _styles.ui.get('buttonPrimaryTextColorActive'), _styles.ui.get('buttonPrimaryColorHover'), _styles.ui.get('buttonPrimaryBorderColorHover'), _styles.ui.get('buttonPrimaryTextColorHover'), _styles.ui.get('buttonPrimaryColorFocus'), _styles.ui.get('buttonPrimaryBorderColorFocus'), _styles.ui.get('buttonPrimaryTextColorFocus'), _styles.ui.get('buttonPrimaryColorActive'), _styles.ui.get('buttonPrimaryBorderColorActive'), _styles.ui.color.destructive, _styles.ui.color.destructive, _styles.ui.color.destructive, _styles.ui.color.destructive, _styles.ui.color.text);
exports.primary = primary;
var secondary = (0, _styles.css)(_templateObject11(), _styles.ui.get('buttonSecondaryColor'), _styles.ui.get('buttonSecondaryBorderColor'), _styles.ui.get('buttonSecondaryTextColor'), _styles.ui.get('buttonSecondaryColorHover'), _styles.ui.get('buttonSecondaryBorderColorHover'), _styles.ui.get('buttonSecondaryColorFocus'), _styles.ui.get('buttonSecondaryBorderColorFocus'), _styles.ui.get('buttonSecondaryTextColorFocus'), _styles.ui.get('buttonSecondaryColorActive'), _styles.ui.get('buttonSecondaryBorderColorActive'), _styles.ui.get('buttonSecondaryTextColorActive'), _styles.ui.color.destructive, _styles.ui.color.destructive, _styles.ui.color.destructive, _styles.ui.color.destructive, _styles.ui.color.text);
exports.secondary = secondary;
var tertiary = (0, _styles.css)(_templateObject12(), _styles.ui.get('buttonTertiaryColor'), _styles.ui.get('buttonTertiaryBorderColor'), _styles.ui.get('buttonTertiaryTextColor'), _styles.ui.get('buttonTertiaryColorHover'), _styles.ui.get('buttonTertiaryBorderColorHover'), _styles.ui.get('buttonTertiaryColorFocus'), _styles.ui.get('buttonTertiaryBorderColorFocus'), _styles.ui.get('buttonTertiaryTextColorFocus'), _styles.ui.get('buttonTertiaryColorActive'), _styles.ui.get('buttonTertiaryBorderColorActive'), _styles.ui.get('buttonTertiaryTextColorActive'), _styles.ui.color.destructive, _styles.ui.color.destructive, _styles.ui.color.text);
exports.tertiary = tertiary;
var link = (0, _styles.css)(_templateObject13(), _styles.ui.get('linkColor'), _styles.ui.get('linkColorActive'), _styles.ui.color.destructive, _styles.ui.color.text);
exports.link = link;
var plainLink = (0, _styles.css)(_templateObject14(), _styles.ui.padding.x(0), _styles.ui.get('linkColor'), _styles.ui.get('linkColorActive'), _styles.ui.color.destructive, _styles.ui.color.text);
exports.plainLink = plainLink;
var subtle = baseButtonStyles.subtle;
exports.subtle = subtle;
var control = baseButtonStyles.control;
exports.control = control;
var subtleControlActiveTransition = (0, _styles.css)({
  transition: ['all', _styles.ui.get('transitionDuration'), _styles.ui.get('transitionTimingFunctionControl')].join(' ')
});
var subtleControlActive = (0, _styles.css)(_templateObject15(), subtleControlActiveTransition, _styles.ui.zIndex('ControlFocus'), _styles.ui.get('buttonControlActiveStateColor'), _styles.ui.get('buttonControlActiveStateColor'), _styles.ui.get('buttonControlActiveStateTextColor'), _styles.ui.get('buttonControlActiveStateColorHover'), _styles.ui.get('buttonControlActiveStateColorFocus'), _styles.ui.get('buttonControlActiveStateBoxShadowFocus'), _styles.ui.get('buttonControlActiveStateBorderColorFocus'), _styles.ui.get('buttonControlActiveStateColorActive'));
exports.subtleControlActive = subtleControlActive;
var subtleControl = (0, _styles.css)(_templateObject16(), baseButtonStyles.subtleControl, subtleControlActive);
exports.subtleControl = subtleControl;
var currentColor = baseButtonStyles.currentColor;
exports.currentColor = currentColor;
