// @wp-g2/components/dist/cjs/Button/Button.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _utils = require("@wp-g2/utils");

var _react = _interopRequireDefault(require("react"));

var _BaseButton = require("../BaseButton");

var _ButtonGroup = require("../ButtonGroup");

var styles = _interopRequireWildcard(require("./Button.styles"));

/**
 * @typedef {'primary' | 'secondary' | 'tertiary' | 'plain' | 'link'} ButtonVariant
 */

/**
 * @typedef OwnProps
 * @property {ButtonVariant} [variant='secondary'] Determinds the `Button` variant to render.
 */

/**
 * @typedef {import('../BaseButton/types').Props & OwnProps} Props
 */

/**
 * @param {import('@wp-g2/create-styles').ViewOwnProps<Props, 'button'>} props
 * @param {import('react').Ref<any>} forwardedRef
 */
function Button(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'Button'),
      children = _useContextSystem.children,
      className = _useContextSystem.className,
      currentColor = _useContextSystem.currentColor,
      icon = _useContextSystem.icon,
      _useContextSystem$isA = _useContextSystem.isActive,
      isActiveProp = _useContextSystem$isA === void 0 ? false : _useContextSystem$isA,
      _useContextSystem$isC = _useContextSystem.isControl,
      isControl = _useContextSystem$isC === void 0 ? false : _useContextSystem$isC,
      _useContextSystem$isS = _useContextSystem.isSubtle,
      isSubtle = _useContextSystem$isS === void 0 ? false : _useContextSystem$isS,
      _useContextSystem$onC = _useContextSystem.onClick,
      onClick = _useContextSystem$onC === void 0 ? _utils.noop : _useContextSystem$onC,
      _useContextSystem$siz = _useContextSystem.size,
      size = _useContextSystem$siz === void 0 ? 'medium' : _useContextSystem$siz,
      _useContextSystem$var = _useContextSystem.variant,
      variant = _useContextSystem$var === void 0 ? 'secondary' : _useContextSystem$var,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["children", "className", "currentColor", "icon", "isActive", "isControl", "isSubtle", "onClick", "size", "variant"]);

  var _useButtonGroupContex = (0, _ButtonGroup.useButtonGroupContext)(),
      buttonGroup = _useButtonGroupContex.buttonGroup,
      enableSelectNone = _useButtonGroupContex.enableSelectNone;

  var isWithinButtonGroup = !!buttonGroup;
  var isButtonGroupActive = isWithinButtonGroup && (buttonGroup == null ? void 0 : buttonGroup.state) === otherProps.value;

  var handleOnClickWithinButtonGroup = _react.default.useCallback(function (event) {
    if (isWithinButtonGroup && enableSelectNone && isButtonGroupActive && buttonGroup) {
      event.preventDefault();
      event.stopPropagation();
      buttonGroup.setState(undefined);
    }
  }, [buttonGroup, enableSelectNone, isButtonGroupActive, isWithinButtonGroup]);

  var handleOnClick = _react.default.useCallback(function (event) {
    onClick(event);
    handleOnClickWithinButtonGroup(event);
  }, [handleOnClickWithinButtonGroup, onClick]);

  var classes = (0, _styles.cx)(styles.Button, styles[variant], styles[size], isControl && styles.control, isSubtle && styles.subtle, isSubtle && isControl && styles.subtleControl, isButtonGroupActive && styles.subtleControlActive, !!icon && !children && styles.icon, currentColor && styles.currentColor, className);
  return /*#__PURE__*/_react.default.createElement(_BaseButton.BaseButton, Object.assign({
    className: classes,
    icon: icon,
    isActive: isActiveProp || isButtonGroupActive,
    onClick: handleOnClick,
    ref: forwardedRef
  }, otherProps), children);
}
/**
 * `Button` is a component used to trigger an action or event, such as submitting a Form, opening a Dialog, canceling an action, or performing a delete operation.
 *
 * @example
 * ```jsx
 * import { Button } from `@wp-g2/components`
 *
 * function Example() {
 *   return <Button variant="secondary">Let It Go</Button>;
 * }
 * ```
 */


var ConnectedButton = (0, _context.contextConnect)(Button, 'Button');
var _default = ConnectedButton;
exports.default = _default;
