// @wordpress/components/build-module/mobile/utils/alignments.native.js
export var WIDE_ALIGNMENTS = {
  alignments: {
    wide: 'wide',
    full: 'full'
  },
  // `innerContainers`: Group of blocks based on `InnerBlocks` component,
  // used to nest other blocks inside
  innerContainers: ['core/group', 'core/columns', 'core/column', 'core/buttons', 'core/button'],
  excludeBlocks: ['core/heading']
};
export var ALIGNMENT_BREAKPOINTS = {
  wide: 1024,
  large: 820,
  medium: 768,
  small: 680,
  mobile: 480
};

var isFullWidth = function isFullWidth(align) {
  return align === WIDE_ALIGNMENTS.alignments.full;
};

var isWideWidth = function isWideWidth(align) {
  return align === WIDE_ALIGNMENTS.alignments.wide;
};

var isWider = function isWider(width, breakpoint) {
  return width > ALIGNMENT_BREAKPOINTS[breakpoint];
};

var isContainerRelated = function isContainerRelated(blockName) {
  return WIDE_ALIGNMENTS.innerContainers.includes(blockName);
};

export var alignmentHelpers = {
  isFullWidth: isFullWidth,
  isWideWidth: isWideWidth,
  isWider: isWider,
  isContainerRelated: isContainerRelated
};
