// @wp-g2/components/dist/cjs/Navigator/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Navigator: true,
  NavigatorBack: true,
  NavigatorButton: true,
  NavigatorLink: true,
  NavigatorRouter: true,
  NavigatorScreen: true,
  NavigatorScreens: true,
  NavigatorSwitch: true,
  withNavigator: true
};
Object.defineProperty(exports, "Navigator", {
  enumerable: true,
  get: function get() {
    return _Navigator.default;
  }
});
Object.defineProperty(exports, "NavigatorBack", {
  enumerable: true,
  get: function get() {
    return _NavigatorBack.default;
  }
});
Object.defineProperty(exports, "NavigatorButton", {
  enumerable: true,
  get: function get() {
    return _NavigatorButton.default;
  }
});
Object.defineProperty(exports, "NavigatorLink", {
  enumerable: true,
  get: function get() {
    return _NavigatorLink.default;
  }
});
Object.defineProperty(exports, "NavigatorRouter", {
  enumerable: true,
  get: function get() {
    return _NavigatorRouter.default;
  }
});
Object.defineProperty(exports, "NavigatorScreen", {
  enumerable: true,
  get: function get() {
    return _NavigatorScreen.default;
  }
});
Object.defineProperty(exports, "NavigatorScreens", {
  enumerable: true,
  get: function get() {
    return _NavigatorScreens.default;
  }
});
Object.defineProperty(exports, "NavigatorSwitch", {
  enumerable: true,
  get: function get() {
    return _NavigatorSwitch.default;
  }
});
Object.defineProperty(exports, "withNavigator", {
  enumerable: true,
  get: function get() {
    return _withNavigator.default;
  }
});

var _Navigator = _interopRequireDefault(require("./Navigator"));

var _NavigatorBack = _interopRequireDefault(require("./NavigatorBack"));

var _NavigatorButton = _interopRequireDefault(require("./NavigatorButton"));

var _NavigatorLink = _interopRequireDefault(require("./NavigatorLink"));

var _NavigatorRouter = _interopRequireDefault(require("./NavigatorRouter"));

var _NavigatorScreen = _interopRequireDefault(require("./NavigatorScreen"));

var _NavigatorScreens = _interopRequireDefault(require("./NavigatorScreens"));

var _NavigatorSwitch = _interopRequireDefault(require("./NavigatorSwitch"));

var _withNavigator = _interopRequireDefault(require("./withNavigator"));

var _Navigator2 = require("./Navigator.Context");

Object.keys(_Navigator2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _Navigator2[key];
    }
  });
});

var _Navigator3 = require("./Navigator.hooks");

Object.keys(_Navigator3).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _Navigator3[key];
    }
  });
});
