// @wp-g2/components/dist/cjs/Navigator/NavigatorButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _react = _interopRequireDefault(require("react"));

var _Button = require("../Button");

var _Router = require("./Router");

function NavigatorButton(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'NavigatorButton'),
      as = _useContextSystem.as,
      children = _useContextSystem.children,
      exact = _useContextSystem.exact,
      href = _useContextSystem.href,
      isBack = _useContextSystem.isBack,
      isPlain = _useContextSystem.isPlain,
      params = _useContextSystem.params,
      showArrow = _useContextSystem.showArrow,
      to = _useContextSystem.to,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["as", "children", "exact", "href", "isBack", "isPlain", "params", "showArrow", "to"]);

  var history = (0, _Router.useHistory)();

  var handleOnClick = function (event) {
    event.preventDefault();

    if (isBack) {
      if (to) {
        history.push(to, {
          isBack: true
        });
      } else {
        history.goBack();
      }
    } else {
      history.push(to);
    }
  };

  if (!to) {
    return /*#__PURE__*/_react.default.createElement(_Button.Button, Object.assign({
      href: href || '#',
      ref: forwardedRef
    }, otherProps, {
      onClick: handleOnClick
    }), children);
  }

  return /*#__PURE__*/_react.default.createElement(_Button.Button, Object.assign({}, otherProps, {
    exact: exact,
    onClick: handleOnClick,
    ref: forwardedRef
  }), children);
}

var _default = (0, _context.contextConnect)(NavigatorButton, 'NavigatorButton');

exports.default = _default;
