// @wordpress/block-editor/build-module/components/block-mobile-toolbar/index.native.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { View } from 'react-native';
/**
 * WordPress dependencies
 */

import { withDispatch, withSelect } from '@wordpress/data';
import { compose } from '@wordpress/compose';
import { useState } from '@wordpress/element';
/**
 * Internal dependencies
 */

import styles from './style.scss';
import BlockMover from '../block-mover';
import BlockActionsMenu from './block-actions-menu';
import { BlockSettingsButton } from '../block-settings';
import { store as blockEditorStore } from '../../store'; // Defined breakpoints are used to get a point when
// `settings` and `mover` controls should be wrapped into `BlockActionsMenu`
// and accessed through `BottomSheet`(Android)/`ActionSheet`(iOS).

var BREAKPOINTS = {
  wrapSettings: 65,
  wrapMover: 150
};

var BlockMobileToolbar = function BlockMobileToolbar(_ref) {
  var clientId = _ref.clientId,
      onDelete = _ref.onDelete,
      isStackedHorizontally = _ref.isStackedHorizontally,
      blockWidth = _ref.blockWidth,
      anchorNodeRef = _ref.anchorNodeRef,
      isFullWidth = _ref.isFullWidth;

  var _useState = useState(null),
      _useState2 = _slicedToArray(_useState, 2),
      fillsLength = _useState2[0],
      setFillsLength = _useState2[1];

  var _useState3 = useState(0),
      _useState4 = _slicedToArray(_useState3, 2),
      appenderWidth = _useState4[0],
      setAppenderWidth = _useState4[1];

  var spacingValue = styles.toolbar.marginLeft * 2;

  function onLayout(_ref2) {
    var nativeEvent = _ref2.nativeEvent;
    var layout = nativeEvent.layout;
    var layoutWidth = Math.floor(layout.width);

    if (layoutWidth !== appenderWidth) {
      setAppenderWidth(nativeEvent.layout.width);
    }
  }

  var wrapBlockSettings = blockWidth < BREAKPOINTS.wrapSettings || appenderWidth - spacingValue < BREAKPOINTS.wrapSettings;
  var wrapBlockMover = blockWidth <= BREAKPOINTS.wrapMover || appenderWidth - spacingValue <= BREAKPOINTS.wrapMover;
  return createElement(View, {
    style: [styles.toolbar, isFullWidth && styles.toolbarFullWidth],
    onLayout: onLayout
  }, !wrapBlockMover && createElement(BlockMover, {
    clientIds: [clientId],
    isStackedHorizontally: isStackedHorizontally
  }), createElement(View, {
    style: styles.spacer
  }), createElement(BlockSettingsButton.Slot, null, function () {
    var fills = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [null];
    setFillsLength(fills.length);
    return wrapBlockSettings ? null : fills[0];
  }), createElement(BlockActionsMenu, {
    clientIds: [clientId],
    wrapBlockMover: wrapBlockMover,
    wrapBlockSettings: wrapBlockSettings && fillsLength,
    isStackedHorizontally: isStackedHorizontally,
    onDelete: onDelete,
    anchorNodeRef: anchorNodeRef
  }));
};

export default compose(withSelect(function (select, _ref3) {
  var clientId = _ref3.clientId;

  var _select = select(blockEditorStore),
      getBlockIndex = _select.getBlockIndex;

  return {
    order: getBlockIndex(clientId)
  };
}), withDispatch(function (dispatch, _ref4) {
  var clientId = _ref4.clientId,
      rootClientId = _ref4.rootClientId,
      onDelete = _ref4.onDelete;

  var _dispatch = dispatch(blockEditorStore),
      removeBlock = _dispatch.removeBlock;

  return {
    onDelete: onDelete || function () {
      return removeBlock(clientId, rootClientId);
    }
  };
}))(BlockMobileToolbar);
