// @wordpress/components/build/custom-gradient-picker/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CustomGradientPicker;

var _element = require("@wordpress/element");

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _lodash = require("lodash");

var _i18n = require("@wordpress/i18n");

var _anglePickerControl = _interopRequireDefault(require("../angle-picker-control"));

var _customGradientBar = _interopRequireDefault(require("../custom-gradient-bar"));

var _flex = require("../flex");

var _selectControl = _interopRequireDefault(require("../select-control"));

var _utils = require("./utils");

var _serializer = require("./serializer");

var _constants = require("./constants");

var _customGradientPickerStyles = require("./styles/custom-gradient-picker-styles");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var GradientAnglePicker = function GradientAnglePicker(_ref) {
  var gradientAST = _ref.gradientAST,
      hasGradient = _ref.hasGradient,
      onChange = _ref.onChange;
  var angle = (0, _lodash.get)(gradientAST, ['orientation', 'value'], _constants.DEFAULT_LINEAR_GRADIENT_ANGLE);

  var onAngleChange = function onAngleChange(newAngle) {
    onChange((0, _serializer.serializeGradient)(_objectSpread(_objectSpread({}, gradientAST), {}, {
      orientation: {
        type: 'angular',
        value: newAngle
      }
    })));
  };

  return (0, _element.createElement)(_anglePickerControl.default, {
    hideLabelFromVision: true,
    onChange: onAngleChange,
    value: hasGradient ? angle : ''
  });
};

var GradientTypePicker = function GradientTypePicker(_ref2) {
  var gradientAST = _ref2.gradientAST,
      hasGradient = _ref2.hasGradient,
      onChange = _ref2.onChange;
  var type = gradientAST.type;

  var onSetLinearGradient = function onSetLinearGradient() {
    onChange((0, _serializer.serializeGradient)(_objectSpread(_objectSpread(_objectSpread({}, gradientAST), gradientAST.orientation ? {} : {
      orientation: _constants.HORIZONTAL_GRADIENT_ORIENTATION
    }), {}, {
      type: 'linear-gradient'
    })));
  };

  var onSetRadialGradient = function onSetRadialGradient() {
    onChange((0, _serializer.serializeGradient)(_objectSpread(_objectSpread({}, (0, _lodash.omit)(gradientAST, ['orientation'])), {}, {
      type: 'radial-gradient'
    })));
  };

  var handleOnChange = function handleOnChange(next) {
    if (next === 'linear-gradient') {
      onSetLinearGradient();
    }

    if (next === 'radial-gradient') {
      onSetRadialGradient();
    }
  };

  return (0, _element.createElement)(_selectControl.default, {
    className: "components-custom-gradient-picker__type-picker",
    label: (0, _i18n.__)('Type'),
    labelPosition: 'side',
    onChange: handleOnChange,
    options: _constants.GRADIENT_OPTIONS,
    value: hasGradient && type
  });
};

function CustomGradientPicker(_ref3) {
  var value = _ref3.value,
      _onChange = _ref3.onChange;
  var gradientAST = (0, _utils.getGradientAstWithDefault)(value); // On radial gradients the bar should display a linear gradient.
  // On radial gradients the bar represents a slice of the gradient from the center until the outside.

  var background = gradientAST.type === 'radial-gradient' ? (0, _utils.getLinearGradientRepresentationOfARadial)(gradientAST) : gradientAST.value;
  var hasGradient = gradientAST.value !== _constants.DEFAULT_GRADIENT; // Control points color option may be hex from presets, custom colors will be rgb.
  // The position should always be a percentage.

  var controlPoints = gradientAST.colorStops.map(function (colorStop) {
    return {
      color: (0, _utils.getStopCssColor)(colorStop),
      position: parseInt(colorStop.length.value)
    };
  });
  return (0, _element.createElement)("div", {
    className: "components-custom-gradient-picker"
  }, (0, _element.createElement)(_customGradientBar.default, {
    background: background,
    hasGradient: hasGradient,
    value: controlPoints,
    onChange: function onChange(newControlPoints) {
      _onChange((0, _serializer.serializeGradient)((0, _utils.getGradientAstWithControlPoints)(gradientAST, newControlPoints)));
    }
  }), (0, _element.createElement)(_flex.Flex, {
    gap: 3,
    className: "components-custom-gradient-picker__ui-line"
  }, (0, _element.createElement)(_customGradientPickerStyles.SelectWrapper, null, (0, _element.createElement)(GradientTypePicker, {
    gradientAST: gradientAST,
    hasGradient: hasGradient,
    onChange: _onChange
  })), (0, _element.createElement)(_customGradientPickerStyles.AccessoryWrapper, null, gradientAST.type === 'linear-gradient' && (0, _element.createElement)(GradientAnglePicker, {
    gradientAST: gradientAST,
    hasGradient: hasGradient,
    onChange: _onChange
  }))));
}
