// @wp-g2/components/dist/es/Dropdown/__stories__/Dropdown.stories.js
import React from 'react';
import { useDialogState } from 'reakit';
import { Button, Divider, Modal, ModalBody, ModalFooter, ModalHeader } from '../../index';
import { Dropdown, DropdownMenu, DropdownMenuItem, DropdownTrigger } from '../index';
export default {
  component: Dropdown,
  title: 'Components/Dropdown'
};
export var WithModal = function () {
  var dialog = useDialogState({
    animated: true
  });
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Dropdown, null, /*#__PURE__*/React.createElement(DropdownTrigger, null, "Dropdown"), /*#__PURE__*/React.createElement(DropdownMenu, null, /*#__PURE__*/React.createElement(DropdownMenuItem, {
    onClick: dialog.show
  }, "One"), /*#__PURE__*/React.createElement(DropdownMenuItem, {
    onClick: dialog.show
  }, "Two"))), /*#__PURE__*/React.createElement(Modal, {
    dialog: dialog
  }, /*#__PURE__*/React.createElement(ModalHeader, null), /*#__PURE__*/React.createElement(ModalBody, null, "Hello"), /*#__PURE__*/React.createElement(ModalFooter, null, /*#__PURE__*/React.createElement(Button, {
    onClick: dialog.hide,
    variant: "primary"
  }, "Save"), /*#__PURE__*/React.createElement(Button, {
    onClick: dialog.hide
  }, "Cancel"))));
};
export var _default = function () {
  return /*#__PURE__*/React.createElement(Dropdown, {
    visible: true
  }, /*#__PURE__*/React.createElement(DropdownTrigger, null, "Dropdown"), /*#__PURE__*/React.createElement(DropdownMenu, null, /*#__PURE__*/React.createElement(DropdownMenuItem, null, "One"), /*#__PURE__*/React.createElement(DropdownMenuItem, null, "Two"), /*#__PURE__*/React.createElement(DropdownMenuItem, null, "Three"), /*#__PURE__*/React.createElement(Divider, null), /*#__PURE__*/React.createElement(DropdownMenuItem, null, "Three"), /*#__PURE__*/React.createElement(DropdownMenuItem, null, "Three"), /*#__PURE__*/React.createElement(DropdownMenuItem, null, "Three"), /*#__PURE__*/React.createElement(DropdownMenuItem, null, "Three"), /*#__PURE__*/React.createElement(DropdownMenuItem, null, "Three"), /*#__PURE__*/React.createElement(DropdownMenuItem, null, "Three"), /*#__PURE__*/React.createElement(DropdownMenuItem, null, "Three")));
};
export var _renderFunction = function () {
  return /*#__PURE__*/React.createElement(Dropdown, {
    visible: true
  }, function (_ref) {
    var onClose = _ref.onClose;
    return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(DropdownTrigger, null, "Dropdown"), /*#__PURE__*/React.createElement(DropdownMenu, null, /*#__PURE__*/React.createElement(DropdownMenuItem, {
      onClick: onClose
    }, "One"), /*#__PURE__*/React.createElement(DropdownMenuItem, null, "Two"), /*#__PURE__*/React.createElement(DropdownMenuItem, null, "Three")));
  });
};
