// @wp-g2/components/dist/cjs/font-size-control/font-size-control-select.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _i18n = require("@wordpress/i18n");

var _context = require("@wp-g2/context");

var _utils = require("@wp-g2/utils");

var _react = _interopRequireDefault(require("react"));

var _Button = require("../Button");

var _FormGroup = require("../FormGroup");

var _HStack = require("../HStack");

var _selectDropdown = require("../select-dropdown");

var _TextInput = require("../TextInput");

var _View = require("../View");

var styles = _interopRequireWildcard(require("./font-size-control-styles"));

function FontSizeControlSelect(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'FontSizeControlSelect'),
      customLabel = _useContextSystem.customLabel,
      disabled = _useContextSystem.disabled,
      inputValue = _useContextSystem.inputValue,
      isDefaultValue = _useContextSystem.isDefaultValue,
      label = _useContextSystem.label,
      max = _useContextSystem.max,
      min = _useContextSystem.min,
      _useContextSystem$onC = _useContextSystem.onChange,
      onChange = _useContextSystem$onC === void 0 ? _utils.noop : _useContextSystem$onC,
      _useContextSystem$onI = _useContextSystem.onInputChange,
      onInputChange = _useContextSystem$onI === void 0 ? _utils.noop : _useContextSystem$onI,
      _useContextSystem$onR = _useContextSystem.onReset,
      onReset = _useContextSystem$onR === void 0 ? _utils.noop : _useContextSystem$onR,
      _useContextSystem$opt = _useContextSystem.options,
      options = _useContextSystem$opt === void 0 ? [] : _useContextSystem$opt,
      size = _useContextSystem.size,
      value = _useContextSystem.value,
      withNumberInput = _useContextSystem.withNumberInput,
      withSelect = _useContextSystem.withSelect,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["customLabel", "disabled", "inputValue", "isDefaultValue", "label", "max", "min", "onChange", "onInputChange", "onReset", "options", "size", "value", "withNumberInput", "withSelect"]);

  var renderItem = _react.default.useCallback(function (_ref) {
    var name = _ref.name,
        style = _ref.style;
    return /*#__PURE__*/_react.default.createElement("span", {
      style: style
    }, name);
  }, []);

  return /*#__PURE__*/_react.default.createElement(_HStack.HStack, {
    alignment: "left",
    wrap: true
  }, withSelect && /*#__PURE__*/_react.default.createElement(_View.View, {
    className: styles.SelectDropdownWrapper
  }, /*#__PURE__*/_react.default.createElement(_FormGroup.FormGroup, {
    label: label
  }, /*#__PURE__*/_react.default.createElement(_selectDropdown.SelectDropdown, Object.assign({
    disabled: disabled,
    max: 260,
    onChange: onChange,
    options: options,
    ref: forwardedRef,
    renderItem: renderItem,
    size: size,
    value: value
  }, otherProps)))), withNumberInput && /*#__PURE__*/_react.default.createElement(_View.View, {
    className: styles.NumberInputWrapper
  }, /*#__PURE__*/_react.default.createElement(_FormGroup.FormGroup, {
    label: customLabel
  }, /*#__PURE__*/_react.default.createElement(_TextInput.TextInput, {
    disabled: disabled,
    max: max,
    min: min,
    onChange: onInputChange,
    size: size,
    type: "number",
    value: inputValue
  }))), /*#__PURE__*/_react.default.createElement(_View.View, {
    className: styles.ResetButtonWrapper
  }, /*#__PURE__*/_react.default.createElement(_Button.Button, {
    disabled: isDefaultValue,
    isBlock: true,
    onClick: onReset
  }, (0, _i18n.__)('Reset'))));
}

var _default = (0, _context.contextConnect)(FontSizeControlSelect, 'FontSizeControlSelect');

exports.default = _default;
