// @wp-g2/components/dist/cjs/font-size-control/font-size-control-slider.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _i18n = require("@wordpress/i18n");

var _utils = require("@wp-g2/utils");

var _react = _interopRequireDefault(require("react"));

var _ControlLabel = require("../ControlLabel");

var _HStack = require("../HStack");

var _Slider = require("../Slider");

var _TextInput = require("../TextInput");

var _View = require("../View");

var _VStack = require("../VStack");

var styles = _interopRequireWildcard(require("./font-size-control-styles"));

function FontSizeControlSlider(props) {
  var _props$label = props.label,
      label = _props$label === void 0 ? (0, _i18n.__)('Custom size') : _props$label,
      disabled = props.disabled,
      min = props.min,
      max = props.max,
      _props$onChange = props.onChange,
      onChange = _props$onChange === void 0 ? _utils.noop : _props$onChange,
      size = props.size,
      value = props.value,
      withSlider = props.withSlider;
  if (!withSlider) return null;
  var controlProps = {
    disabled: disabled,
    min: min,
    max: max,
    onChange: onChange,
    size: size,
    value: value
  };
  return /*#__PURE__*/_react.default.createElement(_VStack.VStack, {
    spacing: 0
  }, /*#__PURE__*/_react.default.createElement(_ControlLabel.ControlLabel, null, label), /*#__PURE__*/_react.default.createElement(_HStack.HStack, {
    wrap: true
  }, /*#__PURE__*/_react.default.createElement(_View.View, {
    className: styles.SliderWrapper
  }, /*#__PURE__*/_react.default.createElement(_Slider.Slider, controlProps)), /*#__PURE__*/_react.default.createElement(_View.View, null, /*#__PURE__*/_react.default.createElement(_TextInput.TextInput, Object.assign({}, controlProps, {
    type: "number"
  })))));
}

var _default = FontSizeControlSlider;
exports.default = _default;
