// @wp-g2/components/dist/es/font-size-control/font-size-control-slider.js
import { __ } from '@wordpress/i18n';
import { noop } from '@wp-g2/utils';
import React from 'react';
import { ControlLabel } from '../ControlLabel';
import { HStack } from '../HStack';
import { Slider } from '../Slider';
import { TextInput } from '../TextInput';
import { View } from '../View';
import { VStack } from '../VStack';
import * as styles from './font-size-control-styles';

function FontSizeControlSlider(props) {
  var _props$label = props.label,
      label = _props$label === void 0 ? __('Custom size') : _props$label,
      disabled = props.disabled,
      min = props.min,
      max = props.max,
      _props$onChange = props.onChange,
      onChange = _props$onChange === void 0 ? noop : _props$onChange,
      size = props.size,
      value = props.value,
      withSlider = props.withSlider;
  if (!withSlider) return null;
  var controlProps = {
    disabled: disabled,
    min: min,
    max: max,
    onChange: onChange,
    size: size,
    value: value
  };
  return /*#__PURE__*/React.createElement(VStack, {
    spacing: 0
  }, /*#__PURE__*/React.createElement(ControlLabel, null, label), /*#__PURE__*/React.createElement(HStack, {
    wrap: true
  }, /*#__PURE__*/React.createElement(View, {
    className: styles.SliderWrapper
  }, /*#__PURE__*/React.createElement(Slider, controlProps)), /*#__PURE__*/React.createElement(View, null, /*#__PURE__*/React.createElement(TextInput, Object.assign({}, controlProps, {
    type: "number"
  })))));
}

export default FontSizeControlSlider;
