// @wordpress/block-editor/build/components/inserter/hooks/use-patterns-state.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _lodash = require("lodash");

var _element = require("@wordpress/element");

var _blocks = require("@wordpress/blocks");

var _data = require("@wordpress/data");

var _i18n = require("@wordpress/i18n");

var _notices = require("@wordpress/notices");

var _store = require("../../../store");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

/**
 * Retrieves the block patterns inserter state.
 *
 * @param {Function} onInsert     function called when inserter a list of blocks.
 * @param {string=}  rootClientId Insertion's root client ID.
 *
 * @return {Array} Returns the patterns state. (patterns, categories, onSelect handler)
 */
var usePatternsState = function usePatternsState(onInsert, rootClientId) {
  var _useSelect = (0, _data.useSelect)(function (select) {
    var _select = select(_store.store),
        __experimentalGetAllowedPatterns = _select.__experimentalGetAllowedPatterns,
        getSettings = _select.getSettings;

    var inserterPatterns = __experimentalGetAllowedPatterns(rootClientId).filter(function (pattern) {
      return !pattern.scope || pattern.scope.inserter;
    });

    return {
      patterns: inserterPatterns,
      patternCategories: getSettings().__experimentalBlockPatternCategories
    };
  }, [rootClientId]),
      patternCategories = _useSelect.patternCategories,
      patterns = _useSelect.patterns;

  var _useDispatch = (0, _data.useDispatch)(_notices.store),
      createSuccessNotice = _useDispatch.createSuccessNotice;

  var onClickPattern = (0, _element.useCallback)(function (pattern, blocks) {
    onInsert((0, _lodash.map)(blocks, function (block) {
      return (0, _blocks.cloneBlock)(block);
    }), pattern.name);
    createSuccessNotice((0, _i18n.sprintf)(
    /* translators: %s: block pattern title. */
    (0, _i18n.__)('Block pattern "%s" inserted.'), pattern.title), {
      type: 'snackbar'
    });
  }, []);
  return [patterns, patternCategories, onClickPattern];
};

var _default = usePatternsState;
exports.default = _default;
