// @wp-g2/components/dist/es/Alert/__stories__/Alert.stories.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import React, { useState } from 'react';
import { Button, Spacer, Text, View } from '../../index';
import { Alert, Alerts } from '../index';
export default {
  component: Alert,
  title: 'Components/Alert'
};
var statuses = ['default', 'success', 'warning', 'info', 'critical'];

var Example = function () {
  var _useState = useState(statuses),
      _useState2 = _slicedToArray(_useState, 2),
      alerts = _useState2[0],
      setAlerts = _useState2[1];

  var remove = function (id) {
    return setAlerts(function (prev) {
      return prev.filter(function (i) {
        return i !== id;
      });
    });
  };

  return /*#__PURE__*/React.createElement(View, {
    css: "\n\t\t\t\tmax-width: 480px;\n\t\t\t\tmargin: auto;\n\t\t\t"
  }, /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(Button, {
    disabled: alerts.length === 5,
    onClick: function reset() {
      return setAlerts(statuses);
    }
  }, "Reset")), /*#__PURE__*/React.createElement(Alerts, null, alerts.map(function (status) {
    return /*#__PURE__*/React.createElement(Alert, {
      isDismissable: true,
      key: status,
      onDismiss: function onDismiss() {
        remove(status);
      },
      status: status,
      title: "Title"
    }, /*#__PURE__*/React.createElement(Text, null, "Alert Content"));
  })));
};

export var _default = function () {
  return /*#__PURE__*/React.createElement(Example, null);
};
