// @wordpress/components/build/navigation/menu/menu-title.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = NavigationMenuTitle;

var _element = require("@wordpress/element");

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

var _animate = require("../../animate");

var _button = _interopRequireDefault(require("../../button"));

var _menuTitleSearch = _interopRequireDefault(require("./menu-title-search"));

var _navigationStyles = require("../styles/navigation-styles");

var _context = require("./context");

var _constants = require("../constants");

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function NavigationMenuTitle(_ref) {
  var hasSearch = _ref.hasSearch,
      onSearch = _ref.onSearch,
      search = _ref.search,
      title = _ref.title,
      titleAction = _ref.titleAction;

  var _useState = (0, _element.useState)(false),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      isSearching = _useState2[0],
      setIsSearching = _useState2[1];

  var _useNavigationMenuCon = (0, _context.useNavigationMenuContext)(),
      menu = _useNavigationMenuCon.menu;

  var searchButtonRef = (0, _element.useRef)();

  if (!title) {
    return null;
  }

  var onCloseSearch = function onCloseSearch() {
    setIsSearching(false); // Wait for the slide-in animation to complete before focusing the search button.
    // eslint-disable-next-line @wordpress/react-no-unsafe-timeout

    setTimeout(function () {
      searchButtonRef.current.focus();
    }, _constants.SEARCH_FOCUS_DELAY);
  };

  var menuTitleId = "components-navigation__menu-title-".concat(menu);
  /* translators: search button label for menu search box. %s: menu title */

  var searchButtonLabel = (0, _i18n.sprintf)((0, _i18n.__)('Search in %s'), title);
  return (0, _element.createElement)(_navigationStyles.MenuTitleUI, {
    className: "components-navigation__menu-title"
  }, !isSearching && (0, _element.createElement)(_navigationStyles.MenuTitleHeadingUI, {
    as: "h2",
    className: "components-navigation__menu-title-heading",
    variant: "title.small"
  }, (0, _element.createElement)("span", {
    id: menuTitleId
  }, title), (hasSearch || titleAction) && (0, _element.createElement)(_navigationStyles.MenuTitleActionsUI, null, titleAction, hasSearch && (0, _element.createElement)(_button.default, {
    isSmall: true,
    isTertiary: true,
    label: searchButtonLabel,
    onClick: function onClick() {
      return setIsSearching(true);
    },
    ref: searchButtonRef
  }, (0, _element.createElement)(_icons.Icon, {
    icon: _icons.search
  })))), isSearching && (0, _element.createElement)("div", {
    className: (0, _animate.getAnimateClassName)({
      type: 'slide-in',
      origin: 'left'
    })
  }, (0, _element.createElement)(_menuTitleSearch.default, {
    onCloseSearch: onCloseSearch,
    onSearch: onSearch,
    search: search,
    title: title
  })));
}
