// @wp-g2/components/dist/es/Draggable/__stories__/BeautifulDnd.stories.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import { useListState } from '@wp-g2/protokit';
import { ui } from '@wp-g2/styles';
import faker from 'faker';
import { Schema } from 'faker-schema';
import React from 'react';
import { DragDropContext, Draggable, Droppable } from 'react-beautiful-dnd';
import { Card, CardBody, Heading, HStack, Image, Spacer, Text, View, VStack } from '../../index';
export default {
  title: 'Components/Draggable'
};
var userSchema = new Schema(function () {
  return {
    avatar: faker.image.avatar(),
    description: faker.lorem.sentence(),
    id: faker.random.uuid(),
    title: faker.name.firstName()
  };
});

function Avatar(_ref) {
  var src = _ref.src;
  return /*#__PURE__*/React.createElement(View, {
    css: {
      height: 32,
      width: 32
    }
  }, /*#__PURE__*/React.createElement(Image, {
    css: [ui.borderRadius.circle, {
      display: 'block'
    }],
    src: src
  }));
}

var UserCard = function (_ref2) {
  var avatar = _ref2.avatar,
      description = _ref2.description,
      title = _ref2.title;
  return /*#__PURE__*/React.createElement(Card, null, /*#__PURE__*/React.createElement(CardBody, null, /*#__PURE__*/React.createElement(HStack, {
    alignment: "topLeft",
    spacing: 4
  }, /*#__PURE__*/React.createElement(Avatar, {
    src: avatar
  }), /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(VStack, {
    spacing: 1
  }, /*#__PURE__*/React.createElement(Text, {
    weight: "bold"
  }, title), /*#__PURE__*/React.createElement(Text, {
    variant: "muted"
  }, description))))));
};

var Example = function () {
  var _useListState = useListState(userSchema.make(10)),
      _useListState2 = _slicedToArray(_useListState, 2),
      users = _useListState2[0],
      usersData = _useListState2[1];

  var _useListState3 = useListState(userSchema.make(10)),
      _useListState4 = _slicedToArray(_useListState3, 2),
      editors = _useListState4[0],
      editorsData = _useListState4[1];

  var states = {
    editors: editors,
    users: users
  };
  var sources = {
    editors: editorsData,
    users: usersData
  };
  return /*#__PURE__*/React.createElement(DragDropContext, {
    onBeforeCapture: function onBeforeCapture(_ref4) {
      var draggableId = _ref4.draggableId;
      var el = document.querySelector("[data-rbd-drag-handle-draggable-id=\"" + draggableId + "\"]");

      if (el) {
        el.style.height = '12px';
      }
    },
    onDragEnd: function onMove(_ref3) {
      var destination = _ref3.destination,
          source = _ref3.source,
          rest = _objectWithoutProperties(_ref3, ["destination", "source"]);

      // Same list
      if (destination.droppableId === source.droppableId) {
        sources[destination.droppableId].move(source.index, destination.index);
        return;
      }

      var item = states[source.droppableId][source.index];
      sources[source.droppableId].remove({
        at: source.index
      });
      sources[destination.droppableId].insert({
        at: destination.index,
        item: item
      });
    }
  }, /*#__PURE__*/React.createElement(HStack, {
    css: [{
      maxWidth: '100%',
      width: 800
    }, ui.alignment.center],
    spacing: 0
  }, /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(Droppable, {
    droppableId: "users"
  }, function (provided, snapshot) {
    return /*#__PURE__*/React.createElement(View, Object.assign({}, provided.droppableProps, {
      css: [ui.padding(5), snapshot.isDraggingOver && ui.background.blue],
      ref: provided.innerRef
    }), /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(Heading, null, "Users")), users.map(function (user, index) {
      return /*#__PURE__*/React.createElement(Draggable, {
        draggableId: user.id,
        index: index,
        key: user.id
      }, function (provided) {
        return /*#__PURE__*/React.createElement("div", Object.assign({
          ref: provided.innerRef
        }, provided.draggableProps, provided.dragHandleProps, {
          style: _objectSpread(_objectSpread({}, provided.draggableProps.style), {}, {
            marginBottom: 8
          })
        }), /*#__PURE__*/React.createElement(UserCard, user));
      });
    }), provided.placeholder);
  })), /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(Droppable, {
    droppableId: "editors"
  }, function (provided, snapshot) {
    return /*#__PURE__*/React.createElement(View, Object.assign({}, provided.droppableProps, {
      css: [ui.padding(5), snapshot.isDraggingOver && ui.background.blue],
      ref: provided.innerRef
    }), /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(Heading, null, "Editors")), editors.map(function (user, index) {
      return /*#__PURE__*/React.createElement(Draggable, {
        draggableId: user.id,
        index: index,
        key: user.id
      }, function (provided) {
        return /*#__PURE__*/React.createElement("div", Object.assign({
          ref: provided.innerRef
        }, provided.draggableProps, provided.dragHandleProps, {
          style: _objectSpread(_objectSpread({}, provided.draggableProps.style), {}, {
            marginBottom: 8
          })
        }), /*#__PURE__*/React.createElement(UserCard, user));
      });
    }), provided.placeholder);
  }))));
};

export var ReactBeautifulDnd = function () {
  return /*#__PURE__*/React.createElement(Example, null);
};
