// @wp-g2/components/dist/es/Draggable/__stories__/DragListContext.stories.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import { useListState } from '@wp-g2/protokit';
import { ui } from '@wp-g2/styles';
import faker from 'faker';
import { Schema } from 'faker-schema';
import React, { useState } from 'react';
import { DragDropContext, Draggable, Droppable } from 'react-beautiful-dnd';
import { Animated, Card, CardBody, Heading, HStack, Image, Scrollable, Spacer, Text, View, VStack } from '../../index';
export default {
  title: 'Components/Draggable'
};
var blockSchema = new Schema(function () {
  return {
    content: faker.lorem.sentences(),
    id: faker.random.uuid(),
    title: 'Paragraph'
  };
});

var BlockDragIndexLine = function () {
  return /*#__PURE__*/React.createElement(View, {
    css: [ui.position.relative]
  }, /*#__PURE__*/React.createElement(View, {
    css: [ui.position.top, ui.borderRadius.round, {
      height: 3,
      width: '100%'
    }, ui.background.admin]
  }));
};

var Example = function () {
  var _useListState = useListState(blockSchema.make(10)),
      _useListState2 = _slicedToArray(_useListState, 2),
      blocks = _useListState2[0],
      setBlocks = _useListState2[1];

  var _useState = useState(),
      _useState2 = _slicedToArray(_useState, 2),
      dragIndex = _useState2[0],
      setDragIndex = _useState2[1];

  var _useState3 = useState(),
      _useState4 = _slicedToArray(_useState3, 2),
      targetIndex = _useState4[0],
      setTargetIndex = _useState4[1];

  var _useState5 = useState(false),
      _useState6 = _slicedToArray(_useState5, 2),
      isDragging = _useState6[0],
      setIsDragging = _useState6[1];

  var setHighlightOn = function (index) {
    setTargetIndex(index);
    setIsDragging(true);
  };

  var setHighlightOff = function () {
    setTargetIndex(undefined);
    setIsDragging(false);
  };

  return /*#__PURE__*/React.createElement(DragDropContext, {
    onDragEnd: function onDragEnd(_ref) {
      var destination = _ref.destination,
          source = _ref.source;
      setBlocks.move(source.index, destination.index);
      setDragIndex(undefined);
      setHighlightOff();
    },
    onDragStart: function onDragStart(_ref2) {
      var source = _ref2.source;
      setDragIndex(source == null ? void 0 : source.index);
      setTargetIndex(source == null ? void 0 : source.index);
      setHighlightOn(source == null ? void 0 : source.index);
    },
    onDragUpdate: function onDragUpdate(_ref3) {
      var destination = _ref3.destination;
      setDragIndex(destination == null ? void 0 : destination.index);
    }
  }, /*#__PURE__*/React.createElement(HStack, {
    css: [ui.frame.width(800), ui.alignment.center],
    spacing: 8
  }, /*#__PURE__*/React.createElement(View, {
    css: ui.frame.width(200)
  }, /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(Heading, {
    size: 4
  }, "Block List")), /*#__PURE__*/React.createElement(Scrollable, {
    css: [ui.frame.height(300), ui.padding(2)]
  }, /*#__PURE__*/React.createElement(Droppable, {
    droppableId: "blockList"
  }, function (provided) {
    return /*#__PURE__*/React.createElement(View, Object.assign({}, provided.droppableProps, {
      ref: provided.innerRef
    }), blocks.map(function (block, index) {
      return /*#__PURE__*/React.createElement(Draggable, {
        draggableId: block.id,
        index: index,
        key: block.id
      }, function (provided) {
        return /*#__PURE__*/React.createElement("div", Object.assign({
          ref: provided.innerRef
        }, provided.draggableProps, provided.dragHandleProps, {
          style: _objectSpread(_objectSpread({}, provided.draggableProps.style), {}, {
            marginBottom: 8
          })
        }), /*#__PURE__*/React.createElement(Card, null, /*#__PURE__*/React.createElement(CardBody, null, /*#__PURE__*/React.createElement(Text, {
          weight: "bold"
        }, block.title))));
      });
    }), provided.placeholder);
  }))), /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(VStack, null, /*#__PURE__*/React.createElement(Heading, {
    size: 2
  }, "Blog Post"), blocks.map(function (block, index) {
    return /*#__PURE__*/React.createElement(React.Fragment, {
      key: block.id
    }, index === dragIndex && /*#__PURE__*/React.createElement(Animated, {
      key: "BlockDragIndexLine",
      layout: true
    }, /*#__PURE__*/React.createElement(BlockDragIndexLine, null)), /*#__PURE__*/React.createElement(Animated, {
      key: block.id,
      layout: true
    }, /*#__PURE__*/React.createElement(View, {
      css: [ui.animation.ease, isDragging && ui.opacity(0.3), targetIndex === index && ui.opacity(1)]
    }, /*#__PURE__*/React.createElement(Text, {
      css: {
        marginBottom: '1em',
        marginTop: '1em'
      },
      isBlock: true,
      size: 16
    }, block.content, ' '))));
  })))));
};

export var DragListContext = function () {
  return /*#__PURE__*/React.createElement(Example, null);
};
