// @wp-g2/components/dist/es/Draggable/__stories__/Sortable.stories.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _inheritsLoose from "@babel/runtime/helpers/esm/inheritsLoose";
import { useListState } from '@wp-g2/protokit';
import faker from 'faker';
import { Schema } from 'faker-schema';
import React, { createRef } from 'react';
import SortableCore from 'sortablejs';
import { Card, CardBody, Heading, Subheading, Text, View, VStack } from '../../index';
export default {
  title: 'Components/Draggable'
};

var Sortable = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(Sortable, _React$Component);

  function Sortable() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;
    _this.nodeRef = /*#__PURE__*/createRef();
    _this.sortable = null;
    return _this;
  }

  var _proto = Sortable.prototype;

  _proto.componentDidMount = function componentDidMount() {
    this.sortable = SortableCore.create(this.nodeRef.current, {
      animation: 150,
      forceFallback: true,
      ghostClass: 'ghost'
    });
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    this.sortable.destroy();
  };

  _proto.render = function render() {
    return /*#__PURE__*/React.createElement(View, {
      ref: this.nodeRef
    }, this.props.children);
  };

  return Sortable;
}(React.Component);

var Draggable = function (props) {
  return /*#__PURE__*/React.createElement(View, Object.assign({
    "data-sortable-item": "true"
  }, props, {
    css: [{
      marginBottom: 8
    }, "&.ghost {opacity: 0}"]
  }));
};

var itemSchema = new Schema(function () {
  return {
    id: faker.random.uuid(),
    name: faker.name.firstName()
  };
});

var Example = function () {
  var _useListState = useListState(itemSchema.make(10)),
      _useListState2 = _slicedToArray(_useListState, 1),
      items = _useListState2[0];

  return /*#__PURE__*/React.createElement(VStack, {
    spacing: 5
  }, /*#__PURE__*/React.createElement(VStack, {
    spacing: 0
  }, /*#__PURE__*/React.createElement(Heading, null, "SortableJS"), /*#__PURE__*/React.createElement(Subheading, null, "Probaby a no go.")), /*#__PURE__*/React.createElement(Sortable, null, items.map(function (item) {
    return /*#__PURE__*/React.createElement(Draggable, {
      key: item.id
    }, /*#__PURE__*/React.createElement(Card, null, /*#__PURE__*/React.createElement(CardBody, null, /*#__PURE__*/React.createElement(Text, null, item.name))));
  })));
};

export var SortableExample = function () {
  return /*#__PURE__*/React.createElement(Example, null);
};
