// @wp-g2/components/dist/es/VStack/__stories__/VStack.stories.js
import React from 'react';
import { HStack } from '../../HStack';
import { Badge, Button, Heading, Placeholder, Spacer, Text, View } from '../../index';
import { VStack } from '../index';
export default {
  component: VStack,
  title: 'Components/VStack'
};
export var _default = function () {
  return /*#__PURE__*/React.createElement(View, {
    css: {
      margin: 'auto',
      maxWidth: 960
    }
  }, /*#__PURE__*/React.createElement(VStack, null, /*#__PURE__*/React.createElement(HStack, {
    justify: "space-between"
  }, /*#__PURE__*/React.createElement(View, null, /*#__PURE__*/React.createElement(Heading, {
    size: 3
  }, "Components: HStack, VStack, ZStack", ' ', /*#__PURE__*/React.createElement(View, {
    as: "span",
    css: {
      fontSize: 'inherit',
      opacity: 0.5
    }
  }, "#22"))), /*#__PURE__*/React.createElement(View, null, /*#__PURE__*/React.createElement(HStack, null, /*#__PURE__*/React.createElement(Button, null, "Edit"), /*#__PURE__*/React.createElement(Button, {
    variant: "primary"
  }, "New Issue")))), /*#__PURE__*/React.createElement(HStack, {
    justify: "flex-start"
  }, /*#__PURE__*/React.createElement(Badge, {
    color: "green"
  }, "Open"), /*#__PURE__*/React.createElement(Text, {
    weight: "bold"
  }, "ItsJonQ"), /*#__PURE__*/React.createElement(Text, null, "opened this issue 2 days ago \xB7 1 comment"))));
};
