// @wp-g2/components/dist/cjs/ButtonGroup/__stories__/ButtonGroup.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._segmented = exports._expanded = exports._default = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _icons = require("@wordpress/icons");

var _react = _interopRequireDefault(require("react"));

var _index = require("../../index");

var _index2 = require("../index");

var _default2 = {
  component: _index2.ButtonGroup,
  title: 'Components/ButtonGroup'
};
exports.default = _default2;

var _default = function () {
  var _React$useState = _react.default.useState('bold'),
      _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
      value = _React$useState2[0],
      setValue = _React$useState2[1];

  return /*#__PURE__*/_react.default.createElement(_index.ListGroups, null, /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(_index.ListGroupHeader, null, "Default"), /*#__PURE__*/_react.default.createElement(_index.FormGroup, null, /*#__PURE__*/_react.default.createElement(_index2.ButtonGroup, {
    onChange: setValue,
    value: value
  }, /*#__PURE__*/_react.default.createElement(_index.Button, {
    icon: _icons.formatBold,
    value: "bold"
  }), /*#__PURE__*/_react.default.createElement(_index.Button, {
    icon: _icons.formatItalic,
    value: "italic"
  }), /*#__PURE__*/_react.default.createElement(_index.Button, {
    icon: _icons.formatUnderline,
    value: "underline"
  })))), /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(_index.ListGroupHeader, null, "Segmented"), /*#__PURE__*/_react.default.createElement(_index.FormGroup, null, /*#__PURE__*/_react.default.createElement(_index2.ButtonGroup, {
    onChange: setValue,
    segmented: true,
    value: value
  }, /*#__PURE__*/_react.default.createElement(_index.Button, {
    icon: _icons.formatBold,
    value: "bold"
  }), /*#__PURE__*/_react.default.createElement(_index.Button, {
    icon: _icons.formatItalic,
    value: "italic"
  }), /*#__PURE__*/_react.default.createElement(_index.Button, {
    icon: _icons.formatUnderline,
    value: "underline"
  })))), /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(_index.ListGroupHeader, null, "Expanded"), /*#__PURE__*/_react.default.createElement(_index.FormGroup, null, /*#__PURE__*/_react.default.createElement(_index2.ButtonGroup, {
    expanded: true,
    onChange: setValue,
    value: value
  }, /*#__PURE__*/_react.default.createElement(_index.Button, {
    icon: _icons.formatBold,
    value: "bold"
  }), /*#__PURE__*/_react.default.createElement(_index.Button, {
    icon: _icons.formatItalic,
    value: "italic"
  }), /*#__PURE__*/_react.default.createElement(_index.Button, {
    icon: _icons.formatUnderline,
    value: "underline"
  })))), /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(_index.ListGroupHeader, null, "Segmented + Expanded"), /*#__PURE__*/_react.default.createElement(_index.FormGroup, null, /*#__PURE__*/_react.default.createElement(_index2.ButtonGroup, {
    expanded: true,
    onChange: setValue,
    segmented: true,
    value: value
  }, /*#__PURE__*/_react.default.createElement(_index.Button, {
    icon: _icons.formatBold,
    value: "bold"
  }), /*#__PURE__*/_react.default.createElement(_index.Button, {
    icon: _icons.formatItalic,
    value: "italic"
  }), /*#__PURE__*/_react.default.createElement(_index.Button, {
    icon: _icons.formatUnderline,
    value: "underline"
  })))));
};

exports._default = _default;

var _expanded = function () {
  var _React$useState3 = _react.default.useState('bold'),
      _React$useState4 = (0, _slicedToArray2.default)(_React$useState3, 2),
      value = _React$useState4[0],
      setValue = _React$useState4[1];

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_index2.ButtonGroup, {
    expanded: true,
    onChange: setValue,
    value: value
  }, /*#__PURE__*/_react.default.createElement(_index.Button, {
    icon: _icons.formatBold,
    value: "bold"
  }), /*#__PURE__*/_react.default.createElement(_index.Button, {
    icon: _icons.formatItalic,
    value: "italic"
  }), /*#__PURE__*/_react.default.createElement(_index.Button, {
    icon: _icons.formatUnderline,
    value: "underline"
  })));
};

exports._expanded = _expanded;

var _segmented = function () {
  var _React$useState5 = _react.default.useState('bold'),
      _React$useState6 = (0, _slicedToArray2.default)(_React$useState5, 2),
      value = _React$useState6[0],
      setValue = _React$useState6[1];

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_index2.ButtonGroup, {
    onChange: setValue,
    segmented: true,
    value: value
  }, /*#__PURE__*/_react.default.createElement(_index.Button, {
    icon: _icons.formatBold,
    value: "bold"
  }), /*#__PURE__*/_react.default.createElement(_index.Button, {
    icon: _icons.formatItalic,
    value: "italic"
  }), /*#__PURE__*/_react.default.createElement(_index.Button, {
    icon: _icons.formatUnderline,
    value: "underline"
  })));
};

exports._segmented = _segmented;
