// @wp-g2/components/dist/cjs/Select/Select.styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Suffix = exports.arrowPadding = exports.placeholder = exports.ArrowWrapperView = exports.select = exports.base = exports.ARROW_WRAPPER_WIDTH = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _styles = require("@wp-g2/styles");

var _Flex = require("../Flex");

function _templateObject6() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tpointer-events: none;\n\tposition: absolute;\n\tright: 0;\n"]);

  _templateObject6 = function () {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tpadding-right: ", ";\n"]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tcolor: ", ";\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\talign-items: center;\n\tbottom: 0;\n\tdisplay: flex;\n\tjustify-content: center;\n\tpadding: 0 5px;\n\tpointer-events: none;\n\tposition: absolute;\n\tright: 0;\n\ttop: 0;\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tpadding-left: 8px;\n\tpadding-right: 30px;\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tpadding: 0;\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var ARROW_WRAPPER_WIDTH = 24;
exports.ARROW_WRAPPER_WIDTH = ARROW_WRAPPER_WIDTH;
var base = (0, _styles.css)(_templateObject());
exports.base = base;
var select = (0, _styles.css)(_templateObject2());
exports.select = select;
var ArrowWrapperView = (0, _styles.styled)(_Flex.FlexItem)(_templateObject3());
exports.ArrowWrapperView = ArrowWrapperView;
var placeholder = (0, _styles.css)(_templateObject4(), _styles.ui.get('colorTextMuted'));
exports.placeholder = placeholder;
var arrowPadding = (0, _styles.css)(_templateObject5(), _styles.ui.value.px(ARROW_WRAPPER_WIDTH));
exports.arrowPadding = arrowPadding;
var Suffix = (0, _styles.css)(_templateObject6());
exports.Suffix = Suffix;
