// @wp-g2/components/dist/es/Card/__stories__/Card.stories.js
import { faker } from '@wp-g2/protokit';
import React from 'react';
import { Button, Divider, Grid, Heading, Spacer, Text, VStack } from '../../index';
import { Card, CardBody, CardFooter } from '../index';
export default {
  component: Card,
  title: 'Components/Card'
};

var ExampleCard = function () {
  return /*#__PURE__*/React.createElement(Card, null, /*#__PURE__*/React.createElement(VStack, {
    expanded: true
  }, /*#__PURE__*/React.createElement(CardBody, null, /*#__PURE__*/React.createElement(VStack, null, /*#__PURE__*/React.createElement(Heading, null, faker.lorem.sentence()), /*#__PURE__*/React.createElement(Text, null, faker.random.arrayElement([faker.lorem.sentence(), faker.lorem.words(), faker.lorem.paragraphs()])))), /*#__PURE__*/React.createElement(Divider, null), /*#__PURE__*/React.createElement(CardBody, null, /*#__PURE__*/React.createElement(VStack, null, /*#__PURE__*/React.createElement(Heading, null, faker.lorem.sentence()), /*#__PURE__*/React.createElement(Text, null, faker.random.arrayElement([faker.lorem.sentence(), faker.lorem.words(), faker.lorem.paragraphs()])))), /*#__PURE__*/React.createElement(Spacer, null), /*#__PURE__*/React.createElement(Divider, null), /*#__PURE__*/React.createElement(CardFooter, null, /*#__PURE__*/React.createElement(Button, null, "Action"))));
};

export var adaptiveHeight = function () {
  return /*#__PURE__*/React.createElement(Grid, {
    columns: [1, 2, 4]
  }, /*#__PURE__*/React.createElement(ExampleCard, null), /*#__PURE__*/React.createElement(ExampleCard, null), /*#__PURE__*/React.createElement(ExampleCard, null), /*#__PURE__*/React.createElement(ExampleCard, null), /*#__PURE__*/React.createElement(ExampleCard, null), /*#__PURE__*/React.createElement(ExampleCard, null), /*#__PURE__*/React.createElement(ExampleCard, null), /*#__PURE__*/React.createElement(ExampleCard, null));
};
export var _default = function () {
  return /*#__PURE__*/React.createElement(Card, {
    elevation: 5,
    isBorderless: true
  }, /*#__PURE__*/React.createElement(CardBody, null, "Card"));
};
