// @wp-g2/components/dist/cjs/Dropdown/__stories__/Dropdown.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._renderFunction = exports._default = exports.WithModal = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reakit = require("reakit");

var _index = require("../../index");

var _index2 = require("../index");

var _default2 = {
  component: _index2.Dropdown,
  title: 'Components/Dropdown'
};
exports.default = _default2;

var WithModal = function () {
  var dialog = (0, _reakit.useDialogState)({
    animated: true
  });
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_index2.Dropdown, null, /*#__PURE__*/_react.default.createElement(_index2.DropdownTrigger, null, "Dropdown"), /*#__PURE__*/_react.default.createElement(_index2.DropdownMenu, null, /*#__PURE__*/_react.default.createElement(_index2.DropdownMenuItem, {
    onClick: dialog.show
  }, "One"), /*#__PURE__*/_react.default.createElement(_index2.DropdownMenuItem, {
    onClick: dialog.show
  }, "Two"))), /*#__PURE__*/_react.default.createElement(_index.Modal, {
    dialog: dialog
  }, /*#__PURE__*/_react.default.createElement(_index.ModalHeader, null), /*#__PURE__*/_react.default.createElement(_index.ModalBody, null, "Hello"), /*#__PURE__*/_react.default.createElement(_index.ModalFooter, null, /*#__PURE__*/_react.default.createElement(_index.Button, {
    onClick: dialog.hide,
    variant: "primary"
  }, "Save"), /*#__PURE__*/_react.default.createElement(_index.Button, {
    onClick: dialog.hide
  }, "Cancel"))));
};

exports.WithModal = WithModal;

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(_index2.Dropdown, {
    visible: true
  }, /*#__PURE__*/_react.default.createElement(_index2.DropdownTrigger, null, "Dropdown"), /*#__PURE__*/_react.default.createElement(_index2.DropdownMenu, null, /*#__PURE__*/_react.default.createElement(_index2.DropdownMenuItem, null, "One"), /*#__PURE__*/_react.default.createElement(_index2.DropdownMenuItem, null, "Two"), /*#__PURE__*/_react.default.createElement(_index2.DropdownMenuItem, null, "Three"), /*#__PURE__*/_react.default.createElement(_index.Divider, null), /*#__PURE__*/_react.default.createElement(_index2.DropdownMenuItem, null, "Three"), /*#__PURE__*/_react.default.createElement(_index2.DropdownMenuItem, null, "Three"), /*#__PURE__*/_react.default.createElement(_index2.DropdownMenuItem, null, "Three"), /*#__PURE__*/_react.default.createElement(_index2.DropdownMenuItem, null, "Three"), /*#__PURE__*/_react.default.createElement(_index2.DropdownMenuItem, null, "Three"), /*#__PURE__*/_react.default.createElement(_index2.DropdownMenuItem, null, "Three"), /*#__PURE__*/_react.default.createElement(_index2.DropdownMenuItem, null, "Three")));
};

exports._default = _default;

var _renderFunction = function () {
  return /*#__PURE__*/_react.default.createElement(_index2.Dropdown, {
    visible: true
  }, function (_ref) {
    var onClose = _ref.onClose;
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_index2.DropdownTrigger, null, "Dropdown"), /*#__PURE__*/_react.default.createElement(_index2.DropdownMenu, null, /*#__PURE__*/_react.default.createElement(_index2.DropdownMenuItem, {
      onClick: onClose
    }, "One"), /*#__PURE__*/_react.default.createElement(_index2.DropdownMenuItem, null, "Two"), /*#__PURE__*/_react.default.createElement(_index2.DropdownMenuItem, null, "Three")));
  });
};

exports._renderFunction = _renderFunction;
