// @wp-g2/components/dist/cjs/BaseButton/BaseButton.styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.split = exports.currentColor = exports.narrow = exports.subtleControl = exports.control = exports.subtle = exports.LoadingOverlay = exports.CaretWrapper = exports.PrefixSuffix = exports.noWrap = exports.Content = exports.loading = exports.icon = exports.xxSmall = exports.xSmall = exports.small = exports.large = exports.rounded = exports.block = exports.destructive = exports.Button = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _styles = require("@wp-g2/styles");

function _templateObject21() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tborder-bottom-left-radius: 0px;\n\tborder-top-left-radius: 0px;\n"]);

  _templateObject21 = function () {
    return data;
  };

  return data;
}

function _templateObject20() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tcolor: currentColor;\n\n\t&:hover,\n\t&:focus {\n\t\tcolor: currentColor;\n\t}\n"]);

  _templateObject20 = function () {
    return data;
  };

  return data;
}

function _templateObject19() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tpadding-left: ", ";\n\tpadding-right: ", ";\n"]);

  _templateObject19 = function () {
    return data;
  };

  return data;
}

function _templateObject18() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbackground-color: transparent;\n\n\t&:hover,\n\t&:focus {\n\t\tbackground-color: ", ";\n\t}\n\n\t&:focus {\n\t\tborder-color: ", ";\n\t}\n\n\t&:active {\n\t\tbackground-color: ", ";\n\t}\n\n\t&[data-active='true'] {\n\t\tbackground: ", ";\n\t\tcolor: ", ";\n\t}\n"]);

  _templateObject18 = function () {
    return data;
  };

  return data;
}

function _templateObject17() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbackground-color: ", ";\n\tborder: 1px solid;\n\tborder-color: ", ";\n\tcolor: ", ";\n\tfont-family: ", ";\n\tfont-size: ", ";\n\n\t&:hover,\n\t&:active,\n\t&:focus {\n\t\tcolor: ", ";\n\t}\n\n\t&:focus {\n\t\tborder-color: ", ";\n\t\tbox-shadow: ", ";\n\t}\n"]);

  _templateObject17 = function () {
    return data;
  };

  return data;
}

function _templateObject16() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tborder-color: ", ";\n\tcolor: ", ";\n\n\t&:hover,\n\t&:active,\n\t&:focus {\n\t\tborder-color: ", ";\n\t\tcolor: ", ";\n\t}\n\n\t&:focus {\n\t\tborder-color: ", ";\n\t}\n"]);

  _templateObject16 = function () {
    return data;
  };

  return data;
}

function _templateObject15() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbottom: 0;\n\tleft: 0;\n\tpointer-events: none;\n\tposition: absolute;\n\tright: 0;\n\ttop: 0;\n"]);

  _templateObject15 = function () {
    return data;
  };

  return data;
}

function _templateObject14() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tmargin-left: 0 !important;\n\tposition: relative;\n\tright: ", ";\n"]);

  _templateObject14 = function () {
    return data;
  };

  return data;
}

function _templateObject13() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\topacity: 1;\n\n\tsvg {\n\t\tdisplay: block;\n\t\tuser-select: none;\n\t}\n"]);

  _templateObject13 = function () {
    return data;
  };

  return data;
}

function _templateObject12() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\twhite-space: nowrap;\n"]);

  _templateObject12 = function () {
    return data;
  };

  return data;
}

function _templateObject11() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tfont-size: ", ";\n\t/* line-height: 0; */\n\topacity: 1;\n"]);

  _templateObject11 = function () {
    return data;
  };

  return data;
}

function _templateObject10() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\topacity: 0;\n"]);

  _templateObject10 = function () {
    return data;
  };

  return data;
}

function _templateObject9() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n"]);

  _templateObject9 = function () {
    return data;
  };

  return data;
}

function _templateObject8() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\tmin-height: ", ";\n\n\t&[data-icon='true'] {\n\t\tmin-width: ", ";\n\t}\n"]);

  _templateObject8 = function () {
    return data;
  };

  return data;
}

function _templateObject7() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\t", ";\n\tmin-height: ", ";\n\n\t&[data-icon='true'] {\n\t\tmin-width: ", ";\n\t}\n"]);

  _templateObject7 = function () {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\t", ";\n\tmin-height: ", ";\n\n\t&[data-icon='true'] {\n\t\tmin-width: ", ";\n\t}\n"]);

  _templateObject6 = function () {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tmin-height: ", ";\n\n\t&[data-icon='true'] {\n\t\tmin-width: ", ";\n\t}\n"]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tdisplay: flex;\n\twidth: 100%;\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tcolor: ", ";\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\talign-items: center;\n\tappearance: none;\n\tbackground-color: transparent;\n\tborder-color: transparent;\n\tborder-radius: ", ";\n\tborder-style: solid;\n\tborder-width: 1px;\n\tbox-shadow: 0 0 0 ", " transparent;\n\tbox-sizing: border-box;\n\tcolor: ", ";\n\tcursor: pointer;\n\tdisplay: inline-flex;\n\tfont-size: ", ";\n\theight: auto;\n\tline-height: 1;\n\tmin-height: ", ";\n\toutline: none;\n\tpadding-bottom: ", ";\n\tpadding-left: ", ";\n\tpadding-right: ", ";\n\tpadding-top: ", ";\n\tposition: relative;\n\ttext-decoration: none;\n\ttouch-action: manipulation; /* Prevents zooming on mobile */\n\tuser-select: none;\n\twidth: auto;\n\n\t&:hover,\n\t&:active,\n\t&:focus {\n\t\ttransition: all ", "\n\t\t\tcubic-bezier(0.12, 0.8, 0.32, 1);\n\t}\n\n\t&[disabled]:not([aria-busy='true']),\n\t&[aria-disabled='true']:not([aria-busy='true']) {\n\t\t", ";\n\t\tcursor: auto;\n\t}\n\n\t&:focus {\n\t\tbox-shadow: ", ";\n\t}\n\n\t&[data-destructive='true'] {\n\t\t&:focus {\n\t\t\tbox-shadow: ", ";\n\t\t}\n\t}\n\n\t&[data-icon='true'] {\n\t\tmin-width: ", ";\n\t}\n\n\tsvg {\n\t\tdisplay: block;\n\t}\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var Button = (0, _styles.css)(_templateObject(), _styles.ui.get('controlBorderRadius'), _styles.ui.get('controlBoxShadowFocusSize'), _styles.ui.color.text, _styles.ui.get('fontSize'), _styles.ui.get('controlHeight'), _styles.ui.space(1), _styles.ui.get('controlPaddingX'), _styles.ui.get('controlPaddingX'), _styles.ui.space(1), _styles.ui.get('transitionDuration'), _styles.ui.opacity.muted, _styles.ui.get('controlBoxShadowFocus'), _styles.ui.get('controlDestructiveBoxShadowFocus'), _styles.ui.get('controlHeight'));
exports.Button = Button;
var destructive = (0, _styles.css)(_templateObject2(), _styles.ui.color.destructive);
exports.destructive = destructive;
var block = (0, _styles.css)(_templateObject3());
exports.block = block;
var rounded = (0, _styles.css)(_templateObject4(), _styles.ui.borderRadius.circle);
exports.rounded = rounded;
var large = (0, _styles.css)(_templateObject5(), _styles.ui.get('controlHeightLarge'), _styles.ui.get('controlHeightLarge'));
exports.large = large;
var small = (0, _styles.css)(_templateObject6(), _styles.ui.padding.y(_styles.ui.space(0.5)), _styles.ui.padding.x(_styles.ui.get('controlPaddingXSmall')), _styles.ui.get('controlHeightSmall'), _styles.ui.get('controlPaddingXSmall'));
exports.small = small;
var xSmall = (0, _styles.css)(_templateObject7(), _styles.ui.padding.y(_styles.ui.space(0.5)), _styles.ui.padding.x(_styles.ui.get('controlPaddingXXSmall')), _styles.ui.get('controlHeightXSmall'), _styles.ui.get('controlHeightXXSmall'));
exports.xSmall = xSmall;
var xxSmall = (0, _styles.css)(_templateObject8(), _styles.ui.padding(0), _styles.ui.get('controlHeightXXSmall'), _styles.ui.get('controlHeightXXSmall'));
exports.xxSmall = xxSmall;
var icon = (0, _styles.css)(_templateObject9(), _styles.ui.padding.x(0));
exports.icon = icon;
var loading = (0, _styles.css)(_templateObject10());
exports.loading = loading;
var Content = (0, _styles.css)(_templateObject11(), _styles.ui.get('fontSize'));
exports.Content = Content;
var noWrap = (0, _styles.css)(_templateObject12());
exports.noWrap = noWrap;
var PrefixSuffix = (0, _styles.css)(_templateObject13());
exports.PrefixSuffix = PrefixSuffix;
var CaretWrapper = (0, _styles.css)(_templateObject14(), _styles.ui.space(-2));
exports.CaretWrapper = CaretWrapper;
var LoadingOverlay = (0, _styles.css)(_templateObject15());
exports.LoadingOverlay = LoadingOverlay;
var subtle = (0, _styles.css)(_templateObject16(), _styles.ui.get('controlBorderColorSubtle'), _styles.ui.color.text, _styles.ui.get('controlBorderColorSubtle'), _styles.ui.color.text, _styles.ui.get('buttonPrimaryBorderColorFocus'));
exports.subtle = subtle;
var control = (0, _styles.css)(_templateObject17(), _styles.ui.get('controlBackgroundColor'), _styles.ui.get('controlBorderColor'), _styles.ui.color.text, _styles.ui.get('fontFamily'), _styles.ui.get('fontSize'), _styles.ui.color.text, _styles.ui.get('buttonPrimaryBorderColorFocus'), _styles.ui.get('controlBoxShadowFocus'));
exports.control = control;
var subtleControl = (0, _styles.css)(_templateObject18(), _styles.ui.get('controlBackgroundColor'), _styles.ui.get('buttonPrimaryBorderColorFocus'), _styles.ui.get('controlBackgroundColorActive'), _styles.ui.get('colorText'), _styles.ui.get('colorTextInverted'));
exports.subtleControl = subtleControl;
var narrow = (0, _styles.css)(_templateObject19(), _styles.ui.get('controlPaddingX'), _styles.ui.get('controlPaddingX'));
exports.narrow = narrow;
var currentColor = (0, _styles.css)(_templateObject20());
exports.currentColor = currentColor;
var split = (0, _styles.css)(_templateObject21());
exports.split = split;
