// @wp-g2/components/dist/cjs/BaseButton/BaseButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _icons = require("@wordpress/icons");

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _reakit = require("reakit");

var _ButtonGroup = require("../ButtonGroup");

var _Elevation = require("../Elevation");

var _Flex = require("../Flex");

var _Icon = require("../Icon");

var _View = require("../View");

var styles = _interopRequireWildcard(require("./BaseButton.styles"));

var _BaseButtonLoadingOverlay = _interopRequireDefault(require("./BaseButtonLoadingOverlay"));

var _useBaseButton2 = require("./useBaseButton");

/**
 * @param {import('@wp-g2/create-styles').ViewOwnProps<import('./types').Props, 'button'>} props
 * @param {import('react').Ref<any>} forwardedRef
 */
function BaseButton(props, forwardedRef) {
  var _useBaseButton = (0, _useBaseButton2.useBaseButton)(props),
      asProp = _useBaseButton.as,
      children = _useBaseButton.children,
      _useBaseButton$disabl = _useBaseButton.disabled,
      disabled = _useBaseButton$disabl === void 0 ? false : _useBaseButton$disabl,
      _useBaseButton$elevat = _useBaseButton.elevation,
      elevation = _useBaseButton$elevat === void 0 ? 0 : _useBaseButton$elevat,
      elevationActive = _useBaseButton.elevationActive,
      elevationFocus = _useBaseButton.elevationFocus,
      elevationHover = _useBaseButton.elevationHover,
      _useBaseButton$hasCar = _useBaseButton.hasCaret,
      hasCaret = _useBaseButton$hasCar === void 0 ? false : _useBaseButton$hasCar,
      href = _useBaseButton.href,
      icon = _useBaseButton.icon,
      _useBaseButton$iconSi = _useBaseButton.iconSize,
      iconSize = _useBaseButton$iconSi === void 0 ? 16 : _useBaseButton$iconSi,
      _useBaseButton$isActi = _useBaseButton.isActive,
      isActive = _useBaseButton$isActi === void 0 ? false : _useBaseButton$isActi,
      _useBaseButton$isDest = _useBaseButton.isDestructive,
      isDestructive = _useBaseButton$isDest === void 0 ? false : _useBaseButton$isDest,
      _useBaseButton$isFocu = _useBaseButton.isFocused,
      isFocused = _useBaseButton$isFocu === void 0 ? false : _useBaseButton$isFocu,
      _useBaseButton$isLoad = _useBaseButton.isLoading,
      isLoading = _useBaseButton$isLoad === void 0 ? false : _useBaseButton$isLoad,
      _useBaseButton$noWrap = _useBaseButton.noWrap,
      noWrap = _useBaseButton$noWrap === void 0 ? true : _useBaseButton$noWrap,
      prefix = _useBaseButton.prefix,
      suffix = _useBaseButton.suffix,
      otherProps = (0, _objectWithoutProperties2.default)(_useBaseButton, ["as", "children", "disabled", "elevation", "elevationActive", "elevationFocus", "elevationHover", "hasCaret", "href", "icon", "iconSize", "isActive", "isDestructive", "isFocused", "isLoading", "noWrap", "prefix", "suffix"]);

  var _useButtonGroupContex = (0, _ButtonGroup.useButtonGroupContext)(),
      buttonGroup = _useButtonGroupContex.buttonGroup;

  var BaseComponent = buttonGroup ? _reakit.Radio : _View.View;
  var as = asProp || (href ? 'a' : 'button');
  return (
    /*#__PURE__*/
    // @ts-ignore No idea why TS is confused about this but ReakitRadio and ReakitButton are definitely renderable
    _react.default.createElement(BaseComponent, Object.assign({
      "aria-busy": isLoading,
      as: as,
      "data-active": isActive,
      "data-destructive": isDestructive,
      "data-focused": isFocused,
      "data-icon": !!icon,
      disabled: disabled || isLoading,
      href: href
    }, buttonGroup || {}, otherProps, {
      ref: forwardedRef
    }), /*#__PURE__*/_react.default.createElement(_BaseButtonLoadingOverlay.default, {
      isLoading: isLoading
    }), prefix && /*#__PURE__*/_react.default.createElement(_Flex.FlexItem, Object.assign({
      as: "span",
      className: (0, _styles.cx)(styles.PrefixSuffix, isLoading && styles.loading)
    }, _styles.ui.$('ButtonPrefix')), prefix), icon && /*#__PURE__*/_react.default.createElement(_Flex.FlexItem, Object.assign({
      as: "span",
      className: (0, _styles.cx)(styles.PrefixSuffix, isLoading && styles.loading)
    }, _styles.ui.$('ButtonIcon')), /*#__PURE__*/_react.default.createElement(_Icon.Icon, {
      icon: icon,
      size: iconSize
    })), children && /*#__PURE__*/_react.default.createElement(_Flex.FlexItem, Object.assign({
      as: "span",
      className: (0, _styles.cx)(styles.Content, isLoading && styles.loading, noWrap && styles.noWrap),
      isBlock: true
    }, _styles.ui.$('ButtonContent')), children), suffix && /*#__PURE__*/_react.default.createElement(_Flex.FlexItem, Object.assign({
      as: "span",
      className: (0, _styles.cx)(styles.PrefixSuffix, isLoading && styles.loading)
    }, _styles.ui.$('ButtonSuffix')), suffix), hasCaret && /*#__PURE__*/_react.default.createElement(_Flex.FlexItem, Object.assign({
      as: "span",
      className: (0, _styles.cx)(styles.CaretWrapper, isLoading && styles.loading)
    }, _styles.ui.$('ButtonCaret')), /*#__PURE__*/_react.default.createElement(_Icon.Icon, {
      icon: _icons.chevronDown,
      size: 16
    })), /*#__PURE__*/_react.default.createElement(_Elevation.Elevation, Object.assign({
      active: elevationActive,
      as: "span",
      focus: elevationFocus,
      hover: elevationHover,
      offset: -1,
      value: elevation
    }, _styles.ui.$('ButtonElevation'))))
  );
}
/**
 * `BaseButton` is a primitive component used to create actionable components (e.g. `Button`).
 */


var ConnectedBaseButton = (0, _context.contextConnect)(BaseButton, 'BaseButton');
var _default = ConnectedBaseButton;
exports.default = _default;
