// @wp-g2/components/dist/cjs/Card/__stories__/Card.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.adaptiveHeight = exports.default = void 0;

var _protokit = require("@wp-g2/protokit");

var _react = _interopRequireDefault(require("react"));

var _index = require("../../index");

var _index2 = require("../index");

var _default2 = {
  component: _index2.Card,
  title: 'Components/Card'
};
exports.default = _default2;

var ExampleCard = function () {
  return /*#__PURE__*/_react.default.createElement(_index2.Card, null, /*#__PURE__*/_react.default.createElement(_index.VStack, {
    expanded: true
  }, /*#__PURE__*/_react.default.createElement(_index2.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.Heading, null, _protokit.faker.lorem.sentence()), /*#__PURE__*/_react.default.createElement(_index.Text, null, _protokit.faker.random.arrayElement([_protokit.faker.lorem.sentence(), _protokit.faker.lorem.words(), _protokit.faker.lorem.paragraphs()])))), /*#__PURE__*/_react.default.createElement(_index.Divider, null), /*#__PURE__*/_react.default.createElement(_index2.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.Heading, null, _protokit.faker.lorem.sentence()), /*#__PURE__*/_react.default.createElement(_index.Text, null, _protokit.faker.random.arrayElement([_protokit.faker.lorem.sentence(), _protokit.faker.lorem.words(), _protokit.faker.lorem.paragraphs()])))), /*#__PURE__*/_react.default.createElement(_index.Spacer, null), /*#__PURE__*/_react.default.createElement(_index.Divider, null), /*#__PURE__*/_react.default.createElement(_index2.CardFooter, null, /*#__PURE__*/_react.default.createElement(_index.Button, null, "Action"))));
};

var adaptiveHeight = function () {
  return /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: [1, 2, 4]
  }, /*#__PURE__*/_react.default.createElement(ExampleCard, null), /*#__PURE__*/_react.default.createElement(ExampleCard, null), /*#__PURE__*/_react.default.createElement(ExampleCard, null), /*#__PURE__*/_react.default.createElement(ExampleCard, null), /*#__PURE__*/_react.default.createElement(ExampleCard, null), /*#__PURE__*/_react.default.createElement(ExampleCard, null), /*#__PURE__*/_react.default.createElement(ExampleCard, null), /*#__PURE__*/_react.default.createElement(ExampleCard, null));
};

exports.adaptiveHeight = adaptiveHeight;

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(_index2.Card, {
    elevation: 5,
    isBorderless: true
  }, /*#__PURE__*/_react.default.createElement(_index2.CardBody, null, "Card"));
};

exports._default = _default;
