// @wordpress/components/build-module/ui/control-group/component.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { contextConnect } from '@wp-g2/context';
/**
 * Internal dependencies
 */

import { Flex } from '../flex';
import { Grid } from '../grid';
import { useControlGroup } from './hook';
/**
 * @param {import('@wp-g2/create-styles').ViewOwnProps<import('./types').Props, 'div'>} props
 * @param {import('react').Ref<any>} forwardedRef
 */

function ControlGroup(props, forwardedRef) {
  var _useControlGroup = useControlGroup(props),
      children = _useControlGroup.children,
      _useControlGroup$dire = _useControlGroup.direction,
      direction = _useControlGroup$dire === void 0 ? 'row' : _useControlGroup$dire,
      templateColumns = _useControlGroup.templateColumns,
      otherProps = _objectWithoutProperties(_useControlGroup, ["children", "direction", "templateColumns"]);

  var isGrid = !!templateColumns;

  if (isGrid) {
    return createElement(Grid, _extends({
      gap: 0,
      templateColumns: templateColumns
    }, otherProps, {
      ref: forwardedRef
    }), children);
  }

  return createElement(Flex, _extends({
    direction: direction,
    gap: "-1px"
  }, otherProps, {
    ref: forwardedRef
  }), children);
}
/**
 * `ControlGroup` is a layout-based component for rendering a group of
 * control-based components, such as `Button`, `Select` or `TextInput`.
 * Control components that render within `ControlGroup` automatically
 * have their borders offset and border-radii rounded.
 *
 * @example
 * ```jsx
 * import { Button, ControlGroup, Select, TextInput } from `@wordpress/components/ui`;
 *
 * function Example() {
 *   return (
 *     <ControlGroup templateColumns="auto 1fr auto">
 *       <Select />
 *       <TextInput placeholder="First name" />
 *       <Button variant="primary" />
 *     </ControlGroup>
 *   );
 * }
 * ```
 */


var ConnectedControlGroup = contextConnect(ControlGroup, 'ControlGroup');
export default ConnectedControlGroup;
