// @wordpress/components/build-module/ui/control-group/styles.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/esm/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7;

/**
 * External dependencies
 */
import { css } from '@wp-g2/styles';
export var first = css(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n\tborder-bottom-right-radius: 0;\n\tborder-top-right-radius: 0;\n"])));
export var middle = css(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n\tborder-radius: 0;\n"])));
export var last = css(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n\tborder-bottom-left-radius: 0;\n\tborder-top-left-radius: 0;\n"])));
export var firstRow = css(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n\tborder-bottom-left-radius: 0;\n\tborder-bottom-right-radius: 0;\n"])));
export var lastRow = css(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["\n\tborder-top-left-radius: 0;\n\tborder-top-right-radius: 0;\n"])));
export var itemFocus = css(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["\n\t> * {\n\t\t&:focus-within {\n\t\t\tz-index: 1;\n\t\t}\n\t}\n"])));
export var itemGrid = css(_templateObject7 || (_templateObject7 = _taggedTemplateLiteral(["\n\t> * + *:not( marquee ) {\n\t\tmargin-left: -1px;\n\t\twidth: calc( 100% + 1px );\n\t}\n"])));
