// @wordpress/block-editor/build-module/components/inserter/menu.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement, Fragment } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { useState, useCallback, useMemo } from '@wordpress/element';
import { VisuallyHidden } from '@wordpress/components';
import { __ } from '@wordpress/i18n';
import { useSelect } from '@wordpress/data';
/**
 * Internal dependencies
 */

import Tips from './tips';
import InserterSearchForm from './search-form';
import InserterPreviewPanel from './preview-panel';
import BlockTypesTab from './block-types-tab';
import BlockPatternsTabs from './block-patterns-tab';
import ReusableBlocksTab from './reusable-blocks-tab';
import InserterSearchResults from './search-results';
import useInsertionPoint from './hooks/use-insertion-point';
import InserterTabs from './tabs';
import { store as blockEditorStore } from '../../store';

function InserterMenu(_ref) {
  var rootClientId = _ref.rootClientId,
      clientId = _ref.clientId,
      isAppender = _ref.isAppender,
      __experimentalInsertionIndex = _ref.__experimentalInsertionIndex,
      onSelect = _ref.onSelect,
      showInserterHelpPanel = _ref.showInserterHelpPanel,
      showMostUsedBlocks = _ref.showMostUsedBlocks,
      _ref$shouldFocusBlock = _ref.shouldFocusBlock,
      shouldFocusBlock = _ref$shouldFocusBlock === void 0 ? true : _ref$shouldFocusBlock;

  var _useState = useState(''),
      _useState2 = _slicedToArray(_useState, 2),
      filterValue = _useState2[0],
      setFilterValue = _useState2[1];

  var _useState3 = useState(null),
      _useState4 = _slicedToArray(_useState3, 2),
      hoveredItem = _useState4[0],
      setHoveredItem = _useState4[1];

  var _useState5 = useState(null),
      _useState6 = _slicedToArray(_useState5, 2),
      selectedPatternCategory = _useState6[0],
      setSelectedPatternCategory = _useState6[1];

  var _useInsertionPoint = useInsertionPoint({
    rootClientId: rootClientId,
    clientId: clientId,
    isAppender: isAppender,
    insertionIndex: __experimentalInsertionIndex,
    shouldFocusBlock: shouldFocusBlock
  }),
      _useInsertionPoint2 = _slicedToArray(_useInsertionPoint, 3),
      destinationRootClientId = _useInsertionPoint2[0],
      onInsertBlocks = _useInsertionPoint2[1],
      onToggleInsertionPoint = _useInsertionPoint2[2];

  var _useSelect = useSelect(function (select) {
    var _getSettings$__experi;

    var _select = select(blockEditorStore),
        __experimentalGetAllowedPatterns = _select.__experimentalGetAllowedPatterns,
        getSettings = _select.getSettings;

    return {
      showPatterns: !destinationRootClientId || !!__experimentalGetAllowedPatterns(destinationRootClientId).length,
      hasReusableBlocks: !!((_getSettings$__experi = getSettings().__experimentalReusableBlocks) !== null && _getSettings$__experi !== void 0 && _getSettings$__experi.length)
    };
  }, [destinationRootClientId]),
      showPatterns = _useSelect.showPatterns,
      hasReusableBlocks = _useSelect.hasReusableBlocks;

  var onInsert = useCallback(function (blocks, meta, shouldForceFocusBlock) {
    onInsertBlocks(blocks, meta, shouldForceFocusBlock);
    onSelect();
  }, [onInsertBlocks, onSelect]);
  var onInsertPattern = useCallback(function (blocks, patternName) {
    onInsertBlocks(blocks, {
      patternName: patternName
    });
    onSelect();
  }, [onInsertBlocks, onSelect]);
  var onHover = useCallback(function (item) {
    onToggleInsertionPoint(!!item);
    setHoveredItem(item);
  }, [onToggleInsertionPoint, setHoveredItem]);
  var onClickPatternCategory = useCallback(function (patternCategory) {
    setSelectedPatternCategory(patternCategory);
  }, [setSelectedPatternCategory]);
  var blocksTab = useMemo(function () {
    return createElement(Fragment, null, createElement("div", {
      className: "block-editor-inserter__block-list"
    }, createElement(BlockTypesTab, {
      rootClientId: destinationRootClientId,
      onInsert: onInsert,
      onHover: onHover,
      showMostUsedBlocks: showMostUsedBlocks
    })), showInserterHelpPanel && createElement("div", {
      className: "block-editor-inserter__tips"
    }, createElement(VisuallyHidden, {
      as: "h2"
    }, __('A tip for using the block editor')), createElement(Tips, null)));
  }, [destinationRootClientId, onInsert, onHover, filterValue, showMostUsedBlocks, showInserterHelpPanel]);
  var patternsTab = useMemo(function () {
    return createElement(BlockPatternsTabs, {
      rootClientId: destinationRootClientId,
      onInsert: onInsertPattern,
      onClickCategory: onClickPatternCategory,
      selectedCategory: selectedPatternCategory
    });
  }, [destinationRootClientId, onInsertPattern, onClickPatternCategory, selectedPatternCategory]);
  var reusableBlocksTab = useMemo(function () {
    return createElement(ReusableBlocksTab, {
      rootClientId: destinationRootClientId,
      onInsert: onInsert,
      onHover: onHover
    });
  }, [destinationRootClientId, onInsert, onHover]);
  var getCurrentTab = useCallback(function (tab) {
    if (tab.name === 'blocks') {
      return blocksTab;
    } else if (tab.name === 'patterns') {
      return patternsTab;
    }

    return reusableBlocksTab;
  }, [blocksTab, patternsTab, reusableBlocksTab]);
  return createElement("div", {
    className: "block-editor-inserter__menu"
  }, createElement("div", {
    className: "block-editor-inserter__main-area"
  }, createElement("div", {
    className: "block-editor-inserter__content"
  }, createElement(InserterSearchForm, {
    onChange: function onChange(value) {
      if (hoveredItem) setHoveredItem(null);
      setFilterValue(value);
    },
    value: filterValue,
    label: __('Search for blocks and patterns'),
    placeholder: __('Search')
  }), !!filterValue && createElement(InserterSearchResults, {
    filterValue: filterValue,
    onSelect: onSelect,
    onHover: onHover,
    rootClientId: rootClientId,
    clientId: clientId,
    isAppender: isAppender,
    showBlockDirectory: true,
    shouldFocusBlock: shouldFocusBlock
  }), !filterValue && (showPatterns || hasReusableBlocks) && createElement(InserterTabs, {
    showPatterns: showPatterns,
    showReusableBlocks: hasReusableBlocks
  }, getCurrentTab), !filterValue && !showPatterns && !hasReusableBlocks && blocksTab)), showInserterHelpPanel && hoveredItem && createElement(InserterPreviewPanel, {
    item: hoveredItem
  }));
}

export default InserterMenu;
