// @wordpress/block-editor/build-module/components/inserter/menu.native.js
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { View } from 'react-native';
/**
 * WordPress dependencies
 */

import { useEffect, useState, useCallback } from '@wordpress/element';
import { useSelect, useDispatch } from '@wordpress/data';
import { createBlock, rawHandler, store as blocksStore } from '@wordpress/blocks';
import { BottomSheet, BottomSheetConsumer, getClipboard } from '@wordpress/components';
/**
 * Internal dependencies
 */

import InserterSearchResults from './search-results';
import InserterSearchForm from './search-form';
import { store as blockEditorStore } from '../../store';
import { searchItems } from './search-items';
var MIN_ITEMS_FOR_SEARCH = 2;

function InserterMenu(_ref) {
  var _onSelect = _ref.onSelect,
      onDismiss = _ref.onDismiss,
      rootClientId = _ref.rootClientId,
      clientId = _ref.clientId,
      isAppender = _ref.isAppender,
      shouldReplaceBlock = _ref.shouldReplaceBlock,
      insertionIndex = _ref.insertionIndex;

  var _useState = useState(''),
      _useState2 = _slicedToArray(_useState, 2),
      filterValue = _useState2[0],
      setFilterValue = _useState2[1];

  var _useState3 = useState(0),
      _useState4 = _slicedToArray(_useState3, 2),
      searchFormHeight = _useState4[0],
      setSearchFormHeight = _useState4[1]; // eslint-disable-next-line no-undef


  var _useState5 = useState(__DEV__),
      _useState6 = _slicedToArray(_useState5, 2),
      showSearchForm = _useState6[0],
      setShowSearchForm = _useState6[1];

  var _useDispatch = useDispatch(blockEditorStore),
      showInsertionPoint = _useDispatch.showInsertionPoint,
      hideInsertionPoint = _useDispatch.hideInsertionPoint,
      clearSelectedBlock = _useDispatch.clearSelectedBlock,
      insertBlock = _useDispatch.insertBlock,
      removeBlock = _useDispatch.removeBlock,
      resetBlocks = _useDispatch.resetBlocks,
      insertDefaultBlock = _useDispatch.insertDefaultBlock;

  var _useSelect = useSelect(function (select) {
    var _select = select(blockEditorStore),
        getInserterItems = _select.getInserterItems,
        getBlockRootClientId = _select.getBlockRootClientId,
        getBlockSelectionEnd = _select.getBlockSelectionEnd,
        selectBlockEditorStore = _objectWithoutProperties(_select, ["getInserterItems", "getBlockRootClientId", "getBlockSelectionEnd"]);

    var targetRootClientId = rootClientId;

    if (!targetRootClientId && !clientId && !isAppender) {
      var end = getBlockSelectionEnd();

      if (end) {
        targetRootClientId = getBlockRootClientId(end) || undefined;
      }
    }

    return {
      items: getInserterItems(targetRootClientId),
      destinationRootClientId: targetRootClientId,
      getBlockOrder: selectBlockEditorStore.getBlockOrder,
      getBlockCount: selectBlockEditorStore.getBlockCount,
      canInsertBlockType: selectBlockEditorStore.canInsertBlockType
    };
  }),
      items = _useSelect.items,
      destinationRootClientId = _useSelect.destinationRootClientId,
      getBlockOrder = _useSelect.getBlockOrder,
      getBlockCount = _useSelect.getBlockCount,
      canInsertBlockType = _useSelect.canInsertBlockType;

  var _useSelect2 = useSelect(function (select) {
    return select(blocksStore);
  }),
      getBlockType = _useSelect2.getBlockType;

  useEffect(function () {
    var _getItems;

    // Show/Hide insertion point on Mount/Dismount
    if (shouldReplaceBlock) {
      var count = getBlockCount(); // Check if there is a rootClientId because that means it is a nested replaceable block
      // and we don't want to clear/reset all blocks.

      if (count === 1 && !rootClientId) {
        // Removing the last block is not possilble with `removeBlock` action.
        // It always inserts a default block if the last of the blocks have been removed.
        clearSelectedBlock();
        resetBlocks([]);
      } else {
        var blockToReplace = getBlockOrder(destinationRootClientId)[insertionIndex];
        removeBlock(blockToReplace, false);
      }
    }

    showInsertionPoint(destinationRootClientId, insertionIndex); // Show search form if there are enough items to filter.

    if (((_getItems = getItems()) === null || _getItems === void 0 ? void 0 : _getItems.length) < MIN_ITEMS_FOR_SEARCH) {
      setShowSearchForm(false);
    }

    return hideInsertionPoint;
  }, []);
  var onClose = useCallback(function () {
    // if should replace but didn't insert any block
    // re-insert default block
    if (shouldReplaceBlock) {
      insertDefaultBlock({}, destinationRootClientId, insertionIndex);
    }

    onDismiss();
  }, [shouldReplaceBlock, destinationRootClientId, insertionIndex]);
  var onInsert = useCallback(function (item) {
    var name = item.name,
        initialAttributes = item.initialAttributes,
        innerBlocks = item.innerBlocks;
    var newBlock = createBlock(name, initialAttributes, innerBlocks);
    insertBlock(newBlock, insertionIndex, destinationRootClientId);
  }, [insertBlock, destinationRootClientId, insertionIndex]);
  /**
   * Processes the inserter items to check
   * if there's any copied block in the clipboard
   * to add it as an extra item
   */

  function getItems() {
    var _clipboardBlock;

    // Filter out reusable blocks (they will be added in another tab)
    var itemsToDisplay = items.filter(function (_ref2) {
      var name = _ref2.name;
      return name !== 'core/block';
    });
    itemsToDisplay = searchItems(itemsToDisplay, filterValue);
    var clipboard = getClipboard();
    var clipboardBlock = rawHandler({
      HTML: clipboard
    })[0];
    var canAddClipboardBlock = canInsertBlockType((_clipboardBlock = clipboardBlock) === null || _clipboardBlock === void 0 ? void 0 : _clipboardBlock.name, destinationRootClientId);

    if (!canAddClipboardBlock) {
      return itemsToDisplay;
    }

    var _getBlockType = getBlockType(clipboardBlock.name),
        icon = _getBlockType.icon,
        name = _getBlockType.name;

    var _clipboardBlock2 = clipboardBlock,
        initialAttributes = _clipboardBlock2.attributes,
        innerBlocks = _clipboardBlock2.innerBlocks;
    clipboardBlock = {
      id: 'clipboard',
      name: name,
      icon: icon,
      initialAttributes: initialAttributes,
      innerBlocks: innerBlocks
    };
    return [clipboardBlock].concat(_toConsumableArray(itemsToDisplay));
  }

  return createElement(BottomSheet, {
    isVisible: true,
    onClose: onClose,
    hideHeader: true,
    hasNavigation: true,
    setMinHeightToMaxHeight: showSearchForm
  }, createElement(BottomSheetConsumer, null, function (_ref3) {
    var listProps = _ref3.listProps,
        safeAreaBottomInset = _ref3.safeAreaBottomInset;
    return createElement(View, null, showSearchForm && createElement(InserterSearchForm, {
      onChange: function onChange(value) {
        setFilterValue(value);
      },
      value: filterValue,
      onLayout: function onLayout(event) {
        var height = event.nativeEvent.layout.height;
        setSearchFormHeight(height);
      }
    }), createElement(InserterSearchResults, {
      items: getItems(),
      onSelect: function onSelect(item) {
        onInsert(item);

        _onSelect(item);
      },
      listProps: listProps,
      safeAreaBottomInset: safeAreaBottomInset,
      searchFormHeight: searchFormHeight
    }));
  }));
}

export default InserterMenu;
