// @wordpress/block-editor/build-module/components/inserter/search-form.js
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import classnames from 'classnames';
/**
 * WordPress dependencies
 */

import { useInstanceId } from '@wordpress/compose';
import { __ } from '@wordpress/i18n';
import { VisuallyHidden, Button } from '@wordpress/components';
import { Icon, search, closeSmall } from '@wordpress/icons';
import { useRef } from '@wordpress/element';

function InserterSearchForm(_ref) {
  var className = _ref.className,
      _onChange = _ref.onChange,
      value = _ref.value,
      label = _ref.label,
      placeholder = _ref.placeholder;
  var instanceId = useInstanceId(InserterSearchForm);
  var searchInput = useRef();
  return createElement("div", {
    className: classnames('block-editor-inserter__search', className)
  }, createElement(VisuallyHidden, {
    as: "label",
    htmlFor: "block-editor-inserter__search-".concat(instanceId)
  }, label || placeholder), createElement("input", {
    ref: searchInput,
    className: "block-editor-inserter__search-input",
    id: "block-editor-inserter__search-".concat(instanceId),
    type: "search",
    placeholder: placeholder,
    onChange: function onChange(event) {
      return _onChange(event.target.value);
    },
    autoComplete: "off",
    value: value || ''
  }), createElement("div", {
    className: "block-editor-inserter__search-icon"
  }, !!value && createElement(Button, {
    icon: closeSmall,
    label: __('Reset search'),
    onClick: function onClick() {
      _onChange('');

      searchInput.current.focus();
    }
  }), !value && createElement(Icon, {
    icon: search
  })));
}

export default InserterSearchForm;
