// @wordpress/block-editor/build/components/block-alignment-control/ui.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _i18n = require("@wordpress/i18n");

var _components = require("@wordpress/components");

var _icons = require("@wordpress/icons");

var _useAvailableAlignments = _interopRequireDefault(require("./use-available-alignments"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var BLOCK_ALIGNMENTS_CONTROLS = {
  left: {
    icon: _icons.positionLeft,
    title: (0, _i18n.__)('Align left')
  },
  center: {
    icon: _icons.positionCenter,
    title: (0, _i18n.__)('Align center')
  },
  right: {
    icon: _icons.positionRight,
    title: (0, _i18n.__)('Align right')
  },
  wide: {
    icon: _icons.stretchWide,
    title: (0, _i18n.__)('Wide width')
  },
  full: {
    icon: _icons.stretchFullWidth,
    title: (0, _i18n.__)('Full width')
  }
};
var DEFAULT_CONTROL = 'center';
var POPOVER_PROPS = {
  isAlternate: true
};

function BlockAlignmentUI(_ref) {
  var value = _ref.value,
      onChange = _ref.onChange,
      controls = _ref.controls,
      isToolbar = _ref.isToolbar,
      _ref$isCollapsed = _ref.isCollapsed,
      isCollapsed = _ref$isCollapsed === void 0 ? true : _ref$isCollapsed,
      _ref$isToolbarButton = _ref.isToolbarButton,
      isToolbarButton = _ref$isToolbarButton === void 0 ? true : _ref$isToolbarButton;
  var enabledControls = (0, _useAvailableAlignments.default)(controls);

  if (enabledControls.length === 0) {
    return null;
  }

  function applyOrUnset(align) {
    return function () {
      return onChange(value === align ? undefined : align);
    };
  }

  var activeAlignmentControl = BLOCK_ALIGNMENTS_CONTROLS[value];
  var defaultAlignmentControl = BLOCK_ALIGNMENTS_CONTROLS[DEFAULT_CONTROL];
  var UIComponent = isToolbar ? _components.ToolbarGroup : _components.DropdownMenu;
  var extraProps = isToolbar ? {
    isCollapsed: isCollapsed
  } : {
    isToolbarButton: isToolbarButton
  };
  return (0, _element.createElement)(UIComponent, (0, _extends2.default)({
    popoverProps: POPOVER_PROPS,
    icon: activeAlignmentControl ? activeAlignmentControl.icon : defaultAlignmentControl.icon,
    label: (0, _i18n.__)('Align'),
    toggleProps: {
      describedBy: (0, _i18n.__)('Change alignment')
    },
    controls: enabledControls.map(function (control) {
      return _objectSpread(_objectSpread({}, BLOCK_ALIGNMENTS_CONTROLS[control]), {}, {
        isActive: value === control,
        role: isCollapsed ? 'menuitemradio' : undefined,
        onClick: applyOrUnset(control)
      });
    })
  }, extraProps));
}

var _default = BlockAlignmentUI;
exports.default = _default;
