// @wordpress/block-editor/build-module/components/alignment-control/ui.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { createElement } from "@wordpress/element";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * External dependencies
 */
import { find } from 'lodash';
/**
 * WordPress dependencies
 */

import { __, isRTL } from '@wordpress/i18n';
import { DropdownMenu, ToolbarGroup } from '@wordpress/components';
import { alignLeft, alignRight, alignCenter } from '@wordpress/icons';
var DEFAULT_ALIGNMENT_CONTROLS = [{
  icon: alignLeft,
  title: __('Align text left'),
  align: 'left'
}, {
  icon: alignCenter,
  title: __('Align text center'),
  align: 'center'
}, {
  icon: alignRight,
  title: __('Align text right'),
  align: 'right'
}];
var POPOVER_PROPS = {
  position: 'bottom right',
  isAlternate: true
};

function AlignmentUI(_ref) {
  var value = _ref.value,
      onChange = _ref.onChange,
      _ref$alignmentControl = _ref.alignmentControls,
      alignmentControls = _ref$alignmentControl === void 0 ? DEFAULT_ALIGNMENT_CONTROLS : _ref$alignmentControl,
      _ref$label = _ref.label,
      label = _ref$label === void 0 ? __('Align') : _ref$label,
      _ref$describedBy = _ref.describedBy,
      describedBy = _ref$describedBy === void 0 ? __('Change text alignment') : _ref$describedBy,
      _ref$isCollapsed = _ref.isCollapsed,
      isCollapsed = _ref$isCollapsed === void 0 ? true : _ref$isCollapsed,
      isToolbar = _ref.isToolbar,
      _ref$isToolbarButton = _ref.isToolbarButton,
      isToolbarButton = _ref$isToolbarButton === void 0 ? true : _ref$isToolbarButton;

  function applyOrUnset(align) {
    return function () {
      return onChange(value === align ? undefined : align);
    };
  }

  var activeAlignment = find(alignmentControls, function (control) {
    return control.align === value;
  });

  function setIcon() {
    if (activeAlignment) return activeAlignment.icon;
    return isRTL() ? alignRight : alignLeft;
  }

  var UIComponent = isToolbar ? ToolbarGroup : DropdownMenu;
  var extraProps = isToolbar ? {
    isCollapsed: isCollapsed
  } : {
    isToolbarButton: isToolbarButton
  };
  return createElement(UIComponent, _extends({
    icon: setIcon(),
    label: label,
    toggleProps: {
      describedBy: describedBy
    },
    popoverProps: POPOVER_PROPS,
    controls: alignmentControls.map(function (control) {
      var align = control.align;
      var isActive = value === align;
      return _objectSpread(_objectSpread({}, control), {}, {
        isActive: isActive,
        role: isCollapsed ? 'menuitemradio' : undefined,
        onClick: applyOrUnset(align)
      });
    })
  }, extraProps));
}

export default AlignmentUI;
