// @wordpress/components/build-module/button/next.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * Internal dependencies
 */
import { withNext } from '../ui/context';
import { TooltipButton as NextButton } from '../ui/tooltip-button';
var Button = process.env.COMPONENT_SYSTEM_PHASE === 1 ? NextButton : undefined;

var getVariant = function getVariant(_ref) {
  var isPrimary = _ref.isPrimary,
      isTertiary = _ref.isTertiary,
      isSecondary = _ref.isSecondary,
      isLink = _ref.isLink,
      isDefault = _ref.isDefault;

  if (isPrimary) {
    return 'primary';
  }

  if (isTertiary) {
    return 'tertiary';
  }

  if (isSecondary) {
    return 'secondary';
  }

  if (isDefault) {
    return 'plain';
  }

  if (isLink) {
    return 'link';
  }

  return 'plain';
};

var adapter = function adapter(_ref2) {
  var isPrimary = _ref2.isPrimary,
      isTertiary = _ref2.isTertiary,
      isSecondary = _ref2.isSecondary,
      isLink = _ref2.isLink,
      isDefault = _ref2.isDefault,
      isSmall = _ref2.isSmall,
      isBusy = _ref2.isBusy,
      label = _ref2.label,
      ariaLabel = _ref2['aria-label'],
      describedBy = _ref2.describedBy,
      shortcut = _ref2.shortcut,
      tooltipPosition = _ref2.tooltipPosition,
      text = _ref2.text,
      props = _objectWithoutProperties(_ref2, ["isPrimary", "isTertiary", "isSecondary", "isLink", "isDefault", "isSmall", "isBusy", "label", "aria-label", "describedBy", "shortcut", "tooltipPosition", "text"]);

  return _objectSpread(_objectSpread({}, props), {}, {
    describedBy: describedBy,
    variant: getVariant({
      isPrimary: isPrimary,
      isTertiary: isTertiary,
      isSecondary: isSecondary,
      isLink: isLink,
      isDefault: isDefault
    }),
    isActive: isBusy,
    isLoading: isBusy,
    size: isSmall ? 'small' : undefined,
    'aria-label': ariaLabel || label,
    pre: text,
    tooltip: {
      content: describedBy ? describedBy : ariaLabel || label,
      position: tooltipPosition,
      shortcut: shortcut
    }
  });
};

export function withNextComponent(Component) {
  return withNext(Component, Button, 'WPComponentsButton', adapter);
}
