// @wp-g2/context/dist/es/__stories__/ComponentsProvider.stories.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { Card, CardBody, Text, View } from '@wp-g2/components';
import { ui } from '@wp-g2/styles';
import React from 'react';
import { ContextSystemProvider } from '../index';
var SomeContext = /*#__PURE__*/React.createContext();

var useSomeContext = function () {
  return React.useContext(SomeContext);
};

export default {
  component: ContextSystemProvider,
  title: 'Context/ContextSystemProvider'
};
var innerContext = {
  Card: {
    css: {
      background: 'white'
    }
  }
};
var InnerContent = /*#__PURE__*/React.memo(function () {
  var state = useSomeContext();
  var style = state % 2 === 0 ? {
    background: 'red'
  } : null;
  return /*#__PURE__*/React.createElement(View, {
    css: style
  }, state);
});
var InnerCard = /*#__PURE__*/React.memo(function () {
  return /*#__PURE__*/React.createElement(View, {
    css: {
      padding: 40
    }
  }, /*#__PURE__*/React.createElement(Card, null, /*#__PURE__*/React.createElement(CardBody, {
    css: [{
      border: '3px solid green'
    }]
  }, /*#__PURE__*/React.createElement(InnerContent, null))));
});
export var Default = function () {
  var _React$useState = React.useState(0),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      state = _React$useState2[0],
      update = _React$useState2[1];

  var value = {
    Card: {
      css: {
        background: state % 2 === 0 ? 'blue' : 'yellow'
      }
    },
    CardBody: {
      css: ["\n\t\t\t\tpadding: 20px;\n\t\t\t\t*:hover > & {\n\t\t\t\t\tpadding: 40px;\n\t\t\t\t}\n\t\t\t", ui.animation.default]
    }
  };
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(SomeContext.Provider, {
    value: state
  }, /*#__PURE__*/React.createElement("button", {
    onClick: function forceUpdate() {
      return update(function (prev) {
        return prev + 1;
      });
    }
  }, "Force Update"), /*#__PURE__*/React.createElement(ContextSystemProvider, {
    value: value
  }, /*#__PURE__*/React.createElement(Card, null, /*#__PURE__*/React.createElement(CardBody, {
    css: [{
      border: '3px solid green'
    }]
  }, /*#__PURE__*/React.createElement(Text, {
    optimizeReadabilityFor: "blue"
  }, "Card"), /*#__PURE__*/React.createElement(ContextSystemProvider, {
    value: innerContext
  }, /*#__PURE__*/React.createElement(InnerCard, null))))), /*#__PURE__*/React.createElement(Card, null, /*#__PURE__*/React.createElement(CardBody, null, /*#__PURE__*/React.createElement(Text, null, "Card")))));
};
