// @wp-g2/components/dist/es/Text/__stories__/Text.stories.js
import { FiScissors } from '@wp-g2/icons';
import React from 'react';
import { Icon, Link } from '../../index';
import { Text } from '../index';
export default {
  component: Text,
  title: 'Components/Text'
};
export var _default = function () {
  return /*#__PURE__*/React.createElement(Text, null, "Hello");
};
export var truncate = function () {
  return /*#__PURE__*/React.createElement(Text, {
    numberOfLines: 2,
    truncate: true
  }, "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut facilisis dictum tortor, eu tincidunt justo scelerisque tincidunt. Duis semper dui id augue malesuada, ut feugiat nisi aliquam. Vestibulum venenatis diam sem, finibus dictum massa semper in. Nulla facilisi. Nunc vulputate faucibus diam, in lobortis arcu ornare vel. In dignissim nunc sed facilisis finibus. Etiam imperdiet mattis arcu, sed rutrum sapien blandit gravida. Aenean sollicitudin neque eget enim blandit, sit amet rutrum leo vehicula. Nunc malesuada ultricies eros ut faucibus. Aliquam erat volutpat. Nulla nec feugiat risus. Vivamus iaculis dui aliquet ante ultricies feugiat. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia curae; Vivamus nec pretium velit, sit amet consectetur ante. Praesent porttitor ex eget fermentum mattis.");
};
export var highlight = function () {
  return /*#__PURE__*/React.createElement(Text, {
    highlightWords: ['con']
  }, "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut facilisis dictum tortor, eu tincidunt justo scelerisque tincidunt. Duis semper dui id augue malesuada, ut feugiat nisi aliquam. Vestibulum venenatis diam sem, finibus dictum massa semper in. Nulla facilisi. Nunc vulputate faucibus diam, in lobortis arcu ornare vel. In dignissim nunc sed facilisis finibus. Etiam imperdiet mattis arcu, sed rutrum sapien blandit gravida. Aenean sollicitudin neque eget enim blandit, sit amet rutrum leo vehicula. Nunc malesuada ultricies eros ut faucibus. Aliquam erat volutpat. Nulla nec feugiat risus. Vivamus iaculis dui aliquet ante ultricies feugiat. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia curae; Vivamus nec pretium velit, sit amet consectetur ante. Praesent porttitor ex eget fermentum mattis.");
};
export var textLinkIcon = function () {
  return /*#__PURE__*/React.createElement(Text, null, "Lorem ipsum dolor sit amet, consectetur adipiscing elit.", ' ', /*#__PURE__*/React.createElement(Link, null, "Link"), /*#__PURE__*/React.createElement(Icon, {
    icon: /*#__PURE__*/React.createElement(FiScissors, null),
    inline: true
  }), ".", ' ', /*#__PURE__*/React.createElement(Link, null, "Link ", /*#__PURE__*/React.createElement(Icon, {
    icon: /*#__PURE__*/React.createElement(FiScissors, null),
    inline: true
  }), ' '));
};
