// @wp-g2/components/dist/cjs/UnitInput/__stories__/UnitInput.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._everything = exports._default = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _utils = require("@wp-g2/utils");

var _react = _interopRequireDefault(require("react"));

var _zustand = _interopRequireDefault(require("zustand"));

var _index = require("../../index");

var _index2 = require("../index");

var _utils2 = require("./utils");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var _default2 = {
  component: _index2.UnitInput,
  title: 'Components/UnitInput'
};
exports.default = _default2;

var everythingInitialState = _utils2.validStyleProps.filter(Boolean).reduce(function (state, key) {
  var _objectSpread2;

  var initialValue = 1;

  if ((0, _utils.isValidCSSValueForProp)(key, '1px')) {
    initialValue = '1px';
  }

  return _objectSpread(_objectSpread({}, state), {}, (_objectSpread2 = {}, _objectSpread2[key] = initialValue, _objectSpread2));
}, {});

var valueStore = (0, _zustand.default)(function (set) {
  return _objectSpread(_objectSpread({}, everythingInitialState), {}, {
    setState: set
  });
});
var useValueStore = valueStore;

var CSSEntry = /*#__PURE__*/_react.default.memo(function (_ref) {
  var prop = _ref.prop;
  var value = useValueStore(function (state) {
    return state[prop];
  });
  return /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    horizontal: true,
    label: prop
  }, /*#__PURE__*/_react.default.createElement(_index2.UnitInput, {
    cssProp: prop,
    onChange: function update(next) {
      var _valueStore$setState;

      valueStore.setState((_valueStore$setState = {}, _valueStore$setState[prop] = next, _valueStore$setState));
    },
    validate: function validate(next) {
      return (0, _utils.isValidCSSValueForProp)(prop, next);
    },
    value: value
  }));
});

var Everything = function () {
  var _useValueStore = useValueStore(),
      setState = _useValueStore.setState,
      state = (0, _objectWithoutProperties2.default)(_useValueStore, ["setState"]);

  return /*#__PURE__*/_react.default.createElement(_index.ListGroup, {
    css: "\n\t\t\t\tpadding-bottom: 20vh;\n\t\t\t"
  }, Object.entries(state).map(function (_ref2) {
    var _ref3 = (0, _slicedToArray2.default)(_ref2, 1),
        key = _ref3[0];

    return /*#__PURE__*/_react.default.createElement(CSSEntry, {
      key: key,
      prop: key
    });
  }));
};

var Example = function () {
  var _React$useState = _react.default.useState('auto'),
      _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
      value = _React$useState2[0],
      setValue = _React$useState2[1];

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_index2.UnitInput, {
    onChange: setValue,
    value: value
  })), /*#__PURE__*/_react.default.createElement("br", null), /*#__PURE__*/_react.default.createElement("br", null), /*#__PURE__*/_react.default.createElement("br", null), /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_index2.UnitInput, {
    arrows: true,
    cssProp: "height",
    onChange: setValue,
    value: value
  })));
};

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(Example, null);
};

exports._default = _default;

var _everything = function () {
  return /*#__PURE__*/_react.default.createElement(Everything, null);
};

exports._everything = _everything;
