// @wp-g2/components/dist/cjs/ButtonGroup/ButtonGroup.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _i18n = require("@wordpress/i18n");

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _utils = require("@wp-g2/utils");

var _react = _interopRequireDefault(require("react"));

var _reakit = require("reakit");

var _ControlGroup = require("../ControlGroup");

var _ButtonGroup = require("./ButtonGroup.Context");

var styles = _interopRequireWildcard(require("./ButtonGroup.styles"));

/**
 * @param {import('@wp-g2/create-styles').ViewOwnProps<import('./types').Props, 'div'>} props
 * @param {import('react').Ref<any>} forwardedRef
 */
function ButtonGroup(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'ButtonGroup'),
      baseId = _useContextSystem.baseId,
      className = _useContextSystem.className,
      children = _useContextSystem.children,
      _useContextSystem$ena = _useContextSystem.enableSelectNone,
      enableSelectNone = _useContextSystem$ena === void 0 ? false : _useContextSystem$ena,
      _useContextSystem$exp = _useContextSystem.expanded,
      expanded = _useContextSystem$exp === void 0 ? false : _useContextSystem$exp,
      _useContextSystem$seg = _useContextSystem.segmented,
      segmented = _useContextSystem$seg === void 0 ? false : _useContextSystem$seg,
      id = _useContextSystem.id,
      _useContextSystem$lab = _useContextSystem.label,
      label = _useContextSystem$lab === void 0 ? (0, _i18n.__)('ButtonGroup') : _useContextSystem$lab,
      value = _useContextSystem.value,
      _useContextSystem$onC = _useContextSystem.onChange,
      onChange = _useContextSystem$onC === void 0 ? _utils.noop : _useContextSystem$onC,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["baseId", "className", "children", "enableSelectNone", "expanded", "segmented", "id", "label", "value", "onChange"]);

  var radio = (0, _reakit.useRadioState)({
    baseId: baseId || id,
    state: value
  });
  (0, _utils.useUpdateEffect)(function () {
    onChange(radio.state);
  }, [radio.state]);
  (0, _utils.useUpdateEffect)(function () {
    radio.setState(value);
  }, [value]);

  var contextProps = _react.default.useMemo(function () {
    return {
      buttonGroup: radio,
      enableSelectNone: enableSelectNone
    };
  }, [enableSelectNone, radio]);

  var contextSystemProps = _react.default.useMemo(function () {
    return {
      Button: {
        isBlock: expanded,
        isSubtle: !segmented,
        isControl: true
      },
      ControlGroup: {
        isItemBlock: expanded
      }
    };
  }, [expanded, segmented]);

  var classes = (0, _styles.cx)(segmented && styles.segmented, expanded && styles.expanded, className);
  var BaseComponent = segmented ? _ControlGroup.ControlGroup : styles.ButtonGroupView;
  return /*#__PURE__*/_react.default.createElement(_ButtonGroup.ButtonGroupContext.Provider, {
    value: contextProps
  }, /*#__PURE__*/_react.default.createElement(_context.ContextSystemProvider, {
    value: contextSystemProps
  }, /*#__PURE__*/_react.default.createElement(_reakit.RadioGroup, Object.assign({
    "aria-label": label,
    as: BaseComponent
  }, otherProps, {
    className: classes,
    ref: forwardedRef
  }), children)));
}
/**
 * `ButtonGroup` is a form component contains and coordinates the checked state of multiple `Button` components.
 *
 * @example
 * ```jsx
 * import { ButtonGroup, Button } from `@wp-g2/components`;
 *
 * function Example() {
 *  return (
 *    <ButtonGroup value="ana">
 *      <Button value="ana">Ana</Button>
 *      <Button value="elsa">Elsa</Button>
 *      <Button value="kristoff">Kristoff</Button>
 *      <Button value="olaf">Olaf</Button>
 *    </ButtonGroup>
 *  );
 * }
 * ```
 */


var ConnectedButtonGroup = (0, _context.contextConnect)(ButtonGroup, 'ButtonGroup');
var _default = ConnectedButtonGroup;
exports.default = _default;
