// @wp-g2/components/dist/cjs/Elevation/useElevation.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useElevation = useElevation;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _utils = require("@wp-g2/utils");

var _react = require("react");

var styles = _interopRequireWildcard(require("./Elevation.styles"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t\t\t*:focus > & {\n\t\t\t\tbox-shadow: ", ";\n\t\t\t}\n\t\t"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t\t\t*:active > & {\n\t\t\t\tbox-shadow: ", ";\n\t\t\t}\n\t\t"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t\t\t*:hover > & {\n\t\t\t\tbox-shadow: ", ";\n\t\t\t}\n\t\t"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

/**
 * @param {import('@wp-g2/create-styles').ViewOwnProps<import('./types').Props, 'div'>} props
 */
function useElevation(props) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'Elevation'),
      active = _useContextSystem.active,
      _useContextSystem$bor = _useContextSystem.borderRadius,
      borderRadius = _useContextSystem$bor === void 0 ? 'inherit' : _useContextSystem$bor,
      className = _useContextSystem.className,
      focus = _useContextSystem.focus,
      hover = _useContextSystem.hover,
      _useContextSystem$isI = _useContextSystem.isInteractive,
      isInteractive = _useContextSystem$isI === void 0 ? false : _useContextSystem$isI,
      _useContextSystem$off = _useContextSystem.offset,
      offset = _useContextSystem$off === void 0 ? 0 : _useContextSystem$off,
      _useContextSystem$val = _useContextSystem.value,
      value = _useContextSystem$val === void 0 ? 0 : _useContextSystem$val,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["active", "borderRadius", "className", "focus", "hover", "isInteractive", "offset", "value"]);

  var classes = (0, _react.useMemo)(function () {
    var hoverValue = _utils.is.defined(hover) ? hover : value * 2;
    var activeValue = _utils.is.defined(active) ? hover : value / 2;

    if (!isInteractive) {
      hoverValue = _utils.is.defined(hover) ? hover : undefined;
      activeValue = _utils.is.defined(active) ? active : undefined;
    }

    var transition = "box-shadow " + _styles.ui.get('transitionDuration') + " " + _styles.ui.get('transitionTimingFunction');

    var sx = {};
    sx.Base = (0, _styles.css)({
      borderRadius: borderRadius,
      bottom: offset,
      boxShadow: (0, _styles.getBoxShadow)(value),
      opacity: _styles.ui.get('elevationIntensity'),
      left: offset,
      right: offset,
      top: offset,
      transition: transition
    });
    sx.hover = (0, _styles.css)(_templateObject(), (0, _styles.getBoxShadow)(hoverValue));
    sx.active = (0, _styles.css)(_templateObject2(), (0, _styles.getBoxShadow)(activeValue));
    sx.focus = (0, _styles.css)(_templateObject3(), (0, _styles.getBoxShadow)(focus));
    return (0, _styles.cx)(styles.Elevation, sx.Elevation, sx.Base, _utils.is.defined(hoverValue) && sx.hover, _utils.is.defined(activeValue) && sx.active, _utils.is.defined(focus) && sx.focus, className);
  }, [active, borderRadius, className, focus, hover, isInteractive, offset, value]);
  return _objectSpread(_objectSpread({}, otherProps), {}, {
    className: classes
  });
}
