// @wordpress/components/build/ui/button/component.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _lodash = require("lodash");

var _baseButton = require("../base-button");

var _buttonGroup = require("../button-group");

var _visuallyHidden = require("../visually-hidden");

var styles = _interopRequireWildcard(require("./styles"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

/**
 * @typedef {'primary' | 'secondary' | 'tertiary' | 'plain' | 'link'} ButtonVariant
 */

/**
 * @typedef OwnProps
 * @property {ButtonVariant} [variant='secondary'] Determinds the `Button` variant to render.
 * @property {string} [describedBy] Text for element pointed to by aria-describedby (id is auto-generated if aria-describedby is not provided).
 */

/**
 * @typedef {import('../base-button/types').Props & OwnProps} Props
 */

/**
 * @param {import('@wp-g2/create-styles').ViewOwnProps<Props, 'button'>} props
 * @param {import('react').Ref<any>} forwardedRef
 */
function Button(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'Button'),
      children = _useContextSystem.children,
      className = _useContextSystem.className,
      currentColor = _useContextSystem.currentColor,
      icon = _useContextSystem.icon,
      _useContextSystem$isA = _useContextSystem.isActive,
      isActiveProp = _useContextSystem$isA === void 0 ? false : _useContextSystem$isA,
      _useContextSystem$isC = _useContextSystem.isControl,
      isControl = _useContextSystem$isC === void 0 ? false : _useContextSystem$isC,
      _useContextSystem$isS = _useContextSystem.isSubtle,
      isSubtle = _useContextSystem$isS === void 0 ? false : _useContextSystem$isS,
      _useContextSystem$onC = _useContextSystem.onClick,
      onClick = _useContextSystem$onC === void 0 ? _lodash.noop : _useContextSystem$onC,
      _useContextSystem$siz = _useContextSystem.size,
      size = _useContextSystem$siz === void 0 ? 'medium' : _useContextSystem$siz,
      _useContextSystem$typ = _useContextSystem.type,
      type = _useContextSystem$typ === void 0 ? 'button' : _useContextSystem$typ,
      _useContextSystem$var = _useContextSystem.variant,
      variant = _useContextSystem$var === void 0 ? 'secondary' : _useContextSystem$var,
      describedBy = _useContextSystem.describedBy,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["children", "className", "currentColor", "icon", "isActive", "isControl", "isSubtle", "onClick", "size", "type", "variant", "describedBy"]);

  var _useButtonGroupContex = (0, _buttonGroup.useButtonGroupContext)(),
      buttonGroup = _useButtonGroupContex.buttonGroup,
      enableSelectNone = _useButtonGroupContex.enableSelectNone;

  var isWithinButtonGroup = !!buttonGroup;
  var isButtonGroupActive = isWithinButtonGroup && (buttonGroup === null || buttonGroup === void 0 ? void 0 : buttonGroup.state) === otherProps.value;
  var isActive = isActiveProp || isButtonGroupActive;
  var isIconOnly = !!icon && !children;
  var descriptionId = describedBy ? (0, _lodash.uniqueId)() : undefined;
  var describedById = otherProps['aria-describedby'] || descriptionId;
  var handleOnClickWithinButtonGroup = (0, _element.useCallback)(function (event) {
    if (isWithinButtonGroup && enableSelectNone && isButtonGroupActive && buttonGroup) {
      event.preventDefault();
      event.stopPropagation();
      buttonGroup.setState(undefined);
    }
  }, [buttonGroup, enableSelectNone, isButtonGroupActive, isWithinButtonGroup]);
  var handleOnClick = (0, _element.useCallback)(function (event) {
    onClick(event);
    handleOnClickWithinButtonGroup(event);
  }, [handleOnClickWithinButtonGroup, onClick]);
  var classes = (0, _styles.cx)(styles.Button, styles[variant], styles[size], isControl && styles.control, isSubtle && styles.subtle, isSubtle && isControl && styles.subtleControl, isButtonGroupActive && styles.subtleControlActive, isIconOnly && styles.icon, currentColor && styles.currentColor, className);
  return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_baseButton.BaseButton, (0, _extends2.default)({
    className: classes,
    icon: icon,
    isActive: isActive,
    onClick: handleOnClick,
    ref: forwardedRef,
    "aria-describedby": describedById,
    type: type
  }, otherProps), children), describedBy && (0, _element.createElement)(_visuallyHidden.VisuallyHidden, null, (0, _element.createElement)("span", {
    id: descriptionId
  }, describedBy)));
}
/**
 * `Button` is a component used to trigger an action or event, such as submitting a Form, opening a Dialog, canceling an action, or performing a delete operation.
 *
 * @example
 * ```jsx
 * import { Button } from `@wordpress/components/ui`;
 *
 * function Example() {
 *   return <Button variant="secondary">Code is Poetry</Button>;
 * }
 * ```
 */


var ConnectedButton = (0, _context.contextConnect)(Button, 'Button');
var _default = ConnectedButton;
exports.default = _default;
