// @wp-g2/components/dist/cjs/ControlGroup/__stories__/ControlGroup.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _index = require("../../index");

var _index2 = require("../index");

var _default2 = {
  component: _index2.ControlGroup,
  title: 'Components/ControlGroup'
};
exports.default = _default2;

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(_index.Container, null, /*#__PURE__*/_react.default.createElement(_index.ListGroups, {
    spacing: 20
  }, /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(_index.ListGroupHeader, null, "Grid"), /*#__PURE__*/_react.default.createElement(_index2.ControlGroup, {
    templateColumns: "auto 1fr 1fr auto"
  }, /*#__PURE__*/_react.default.createElement(_index.Select, {
    options: [{
      label: 'Ms.',
      value: 'ms'
    }]
  }), /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    placeholder: "First name"
  }), /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    placeholder: "Last name"
  }), /*#__PURE__*/_react.default.createElement(_index.Button, {
    variant: "primary"
  }, "Submit"))), /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(_index.ListGroupHeader, null, "Flex"), /*#__PURE__*/_react.default.createElement(_index2.ControlGroup, null, /*#__PURE__*/_react.default.createElement(_index.Select, {
    options: [{
      label: 'Ms.',
      value: 'ms'
    }]
  }), /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    placeholder: "First name"
  }), /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    placeholder: "Last name"
  }), /*#__PURE__*/_react.default.createElement(_index.Button, {
    variant: "primary"
  }, "Submit")))));
};

exports._default = _default;
