// @wp-g2/components/dist/cjs/__stories__/WIP/ContextPopover.stories.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _icons = require("@wp-g2/icons");

var _protokit = require("@wp-g2/protokit");

var _styles = require("@wp-g2/styles");

var _utils = require("@wp-g2/utils");

var _react = _interopRequireWildcard(require("react"));

var _index = require("../../index");

var _default2 = {
  title: 'Examples/WIP/ContextPopover'
};
exports.default = _default2;
var colorSchema = new _protokit.Schema(function () {
  return {
    color: _protokit.faker.internet.color()
  };
});
var baseColors = colorSchema.make(34);
var themeColors = colorSchema.make(50);
var customColors = colorSchema.make(8);

var TabbedColorControls = function (_ref) {
  var _ref$colors = _ref.colors,
      colors = _ref$colors === void 0 ? baseColors : _ref$colors;
  return /*#__PURE__*/_react.default.createElement(_index.Tabs, null, /*#__PURE__*/_react.default.createElement(_index.TabList, null, /*#__PURE__*/_react.default.createElement(_index.Tab, {
    size: "xSmall"
  }, "Solid"), /*#__PURE__*/_react.default.createElement(_index.Tab, {
    size: "xSmall"
  }, "Gradient")), /*#__PURE__*/_react.default.createElement(_index.TabPanel, null, /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    pt: 3
  }, /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: 7,
    gap: 2
  }, colors.map(function (color, index) {
    return /*#__PURE__*/_react.default.createElement(_index.ColorCircle, Object.assign({}, color, {
      color: "hsl(" + (360 - index * 10) + ", 100%, 50%)"
    }));
  })))), /*#__PURE__*/_react.default.createElement(_index.TabPanel, null, /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    pt: 3
  }, /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: 7,
    gap: 2
  }, colors.map(function (color, index) {
    return /*#__PURE__*/_react.default.createElement(_index.ColorCircle, Object.assign({}, color, {
      color: "hsl(" + index * 10 + ", 100%, 50%)"
    }));
  })))));
};

var ColorControls = function (_ref2) {
  var limitColors = _ref2.limitColors,
      showTabs = _ref2.showTabs;

  var _useState = (0, _react.useState)(false),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      popoverState = _useState2[0],
      setPopoverState = _useState2[1];

  var _chunk = (0, _utils.chunk)(baseColors, limitColors ? 100 : 14),
      _chunk2 = (0, _slicedToArray2.default)(_chunk, 1),
      baseColorsData = _chunk2[0];

  var _chunk3 = (0, _utils.chunk)(themeColors, limitColors ? 100 : 8),
      _chunk4 = (0, _slicedToArray2.default)(_chunk3, 1),
      themeColorsData = _chunk4[0];

  var _chunk5 = (0, _utils.chunk)(customColors, limitColors ? 100 : 2),
      _chunk6 = (0, _slicedToArray2.default)(_chunk5, 1),
      customColorsData = _chunk6[0];

  return /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    m: 0,
    px: 1,
    py: 2
  }, /*#__PURE__*/_react.default.createElement(_index.ListGroups, null, /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(_index.ListGroupHeader, null, "Core", !showTabs && /*#__PURE__*/_react.default.createElement(_index.Select, {
    isSubtle: true,
    size: "small"
  }, /*#__PURE__*/_react.default.createElement("option", null, "Solid"), /*#__PURE__*/_react.default.createElement("option", null, "Gradient"))), showTabs ? /*#__PURE__*/_react.default.createElement(TabbedColorControls, {
    colors: baseColorsData
  }) : /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: 7,
    gap: 2
  }, baseColorsData.map(function (color, index) {
    return /*#__PURE__*/_react.default.createElement(_index.ColorCircle, Object.assign({}, color, {
      color: "hsl(" + (360 - index * 10) + ", 100%, 50%)"
    }));
  }))), /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(_index.ListGroupHeader, null, "Theme", !showTabs && /*#__PURE__*/_react.default.createElement(_index.Select, {
    isSubtle: true,
    size: "small"
  }, /*#__PURE__*/_react.default.createElement("option", null, "Solid"), /*#__PURE__*/_react.default.createElement("option", null, "Gradient"))), /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: 7,
    gap: 2
  }, themeColorsData.map(function (color) {
    return /*#__PURE__*/_react.default.createElement(_index.ColorCircle, color);
  }))), /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(_index.ListGroupHeader, null, "Custom", !showTabs && /*#__PURE__*/_react.default.createElement(_index.Select, {
    isSubtle: true,
    size: "small"
  }, /*#__PURE__*/_react.default.createElement("option", null, "Solid"), /*#__PURE__*/_react.default.createElement("option", null, "Gradient"))), /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: 7,
    gap: 2
  }, customColorsData.map(function (color) {
    return /*#__PURE__*/_react.default.createElement(_index.ColorCircle, color);
  }), /*#__PURE__*/_react.default.createElement(_index.View, null, /*#__PURE__*/_react.default.createElement(_index.Popover, {
    css: [{
      marginLeft: 39
    }],
    maxWidth: "262px",
    onVisibleChange: setPopoverState,
    placement: "top",
    trigger: /*#__PURE__*/_react.default.createElement(_index.Button, {
      icon: /*#__PURE__*/_react.default.createElement(_icons.FiPlus, null),
      isControl: true,
      isRounded: true,
      isSubtle: true,
      size: "xSmall"
    }),
    visible: popoverState
  }, /*#__PURE__*/_react.default.createElement(_index.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.ColorPicker, null))))))));
};

var ColorSetting = function (_ref3) {
  var color = _ref3.color,
      limitColors = _ref3.limitColors,
      onVisibleChange = _ref3.onVisibleChange,
      showElevation = _ref3.showElevation,
      showTabs = _ref3.showTabs,
      title = _ref3.title;

  var _useState3 = (0, _react.useState)(false),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      visible = _useState4[0],
      setVisible = _useState4[1];

  return /*#__PURE__*/_react.default.createElement(_index.Collapsible, {
    css: [_styles.ui.position.relative, _styles.ui.zIndex(visible ? 10 : 0)],
    onVisibleChange: function handleOnVisibleChange(next) {
      setVisible(next);
      onVisibleChange(next);
    },
    visible: visible
  }, /*#__PURE__*/_react.default.createElement(_index.Elevation, {
    offset: -8,
    value: showElevation && visible ? 4 : 0
  }), /*#__PURE__*/_react.default.createElement(_index.HStack, {
    css: [_styles.ui.hover(_styles.ui.$('ColorAction').css({
      opacity: 1
    }))]
  }, /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.CollapsibleTrigger, {
    as: _index.ColorControl,
    color: color
  }, title)), /*#__PURE__*/_react.default.createElement(_index.View, null, /*#__PURE__*/_react.default.createElement(_index.Button, Object.assign({}, _styles.ui.$('ColorAction'), {
    className: "action",
    css: "\n\t\t\t\t\t\t\topacity: 0;\n\t\t\t\t\t\t\t&:focus {\n\t\t\t\t\t\t\t\topacity: 1;\n\t\t\t\t\t\t\t}\n\t\t\t\t\t\t",
    icon: /*#__PURE__*/_react.default.createElement(_icons.FiMinus, null),
    isRounded: true,
    isSubtle: true,
    size: "small"
  })))), /*#__PURE__*/_react.default.createElement(_index.CollapsibleContent, null, /*#__PURE__*/_react.default.createElement(ColorControls, {
    limitColors: limitColors,
    showTabs: showTabs
  })));
};

var Example = function () {
  var _useState5 = (0, _react.useState)(false),
      _useState6 = (0, _slicedToArray2.default)(_useState5, 2),
      visible = _useState6[0],
      setVisible = _useState6[1];

  var _useState7 = (0, _react.useState)(false),
      _useState8 = (0, _slicedToArray2.default)(_useState7, 2),
      showElevation = _useState8[0],
      setShowElevation = _useState8[1];

  var _useState9 = (0, _react.useState)(false),
      _useState10 = (0, _slicedToArray2.default)(_useState9, 2),
      showTabs = _useState10[0],
      setShowTabs = _useState10[1];

  var _useState11 = (0, _react.useState)(false),
      _useState12 = (0, _slicedToArray2.default)(_useState11, 2),
      limitColors = _useState12[0],
      setLimitColors = _useState12[1];

  return /*#__PURE__*/_react.default.createElement(_index.Grid, null, /*#__PURE__*/_react.default.createElement(_index.View, null, /*#__PURE__*/_react.default.createElement(_index.Container, {
    width: 270
  }, /*#__PURE__*/_react.default.createElement(_index.Card, null, /*#__PURE__*/_react.default.createElement(_index.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(_index.ListGroupHeader, null, "Settings"), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Elevation"
  }, /*#__PURE__*/_react.default.createElement(_index.Switch, {
    checked: showElevation,
    onChange: setShowElevation
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Use Tabs"
  }, /*#__PURE__*/_react.default.createElement(_index.Switch, {
    checked: showTabs,
    onChange: setShowTabs
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "COLORS!"
  }, /*#__PURE__*/_react.default.createElement(_index.Switch, {
    checked: limitColors,
    onChange: setLimitColors
  }))))))), /*#__PURE__*/_react.default.createElement(_index.Container, {
    css: [_styles.ui.position.relative, {
      minHeight: '80vh'
    }],
    width: 270
  }, /*#__PURE__*/_react.default.createElement(_index.Surface, {
    css: [_styles.ui.opacity(visible ? 0.8 : 0), {
      pointerEvents: visible ? null : 'none'
    }, _styles.ui.position.full, _styles.ui.zIndex(1), {
      minHeight: '80vh'
    }]
  }), /*#__PURE__*/_react.default.createElement(_index.Card, {
    css: [{
      minHeight: '80vh'
    }],
    isRounded: false
  }, /*#__PURE__*/_react.default.createElement(_index.Panel, {
    visible: true
  }, /*#__PURE__*/_react.default.createElement(_index.PanelHeader, {
    title: "Color"
  }), /*#__PURE__*/_react.default.createElement(_index.PanelBody, null, /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    pb: 3
  }, /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(ColorSetting, {
    color: "white",
    limitColors: limitColors,
    onVisibleChange: setVisible,
    showElevation: showElevation,
    showTabs: showTabs,
    title: "Background"
  }), /*#__PURE__*/_react.default.createElement(ColorSetting, {
    color: "black",
    limitColors: limitColors,
    onVisibleChange: setVisible,
    showElevation: showElevation,
    showTabs: showTabs,
    title: "Text"
  }), /*#__PURE__*/_react.default.createElement(ColorSetting, {
    color: "pink",
    limitColors: limitColors,
    onVisibleChange: setVisible,
    showElevation: showElevation,
    showTabs: showTabs,
    title: "Accent"
  }))))))));
};

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(Example, null);
};

exports._default = _default;
