// @wp-g2/components/dist/cjs/__stories__/WIP/ContextRendering.stories.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireWildcard(require("react"));

var _reactFrameComponent = _interopRequireDefault(require("react-frame-component"));

var _index = require("../../index");

var _default2 = {
  title: 'Examples/WIP/ContextRendering'
};
exports.default = _default2;

var Cluster = function () {
  return /*#__PURE__*/_react.default.createElement(_index.View, null, /*#__PURE__*/_react.default.createElement(_index.Card, null, /*#__PURE__*/_react.default.createElement(_index.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "First Name"
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    value: "Elsa"
  })))));
};

var Section = function (_ref) {
  var children = _ref.children,
      title = _ref.title,
      _ref$works = _ref.works,
      works = _ref$works === void 0 ? true : _ref$works;
  return /*#__PURE__*/_react.default.createElement(_index.Surface, {
    border: true,
    css: {
      padding: 20,
      paddingTop: 12
    }
  }, /*#__PURE__*/_react.default.createElement(_index.VStack, {
    alignment: "left"
  }, /*#__PURE__*/_react.default.createElement(_index.HStack, null, /*#__PURE__*/_react.default.createElement(_index.Subheading, null, title), /*#__PURE__*/_react.default.createElement(_index.Text, null, works ? '✅' : '❌')), children));
};

var Example = function () {
  var _useState = (0, _react.useState)(false),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      now = _useState2[0],
      setNow = _useState2[1];

  (0, _react.useEffect)(function () {
    setTimeout(function () {
      setNow(true);
    }, 3000);
  }, []);
  return /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: 3
  }, /*#__PURE__*/_react.default.createElement(Section, {
    title: "Default"
  }, /*#__PURE__*/_react.default.createElement(Cluster, null), /*#__PURE__*/_react.default.createElement(_index.Alerts, null, now && /*#__PURE__*/_react.default.createElement(_index.Alert, {
    status: "success"
  }, /*#__PURE__*/_react.default.createElement(_index.Text, null, "New Element")))), /*#__PURE__*/_react.default.createElement(Section, {
    title: "Dark"
  }, /*#__PURE__*/_react.default.createElement(_styles.ThemeProvider, {
    isDark: true
  }, /*#__PURE__*/_react.default.createElement(_index.Surface, null, /*#__PURE__*/_react.default.createElement(Cluster, null), /*#__PURE__*/_react.default.createElement(_index.Alerts, null, now && /*#__PURE__*/_react.default.createElement(_index.Alert, {
    status: "success"
  }, /*#__PURE__*/_react.default.createElement(_index.Text, null, "New Element")))))), /*#__PURE__*/_react.default.createElement(Section, {
    title: "Custom Context"
  }, /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.Subheading, null, "Inside Context"), /*#__PURE__*/_react.default.createElement(_context.ContextSystemProvider, {
    value: {
      FormGroup: {
        horizontal: false
      }
    }
  }, /*#__PURE__*/_react.default.createElement(Cluster, null)), /*#__PURE__*/_react.default.createElement(_index.Subheading, null, "Outside Context"), /*#__PURE__*/_react.default.createElement(Cluster, null))), /*#__PURE__*/_react.default.createElement(Section, {
    title: "iFrame"
  }, /*#__PURE__*/_react.default.createElement(_reactFrameComponent.default, {
    style: {
      width: '100%'
    }
  }, /*#__PURE__*/_react.default.createElement(_styles.StyleFrameProvider, null, /*#__PURE__*/_react.default.createElement(Cluster, null), /*#__PURE__*/_react.default.createElement(_index.Alerts, null, now && /*#__PURE__*/_react.default.createElement(_index.Alert, {
    status: "success"
  }, /*#__PURE__*/_react.default.createElement(_index.Text, null, "New Element")))))), /*#__PURE__*/_react.default.createElement(Section, {
    title: "Dark + iFrame"
  }, /*#__PURE__*/_react.default.createElement(_reactFrameComponent.default, {
    style: {
      width: '100%'
    }
  }, /*#__PURE__*/_react.default.createElement(_styles.StyleFrameProvider, null, /*#__PURE__*/_react.default.createElement(_styles.ThemeProvider, {
    isDark: true
  }, /*#__PURE__*/_react.default.createElement(_index.Surface, null, /*#__PURE__*/_react.default.createElement(Cluster, null), /*#__PURE__*/_react.default.createElement(_index.Alerts, null, now && /*#__PURE__*/_react.default.createElement(_index.Alert, {
    status: "success"
  }, /*#__PURE__*/_react.default.createElement(_index.Text, null, "New Element")))))))), /*#__PURE__*/_react.default.createElement(Section, {
    title: "Custom Context + iFrame"
  }, /*#__PURE__*/_react.default.createElement(_reactFrameComponent.default, {
    style: {
      width: '100%'
    }
  }, /*#__PURE__*/_react.default.createElement(_styles.StyleFrameProvider, null, /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.Subheading, null, "Inside Context"), /*#__PURE__*/_react.default.createElement(_context.ContextSystemProvider, {
    value: {
      FormGroup: {
        horizontal: false
      }
    }
  }, /*#__PURE__*/_react.default.createElement(Cluster, null)), /*#__PURE__*/_react.default.createElement(_index.Subheading, null, "Outside Context"), /*#__PURE__*/_react.default.createElement(Cluster, null))))), /*#__PURE__*/_react.default.createElement(Section, {
    title: "Popover"
  }, /*#__PURE__*/_react.default.createElement(_index.Card, null, /*#__PURE__*/_react.default.createElement(_index.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.Popover, {
    hideOnClickOutside: false,
    placement: "bottom-start",
    trigger: /*#__PURE__*/_react.default.createElement(_index.Button, null, "Popover"),
    visible: false
  }, /*#__PURE__*/_react.default.createElement(Cluster, null))))), /*#__PURE__*/_react.default.createElement(Section, {
    title: "Popover + Dark"
  }, /*#__PURE__*/_react.default.createElement(_styles.ThemeProvider, {
    isDark: true
  }, /*#__PURE__*/_react.default.createElement(_index.Card, null, /*#__PURE__*/_react.default.createElement(_index.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.Popover, {
    hideOnClickOutside: false,
    placement: "bottom-start",
    trigger: /*#__PURE__*/_react.default.createElement(_index.Button, null, "Popover"),
    visible: false
  }, /*#__PURE__*/_react.default.createElement(Cluster, null)))))), /*#__PURE__*/_react.default.createElement(_index.View, null));
};

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(Example, null);
};

exports._default = _default;
