// @wp-g2/components/dist/cjs/__stories__/WIP/FormElementStates.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _react = _interopRequireDefault(require("react"));

var _index = require("../../index");

var _default2 = {
  title: 'Examples/WIP/FormElementStates'
};
exports.default = _default2;

var ControlCluster = function (_ref) {
  var heading = _ref.heading,
      inputValue = _ref.inputValue,
      selectValue = _ref.selectValue,
      setInputValue = _ref.setInputValue,
      setSelectValue = _ref.setSelectValue,
      setSwitchValue = _ref.setSwitchValue,
      switchValue = _ref.switchValue,
      otherProps = (0, _objectWithoutProperties2.default)(_ref, ["heading", "inputValue", "selectValue", "setInputValue", "setSelectValue", "setSwitchValue", "switchValue"]);
  return /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(_index.ListGroupHeader, null, heading), /*#__PURE__*/_react.default.createElement(_context.ContextSystemProvider, {
    value: {
      FormGroup: {
        horizontal: false
      }
    }
  }, /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: 8
  }, /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Unit Input"
  }, /*#__PURE__*/_react.default.createElement(_index.UnitInput, Object.assign({}, otherProps, {
    onChange: setInputValue,
    value: inputValue
  }))), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Text Input"
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, Object.assign({}, otherProps, {
    onChange: setInputValue,
    value: inputValue
  }))), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Text Input + Stepper"
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, Object.assign({}, otherProps, {
    arrows: "stepper",
    onChange: setInputValue,
    type: "number",
    value: inputValue
  }))), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Slider"
  }, /*#__PURE__*/_react.default.createElement(_index.Slider, Object.assign({}, otherProps, {
    onChange: setInputValue,
    value: inputValue
  }))), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Select"
  }, /*#__PURE__*/_react.default.createElement(_index.Select, Object.assign({}, otherProps, {
    onChange: setSelectValue,
    value: selectValue
  }), /*#__PURE__*/_react.default.createElement("option", {
    label: "One",
    value: "one"
  }), /*#__PURE__*/_react.default.createElement("option", {
    label: "Two",
    value: "two"
  }), /*#__PURE__*/_react.default.createElement("option", {
    label: "Three",
    value: "three"
  }))), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Button"
  }, /*#__PURE__*/_react.default.createElement(_index.Button, {
    disabled: otherProps.disabled,
    isBlock: true,
    isDestructive: otherProps.error,
    isFocused: otherProps.isFocused,
    variant: "primary"
  }, "Button")), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Switch"
  }, /*#__PURE__*/_react.default.createElement(_index.Switch, {
    checked: switchValue,
    disabled: otherProps.disabled,
    isDestructive: otherProps.error,
    isFocused: otherProps.isFocused,
    onChange: setSwitchValue
  })))));
};

var Example = function () {
  var _React$useState = _react.default.useState(1),
      _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
      inputValue = _React$useState2[0],
      setInputValue = _React$useState2[1];

  var _React$useState3 = _react.default.useState('one'),
      _React$useState4 = (0, _slicedToArray2.default)(_React$useState3, 2),
      selectValue = _React$useState4[0],
      setSelectValue = _React$useState4[1];

  var _React$useState5 = _react.default.useState(false),
      _React$useState6 = (0, _slicedToArray2.default)(_React$useState5, 2),
      switchValue = _React$useState6[0],
      setSwitchValue = _React$useState6[1];

  var clusterProps = {
    inputValue: inputValue,
    setInputValue: setInputValue,
    selectValue: selectValue,
    setSelectValue: setSelectValue,
    switchValue: switchValue,
    setSwitchValue: setSwitchValue
  };
  return /*#__PURE__*/_react.default.createElement(_index.Container, {
    css: "\n\t\t\t\tmargin-top: 10vh;\n\t\t\t\tmargin-bottom: 20vh;\n\t\t\t"
  }, /*#__PURE__*/_react.default.createElement(_index.ListGroups, {
    separator: true,
    spacing: 12
  }, /*#__PURE__*/_react.default.createElement(ControlCluster, Object.assign({}, clusterProps, {
    heading: "Default"
  })), /*#__PURE__*/_react.default.createElement(ControlCluster, Object.assign({}, clusterProps, {
    heading: "Focused",
    isFocused: true
  })), /*#__PURE__*/_react.default.createElement(ControlCluster, Object.assign({}, clusterProps, {
    disabled: true,
    heading: "Disabled"
  })), /*#__PURE__*/_react.default.createElement(ControlCluster, Object.assign({}, clusterProps, {
    error: true,
    heading: "Error"
  })), /*#__PURE__*/_react.default.createElement(ControlCluster, Object.assign({}, clusterProps, {
    error: true,
    heading: "Error + Focused",
    isFocused: true
  })), /*#__PURE__*/_react.default.createElement(ControlCluster, Object.assign({}, clusterProps, {
    disabled: true,
    error: true,
    heading: "Error + Disabled"
  }))));
};

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(Example, null);
};

exports._default = _default;
