// @wp-g2/components/dist/cjs/__stories__/WIP/FormElementsInPanel.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _index = require("../../index");

var _default2 = {
  title: 'Examples/WIP/FormElementInPanel'
};
exports.default = _default2;

var Example = function () {
  return /*#__PURE__*/_react.default.createElement(_index.Container, {
    css: "\n\t\t\t\tmargin-top: 10vh;\n\t\t\t\tmargin-bottom: 20vh;\n\t\t\t",
    width: 280
  }, /*#__PURE__*/_react.default.createElement(_index.Surface, {
    border: true
  }, /*#__PURE__*/_react.default.createElement(_index.Panel, {
    visible: true
  }, /*#__PURE__*/_react.default.createElement(_index.PanelHeader, null, "Status & visibility"), /*#__PURE__*/_react.default.createElement(_index.PanelBody, null, /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    horizontal: true,
    label: "Visibility"
  }, /*#__PURE__*/_react.default.createElement(_index.Button, null, "Public")), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    horizontal: true,
    label: "Publish"
  }, /*#__PURE__*/_react.default.createElement(_index.Button, null, "Immediately")), /*#__PURE__*/_react.default.createElement(_index.FormGroup, null, /*#__PURE__*/_react.default.createElement(_index.Checkbox, {
    label: "Stick to the top of the blog"
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, null, /*#__PURE__*/_react.default.createElement(_index.Checkbox, {
    label: "Pending review"
  })), /*#__PURE__*/_react.default.createElement(_index.Button, {
    isDestructive: true
  }, "Move to trash"))))));
};

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(Example, null);
};

exports._default = _default;
