// @wp-g2/components/dist/cjs/__stories__/WIP/GlobalStylesSidebar.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _context = require("@wp-g2/context");

var _icons = require("@wp-g2/icons");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _index = require("../../index");

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbottom: 0;\n\tleft: 0;\n\tright: 0;\n\ttop: 0;\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var _default2 = {
  title: 'Examples/WIP/GlobalStylesSidebar'
};
exports.default = _default2;
var Screen = (0, _styles.styled)(_index.Surface, {
  props: {
    variant: 'tertiary'
  }
})(_templateObject());

var SettingLink = function (_ref) {
  var prefix = _ref.prefix,
      title = _ref.title,
      to = _ref.to;
  return /*#__PURE__*/_react.default.createElement(_index.NavigatorLink, {
    to: to
  }, /*#__PURE__*/_react.default.createElement(_index.MenuItem, {
    prefix: prefix,
    showArrow: true
  }, /*#__PURE__*/_react.default.createElement(_index.Text, {
    weight: 600
  }, title)));
};

var GlobalStylesScreen = function () {
  return /*#__PURE__*/_react.default.createElement(Screen, {
    css: "\n\t\t\t\tpadding-top: 40px;\n\t\t\t"
  }, /*#__PURE__*/_react.default.createElement(_index.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(SettingLink, {
    prefix: /*#__PURE__*/_react.default.createElement(_index.Icon, {
      icon: /*#__PURE__*/_react.default.createElement(_icons.FiDroplet, null)
    }),
    title: "Colors",
    to: "Colors"
  }), /*#__PURE__*/_react.default.createElement(SettingLink, {
    prefix: /*#__PURE__*/_react.default.createElement(_index.Icon, {
      icon: /*#__PURE__*/_react.default.createElement(_icons.FiType, null)
    }),
    title: "Typography",
    to: "Typography"
  }), /*#__PURE__*/_react.default.createElement(SettingLink, {
    prefix: /*#__PURE__*/_react.default.createElement(_index.Icon, {
      icon: /*#__PURE__*/_react.default.createElement(_icons.FiGrid, null)
    }),
    title: "Spacing"
  }))), /*#__PURE__*/_react.default.createElement(_index.Divider, null), /*#__PURE__*/_react.default.createElement(_index.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(_index.ListGroupHeader, null, "Blocks"), /*#__PURE__*/_react.default.createElement(SettingLink, {
    title: "Button",
    to: "Buttons"
  }), /*#__PURE__*/_react.default.createElement(SettingLink, {
    title: "Forms"
  }), /*#__PURE__*/_react.default.createElement(SettingLink, {
    title: "Images"
  }))));
};

var ColorsScreen = function () {
  return /*#__PURE__*/_react.default.createElement(Screen, {
    css: "\n\t\t\t\tpadding-top: 60px;\n\t\t\t"
  }, /*#__PURE__*/_react.default.createElement(_index.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(_index.ListGroupHeader, null, "Defaults"), /*#__PURE__*/_react.default.createElement(_index.ColorControl, {
    color: "#fff"
  }, "Site Background"), /*#__PURE__*/_react.default.createElement(_index.ColorControl, {
    color: "#000"
  }, "Body Text"), /*#__PURE__*/_react.default.createElement(_index.ColorControl, {
    color: "#0f6"
  }, "Accent"))), /*#__PURE__*/_react.default.createElement(_index.Divider, null), /*#__PURE__*/_react.default.createElement(_index.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(_index.ListGroupHeader, null, "Theme"), /*#__PURE__*/_react.default.createElement(_index.ColorControl, {
    color: "#111"
  }, "Dark Gray"), /*#__PURE__*/_react.default.createElement(_index.ColorControl, {
    color: "#05f"
  }, "Blues"))), /*#__PURE__*/_react.default.createElement(_index.Divider, null), /*#__PURE__*/_react.default.createElement(_index.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(_index.ListGroupHeader, null, "Custom", /*#__PURE__*/_react.default.createElement(_index.Button, {
    icon: /*#__PURE__*/_react.default.createElement(_icons.FiPlus, null),
    isControl: true,
    isSubtle: true,
    size: "small"
  })), /*#__PURE__*/_react.default.createElement(_index.ColorControl, {
    color: "#f00"
  }, "Red"), /*#__PURE__*/_react.default.createElement(_index.ColorControl, {
    color: "#0f0"
  }, "Green"))));
};

var TypographyScreen = function () {
  var headings = [{
    label: /*#__PURE__*/_react.default.createElement(_index.Text, null, "H1"),
    value: 'h1'
  }, {
    label: /*#__PURE__*/_react.default.createElement(_index.Text, null, "H2"),
    value: 'h2'
  }, {
    label: /*#__PURE__*/_react.default.createElement(_index.Text, null, "H3"),
    value: 'h3'
  }, {
    label: /*#__PURE__*/_react.default.createElement(_index.Text, null, "H4"),
    value: 'h4'
  }, {
    label: /*#__PURE__*/_react.default.createElement(_index.Text, null, "H5"),
    value: 'h5'
  }, {
    label: /*#__PURE__*/_react.default.createElement(_index.Text, null, "H6"),
    value: 'h6'
  }];
  return /*#__PURE__*/_react.default.createElement(Screen, {
    css: "\n\t\t\t\tpadding-top: 60px;\n\t\t\t"
  }, /*#__PURE__*/_react.default.createElement(_index.Panel, {
    visible: true
  }, /*#__PURE__*/_react.default.createElement(_index.PanelHeader, {
    title: "Heading"
  }), /*#__PURE__*/_react.default.createElement(_index.PanelBody, null, /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    mb: 3
  }, /*#__PURE__*/_react.default.createElement(_index.SegmentedControl, {
    isBlock: true,
    options: headings
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Family"
  }, /*#__PURE__*/_react.default.createElement(_index.Select, {
    options: [{
      value: 'georgia',
      label: 'Georgia'
    }]
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Size"
  }, /*#__PURE__*/_react.default.createElement(_index.Grid, null, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    value: 32
  }), /*#__PURE__*/_react.default.createElement(_index.Slider, null))), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Weight"
  }, /*#__PURE__*/_react.default.createElement(_index.Select, {
    options: [{
      value: '500',
      label: '500'
    }]
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Transform"
  }, /*#__PURE__*/_react.default.createElement(_index.Select, {
    options: [{
      value: 'None',
      label: 'None'
    }]
  }))))), /*#__PURE__*/_react.default.createElement(_index.Panel, {
    visible: true
  }, /*#__PURE__*/_react.default.createElement(_index.PanelHeader, {
    title: "Base Text"
  }), /*#__PURE__*/_react.default.createElement(_index.PanelBody, null, /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Family"
  }, /*#__PURE__*/_react.default.createElement(_index.Select, {
    options: [{
      value: 'georgia',
      label: 'Georgia'
    }]
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Size"
  }, /*#__PURE__*/_react.default.createElement(_index.Grid, null, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    value: 32
  }), /*#__PURE__*/_react.default.createElement(_index.Slider, null))), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Weight"
  }, /*#__PURE__*/_react.default.createElement(_index.Select, {
    options: [{
      value: '500',
      label: '500'
    }]
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Transform"
  }, /*#__PURE__*/_react.default.createElement(_index.Select, {
    options: [{
      value: 'None',
      label: 'None'
    }]
  }))))));
};

var ButtonsScreen = function () {
  return /*#__PURE__*/_react.default.createElement(Screen, {
    css: "\n\t\t\t\tpadding-top: 60px;\n\t\t\t"
  }, /*#__PURE__*/_react.default.createElement(_index.Panel, {
    visible: true
  }, /*#__PURE__*/_react.default.createElement(_index.PanelHeader, {
    title: "Typography"
  }), /*#__PURE__*/_react.default.createElement(_index.PanelBody, null, /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Family"
  }, /*#__PURE__*/_react.default.createElement(_index.Select, {
    options: [{
      value: 'georgia',
      label: 'Georgia'
    }]
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Size"
  }, /*#__PURE__*/_react.default.createElement(_index.Grid, null, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    value: 32
  }), /*#__PURE__*/_react.default.createElement(_index.Slider, null))), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Weight"
  }, /*#__PURE__*/_react.default.createElement(_index.Select, {
    options: [{
      value: '500',
      label: '500'
    }]
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Transform"
  }, /*#__PURE__*/_react.default.createElement(_index.Select, {
    options: [{
      value: 'None',
      label: 'None'
    }]
  }))))), /*#__PURE__*/_react.default.createElement(_index.Panel, {
    visible: true
  }, /*#__PURE__*/_react.default.createElement(_index.PanelHeader, {
    title: "Color"
  }), /*#__PURE__*/_react.default.createElement(_index.PanelBody, null, /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(_index.ColorControl, {
    color: "#111"
  }, "Text"), /*#__PURE__*/_react.default.createElement(_index.ColorControl, {
    color: "#05f"
  }, "Background")))), /*#__PURE__*/_react.default.createElement(_index.Panel, null, /*#__PURE__*/_react.default.createElement(_index.PanelHeader, {
    title: "Behaviour"
  })), /*#__PURE__*/_react.default.createElement(_index.Panel, null, /*#__PURE__*/_react.default.createElement(_index.PanelHeader, {
    title: "Advanced"
  })));
};

var useNavigatorTitle = function (path) {
  var _useNavigatorLocation = (0, _index.useNavigatorLocation)(),
      pathname = _useNavigatorLocation.pathname;

  var screen = screens.find(function (screen) {
    return screen.path === (path || pathname);
  });
  return screen ? screen.title : '';
};

var usePreviousScreen = function () {
  var _useNavigatorHistory = (0, _index.useNavigatorHistory)(),
      entries = _useNavigatorHistory.entries;

  var _useNavigatorLocation2 = (0, _index.useNavigatorLocation)(),
      pathname = _useNavigatorLocation2.pathname;

  var current = entries.find(function (entry) {
    return entry.pathname === pathname;
  });
  var currentIndex = entries.indexOf(current);
  var previous = entries[currentIndex - 1];
  if (!previous) return;
  return screens.find(function (screen) {
    return screen.path === previous.pathname;
  });
};

var ANIMATION_SPEED = 0.14;

var GlobalStylesHeader = function () {
  var title = useNavigatorTitle();
  var previousScreen = usePreviousScreen();
  var hasPreviousScreen = !!previousScreen;
  return /*#__PURE__*/_react.default.createElement(_index.Surface, {
    borderBottom: true,
    css: [_styles.ui.padding(3), {
      position: 'absolute',
      top: 0,
      left: 0,
      right: 0,
      zIndex: 10
    }, {
      overflow: 'hidden'
    }, _styles.ui.frame.height(hasPreviousScreen ? 60 : 40), _styles.ui.animation.default, _styles.ui.animation.duration(ANIMATION_SPEED)],
    variant: "tertiary"
  }, /*#__PURE__*/_react.default.createElement(_index.View, {
    css: [{
      pointerEvents: hasPreviousScreen ? 'auto' : 'none'
    }, _styles.ui.opacity(hasPreviousScreen ? 1 : 0), _styles.ui.offset.x(hasPreviousScreen ? 0 : -20), _styles.ui.animation.default, _styles.ui.animation.delay(ANIMATION_SPEED * 0.5), _styles.ui.animation.duration(ANIMATION_SPEED)]
  }, previousScreen && /*#__PURE__*/_react.default.createElement(_index.NavigatorLink, {
    isBack: true
  }, /*#__PURE__*/_react.default.createElement(_index.HStack, {
    alignment: "left",
    gap: 1
  }, /*#__PURE__*/_react.default.createElement(_index.Icon, {
    icon: /*#__PURE__*/_react.default.createElement(_icons.FiChevronLeft, null),
    size: 12
  }), /*#__PURE__*/_react.default.createElement(_index.Text, {
    color: "currentColor",
    isBlock: true,
    lineHeight: 1,
    size: "caption"
  }, "Back to ", previousScreen.title)))), /*#__PURE__*/_react.default.createElement(_index.Heading, {
    css: [_styles.ui.position.bottom, _styles.ui.offset({
      y: -12
    })],
    size: 5
  }, title));
};

var screens = [{
  component: GlobalStylesScreen,
  path: 'GlobalStyles',
  title: 'Global Styles'
}, {
  component: ColorsScreen,
  path: 'Colors',
  title: 'Colors'
}, {
  component: TypographyScreen,
  path: 'Typography',
  title: 'Typography'
}, {
  component: ButtonsScreen,
  path: 'Buttons',
  title: 'Buttons'
}];

var Sidebar = function (_ref2) {
  var children = _ref2.children;
  return /*#__PURE__*/_react.default.createElement(_context.ContextSystemProvider, {
    value: {
      Grid: {
        gap: 2
      },
      Icon: {
        size: 16
      }
    }
  }, /*#__PURE__*/_react.default.createElement(_index.View, {
    css: "\n\t\t\t\t\twidth: 280px;\n\t\t\t\t\tposition: absolute;\n\t\t\t\t\theight: 100vh;\n\t\t\t\t\ttop: 0;\n\t\t\t\t\tright: 0;\n\t\t\t\t\tborder-left: 1px solid " + _styles.ui.get('colorDivider') + ";\n\t\t\t\t"
  }, children));
};

var Example = function () {
  // const initialPath = 'Typography';
  return /*#__PURE__*/_react.default.createElement(_index.Navigator, {
    initialPath: 'GlobalStyles'
  }, /*#__PURE__*/_react.default.createElement(GlobalStylesHeader, null), /*#__PURE__*/_react.default.createElement(_index.NavigatorScreens, {
    css: [_styles.ui.frame.height('auto')]
  }, screens.map(function (screen) {
    return /*#__PURE__*/_react.default.createElement(_index.NavigatorScreen, Object.assign({}, screen, {
      animationEnterDelay: 0,
      animationEnterDuration: ANIMATION_SPEED,
      animationExitDuration: ANIMATION_SPEED,
      key: screen.path
    }));
  })));
};

var _default = function () {
  return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(Sidebar, null, /*#__PURE__*/_react.default.createElement(Example, null)));
};

exports._default = _default;
