// @wp-g2/components/dist/cjs/__stories__/WIP/Interpolation.stories.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _styles = require("@wp-g2/styles");

var _utils = require("@wp-g2/utils");

var _react = _interopRequireWildcard(require("react"));

var _index = require("../../index");

var _default2 = {
  title: 'Examples/WIP/Utils/Interpolate'
};
exports.default = _default2;

var Circle = function (props) {
  return /*#__PURE__*/_react.default.createElement(_index.View, Object.assign({}, props, {
    css: [_styles.ui.background.blue, _styles.ui.frame({
      height: 100,
      width: 100
    }), _styles.ui.borderRadius.circle]
  }));
};

var Example = function () {
  var _useState = (0, _react.useState)({
    x: 0,
    y: 0
  }),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      position = _useState2[0],
      setPosition = _useState2[1];

  (0, _react.useEffect)(function () {
    var updatePosition = function (event) {
      setPosition({
        x: event.clientX,
        y: event.clientY
      });
    };

    document.addEventListener('mousemove', updatePosition);
    return function () {
      document.removeEventListener('mousemove', updatePosition);
    };
  }, []);
  var top = position.y - 50;
  var opacity = (0, _utils.interpolate)(position.x, [0, window.innerWidth], [1, 0]);
  var backgroundColor = (0, _utils.interpolate)(position.x, [0, window.innerWidth], ['red', 'blue']);
  return /*#__PURE__*/_react.default.createElement(Circle, {
    style: {
      backgroundColor: backgroundColor,
      left: position.x - 50,
      opacity: opacity,
      position: 'fixed',
      top: top
    }
  });
};

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(Example, null);
};

exports._default = _default;
