// @wp-g2/components/dist/cjs/__stories__/WIP/Lightbar.stories.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _Button = require("../../Button");

var _FormGroup = require("../../FormGroup");

var _Heading = require("../../Heading");

var _Slider = require("../../Slider");

var _Text = require("../../Text");

var _VStack = require("../../VStack");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var _default2 = {
  title: 'Examples/WIP/Performance/Lightbar'
};
exports.default = _default2;
var REQUEST_ANIMATION_FRAME_DELTA = 'REQUEST_ANIMATION_FRAME_DELTA';
var TOGGLE_START = 'TOGGLE_START';
var SET_WIDTH = 'SET_WIDTH';
var SET_SPEED = 'SET_SPEED';
var RECEIVE_API_STATE = 'RECEIVE_API_STATE';

var requestAnimationFrameDelta = function (delta) {
  return {
    type: REQUEST_ANIMATION_FRAME_DELTA,
    delta: delta
  };
};

var toggleStart = function () {
  return {
    type: TOGGLE_START
  };
};

var setWidth = function (width) {
  return {
    type: SET_WIDTH,
    width: width
  };
};

var setSpeed = function (speed) {
  return {
    type: SET_SPEED,
    speed: speed
  };
};

var receiveApiState = function (_ref) {
  var isStarted = _ref.isStarted,
      isStopping = _ref.isStopping,
      speed = _ref.speed,
      stoppingBounceCount = _ref.stoppingBounceCount,
      width = _ref.width;
  return {
    type: RECEIVE_API_STATE,
    receivedState: {
      isStarted: isStarted,
      width: width,
      speed: speed,
      isStopping: isStopping,
      stoppingBounceCount: stoppingBounceCount
    }
  };
};

var minMarginLeft = 0;

var reverseMovementDirection = function (state) {
  var nextMovementDirection = state.movementDirection === 'right' ? 'left' : 'right';
  var nextStoppingBounceCount = state.isStopping ? state.stoppingBounceCount + 1 : state.stoppingBounceCount;
  return _objectSpread(_objectSpread({}, state), {}, {
    movementDirection: nextMovementDirection,
    stoppingBounceCount: nextStoppingBounceCount
  });
};

var bounceLeftSide = function (state, nextMarginLeft) {
  var bouncedMarginLeft = Math.abs(nextMarginLeft);
  return reverseMovementDirection(_objectSpread(_objectSpread({}, state), {}, {
    marginLeft: bouncedMarginLeft
  }));
};

var moveLeft = function (state, pctDelta) {
  var nextMarginLeft = state.marginLeft - pctDelta;

  if (nextMarginLeft < minMarginLeft) {
    return bounceLeftSide(state, nextMarginLeft);
  } else {
    return _objectSpread(_objectSpread({}, state), {}, {
      marginLeft: nextMarginLeft
    });
  }
};

var bounceRightSide = function (state, nextMarginLeft) {
  var bouncedMarginLeft = state.maxMarginLeft - (nextMarginLeft - state.maxMarginLeft);
  return reverseMovementDirection(_objectSpread(_objectSpread({}, state), {}, {
    marginLeft: bouncedMarginLeft
  }));
};

var moveRight = function (state, pctDelta) {
  var nextMarginLeft = state.marginLeft + pctDelta;

  if (nextMarginLeft > state.maxMarginLeft) {
    return bounceRightSide(state, nextMarginLeft);
  } else {
    return _objectSpread(_objectSpread({}, state), {}, {
      marginLeft: nextMarginLeft
    });
  }
};

var stop = function (state) {
  return _objectSpread(_objectSpread({}, state), {}, {
    isStarted: false,
    isStopping: false,
    stoppingBounceCount: 0
  });
};

var stopIfNeeded = function (state) {
  if (!state.isStopping || state.stoppingBounceCount < 2) {
    return state;
  }

  var midpoint = state.maxMarginLeft / 2;
  var tolerance = state.maxMarginLeft * 0.02;
  var isAtMidpoint = state.marginLeft > midpoint - tolerance / 2 && state.marginLeft < midpoint + tolerance / 2;

  if (isAtMidpoint) {
    return stop(state);
  }

  return state;
};

var moveLightbar = function (state, delta) {
  var pctDelta = delta * (state.maxMarginLeft / state.speed);
  var nextState = state.movementDirection === 'right' ? moveRight(state, pctDelta) : moveLeft(state, pctDelta);
  return stopIfNeeded(nextState);
};

var startStopping = function (state) {
  return _objectSpread(_objectSpread({}, state), {}, {
    isStarted: false,
    isStopping: true,
    stoppingBounceCount: 0
  });
};

var start = function (state) {
  return _objectSpread(_objectSpread({}, state), {}, {
    isStarted: true
  });
};

var initialState = {
  isStarted: false,
  isStopping: false,
  stoppingBounceCount: 0,
  speed: 1000,
  movementDirection: 'right',
  maxMarginLeft: 98,
  marginLeft: 0,
  width: 2
};

var reducer = function (state, action) {
  switch (action.type) {
    case REQUEST_ANIMATION_FRAME_DELTA:
      {
        if (state.isStarted || state.isStopping) {
          return moveLightbar(state, action.delta);
        }

        return state;
      }

    case SET_SPEED:
      {
        return _objectSpread(_objectSpread({}, state), {}, {
          speed: action.speed
        });
      }

    case SET_WIDTH:
      {
        return _objectSpread(_objectSpread({}, state), {}, {
          width: action.width,
          maxMarginLeft: 100 - action.width
        });
      }

    case TOGGLE_START:
      {
        return state.isStarted ? startStopping(state) : start(state);
      }

    case RECEIVE_API_STATE:
      {
        return _objectSpread(_objectSpread({}, state), action.receivedState);
      }

    default:
      return state;
  }
};

var useLightbarReducer = function () {
  var _useReducer = (0, _react.useReducer)(reducer, initialState),
      _useReducer2 = (0, _slicedToArray2.default)(_useReducer, 2),
      state = _useReducer2[0],
      dispatch = _useReducer2[1];

  return [state, dispatch];
};

var useRequestAnimationFrame = function (dispatch) {
  var previousTimestampRef = (0, _react.useRef)();
  var requestRef = (0, _react.useRef)();
  var dispatchDelta = (0, _react.useCallback)(function (timestamp) {
    if (previousTimestampRef.current) {
      var delta = timestamp - previousTimestampRef.current;
      dispatch(requestAnimationFrameDelta(delta));
    }

    previousTimestampRef.current = timestamp;
    requestAnimationFrame(dispatchDelta);
  }, [dispatch]);
  (0, _react.useEffect)(function () {
    requestRef.current = requestAnimationFrame(dispatchDelta);
    return function () {
      return cancelAnimationFrame(requestRef.current);
    };
  }, [dispatchDelta]);
};

var Lightbar = function (_ref2) {
  var marginLeft = _ref2.marginLeft,
      width = _ref2.width;
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      border: '2px solid black',
      width: '98%',
      height: '250px',
      margin: 'auto'
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      backgroundColor: 'black',
      height: '100%',
      width: width + "%",
      marginLeft: marginLeft + "%"
    }
  }));
};

var sessionId = 'ABC-123';

var TherapistSettings = function (_ref3) {
  var dispatch = _ref3.dispatch,
      isStarted = _ref3.isStarted,
      isStopping = _ref3.isStopping,
      speed = _ref3.speed,
      width = _ref3.width;
  var handleToggleStart = (0, _react.useCallback)(function () {
    return dispatch(toggleStart());
  }, [dispatch]);
  var handleSetSpeed = (0, _react.useCallback)(function (speed) {
    return dispatch(setSpeed(speed));
  }, [dispatch]);
  var handleSetWidth = (0, _react.useCallback)(function (width) {
    return dispatch(setWidth(width));
  }, [dispatch]);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_FormGroup.FormGroup, {
    label: "Light speed"
  }, /*#__PURE__*/_react.default.createElement(_Slider.Slider, {
    max: 3000,
    min: 100,
    onChange: handleSetSpeed,
    value: speed
  })), /*#__PURE__*/_react.default.createElement(_FormGroup.FormGroup, {
    label: "Light width"
  }, /*#__PURE__*/_react.default.createElement(_Slider.Slider, {
    max: 60,
    min: 20,
    onChange: handleSetWidth,
    value: width
  })), /*#__PURE__*/_react.default.createElement(_Button.Button, {
    css: {
      margin: 'auto'
    },
    disabled: isStopping,
    onClick: handleToggleStart
  }, isStarted ? 'Stop' : 'Start'));
};

var _default = function () {
  var _useLightbarReducer = useLightbarReducer(),
      _useLightbarReducer2 = (0, _slicedToArray2.default)(_useLightbarReducer, 2),
      lightbarState = _useLightbarReducer2[0],
      dispatch = _useLightbarReducer2[1];

  useRequestAnimationFrame(dispatch);
  return /*#__PURE__*/_react.default.createElement(_VStack.VStack, {
    alignment: "center"
  }, /*#__PURE__*/_react.default.createElement(_Heading.Heading, {
    size: 1
  }, "EMDR Lightbar"), /*#__PURE__*/_react.default.createElement(_Text.Text, null, "Session ID to share with your client: ", sessionId), /*#__PURE__*/_react.default.createElement(TherapistSettings, Object.assign({}, lightbarState, {
    dispatch: dispatch
  })), /*#__PURE__*/_react.default.createElement(Lightbar, lightbarState));
};

exports._default = _default;
