// @wp-g2/components/dist/cjs/__stories__/WIP/ListGroups.stories.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _context = require("@wp-g2/context");

var _icons = require("@wp-g2/icons");

var _styles = require("@wp-g2/styles");

var _utils = require("@wp-g2/utils");

var _react = _interopRequireWildcard(require("react"));

var _index = require("../../index");

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", " {\n\t\t", " {\n\t\t\tposition: relative;\n\n\t\t\t&::before {\n\t\t\t\twidth: 12px;\n\t\t\t\tcontent: '';\n\t\t\t\tposition: absolute;\n\t\t\t\ttop: 0;\n\t\t\t\tbottom: 0;\n\t\t\t\tleft: 50%;\n\t\t\t\ttransform: translateX(-50%);\n\t\t\t\tbackground: blue;\n\t\t\t\topacity: 0.05;\n\t\t\t\tpointer-events: none;\n\t\t\t\tz-index: 1;\n\t\t\t}\n\t\t}\n\t}\n\n\t", ",\n\t", " {\n\t\t> * {\n\t\t\tposition: relative;\n\t\t\t&::before {\n\t\t\t\tbackground: pink;\n\t\t\t\tbottom: calc(", " * -1);\n\t\t\t\tcontent: '';\n\t\t\t\theight: ", ";\n\t\t\t\tleft: 0;\n\t\t\t\topacity: 0.15;\n\t\t\t\tpointer-events: none;\n\t\t\t\tposition: absolute;\n\t\t\t\tright: 0;\n\t\t\t\tz-index: 1;\n\t\t\t}\n\n\t\t\t&:last-child {\n\t\t\t\t&::before {\n\t\t\t\t\tdisplay: none;\n\t\t\t\t}\n\t\t\t}\n\t\t}\n\t}\n\t", " {\n\t\t> * {\n\t\t\tposition: relative;\n\t\t\t&::before {\n\t\t\t\tbackground: pink;\n\t\t\t\tright: calc(", " * -1);\n\t\t\t\tcontent: '';\n\t\t\t\theight: ", ";\n\t\t\t\ttop: 0;\n\t\t\t\tbottom: 0;\n\t\t\t\topacity: 0.15;\n\t\t\t\tpointer-events: none;\n\t\t\t\tposition: absolute;\n\t\t\t\tz-index: 1;\n\t\t\t}\n\n\t\t\t&:last-child {\n\t\t\t\t&::before {\n\t\t\t\t\tdisplay: none;\n\t\t\t\t}\n\t\t\t}\n\t\t}\n\t}\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var _default2 = {
  title: 'Examples/WIP/ListGroups'
};
exports.default = _default2;

var selector = function (target) {
  var key = Object.keys((0, _context.ns)());
  var n = (0, _context.ns)(target);
  return "[" + key + "=\"" + n[key] + "\"]";
};

var SpacingVisualizerView = _styles.styled.div(_templateObject(), selector('GridItem'), selector('Container'), selector('VStack'), selector('ListGroup'), _styles.ui.get('HStackSpacing'), _styles.ui.get('HStackSpacing'), selector('HStack'), _styles.ui.get('HStackSpacing'), _styles.ui.get('HStackSpacing'));

var SpacingVisualizer = function (_ref) {
  var children = _ref.children,
      visualize = _ref.visualize;
  if (!visualize) return children;
  return /*#__PURE__*/_react.default.createElement(SpacingVisualizerView, null, children);
};

var BaseGridItem = function (_ref2, forwardedRef) {
  var children = _ref2.children,
      props = (0, _objectWithoutProperties2.default)(_ref2, ["children"]);
  return /*#__PURE__*/_react.default.createElement(_index.Surface, Object.assign({
    border: true,
    css: [_styles.ui.frame.height('100%'), _styles.ui.padding(4)],
    variant: "tertiary"
  }, props, {
    ref: forwardedRef
  }), /*#__PURE__*/_react.default.createElement(_index.Container, {
    width: 300
  }, children));
};

var GridItem = (0, _context.contextConnect)(BaseGridItem, 'GridItem');

var ExampleCluster = function (_ref3) {
  var _ref3$horizontal = _ref3.horizontal,
      horizontal = _ref3$horizontal === void 0 ? true : _ref3$horizontal;
  return /*#__PURE__*/_react.default.createElement(GridItem, null, /*#__PURE__*/_react.default.createElement(_context.ContextSystemProvider, {
    value: {
      FormGroup: {
        horizontal: horizontal
      }
    }
  }, /*#__PURE__*/_react.default.createElement(_index.VStack, {
    spacing: 8
  }, /*#__PURE__*/_react.default.createElement(_index.View, null, /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.Subheading, null, "Site")), /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Title"
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    placeholder: "Some Title"
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Description"
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    minRows: 3,
    multiline: true,
    placeholder: "Description..."
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Socials"
  }, /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: 2
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    placeholder: "Twitter"
  }), /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    placeholder: "Instagram"
  }))), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Base Colors"
  }, /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: 2
  }, /*#__PURE__*/_react.default.createElement(_index.ColorControl, {
    color: "black"
  }, "Text"), /*#__PURE__*/_react.default.createElement(_index.ColorControl, {
    color: "#ddd"
  }, "Background"))), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Mobile Zoom"
  }, /*#__PURE__*/_react.default.createElement(_index.Switch, null)))), /*#__PURE__*/_react.default.createElement(_index.View, null, /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.Subheading, null, "Colors")), !horizontal ? /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.ColorControl, {
    color: "black"
  }, "Text"), /*#__PURE__*/_react.default.createElement(_index.ColorControl, {
    color: "#ddd"
  }, "Background"), /*#__PURE__*/_react.default.createElement(_index.ColorControl, {
    color: "#0f5"
  }, "Primary"), /*#__PURE__*/_react.default.createElement(_index.ColorControl, {
    color: "#0ff"
  }, "Accent")) : /*#__PURE__*/_react.default.createElement(_index.Grid, null, /*#__PURE__*/_react.default.createElement(_index.ColorControl, {
    color: "black"
  }, "Text"), /*#__PURE__*/_react.default.createElement(_index.ColorControl, {
    color: "#ddd"
  }, "Background"), /*#__PURE__*/_react.default.createElement(_index.ColorControl, {
    color: "#0f5"
  }, "Primary"), /*#__PURE__*/_react.default.createElement(_index.ColorControl, {
    color: "#0ff"
  }, "Accent"))))));
};

var ExampleCluster2 = function (props) {
  return /*#__PURE__*/_react.default.createElement(_index.ListGroups, props, /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(_index.ListGroupHeader, null, "Title", /*#__PURE__*/_react.default.createElement(_index.Button, {
    size: "xSmall",
    variant: "tertiary"
  }, "Edit")), /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: [1, 2]
  }, /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Action"
  }, /*#__PURE__*/_react.default.createElement(_index.Switch, null)), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Action"
  }, /*#__PURE__*/_react.default.createElement(_index.Switch, null)), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Action"
  }, /*#__PURE__*/_react.default.createElement(_index.Switch, null)), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Action"
  }, /*#__PURE__*/_react.default.createElement(_index.Switch, null)), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Action"
  }, /*#__PURE__*/_react.default.createElement(_index.Switch, null)))), /*#__PURE__*/_react.default.createElement(_index.ListGroup, {
    separator: false
  }, /*#__PURE__*/_react.default.createElement(_index.ListGroupHeader, null, "Title"), /*#__PURE__*/_react.default.createElement(_index.NavigatorLink, {
    href: "#",
    isBack: true
  }, /*#__PURE__*/_react.default.createElement(_index.MenuItem, {
    prefix: /*#__PURE__*/_react.default.createElement(_index.Icon, {
      icon: /*#__PURE__*/_react.default.createElement(_icons.FiChrome, null)
    })
  }, /*#__PURE__*/_react.default.createElement(_index.VStack, {
    spacing: 0
  }, /*#__PURE__*/_react.default.createElement(_index.Text, null, "Action"), /*#__PURE__*/_react.default.createElement(_index.Text, {
    size: "caption"
  }, "Help text.")))), /*#__PURE__*/_react.default.createElement(_index.NavigatorLink, {
    href: "#"
  }, /*#__PURE__*/_react.default.createElement(_index.MenuItem, {
    prefix: /*#__PURE__*/_react.default.createElement(_index.Icon, {
      icon: /*#__PURE__*/_react.default.createElement(_icons.FiCompass, null)
    })
  }, "Action")), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Title"
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    placeholder: "Some Title"
  })), /*#__PURE__*/_react.default.createElement(_index.NavigatorLink, {
    href: "#"
  }, /*#__PURE__*/_react.default.createElement(_index.MenuItem, {
    as: "button",
    prefix: /*#__PURE__*/_react.default.createElement(_index.Icon, {
      icon: /*#__PURE__*/_react.default.createElement(_icons.FiCompass, null)
    }),
    suffix: /*#__PURE__*/_react.default.createElement(_index.Text, {
      isBlock: true,
      size: "caption"
    }, "Detail")
  }, "Action")), /*#__PURE__*/_react.default.createElement(_index.NavigatorLink, {
    href: "#"
  }, /*#__PURE__*/_react.default.createElement(_index.MenuItem, {
    as: "button",
    suffix: /*#__PURE__*/_react.default.createElement(_index.Text, {
      isBlock: true,
      size: "caption"
    }, "Detail")
  }, "Action")), /*#__PURE__*/_react.default.createElement(_index.MenuItem, {
    as: "button",
    suffix: /*#__PURE__*/_react.default.createElement(_index.Text, {
      isBlock: true,
      size: "caption"
    }, "Detail")
  }, "Action"), /*#__PURE__*/_react.default.createElement(_index.ListGroupFooter, null, "Some footer caption")));
};

var AllTheThings = function () {
  var _useLocalState = (0, _utils.useLocalState)('WIP/ListGroups/SpacingVisualizer', false),
      _useLocalState2 = (0, _slicedToArray2.default)(_useLocalState, 2),
      visualize = _useLocalState2[0],
      setVisualize = _useLocalState2[1];

  var _useState = (0, _react.useState)(4),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      gridBase = _useState2[0],
      setGridBase = _useState2[1];

  return /*#__PURE__*/_react.default.createElement(_styles.ThemeProvider, {
    theme: {
      gridBase: _styles.ui.value.px(gridBase)
    }
  }, /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    mb: 5
  }, /*#__PURE__*/_react.default.createElement(_index.HStack, null, /*#__PURE__*/_react.default.createElement(_index.Heading, {
    size: 1
  }, "ListGroup Examples"), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    isInline: true,
    label: "Grid Base"
  }, /*#__PURE__*/_react.default.createElement(_index.SegmentedControl, {
    onChange: setGridBase,
    options: [{
      value: 2,
      label: '2px'
    }, {
      value: 4,
      label: '4px'
    }, {
      value: 6,
      label: '6px'
    }, {
      value: 8,
      label: '8px'
    }, {
      value: 10,
      label: '10px'
    }, {
      value: 12,
      label: '12px'
    }],
    value: gridBase
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    isInline: true,
    label: "Show Visualizer"
  }, /*#__PURE__*/_react.default.createElement(_index.Switch, {
    checked: visualize,
    onChange: setVisualize
  })))), /*#__PURE__*/_react.default.createElement(SpacingVisualizer, {
    visualize: visualize
  }, /*#__PURE__*/_react.default.createElement(_index.Grid, {
    align: "flex-start",
    columns: [1, 2, 4],
    gap: 5
  }, /*#__PURE__*/_react.default.createElement(ExampleCluster, null), /*#__PURE__*/_react.default.createElement(ExampleCluster, {
    horizontal: false
  }), /*#__PURE__*/_react.default.createElement(GridItem, null, /*#__PURE__*/_react.default.createElement(ExampleCluster2, null)), /*#__PURE__*/_react.default.createElement(GridItem, {
    css: [_styles.ui.background.secondary]
  }, /*#__PURE__*/_react.default.createElement(ExampleCluster2, {
    inset: true
  })))));
};

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(AllTheThings, null);
};

exports._default = _default;
