// @wp-g2/components/dist/cjs/__stories__/WIP/Performance.stories.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _protokit = require("@wp-g2/protokit");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireWildcard(require("react"));

var _index = require("../../index");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\talign-items: center;\n\tbottom: 0;\n\tdisplay: flex;\n\tjustify-content: center;\n\tleft: 0;\n\tposition: fixed;\n\tright: 0;\n\ttop: 0;\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var _default2 = {
  title: 'Examples/WIP/Performance'
};
exports.default = _default2;
var Frame = (0, _styles.styled)(_index.Surface)(_templateObject());
var dimensionSchema = new _protokit.Schema(function () {
  return {
    title: '',
    x: 0,
    y: 1,
    z: 0
  };
});

var SliderTextInput = /*#__PURE__*/_react.default.memo(function (_ref) {
  var onChange = _ref.onChange,
      prop = _ref.prop,
      value = _ref.value;

  var handleOnChange = _react.default.useCallback(function (next) {
    var _onChange;

    onChange((_onChange = {}, _onChange[prop] = next, _onChange));
  }, [onChange, prop]);

  return /*#__PURE__*/_react.default.createElement(_index.Grid, null, /*#__PURE__*/_react.default.createElement(_index.Slider, {
    onChange: handleOnChange,
    value: value
  }), /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    onChange: handleOnChange,
    type: "number",
    value: value
  }));
});

var DimensionCard = /*#__PURE__*/_react.default.memo(function (_ref2) {
  var id = _ref2.id,
      onChange = _ref2.onChange,
      title = _ref2.title,
      x = _ref2.x,
      y = _ref2.y,
      z = _ref2.z;

  var updateTitle = _react.default.useCallback(function (next) {
    onChange({
      id: id,
      title: next
    });
  }, [onChange, id]);

  var updateValue = _react.default.useCallback(function (next) {
    onChange(_objectSpread({
      id: id
    }, next));
  }, [onChange, id]);

  return /*#__PURE__*/_react.default.createElement(_index.Card, null, /*#__PURE__*/_react.default.createElement(_index.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, /*#__PURE__*/_react.default.createElement(_index.ListGroupHeader, null, "Dimensions"), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Title"
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    onChange: updateTitle,
    value: title
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "x"
  }, /*#__PURE__*/_react.default.createElement(SliderTextInput, {
    onChange: updateValue,
    prop: "x",
    value: x
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "y"
  }, /*#__PURE__*/_react.default.createElement(SliderTextInput, {
    onChange: updateValue,
    prop: "y",
    value: y
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "z"
  }, /*#__PURE__*/_react.default.createElement(SliderTextInput, {
    onChange: updateValue,
    prop: "z",
    value: z
  })))));
});

var Example = function () {
  var _useState = (0, _react.useState)([].concat(dimensionSchema.make(1))),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      dimensions = _useState2[0],
      setDimensions = _useState2[1];

  var updateDimension = _react.default.useCallback(function (next) {
    setDimensions(function (prev) {
      return prev.map(function (d) {
        if (d.id !== next.id) return d;
        return _objectSpread(_objectSpread({}, d), next);
      });
    });
  }, []);

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_index.Container, {
    css: (0, _styles.css)([_styles.ui.position.relative, _styles.ui.zIndex(2)]),
    width: 600
  }, /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.HStack, null, /*#__PURE__*/_react.default.createElement(_index.Button, {
    onClick: function addDimension() {
      setDimensions(function (prev) {
        return [].concat(prev, [dimensionSchema.makeOne()]);
      });
    },
    variant: "primary"
  }, "Add New")), /*#__PURE__*/_react.default.createElement(_index.ListGroup, null, dimensions.map(function (item) {
    return /*#__PURE__*/_react.default.createElement(DimensionCard, Object.assign({}, item, {
      onChange: updateDimension
    }));
  })))), /*#__PURE__*/_react.default.createElement(Frame, {
    variant: "dotted"
  }));
};

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(Example, null);
};

exports._default = _default;
