// @wp-g2/components/dist/cjs/__stories__/WIP/PerformanceNative.stories.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _protokit = require("@wp-g2/protokit");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireWildcard(require("react"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var _default2 = {
  title: 'Examples/WIP/PerformanceNative'
};
exports.default = _default2;
var dimensionSchema = new _protokit.Schema(function () {
  return {
    title: '',
    x: 0,
    y: 1,
    z: 0
  };
});

var SomeContext = /*#__PURE__*/_react.default.createContext({});

var useSomeContext = function () {
  return _react.default.useContext(SomeContext);
};

var OtherContext = /*#__PURE__*/_react.default.createContext({});

var useOtherContext = function () {
  return _react.default.useContext(OtherContext);
};

var SliderTextInput = /*#__PURE__*/_react.default.memo(function (_ref) {
  var onChange = _ref.onChange,
      prop = _ref.prop,
      value = _ref.value;

  var _useSomeContext = useSomeContext(),
      slider = _useSomeContext.slider;

  var handleOnChange = _react.default.useCallback(function (next) {
    var _onChange;

    onChange((_onChange = {}, _onChange[prop] = next.target.value, _onChange));
  }, [onChange, prop]);

  return /*#__PURE__*/_react.default.createElement("div", null, slider && /*#__PURE__*/_react.default.createElement("input", {
    onChange: handleOnChange,
    type: "range",
    value: value
  }), /*#__PURE__*/_react.default.createElement("input", {
    onChange: handleOnChange,
    type: "number",
    value: value
  }));
});

var TestFormGroup = /*#__PURE__*/_react.default.memo(function (_ref2) {
  var children = _ref2.children,
      label = _ref2.label;

  var contextValue = _react.default.useMemo(function () {
    return {
      slider: true
    };
  }, []);

  return /*#__PURE__*/_react.default.createElement(SomeContext.Provider, {
    value: contextValue
  }, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("label", null, label, children)))));
});

var DimensionCard = /*#__PURE__*/_react.default.memo(function (_ref3) {
  var id = _ref3.id,
      onChange = _ref3.onChange,
      title = _ref3.title,
      x = _ref3.x,
      y = _ref3.y,
      z = _ref3.z;

  var updateTitle = _react.default.useCallback(function (next) {
    onChange({
      id: id,
      title: next.target.value
    });
  }, [onChange, id]);

  var updateValue = _react.default.useCallback(function (next) {
    onChange(_objectSpread({
      id: id
    }, next));
  }, [onChange, id]);

  return /*#__PURE__*/_react.default.createElement("li", null, /*#__PURE__*/_react.default.createElement("h5", null, "Dimensions"), /*#__PURE__*/_react.default.createElement(TestFormGroup, {
    label: "Title"
  }, /*#__PURE__*/_react.default.createElement("input", {
    onChange: updateTitle,
    value: title
  })), /*#__PURE__*/_react.default.createElement(TestFormGroup, {
    label: "x"
  }, /*#__PURE__*/_react.default.createElement(SliderTextInput, {
    onChange: updateValue,
    prop: "x",
    value: x
  })), /*#__PURE__*/_react.default.createElement(TestFormGroup, {
    label: "y"
  }, /*#__PURE__*/_react.default.createElement(SliderTextInput, {
    onChange: updateValue,
    prop: "y",
    value: y
  })), /*#__PURE__*/_react.default.createElement(TestFormGroup, {
    label: "z"
  }, /*#__PURE__*/_react.default.createElement(SliderTextInput, {
    onChange: updateValue,
    prop: "z",
    value: z
  })));
});

var Example = function () {
  var _useState = (0, _react.useState)([].concat(dimensionSchema.make(1))),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      dimensions = _useState2[0],
      setDimensions = _useState2[1];

  var updateDimension = _react.default.useCallback(function (next) {
    setDimensions(function (prev) {
      return prev.map(function (d) {
        if (d.id !== next.id) return d;
        return _objectSpread(_objectSpread({}, d), next);
      });
    });
  }, []);

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("button", {
    onClick: function addDimension() {
      setDimensions(function (prev) {
        return [].concat(prev, [dimensionSchema.makeOne()]);
      });
    }
  }, "Add New"), /*#__PURE__*/_react.default.createElement(OtherContext.Provider, {
    value: {
      ok: true
    }
  }, /*#__PURE__*/_react.default.createElement("ul", null, dimensions.map(function (item) {
    return /*#__PURE__*/_react.default.createElement(DimensionCard, Object.assign({}, item, {
      onChange: updateDimension
    }));
  }))));
};

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(Example, null);
};

exports._default = _default;
