// @wp-g2/components/dist/cjs/__stories__/WIP/Protokit.stories.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _protokit = require("@wp-g2/protokit");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireWildcard(require("react"));

var _index = require("../../index");

var _default2 = {
  title: 'Examples/WIP/Protokit/ListData'
};
exports.default = _default2;
var itemSchema = new _protokit.Schema(function () {
  return {
    description: _protokit.faker.lorem.paragraph(),
    title: _protokit.faker.name.firstName()
  };
});

var Example = function () {
  var _useState = (0, _react.useState)(false),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      isLoading = _useState2[0],
      setIsLoading = _useState2[1];

  var _useState3 = (0, _react.useState)(false),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      isError = _useState4[0],
      setIsError = _useState4[1];

  var _useState5 = (0, _react.useState)(false),
      _useState6 = (0, _slicedToArray2.default)(_useState5, 2),
      fail = _useState6[0],
      setFail = _useState6[1];

  var _useState7 = (0, _react.useState)(''),
      _useState8 = (0, _slicedToArray2.default)(_useState7, 2),
      search = _useState8[0],
      setSearch = _useState8[1];

  var _useListData = (0, _protokit.useListData)({
    initialCount: 20,
    schema: itemSchema
  }),
      _useListData2 = (0, _slicedToArray2.default)(_useListData, 2),
      data = _useListData2[0],
      fns = _useListData2[1];

  var handleOnLoadMore = /*#__PURE__*/function () {
    var _ref = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee() {
      return _regenerator.default.wrap(function (_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              setIsLoading(true);
              setIsError(false);
              _context.prev = 2;
              _context.next = 5;
              return (0, _protokit.mockRequest)({
                data: function () {
                  return fns.loadMore({
                    limit: 10
                  });
                },
                status: fail ? 400 : 200,
                timeout: 300
              });

            case 5:
              _context.next = 10;
              break;

            case 7:
              _context.prev = 7;
              _context.t0 = _context["catch"](2);
              setIsError(true);

            case 10:
              setIsLoading(false);

            case 11:
            case "end":
              return _context.stop();
          }
        }
      }, _callee, null, [[2, 7]]);
    }));

    return function () {
      return _ref.apply(this, arguments);
    };
  }();

  return /*#__PURE__*/_react.default.createElement(_index.Container, {
    width: 800
  }, /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.Alerts, null, isError && /*#__PURE__*/_react.default.createElement(_index.Alert, {
    isDismissable: true,
    onDismiss: function onDismiss() {
      return setIsError(false);
    },
    status: "critical"
  }, /*#__PURE__*/_react.default.createElement(_index.Text, null, "Bad data :("))), /*#__PURE__*/_react.default.createElement(_index.HStack, null, /*#__PURE__*/_react.default.createElement(_index.HStack, {
    alignment: "left"
  }, /*#__PURE__*/_react.default.createElement(_index.Button, {
    disabled: isLoading,
    onClick: handleOnLoadMore
  }, "Load More"), /*#__PURE__*/_react.default.createElement(_index.SearchInput, {
    onChange: function onChange(next) {
      setSearch(next);
      fns.search(next);
    },
    value: search
  }), isLoading && /*#__PURE__*/_react.default.createElement(_index.Spinner, null)), /*#__PURE__*/_react.default.createElement(_index.Spacer, null), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Force Fail",
    templateColumns: "1fr auto"
  }, /*#__PURE__*/_react.default.createElement(_index.Switch, {
    checked: fail,
    onChange: setFail
  }))), /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: [2, 3]
  }, /*#__PURE__*/_react.default.createElement(_index.AnimatedContainer, null, data.map(function (item, index) {
    return /*#__PURE__*/_react.default.createElement(_index.Animated, {
      auto: true,
      key: item.id,
      transition: {// This should feel easier...
        // delay: 0.05 * (index - data.length + limit),
      }
    }, /*#__PURE__*/_react.default.createElement(_index.Card, {
      css: [_styles.ui.frame.height(100), {
        cursor: 'pointer',
        userSelect: 'none',
        zIndex: 2
      }, _styles.ui.hover([_styles.ui.scale(1.1), _styles.ui.offset.y(-10), _styles.ui.shadow({
        radius: 20
      })]), _styles.ui.active(_styles.ui.scale(0.9)), _styles.ui.animation.default],
      onClick: function onClick() {
        return fns.remove({
          id: item.id
        });
      }
    }, /*#__PURE__*/_react.default.createElement(_index.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.Text, {
      weight: "bold"
    }, index, ". ", item.title), /*#__PURE__*/_react.default.createElement(_index.Text, {
      numberOfLines: 2,
      size: "caption",
      truncate: true,
      variant: "muted"
    }, item.description)))));
  })))));
};

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(Example, null);
};

exports._default = _default;
