// @wp-g2/components/dist/cjs/__stories__/WIP/ProtokitPlugin.stories.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _protokit = require("@wp-g2/protokit");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireWildcard(require("react"));

var _reakit = require("reakit");

var _index = require("../../index");

var _default2 = {
  title: 'Examples/WIP/Protokit/SamplePlugin'
};
exports.default = _default2;
var userSchema = new _protokit.Schema(function () {
  return {
    biography: _protokit.faker.lorem.paragraph(),
    firstName: _protokit.faker.name.firstName(),
    lastName: _protokit.faker.name.lastName(),
    userName: _protokit.faker.internet.userName(),
    avatar: _protokit.faker.random.arrayElement([_protokit.faker.image.avatar(), null]),
    role: _protokit.faker.random.arrayElement(['user', 'editor', 'admin'])
  };
});

var UserCard = function (_ref) {
  var user = _ref.user,
      props = (0, _objectWithoutProperties2.default)(_ref, ["user"]);
  return /*#__PURE__*/_react.default.createElement(_index.Card, props, /*#__PURE__*/_react.default.createElement(_index.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.HStack, {
    alignment: "topLeft",
    spacing: 3
  }, /*#__PURE__*/_react.default.createElement(_index.Avatar, {
    name: user.firstName + " " + user.lastName,
    size: "large",
    src: user.avatar
  }), /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.VStack, {
    spacing: 1
  }, /*#__PURE__*/_react.default.createElement(_index.Text, {
    weight: "bold"
  }, user.firstName, " ", user.lastName), /*#__PURE__*/_react.default.createElement(_index.Text, {
    size: 11,
    variant: "muted"
  }, user.userName)), /*#__PURE__*/_react.default.createElement(_index.Badge, null, user.role)))));
};

var EditUser = function (_ref2) {
  var onClose = _ref2.onClose,
      userData = _ref2.user,
      usersFn = _ref2.usersFn;

  var _useData = (0, _protokit.useData)(userData),
      _useData2 = (0, _slicedToArray2.default)(_useData, 2),
      user = _useData2[0],
      userFns = _useData2[1];

  var name = user.firstName + " " + user.lastName;
  var bindToField = userFns.bindToField,
      hasChanges = userFns.hasChanges,
      reset = userFns.reset;

  var _useState = (0, _react.useState)(false),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      isLoading = _useState2[0],
      setIsLoading = _useState2[1];

  var _useState3 = (0, _react.useState)(),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      error = _useState4[0],
      setError = _useState4[1];

  var applyChanges = /*#__PURE__*/function () {
    var _ref3 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee() {
      return _regenerator.default.wrap(function (_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              setIsLoading(true);
              setError();
              _context.prev = 2;
              _context.next = 5;
              return (0, _protokit.mockRequest)({
                action: function () {
                  usersFn.update(user);
                },
                // status: 400,
                message: 'Could not save user data.',
                timeout: 300
              });

            case 5:
              onClose();
              _context.next = 11;
              break;

            case 8:
              _context.prev = 8;
              _context.t0 = _context["catch"](2);
              setError(_context.t0.message);

            case 11:
              setIsLoading(false);

            case 12:
            case "end":
              return _context.stop();
          }
        }
      }, _callee, null, [[2, 8]]);
    }));

    return function () {
      return _ref3.apply(this, arguments);
    };
  }();

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_index.ModalHeader, {
    title: "Edit User"
  }), /*#__PURE__*/_react.default.createElement(_index.ModalBody, null, /*#__PURE__*/_react.default.createElement(_index.Alerts, null, !!error && /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.Alert, {
    status: "critical"
  }, /*#__PURE__*/_react.default.createElement(_index.Text, null, error)))), /*#__PURE__*/_react.default.createElement(_index.Grid, {
    templateColumns: '120px 1fr'
  }, /*#__PURE__*/_react.default.createElement(_index.View, {
    css: [_styles.ui.alignment.center]
  }, /*#__PURE__*/_react.default.createElement(_index.Avatar, {
    name: name,
    size: 80,
    src: user.avatar
  })), /*#__PURE__*/_react.default.createElement(_index.View, null, /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "First name"
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, bindToField('firstName'))), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Last name"
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, bindToField('lastName'))), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "User name"
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, bindToField('userName'))), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Role"
  }, /*#__PURE__*/_react.default.createElement(_index.Select, Object.assign({}, bindToField('role'), {
    options: [{
      value: 'user',
      label: 'User'
    }, {
      value: 'editor',
      label: 'Editor'
    }, {
      value: 'admin',
      label: 'Admin'
    }]
  })))))), /*#__PURE__*/_react.default.createElement(_index.ModalFooter, null, /*#__PURE__*/_react.default.createElement(_index.Button, {
    onClick: function onClick() {
      onClose();
      reset();
    },
    variant: "tertiary"
  }, "Discard Changes"), /*#__PURE__*/_react.default.createElement(_index.Button, {
    disabled: !hasChanges(),
    isLoading: isLoading,
    onClick: applyChanges,
    variant: "primary"
  }, "Save Changes")));
};

function Example() {
  var _useState5 = (0, _react.useState)(''),
      _useState6 = (0, _slicedToArray2.default)(_useState5, 2),
      query = _useState6[0],
      setQuery = _useState6[1];

  var _useListData = (0, _protokit.useListData)({
    schema: userSchema
  }),
      _useListData2 = (0, _slicedToArray2.default)(_useListData, 2),
      users = _useListData2[0],
      usersFn = _useListData2[1];

  var _useState7 = (0, _react.useState)(),
      _useState8 = (0, _slicedToArray2.default)(_useState7, 2),
      currentUser = _useState8[0],
      setCurrentUser = _useState8[1];

  var dialog = (0, _reakit.useDialogState)({
    animated: true
  });
  return /*#__PURE__*/_react.default.createElement(_index.Container, {
    width: 800
  }, /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    mb: 6
  }, /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.Heading, {
    size: 1
  }, "Users"), /*#__PURE__*/_react.default.createElement(_index.HStack, null, /*#__PURE__*/_react.default.createElement(_index.SearchInput, {
    onChange: function onChange(next) {
      usersFn.search(next);
      setQuery(next);
    },
    value: query
  }), /*#__PURE__*/_react.default.createElement(_index.Spacer, null), /*#__PURE__*/_react.default.createElement(_index.Button, {
    onClick: usersFn.loadMore
  }, "Load More")))), /*#__PURE__*/_react.default.createElement(_index.Grid, {
    columns: 3
  }, /*#__PURE__*/_react.default.createElement(_index.AnimatedContainer, null, users.map(function (user) {
    return /*#__PURE__*/_react.default.createElement(_index.Animated, {
      auto: true,
      key: user.id
    }, /*#__PURE__*/_react.default.createElement(UserCard, {
      css: [{
        cursor: 'pointer'
      }, _styles.ui.animation.default, _styles.ui.zIndex(1), _styles.ui.hover([_styles.ui.scale(1.05)]), _styles.ui.active([_styles.ui.scale(0.92)])],
      onClick: function onClick() {
        setCurrentUser(user);
        dialog.show();
      },
      user: user
    }));
  }))), /*#__PURE__*/_react.default.createElement(_index.Modal, {
    dialog: dialog
  }, /*#__PURE__*/_react.default.createElement(EditUser, {
    onClose: dialog.hide,
    user: currentUser,
    usersFn: usersFn
  })));
}

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(Example, null);
};

exports._default = _default;
