// @wp-g2/components/dist/cjs/__stories__/WIP/UserProfilesPlugin.stories.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = _default;
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _context = require("@wp-g2/context");

var _protokit = require("@wp-g2/protokit");

var _styles = require("@wp-g2/styles");

var _utils = require("@wp-g2/utils");

var _react = _interopRequireWildcard(require("react"));

var _reactBeautifulDnd = require("react-beautiful-dnd");

var _index = require("../../index");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var _default2 = {
  title: 'Examples/WIP/UserProfilePlugin'
};
exports.default = _default2;
var AppContext = /*#__PURE__*/(0, _react.createContext)({});

var useAppContext = function () {
  return (0, _react.useContext)(AppContext);
};

var ContextSystemProviderValue = {
  FormGroup: {
    alignLabel: 'right',
    gap: 7
  },
  Radio: {
    gap: 2
  }
};
var userSchema = new _protokit.Schema(function () {
  return {
    avatar: _protokit.faker.image.avatar(),
    firstName: _protokit.faker.name.firstName(),
    lastName: _protokit.faker.name.lastName(),
    userName: _protokit.faker.internet.userName(),
    nickName: _protokit.faker.internet.userName(),
    displayName: _protokit.faker.internet.userName(),
    website: _protokit.faker.internet.url(),
    biography: _protokit.faker.lorem.sentences(),
    avatarRating: 'G'
  };
});
var SectionComponents = {
  name: NameSection,
  about: AboutSection,
  status: StatusSection,
  avatar: AvatarSection
};

function Section(_ref) {
  var children = _ref.children;
  return children;
}

function SectionWrapper(_ref2) {
  var id = _ref2.id;

  var _useAppContext = useAppContext(),
      sectionsData = _useAppContext.sections;

  var sections = sectionsData[id];
  return /*#__PURE__*/_react.default.createElement(_reactBeautifulDnd.Droppable, {
    droppableId: id
  }, function (provided) {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_index.VStack, Object.assign({
      ref: provided.innerRef
    }, provided.droppableProps, {
      spacing: 6
    }), sections.map(function (section, index) {
      var Component = SectionComponents[section];
      return /*#__PURE__*/_react.default.createElement(_reactBeautifulDnd.Draggable, {
        draggableId: section,
        index: index,
        key: section
      }, function (provided) {
        return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_index.View, Object.assign({
          ref: provided.innerRef
        }, provided.draggableProps, provided.dragHandleProps), /*#__PURE__*/_react.default.createElement(Component, null)));
      });
    }), provided.placeholder), /*#__PURE__*/_react.default.createElement(_index.Spacer, {
      pb: 4
    }));
  });
}

function NameSection() {
  var _useAppContext2 = useAppContext(),
      update = _useAppContext2.update,
      user = _useAppContext2.user;

  return /*#__PURE__*/_react.default.createElement(Section, null, /*#__PURE__*/_react.default.createElement(_index.Card, null, /*#__PURE__*/_react.default.createElement(_index.CardHeader, null, /*#__PURE__*/_react.default.createElement(_index.Heading, {
    size: 5
  }, "Name")), /*#__PURE__*/_react.default.createElement(_index.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.Container, {
    alignment: "left",
    width: 640
  }, /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "User Name"
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    onChange: update('userName'),
    value: user.userName
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "First Name"
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    onChange: update('firstName'),
    value: user.firstName
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Last Name"
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    onChange: update('lastName'),
    value: user.lastName
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Nickname (required)"
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    onChange: update('nickName'),
    value: user.nickName
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Display Name"
  }, /*#__PURE__*/_react.default.createElement(_index.Select, {
    options: [{
      value: user.userName,
      label: user.userName
    }, {
      value: user.nickName,
      label: user.nickName
    }, {
      value: user.firstName,
      label: user.firstName
    }]
  }))))));
}

function AboutSection() {
  var _useAppContext3 = useAppContext(),
      update = _useAppContext3.update,
      user = _useAppContext3.user;

  return /*#__PURE__*/_react.default.createElement(Section, null, /*#__PURE__*/_react.default.createElement(_index.Card, null, /*#__PURE__*/_react.default.createElement(_index.CardHeader, null, /*#__PURE__*/_react.default.createElement(_index.Heading, {
    size: 5
  }, "About")), /*#__PURE__*/_react.default.createElement(_index.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.Container, {
    alignment: "left",
    width: 640
  }, /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Website"
  }, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    onChange: update('website'),
    value: user.website
  })), /*#__PURE__*/_react.default.createElement(_index.FormGroup, {
    label: "Biography"
  }, /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.TextInput, {
    maxRows: 8,
    minRows: 4,
    multiline: 4,
    onChange: update('biography'),
    value: user.biography
  }), /*#__PURE__*/_react.default.createElement(_index.Text, {
    variant: "muted"
  }, "Share a little biographical information to fill out your profile. This may be shown publicly.")))))));
}

function StatusSection() {
  return /*#__PURE__*/_react.default.createElement(Section, null, /*#__PURE__*/_react.default.createElement(_index.Card, null, /*#__PURE__*/_react.default.createElement(_index.CardHeader, null, /*#__PURE__*/_react.default.createElement(_index.Heading, {
    size: 5
  }, "Status")), /*#__PURE__*/_react.default.createElement(_index.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.Text, null, "Registered on: ", /*#__PURE__*/_react.default.createElement("strong", null, "August 25, 2020 7:54 pm"))), /*#__PURE__*/_react.default.createElement(_index.CardFooter, {
    justify: "flex-end"
  }, /*#__PURE__*/_react.default.createElement(_index.Button, null, "View User"), /*#__PURE__*/_react.default.createElement(_index.Button, {
    variant: "primary"
  }, "Update"))));
}

function AvatarSection() {
  var _useAppContext4 = useAppContext(),
      update = _useAppContext4.update,
      user = _useAppContext4.user;

  return /*#__PURE__*/_react.default.createElement(Section, null, /*#__PURE__*/_react.default.createElement(_index.Card, null, /*#__PURE__*/_react.default.createElement(_index.CardHeader, null, /*#__PURE__*/_react.default.createElement(_index.Heading, {
    size: 5
  }, "Avatar")), /*#__PURE__*/_react.default.createElement(_index.CardBody, null, /*#__PURE__*/_react.default.createElement(_index.VStack, {
    spacing: 4
  }, /*#__PURE__*/_react.default.createElement(_index.HStack, {
    alignment: "center"
  }, /*#__PURE__*/_react.default.createElement(_index.Avatar, {
    size: 80,
    src: user.avatar
  })), /*#__PURE__*/_react.default.createElement(_index.HStack, {
    alignment: "left"
  }, /*#__PURE__*/_react.default.createElement(_index.Button, {
    onClick: function onClick() {
      return update('avatar')(_protokit.faker.image.avatar());
    }
  }, "Choose From Media"), /*#__PURE__*/_react.default.createElement(_index.Button, {
    isDestructive: true,
    onClick: function onClick() {
      return update('avatar')(null);
    },
    variant: "tertiary"
  }, "Remove")), /*#__PURE__*/_react.default.createElement(_index.RadioGroup, {
    onChange: update('avatarRating'),
    value: user.avatarRating
  }, /*#__PURE__*/_react.default.createElement(_index.HStack, {
    alignment: "left",
    spacing: 5
  }, /*#__PURE__*/_react.default.createElement(_index.Radio, {
    label: "G",
    value: "G"
  }), /*#__PURE__*/_react.default.createElement(_index.Radio, {
    label: "PR",
    value: "PG"
  }), /*#__PURE__*/_react.default.createElement(_index.Radio, {
    label: "R",
    value: "R"
  }), /*#__PURE__*/_react.default.createElement(_index.Radio, {
    label: "X",
    value: "X"
  })))))));
}

function ProfilePanel() {
  return /*#__PURE__*/_react.default.createElement(SectionWrapper, {
    id: "main"
  });
}

function AccountPanel() {
  return /*#__PURE__*/_react.default.createElement(_index.Text, null, "Account");
}

function OptionsPanel() {
  return /*#__PURE__*/_react.default.createElement(_index.Text, null, "Options");
}

function PermissionsPanel() {
  return /*#__PURE__*/_react.default.createElement(_index.Text, null, "Permissions");
}

function Sidebar() {
  return /*#__PURE__*/_react.default.createElement(SectionWrapper, {
    id: "sidebar"
  });
}

function App() {
  var _useState = (0, _react.useState)(userSchema.makeOne()),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      user = _useState2[0],
      setUser = _useState2[1];

  var _useListState = (0, _utils.useListState)(['name', 'about']),
      _useListState2 = (0, _slicedToArray2.default)(_useListState, 2),
      main = _useListState2[0],
      setMain = _useListState2[1];

  var _useListState3 = (0, _utils.useListState)(['status', 'avatar']),
      _useListState4 = (0, _slicedToArray2.default)(_useListState3, 2),
      sidebar = _useListState4[0],
      setSidebar = _useListState4[1];

  return /*#__PURE__*/_react.default.createElement(AppContext.Provider, {
    value: {
      user: user,
      setUser: setUser,
      update: function update(key) {
        return function (next) {
          return setUser(function (prev) {
            var _objectSpread2;

            return _objectSpread(_objectSpread({}, prev), {}, (_objectSpread2 = {}, _objectSpread2[key] = next, _objectSpread2));
          });
        };
      },
      sections: {
        main: main,
        sidebar: sidebar
      }
    }
  }, /*#__PURE__*/_react.default.createElement(_reactBeautifulDnd.DragDropContext, {
    onDragEnd: function onDragEnd(_ref3) {
      var destination = _ref3.destination,
          source = _ref3.source;
      var updateFns = {
        main: setMain,
        sidebar: setSidebar
      }; // Sorting content

      if ((destination == null ? void 0 : destination.droppableId) === (source == null ? void 0 : source.droppableId)) {
        var updateFn = updateFns[source == null ? void 0 : source.droppableId];

        if (updateFn) {
          updateFn.move(source.index, destination.index);
        }

        return;
      }

      var prevSection = updateFns[source == null ? void 0 : source.droppableId];
      var nextSection = updateFns[destination == null ? void 0 : destination.droppableId];
      var next = prevSection.find({
        at: source.index
      });

      if (next && nextSection) {
        prevSection.remove({
          at: source.index
        });
        nextSection.insert({
          at: destination.index,
          item: next
        });
      }
    }
  }, /*#__PURE__*/_react.default.createElement(_index.Container, null, /*#__PURE__*/_react.default.createElement(_index.Tabs, null, /*#__PURE__*/_react.default.createElement(_index.TabList, null, /*#__PURE__*/_react.default.createElement(_index.HStack, {
    alignment: "left"
  }, /*#__PURE__*/_react.default.createElement(_index.Tab, {
    css: [_styles.ui.padding.x(3)]
  }, "Profile"), /*#__PURE__*/_react.default.createElement(_index.Tab, {
    css: [_styles.ui.padding.x(3)]
  }, "Account"), /*#__PURE__*/_react.default.createElement(_index.Tab, {
    css: [_styles.ui.padding.x(3)]
  }, "Options"), /*#__PURE__*/_react.default.createElement(_index.Tab, {
    css: [_styles.ui.padding.x(3)]
  }, "Permission"))), /*#__PURE__*/_react.default.createElement(_index.Surface, {
    css: [_styles.ui.padding(5)],
    variant: "secondary"
  }, /*#__PURE__*/_react.default.createElement(_context.ContextSystemProvider, {
    value: ContextSystemProviderValue
  }, /*#__PURE__*/_react.default.createElement(_index.Grid, {
    gap: 28,
    templateColumns: "1fr 280px"
  }, /*#__PURE__*/_react.default.createElement(_index.View, null, /*#__PURE__*/_react.default.createElement(_index.TabPanel, null, /*#__PURE__*/_react.default.createElement(ProfilePanel, null)), /*#__PURE__*/_react.default.createElement(_index.TabPanel, null, /*#__PURE__*/_react.default.createElement(AccountPanel, null)), /*#__PURE__*/_react.default.createElement(_index.TabPanel, null, /*#__PURE__*/_react.default.createElement(OptionsPanel, null)), /*#__PURE__*/_react.default.createElement(_index.TabPanel, null, /*#__PURE__*/_react.default.createElement(PermissionsPanel, null))), /*#__PURE__*/_react.default.createElement(_index.View, null, /*#__PURE__*/_react.default.createElement(Sidebar, null)))))))));
}

function _default() {
  return /*#__PURE__*/_react.default.createElement(App, null);
}
