// @wp-g2/components/dist/cjs/SearchInput/SearchInputClearButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

var _utils = require("@wp-g2/utils");

var _react = _interopRequireDefault(require("react"));

var _Button = require("../Button");

var _View = require("../View");

function SearchInputClearButton(_ref) {
  var _ref$onClick = _ref.onClick,
      onClick = _ref$onClick === void 0 ? _utils.noop : _ref$onClick,
      value = _ref.value,
      otherProps = (0, _objectWithoutProperties2.default)(_ref, ["onClick", "value"]);
  var hasValue = !(0, _utils.isValueEmpty)(value);
  return /*#__PURE__*/_react.default.createElement(_View.View, {
    css: {
      opacity: hasValue ? 1 : 0,
      pointerEvents: hasValue ? 'cursor' : 'none'
    }
  }, /*#__PURE__*/_react.default.createElement(_Button.Button, Object.assign({
    "aria-label": (0, _i18n.__)('Clear'),
    icon: _icons.closeSmall,
    iconSize: 10,
    isRounded: true,
    onClick: function handleOnClick(event) {
      event.stopPropagation();
      onClick(event);
    },
    size: "xxSmall",
    tabIndex: -1
  }, otherProps)));
}

var _default = /*#__PURE__*/_react.default.memo(SearchInputClearButton);

exports.default = _default;
