// @wp-g2/components/dist/cjs/SearchInput/SearchInput.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _utils = require("@wp-g2/utils");

var _react = _interopRequireWildcard(require("react"));

var _Icon = require("../Icon");

var _Spinner = require("../Spinner");

var _Text = require("../Text");

var _TextInput = require("../TextInput");

var _View = require("../View");

var styles = _interopRequireWildcard(require("./SearchInput.styles"));

var _SearchInputClearButton = _interopRequireDefault(require("./SearchInputClearButton"));

function SearchInput(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'SearchInput'),
      className = _useContextSystem.className,
      _useContextSystem$def = _useContextSystem.defaultValue,
      defaultValue = _useContextSystem$def === void 0 ? '' : _useContextSystem$def,
      _useContextSystem$isL = _useContextSystem.isLoading,
      isLoading = _useContextSystem$isL === void 0 ? false : _useContextSystem$isL,
      _useContextSystem$pla = _useContextSystem.placeholder,
      placeholder = _useContextSystem$pla === void 0 ? (0, _i18n.__)('Search...') : _useContextSystem$pla,
      prefix = _useContextSystem.prefix,
      _useContextSystem$onC = _useContextSystem.onChange,
      onChangeProp = _useContextSystem$onC === void 0 ? _utils.noop : _useContextSystem$onC,
      _useContextSystem$onC2 = _useContextSystem.onClear,
      onClear = _useContextSystem$onC2 === void 0 ? _utils.noop : _useContextSystem$onC2,
      suffix = _useContextSystem.suffix,
      valueProp = _useContextSystem.value,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["className", "defaultValue", "isLoading", "placeholder", "prefix", "onChange", "onClear", "suffix", "value"]);

  var _useControlledValue = (0, _utils.useControlledValue)({
    value: valueProp,
    onChange: onChangeProp,
    defaultValue: defaultValue
  }),
      _useControlledValue2 = (0, _slicedToArray2.default)(_useControlledValue, 2),
      value = _useControlledValue2[0],
      onChange = _useControlledValue2[1];

  var textInputRef = (0, _react.useRef)();
  var handleOnChange = (0, _react.useCallback)(function (next, changeProps) {
    onChange(next, changeProps);
  }, [onChange]);
  var handleOnClearClick = (0, _react.useCallback)(function (event) {
    onChange('', {
      event: event
    });
    onClear(event);

    if (textInputRef.current) {
      textInputRef.current.focus();
    }
  }, [onChange, onClear]);
  var classes = (0, _styles.cx)(styles.SearchInput, className);
  return /*#__PURE__*/_react.default.createElement(_TextInput.TextInput, Object.assign({
    className: classes,
    isCommitOnBlurOrEnter: false,
    onChange: handleOnChange,
    placeholder: placeholder,
    prefix: /*#__PURE__*/_react.default.createElement(SearchPrefix, {
      isLoading: isLoading,
      prefix: prefix
    }),
    ref: (0, _utils.mergeRefs)([forwardedRef, textInputRef]),
    suffix: /*#__PURE__*/_react.default.createElement(SearchSuffix, {
      onClick: handleOnClearClick,
      suffix: suffix,
      value: !!value
    }),
    type: "search",
    value: value
  }, otherProps));
}

var SearchSuffix = /*#__PURE__*/_react.default.memo(function (_ref) {
  var onClick = _ref.onClick,
      suffix = _ref.suffix,
      value = _ref.value;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, suffix, /*#__PURE__*/_react.default.createElement(_SearchInputClearButton.default, {
    onClick: onClick,
    value: value ? true : undefined
  }));
});

var SearchPrefix = /*#__PURE__*/_react.default.memo(function (_ref2) {
  var _ref2$isLoading = _ref2.isLoading,
      isLoading = _ref2$isLoading === void 0 ? false : _ref2$isLoading,
      prefix = _ref2.prefix;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_View.View, {
    css: [_styles.ui.alignment.content.center, _styles.ui.opacity(0.5), _styles.ui.margin.right(-1)]
  }, isLoading ? /*#__PURE__*/_react.default.createElement(_Spinner.Spinner, {
    size: 16
  }) : /*#__PURE__*/_react.default.createElement(_Text.Text, null, /*#__PURE__*/_react.default.createElement(_Icon.Icon, {
    icon: _icons.search,
    size: 16
  }))), prefix);
});

var _default = (0, _context.contextConnect)(SearchInput, 'SearchInput');

exports.default = _default;
