// @wp-g2/components/dist/cjs/Popover/PopoverContent.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _reakit = require("reakit");

var _Card = require("../Card");

var _View = require("../View");

var _Popover = require("./Popover.Context");

var styles = _interopRequireWildcard(require("./Popover.styles"));

function PopoverContent(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'PopoverContent'),
      children = _useContextSystem.children,
      className = _useContextSystem.className,
      _useContextSystem$ele = _useContextSystem.elevation,
      elevation = _useContextSystem$ele === void 0 ? 5 : _useContextSystem$ele,
      _useContextSystem$max = _useContextSystem.maxWidth,
      maxWidth = _useContextSystem$max === void 0 ? 360 : _useContextSystem$max,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["children", "className", "elevation", "maxWidth"]);

  var _usePopoverContext = (0, _Popover.usePopoverContext)(),
      label = _usePopoverContext.label,
      popover = _usePopoverContext.popover;

  var classes = (0, _styles.cx)(styles.PopoverContent, (0, _styles.css)({
    maxWidth: maxWidth
  }), className);
  var showContent = popover.visible || popover.animating;
  return /*#__PURE__*/_react.default.createElement(_reakit.Popover, Object.assign({
    "aria-label": label,
    as: _View.View,
    className: classes
  }, otherProps, popover), showContent && /*#__PURE__*/_react.default.createElement(_Card.Card, {
    elevation: elevation,
    ref: forwardedRef
  }, children));
}

var _default = (0, _context.contextConnect)(PopoverContent, 'PopoverContent');

exports.default = _default;
