// @wp-g2/components/dist/cjs/Popover/Popover.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _utils = require("@wp-g2/utils");

var _react = _interopRequireWildcard(require("react"));

var _reakit = require("reakit");

var _Portal = require("../Portal");

var _Popover = require("./Popover.Context");

var _Popover2 = require("./Popover.utils");

var _PopoverContent = _interopRequireDefault(require("./PopoverContent"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function Popover(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'Popover'),
      _useContextSystem$ani = _useContextSystem.animated,
      animated = _useContextSystem$ani === void 0 ? true : _useContextSystem$ani,
      _useContextSystem$ani2 = _useContextSystem.animationDuration,
      animationDuration = _useContextSystem$ani2 === void 0 ? 160 : _useContextSystem$ani2,
      baseId = _useContextSystem.baseId,
      children = _useContextSystem.children,
      _useContextSystem$ele = _useContextSystem.elevation,
      elevation = _useContextSystem$ele === void 0 ? 5 : _useContextSystem$ele,
      id = _useContextSystem.id,
      label = _useContextSystem.label,
      _useContextSystem$max = _useContextSystem.maxWidth,
      maxWidth = _useContextSystem$max === void 0 ? 360 : _useContextSystem$max,
      _useContextSystem$onV = _useContextSystem.onVisibleChange,
      onVisibleChange = _useContextSystem$onV === void 0 ? _utils.noop : _useContextSystem$onV,
      placement = _useContextSystem.placement,
      state = _useContextSystem.state,
      trigger = _useContextSystem.trigger,
      visible = _useContextSystem.visible,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["animated", "animationDuration", "baseId", "children", "elevation", "id", "label", "maxWidth", "onVisibleChange", "placement", "state", "trigger", "visible"]);

  var _popover = (0, _reakit.usePopoverState)(_objectSpread({
    animated: animated ? animationDuration : undefined,
    baseId: baseId || id,
    placement: placement,
    visible: visible
  }, otherProps));

  var popover = state || _popover;
  var resizeListener = (0, _Popover2.usePopoverResizeUpdater)({
    onResize: popover.unstable_update
  });
  var uniqueId = "popover-" + popover.baseId;
  var labelId = label || uniqueId;

  var contextProps = _react.default.useMemo(function () {
    return {
      label: labelId,
      popover: popover
    };
  }, [labelId, popover]);

  var triggerContent = (0, _react.useCallback)(function (triggerProps) {
    return /*#__PURE__*/_react.default.cloneElement(trigger, triggerProps);
  }, [trigger]);
  (0, _utils.useUpdateEffect)(function () {
    onVisibleChange(popover.visible);
  }, [popover.visible]);
  return /*#__PURE__*/_react.default.createElement(_Popover.PopoverContext.Provider, {
    value: contextProps
  }, trigger && /*#__PURE__*/_react.default.createElement(_reakit.PopoverDisclosure, Object.assign({}, popover, {
    ref: trigger.ref
  }, trigger.props), triggerContent), /*#__PURE__*/_react.default.createElement(_Portal.Portal, null, /*#__PURE__*/_react.default.createElement(_PopoverContent.default, Object.assign({
    ref: forwardedRef
  }, otherProps, {
    elevation: elevation,
    maxWidth: maxWidth
  }), resizeListener, children)));
}

var _default = (0, _context.contextConnect)(Popover, 'Popover');

exports.default = _default;
