// @wordpress/components/build-module/ui/popover/component.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import { createElement } from "@wordpress/element";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * External dependencies
 */
import { contextConnect, useContextSystem } from '@wp-g2/context';
import { noop, useUpdateEffect } from '@wp-g2/utils'; // eslint-disable-next-line no-restricted-imports

import { PopoverDisclosure, usePopoverState } from 'reakit';
/**
 * WordPress dependencies
 */

import { useCallback, useMemo, cloneElement } from '@wordpress/element';
/**
 * Internal dependencies
 */

import { Portal } from '../portal';
import { PopoverContext } from './context';
import { usePopoverResizeUpdater } from './utils';
import PopoverContent from './content';
/**
 *
 * @param {import('@wp-g2/create-styles').ViewOwnProps<import('./types').Props, 'div'>} props
 * @param {import('react').Ref<any>} forwardedRef
 */

function Popover(props, forwardedRef) {
  var _useContextSystem = useContextSystem(props, 'Popover'),
      _useContextSystem$ani = _useContextSystem.animated,
      animated = _useContextSystem$ani === void 0 ? true : _useContextSystem$ani,
      _useContextSystem$ani2 = _useContextSystem.animationDuration,
      animationDuration = _useContextSystem$ani2 === void 0 ? 160 : _useContextSystem$ani2,
      baseId = _useContextSystem.baseId,
      children = _useContextSystem.children,
      _useContextSystem$ele = _useContextSystem.elevation,
      elevation = _useContextSystem$ele === void 0 ? 5 : _useContextSystem$ele,
      id = _useContextSystem.id,
      label = _useContextSystem.label,
      _useContextSystem$max = _useContextSystem.maxWidth,
      maxWidth = _useContextSystem$max === void 0 ? 360 : _useContextSystem$max,
      _useContextSystem$onV = _useContextSystem.onVisibleChange,
      onVisibleChange = _useContextSystem$onV === void 0 ? noop : _useContextSystem$onV,
      placement = _useContextSystem.placement,
      state = _useContextSystem.state,
      trigger = _useContextSystem.trigger,
      visible = _useContextSystem.visible,
      otherProps = _objectWithoutProperties(_useContextSystem, ["animated", "animationDuration", "baseId", "children", "elevation", "id", "label", "maxWidth", "onVisibleChange", "placement", "state", "trigger", "visible"]);

  var _popover = usePopoverState(_objectSpread({
    animated: animated ? animationDuration : undefined,
    baseId: baseId || id,
    placement: placement,
    visible: visible
  }, otherProps));

  var popover = state || _popover;
  var resizeListener = usePopoverResizeUpdater({
    onResize: popover.unstable_update
  });
  var uniqueId = "popover-".concat(popover.baseId);
  var labelId = label || uniqueId;
  var contextProps = useMemo(function () {
    return {
      label: labelId,
      popover: popover
    };
  }, [labelId, popover]);
  var triggerContent = useCallback(function (triggerProps) {
    return cloneElement(trigger, triggerProps);
  }, [trigger]);
  useUpdateEffect(function () {
    onVisibleChange(popover.visible);
  }, [popover.visible]);
  return createElement(PopoverContext.Provider, {
    value: contextProps
  }, trigger && createElement(PopoverDisclosure, _extends({}, popover, {
    ref: trigger.ref
  }, trigger.props), triggerContent), createElement(Portal, null, createElement(PopoverContent, _extends({
    ref: forwardedRef
  }, otherProps, {
    elevation: elevation,
    maxWidth: maxWidth
  }), resizeListener, children)));
}
/**
 * `Popover` is a component to render a floating content modal.
 * It is similar in purpose to a tooltip, but renders content of any sort,
 * not only simple text.
 *
 * @example
 * ```jsx
 * import { Button, Popover, View, Text } from `@wordpress/components/ui`;
 
 * function Example() {
 * 	return (
 * 		<Popover trigger={ <Button>Popover</Button> }>
 * 			<View>
 * 				<Text>Code is Poetry</Text>
 * 			</View>
 * 		</Popover>
 * 	);
 * }
 * ```
 */


export default contextConnect(Popover, 'Popover');
