// @wp-g2/components/dist/es/ColorPicker/__stories__/ColorPicker.stories.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { ui } from '@wp-g2/styles';
import React from 'react';
import tinyColor from 'tinycolor2';
import { Container, Grid, Heading, VStack } from '../../index';
import { ColorPicker } from '../index';
export default {
  component: ColorPicker,
  title: 'Components/ColorPicker'
};

var Example = function () {
  var _React$useState = React.useState('red'),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      color = _React$useState2[0],
      setColor = _React$useState2[1];

  return /*#__PURE__*/React.createElement(Container, null, /*#__PURE__*/React.createElement("p", null, /*#__PURE__*/React.createElement("strong", null, color)), /*#__PURE__*/React.createElement(VStack, {
    spacing: 10
  }, /*#__PURE__*/React.createElement(Grid, {
    gap: 8
  }, /*#__PURE__*/React.createElement(ColorPicker, {
    color: color,
    disableAlpha: false,
    inputType: "rgb",
    onChange: function handleOnChange(next) {
      setColor(next);
    },
    width: 300
  }), /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("button", {
    onClick: function onClick() {
      return setColor(tinyColor.random().toRgbString());
    }
  }, "Set Random Color")))));
};

export var _default = function () {
  return /*#__PURE__*/React.createElement(Example, null);
};
export var BasicTest = function () {
  var _React$useState3 = React.useState('red'),
      _React$useState4 = _slicedToArray(_React$useState3, 2),
      color = _React$useState4[0],
      setColor = _React$useState4[1];

  return /*#__PURE__*/React.createElement(Container, null, /*#__PURE__*/React.createElement("p", null, /*#__PURE__*/React.createElement("strong", null, color)), /*#__PURE__*/React.createElement(VStack, {
    spacing: 10
  }, /*#__PURE__*/React.createElement(Grid, {
    gap: 8
  }, /*#__PURE__*/React.createElement(ColorPicker, {
    color: color,
    onChange: function handleOnChange(next) {
      setColor(next);
    },
    width: 300
  }), /*#__PURE__*/React.createElement("input", {
    onChange: function onChange(e) {
      return setColor(e.target.value);
    },
    type: "color",
    value: ui.color(color).toHexString()
  }))));
};
export var HTMLPickerTest = function () {
  var _React$useState5 = React.useState('#ff0000'),
      _React$useState6 = _slicedToArray(_React$useState5, 2),
      color = _React$useState6[0],
      setColor = _React$useState6[1];

  return /*#__PURE__*/React.createElement(VStack, null, /*#__PURE__*/React.createElement(VStack, null, /*#__PURE__*/React.createElement(VStack, null, /*#__PURE__*/React.createElement(VStack, null, /*#__PURE__*/React.createElement(VStack, null, /*#__PURE__*/React.createElement(Heading, {
    size: 1
  }, /*#__PURE__*/React.createElement("span", {
    "aria-label": "wave",
    role: "img"
  }, "\uD83D\uDC4B"), ' ', color), /*#__PURE__*/React.createElement("input", {
    onChange: function onChange(e) {
      return setColor(e.target.value);
    },
    type: "color"
  }))))));
};
