// @wordpress/components/build/ui/__storybook-utils/example-grid.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ExampleMetaContent = exports.ExampleGridItem = exports.ExampleGrid = void 0;

var _element = require("@wordpress/element");

var _components = require("@wp-g2/components");

var _grid = require("../grid");

var _surface = require("../surface");

var _text = require("../text");

var _view = require("../view");

var _vStack = require("../v-stack");

/**
 * External dependencies
 */

/**
 * Internal dependencies
 */
var ExampleGrid = function ExampleGrid(_ref) {
  var children = _ref.children;
  return (0, _element.createElement)(_components.Container, null, (0, _element.createElement)(_grid.Grid, {
    columns: [2, 2, 4],
    gap: "1px"
  }, children));
};

exports.ExampleGrid = ExampleGrid;

var ExampleGridItem = function ExampleGridItem(_ref2) {
  var children = _ref2.children;
  return (0, _element.createElement)(_surface.Surface, {
    variant: "secondary",
    css: {
      padding: 20
    }
  }, (0, _element.createElement)(_vStack.VStack, {
    alignment: "center"
  }, children));
};

exports.ExampleGridItem = ExampleGridItem;

var ExampleMetaContent = function ExampleMetaContent(_ref3) {
  var title = _ref3.title,
      _ref3$items = _ref3.items,
      items = _ref3$items === void 0 ? [] : _ref3$items;
  return (0, _element.createElement)(_grid.Grid, {
    templateColumns: "60px 1fr"
  }, (0, _element.createElement)(_view.View, null, (0, _element.createElement)(_text.Text, {
    size: "caption",
    weight: "bold",
    align: "right",
    isBlock: true
  }, title)), (0, _element.createElement)(_vStack.VStack, {
    spacing: 1
  }, items.map(function (item, index) {
    return (0, _element.createElement)(_view.View, {
      key: index
    }, (0, _element.createElement)(_text.Text, {
      size: "caption",
      isBlock: true
    }, item));
  })));
};

exports.ExampleMetaContent = ExampleMetaContent;
