// @wp-g2/styles/dist/cjs/presets/alignments.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.alignment = void 0;

var _styleSystem = require("../style-system");

var content = {
  bottom: (0, _styleSystem.css)({
    alignItems: 'flex-end',
    display: 'flex',
    justifyContent: 'center'
  }),
  center: (0, _styleSystem.css)({
    alignItems: 'center',
    display: 'flex',
    justifyContent: 'center'
  }),
  left: (0, _styleSystem.css)({
    alignItems: 'center',
    display: 'flex',
    justifyContent: 'flex-start'
  }),
  right: (0, _styleSystem.css)({
    alignItems: 'center',
    display: 'flex',
    justifyContent: 'flex-end'
  }),
  top: (0, _styleSystem.css)({
    alignItems: 'flex-start',
    display: 'flex',
    justifyContent: 'center'
  })
};
var alignment = {
  center: (0, _styleSystem.css)({
    margin: 'auto'
  }),
  content: content,
  left: (0, _styleSystem.css)({
    marginRight: 'auto'
  }),
  right: (0, _styleSystem.css)({
    marginLeft: 'auto'
  })
};
exports.alignment = alignment;
