// @wp-g2/styles/dist/cjs/presets/borders.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.borderRadius = borderRadius;
exports.border = void 0;

var _core = require("../core");

var _styleSystem = require("../style-system");

var control = {
  default: (0, _styleSystem.css)({
    border: "1px solid " + (0, _core.get)('controlBorderColor')
  }),
  focus: (0, _styleSystem.css)({
    border: "1px solid " + (0, _core.get)('colorAdmin'),
    boxShadow: (0, _core.get)('controlBoxShadowFocus')
  }),
  subtle: (0, _styleSystem.css)({
    border: "1px solid " + (0, _core.get)('controlBorderSubtleColor')
  })
};
var border = {
  all: (0, _styleSystem.css)({
    border: "1px solid " + (0, _core.get)('surfaceBorderColor')
  }),
  bottom: (0, _styleSystem.css)({
    borderBottom: "1px solid " + (0, _core.get)('surfaceBorderColor')
  }),
  control: control,
  left: (0, _styleSystem.css)({
    borderLeft: "1px solid " + (0, _core.get)('surfaceBorderColor')
  }),
  right: (0, _styleSystem.css)({
    borderRight: "1px solid " + (0, _core.get)('surfaceBorderColor')
  }),
  top: (0, _styleSystem.css)({
    borderTop: "1px solid " + (0, _core.get)('surfaceBorderColor')
  })
};
/**
 *
 * @param {import('create-emotion').ObjectInterpolation<any>['borderRadius']} value
 */

exports.border = border;

function borderRadius(value) {
  return (0, _styleSystem.css)({
    borderRadius: value
  });
}

borderRadius.none = borderRadius(0);
borderRadius.round = borderRadius((0, _core.get)('controlBorderRadius'));
borderRadius.circle = borderRadius(99999);
