// @wp-g2/styles/dist/cjs/presets/positions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.position = void 0;

var _styleSystem = require("../style-system");

var position = {
  absolute: (0, _styleSystem.css)({
    position: 'absolute'
  }),
  fixed: (0, _styleSystem.css)({
    position: 'fixed'
  }),
  bottom: (0, _styleSystem.css)({
    bottom: 0,
    position: 'absolute'
  }),
  bottomLeft: (0, _styleSystem.css)({
    bottom: 0,
    left: 0,
    position: 'absolute'
  }),
  bottomRight: (0, _styleSystem.css)({
    bottom: 0,
    position: 'absolute',
    right: 0
  }),
  full: (0, _styleSystem.css)({
    bottom: 0,
    left: 0,
    position: 'absolute',
    right: 0,
    top: 0
  }),
  left: (0, _styleSystem.css)({
    left: 0,
    position: 'absolute'
  }),
  relative: (0, _styleSystem.css)({
    position: 'relative'
  }),
  right: (0, _styleSystem.css)({
    position: 'absolute',
    right: 0
  }),
  sticky: (0, _styleSystem.css)({
    position: 'sticky',
    top: 0
  }),
  top: (0, _styleSystem.css)({
    position: 'absolute',
    top: 0
  }),
  topLeft: (0, _styleSystem.css)({
    left: 0,
    position: 'absolute',
    top: 0
  }),
  topRight: (0, _styleSystem.css)({
    position: 'absolute',
    right: 0,
    top: 0
  })
};
exports.position = position;
