// @wp-g2/styles/dist/cjs/presets/ui.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ui = void 0;

var _alignments = require("./alignments");

var _animations = require("./animations");

var _backgrounds = require("./backgrounds");

var _borders = require("./borders");

var _browsers = require("./browsers");

var _colors = require("./colors");

var _css = require("./css");

var _dimensions = require("./dimensions");

var _flow = require("./flow");

var _fonts = require("./fonts");

var _get = require("./get");

var _interactions = require("./interactions");

var _modes = require("./modes");

var _opacity = require("./opacity");

var _positions = require("./positions");

var _shadows = require("./shadows");

var _spacing = require("./spacing");

var _styleQuery = require("./style-query");

var _transforms = require("./transforms");

var _values = require("./values");

var _zIndex = require("./z-index");

var ui = {
  $: _styleQuery.$,
  active: _interactions.active,
  alignment: _alignments.alignment,
  animation: _animations.animation,
  background: _backgrounds.background,
  border: _borders.border,
  borderRadius: _borders.borderRadius,
  browsers: _browsers.browsers,
  color: _colors.color,
  createToken: _get.createToken,
  css: _css.css,
  flow: _flow.flow,
  focus: _interactions.focus,
  font: _fonts.font,
  frame: _dimensions.frame,
  get: _get.get,
  getTokenValue: _get.getTokenValue,
  hover: _interactions.hover,
  margin: _spacing.margin,
  mode: _modes.mode,
  offset: _transforms.offset,
  opacity: _opacity.opacity,
  padding: _spacing.padding,
  position: _positions.position,
  rotate: _transforms.rotate,
  scale: _transforms.scale,
  scaleX: _transforms.scaleX,
  scaleY: _transforms.scaleY,
  shadow: _shadows.shadow,
  space: _spacing.space,
  value: _values.value,
  zIndex: _zIndex.zIndex
};
exports.ui = ui;
