// @wordpress/block-editor/build-module/components/inserter-list-item/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import classnames from 'classnames';
/**
 * WordPress dependencies
 */

import { useMemo, useRef, memo } from '@wordpress/element';
import { createBlock, createBlocksFromInnerBlocksTemplate } from '@wordpress/blocks';
import { ENTER } from '@wordpress/keycodes';
/**
 * Internal dependencies
 */

import BlockIcon from '../block-icon';
import { InserterListboxItem } from '../inserter-listbox';
import InserterDraggableBlocks from '../inserter-draggable-blocks';
/**
 * Return true if platform is MacOS.
 *
 * @param {Object} _window   window object by default; used for DI testing.
 *
 * @return {boolean} True if MacOS; false otherwise.
 */

function isAppleOS() {
  var _window = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : window;

  var platform = _window.navigator.platform;
  return platform.indexOf('Mac') !== -1 || ['iPad', 'iPhone'].includes(platform);
}

function InserterListItem(_ref) {
  var className = _ref.className,
      isFirst = _ref.isFirst,
      item = _ref.item,
      onSelect = _ref.onSelect,
      onHover = _ref.onHover,
      isDraggable = _ref.isDraggable,
      props = _objectWithoutProperties(_ref, ["className", "isFirst", "item", "onSelect", "onHover", "isDraggable"]);

  var isDragging = useRef(false);
  var itemIconStyle = item.icon ? {
    backgroundColor: item.icon.background,
    color: item.icon.foreground
  } : {};
  var blocks = useMemo(function () {
    return [createBlock(item.name, item.initialAttributes, createBlocksFromInnerBlocksTemplate(item.innerBlocks))];
  }, [item.name, item.initialAttributes, item.initialAttributes]);
  return createElement(InserterDraggableBlocks, {
    isEnabled: isDraggable && !item.disabled,
    blocks: blocks,
    icon: item.icon
  }, function (_ref2) {
    var draggable = _ref2.draggable,
        _onDragStart = _ref2.onDragStart,
        _onDragEnd = _ref2.onDragEnd;
    return createElement("div", {
      className: "block-editor-block-types-list__list-item",
      draggable: draggable,
      onDragStart: function onDragStart(event) {
        isDragging.current = true;

        if (_onDragStart) {
          onHover(null);

          _onDragStart(event);
        }
      },
      onDragEnd: function onDragEnd(event) {
        isDragging.current = false;

        if (_onDragEnd) {
          _onDragEnd(event);
        }
      }
    }, createElement(InserterListboxItem, _extends({
      isFirst: isFirst,
      className: classnames('block-editor-block-types-list__item', className),
      disabled: item.isDisabled,
      onClick: function onClick(event) {
        event.preventDefault();
        onSelect(item, isAppleOS() ? event.metaKey : event.ctrlKey);
        onHover(null);
      },
      onKeyDown: function onKeyDown(event) {
        var keyCode = event.keyCode;

        if (keyCode === ENTER) {
          event.preventDefault();
          onSelect(item, isAppleOS() ? event.metaKey : event.ctrlKey);
          onHover(null);
        }
      },
      onFocus: function onFocus() {
        if (isDragging.current) {
          return;
        }

        onHover(item);
      },
      onMouseEnter: function onMouseEnter() {
        if (isDragging.current) {
          return;
        }

        onHover(item);
      },
      onMouseLeave: function onMouseLeave() {
        return onHover(null);
      },
      onBlur: function onBlur() {
        return onHover(null);
      }
    }, props), createElement("span", {
      className: "block-editor-block-types-list__item-icon",
      style: itemIconStyle
    }, createElement(BlockIcon, {
      icon: item.icon,
      showColors: true
    })), createElement("span", {
      className: "block-editor-block-types-list__item-title"
    }, item.title)));
  });
}

export default memo(InserterListItem);
