// @wordpress/components/build-module/mobile/link-settings/index.native.js
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement, Fragment } from "@wordpress/element";
import _regeneratorRuntime from "@babel/runtime/regenerator";

/**
 * External dependencies
 */
import { Platform, Clipboard } from 'react-native';
/**
 * WordPress dependencies
 */

import { compose } from '@wordpress/compose';
import { withSelect } from '@wordpress/data';
import { isURL, prependHTTP } from '@wordpress/url';
import { useEffect, useState, useRef, useContext, useCallback } from '@wordpress/element';
import { link, external } from '@wordpress/icons';
/**
 * Internal dependencies
 */

import BottomSheet from '../bottom-sheet';
import { BottomSheetContext } from '../bottom-sheet/bottom-sheet-context';
import PanelBody from '../../panel/body';
import TextControl from '../../text-control';
import ToggleControl from '../../toggle-control';
import FooterMessageControl from '../../footer-message-control';
import PanelActions from '../../panel/actions';
import LinkRelIcon from './link-rel';
import styles from './style.scss';
var NEW_TAB_REL = 'noreferrer noopener';

function LinkSettings(_ref) {
  var isVisible = _ref.isVisible,
      onClose = _ref.onClose,
      setAttributes = _ref.setAttributes,
      onEmptyURL = _ref.onEmptyURL,
      options = _ref.options,
      withBottomSheet = _ref.withBottomSheet,
      actions = _ref.actions,
      editorSidebarOpened = _ref.editorSidebarOpened,
      showIcon = _ref.showIcon,
      onLinkCellPressed = _ref.onLinkCellPressed,
      urlValue = _ref.urlValue,
      url = _ref.url,
      label = _ref.label,
      linkTarget = _ref.linkTarget,
      rel = _ref.rel;

  var _useState = useState(''),
      _useState2 = _slicedToArray(_useState, 2),
      urlInputValue = _useState2[0],
      setUrlInputValue = _useState2[1];

  var _useState3 = useState(''),
      _useState4 = _slicedToArray(_useState3, 2),
      labelInputValue = _useState4[0],
      setLabelInputValue = _useState4[1];

  var _useState5 = useState(''),
      _useState6 = _slicedToArray(_useState5, 2),
      linkRelInputValue = _useState6[0],
      setLinkRelInputValue = _useState6[1];

  var prevEditorSidebarOpenedRef = useRef();

  var _useContext = useContext(BottomSheetContext),
      onHandleClosingBottomSheet = _useContext.onHandleClosingBottomSheet;

  useEffect(function () {
    if (onHandleClosingBottomSheet) {
      onHandleClosingBottomSheet(onCloseSettingsSheet);
    }
  }, [urlInputValue, labelInputValue, linkRelInputValue]);
  useEffect(function () {
    prevEditorSidebarOpenedRef.current = editorSidebarOpened;
  });
  var prevEditorSidebarOpened = prevEditorSidebarOpenedRef.current;
  useEffect(function () {
    if (url !== urlInputValue) {
      setUrlInputValue(url || '');
    }
  }, [url]);
  useEffect(function () {
    setLabelInputValue(label || '');
  }, [label]);
  useEffect(function () {
    setLinkRelInputValue(rel || '');
  }, [rel]);
  useEffect(function () {
    var isSettingSheetOpen = isVisible || editorSidebarOpened;

    if (options.url.autoFill && isSettingSheetOpen && !url) {
      getURLFromClipboard();
    }

    if (prevEditorSidebarOpened && !editorSidebarOpened) {
      onSetAttributes();
    }
  }, [editorSidebarOpened, isVisible]);
  useEffect(function () {
    if (!urlValue && onEmptyURL) {
      onEmptyURL();
    }

    if (prependHTTP(urlValue) !== url) {
      setAttributes({
        url: prependHTTP(urlValue)
      });
    }
  }, [urlValue]);
  var onChangeURL = useCallback(function (value) {
    if (!value && onEmptyURL) {
      onEmptyURL();
    }

    setUrlInputValue(value);
  }, [onEmptyURL]);
  var onChangeLabel = useCallback(function (value) {
    setLabelInputValue(value);
  }, []);
  var onSetAttributes = useCallback(function () {
    var newURL = prependHTTP(urlInputValue);

    if (url !== newURL || labelInputValue !== label || linkRelInputValue !== rel) {
      setAttributes({
        url: newURL,
        label: labelInputValue,
        rel: linkRelInputValue
      });
    }
  }, [urlInputValue, labelInputValue, linkRelInputValue, setAttributes]);
  var onCloseSettingsSheet = useCallback(function () {
    onSetAttributes();

    if (onClose) {
      onClose();
    }
  }, [onClose, onSetAttributes]);
  var onChangeOpenInNewTab = useCallback(function (value) {
    var newLinkTarget = value ? '_blank' : undefined;
    var updatedRel = linkRelInputValue;

    if (newLinkTarget && !linkRelInputValue) {
      updatedRel = NEW_TAB_REL;
    } else if (!newLinkTarget && linkRelInputValue === NEW_TAB_REL) {
      updatedRel = undefined;
    }

    setAttributes({
      linkTarget: newLinkTarget,
      rel: updatedRel
    });
  }, [linkRelInputValue]);
  var onChangeLinkRel = useCallback(function (value) {
    setLinkRelInputValue(value);
  }, []);

  function getURLFromClipboard() {
    return _getURLFromClipboard.apply(this, arguments);
  }

  function _getURLFromClipboard() {
    _getURLFromClipboard = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
      var clipboardText;
      return _regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              _context.next = 2;
              return Clipboard.getString();

            case 2:
              clipboardText = _context.sent;

              if (clipboardText) {
                _context.next = 5;
                break;
              }

              return _context.abrupt("return");

            case 5:
              if (isURL(clipboardText)) {
                _context.next = 7;
                break;
              }

              return _context.abrupt("return");

            case 7:
              setAttributes({
                url: clipboardText
              });

            case 8:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }));
    return _getURLFromClipboard.apply(this, arguments);
  }

  function getSettings() {
    return createElement(Fragment, null, options.url && (onLinkCellPressed ? createElement(BottomSheet.LinkCell, {
      showIcon: showIcon,
      value: url,
      onPress: onLinkCellPressed
    }) : createElement(TextControl, {
      icon: showIcon && link,
      label: options.url.label,
      value: urlInputValue,
      valuePlaceholder: options.url.placeholder,
      onChange: onChangeURL,
      onSubmit: onCloseSettingsSheet,
      autoCapitalize: "none",
      autoCorrect: false // eslint-disable-next-line jsx-a11y/no-autofocus
      ,
      autoFocus: Platform.OS === 'ios' && options.url.autoFocus,
      keyboardType: "url"
    })), options.linkLabel && createElement(TextControl, {
      label: options.linkLabel.label,
      value: labelInputValue,
      valuePlaceholder: options.linkLabel.placeholder,
      onChange: onChangeLabel
    }), !!urlInputValue && createElement(Fragment, null, options.openInNewTab && createElement(ToggleControl, {
      icon: showIcon && external,
      label: options.openInNewTab.label,
      checked: linkTarget === '_blank',
      onChange: onChangeOpenInNewTab
    }), options.linkRel && createElement(TextControl, {
      icon: showIcon && LinkRelIcon,
      label: options.linkRel.label,
      value: linkRelInputValue,
      valuePlaceholder: options.linkRel.placeholder,
      onChange: onChangeLinkRel,
      onSubmit: onCloseSettingsSheet,
      autoCapitalize: "none",
      autoCorrect: false,
      keyboardType: "default"
    })));
  }

  if (!withBottomSheet) {
    return getSettings();
  }

  return createElement(Fragment, null, createElement(PanelBody, {
    style: styles.linkSettingsPanel
  }, getSettings()), options.footer && createElement(PanelBody, {
    style: styles.linkSettingsPanel
  }, createElement(FooterMessageControl, {
    label: options.footer.label,
    textAlign: "left"
  })), actions && createElement(PanelActions, {
    actions: actions
  }));
}

export default compose([withSelect(function (select) {
  var _select = select('core/edit-post'),
      isEditorSidebarOpened = _select.isEditorSidebarOpened;

  return {
    editorSidebarOpened: isEditorSidebarOpened()
  };
})])(LinkSettings);
