// @wp-g2/components/dist/cjs/ControlGroup/ControlGroup.styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.itemGrid = exports.itemFocus = exports.lastRow = exports.firstRow = exports.last = exports.middle = exports.first = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _styles = require("@wp-g2/styles");

function _templateObject7() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t> * + *:not(marquee) {\n\t\tmargin-left: -1px;\n\t\twidth: calc(100% + 1px);\n\t}\n"]);

  _templateObject7 = function () {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t> * {\n\t\t&:focus-within {\n\t\t\tz-index: 1;\n\t\t}\n\t}\n"]);

  _templateObject6 = function () {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tborder-top-left-radius: 0;\n\tborder-top-right-radius: 0;\n"]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tborder-bottom-left-radius: 0;\n\tborder-bottom-right-radius: 0;\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tborder-bottom-left-radius: 0;\n\tborder-top-left-radius: 0;\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tborder-radius: 0;\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tborder-bottom-right-radius: 0;\n\tborder-top-right-radius: 0;\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var first = (0, _styles.css)(_templateObject());
exports.first = first;
var middle = (0, _styles.css)(_templateObject2());
exports.middle = middle;
var last = (0, _styles.css)(_templateObject3());
exports.last = last;
var firstRow = (0, _styles.css)(_templateObject4());
exports.firstRow = firstRow;
var lastRow = (0, _styles.css)(_templateObject5());
exports.lastRow = lastRow;
var itemFocus = (0, _styles.css)(_templateObject6());
exports.itemFocus = itemFocus;
var itemGrid = (0, _styles.css)(_templateObject7());
exports.itemGrid = itemGrid;
