// @wp-g2/components/dist/cjs/ControlGroup/ControlGroup.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _react = _interopRequireDefault(require("react"));

var _Flex = require("../Flex");

var _Grid = require("../Grid");

var _useControlGroup2 = require("./useControlGroup");

/**
 * @param {import('@wp-g2/create-styles').ViewOwnProps<import('./types').Props, 'div'>} props
 * @param {import('react').Ref<any>} forwardedRef
 */
function ControlGroup(props, forwardedRef) {
  var _useControlGroup = (0, _useControlGroup2.useControlGroup)(props),
      children = _useControlGroup.children,
      _useControlGroup$dire = _useControlGroup.direction,
      direction = _useControlGroup$dire === void 0 ? 'row' : _useControlGroup$dire,
      templateColumns = _useControlGroup.templateColumns,
      otherProps = (0, _objectWithoutProperties2.default)(_useControlGroup, ["children", "direction", "templateColumns"]);

  if (!!templateColumns) {
    return /*#__PURE__*/_react.default.createElement(_Grid.Grid, Object.assign({
      gap: 0,
      templateColumns: templateColumns
    }, otherProps, {
      ref: forwardedRef
    }), children);
  }

  return /*#__PURE__*/_react.default.createElement(_Flex.Flex, Object.assign({
    direction: direction,
    gap: "-1px"
  }, otherProps, {
    ref: forwardedRef
  }), children);
}
/**
 * `ControlGroup` is a layout component that contains control elements (e.g. `TextInput` or `Select`).
 *
 * @example
 * ```jsx
 * import { Button, ControlGroup, TextInput } from `@wp-g2/components`
 * import { ui } from `@wp-g2/styles`
 *
 * function Example() {
 *   return (
 *     <ControlGroup>
 *       <TextInput placeholder="First name" />
 *       <Button variant="primary" />
 *     </ControlGroup>
 *   );
 * }
 * ```
 */


var ConnectedControlGroup = (0, _context.contextConnect)(ControlGroup, 'ControlGroup');
var _default = ConnectedControlGroup;
exports.default = _default;
