// @wordpress/components/build-module/mobile/bottom-sheet/stepper-cell/index.native.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import { createElement } from "@wordpress/element";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

/**
 * External dependencies
 */
import { AccessibilityInfo, View, Platform } from 'react-native';
/**
 * WordPress dependencies
 */

import { __, sprintf } from '@wordpress/i18n';
import { Component } from '@wordpress/element';
import { withPreferredColorScheme } from '@wordpress/compose';
/**
 * Internal dependencies
 */

import Cell from '../cell';
import Stepper from './stepper';
import styles from './style.scss';
import RangeTextInput from '../range-text-input';
import { toFixed } from '../../utils';
var STEP_DELAY = 200;
var DEFAULT_STEP = 1;
var isIOS = Platform.OS === 'ios';

var BottomSheetStepperCell = /*#__PURE__*/function (_Component) {
  _inherits(BottomSheetStepperCell, _Component);

  var _super = _createSuper(BottomSheetStepperCell);

  function BottomSheetStepperCell(props) {
    var _this;

    _classCallCheck(this, BottomSheetStepperCell);

    _this = _super.call(this, props);
    _this.announceValue = _this.announceValue.bind(_assertThisInitialized(_this));
    _this.onDecrementValue = _this.onDecrementValue.bind(_assertThisInitialized(_this));
    _this.onDecrementValuePressIn = _this.onDecrementValuePressIn.bind(_assertThisInitialized(_this));
    _this.onIncrementValue = _this.onIncrementValue.bind(_assertThisInitialized(_this));
    _this.onIncrementValuePressIn = _this.onIncrementValuePressIn.bind(_assertThisInitialized(_this));
    _this.onPressOut = _this.onPressOut.bind(_assertThisInitialized(_this));
    var value = props.value,
        defaultValue = props.defaultValue,
        min = props.min;
    var initialValue = value || defaultValue || min;
    _this.state = {
      inputValue: initialValue,
      stepperValue: initialValue
    };
    return _this;
  }

  _createClass(BottomSheetStepperCell, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      clearTimeout(this.timeout);
      clearInterval(this.interval);
      clearTimeout(this.timeoutAnnounceValue);
    }
  }, {
    key: "onIncrementValue",
    value: function onIncrementValue() {
      var _this$props = this.props,
          step = _this$props.step,
          max = _this$props.max,
          onChange = _this$props.onChange,
          value = _this$props.value,
          decimalNum = _this$props.decimalNum;
      var newValue = toFixed(value + step, decimalNum);

      if (newValue <= max || max === undefined) {
        onChange(newValue);
        this.setState({
          inputValue: newValue
        });
        this.announceValue(newValue);
      }
    }
  }, {
    key: "onDecrementValue",
    value: function onDecrementValue() {
      var _this$props2 = this.props,
          step = _this$props2.step,
          min = _this$props2.min,
          onChange = _this$props2.onChange,
          value = _this$props2.value,
          decimalNum = _this$props2.decimalNum;
      var newValue = toFixed(value - step, decimalNum);

      if (newValue >= min) {
        onChange(newValue);
        this.setState({
          inputValue: newValue
        });
        this.announceValue(newValue);
      }
    }
  }, {
    key: "onIncrementValuePressIn",
    value: function onIncrementValuePressIn() {
      var _this2 = this;

      this.onIncrementValue();
      this.timeout = setTimeout(function () {
        _this2.startPressInterval(_this2.onIncrementValue);
      }, 500);
    }
  }, {
    key: "onDecrementValuePressIn",
    value: function onDecrementValuePressIn() {
      var _this3 = this;

      this.onDecrementValue();
      this.timeout = setTimeout(function () {
        _this3.startPressInterval(_this3.onDecrementValue);
      }, 500);
    }
  }, {
    key: "onPressOut",
    value: function onPressOut() {
      clearTimeout(this.timeout);
      clearInterval(this.interval);
    }
  }, {
    key: "startPressInterval",
    value: function startPressInterval(callback) {
      var _this4 = this;

      var speed = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : STEP_DELAY;
      var counter = 0;
      this.interval = setInterval(function () {
        callback();
        counter += 1;

        if (counter === 10) {
          clearInterval(_this4.interval);

          _this4.startPressInterval(callback, speed / 2);
        }
      }, speed);
    }
  }, {
    key: "announceValue",
    value: function announceValue(value) {
      var _this$props3 = this.props,
          label = _this$props3.label,
          _this$props3$unitLabe = _this$props3.unitLabel,
          unitLabel = _this$props3$unitLabe === void 0 ? '' : _this$props3$unitLabe;

      if (isIOS) {
        // On Android it triggers the accessibilityLabel with the value change
        clearTimeout(this.timeoutAnnounceValue);
        this.timeoutAnnounceValue = setTimeout(function () {
          AccessibilityInfo.announceForAccessibility("".concat(value, " ").concat(unitLabel, " ").concat(label));
        }, 300);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this5 = this;

      var _this$props4 = this.props,
          label = _this$props4.label,
          _this$props4$settingL = _this$props4.settingLabel,
          settingLabel = _this$props4$settingL === void 0 ? 'Value' : _this$props4$settingL,
          _this$props4$unitLabe = _this$props4.unitLabel,
          unitLabel = _this$props4$unitLabe === void 0 ? '' : _this$props4$unitLabe,
          icon = _this$props4.icon,
          min = _this$props4.min,
          max = _this$props4.max,
          value = _this$props4.value,
          separatorType = _this$props4.separatorType,
          children = _this$props4.children,
          _this$props4$shouldDi = _this$props4.shouldDisplayTextInput,
          shouldDisplayTextInput = _this$props4$shouldDi === void 0 ? false : _this$props4$shouldDi,
          preview = _this$props4.preview,
          onChange = _this$props4.onChange,
          openUnitPicker = _this$props4.openUnitPicker,
          decimalNum = _this$props4.decimalNum,
          cellContainerStyle = _this$props4.cellContainerStyle;
      var inputValue = this.state.inputValue;
      var isMinValue = value === min;
      var isMaxValue = value === max;
      var labelStyle = [styles.cellLabel, !icon ? styles.cellLabelNoIcon : {}];

      var getAccessibilityHint = function getAccessibilityHint() {
        return openUnitPicker ? __('double-tap to change unit') : '';
      };

      var accessibilityLabel = sprintf(
      /* translators: accessibility text. Inform about current value. %1$s: Control label %2$s: setting label (example: width), %3$s: Current value. %4$s: value measurement unit (example: pixels) */
      __('%1$s. %2$s is %3$s %4$s.'), label, settingLabel, value, unitLabel);
      var containerStyle = [styles.rowContainer, isIOS ? styles.containerIOS : styles.containerAndroid];
      return createElement(View, {
        accessible: true,
        accessibilityRole: "adjustable",
        accessibilityLabel: accessibilityLabel,
        accessibilityHint: getAccessibilityHint(),
        accessibilityActions: [{
          name: 'increment'
        }, {
          name: 'decrement'
        }, {
          name: 'activate'
        }],
        onAccessibilityAction: function onAccessibilityAction(event) {
          switch (event.nativeEvent.actionName) {
            case 'increment':
              _this5.onIncrementValue();

              break;

            case 'decrement':
              _this5.onDecrementValue();

              break;

            case 'activate':
              if (openUnitPicker) {
                openUnitPicker();
              }

              break;
          }
        }
      }, createElement(View, {
        importantForAccessibility: "no-hide-descendants"
      }, createElement(Cell, {
        accessible: false,
        cellContainerStyle: [styles.cellContainerStyle, preview && styles.columnContainer, cellContainerStyle],
        cellRowContainerStyle: preview ? containerStyle : styles.cellRowStyles,
        disabled: true,
        editable: false,
        icon: icon,
        label: label,
        labelStyle: labelStyle,
        leftAlign: true,
        separatorType: separatorType
      }, createElement(View, {
        style: preview && containerStyle
      }, preview, createElement(Stepper, {
        isMaxValue: isMaxValue,
        isMinValue: isMinValue,
        onPressInDecrement: this.onDecrementValuePressIn,
        onPressInIncrement: this.onIncrementValuePressIn,
        onPressOut: this.onPressOut,
        value: value,
        shouldDisplayTextInput: shouldDisplayTextInput
      }, shouldDisplayTextInput && createElement(RangeTextInput, {
        label: label,
        onChange: onChange,
        defaultValue: "".concat(inputValue),
        value: inputValue,
        min: min,
        step: 1,
        decimalNum: decimalNum
      }, children))))));
    }
  }]);

  return BottomSheetStepperCell;
}(Component);

BottomSheetStepperCell.defaultProps = {
  step: DEFAULT_STEP
};
export default withPreferredColorScheme(BottomSheetStepperCell);
