// @wordpress/components/build/ui/flex/use-flex.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useFlex = useFlex;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _element = require("@wordpress/element");

var styles = _interopRequireWildcard(require("./styles"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * @param {import('@wp-g2/create-styles').ViewOwnProps<import('./types').FlexProps, 'div'>} props
 */
function useFlex(props) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'Flex'),
      _useContextSystem$ali = _useContextSystem.align,
      align = _useContextSystem$ali === void 0 ? 'center' : _useContextSystem$ali,
      className = _useContextSystem.className,
      _useContextSystem$dir = _useContextSystem.direction,
      directionProp = _useContextSystem$dir === void 0 ? 'row' : _useContextSystem$dir,
      _useContextSystem$exp = _useContextSystem.expanded,
      expanded = _useContextSystem$exp === void 0 ? true : _useContextSystem$exp,
      _useContextSystem$gap = _useContextSystem.gap,
      gap = _useContextSystem$gap === void 0 ? 2 : _useContextSystem$gap,
      _useContextSystem$jus = _useContextSystem.justify,
      justify = _useContextSystem$jus === void 0 ? 'space-between' : _useContextSystem$jus,
      _useContextSystem$wra = _useContextSystem.wrap,
      wrap = _useContextSystem$wra === void 0 ? false : _useContextSystem$wra,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["align", "className", "direction", "expanded", "gap", "justify", "wrap"]);

  var directionAsArray = Array.isArray(directionProp) ? directionProp : [directionProp];
  var direction = (0, _styles.useResponsiveValue)(directionAsArray);
  var isColumn = typeof direction === 'string' && !!direction.includes('column');
  var isReverse = typeof direction === 'string' && direction.includes('reverse');
  var classes = (0, _element.useMemo)(function () {
    var _css;

    var sx = {};
    sx.Base = (0, _styles.css)((_css = {}, (0, _defineProperty2.default)(_css, _styles.ui.createToken('flexGap'), _styles.ui.space(gap)), (0, _defineProperty2.default)(_css, _styles.ui.createToken('flexItemDisplay'), isColumn ? 'block' : undefined), (0, _defineProperty2.default)(_css, "alignItems", isColumn ? 'normal' : align), (0, _defineProperty2.default)(_css, "flexDirection", direction), (0, _defineProperty2.default)(_css, "flexWrap", wrap ? 'wrap' : undefined), (0, _defineProperty2.default)(_css, "justifyContent", justify), (0, _defineProperty2.default)(_css, "height", isColumn && expanded ? '100%' : undefined), (0, _defineProperty2.default)(_css, "width", !isColumn && expanded ? '100%' : undefined), (0, _defineProperty2.default)(_css, "marginBottom", wrap ? "calc(".concat(_styles.ui.space(gap), " * -1)") : undefined), _css));
    sx.Items = (0, _styles.css)({
      /**
       * Workaround to optimize DOM rendering.
       * We'll enhance alignment with naive parent flex assumptions.
       *
       * Trade-off:
       * Far less DOM less. However, UI rendering is not as reliable.
       */
      '> * + *:not(marquee)': {
        marginTop: isColumn ? _styles.ui.space(gap) : undefined,
        marginRight: !isColumn && isReverse ? _styles.ui.space(gap) : undefined,
        marginLeft: !isColumn && !isReverse ? _styles.ui.space(gap) : undefined
      }
    });
    sx.WrapItems = (0, _styles.css)({
      '> *:not(marquee)': {
        marginBottom: _styles.ui.space(gap),
        marginLeft: !isColumn && isReverse ? _styles.ui.space(gap) : undefined,
        marginRight: !isColumn && !isReverse ? _styles.ui.space(gap) : undefined
      },
      '> *:last-child:not(marquee)': {
        marginLeft: !isColumn && isReverse ? 0 : undefined,
        marginRight: !isColumn && !isReverse ? 0 : undefined
      }
    });
    return (0, _styles.cx)(styles.Flex, sx.Base, wrap ? sx.WrapItems : sx.Items, isColumn ? styles.ItemsColumn : styles.ItemsRow, className);
  }, [align, className, direction, expanded, gap, isColumn, isReverse, justify, wrap]);
  return _objectSpread(_objectSpread({}, otherProps), {}, {
    className: classes
  });
}
