// @wp-g2/components/dist/es/Avatar/__stories__/Avatar.stories.js
import { faker } from '@wp-g2/protokit';
import React from 'react';
import { HStack, View, VStack, ZStack } from '../../index';
import { AVATAR_SIZES } from '../Avatar.utils';
import { Avatar } from '../index';
export default {
  component: Avatar,
  title: 'Components/Avatar'
}; // function StatusDot({ color = 'green' }) {
// 	const styles = {
// 		border: '2px solid white',
// 		zIndex: 1,
// 	};
// 	return (
// 		<View
// 			css={[
// 				{ background: color },
// 				ui.offset({ x: 4, y: -4 }),
// 				ui.borderRadius.circle,
// 				ui.frame({ height: 10, width: 10 }),
// 				ui.position.topRight,
// 				styles,
// 			]}
// 		/>
// 	);
// }

function AvatarGroup(_ref) {
  var children = _ref.children;
  return /*#__PURE__*/React.createElement(ZStack, {
    isLayered: false,
    isReversed: true,
    offset: 8
  }, children);
}

export var _default = function () {
  var avatars = Object.keys(AVATAR_SIZES);
  return /*#__PURE__*/React.createElement(VStack, {
    spacing: 8
  }, /*#__PURE__*/React.createElement(HStack, {
    alignment: "left"
  }, avatars.map(function (size, index) {
    var randomName = faker.name.firstName() + " " + faker.name.lastName();
    var avatarImage = faker.image.avatar();
    return /*#__PURE__*/React.createElement(View, {
      key: index
    }, /*#__PURE__*/React.createElement(Avatar, {
      color: faker.random.arrayElement(['var(--wp-g2-green-500)', 'var(--wp-g2-red-500)', 'var(--wp-g2-orange-500)', undefined]),
      key: size,
      name: randomName,
      size: size,
      src: faker.random.arrayElement([avatarImage, undefined]),
      status: faker.random.arrayElement(['green', 'red', 'orange', undefined])
    }));
  })), /*#__PURE__*/React.createElement(HStack, {
    alignment: "left"
  }, avatars.map(function (size, index) {
    var randomName = "" + faker.name.firstName();
    var avatarImage = faker.image.avatar();
    return /*#__PURE__*/React.createElement(View, {
      key: index
    }, /*#__PURE__*/React.createElement(Avatar, {
      color: faker.random.arrayElement(['green', 'red', 'orange', undefined]),
      key: size,
      name: randomName,
      size: size,
      src: faker.random.arrayElement([avatarImage, undefined]),
      status: faker.random.arrayElement(['green', 'red', 'orange', undefined]),
      variant: "square"
    }));
  })), /*#__PURE__*/React.createElement(AvatarGroup, null, avatars.map(function (size, index) {
    var randomName = "" + faker.name.firstName();
    var avatarImage = faker.image.avatar();
    return /*#__PURE__*/React.createElement(View, {
      key: index
    }, /*#__PURE__*/React.createElement(Avatar, {
      key: size,
      name: randomName,
      src: avatarImage,
      variant: "circle"
    }));
  })));
};
