// @wp-g2/components/dist/cjs/Button/__stories__/Button.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _index = require("../../index");

var _index2 = require("../index");

var _default2 = {
  component: _index2.Button,
  title: 'Components/Button'
};
exports.default = _default2;

var Buttons = function (props) {
  return /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.Flex, {
    direction: "row",
    justify: "left",
    spacing: 4
  }, /*#__PURE__*/_react.default.createElement(_index.Text, {
    size: 12,
    variant: "muted"
  }, "Large"), /*#__PURE__*/_react.default.createElement(_index2.Button, Object.assign({
    size: "large"
  }, props), "Button"), /*#__PURE__*/_react.default.createElement(_index.Text, {
    size: 12,
    variant: "muted"
  }, "Medium"), /*#__PURE__*/_react.default.createElement(_index2.Button, Object.assign({
    size: "medium"
  }, props), "Button"), /*#__PURE__*/_react.default.createElement(_index.Text, {
    size: 12,
    variant: "muted"
  }, "Small"), /*#__PURE__*/_react.default.createElement(_index2.Button, Object.assign({
    size: "small"
  }, props), "Button")));
};

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(Buttons, {
    variant: "primary"
  }), /*#__PURE__*/_react.default.createElement(Buttons, {
    variant: "secondary"
  }), /*#__PURE__*/_react.default.createElement(Buttons, {
    variant: "tertiary"
  }), /*#__PURE__*/_react.default.createElement(Buttons, {
    isDestructive: true,
    variant: "primary"
  }), /*#__PURE__*/_react.default.createElement(Buttons, {
    isDestructive: true,
    variant: "secondary"
  }), /*#__PURE__*/_react.default.createElement(Buttons, {
    isDestructive: true,
    variant: "tertiary"
  }), /*#__PURE__*/_react.default.createElement(Buttons, {
    isSubtle: true,
    variant: "secondary"
  }), /*#__PURE__*/_react.default.createElement(Buttons, {
    isControl: true
  }), /*#__PURE__*/_react.default.createElement(Buttons, {
    isControl: true,
    isSubtle: true
  }), /*#__PURE__*/_react.default.createElement(Buttons, {
    elevation: 2,
    prefix: /*#__PURE__*/_react.default.createElement(_index.Text, null, "Prefix"),
    suffix: /*#__PURE__*/_react.default.createElement(_index.Text, null, "Suffix"),
    variant: "secondary"
  }), /*#__PURE__*/_react.default.createElement(Buttons, {
    elevation: 2,
    isLoading: true,
    prefix: /*#__PURE__*/_react.default.createElement(_index.Text, null, "Prefix"),
    suffix: /*#__PURE__*/_react.default.createElement(_index.Text, null, "Suffix"),
    variant: "secondary"
  }));
};

exports._default = _default;
