// @wp-g2/components/dist/cjs/Select/__stories__/Select.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._multiple = exports._groups = exports._default = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireDefault(require("react"));

var _index = require("../index");

var _default2 = {
  component: _index.Select,
  title: 'Components/Select'
};
exports.default = _default2;

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(_index.Select, {
    options: [{
      label: 'Today',
      value: 'today'
    }, {
      label: 'Yesterday',
      value: 'yesterday'
    }, {
      label: 'Last 7 days',
      value: 'lastWeek'
    }],
    required: true
  });
};

exports._default = _default;

var _groups = function () {
  var _React$useState = _react.default.useState(''),
      _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
      value = _React$useState2[0],
      setValue = _React$useState2[1];

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_index.Select, {
    onChange: setValue,
    options: [{
      label: 'Frozen',
      options: [{
        label: 'Frozen Heart',
        value: 'frozen-heart'
      }, {
        label: 'Do You Want To Build A Snowman?',
        value: 'do-you-want-to-build-a-snowman'
      }, {
        label: 'First Time In Forever',
        value: 'first-time-in-forever'
      }]
    }, {
      label: 'Frozen 2',
      options: [{
        label: 'All Is Found',
        value: 'all-is-found'
      }, {
        label: 'Some Things Never Change',
        value: 'some-things-never-change'
      }, {
        label: 'Into The Unknown',
        value: 'into-the-unknown'
      }]
    }],
    placeholder: "Select a song",
    value: value
  }));
};

exports._groups = _groups;

var _multiple = function () {
  var _React$useState3 = _react.default.useState([]),
      _React$useState4 = (0, _slicedToArray2.default)(_React$useState3, 2),
      value = _React$useState4[0],
      setValue = _React$useState4[1];

  return /*#__PURE__*/_react.default.createElement(_index.Select, {
    multiple: true,
    onChange: setValue,
    options: [{
      label: 'Today',
      value: 'today'
    }, {
      label: 'Yesterday',
      value: 'yesterday'
    }, {
      label: 'Last 7 days',
      value: 'lastWeek'
    }],
    value: value
  });
};

exports._multiple = _multiple;
