// @wordpress/block-editor/build-module/components/media-upload/index.native.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import { createElement } from "@wordpress/element";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

/**
 * WordPress dependencies
 */
import { Component } from '@wordpress/element';
import { __ } from '@wordpress/i18n';
import { Picker } from '@wordpress/components';
import { getOtherMediaOptions, requestMediaPicker, mediaSources } from '@wordpress/react-native-bridge';
import { capturePhoto, captureVideo, image, wordpress, mobile } from '@wordpress/icons';
export var MEDIA_TYPE_IMAGE = 'image';
export var MEDIA_TYPE_VIDEO = 'video';
export var MEDIA_TYPE_AUDIO = 'audio';
export var MEDIA_TYPE_ANY = 'any';
export var OPTION_TAKE_VIDEO = __('Take a Video');
export var OPTION_TAKE_PHOTO = __('Take a Photo');
export var OPTION_TAKE_PHOTO_OR_VIDEO = __('Take a Photo or Video');
export var MediaUpload = /*#__PURE__*/function (_Component) {
  _inherits(MediaUpload, _Component);

  var _super = _createSuper(MediaUpload);

  function MediaUpload(props) {
    var _this;

    _classCallCheck(this, MediaUpload);

    _this = _super.call(this, props);
    _this.onPickerPresent = _this.onPickerPresent.bind(_assertThisInitialized(_this));
    _this.onPickerSelect = _this.onPickerSelect.bind(_assertThisInitialized(_this));
    _this.getAllSources = _this.getAllSources.bind(_assertThisInitialized(_this));
    _this.state = {
      otherMediaOptions: []
    };
    return _this;
  }

  _createClass(MediaUpload, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this2 = this;

      var _this$props$allowedTy = this.props.allowedTypes,
          allowedTypes = _this$props$allowedTy === void 0 ? [] : _this$props$allowedTy;
      getOtherMediaOptions(allowedTypes, function (otherMediaOptions) {
        var otherMediaOptionsWithIcons = otherMediaOptions.map(function (option) {
          return _objectSpread(_objectSpread({}, option), {}, {
            requiresModal: true,
            types: allowedTypes,
            id: option.value
          });
        });

        _this2.setState({
          otherMediaOptions: otherMediaOptionsWithIcons
        });
      });
    }
  }, {
    key: "getAllSources",
    value: function getAllSources() {
      var cameraImageSource = {
        id: mediaSources.deviceCamera,
        // ID is the value sent to native
        value: mediaSources.deviceCamera + '-IMAGE',
        // This is needed to diferenciate image-camera from video-camera sources.
        label: __('Take a Photo'),
        requiresModal: true,
        types: [MEDIA_TYPE_IMAGE],
        icon: capturePhoto
      };
      var cameraVideoSource = {
        id: mediaSources.deviceCamera,
        value: mediaSources.deviceCamera,
        label: __('Take a Video'),
        requiresModal: true,
        types: [MEDIA_TYPE_VIDEO],
        icon: captureVideo
      };
      var deviceLibrarySource = {
        id: mediaSources.deviceLibrary,
        value: mediaSources.deviceLibrary,
        label: __('Choose from device'),
        requiresModal: true,
        types: [MEDIA_TYPE_IMAGE, MEDIA_TYPE_VIDEO],
        icon: image
      };
      var siteLibrarySource = {
        id: mediaSources.siteMediaLibrary,
        value: mediaSources.siteMediaLibrary,
        label: __('WordPress Media Library'),
        requiresModal: true,
        types: [MEDIA_TYPE_IMAGE, MEDIA_TYPE_VIDEO, MEDIA_TYPE_AUDIO, MEDIA_TYPE_ANY],
        icon: wordpress,
        mediaLibrary: true
      };
      var internalSources = [deviceLibrarySource, cameraImageSource, cameraVideoSource, siteLibrarySource];
      return internalSources.concat(this.state.otherMediaOptions);
    }
  }, {
    key: "getMediaOptionsItems",
    value: function getMediaOptionsItems() {
      var _this3 = this;

      var _this$props = this.props,
          _this$props$allowedTy2 = _this$props.allowedTypes,
          allowedTypes = _this$props$allowedTy2 === void 0 ? [] : _this$props$allowedTy2,
          __experimentalOnlyMediaLibrary = _this$props.__experimentalOnlyMediaLibrary;
      return this.getAllSources().filter(function (source) {
        return __experimentalOnlyMediaLibrary ? source.mediaLibrary : allowedTypes.some(function (allowedType) {
          return source.types.includes(allowedType);
        });
      }).map(function (source) {
        return _objectSpread(_objectSpread({}, source), {}, {
          icon: source.icon || _this3.getChooseFromDeviceIcon()
        });
      });
    }
  }, {
    key: "getChooseFromDeviceIcon",
    value: function getChooseFromDeviceIcon() {
      return mobile;
    }
  }, {
    key: "onPickerPresent",
    value: function onPickerPresent() {
      if (this.picker) {
        this.picker.presentPicker();
      }
    }
  }, {
    key: "onPickerSelect",
    value: function onPickerSelect(value) {
      var _this$props2 = this.props,
          _this$props2$allowedT = _this$props2.allowedTypes,
          allowedTypes = _this$props2$allowedT === void 0 ? [] : _this$props2$allowedT,
          onSelect = _this$props2.onSelect,
          _this$props2$multiple = _this$props2.multiple,
          multiple = _this$props2$multiple === void 0 ? false : _this$props2$multiple;
      var mediaSource = this.getAllSources().filter(function (source) {
        return source.value === value;
      }).shift();
      var types = allowedTypes.filter(function (type) {
        return mediaSource.types.includes(type);
      });
      requestMediaPicker(mediaSource.id, types, multiple, function (media) {
        if (multiple && media || media && media.id) {
          onSelect(media);
        }
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this4 = this;

      var _this$props3 = this.props,
          _this$props3$allowedT = _this$props3.allowedTypes,
          allowedTypes = _this$props3$allowedT === void 0 ? [] : _this$props3$allowedT,
          isReplacingMedia = _this$props3.isReplacingMedia,
          multiple = _this$props3.multiple;
      var isOneType = allowedTypes.length === 1;
      var isImage = isOneType && allowedTypes.includes(MEDIA_TYPE_IMAGE);
      var isVideo = isOneType && allowedTypes.includes(MEDIA_TYPE_VIDEO);
      var isAudio = isOneType && allowedTypes.includes(MEDIA_TYPE_AUDIO);
      var isAnyType = isOneType && allowedTypes.includes(MEDIA_TYPE_ANY);
      var isImageOrVideo = allowedTypes.length === 2 && allowedTypes.includes(MEDIA_TYPE_IMAGE) && allowedTypes.includes(MEDIA_TYPE_VIDEO);
      var pickerTitle;

      if (isImage) {
        if (isReplacingMedia) {
          pickerTitle = __('Replace image');
        } else {
          pickerTitle = multiple ? __('Choose images') : __('Choose image');
        }
      } else if (isVideo) {
        if (isReplacingMedia) {
          pickerTitle = __('Replace video');
        } else {
          pickerTitle = __('Choose video');
        }
      } else if (isImageOrVideo) {
        if (isReplacingMedia) {
          pickerTitle = __('Replace image or video');
        } else {
          pickerTitle = __('Choose image or video');
        }
      } else if (isAudio) {
        if (isReplacingMedia) {
          pickerTitle = __('Replace audio');
        } else {
          pickerTitle = __('Choose audio');
        }
      } else if (isAnyType) {
        pickerTitle = __('Choose file');

        if (isReplacingMedia) {
          pickerTitle = __('Replace file');
        } else {
          pickerTitle = __('Choose file');
        }
      }

      var getMediaOptions = function getMediaOptions() {
        return createElement(Picker, {
          title: pickerTitle,
          hideCancelButton: true,
          ref: function ref(instance) {
            return _this4.picker = instance;
          },
          options: _this4.getMediaOptionsItems(),
          onChange: _this4.onPickerSelect
        });
      };

      return this.props.render({
        open: this.onPickerPresent,
        getMediaOptions: getMediaOptions
      });
    }
  }]);

  return MediaUpload;
}(Component);
export default MediaUpload;
