// @wp-g2/create-styles/dist/cjs/create-style-system/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.INTERPOLATION_CLASS_NAME = exports.MODE_SPECIFICITY_COMPOUND_LEVEL = exports.DARK_HIGH_CONTRAST_MODE_MODE_ATTR = exports.REDUCED_MOTION_MODE_ATTR = exports.COLOR_BLIND_MODE_ATTR = exports.HIGH_CONTRAST_MODE_MODE_ATTR = exports.DARK_MODE_ATTR = exports.REDUCED_MOTION_MODE_ATTR_PROP = exports.COLOR_BLIND_MODE_ATTR_PROP = exports.HIGH_CONTRAST_MODE_ATTR_PROP = exports.DARK_MODE_ATTR_PROP = exports.NAMESPACE = void 0;

/**
 * Uses the the prefix for the CSS variables compiled by the system.
 */
var NAMESPACE = '--wp-g2';
exports.NAMESPACE = NAMESPACE;
var DARK_MODE_ATTR_PROP = 'data-system-ui-mode';
exports.DARK_MODE_ATTR_PROP = DARK_MODE_ATTR_PROP;
var HIGH_CONTRAST_MODE_ATTR_PROP = 'data-system-ui-contrast-mode';
exports.HIGH_CONTRAST_MODE_ATTR_PROP = HIGH_CONTRAST_MODE_ATTR_PROP;
var COLOR_BLIND_MODE_ATTR_PROP = 'data-system-ui-color-blind-mode';
exports.COLOR_BLIND_MODE_ATTR_PROP = COLOR_BLIND_MODE_ATTR_PROP;
var REDUCED_MOTION_MODE_ATTR_PROP = 'data-system-ui-reduced-motion-mode';
exports.REDUCED_MOTION_MODE_ATTR_PROP = REDUCED_MOTION_MODE_ATTR_PROP;
var DARK_MODE_ATTR = "[" + DARK_MODE_ATTR_PROP + "=\"dark\"]";
exports.DARK_MODE_ATTR = DARK_MODE_ATTR;
var HIGH_CONTRAST_MODE_MODE_ATTR = "[" + HIGH_CONTRAST_MODE_ATTR_PROP + "=\"high\"]";
exports.HIGH_CONTRAST_MODE_MODE_ATTR = HIGH_CONTRAST_MODE_MODE_ATTR;
var COLOR_BLIND_MODE_ATTR = "[" + COLOR_BLIND_MODE_ATTR_PROP + "=\"true\"]";
exports.COLOR_BLIND_MODE_ATTR = COLOR_BLIND_MODE_ATTR;
var REDUCED_MOTION_MODE_ATTR = "[" + REDUCED_MOTION_MODE_ATTR_PROP + "=\"true\"]";
exports.REDUCED_MOTION_MODE_ATTR = REDUCED_MOTION_MODE_ATTR;
var DARK_HIGH_CONTRAST_MODE_MODE_ATTR = "" + DARK_MODE_ATTR + HIGH_CONTRAST_MODE_MODE_ATTR;
exports.DARK_HIGH_CONTRAST_MODE_MODE_ATTR = DARK_HIGH_CONTRAST_MODE_MODE_ATTR;
var MODE_SPECIFICITY_COMPOUND_LEVEL = 3;
exports.MODE_SPECIFICITY_COMPOUND_LEVEL = MODE_SPECIFICITY_COMPOUND_LEVEL;
var INTERPOLATION_CLASS_NAME = '__interpolationClassName__';
exports.INTERPOLATION_CLASS_NAME = INTERPOLATION_CLASS_NAME;
