// @wp-g2/components/dist/cjs/Modal/__stories__/Modal.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Controlled = exports._default = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _index = require("../../index");

var _index2 = require("../index");

var _default2 = {
  component: _index2.Modal,
  title: 'Components/Modal'
};
exports.default = _default2;

var CloseButton = /*#__PURE__*/_react.default.forwardRef(function (_ref, ref) {
  var _ref$children = _ref.children,
      children = _ref$children === void 0 ? 'Done' : _ref$children,
      _ref$variant = _ref.variant,
      variant = _ref$variant === void 0 ? 'primary' : _ref$variant;

  var _useModalContext = (0, _index2.useModalContext)(),
      dialog = _useModalContext.dialog;

  return /*#__PURE__*/_react.default.createElement(_index.Button, {
    onClick: dialog.hide,
    ref: ref,
    variant: variant
  }, children);
});

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(_index2.Modal, {
    trigger: /*#__PURE__*/_react.default.createElement(_index2.ModalTrigger, null, "Open"),
    visible: true
  }, /*#__PURE__*/_react.default.createElement(_index2.ModalHeader, {
    title: 'Modal Title'
  }), /*#__PURE__*/_react.default.createElement(_index2.ModalBody, null, /*#__PURE__*/_react.default.createElement("h2", null, "First"), /*#__PURE__*/_react.default.createElement(_index2.Modal, {
    trigger: /*#__PURE__*/_react.default.createElement(_index2.ModalTrigger, null, "Open")
  }, /*#__PURE__*/_react.default.createElement(_index2.ModalHeader, {
    title: 'Inner Modal Title'
  }), /*#__PURE__*/_react.default.createElement(_index2.ModalBody, null, /*#__PURE__*/_react.default.createElement("h2", null, "Second"), /*#__PURE__*/_react.default.createElement(_index2.Modal, {
    trigger: /*#__PURE__*/_react.default.createElement(_index2.ModalTrigger, null, "Open")
  }, /*#__PURE__*/_react.default.createElement(_index2.ModalHeader, {
    title: 'Third Inner Modal Title'
  }), /*#__PURE__*/_react.default.createElement(_index2.ModalBody, null, /*#__PURE__*/_react.default.createElement("h2", null, "Third")), /*#__PURE__*/_react.default.createElement(_index2.ModalFooter, null, /*#__PURE__*/_react.default.createElement(CloseButton, null)))))));
};

exports._default = _default;

var Controlled = function () {
  var dialog = (0, _index2.useModalState)();
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_index.Button, {
    onClick: dialog.toggle
  }, "Secondary toggle"), /*#__PURE__*/_react.default.createElement(_index2.Modal, {
    state: dialog,
    trigger: /*#__PURE__*/_react.default.createElement(_index2.ModalTrigger, null, "Open")
  }, /*#__PURE__*/_react.default.createElement(_index2.ModalHeader, {
    title: 'Modal Title'
  }), /*#__PURE__*/_react.default.createElement(_index2.ModalBody, null, /*#__PURE__*/_react.default.createElement("h2", null, "Controlled Modal"))));
};

exports.Controlled = Controlled;
