// @wp-g2/components/dist/es/ZStack/__stories__/ZStack.stories.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import React, { useState } from 'react';
import { Animated, Elevation, HStack, Placeholder, View } from '../../index';
import { ZStack } from '../index';
export default {
  component: ZStack,
  title: 'Components/ZStack'
};

var Avatar = function () {
  return /*#__PURE__*/React.createElement(Animated, {
    layout: true
  }, /*#__PURE__*/React.createElement(Placeholder, {
    css: "\n\t\t\t\t\tborder: 3px solid black;\n\t\t\t\t\tborder-radius: 9999px;\n\t\t\t\t",
    height: 48,
    width: 48
  }), /*#__PURE__*/React.createElement(Elevation, {
    borderRadius: 9999,
    isInteractive: false,
    value: 3
  }));
};

var AnimatedAvatars = function () {
  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      isHover = _useState2[0],
      setIsHover = _useState2[1];

  var offset = isHover ? 0 : 20;
  return /*#__PURE__*/React.createElement(HStack, null, /*#__PURE__*/React.createElement(Animated, {
    onHoverEnd: function onHoverEnd() {
      return setIsHover(false);
    },
    onHoverStart: function onHoverStart() {
      return setIsHover(true);
    }
  }, /*#__PURE__*/React.createElement(ZStack, {
    isLayered: false,
    offset: offset
  }, /*#__PURE__*/React.createElement(Avatar, null), /*#__PURE__*/React.createElement(Avatar, null), /*#__PURE__*/React.createElement(Avatar, null), /*#__PURE__*/React.createElement(Avatar, null))));
};

export var _default = function () {
  return /*#__PURE__*/React.createElement(View, null, /*#__PURE__*/React.createElement(AnimatedAvatars, null));
};
