// @wp-g2/components/dist/cjs/Viewport/Viewport.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Viewport = Viewport;
exports.ViewportMobile = ViewportMobile;
exports.ViewportPhablet = ViewportPhablet;
exports.ViewportPhabletOnly = ViewportPhabletOnly;
exports.ViewportTablet = ViewportTablet;
exports.ViewportTabletOnly = ViewportTabletOnly;
exports.ViewportDesktop = ViewportDesktop;

var _utils = require("@wp-g2/utils");

var _react = _interopRequireDefault(require("react"));

var _Viewport = require("./Viewport.utils");

function Viewport(_ref) {
  var _ref$breakpoint = _ref.breakpoint,
      breakpoint = _ref$breakpoint === void 0 ? 'md' : _ref$breakpoint,
      children = _ref.children,
      media = _ref.media;
  var visible = (0, _utils.useMediaQuery)(media || (0, _Viewport.getBreakpointValue)(breakpoint));
  return visible ? children : null;
}

function ViewportMobile(props) {
  return /*#__PURE__*/_react.default.createElement(Viewport, Object.assign({}, props, {
    media: {
      maxWidth: _Viewport.BREAKPOINTS.xs
    }
  }));
}

function ViewportPhablet(props) {
  return /*#__PURE__*/_react.default.createElement(Viewport, Object.assign({}, props, {
    breakpoint: "sm"
  }));
}

function ViewportPhabletOnly(props) {
  var media = {
    maxWidth: _Viewport.BREAKPOINTS.md,
    minWidth: _Viewport.BREAKPOINTS.sm
  };
  return /*#__PURE__*/_react.default.createElement(Viewport, Object.assign({}, props, {
    media: media
  }));
}

function ViewportTablet(props) {
  return /*#__PURE__*/_react.default.createElement(Viewport, Object.assign({}, props, {
    breakpoint: "md"
  }));
}

function ViewportTabletOnly(props) {
  var media = {
    maxWidth: _Viewport.BREAKPOINTS.lg,
    minWidth: _Viewport.BREAKPOINTS.md
  };
  return /*#__PURE__*/_react.default.createElement(Viewport, Object.assign({}, props, {
    media: media
  }));
}

function ViewportDesktop(props) {
  return /*#__PURE__*/_react.default.createElement(Viewport, Object.assign({}, props, {
    breakpoint: "lg"
  }));
}
