// @wordpress/components/build-module/ui/control-label/styles.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/esm/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7;

/**
 * External dependencies
 */
import { css, getHighDpi, ui } from '@wp-g2/styles';
var lineHeight = "calc(".concat(ui.get('fontSize'), " * 1.2)");
/* eslint-disable jsdoc/valid-types */

/**
 * @param {Parameters<typeof ui.get>[0]} size The padding size.
 */

/* eslint-enable jsdoc/valid-types */

function getPadding(size) {
  return "calc((".concat(ui.get(size), " - ").concat(lineHeight, ") / 2)");
}

var highDpiAdjust = getHighDpi(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n\t> * {\n\t\tposition: relative;\n\t\ttop: 0.5px;\n\t}\n"])));
export var ControlLabel = css(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n\tdisplay: inline-block;\n\tline-height: ", ";\n\tmargin: 0;\n\tmax-width: 100%;\n\tpadding-bottom: ", ";\n\tpadding-top: ", ";\n\n\t&:active {\n\t\tuser-select: none;\n\t}\n\n\t", ";\n"])), lineHeight, getPadding('controlHeight'), getPadding('controlHeight'), highDpiAdjust);
export var large = css(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n\tpadding-bottom: ", ";\n\tpadding-top: ", ";\n"])), getPadding('controlHeightLarge'), getPadding('controlHeightLarge'));
export var medium = css(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n\tpadding-bottom: ", ";\n\tpadding-top: ", ";\n"])), getPadding('controlHeight'), getPadding('controlHeight'));
export var small = css(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["\n\tpadding-bottom: ", ";\n\tpadding-top: ", ";\n"])), getPadding('controlHeightSmall'), getPadding('controlHeightSmall'));
export var inline = css(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["\n\tdisplay: inline-block;\n\tvertical-align: middle;\n"])));
export var block = css(_templateObject7 || (_templateObject7 = _taggedTemplateLiteral(["\n\tdisplay: block;\n"])));
