// @wp-g2/components/dist/cjs/Text/Text.styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.upperCase = exports.highlighterText = exports.muted = exports.destructive = exports.positive = exports.block = exports.Text = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _styles = require("@wp-g2/styles");

function _templateObject7() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\ttext-transform: uppercase;\n"]);

  _templateObject7 = function () {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tmark {\n\t\tbackground: ", ";\n\t\tborder-radius: 2px;\n\t\tbox-shadow: 0 0 0 1px rgba(0, 0, 0, 0.05) inset,\n\t\t\t0 -1px 0 rgba(0, 0, 0, 0.1) inset;\n\t}\n"]);

  _templateObject6 = function () {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tcolor: ", ";\n"]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tcolor: ", ";\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tcolor: ", ";\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tdisplay: block;\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tcolor: ", ";\n\tline-height: ", ";\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var Text = (0, _styles.css)(_templateObject(), _styles.ui.color.text, _styles.ui.get('fontLineHeightBase'));
exports.Text = Text;
var block = (0, _styles.css)(_templateObject2());
exports.block = block;
var positive = (0, _styles.css)(_templateObject3(), _styles.ui.color.positive);
exports.positive = positive;
var destructive = (0, _styles.css)(_templateObject4(), _styles.ui.color.destructive);
exports.destructive = destructive;
var muted = (0, _styles.css)(_templateObject5(), _styles.ui.get('colorTextMuted'));
exports.muted = muted;
var highlighterText = (0, _styles.css)(_templateObject6(), _styles.ui.get('yellowRgba70'));
exports.highlighterText = highlighterText;
var upperCase = (0, _styles.css)(_templateObject7());
exports.upperCase = upperCase;
