// @wp-g2/components/dist/es/Button/__stories__/Button.stories.js
import React from 'react';
import { Flex, Spacer, Text } from '../../index';
import { Button } from '../index';
export default {
  component: Button,
  title: 'Components/Button'
};

var Buttons = function (props) {
  return /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(Flex, {
    direction: "row",
    justify: "left",
    spacing: 4
  }, /*#__PURE__*/React.createElement(Text, {
    size: 12,
    variant: "muted"
  }, "Large"), /*#__PURE__*/React.createElement(Button, Object.assign({
    size: "large"
  }, props), "Button"), /*#__PURE__*/React.createElement(Text, {
    size: 12,
    variant: "muted"
  }, "Medium"), /*#__PURE__*/React.createElement(Button, Object.assign({
    size: "medium"
  }, props), "Button"), /*#__PURE__*/React.createElement(Text, {
    size: 12,
    variant: "muted"
  }, "Small"), /*#__PURE__*/React.createElement(Button, Object.assign({
    size: "small"
  }, props), "Button")));
};

export var _default = function () {
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Buttons, {
    variant: "primary"
  }), /*#__PURE__*/React.createElement(Buttons, {
    variant: "secondary"
  }), /*#__PURE__*/React.createElement(Buttons, {
    variant: "tertiary"
  }), /*#__PURE__*/React.createElement(Buttons, {
    isDestructive: true,
    variant: "primary"
  }), /*#__PURE__*/React.createElement(Buttons, {
    isDestructive: true,
    variant: "secondary"
  }), /*#__PURE__*/React.createElement(Buttons, {
    isDestructive: true,
    variant: "tertiary"
  }), /*#__PURE__*/React.createElement(Buttons, {
    isSubtle: true,
    variant: "secondary"
  }), /*#__PURE__*/React.createElement(Buttons, {
    isControl: true
  }), /*#__PURE__*/React.createElement(Buttons, {
    isControl: true,
    isSubtle: true
  }), /*#__PURE__*/React.createElement(Buttons, {
    elevation: 2,
    prefix: /*#__PURE__*/React.createElement(Text, null, "Prefix"),
    suffix: /*#__PURE__*/React.createElement(Text, null, "Suffix"),
    variant: "secondary"
  }), /*#__PURE__*/React.createElement(Buttons, {
    elevation: 2,
    isLoading: true,
    prefix: /*#__PURE__*/React.createElement(Text, null, "Prefix"),
    suffix: /*#__PURE__*/React.createElement(Text, null, "Suffix"),
    variant: "secondary"
  }));
};
