// @wp-g2/components/dist/cjs/TextInput/__stories__/TextInput.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.madLibs = exports.inlineRendering = exports.custom = exports.multiline = exports.numberStepper = exports.NumberWithExternalChange = exports.Number = exports._default = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireDefault(require("react"));

var _reactNumberFormat = _interopRequireDefault(require("react-number-format"));

var _Container = require("../../Container");

var _index = require("../../index");

var _index2 = require("../index");

var _default2 = {
  component: _index2.TextInput,
  title: 'Components/TextInput'
};
exports.default = _default2;

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(_index2.TextInput, {
    placeholder: "Into The Unknown"
  });
};

exports._default = _default;

var Number = function () {
  var _React$useState = _react.default.useState(0),
      _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
      value = _React$useState2[0],
      setValue = _React$useState2[1];

  return /*#__PURE__*/_react.default.createElement(_Container.Container, {
    css: "\n\t\t\t\tmargin-top: 20vh;\n\t\t\t",
    width: 480
  }, /*#__PURE__*/_react.default.createElement(_index2.TextInput, {
    max: 100,
    min: 0,
    onChange: setValue,
    type: "number",
    value: value
  }));
};

exports.Number = Number;

var NumberWithExternalChange = function () {
  var _React$useState3 = _react.default.useState(),
      _React$useState4 = (0, _slicedToArray2.default)(_React$useState3, 2),
      value = _React$useState4[0],
      setValue = _React$useState4[1];

  return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("div", null, "Value: ", value), /*#__PURE__*/_react.default.createElement("button", {
    onClick: function onClick() {
      return setValue(10);
    }
  }, "Make ten"), /*#__PURE__*/_react.default.createElement(_index2.TextInput, {
    onChange: setValue,
    type: "number",
    value: value
  }));
};

exports.NumberWithExternalChange = NumberWithExternalChange;

var numberStepper = function () {
  return /*#__PURE__*/_react.default.createElement(_Container.Container, {
    css: "\n\t\t\t\tmargin-top: 20vh;\n\t\t\t",
    width: 480
  }, /*#__PURE__*/_react.default.createElement(_index2.TextInput, {
    arrows: "stepper",
    type: "number",
    value: "1"
  }));
};

exports.numberStepper = numberStepper;

var multiline = function () {
  return /*#__PURE__*/_react.default.createElement(_index2.TextInput, {
    maxRows: 6,
    minRows: 3,
    multiline: true
  });
};

exports.multiline = multiline;

var custom = function () {
  return /*#__PURE__*/_react.default.createElement(_index.VStack, null, /*#__PURE__*/_react.default.createElement(_index.Text, null, "Rendered using ", /*#__PURE__*/_react.default.createElement("code", null, "react-number-format")), /*#__PURE__*/_react.default.createElement(_index2.TextInput, {
    as: _reactNumberFormat.default,
    prefix: '$',
    thousandSeparator: true
  }));
};

exports.custom = custom;

var inlineRendering = function () {
  return /*#__PURE__*/_react.default.createElement(_index.VStack, {
    css: "\n\t\t\t\tmargin: auto;\n\t\t\t\twidth: 320px;\n\t\t\t"
  }, /*#__PURE__*/_react.default.createElement(_index.Text, {
    adjustLineHeightForInnerControls: true
  }, "My site name is ", /*#__PURE__*/_react.default.createElement(_index2.TextInput, {
    isInline: true
  })));
};

exports.inlineRendering = inlineRendering;

var MadLib = function (_ref) {
  var type = _ref.type;
  return /*#__PURE__*/_react.default.createElement(_index2.TextInput, {
    isInline: true,
    placeholder: type
  });
};

var Adjective = function () {
  return /*#__PURE__*/_react.default.createElement(MadLib, {
    type: "Adjective"
  });
};

var Noun = function () {
  return /*#__PURE__*/_react.default.createElement(MadLib, {
    type: "Noun"
  });
};

var PluralNoun = function () {
  return /*#__PURE__*/_react.default.createElement(MadLib, {
    isInline: true,
    type: "Plural noun"
  });
};

var Gerund = function () {
  return /*#__PURE__*/_react.default.createElement(MadLib, {
    isInline: true,
    type: "Gerund"
  });
};
/**
 * Adapted from https://www.madlibs.com/printables/
 */


var madLibs = function () {
  return /*#__PURE__*/_react.default.createElement(_index.VStack, {
    css: "\n\t\t\tmargin: auto;\n\t\t\ttext-align: justify;\n\t\t"
  }, /*#__PURE__*/_react.default.createElement(_index.Text, {
    adjustLineHeightForInnerControls: true
  }, "A vacation is when you take a trip to some ", /*#__PURE__*/_react.default.createElement(Adjective, null), " place with your ", /*#__PURE__*/_react.default.createElement(Adjective, null), " family. Usually you go to some place that is near a/an ", /*#__PURE__*/_react.default.createElement(Noun, null), " or up on a/an ", /*#__PURE__*/_react.default.createElement(Noun, null), ". A good vacation place is one where you can ride ", /*#__PURE__*/_react.default.createElement(PluralNoun, null), " or play ", /*#__PURE__*/_react.default.createElement(MadLib, {
    type: "Game"
  }), " or go birding for ", /*#__PURE__*/_react.default.createElement(MadLib, {
    type: "Bird species"
  }), ". I like to spend my time ", /*#__PURE__*/_react.default.createElement(Gerund, null), " or ", /*#__PURE__*/_react.default.createElement(Gerund, null), ". When parents go on a vacation, they spend their time eating three ", /*#__PURE__*/_react.default.createElement(PluralNoun, null), " a day. Last summer, my little sibling fell in a/an ", /*#__PURE__*/_react.default.createElement(Noun, null), " and got poison", ' ', /*#__PURE__*/_react.default.createElement(MadLib, {
    type: "Plant"
  }), " all over their", ' ', /*#__PURE__*/_react.default.createElement(MadLib, {
    type: "Part of the body"
  }), ". My family is going to go to (the) ", /*#__PURE__*/_react.default.createElement(MadLib, {
    type: "Place"
  }), ", and I will practice ", /*#__PURE__*/_react.default.createElement(Gerund, null), ". Parents need vacations more than kids because parents are always very ", /*#__PURE__*/_react.default.createElement(Adjective, null), " and because they have to work", ' ', /*#__PURE__*/_react.default.createElement(MadLib, {
    type: "Number"
  }), " hours every day all year making enough", ' ', /*#__PURE__*/_react.default.createElement(PluralNoun, null), " to pay for the vacation."));
};

exports.madLibs = madLibs;
