// @wp-g2/components/dist/cjs/ControlLabel/ControlLabel.styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.block = exports.inline = exports.small = exports.large = exports.ControlLabel = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _styles = require("@wp-g2/styles");

function _templateObject6() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tdisplay: block;\n"]);

  _templateObject6 = function () {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tdisplay: inline-block;\n\tvertical-align: middle;\n"]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tpadding-bottom: ", ";\n\tpadding-top: ", ";\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tpadding-bottom: ", ";\n\tpadding-top: ", ";\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tdisplay: inline-block;\n\tline-height: ", ";\n\tmargin: 0;\n\tmax-width: 100%;\n\tpadding-bottom: ", ";\n\tpadding-top: ", ";\n\n\t&:active {\n\t\tuser-select: none;\n\t}\n\n\t", ";\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t> * {\n\t\tposition: relative;\n\t\ttop: 0.5px;\n\t}\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var lineHeight = "calc(" + _styles.ui.get('fontSize') + " * 1.2)";

function getPadding(size) {
  return "calc((" + _styles.ui.get(size) + " - " + lineHeight + ") / 2)";
}

var highDpiAdjust = (0, _styles.getHighDpi)((0, _styles.css)(_templateObject()));
var ControlLabel = (0, _styles.css)(_templateObject2(), lineHeight, getPadding('controlHeight'), getPadding('controlHeight'), highDpiAdjust);
exports.ControlLabel = ControlLabel;
var large = (0, _styles.css)(_templateObject3(), getPadding('controlHeightLarge'), getPadding('controlHeightLarge'));
exports.large = large;
var small = (0, _styles.css)(_templateObject4(), getPadding('controlHeightSmall'), getPadding('controlHeightSmall'));
exports.small = small;
var inline = (0, _styles.css)(_templateObject5());
exports.inline = inline;
var block = (0, _styles.css)(_templateObject6());
exports.block = block;
