// @wp-g2/components/dist/cjs/UnitInput/UnitInputSelect.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _i18n = require("@wordpress/i18n");

var _styles = require("@wp-g2/styles");

require("@wp-g2/utils");

var _react = _interopRequireDefault(require("react"));

var _View = require("../View");

var styles = _interopRequireWildcard(require("./UnitInput.styles"));

var _UnitInput2 = require("./UnitInput.utils");

function UnitInputSelect(_ref) {
  var disabled = _ref.disabled,
      onSelectChange = _ref.onSelectChange,
      unit = _ref.unit;

  var _React$useState = _react.default.useState(false),
      _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
      isFocused = _React$useState2[0],
      setFocused = _React$useState2[1];

  var handleOnChange = _react.default.useCallback(function (event) {
    onSelectChange(event.target.value);
  }, [onSelectChange]);

  var handleOnBlur = _react.default.useCallback(function () {
    return setFocused(false);
  }, []);

  var handleOnFocused = _react.default.useCallback(function () {
    return setFocused(true);
  }, []);

  var handleOnStopPropagation = _react.default.useCallback(function (event) {
    event.stopPropagation();
  }, []);

  if (!unit) return null;
  return /*#__PURE__*/_react.default.createElement(_View.View, {
    className: styles.UnitInputSelect
  }, /*#__PURE__*/_react.default.createElement(_View.View, {
    className: (0, _styles.cx)([styles.UnitInputSelectUnit, isFocused && styles.unitSelectFocused, disabled && styles.disabled])
  }, unit, /*#__PURE__*/_react.default.createElement(_View.View, Object.assign({}, _styles.ui.$('UnitInputSelect'), {
    as: "select",
    className: styles.UnitInputSelectElement,
    disabled: disabled,
    onBlur: handleOnBlur,
    onChange: handleOnChange,
    onClick: handleOnStopPropagation,
    onFocus: handleOnFocused,
    onMouseDown: handleOnStopPropagation,
    title: (0, _i18n.__)('Change unit'),
    value: unit
  }), _UnitInput2.UNITS.map(function (i) {
    return /*#__PURE__*/_react.default.createElement("option", {
      key: i,
      value: i
    }, i);
  }))));
}

var _default = /*#__PURE__*/_react.default.memo(UnitInputSelect);

exports.default = _default;
