// @wp-g2/components/dist/cjs/UnitInput/UnitInput.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UnitInput = UnitInput;
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _TextInput = require("../TextInput");

var _UnitInputSelect = _interopRequireDefault(require("./UnitInputSelect"));

var _useUnitInput2 = require("./useUnitInput");

/**
 * @param {import('@wp-g2/create-styles').ViewOwnProps<import('./useUnitInput').Props, 'input'>} props
 * @param {import('react').Ref<any>} forwardedRef
 */
function UnitInput(props, forwardedRef) {
  var _useUnitInput = (0, _useUnitInput2.useUnitInput)(props),
      disabled = _useUnitInput.disabled,
      onSelectChange = _useUnitInput.onSelectChange,
      unit = _useUnitInput.unit,
      value = _useUnitInput.value,
      otherProps = (0, _objectWithoutProperties2.default)(_useUnitInput, ["disabled", "onSelectChange", "unit", "value"]);

  var unitSelect = _react.default.useMemo(function () {
    return /*#__PURE__*/_react.default.createElement(_UnitInputSelect.default, {
      disabled: disabled,
      onSelectChange: onSelectChange,
      unit: unit
    });
  }, [disabled, onSelectChange, unit]);

  var suffix = unit ? unitSelect : null;
  return /*#__PURE__*/_react.default.createElement(_TextInput.TextInput, Object.assign({}, otherProps, _styles.ui.$('UnitInput'), {
    disabled: disabled,
    format: "number",
    ref: forwardedRef,
    suffix: suffix,
    type: "text",
    value: value
  }));
}
/**
 * `UnitInput` is a form component that users can enter (CSS) unit values into.
 *
 * @example
 * ```jsx
 * import { UnitInput } from `@wp-g2/components`
 *
 * function Example() {
 *   return <UnitInput value="50%" min={0} max={100} />
 * }
 * ```
 */


var ConnectedUnitInput = (0, _context.contextConnect)(UnitInput, 'UnitInput');
var _default = ConnectedUnitInput;
exports.default = _default;
