// @wordpress/components/build-module/ui/base-button/component.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { contextConnect } from '@wp-g2/context';
import { cx, ui } from '@wp-g2/styles'; // eslint-disable-next-line no-restricted-imports

import { Radio as ReakitRadio } from 'reakit';
/**
 * WordPress dependencies
 */

import { Icon, chevronDown } from '@wordpress/icons';
/**
 * Internal dependencies
 */

import { useButtonGroupContext } from '../button-group';
import { Elevation } from '../elevation';
import { FlexItem } from '../flex';
import { View } from '../view';
import * as styles from './styles';
import LoadingOverlay from './loading-overlay';
import { useBaseButton } from './hook';
/**
 * @param {import('@wp-g2/create-styles').ViewOwnProps<import('./types').Props, 'button'>} props
 * @param {import('react').Ref<any>} forwardedRef
 */

function BaseButton(props, forwardedRef) {
  var _useBaseButton = useBaseButton(props),
      asProp = _useBaseButton.as,
      children = _useBaseButton.children,
      _useBaseButton$disabl = _useBaseButton.disabled,
      disabled = _useBaseButton$disabl === void 0 ? false : _useBaseButton$disabl,
      _useBaseButton$elevat = _useBaseButton.elevation,
      elevation = _useBaseButton$elevat === void 0 ? 0 : _useBaseButton$elevat,
      elevationActive = _useBaseButton.elevationActive,
      elevationFocus = _useBaseButton.elevationFocus,
      elevationHover = _useBaseButton.elevationHover,
      _useBaseButton$hasCar = _useBaseButton.hasCaret,
      hasCaret = _useBaseButton$hasCar === void 0 ? false : _useBaseButton$hasCar,
      href = _useBaseButton.href,
      icon = _useBaseButton.icon,
      _useBaseButton$iconPo = _useBaseButton.iconPosition,
      iconPosition = _useBaseButton$iconPo === void 0 ? 'left' : _useBaseButton$iconPo,
      _useBaseButton$iconSi = _useBaseButton.iconSize,
      iconSize = _useBaseButton$iconSi === void 0 ? 16 : _useBaseButton$iconSi,
      _useBaseButton$isActi = _useBaseButton.isActive,
      isActive = _useBaseButton$isActi === void 0 ? false : _useBaseButton$isActi,
      _useBaseButton$isDest = _useBaseButton.isDestructive,
      isDestructive = _useBaseButton$isDest === void 0 ? false : _useBaseButton$isDest,
      _useBaseButton$isFocu = _useBaseButton.isFocused,
      isFocused = _useBaseButton$isFocu === void 0 ? false : _useBaseButton$isFocu,
      _useBaseButton$isLoad = _useBaseButton.isLoading,
      isLoading = _useBaseButton$isLoad === void 0 ? false : _useBaseButton$isLoad,
      _useBaseButton$noWrap = _useBaseButton.noWrap,
      noWrap = _useBaseButton$noWrap === void 0 ? true : _useBaseButton$noWrap,
      pre = _useBaseButton.pre,
      suffix = _useBaseButton.suffix,
      otherProps = _objectWithoutProperties(_useBaseButton, ["as", "children", "disabled", "elevation", "elevationActive", "elevationFocus", "elevationHover", "hasCaret", "href", "icon", "iconPosition", "iconSize", "isActive", "isDestructive", "isFocused", "isLoading", "noWrap", "pre", "suffix"]);

  var _useButtonGroupContex = useButtonGroupContext(),
      buttonGroup = _useButtonGroupContex.buttonGroup;

  var buttonGroupState = buttonGroup || {};
  var BaseComponent = buttonGroup ? ReakitRadio : View;
  var as = asProp || (href && !disabled ? 'a' : 'button');
  return (// @ts-ignore No idea why TS is confused about this but ReakitRadio and View are definitely renderable
    createElement(BaseComponent, _extends({
      "aria-busy": isLoading,
      as: as,
      "data-active": isActive,
      "data-destructive": isDestructive,
      "data-focused": isFocused,
      "data-icon": !!icon,
      disabled: disabled || isLoading,
      href: href
    }, buttonGroupState, otherProps, {
      ref: forwardedRef
    }), createElement(LoadingOverlay, {
      isLoading: isLoading
    }), pre && createElement(FlexItem, _extends({
      as: "span",
      className: cx(styles.PrefixSuffix, isLoading && styles.loading)
    }, ui.$('ButtonPrefix')), pre), icon && iconPosition === 'left' && createElement(FlexItem, _extends({
      as: "span",
      className: cx(styles.PrefixSuffix, isLoading && styles.loading)
    }, ui.$('ButtonIcon')), createElement(Icon, {
      icon: icon,
      size: iconSize
    })), children && createElement(FlexItem, _extends({
      as: "span",
      className: cx(styles.Content, isLoading && styles.loading, noWrap && styles.noWrap),
      isBlock: true
    }, ui.$('ButtonContent')), children), icon && iconPosition === 'right' && createElement(FlexItem, _extends({
      as: "span",
      className: cx(styles.PrefixSuffix, isLoading && styles.loading)
    }, ui.$('ButtonIcon')), createElement(Icon, {
      icon: icon,
      size: iconSize
    })), suffix && createElement(FlexItem, _extends({
      as: "span",
      className: cx(styles.PrefixSuffix, isLoading && styles.loading)
    }, ui.$('ButtonSuffix')), suffix), hasCaret && createElement(FlexItem, _extends({
      as: "span",
      className: cx(styles.CaretWrapper, isLoading && styles.loading)
    }, ui.$('ButtonCaret')), createElement(Icon, {
      icon: chevronDown,
      size: 16
    })), createElement(Elevation, _extends({
      active: elevationActive,
      as: "span",
      focus: elevationFocus,
      hover: elevationHover,
      offset: -1,
      value: elevation
    }, ui.$('ButtonElevation'))))
  );
}
/**
 * `BaseButton` is a primitive component used to create actionable components (e.g. `Button`).
 */


var ConnectedBaseButton = contextConnect(BaseButton, 'BaseButton');
export default ConnectedBaseButton;
