// @wp-g2/components/dist/cjs/SegmentedControl/SegmentedControl.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _i18n = require("@wordpress/i18n");

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _utils = require("@wp-g2/utils");

var _react = _interopRequireWildcard(require("react"));

var _reakit = require("reakit");

var _View = require("../View");

var styles = _interopRequireWildcard(require("./SegmentedControl.styles"));

var _SegmentedControlBackdrop = _interopRequireDefault(require("./SegmentedControlBackdrop"));

var _SegmentedControlButton = _interopRequireDefault(require("./SegmentedControlButton"));

function SegmentControl(props, forwardedRef) {
  var _options$;

  var _useContextSystem = (0, _context.useContextSystem)(props, 'SegmentedControl'),
      className = _useContextSystem.className,
      baseId = _useContextSystem.baseId,
      _useContextSystem$isA = _useContextSystem.isAdaptiveWidth,
      isAdaptiveWidth = _useContextSystem$isA === void 0 ? false : _useContextSystem$isA,
      _useContextSystem$isB = _useContextSystem.isBlock,
      isBlock = _useContextSystem$isB === void 0 ? false : _useContextSystem$isB,
      id = _useContextSystem.id,
      _useContextSystem$lab = _useContextSystem.label,
      label = _useContextSystem$lab === void 0 ? (0, _i18n.__)('SegmentControl') : _useContextSystem$lab,
      _useContextSystem$opt = _useContextSystem.options,
      options = _useContextSystem$opt === void 0 ? [] : _useContextSystem$opt,
      _useContextSystem$onC = _useContextSystem.onChange,
      onChange = _useContextSystem$onC === void 0 ? _utils.noop : _useContextSystem$onC,
      _useContextSystem$siz = _useContextSystem.size,
      size = _useContextSystem$siz === void 0 ? 'medium' : _useContextSystem$siz,
      value = _useContextSystem.value,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["className", "baseId", "isAdaptiveWidth", "isBlock", "id", "label", "options", "onChange", "size", "value"]);

  var containerRef = (0, _react.useRef)();

  var _useResizeAware = (0, _utils.useResizeAware)(),
      _useResizeAware2 = (0, _slicedToArray2.default)(_useResizeAware, 2),
      resizeListener = _useResizeAware2[0],
      sizes = _useResizeAware2[1];

  var radio = (0, _reakit.useRadioState)({
    baseId: baseId || id,
    unstable_virtual: true,
    state: value || ((_options$ = options[0]) == null ? void 0 : _options$.value)
  }); // Propagate radio.state change

  (0, _utils.useUpdateEffect)(function () {
    onChange(radio.state);
  }, [radio.state]); // Sync incoming value with radio.state

  (0, _utils.useUpdateEffect)(function () {
    if (value !== radio.state) {
      radio.setState(value);
    }
  }, [value]);
  var classes = (0, _styles.cx)(styles.SegmentedControl, isBlock && styles.block, styles[size], className);
  return /*#__PURE__*/_react.default.createElement(_reakit.RadioGroup, Object.assign({}, radio, {
    "aria-label": label,
    as: _View.View,
    className: classes
  }, otherProps, {
    ref: (0, _utils.mergeRefs)([containerRef, forwardedRef])
  }), resizeListener, /*#__PURE__*/_react.default.createElement(_SegmentedControlBackdrop.default, Object.assign({}, radio, {
    containerRef: containerRef,
    containerWidth: sizes.width
  })), options.map(function (option, index) {
    var showSeparator = getShowSeparator(radio, index);
    return /*#__PURE__*/_react.default.createElement(_SegmentedControlButton.default, Object.assign({}, radio, option, {
      isBlock: !isAdaptiveWidth,
      key: option.value || index,
      showSeparator: showSeparator
    }));
  }));
}

function getShowSeparator(radio, index) {
  var _items$index, _items;

  var currentId = radio.currentId,
      items = radio.items;
  var isLast = index === items.length - 1;
  var isActive = ((_items$index = items[index]) == null ? void 0 : _items$index.id) === currentId;
  var isNextActive = ((_items = items[index + 1]) == null ? void 0 : _items.id) === currentId;
  var showSeparator = true;

  if (items.length < 3) {
    showSeparator = false;
  }

  if (isActive || isNextActive || isLast) {
    showSeparator = false;
  }

  return showSeparator;
}

var _default = (0, _context.contextConnect)(SegmentControl, 'SegmentControl');

exports.default = _default;
