// @wp-g2/components/dist/cjs/SegmentedControl/SegmentedControlBackdrop.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _styles = require("@wp-g2/styles");

var _react = _interopRequireWildcard(require("react"));

var styles = _interopRequireWildcard(require("./SegmentedControl.styles"));

var BackdropView = styles.BackdropView;

function SegmentedControlBackdrop(_ref) {
  var containerRef = _ref.containerRef,
      containerWidth = _ref.containerWidth,
      state = _ref.state;

  var _useState = (0, _react.useState)(0),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      left = _useState2[0],
      setLeft = _useState2[1];

  var _useState3 = (0, _react.useState)(0),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      width = _useState4[0],
      setWidth = _useState4[1];

  var _useState5 = (0, _react.useState)(false),
      _useState6 = (0, _slicedToArray2.default)(_useState5, 2),
      canAnimate = _useState6[0],
      setCanAnimate = _useState6[1];

  (0, _react.useEffect)(function () {
    var containerNode = containerRef == null ? void 0 : containerRef.current;
    if (!containerNode) return;
    /**
     * Workaround for Reakit
     */

    var targetNode = containerNode.querySelector("[data-value=\"" + state + "\"]");
    if (!targetNode) return;

    var _containerNode$getBou = containerNode.getBoundingClientRect(),
        parentX = _containerNode$getBou.x;

    var _targetNode$getBoundi = targetNode.getBoundingClientRect(),
        offsetWidth = _targetNode$getBoundi.width,
        x = _targetNode$getBoundi.x;

    setLeft(x - parentX - 1);
    setWidth(offsetWidth);

    if (!canAnimate) {
      requestAnimationFrame(function () {
        setCanAnimate(true);
      });
    }
  }, [canAnimate, containerRef, containerWidth, state]);
  return /*#__PURE__*/_react.default.createElement(BackdropView, Object.assign({
    role: "presentation"
  }, _styles.ui.$('SegmentedControlBackdrop'), {
    style: {
      transform: "translateX(" + left + "px)",
      transition: canAnimate ? null : 'none',
      width: width
    }
  }));
}

var _default = /*#__PURE__*/_react.default.memo(SegmentedControlBackdrop);

exports.default = _default;
