// @wp-g2/components/dist/cjs/SegmentedControl/SegmentedControl.styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.small = exports.large = exports.LabelPlaceholderView = exports.separatorActive = exports.SeparatorView = exports.buttonActive = exports.ButtonContentView = exports.ButtonView = exports.labelBlock = exports.LabelView = exports.BackdropView = exports.block = exports.SegmentedControl = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _styles = require("@wp-g2/styles");

function _templateObject13() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tmin-height: ", ";\n"]);

  _templateObject13 = function () {
    return data;
  };

  return data;
}

function _templateObject12() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tmin-height: ", ";\n"]);

  _templateObject12 = function () {
    return data;
  };

  return data;
}

function _templateObject11() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tfont-size: ", ";\n\tfont-weight: bold;\n\theight: 0;\n\toverflow: hidden;\n\tvisibility: hidden;\n"]);

  _templateObject11 = function () {
    return data;
  };

  return data;
}

function _templateObject10() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbackground: transparent;\n"]);

  _templateObject10 = function () {
    return data;
  };

  return data;
}

function _templateObject9() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbackground: ", ";\n\theight: calc(100% - 4px - 4px);\n\tposition: absolute;\n\tright: 0;\n\ttop: 4px;\n\ttransition: background ", " linear;\n\twidth: 1px;\n"]);

  _templateObject9 = function () {
    return data;
  };

  return data;
}

function _templateObject8() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tcolor: ", ";\n\tfont-weight: bold;\n"]);

  _templateObject8 = function () {
    return data;
  };

  return data;
}

function _templateObject7() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tfont-size: ", ";\n\tleft: 50%;\n\tline-height: 1;\n\tposition: absolute;\n\ttop: 50%;\n\ttransform: translate(-50%, -50%);\n"]);

  _templateObject7 = function () {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\talign-items: center;\n\tappearance: none;\n\tbackground: transparent;\n\tborder: none;\n\tborder-radius: ", ";\n\tcolor: ", ";\n\tcursor: pointer;\n\tdisplay: flex;\n\theight: 100%;\n\tjustify-content: center;\n\tline-height: 100%;\n\toutline: none;\n\tpadding: 0 12px;\n\tposition: relative;\n\ttext-align: center;\n\ttransition: background ", " linear,\n\t\tcolor ", " linear,\n\t\tfont-weight 60ms linear;\n\tuser-select: none;\n\twidth: 100%;\n\tz-index: 2;\n\n\t&::-moz-focus-inner {\n\t\tborder: 0;\n\t}\n\n\t&:active {\n\t\tbackground: ", ";\n\t}\n"]);

  _templateObject6 = function () {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tflex: 1;\n"]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tdisplay: inline-flex;\n\tmax-width: 100%;\n\tmin-width: 0;\n\tposition: relative;\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbackground: ", ";\n\tborder: 1px solid ", ";\n\tborder-radius: ", ";\n\tbox-shadow: ", ";\n\theight: calc(", " - 4px);\n\tleft: 0;\n\tposition: absolute;\n\ttransition: all ", " ease;\n\tz-index: 1;\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tdisplay: flex;\n\twidth: 100%;\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbackground: ", ";\n\tborder: 1px solid;\n\tborder-color: ", ";\n\tborder-radius: calc(", " + 1px);\n\tdisplay: inline-flex;\n\tmin-height: ", ";\n\tmin-width: 0;\n\tpadding: 1px;\n\tposition: relative;\n\ttransition: all ", " linear;\n\n\t&:hover {\n\t\tborder-color: ", ";\n\t}\n\n\t&:focus {\n\t\tborder-color: ", ";\n\t\tbox-shadow: ", ";\n\t\toutline: none;\n\t\tz-index: 1;\n\t}\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var SegmentedControl = (0, _styles.css)(_templateObject(), _styles.ui.get('segmentedControlBackgroundColor'), _styles.ui.get('segmentedControlBorderColor'), _styles.ui.get('controlBorderRadius'), _styles.ui.get('controlHeight'), _styles.ui.get('transitionDurationFastest'), _styles.ui.get('controlBorderColorHover'), _styles.ui.color.admin, _styles.ui.get('controlBoxShadowFocus'));
exports.SegmentedControl = SegmentedControl;
var block = (0, _styles.css)(_templateObject2());
exports.block = block;

var BackdropView = _styles.styled.div(_templateObject3(), _styles.ui.get('segmentedControlBackdropBackgroundColor'), _styles.ui.get('segmentedControlBackdropBorderColor'), _styles.ui.get('controlBorderRadius'), _styles.ui.get('segmentedControlBackdropBoxShadow'), _styles.ui.get('controlHeight'), _styles.ui.get('transitionDurationFast'));

exports.BackdropView = BackdropView;

var LabelView = _styles.styled.div(_templateObject4());

exports.LabelView = LabelView;
var labelBlock = (0, _styles.css)(_templateObject5());
exports.labelBlock = labelBlock;

var ButtonView = _styles.styled.button(_templateObject6(), _styles.ui.get('controlBorderRadius'), _styles.ui.color.text, _styles.ui.get('transitionDurationFast'), _styles.ui.get('transitionDurationFast'), _styles.ui.get('segmentedControlButtonColorActive'));

exports.ButtonView = ButtonView;

var ButtonContentView = _styles.styled.div(_templateObject7(), _styles.ui.get('segmentedControlFontSize'));

exports.ButtonContentView = ButtonContentView;
var buttonActive = (0, _styles.css)(_templateObject8(), _styles.ui.get('controlTextActiveColor'));
exports.buttonActive = buttonActive;

var SeparatorView = _styles.styled.div(_templateObject9(), _styles.ui.get('colorDivider'), _styles.ui.get('transitionDuration'));

exports.SeparatorView = SeparatorView;
var separatorActive = (0, _styles.css)(_templateObject10());
exports.separatorActive = separatorActive;

var LabelPlaceholderView = _styles.styled.div(_templateObject11(), _styles.ui.get('segmentedControlFontSize'));

exports.LabelPlaceholderView = LabelPlaceholderView;
var large = (0, _styles.css)(_templateObject12(), _styles.ui.get('controlHeightLarge'));
exports.large = large;
var small = (0, _styles.css)(_templateObject13(), _styles.ui.get('controlHeightSmall'));
exports.small = small;
