// @wp-g2/components/dist/es/ClipboardButton/__stories__/ClipboardButton.stories.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import React from 'react';
import { ClipboardButton } from '../index';
export default {
  component: ClipboardButton,
  title: 'Components/ClipboardButton'
};

var Example = function () {
  var _React$useState = React.useState(false),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      hasCopied = _React$useState2[0],
      setHasCopied = _React$useState2[1];

  var _React$useState3 = React.useState(false),
      _React$useState4 = _slicedToArray(_React$useState3, 2),
      hasFinishedCopying = _React$useState4[0],
      setHasFinishedCopying = _React$useState4[1];

  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(ClipboardButton, {
    onCopy: function onCopy() {
      return setHasCopied(true);
    },
    onFinishCopy: function onFinishCopy() {
      return setHasFinishedCopying(true);
    },
    text: "https://wordpress.org",
    variant: "primary"
  }, "Copy \"https://wordpress.org\""), /*#__PURE__*/React.createElement("div", null, hasCopied && !hasFinishedCopying && 'Copying...', hasFinishedCopying && 'Success!!!!'));
};

export var _default = function () {
  return /*#__PURE__*/React.createElement(Example, null);
};
