// @wp-g2/components/dist/es/__stories__/WIP/FormElementsInPanel.stories.js
import React from 'react';
import { Button, Checkbox, Container, FormGroup, ListGroup, Panel, PanelBody, PanelHeader, Surface, Text } from '../../index';
export default {
  title: 'Examples/WIP/FormElementInPanel'
};

var Example = function () {
  return /*#__PURE__*/React.createElement(Container, {
    css: "\n\t\t\t\tmargin-top: 10vh;\n\t\t\t\tmargin-bottom: 20vh;\n\t\t\t",
    width: 280
  }, /*#__PURE__*/React.createElement(Surface, {
    border: true
  }, /*#__PURE__*/React.createElement(Panel, {
    visible: true
  }, /*#__PURE__*/React.createElement(PanelHeader, null, "Status & visibility"), /*#__PURE__*/React.createElement(PanelBody, null, /*#__PURE__*/React.createElement(ListGroup, null, /*#__PURE__*/React.createElement(FormGroup, {
    horizontal: true,
    label: "Visibility"
  }, /*#__PURE__*/React.createElement(Button, null, "Public")), /*#__PURE__*/React.createElement(FormGroup, {
    horizontal: true,
    label: "Publish"
  }, /*#__PURE__*/React.createElement(Button, null, "Immediately")), /*#__PURE__*/React.createElement(FormGroup, null, /*#__PURE__*/React.createElement(Checkbox, {
    label: "Stick to the top of the blog"
  })), /*#__PURE__*/React.createElement(FormGroup, null, /*#__PURE__*/React.createElement(Checkbox, {
    label: "Pending review"
  })), /*#__PURE__*/React.createElement(Button, {
    isDestructive: true
  }, "Move to trash"))))));
};

export var _default = function () {
  return /*#__PURE__*/React.createElement(Example, null);
};
