// @wp-g2/components/dist/es/__stories__/WIP/Interpolation.stories.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { ui } from '@wp-g2/styles';
import { interpolate } from '@wp-g2/utils';
import React, { useEffect, useState } from 'react';
import { View } from '../../index';
export default {
  title: 'Examples/WIP/Utils/Interpolate'
};

var Circle = function (props) {
  return /*#__PURE__*/React.createElement(View, Object.assign({}, props, {
    css: [ui.background.blue, ui.frame({
      height: 100,
      width: 100
    }), ui.borderRadius.circle]
  }));
};

var Example = function () {
  var _useState = useState({
    x: 0,
    y: 0
  }),
      _useState2 = _slicedToArray(_useState, 2),
      position = _useState2[0],
      setPosition = _useState2[1];

  useEffect(function () {
    var updatePosition = function (event) {
      setPosition({
        x: event.clientX,
        y: event.clientY
      });
    };

    document.addEventListener('mousemove', updatePosition);
    return function () {
      document.removeEventListener('mousemove', updatePosition);
    };
  }, []);
  var top = position.y - 50;
  var opacity = interpolate(position.x, [0, window.innerWidth], [1, 0]);
  var backgroundColor = interpolate(position.x, [0, window.innerWidth], ['red', 'blue']);
  return /*#__PURE__*/React.createElement(Circle, {
    style: {
      backgroundColor: backgroundColor,
      left: position.x - 50,
      opacity: opacity,
      position: 'fixed',
      top: top
    }
  });
};

export var _default = function () {
  return /*#__PURE__*/React.createElement(Example, null);
};
