// @wp-g2/components/dist/es/__stories__/WIP/PerformanceNative.stories.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import { Schema } from '@wp-g2/protokit';
import { css, styled, ui } from '@wp-g2/styles';
import React, { useState } from 'react';
export default {
  title: 'Examples/WIP/PerformanceNative'
};
var dimensionSchema = new Schema(function () {
  return {
    title: '',
    x: 0,
    y: 1,
    z: 0
  };
});
var SomeContext = /*#__PURE__*/React.createContext({});

var useSomeContext = function () {
  return React.useContext(SomeContext);
};

var OtherContext = /*#__PURE__*/React.createContext({});

var useOtherContext = function () {
  return React.useContext(OtherContext);
};

var SliderTextInput = /*#__PURE__*/React.memo(function (_ref) {
  var onChange = _ref.onChange,
      prop = _ref.prop,
      value = _ref.value;

  var _useSomeContext = useSomeContext(),
      slider = _useSomeContext.slider;

  var handleOnChange = React.useCallback(function (next) {
    var _onChange;

    onChange((_onChange = {}, _onChange[prop] = next.target.value, _onChange));
  }, [onChange, prop]);
  return /*#__PURE__*/React.createElement("div", null, slider && /*#__PURE__*/React.createElement("input", {
    onChange: handleOnChange,
    type: "range",
    value: value
  }), /*#__PURE__*/React.createElement("input", {
    onChange: handleOnChange,
    type: "number",
    value: value
  }));
});
var TestFormGroup = /*#__PURE__*/React.memo(function (_ref2) {
  var children = _ref2.children,
      label = _ref2.label;
  var contextValue = React.useMemo(function () {
    return {
      slider: true
    };
  }, []);
  return /*#__PURE__*/React.createElement(SomeContext.Provider, {
    value: contextValue
  }, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("label", null, label, children)))));
});
var DimensionCard = /*#__PURE__*/React.memo(function (_ref3) {
  var id = _ref3.id,
      onChange = _ref3.onChange,
      title = _ref3.title,
      x = _ref3.x,
      y = _ref3.y,
      z = _ref3.z;
  var updateTitle = React.useCallback(function (next) {
    onChange({
      id: id,
      title: next.target.value
    });
  }, [onChange, id]);
  var updateValue = React.useCallback(function (next) {
    onChange(_objectSpread({
      id: id
    }, next));
  }, [onChange, id]);
  return /*#__PURE__*/React.createElement("li", null, /*#__PURE__*/React.createElement("h5", null, "Dimensions"), /*#__PURE__*/React.createElement(TestFormGroup, {
    label: "Title"
  }, /*#__PURE__*/React.createElement("input", {
    onChange: updateTitle,
    value: title
  })), /*#__PURE__*/React.createElement(TestFormGroup, {
    label: "x"
  }, /*#__PURE__*/React.createElement(SliderTextInput, {
    onChange: updateValue,
    prop: "x",
    value: x
  })), /*#__PURE__*/React.createElement(TestFormGroup, {
    label: "y"
  }, /*#__PURE__*/React.createElement(SliderTextInput, {
    onChange: updateValue,
    prop: "y",
    value: y
  })), /*#__PURE__*/React.createElement(TestFormGroup, {
    label: "z"
  }, /*#__PURE__*/React.createElement(SliderTextInput, {
    onChange: updateValue,
    prop: "z",
    value: z
  })));
});

var Example = function () {
  var _useState = useState([].concat(dimensionSchema.make(1))),
      _useState2 = _slicedToArray(_useState, 2),
      dimensions = _useState2[0],
      setDimensions = _useState2[1];

  var updateDimension = React.useCallback(function (next) {
    setDimensions(function (prev) {
      return prev.map(function (d) {
        if (d.id !== next.id) return d;
        return _objectSpread(_objectSpread({}, d), next);
      });
    });
  }, []);
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("button", {
    onClick: function addDimension() {
      setDimensions(function (prev) {
        return [].concat(prev, [dimensionSchema.makeOne()]);
      });
    }
  }, "Add New"), /*#__PURE__*/React.createElement(OtherContext.Provider, {
    value: {
      ok: true
    }
  }, /*#__PURE__*/React.createElement("ul", null, dimensions.map(function (item) {
    return /*#__PURE__*/React.createElement(DimensionCard, Object.assign({}, item, {
      onChange: updateDimension
    }));
  }))));
};

export var _default = function () {
  return /*#__PURE__*/React.createElement(Example, null);
};
