// @wp-g2/components/dist/es/__stories__/WIP/Toolbar.stories.js
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _taggedTemplateLiteralLoose from "@babel/runtime/helpers/esm/taggedTemplateLiteralLoose";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _templateObject7() {
  var data = _taggedTemplateLiteralLoose(["\n\tborder-top: 1px solid ", ";\n\tmargin: 8px -8px;\n"]);

  _templateObject7 = function () {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = _taggedTemplateLiteralLoose(["\n\tappearance: none;\n\tbackground: none;\n\tborder: none;\n\tborder-radius: 2px;\n\tcolor: ", ";\n\tcursor: pointer;\n\tdisplay: block;\n\toutline: none;\n\tpadding: 8px 16px;\n\ttext-align: left;\n\twidth: 100%;\n\n\t&:hover {\n\t\tcolor: ", ";\n\t}\n\n\t&:focus {\n\t\tbox-shadow: 0 0 0 2px ", ";\n\t}\n"]);

  _templateObject6 = function () {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = _taggedTemplateLiteralLoose(["\n\tbackground: ", ";\n\tborder: 1px solid ", ";\n\tborder-radius: 2px;\n\tmin-width: 200px;\n\toutline: none;\n\tpadding: 8px;\n"]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = _taggedTemplateLiteralLoose(["\n\tposition: relative;\n\tz-index: 1;\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = _taggedTemplateLiteralLoose(["\n\tborder-right: 1px solid ", ";\n\tdisplay: flex;\n\n\t&:last-of-type {\n\t\tborder-right: none;\n\t}\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteralLoose(["\n\talign-items: center;\n\tappearance: none;\n\tbackground: ", ";\n\tborder: none;\n\tcolor: ", ";\n\tcursor: pointer;\n\tdisplay: flex;\n\theight: 48px;\n\tjustify-content: center;\n\tmin-width: 36px;\n\toutline: none;\n\tpadding: 6px;\n\tposition: relative;\n\n\t&:first-of-type {\n    border-bottom-left-radius: 3px;\n    border-top-left-radius: 3px;\n    min-width: 42px;\n\t\tpadding-left: 12px;\n\t}\n\n  &:last-of-type {\n    border-bottom-right-radius: 3px;\n    border-top-right-radius: 3px;\n\t\tmin-width: 42px;\n    padding-right: 12px;\n\t}\n\n  &:first-of-type::last-of-type {\n\t\tmin-width: 42px;\n    padding-left: 12px;\n\t  padding-right: 12px;\n\t}\n\n  &:hover {\n    color: ", ";\n  }\n\n\t&::before {\n\t\tborder-radius: 2px;\n\t\tcontent: '';\n\t\theight: 30px;\n\t\tleft: 0;\n\t\tpointer-events: none;\n\t\tposition: absolute;\n\t\ttop: 50%;\n\t\ttransform: translate(3px, -50%);\n\t\twidth: calc(100% - 6px);\n\t}\n\n\t&:first-of-type {\n    &::before {\n      transform: translate(10px, -50%);\n\t\t  width: calc(100% - 14px);\n    }\n  }\n\n  &:last-of-type {\n    &::before {\n      transform: translate(4px, -50%);\n\t\t  width: calc(100% - 14px);\n    }\n  }\n\n  &:first-of-type:last-of-type {\n    &::before {\n      transform: translate(8px, -50%);\n\t\t  width: calc(100% - 16px);\n    }\n  }\n\n\t&:focus {\n\t\t/* color: ", ";\n\n\t\t&::before {\n\t\t\tbackground: black;\n\t\t} */\n    &::before {\n      box-shadow: 0 0 0 2px ", ";\n    }\n\t}\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteralLoose(["\n\tbackground: ", ";\n\tborder: 1px solid ", ";\n\tborder-radius: 2px;\n\tdisplay: inline-flex;\n\tpadding: 0;\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

import * as IconSet from '@wp-g2/icons';
import { styled, ui } from '@wp-g2/styles';
import { is } from '@wp-g2/utils';
import React from 'react';
import { Menu, MenuButton, MenuItem, MenuSeparator, Toolbar as ReakitToolbar, ToolbarItem as ReakitToolbarItem, useMenuState, useToolbarState } from 'reakit';
import { Icon, Tooltip } from '../../index';
export default {
  title: 'Examples/WIP/Toolbar'
};
var ToolbarContext = /*#__PURE__*/React.createContext({});

var useToolbarContext = function () {
  return React.useContext(ToolbarContext);
};

var ToolbarView = styled.div(_templateObject(), ui.get('surfaceColor'), ui.color.text);
var ToolbarButtonView = styled.button(_templateObject2(), ui.get('surfaceColor'), ui.color.text, ui.color.admin, ui.color.textInverted, ui.color.admin);
var ToolbarGroupView = styled.div(_templateObject3(), ui.color.text);
var ToolbarButtonContentView = styled.span(_templateObject4());
var ToolbarDropdownView = styled.div(_templateObject5(), ui.get('surfaceColor'), ui.color.text);
var ToolbarDropdownItemView = styled.button(_templateObject6(), ui.color.text, ui.color.admin, ui.color.admin);
var ToolbarSeparatorView = styled.div(_templateObject7(), ui.color.text);
var MoreItems = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var menu = useMenuState({
    placement: 'bottom-start',
    gutter: 4,
    loop: true
  });
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Tooltip, {
    content: "More options"
  }, /*#__PURE__*/React.createElement(MenuButton, Object.assign({}, menu, props, {
    "aria-label": "More items",
    as: BaseToolbarItem,
    icon: /*#__PURE__*/React.createElement(IconSet.FiMoreVertical, null),
    ref: ref
  }))), /*#__PURE__*/React.createElement(Menu, Object.assign({}, menu, {
    "aria-label": "More items",
    as: ToolbarDropdownView
  }), /*#__PURE__*/React.createElement(MenuItem, Object.assign({}, menu, {
    as: ToolbarDropdownItemView
  }), "Show Block Settings"), /*#__PURE__*/React.createElement(MenuItem, Object.assign({}, menu, {
    as: ToolbarDropdownItemView
  }), "Copy"), /*#__PURE__*/React.createElement(MenuItem, Object.assign({}, menu, {
    as: ToolbarDropdownItemView
  }), "Duplicate"), /*#__PURE__*/React.createElement(MenuItem, Object.assign({}, menu, {
    as: ToolbarDropdownItemView
  }), "Insert Before"), /*#__PURE__*/React.createElement(MenuItem, Object.assign({}, menu, {
    as: ToolbarDropdownItemView
  }), "Insert After"), /*#__PURE__*/React.createElement(MenuItem, Object.assign({}, menu, {
    as: ToolbarDropdownItemView
  }), "Move To"), /*#__PURE__*/React.createElement(MenuItem, Object.assign({}, menu, {
    as: ToolbarDropdownItemView
  }), "Edit as HTML"), /*#__PURE__*/React.createElement(MenuSeparator, Object.assign({}, menu, {
    as: ToolbarSeparatorView
  })), /*#__PURE__*/React.createElement(MenuItem, Object.assign({}, menu, {
    as: ToolbarDropdownItemView
  }), "Group"), /*#__PURE__*/React.createElement(MenuSeparator, Object.assign({}, menu, {
    as: ToolbarSeparatorView
  })), /*#__PURE__*/React.createElement(MenuItem, Object.assign({}, menu, {
    as: ToolbarDropdownItemView
  }), "Remove Block")));
});

function iconStringToToolbarComponents(icons) {
  return function (items, mapIndex) {
    if (!Array.isArray(items)) return null;
    return /*#__PURE__*/React.createElement(ToolbarGroup, {
      key: mapIndex
    }, items.map(function (icon, index) {
      var ToolbarIcon = icons[icon];
      if (!ToolbarIcon) return null;
      return /*#__PURE__*/React.createElement(ToolbarItem, {
        icon: /*#__PURE__*/React.createElement(ToolbarIcon, null),
        key: index
      });
    }));
  };
}

var templateToIconString = function (item) {
  if (typeof item === 'string') {
    var next = item.replace(/ /g, '');
    return "Fi" + next;
  }

  if (Array.isArray(item)) return item.map(templateToIconString).filter(Boolean);
  return undefined;
};

function generateToolbarComponentsFromTemplate(template, icons) {
  if (typeof template !== 'string') return null;
  var items = template.split('|').map(function (chunk) {
    return chunk.split('-');
  }).map(templateToIconString).map(iconStringToToolbarComponents(icons));
  return items;
}

function Toolbar(_ref) {
  var children = _ref.children,
      _ref$icons = _ref.icons,
      icons = _ref$icons === void 0 ? {} : _ref$icons,
      template = _ref.template;
  var toolbar = useToolbarState({
    loop: true
  });

  var iconSet = _objectSpread(_objectSpread({}, IconSet), icons);

  var content = template ? generateToolbarComponentsFromTemplate(template, iconSet) : null;
  return /*#__PURE__*/React.createElement(ToolbarContext.Provider, {
    value: {
      toolbar: toolbar
    }
  }, /*#__PURE__*/React.createElement(ReakitToolbar, Object.assign({}, toolbar, {
    "aria-label": "My toolbar",
    as: ToolbarView
  }), content, children));
}

var BaseToolbarItem = /*#__PURE__*/React.forwardRef(function (_ref2, ref) {
  var icon = _ref2.icon,
      props = _objectWithoutProperties(_ref2, ["icon"]);

  return /*#__PURE__*/React.createElement(ToolbarButtonView, Object.assign({
    ref: ref
  }, props), /*#__PURE__*/React.createElement(ToolbarButtonContentView, null, /*#__PURE__*/React.createElement(Icon, {
    icon: icon
  })));
});
var ToolbarItem = /*#__PURE__*/React.forwardRef(function (_ref3, ref) {
  var icon = _ref3.icon;

  var _useToolbarContext = useToolbarContext(),
      toolbar = _useToolbarContext.toolbar;

  return /*#__PURE__*/React.createElement(ReakitToolbarItem, Object.assign({}, toolbar, {
    as: BaseToolbarItem,
    icon: icon,
    ref: ref
  }));
});

function ToolbarGroup(_ref4) {
  var children = _ref4.children;
  return /*#__PURE__*/React.createElement(ToolbarGroupView, null, children);
}

function MoreExample() {
  var _useToolbarContext2 = useToolbarContext(),
      toolbar = _useToolbarContext2.toolbar;

  return /*#__PURE__*/React.createElement(ReakitToolbarItem, Object.assign({}, toolbar, {
    as: MoreItems
  }));
}

function Example() {
  return /*#__PURE__*/React.createElement(Toolbar, {
    template: "Activity | AlignLeft - AlignCenter - AlignRight | Bell"
  }, /*#__PURE__*/React.createElement(ToolbarGroup, null, /*#__PURE__*/React.createElement(ToolbarItem, {
    icon: /*#__PURE__*/React.createElement(IconSet.FiArrowUp, null)
  })), /*#__PURE__*/React.createElement(MoreExample, null));
}

export var _default = function () {
  return /*#__PURE__*/React.createElement(Example, null);
};
