// @wp-g2/components/dist/cjs/HelpTip/HelpTip.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _icons = require("@wordpress/icons");

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _Icon = require("../Icon");

var _Tooltip = require("../Tooltip");

var _View = require("../View");

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t\t\t\t\t\tdisplay: inline-flex;\n\t\t\t\t\t\topacity: 0.5;\n\t\t\t\t\t"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t\t\t\t\tcursor: pointer;\n\t\t\t\t\tpadding-left: ", ";\n\t\t\t\t\tpadding-right: ", ";\n\t\t\t\t\tvertical-align: middle;\n\t\t\t\t"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

function HelpTip(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'HelpTip'),
      as = _useContextSystem.as,
      children = _useContextSystem.children,
      _useContextSystem$ico = _useContextSystem.iconSize,
      iconSize = _useContextSystem$ico === void 0 ? 14 : _useContextSystem$ico,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["as", "children", "iconSize"]);

  var handleOnClick = _react.default.useCallback(function (event) {
    event.preventDefault();
  }, []);

  return /*#__PURE__*/_react.default.createElement(_Tooltip.Tooltip, Object.assign({}, otherProps, {
    content: children,
    ref: forwardedRef
  }), /*#__PURE__*/_react.default.createElement(_View.View, {
    as: as || 'span',
    css: (0, _styles.css)(_templateObject(), _styles.ui.space(1), _styles.ui.space(1)),
    onClick: handleOnClick
  }, /*#__PURE__*/_react.default.createElement(_Icon.Icon, {
    css: (0, _styles.css)(_templateObject2()),
    fill: _styles.ui.get('colorText'),
    icon: _icons.help,
    size: iconSize
  })));
}

var _default = (0, _context.contextConnect)(HelpTip, 'HelpTip');

exports.default = _default;
