// @wp-g2/styles/dist/es/presets/spacing.js
import { space } from '../mixins/space';
import { css } from '../style-system';
export { space } from '../mixins/space';
/**
 * @param {import('react').ReactText} value
 */

export function margin(value) {
  return css({
    margin: space(value)
  });
}

margin.x = function (
/** @type {import('react').ReactText} */
v) {
  return css({
    marginLeft: space(v),
    marginRight: space(v)
  });
};

margin.y = function (
/** @type {import('react').ReactText} */
v) {
  return css({
    marginBottom: space(v),
    marginTop: space(v)
  });
};

margin.top = function (
/** @type {import('react').ReactText} */
v) {
  return css({
    marginTop: space(v)
  });
};

margin.bottom = function (
/** @type {import('react').ReactText} */
v) {
  return css({
    marginBottom: space(v)
  });
};

margin.left = function (
/** @type {import('react').ReactText} */
v) {
  return css({
    marginLeft: space(v)
  });
};

margin.right = function (
/** @type {import('react').ReactText} */
v) {
  return css({
    marginRight: space(v)
  });
};
/**
 * @param {import('react').ReactText} value
 */


export function padding(value) {
  return css({
    padding: space(value)
  });
}

padding.x = function (
/** @type {import('react').ReactText} */
v) {
  return css({
    paddingLeft: space(v),
    paddingRight: space(v)
  });
};

padding.y = function (
/** @type {import('react').ReactText} */
v) {
  return css({
    paddingBottom: space(v),
    paddingTop: space(v)
  });
};

padding.top = function (
/** @type {import('react').ReactText} */
v) {
  return css({
    paddingTop: space(v)
  });
};

padding.bottom = function (
/** @type {import('react').ReactText} */
v) {
  return css({
    paddingBottom: space(v)
  });
};

padding.left = function (
/** @type {import('react').ReactText} */
v) {
  return css({
    paddingLeft: space(v)
  });
};

padding.right = function (
/** @type {import('react').ReactText} */
v) {
  return css({
    paddingRight: space(v)
  });
};
