// @wordpress/block-editor/build/components/block-types-list/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _blocks = require("@wordpress/blocks");

var _inserterListItem = _interopRequireDefault(require("../inserter-list-item"));

var _inserterListbox = require("../inserter-listbox");

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function chunk(array, size) {
  var chunks = [];

  for (var i = 0, j = array.length; i < j; i += size) {
    chunks.push(array.slice(i, i + size));
  }

  return chunks;
}

function BlockTypesList(_ref) {
  var _ref$items = _ref.items,
      items = _ref$items === void 0 ? [] : _ref$items,
      onSelect = _ref.onSelect,
      _ref$onHover = _ref.onHover,
      onHover = _ref$onHover === void 0 ? function () {} : _ref$onHover,
      children = _ref.children,
      label = _ref.label,
      _ref$isDraggable = _ref.isDraggable,
      isDraggable = _ref$isDraggable === void 0 ? true : _ref$isDraggable;
  return (0, _element.createElement)(_inserterListbox.InserterListboxGroup, {
    className: "block-editor-block-types-list",
    "aria-label": label
  }, chunk(items, 3).map(function (row, i) {
    return (0, _element.createElement)(_inserterListbox.InserterListboxRow, {
      key: i
    }, row.map(function (item, j) {
      return (0, _element.createElement)(_inserterListItem.default, {
        key: item.id,
        item: item,
        className: (0, _blocks.getBlockMenuDefaultClassName)(item.id),
        onSelect: onSelect,
        onHover: onHover,
        isDraggable: isDraggable,
        isFirst: i === 0 && j === 0
      });
    }));
  }), children);
}

var _default = BlockTypesList;
exports.default = _default;
