// @wordpress/components/build/input-control/input-base.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InputBase = InputBase;
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _compose = require("@wordpress/compose");

var _backdrop = _interopRequireDefault(require("./backdrop"));

var _label = _interopRequireDefault(require("./label"));

var _inputControlStyles = require("./styles/input-control-styles");

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function useUniqueId(idProp) {
  var instanceId = (0, _compose.useInstanceId)(InputBase);
  var id = "input-base-control-".concat(instanceId);
  return idProp || id;
} // Adapter to map props for the new ui/flex compopnent.


function getUIFlexProps(_ref) {
  var labelPosition = _ref.labelPosition;
  var props = {};

  switch (labelPosition) {
    case 'top':
      props.direction = 'column';
      props.gap = 0;
      break;

    case 'bottom':
      props.direction = 'column-reverse';
      props.gap = 0;
      break;

    case 'edge':
      props.justify = 'space-between';
      break;
  }

  return props;
}

function InputBase(_ref2, ref) {
  var __unstableInputWidth = _ref2.__unstableInputWidth,
      children = _ref2.children,
      className = _ref2.className,
      _ref2$disabled = _ref2.disabled,
      disabled = _ref2$disabled === void 0 ? false : _ref2$disabled,
      _ref2$hideLabelFromVi = _ref2.hideLabelFromVision,
      hideLabelFromVision = _ref2$hideLabelFromVi === void 0 ? false : _ref2$hideLabelFromVi,
      labelPosition = _ref2.labelPosition,
      idProp = _ref2.id,
      _ref2$isFocused = _ref2.isFocused,
      isFocused = _ref2$isFocused === void 0 ? false : _ref2$isFocused,
      label = _ref2.label,
      prefix = _ref2.prefix,
      _ref2$size = _ref2.size,
      size = _ref2$size === void 0 ? 'default' : _ref2$size,
      suffix = _ref2.suffix,
      props = (0, _objectWithoutProperties2.default)(_ref2, ["__unstableInputWidth", "children", "className", "disabled", "hideLabelFromVision", "labelPosition", "id", "isFocused", "label", "prefix", "size", "suffix"]);
  var id = useUniqueId(idProp);
  var hideLabel = hideLabelFromVision || !label;
  return (0, _element.createElement)(_inputControlStyles.Root, (0, _extends2.default)({}, props, getUIFlexProps({
    labelPosition: labelPosition
  }), {
    className: className,
    isFocused: isFocused,
    labelPosition: labelPosition,
    ref: ref,
    __unstableVersion: "next"
  }), (0, _element.createElement)(_inputControlStyles.LabelWrapper, null, (0, _element.createElement)(_label.default, {
    className: "components-input-control__label",
    hideLabelFromVision: hideLabelFromVision,
    labelPosition: labelPosition,
    htmlFor: id,
    size: size
  }, label)), (0, _element.createElement)(_inputControlStyles.Container, {
    __unstableInputWidth: __unstableInputWidth,
    className: "components-input-control__container",
    disabled: disabled,
    hideLabel: hideLabel,
    isFocused: isFocused,
    labelPosition: labelPosition
  }, prefix && (0, _element.createElement)(_inputControlStyles.Prefix, {
    className: "components-input-control__prefix"
  }, prefix), children, suffix && (0, _element.createElement)(_inputControlStyles.Suffix, {
    className: "components-input-control__suffix"
  }, suffix), (0, _element.createElement)(_backdrop.default, {
    "aria-hidden": "true",
    disabled: disabled,
    isFocused: isFocused,
    label: label,
    size: size
  })));
}

var _default = (0, _element.forwardRef)(InputBase);

exports.default = _default;
