// @wp-g2/components/dist/cjs/Dropdown/DropdownMenuCard.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _Card = require("../Card");

var _Scrollable = require("../Scrollable");

var styles = _interopRequireWildcard(require("./Dropdown.styles"));

/**
 * @param {import('@wp-g2/create-styles').ViewOwnProps<import('../Card/types').CardProps, 'div'>} props
 * @param {import('react').Ref<any>} forwardedRef
 */
function DropdownMenuCard(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'DropdownMenuCard'),
      children = _useContextSystem.children,
      className = _useContextSystem.className,
      _useContextSystem$ele = _useContextSystem.elevation,
      elevation = _useContextSystem$ele === void 0 ? 3 : _useContextSystem$ele,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["children", "className", "elevation"]);

  var classes = (0, _styles.cx)(styles.Card, className);
  return /*#__PURE__*/_react.default.createElement(_Card.Card, Object.assign({
    className: classes,
    elevation: elevation
  }, otherProps, {
    ref: forwardedRef
  }), /*#__PURE__*/_react.default.createElement(_Scrollable.Scrollable, {
    className: styles.Scrollable
  }, /*#__PURE__*/_react.default.createElement(_context.ContextSystemProvider, {
    value: {
      MenuItem: {
        isOffset: false
      }
    }
  }, children)));
}

var _default = (0, _context.contextConnect)(DropdownMenuCard, 'DropdownMenuCard');

exports.default = _default;
