// @wp-g2/components/dist/es/Animated/__stories__/AnimatedView.stories.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import faker from 'faker';
import { Schema } from 'faker-schema';
import React, { useState } from 'react';
import { Button, Card, CloseButton, Divider, Flex, Spacer, Text, View } from '../../index';
import { Animated, AnimatedContainer } from '../index';
export default {
  component: Animated,
  title: 'Components/Animated'
};
var itemSchema = new Schema(function () {
  return {
    id: faker.random.uuid(),
    name: faker.name.firstName()
  };
});

var Example = function () {
  var _useState = useState(itemSchema.make(5)),
      _useState2 = _slicedToArray(_useState, 2),
      items = _useState2[0],
      setItems = _useState2[1];

  var _useState3 = useState(true),
      _useState4 = _slicedToArray(_useState3, 2),
      isEditable = _useState4[0],
      setIsEditable = _useState4[1];

  var remove = function (id) {
    setItems(items.filter(function (item) {
      return item.id !== id;
    }));
  };

  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(View, {
    css: {
      margin: '20px auto',
      width: 300
    }
  }, /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(Flex, null, /*#__PURE__*/React.createElement(Button, {
    onClick: function add() {
      setItems([itemSchema.makeOne()].concat(items));
    }
  }, "Add Item"), /*#__PURE__*/React.createElement(Button, {
    onClick: function onClick() {
      return setIsEditable(!isEditable);
    },
    variant: "primary"
  }, isEditable ? 'Done' : 'Edit'))), /*#__PURE__*/React.createElement(Card, {
    css: {
      overflow: 'hidden'
    }
  }, /*#__PURE__*/React.createElement(AnimatedContainer, {
    initial: false
  }, items.map(function (item, index) {
    return /*#__PURE__*/React.createElement(Animated, {
      auto: true,
      key: item.id
    }, /*#__PURE__*/React.createElement(View, {
      css: {
        padding: '12px 8px',
        position: 'relative'
      }
    }, /*#__PURE__*/React.createElement(Animated, {
      animate: {
        paddingRight: isEditable ? 48 : 0
      },
      exit: {
        opacity: 0,
        x: -100
      }
    }, /*#__PURE__*/React.createElement(Text, null, item.name)), /*#__PURE__*/React.createElement(Animated, {
      animate: {
        opacity: isEditable ? 1 : 0
      },
      css: {
        pointerEvents: !isEditable ? 'none' : null,
        position: 'absolute',
        right: 8,
        top: 8
      },
      transition: {
        duration: 0.2
      }
    }, /*#__PURE__*/React.createElement(CloseButton, {
      onClick: function onClick() {
        return remove(item.id);
      },
      size: "small",
      variant: "tertiary"
    }))), index !== items.length - 1 && /*#__PURE__*/React.createElement(Divider, {
      m: 0
    }));
  })))));
};

export var _default = function () {
  return /*#__PURE__*/React.createElement(Example, null);
};
