// @wp-g2/components/dist/cjs/Collapsible/__stories__/Collapsible.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _Button = require("../../Button");

var _Placeholder = require("../../Placeholder");

var _Surface = require("../../Surface");

var _index = require("../index");

var _default2 = {
  component: _index.Collapsible,
  title: 'Components/Collapsible'
};
exports.default = _default2;

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(_index.Collapsible, {
    css: {
      height: 400,
      width: 300
    }
  }, /*#__PURE__*/_react.default.createElement(_index.CollapsibleTrigger, {
    as: _Button.Button
  }, "Toggle"), /*#__PURE__*/_react.default.createElement(_index.CollapsibleContent, null, /*#__PURE__*/_react.default.createElement(_Surface.Surface, null, /*#__PURE__*/_react.default.createElement(_Placeholder.Placeholder, {
    height: 500
  }))));
};

exports._default = _default;
