// @wp-g2/components/dist/cjs/Text/__stories__/Text.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.textLinkIcon = exports.highlight = exports.truncate = exports._default = exports.default = void 0;

var _icons = require("@wp-g2/icons");

var _react = _interopRequireDefault(require("react"));

var _index = require("../../index");

var _index2 = require("../index");

var _default2 = {
  component: _index2.Text,
  title: 'Components/Text'
};
exports.default = _default2;

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(_index2.Text, null, "Hello");
};

exports._default = _default;

var truncate = function () {
  return /*#__PURE__*/_react.default.createElement(_index2.Text, {
    numberOfLines: 2,
    truncate: true
  }, "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut facilisis dictum tortor, eu tincidunt justo scelerisque tincidunt. Duis semper dui id augue malesuada, ut feugiat nisi aliquam. Vestibulum venenatis diam sem, finibus dictum massa semper in. Nulla facilisi. Nunc vulputate faucibus diam, in lobortis arcu ornare vel. In dignissim nunc sed facilisis finibus. Etiam imperdiet mattis arcu, sed rutrum sapien blandit gravida. Aenean sollicitudin neque eget enim blandit, sit amet rutrum leo vehicula. Nunc malesuada ultricies eros ut faucibus. Aliquam erat volutpat. Nulla nec feugiat risus. Vivamus iaculis dui aliquet ante ultricies feugiat. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia curae; Vivamus nec pretium velit, sit amet consectetur ante. Praesent porttitor ex eget fermentum mattis.");
};

exports.truncate = truncate;

var highlight = function () {
  return /*#__PURE__*/_react.default.createElement(_index2.Text, {
    highlightWords: ['con']
  }, "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut facilisis dictum tortor, eu tincidunt justo scelerisque tincidunt. Duis semper dui id augue malesuada, ut feugiat nisi aliquam. Vestibulum venenatis diam sem, finibus dictum massa semper in. Nulla facilisi. Nunc vulputate faucibus diam, in lobortis arcu ornare vel. In dignissim nunc sed facilisis finibus. Etiam imperdiet mattis arcu, sed rutrum sapien blandit gravida. Aenean sollicitudin neque eget enim blandit, sit amet rutrum leo vehicula. Nunc malesuada ultricies eros ut faucibus. Aliquam erat volutpat. Nulla nec feugiat risus. Vivamus iaculis dui aliquet ante ultricies feugiat. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia curae; Vivamus nec pretium velit, sit amet consectetur ante. Praesent porttitor ex eget fermentum mattis.");
};

exports.highlight = highlight;

var textLinkIcon = function () {
  return /*#__PURE__*/_react.default.createElement(_index2.Text, null, "Lorem ipsum dolor sit amet, consectetur adipiscing elit.", ' ', /*#__PURE__*/_react.default.createElement(_index.Link, null, "Link"), /*#__PURE__*/_react.default.createElement(_index.Icon, {
    icon: /*#__PURE__*/_react.default.createElement(_icons.FiScissors, null),
    inline: true
  }), ".", ' ', /*#__PURE__*/_react.default.createElement(_index.Link, null, "Link ", /*#__PURE__*/_react.default.createElement(_index.Icon, {
    icon: /*#__PURE__*/_react.default.createElement(_icons.FiScissors, null),
    inline: true
  }), ' '));
};

exports.textLinkIcon = textLinkIcon;
