// @wp-g2/components/dist/cjs/select-dropdown/__stories__/index.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._preview = exports._inlineRendering = exports._default = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _Grid = require("../../Grid");

var _Text = require("../../Text");

var _View = require("../../View");

var _index = require("../index");

var _default2 = {
  component: _index.SelectDropdown,
  title: 'Components/SelectDropdown'
};
exports.default = _default2;
var options = [{
  name: 'Small',
  value: 'small',
  size: 11
}, {
  name: 'Medium',
  value: 'medium',
  size: 15
}, {
  name: 'Large',
  value: 'large',
  size: 21
}];

var _default = function () {
  var _React$useState = _react.default.useState(options[1]),
      _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
      value = _React$useState2[0],
      setValue = _React$useState2[1];

  var handleOnChange = function (next) {
    console.log(next.selectedItem);
    setValue(next.selectedItem);
  };

  var renderItem = _react.default.useCallback(function (_ref) {
    var name = _ref.name,
        size = _ref.size;
    return /*#__PURE__*/_react.default.createElement("div", {
      style: {
        fontSize: size
      }
    }, name);
  }, []);

  return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_Grid.Grid, null, /*#__PURE__*/_react.default.createElement(_index.SelectDropdown, {
    isPreviewable: true,
    onChange: handleOnChange,
    options: options,
    placeholder: "Element",
    renderItem: renderItem,
    value: value
  }), /*#__PURE__*/_react.default.createElement(_index.SelectDropdown, {
    onChange: handleOnChange,
    options: options,
    placeholder: "Element",
    renderItem: renderItem,
    value: value
  })));
};

exports._default = _default;

var _inlineRendering = function () {
  var _React$useState3 = _react.default.useState(options[1]),
      _React$useState4 = (0, _slicedToArray2.default)(_React$useState3, 2),
      value = _React$useState4[0],
      setValue = _React$useState4[1];

  var handleOnChange = function (next) {
    return setValue(next.selectedItem);
  };

  var renderItem = _react.default.useCallback(function (_ref2) {
    var name = _ref2.name,
        size = _ref2.size;
    return /*#__PURE__*/_react.default.createElement("div", {
      style: {
        fontSize: size
      }
    }, name);
  }, []);

  return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_Text.Text, {
    adjustLineHeightForInnerControls: true,
    size: value.size
  }, "A vacation is when you take a trip to some", ' ', /*#__PURE__*/_react.default.createElement(_index.SelectDropdown, {
    isInline: true,
    onChange: handleOnChange,
    options: options,
    placeholder: "Element",
    renderItem: renderItem,
    value: value
  }), ' ', "place with your. A vacation is when you take a trip to some", ' ', /*#__PURE__*/_react.default.createElement(_index.SelectDropdown, {
    isInline: true,
    onChange: handleOnChange,
    options: options,
    placeholder: "Element",
    renderItem: renderItem,
    value: value
  }), ' ', "place with your."));
};

exports._inlineRendering = _inlineRendering;
var presets = [{
  name: '0px',
  value: '0px',
  size: 0
}, {
  name: '4px',
  value: '4px',
  size: 4
}, {
  name: '8px',
  value: '8px',
  size: 8
}, {
  name: '12px',
  value: '12px',
  size: 12
}, {
  name: '16px',
  value: '16px',
  size: 16
}, {
  name: '20px',
  value: '20px',
  size: 20
}];

var _preview = function () {
  var _React$useState5 = _react.default.useState(presets[2]),
      _React$useState6 = (0, _slicedToArray2.default)(_React$useState5, 2),
      value = _React$useState6[0],
      setValue = _React$useState6[1];

  return /*#__PURE__*/_react.default.createElement(_Grid.Grid, null, /*#__PURE__*/_react.default.createElement(_index.SelectDropdown, {
    isPreviewable: true,
    onChange: function handleOnChange(next) {
      setValue(next.selectedItem);
    },
    options: presets,
    placeholder: "Element",
    value: value
  }), /*#__PURE__*/_react.default.createElement(_View.View, null, /*#__PURE__*/_react.default.createElement(_View.View, {
    css: {
      background: 'rgba(0, 0, 255, 0.1)',
      border: '2px solid rgba(0, 0, 255, 0.1)',
      display: 'inline-block'
    },
    style: {
      padding: value == null ? void 0 : value.size
    }
  }, /*#__PURE__*/_react.default.createElement(_View.View, {
    css: [{
      background: 'blue',
      color: 'white',
      width: 50,
      height: 50
    }, _styles.ui.alignment.content.center]
  }, value.value))));
};

exports._preview = _preview;
