// @wp-g2/components/dist/es/Panel/__stories__/Panel.stories.js
import React from 'react';
import { Grid, Spacer, Text, TextInput } from '../../index';
import { Panel, PanelBody, PanelHeader } from '../index';
export default {
  component: Panel,
  title: 'Components/Panel'
};

var InputSuffix = function (props) {
  return /*#__PURE__*/React.createElement(Text, Object.assign({
    css: {
      userSelect: 'none'
    },
    isBlock: true,
    lineHeight: 1,
    size: 11,
    variant: "muted"
  }, props));
};

export var _default = function () {
  return /*#__PURE__*/React.createElement(Panel, {
    css: {
      width: 300
    },
    visible: true
  }, /*#__PURE__*/React.createElement(PanelHeader, {
    title: "Title"
  }), /*#__PURE__*/React.createElement(PanelBody, null, /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(Grid, {
    columns: 2
  }, /*#__PURE__*/React.createElement(TextInput, {
    suffix: /*#__PURE__*/React.createElement(InputSuffix, null, "W"),
    value: 300
  }), /*#__PURE__*/React.createElement(TextInput, {
    suffix: /*#__PURE__*/React.createElement(InputSuffix, null, "H"),
    value: 200
  }))), /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(Grid, {
    columns: 3
  }, /*#__PURE__*/React.createElement(TextInput, {
    suffix: /*#__PURE__*/React.createElement(InputSuffix, null, "X")
  }), /*#__PURE__*/React.createElement(TextInput, {
    suffix: /*#__PURE__*/React.createElement(InputSuffix, null, "Y")
  }), /*#__PURE__*/React.createElement(TextInput, {
    suffix: /*#__PURE__*/React.createElement(InputSuffix, null, "Z")
  }))), /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(Grid, {
    templateColumns: "2fr 1fr"
  }, /*#__PURE__*/React.createElement(TextInput, null), /*#__PURE__*/React.createElement(TextInput, {
    suffix: /*#__PURE__*/React.createElement(InputSuffix, null, "%")
  })))));
};
