// @wordpress/components/build/ui/scrollable/styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.scrollAuto = exports.scrollY = exports.scrollX = exports.smoothScroll = exports.Content = exports.Scrollable = exports.scrollableScrollbar = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _styles = require("@wp-g2/styles");

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7;

var scrollableScrollbar = (0, _styles.css)(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2.default)(["\n\t@media only screen and ( min-device-width: 40em ) {\n\t\t&::-webkit-scrollbar {\n\t\t\theight: 12px;\n\t\t\twidth: 12px;\n\t\t}\n\n\t\t&::-webkit-scrollbar-track {\n\t\t\tbackground-color: transparent;\n\t\t}\n\n\t\t&::-webkit-scrollbar-track {\n\t\t\tbackground: ", ";\n\t\t\tborder-radius: 8px;\n\t\t}\n\n\t\t&::-webkit-scrollbar-thumb {\n\t\t\tbackground-clip: padding-box;\n\t\t\tbackground-color: ", ";\n\t\t\tborder: 2px solid rgba( 0, 0, 0, 0 );\n\t\t\tborder-radius: 7px;\n\t\t}\n\n\t\t&:hover::-webkit-scrollbar-thumb {\n\t\t\tbackground-color: ", ";\n\t\t}\n\t}\n"])), _styles.ui.get('colorScrollbarTrack'), _styles.ui.get('colorScrollbarThumb'), _styles.ui.get('colorScrollbarThumbHover'));
exports.scrollableScrollbar = scrollableScrollbar;
var Scrollable = (0, _styles.css)(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2.default)(["\n\theight: 100%;\n"])));
exports.Scrollable = Scrollable;
var Content = (0, _styles.css)(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2.default)(["\n\tposition: relative;\n"])));
exports.Content = Content;
var smoothScroll = (0, _styles.css)(_templateObject4 || (_templateObject4 = (0, _taggedTemplateLiteral2.default)(["\n\tscroll-behavior: smooth;\n"])));
exports.smoothScroll = smoothScroll;
var scrollX = (0, _styles.css)(_templateObject5 || (_templateObject5 = (0, _taggedTemplateLiteral2.default)(["\n\toverflow-x: auto;\n\toverflow-y: hidden;\n"])));
exports.scrollX = scrollX;
var scrollY = (0, _styles.css)(_templateObject6 || (_templateObject6 = (0, _taggedTemplateLiteral2.default)(["\n\toverflow-x: hidden;\n\toverflow-y: auto;\n"])));
exports.scrollY = scrollY;
var scrollAuto = (0, _styles.css)(_templateObject7 || (_templateObject7 = (0, _taggedTemplateLiteral2.default)(["\n\toverflow-y: auto;\n"])));
exports.scrollAuto = scrollAuto;
