// @wp-g2/components/dist/cjs/Tabs/Tabs.styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TabIndicatorView = exports.TabPanelView = exports.xSmall = exports.small = exports.TabView = exports.TabListView = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _styles = require("@wp-g2/styles");

function _templateObject6() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tbackground: ", ";\n\tbottom: -1px;\n\theight: 2px;\n\tleft: 0;\n\tposition: absolute;\n\ttransition: all ", " ease;\n\tz-index: 1;\n"]);

  _templateObject6 = function () {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\toutline: none;\n"]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tmin-height: ", ";\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tmin-height: ", ";\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\talign-items: center;\n\tappearance: none;\n\tbackground: transparent;\n\tborder: none;\n\tcolor: ", ";\n\tcursor: pointer;\n\tflex: 1;\n\tjustify-content: center;\n\tmin-height: ", ";\n\toutline: none;\n\tpadding: 4px;\n\ttransition: color ", " linear;\n\n\t&[aria-selected='true'] {\n\t\tcolor: ", ";\n\t}\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tborder-bottom: 1px solid ", ";\n\tdisplay: flex;\n\toutline: none;\n\tposition: relative;\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var TabListView = _styles.styled.div(_templateObject(), _styles.ui.get('colorDivider'));

exports.TabListView = TabListView;

var TabView = _styles.styled.button(_templateObject2(), _styles.ui.color.text, _styles.ui.get('controlHeightXLarge'), _styles.ui.get('transitionDuration'), _styles.ui.color.admin);

exports.TabView = TabView;
var small = (0, _styles.css)(_templateObject3(), _styles.ui.get('controlHeightLarge'));
exports.small = small;
var xSmall = (0, _styles.css)(_templateObject4(), _styles.ui.get('controlHeight'));
exports.xSmall = xSmall;

var TabPanelView = _styles.styled.div(_templateObject5());

exports.TabPanelView = TabPanelView;

var TabIndicatorView = _styles.styled.div(_templateObject6(), _styles.ui.color.admin, _styles.ui.get('transitionDuration'));

exports.TabIndicatorView = TabIndicatorView;
