// @wp-g2/components/dist/es/Modal/__stories__/Modal.stories.js
import React from 'react';
import { Button } from '../../index';
import { Modal, ModalBody, ModalFooter, ModalHeader, ModalTrigger, useModalContext, useModalState } from '../index';
export default {
  component: Modal,
  title: 'Components/Modal'
};
var CloseButton = /*#__PURE__*/React.forwardRef(function (_ref, ref) {
  var _ref$children = _ref.children,
      children = _ref$children === void 0 ? 'Done' : _ref$children,
      _ref$variant = _ref.variant,
      variant = _ref$variant === void 0 ? 'primary' : _ref$variant;

  var _useModalContext = useModalContext(),
      dialog = _useModalContext.dialog;

  return /*#__PURE__*/React.createElement(Button, {
    onClick: dialog.hide,
    ref: ref,
    variant: variant
  }, children);
});
export var _default = function () {
  return /*#__PURE__*/React.createElement(Modal, {
    trigger: /*#__PURE__*/React.createElement(ModalTrigger, null, "Open"),
    visible: true
  }, /*#__PURE__*/React.createElement(ModalHeader, {
    title: 'Modal Title'
  }), /*#__PURE__*/React.createElement(ModalBody, null, /*#__PURE__*/React.createElement("h2", null, "First"), /*#__PURE__*/React.createElement(Modal, {
    trigger: /*#__PURE__*/React.createElement(ModalTrigger, null, "Open")
  }, /*#__PURE__*/React.createElement(ModalHeader, {
    title: 'Inner Modal Title'
  }), /*#__PURE__*/React.createElement(ModalBody, null, /*#__PURE__*/React.createElement("h2", null, "Second"), /*#__PURE__*/React.createElement(Modal, {
    trigger: /*#__PURE__*/React.createElement(ModalTrigger, null, "Open")
  }, /*#__PURE__*/React.createElement(ModalHeader, {
    title: 'Third Inner Modal Title'
  }), /*#__PURE__*/React.createElement(ModalBody, null, /*#__PURE__*/React.createElement("h2", null, "Third")), /*#__PURE__*/React.createElement(ModalFooter, null, /*#__PURE__*/React.createElement(CloseButton, null)))))));
};
export var Controlled = function () {
  var dialog = useModalState();
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Button, {
    onClick: dialog.toggle
  }, "Secondary toggle"), /*#__PURE__*/React.createElement(Modal, {
    state: dialog,
    trigger: /*#__PURE__*/React.createElement(ModalTrigger, null, "Open")
  }, /*#__PURE__*/React.createElement(ModalHeader, {
    title: 'Modal Title'
  }), /*#__PURE__*/React.createElement(ModalBody, null, /*#__PURE__*/React.createElement("h2", null, "Controlled Modal"))));
};
