// @wordpress/block-editor/build-module/components/inserter-listbox/group.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { forwardRef, useEffect, useState } from '@wordpress/element';
import { __ } from '@wordpress/i18n';
import { speak } from '@wordpress/a11y';

function InserterListboxGroup(props, ref) {
  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      shouldSpeak = _useState2[0],
      setShouldSpeak = _useState2[1];

  useEffect(function () {
    if (shouldSpeak) {
      speak(__('Use left and right arrow keys to move through blocks'));
    }
  }, [shouldSpeak]);
  return createElement("div", _extends({
    ref: ref,
    role: "listbox",
    "aria-orientation": "horizontal",
    onFocus: function onFocus() {
      setShouldSpeak(true);
    },
    onBlur: function onBlur(event) {
      var focusingOutsideGroup = !event.currentTarget.contains(event.relatedTarget);

      if (focusingOutsideGroup) {
        setShouldSpeak(false);
      }
    }
  }, props));
}

export default forwardRef(InserterListboxGroup);
