// @wp-g2/components/dist/es/UnitInput/__stories__/UnitInput.stories.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import { isValidCSSValueForProp } from '@wp-g2/utils';
import React from 'react';
import createStore from 'zustand';
import { FormGroup, ListGroup } from '../../index';
import { UnitInput } from '../index';
import { validStyleProps } from './utils';
export default {
  component: UnitInput,
  title: 'Components/UnitInput'
};
var everythingInitialState = validStyleProps.filter(Boolean).reduce(function (state, key) {
  var _objectSpread2;

  var initialValue = 1;

  if (isValidCSSValueForProp(key, '1px')) {
    initialValue = '1px';
  }

  return _objectSpread(_objectSpread({}, state), {}, (_objectSpread2 = {}, _objectSpread2[key] = initialValue, _objectSpread2));
}, {});
var valueStore = createStore(function (set) {
  return _objectSpread(_objectSpread({}, everythingInitialState), {}, {
    setState: set
  });
});
var useValueStore = valueStore;
var CSSEntry = /*#__PURE__*/React.memo(function (_ref) {
  var prop = _ref.prop;
  var value = useValueStore(function (state) {
    return state[prop];
  });
  return /*#__PURE__*/React.createElement(FormGroup, {
    horizontal: true,
    label: prop
  }, /*#__PURE__*/React.createElement(UnitInput, {
    cssProp: prop,
    onChange: function update(next) {
      var _valueStore$setState;

      valueStore.setState((_valueStore$setState = {}, _valueStore$setState[prop] = next, _valueStore$setState));
    },
    validate: function validate(next) {
      return isValidCSSValueForProp(prop, next);
    },
    value: value
  }));
});

var Everything = function () {
  var _useValueStore = useValueStore(),
      setState = _useValueStore.setState,
      state = _objectWithoutProperties(_useValueStore, ["setState"]);

  return /*#__PURE__*/React.createElement(ListGroup, {
    css: "\n\t\t\t\tpadding-bottom: 20vh;\n\t\t\t"
  }, Object.entries(state).map(function (_ref2) {
    var _ref3 = _slicedToArray(_ref2, 1),
        key = _ref3[0];

    return /*#__PURE__*/React.createElement(CSSEntry, {
      key: key,
      prop: key
    });
  }));
};

var Example = function () {
  var _React$useState = React.useState('auto'),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      value = _React$useState2[0],
      setValue = _React$useState2[1];

  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(UnitInput, {
    onChange: setValue,
    value: value
  })), /*#__PURE__*/React.createElement("br", null), /*#__PURE__*/React.createElement("br", null), /*#__PURE__*/React.createElement("br", null), /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(UnitInput, {
    arrows: true,
    cssProp: "height",
    onChange: setValue,
    value: value
  })));
};

export var _default = function () {
  return /*#__PURE__*/React.createElement(Example, null);
};
export var _everything = function () {
  return /*#__PURE__*/React.createElement(Everything, null);
};
