// @wp-g2/components/dist/cjs/Animated/__stories__/AnimatedView.stories.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _faker = _interopRequireDefault(require("faker"));

var _fakerSchema = require("faker-schema");

var _react = _interopRequireWildcard(require("react"));

var _index = require("../../index");

var _index2 = require("../index");

var _default2 = {
  component: _index2.Animated,
  title: 'Components/Animated'
};
exports.default = _default2;
var itemSchema = new _fakerSchema.Schema(function () {
  return {
    id: _faker.default.random.uuid(),
    name: _faker.default.name.firstName()
  };
});

var Example = function () {
  var _useState = (0, _react.useState)(itemSchema.make(5)),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      items = _useState2[0],
      setItems = _useState2[1];

  var _useState3 = (0, _react.useState)(true),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      isEditable = _useState4[0],
      setIsEditable = _useState4[1];

  var remove = function (id) {
    setItems(items.filter(function (item) {
      return item.id !== id;
    }));
  };

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_index.View, {
    css: {
      margin: '20px auto',
      width: 300
    }
  }, /*#__PURE__*/_react.default.createElement(_index.Spacer, null, /*#__PURE__*/_react.default.createElement(_index.Flex, null, /*#__PURE__*/_react.default.createElement(_index.Button, {
    onClick: function add() {
      setItems([itemSchema.makeOne()].concat(items));
    }
  }, "Add Item"), /*#__PURE__*/_react.default.createElement(_index.Button, {
    onClick: function onClick() {
      return setIsEditable(!isEditable);
    },
    variant: "primary"
  }, isEditable ? 'Done' : 'Edit'))), /*#__PURE__*/_react.default.createElement(_index.Card, {
    css: {
      overflow: 'hidden'
    }
  }, /*#__PURE__*/_react.default.createElement(_index2.AnimatedContainer, {
    initial: false
  }, items.map(function (item, index) {
    return /*#__PURE__*/_react.default.createElement(_index2.Animated, {
      auto: true,
      key: item.id
    }, /*#__PURE__*/_react.default.createElement(_index.View, {
      css: {
        padding: '12px 8px',
        position: 'relative'
      }
    }, /*#__PURE__*/_react.default.createElement(_index2.Animated, {
      animate: {
        paddingRight: isEditable ? 48 : 0
      },
      exit: {
        opacity: 0,
        x: -100
      }
    }, /*#__PURE__*/_react.default.createElement(_index.Text, null, item.name)), /*#__PURE__*/_react.default.createElement(_index2.Animated, {
      animate: {
        opacity: isEditable ? 1 : 0
      },
      css: {
        pointerEvents: !isEditable ? 'none' : null,
        position: 'absolute',
        right: 8,
        top: 8
      },
      transition: {
        duration: 0.2
      }
    }, /*#__PURE__*/_react.default.createElement(_index.CloseButton, {
      onClick: function onClick() {
        return remove(item.id);
      },
      size: "small",
      variant: "tertiary"
    }))), index !== items.length - 1 && /*#__PURE__*/_react.default.createElement(_index.Divider, {
      m: 0
    }));
  })))));
};

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(Example, null);
};

exports._default = _default;
