// @wp-g2/components/dist/cjs/ExternalLink/ExternalLink.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

var _context = require("@wp-g2/context");

var _react = _interopRequireDefault(require("react"));

var _Icon = require("../Icon");

var _Link = require("../Link");

var _VisuallyHidden = require("../VisuallyHidden");

var styles = _interopRequireWildcard(require("./ExternalLink.styles"));

/**
 * @typedef {import('../Link').Props} Props
 */

/**
 *
 * @param {import('@wp-g2/create-styles').ViewOwnProps<Props, 'a'>} props
 * @param {import('react').Ref<any>} forwardedRef
 */
function ExternalLink(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'ExternalLink'),
      children = _useContextSystem.children,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["children"]);

  return /*#__PURE__*/_react.default.createElement(_Link.Link, Object.assign({
    ref: forwardedRef // eslint-disable-next-line react/jsx-no-target-blank
    ,
    target: "_blank"
  }, otherProps), /*#__PURE__*/_react.default.createElement(_VisuallyHidden.VisuallyHidden, null, (0, _i18n.__)('(opens in a new tab)')), children, /*#__PURE__*/_react.default.createElement(_Icon.Icon, {
    className: styles.Icon,
    fill: "currentColor",
    icon: _icons.external,
    inline: true,
    size: "15"
  }));
}

var _default = (0, _context.contextConnect)(ExternalLink, 'ExternalLink');

exports.default = _default;
