// @wp-g2/components/dist/cjs/Badge/Badge.styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getBackground = getBackground;
exports.getBackgroundText = getBackgroundText;
exports.rounded = exports.text = exports.truncate = exports.BadgeView = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _styles = require("@wp-g2/styles");

function _templateObject6() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t\t", "\n\t"]);

  _templateObject6 = function () {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t\t", "\n\t"]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\tjustify-content: center;\n\tmin-width: 18px;\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tpadding: 4px 0;\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tmax-width: 120px;\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tborder-radius: ", ";\n\tbox-shadow: 0 0 0 1px ", " inset;\n\tcursor: default;\n\tdisplay: flex;\n\theight: 18px;\n\tline-height: 1;\n\tpadding: 0 4px;\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var BadgeView = _styles.styled.div(_templateObject(), _styles.ui.get('controlBorderRadius'), _styles.ui.get('surfaceBorderColor'));

exports.BadgeView = BadgeView;
var truncate = (0, _styles.css)(_templateObject2());
exports.truncate = truncate;
var text = (0, _styles.css)(_templateObject3());
exports.text = text;
var rounded = (0, _styles.css)(_templateObject4(), _styles.ui.borderRadius.circle);
exports.rounded = rounded;

function getBackground(_ref) {
  var color = _ref.color,
      isBold = _ref.isBold;
  return (0, _styles.css)(_templateObject5(), (0, _styles.getBackgroundColor)(color, {
    isBold: isBold
  }));
}

function getBackgroundText(_ref2) {
  var color = _ref2.color,
      isBold = _ref2.isBold;
  return (0, _styles.css)(_templateObject6(), (0, _styles.getBackgroundColorText)(color, {
    isBold: isBold
  }));
}
