// @wp-g2/create-styles/dist/cjs/__stories__/styledComponentInterpolation.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _react = _interopRequireDefault(require("react"));

var _index = require("../index");

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tcolor: #aaa;\n\t", ";\n\t", ";\n\tfont-size: 40px;\n\tfont-weight: bold;\n\tpadding: 20px;\n\t", "\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var _default2 = {
  title: 'create-styles/StyledComponent/StringInterpolation'
};
exports.default = _default2;

var _createStyleSystem = (0, _index.createStyleSystem)({
  baseStyles: {
    boxSizing: 'border-box',
    fontFamily: 'Arial'
  }
}),
    View = _createStyleSystem.View,
    styled = _createStyleSystem.styled;

var happyMixin = function (props) {
  return props.happy ? "color: green" : "color: blue";
};

var niceMixin = function (props) {
  return props.nice ? "background-color: yellow" : "background-color: #ddd";
};

var FriendlyExample = styled.div(_templateObject(), happyMixin, niceMixin, function (props) {
  return props.title && "font-size: 90px";
});

var App = function () {
  return /*#__PURE__*/_react.default.createElement(View, null, /*#__PURE__*/_react.default.createElement(FriendlyExample, {
    happy: false,
    nice: false,
    title: "Weee"
  }, "Hello"));
};

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(App, null);
};

exports._default = _default;
