// @wp-g2/components/dist/cjs/Flex/__stories__/Flex.stories.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._default = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _index = require("../../index");

var _index2 = require("../index");

var _default2 = {
  component: _index2.Flex,
  title: 'Components/Flex'
};
exports.default = _default2;

var ItemView = function (props) {
  return /*#__PURE__*/_react.default.createElement(_index.Placeholder, props);
};

var _default = function () {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_index.Spacer, {
    mb: 4
  }, /*#__PURE__*/_react.default.createElement(_index2.Flex, {
    gap: 3
  }, /*#__PURE__*/_react.default.createElement(ItemView, null, "Item"), /*#__PURE__*/_react.default.createElement(ItemView, null, "Item"))), /*#__PURE__*/_react.default.createElement(_index2.Flex, {
    direction: ['column', 'row'],
    gap: 3
  }, /*#__PURE__*/_react.default.createElement(ItemView, {
    css: {
      width: '180px'
    }
  }, "Item"), /*#__PURE__*/_react.default.createElement(_index2.FlexItem, {
    isBlock: true
  }, /*#__PURE__*/_react.default.createElement(ItemView, null, "Item")), /*#__PURE__*/_react.default.createElement(ItemView, null, "Item"), /*#__PURE__*/_react.default.createElement(ItemView, null, "Item")));
};

exports._default = _default;
