// @wordpress/components/build/flex/next.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.withNextFlex = withNextFlex;
exports.withNextFlexItem = withNextFlexItem;
exports.withNextFlexBlock = withNextFlexBlock;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("../ui/context");

var _flex = require("../ui/flex");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var Flex = process.env.COMPONENT_SYSTEM_PHASE === 1 ? _flex.Flex : undefined;
var FlexBlock = process.env.COMPONENT_SYSTEM_PHASE === 1 ? _flex.FlexBlock : undefined;
var FlexItem = process.env.COMPONENT_SYSTEM_PHASE === 1 ? _flex.FlexItem : undefined;
/**
 * @param {import('./index').Props} props Current props.
 * @return {import('../ui/flex/types').FlexProps} Next props.
 */

var flexAdapter = function flexAdapter(_ref) {
  var isReversed = _ref.isReversed,
      restProps = (0, _objectWithoutProperties2.default)(_ref, ["isReversed"]);
  return _objectSpread(_objectSpread({
    // There's no equivalent for `direction` on the original component so we can just translate `isReversed` to it
    direction: isReversed ? 'row-reverse' : 'row'
  }, restProps), {}, {
    // There's an HTML attribute named `wrap` that will exist in `restProps` so we need to set it to undefined so the default behavior of the next component takes over
    wrap: undefined
  });
};
/**
 * @param {import('./item').Props} props Current props.
 * @return {import('../ui/flex/types').FlexItemProps} Next props.
 */


var flexItemAdapter = function flexItemAdapter(props) {
  return _objectSpread(_objectSpread({}, props), {}, {
    // ensure these are undefined so the default takes over
    isBlock: undefined,
    display: undefined
  });
};
/**
 * @param {import('./block').Props} props Current props.
 * @return {import('../ui/flex/types').FlexBlockProps} Next props.
 */


var flexBlockAdapter = function flexBlockAdapter(props) {
  return _objectSpread(_objectSpread({}, props), {}, {
    // ensure this is undefined so the default takes over
    display: undefined
  });
};
/* eslint-disable jsdoc/valid-types */

/**
 * @param {import('react').ForwardRefExoticComponent<import('./index').Props>} Current
 */

/* eslint-enable jsdoc/valid-types */


function withNextFlex(Current) {
  return (0, _context.withNext)(Current, Flex, 'WPComponentsFlex', flexAdapter);
}
/* eslint-disable jsdoc/valid-types */

/**
 * @param {import('react').ForwardRefExoticComponent<import('./item').Props>} Current
 */

/* eslint-enable jsdoc/valid-types */


function withNextFlexItem(Current) {
  return (0, _context.withNext)(Current, FlexItem, 'WPComponentsFlex', flexItemAdapter);
}
/* eslint-disable jsdoc/valid-types */

/**
 * @param {import('react').ForwardRefExoticComponent<import('./block').Props>} Current
 */

/* eslint-enable jsdoc/valid-types */


function withNextFlexBlock(Current) {
  return (0, _context.withNext)(Current, FlexBlock, 'WPComponentsFlex', flexBlockAdapter);
}
