// @wp-g2/components/dist/es/__stories__/Controls.stories.js
import { alignCenter, alignLeft, alignRight, desktop, fullscreen, mobile, tablet } from '@wordpress/icons';
import { ContextSystemProvider } from '@wp-g2/context';
import { FiSquare } from '@wp-g2/icons';
import React from 'react';
import { Sidebar } from '../__fixtures__/components';
import { Button, Divider, Flex, FlexBlock, FormGroup, Grid, Icon, Panel, PanelBody, PanelHeader, SegmentedControl, Select, Slider, Spacer, Stepper, Switch, Text, TextInput, View, VStack } from '../index';
export default {
  title: 'Examples'
};

var SuffixLabel = function (props) {
  return /*#__PURE__*/React.createElement(Text, Object.assign({
    isBlock: true,
    size: 10,
    variant: "muted"
  }, props, {
    css: {
      userSelect: 'none'
    }
  }));
};

var ResponsiveControls = function () {
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Flex, {
    justify: "center"
  }, /*#__PURE__*/React.createElement(SegmentedControl, {
    options: [{
      label: /*#__PURE__*/React.createElement(Icon, {
        icon: mobile
      }),
      value: 'phone'
    }, {
      label: /*#__PURE__*/React.createElement(Icon, {
        icon: tablet
      }),
      value: 'tablet'
    }, {
      label: /*#__PURE__*/React.createElement(Icon, {
        icon: desktop
      }),
      value: 'desktop'
    }]
  })));
};

var LayoutControls = function () {
  return /*#__PURE__*/React.createElement(Panel, {
    visible: true
  }, /*#__PURE__*/React.createElement(PanelHeader, {
    title: "Layout"
  }), /*#__PURE__*/React.createElement(PanelBody, null, /*#__PURE__*/React.createElement(VStack, null, /*#__PURE__*/React.createElement(FormGroup, {
    label: "Dimensions"
  }, /*#__PURE__*/React.createElement(Grid, {
    columns: 2
  }, /*#__PURE__*/React.createElement(TextInput, {
    suffix: /*#__PURE__*/React.createElement(SuffixLabel, null, "W")
  }), /*#__PURE__*/React.createElement(TextInput, {
    id: "d-h",
    suffix: /*#__PURE__*/React.createElement(SuffixLabel, null, "H")
  }))), /*#__PURE__*/React.createElement(FormGroup, {
    label: "Align"
  }, /*#__PURE__*/React.createElement(SegmentedControl, {
    isBlock: true,
    options: [{
      label: /*#__PURE__*/React.createElement(Icon, {
        icon: alignLeft
      }),
      value: 'left'
    }, {
      label: /*#__PURE__*/React.createElement(Icon, {
        icon: alignCenter
      }),
      value: 'center'
    }, {
      label: /*#__PURE__*/React.createElement(Icon, {
        icon: alignRight
      }),
      value: 'right'
    }]
  })), /*#__PURE__*/React.createElement(FormGroup, {
    label: "Position"
  }, /*#__PURE__*/React.createElement(Grid, {
    columns: 2
  }, /*#__PURE__*/React.createElement(TextInput, {
    suffix: /*#__PURE__*/React.createElement(SuffixLabel, null, "X")
  }), /*#__PURE__*/React.createElement(TextInput, {
    id: "pos-y",
    suffix: /*#__PURE__*/React.createElement(SuffixLabel, null, "Y")
  }))), /*#__PURE__*/React.createElement(FormGroup, {
    label: "Distribution"
  }, /*#__PURE__*/React.createElement(Select, {
    options: [{
      label: 'Center',
      value: 'center'
    }]
  })), /*#__PURE__*/React.createElement(FormGroup, {
    label: "Padding"
  }, /*#__PURE__*/React.createElement(View, null, /*#__PURE__*/React.createElement(Spacer, null, /*#__PURE__*/React.createElement(Grid, {
    columns: 2
  }, /*#__PURE__*/React.createElement(TextInput, {
    suffix: /*#__PURE__*/React.createElement(SuffixLabel, null, "PX")
  }), /*#__PURE__*/React.createElement(Stepper, {
    direction: "horizontal"
  }))), /*#__PURE__*/React.createElement(SegmentedControl, {
    isBlock: true,
    options: [{
      label: /*#__PURE__*/React.createElement(Icon, {
        icon: /*#__PURE__*/React.createElement(FiSquare, null)
      }),
      // @todo this is missing a suitable wp/icons replacement
      value: 'all'
    }, {
      label: /*#__PURE__*/React.createElement(Icon, {
        icon: fullscreen
      }),
      value: 'sides'
    }]
  }))), /*#__PURE__*/React.createElement(FormGroup, {
    label: "Overflow"
  }, /*#__PURE__*/React.createElement(SegmentedControl, {
    isBlock: true,
    options: [{
      label: 'Show',
      value: 'show'
    }, {
      label: 'Hide',
      value: 'hide'
    }]
  })))));
};

var TypographyControls = function () {
  return /*#__PURE__*/React.createElement(Panel, {
    visible: true
  }, /*#__PURE__*/React.createElement(PanelHeader, {
    title: "Typography"
  }), /*#__PURE__*/React.createElement(PanelBody, null, /*#__PURE__*/React.createElement(VStack, null, /*#__PURE__*/React.createElement(FormGroup, {
    label: "Font"
  }, /*#__PURE__*/React.createElement(Select, {
    options: [{
      label: 'Inter',
      value: 'inter'
    }]
  })), /*#__PURE__*/React.createElement(FormGroup, {
    label: "Style"
  }, /*#__PURE__*/React.createElement(Select, {
    options: [{
      label: 'Medium',
      value: 'medium'
    }]
  })), /*#__PURE__*/React.createElement(FormGroup, {
    label: "Size"
  }, /*#__PURE__*/React.createElement(Flex, {
    gap: 2
  }, /*#__PURE__*/React.createElement(FlexBlock, null, /*#__PURE__*/React.createElement(TextInput, {
    value: "16"
  })), /*#__PURE__*/React.createElement(FlexBlock, null, /*#__PURE__*/React.createElement(Slider, null)))), /*#__PURE__*/React.createElement(FormGroup, {
    label: "Transform"
  }, /*#__PURE__*/React.createElement(Select, {
    options: [{
      label: 'None',
      value: 'none'
    }]
  })), /*#__PURE__*/React.createElement(FormGroup, {
    label: "Dropcap"
  }, /*#__PURE__*/React.createElement(Flex, {
    justify: "flex-end"
  }, /*#__PURE__*/React.createElement(Switch, null))))));
};

var ActionsControl = function () {
  return /*#__PURE__*/React.createElement(Panel, {
    visible: true
  }, /*#__PURE__*/React.createElement(PanelHeader, {
    title: "Actions"
  }), /*#__PURE__*/React.createElement(PanelBody, null, /*#__PURE__*/React.createElement(Grid, {
    columns: 2
  }, /*#__PURE__*/React.createElement(Button, null, "Save Draft"), /*#__PURE__*/React.createElement(Button, {
    variant: "primary"
  }, "Publish"))));
};

export var Controls = function () {
  return /*#__PURE__*/React.createElement(ContextSystemProvider, {
    value: {
      FormGroup: {
        horizontal: true
      }
    }
  }, /*#__PURE__*/React.createElement(Sidebar, null, /*#__PURE__*/React.createElement(Spacer, null), /*#__PURE__*/React.createElement(ResponsiveControls, null), /*#__PURE__*/React.createElement(Divider, {
    mb: 0,
    mt: 2
  }), /*#__PURE__*/React.createElement(LayoutControls, null), /*#__PURE__*/React.createElement(TypographyControls, null), /*#__PURE__*/React.createElement(ActionsControl, null)));
};
