// @wordpress/components/build/mobile/bottom-sheet/stepper-cell/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _reactNative = require("react-native");

var _i18n = require("@wordpress/i18n");

var _compose = require("@wordpress/compose");

var _cell = _interopRequireDefault(require("../cell"));

var _stepper = _interopRequireDefault(require("./stepper"));

var _style = _interopRequireDefault(require("./style.scss"));

var _rangeTextInput = _interopRequireDefault(require("../range-text-input"));

var _utils = require("../../utils");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var STEP_DELAY = 200;
var DEFAULT_STEP = 1;
var isIOS = _reactNative.Platform.OS === 'ios';

var BottomSheetStepperCell = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(BottomSheetStepperCell, _Component);

  var _super = _createSuper(BottomSheetStepperCell);

  function BottomSheetStepperCell(props) {
    var _this;

    (0, _classCallCheck2.default)(this, BottomSheetStepperCell);
    _this = _super.call(this, props);
    _this.announceValue = _this.announceValue.bind((0, _assertThisInitialized2.default)(_this));
    _this.onDecrementValue = _this.onDecrementValue.bind((0, _assertThisInitialized2.default)(_this));
    _this.onDecrementValuePressIn = _this.onDecrementValuePressIn.bind((0, _assertThisInitialized2.default)(_this));
    _this.onIncrementValue = _this.onIncrementValue.bind((0, _assertThisInitialized2.default)(_this));
    _this.onIncrementValuePressIn = _this.onIncrementValuePressIn.bind((0, _assertThisInitialized2.default)(_this));
    _this.onPressOut = _this.onPressOut.bind((0, _assertThisInitialized2.default)(_this));
    var value = props.value,
        defaultValue = props.defaultValue,
        min = props.min;
    var initialValue = value || defaultValue || min;
    _this.state = {
      inputValue: initialValue,
      stepperValue: initialValue
    };
    return _this;
  }

  (0, _createClass2.default)(BottomSheetStepperCell, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      clearTimeout(this.timeout);
      clearInterval(this.interval);
      clearTimeout(this.timeoutAnnounceValue);
    }
  }, {
    key: "onIncrementValue",
    value: function onIncrementValue() {
      var _this$props = this.props,
          step = _this$props.step,
          max = _this$props.max,
          onChange = _this$props.onChange,
          value = _this$props.value,
          decimalNum = _this$props.decimalNum;
      var newValue = (0, _utils.toFixed)(value + step, decimalNum);

      if (newValue <= max || max === undefined) {
        onChange(newValue);
        this.setState({
          inputValue: newValue
        });
        this.announceValue(newValue);
      }
    }
  }, {
    key: "onDecrementValue",
    value: function onDecrementValue() {
      var _this$props2 = this.props,
          step = _this$props2.step,
          min = _this$props2.min,
          onChange = _this$props2.onChange,
          value = _this$props2.value,
          decimalNum = _this$props2.decimalNum;
      var newValue = (0, _utils.toFixed)(value - step, decimalNum);

      if (newValue >= min) {
        onChange(newValue);
        this.setState({
          inputValue: newValue
        });
        this.announceValue(newValue);
      }
    }
  }, {
    key: "onIncrementValuePressIn",
    value: function onIncrementValuePressIn() {
      var _this2 = this;

      this.onIncrementValue();
      this.timeout = setTimeout(function () {
        _this2.startPressInterval(_this2.onIncrementValue);
      }, 500);
    }
  }, {
    key: "onDecrementValuePressIn",
    value: function onDecrementValuePressIn() {
      var _this3 = this;

      this.onDecrementValue();
      this.timeout = setTimeout(function () {
        _this3.startPressInterval(_this3.onDecrementValue);
      }, 500);
    }
  }, {
    key: "onPressOut",
    value: function onPressOut() {
      clearTimeout(this.timeout);
      clearInterval(this.interval);
    }
  }, {
    key: "startPressInterval",
    value: function startPressInterval(callback) {
      var _this4 = this;

      var speed = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : STEP_DELAY;
      var counter = 0;
      this.interval = setInterval(function () {
        callback();
        counter += 1;

        if (counter === 10) {
          clearInterval(_this4.interval);

          _this4.startPressInterval(callback, speed / 2);
        }
      }, speed);
    }
  }, {
    key: "announceValue",
    value: function announceValue(value) {
      var _this$props3 = this.props,
          label = _this$props3.label,
          _this$props3$unitLabe = _this$props3.unitLabel,
          unitLabel = _this$props3$unitLabe === void 0 ? '' : _this$props3$unitLabe;

      if (isIOS) {
        // On Android it triggers the accessibilityLabel with the value change
        clearTimeout(this.timeoutAnnounceValue);
        this.timeoutAnnounceValue = setTimeout(function () {
          _reactNative.AccessibilityInfo.announceForAccessibility("".concat(value, " ").concat(unitLabel, " ").concat(label));
        }, 300);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this5 = this;

      var _this$props4 = this.props,
          label = _this$props4.label,
          _this$props4$settingL = _this$props4.settingLabel,
          settingLabel = _this$props4$settingL === void 0 ? 'Value' : _this$props4$settingL,
          _this$props4$unitLabe = _this$props4.unitLabel,
          unitLabel = _this$props4$unitLabe === void 0 ? '' : _this$props4$unitLabe,
          icon = _this$props4.icon,
          min = _this$props4.min,
          max = _this$props4.max,
          value = _this$props4.value,
          separatorType = _this$props4.separatorType,
          children = _this$props4.children,
          _this$props4$shouldDi = _this$props4.shouldDisplayTextInput,
          shouldDisplayTextInput = _this$props4$shouldDi === void 0 ? false : _this$props4$shouldDi,
          preview = _this$props4.preview,
          onChange = _this$props4.onChange,
          openUnitPicker = _this$props4.openUnitPicker,
          decimalNum = _this$props4.decimalNum,
          cellContainerStyle = _this$props4.cellContainerStyle;
      var inputValue = this.state.inputValue;
      var isMinValue = value === min;
      var isMaxValue = value === max;
      var labelStyle = [_style.default.cellLabel, !icon ? _style.default.cellLabelNoIcon : {}];

      var getAccessibilityHint = function getAccessibilityHint() {
        return openUnitPicker ? (0, _i18n.__)('double-tap to change unit') : '';
      };

      var accessibilityLabel = (0, _i18n.sprintf)(
      /* translators: accessibility text. Inform about current value. %1$s: Control label %2$s: setting label (example: width), %3$s: Current value. %4$s: value measurement unit (example: pixels) */
      (0, _i18n.__)('%1$s. %2$s is %3$s %4$s.'), label, settingLabel, value, unitLabel);
      var containerStyle = [_style.default.rowContainer, isIOS ? _style.default.containerIOS : _style.default.containerAndroid];
      return (0, _element.createElement)(_reactNative.View, {
        accessible: true,
        accessibilityRole: "adjustable",
        accessibilityLabel: accessibilityLabel,
        accessibilityHint: getAccessibilityHint(),
        accessibilityActions: [{
          name: 'increment'
        }, {
          name: 'decrement'
        }, {
          name: 'activate'
        }],
        onAccessibilityAction: function onAccessibilityAction(event) {
          switch (event.nativeEvent.actionName) {
            case 'increment':
              _this5.onIncrementValue();

              break;

            case 'decrement':
              _this5.onDecrementValue();

              break;

            case 'activate':
              if (openUnitPicker) {
                openUnitPicker();
              }

              break;
          }
        }
      }, (0, _element.createElement)(_reactNative.View, {
        importantForAccessibility: "no-hide-descendants"
      }, (0, _element.createElement)(_cell.default, {
        accessible: false,
        cellContainerStyle: [_style.default.cellContainerStyle, preview && _style.default.columnContainer, cellContainerStyle],
        cellRowContainerStyle: preview ? containerStyle : _style.default.cellRowStyles,
        disabled: true,
        editable: false,
        icon: icon,
        label: label,
        labelStyle: labelStyle,
        leftAlign: true,
        separatorType: separatorType
      }, (0, _element.createElement)(_reactNative.View, {
        style: preview && containerStyle
      }, preview, (0, _element.createElement)(_stepper.default, {
        isMaxValue: isMaxValue,
        isMinValue: isMinValue,
        onPressInDecrement: this.onDecrementValuePressIn,
        onPressInIncrement: this.onIncrementValuePressIn,
        onPressOut: this.onPressOut,
        value: value,
        shouldDisplayTextInput: shouldDisplayTextInput
      }, shouldDisplayTextInput && (0, _element.createElement)(_rangeTextInput.default, {
        label: label,
        onChange: onChange,
        defaultValue: "".concat(inputValue),
        value: inputValue,
        min: min,
        step: 1,
        decimalNum: decimalNum
      }, children))))));
    }
  }]);
  return BottomSheetStepperCell;
}(_element.Component);

BottomSheetStepperCell.defaultProps = {
  step: DEFAULT_STEP
};

var _default = (0, _compose.withPreferredColorScheme)(BottomSheetStepperCell);

exports.default = _default;
