// @wp-g2/components/dist/cjs/Truncate/useTruncate.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useTruncate = useTruncate;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _react = require("react");

var styles = _interopRequireWildcard(require("./Truncate.styles"));

var _Truncate2 = require("./Truncate.utils");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t\t\t-webkit-box-orient: vertical;\n\t\t\t-webkit-line-clamp: ", ";\n\t\t\tdisplay: -webkit-box;\n\t\t\toverflow: hidden;\n\t\t"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

/**
 * @typedef Props
 * @property {string} [ellipsis='...']
 * @property {'auto' | 'head' | 'tail' | 'middle' | 'none'} [ellipsizeMode='auto']
 * @property {number} [limit=0]
 * @property {number} [numberOfLines=0]
 */

/**
 * @param {import('@wp-g2/create-styles').ViewOwnProps<Props, 'span'>} props
 */
function useTruncate(props) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'Truncate'),
      className = _useContextSystem.className,
      children = _useContextSystem.children,
      _useContextSystem$ell = _useContextSystem.ellipsis,
      ellipsis = _useContextSystem$ell === void 0 ? _Truncate2.TRUNCATE_ELLIPSIS : _useContextSystem$ell,
      _useContextSystem$ell2 = _useContextSystem.ellipsizeMode,
      ellipsizeMode = _useContextSystem$ell2 === void 0 ? _Truncate2.TRUNCATE_TYPE.auto : _useContextSystem$ell2,
      _useContextSystem$lim = _useContextSystem.limit,
      limit = _useContextSystem$lim === void 0 ? 0 : _useContextSystem$lim,
      _useContextSystem$num = _useContextSystem.numberOfLines,
      numberOfLines = _useContextSystem$num === void 0 ? 0 : _useContextSystem$num,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["className", "children", "ellipsis", "ellipsizeMode", "limit", "numberOfLines"]);

  var truncatedContent = (0, _Truncate2.truncateContent)(typeof children === 'string' ? children : '', {
    ellipsis: ellipsis,
    ellipsizeMode: ellipsizeMode,
    limit: limit
  });
  var shouldTruncate = ellipsizeMode === _Truncate2.TRUNCATE_TYPE.auto && ellipsizeMode !== _Truncate2.TRUNCATE_TYPE.none;
  var classes = (0, _react.useMemo)(function () {
    var sx = {};
    sx.numberOfLines = (0, _styles.css)(_templateObject(), numberOfLines);
    return (0, _styles.cx)(shouldTruncate && !numberOfLines && styles.Truncate, shouldTruncate && !!numberOfLines && sx.numberOfLines, className);
  }, [className, numberOfLines, shouldTruncate]);
  return _objectSpread(_objectSpread({}, otherProps), {}, {
    className: classes,
    children: truncatedContent
  });
}
