// @wordpress/components/build/ui/card/component.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _elevation = require("../elevation");

var _view = require("../view");

var styles = _interopRequireWildcard(require("./styles"));

var _hook = require("./hook");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

/**
 * @param {import('@wp-g2/create-styles').ViewOwnProps<import('./types').CardProps, 'div'>} props
 * @param {import('react').Ref<any>} forwardedRef
 */
function Card(props, forwardedRef) {
  var _useCard = (0, _hook.useCard)(props),
      children = _useCard.children,
      elevation = _useCard.elevation,
      _useCard$isRounded = _useCard.isRounded,
      isRounded = _useCard$isRounded === void 0 ? true : _useCard$isRounded,
      otherProps = (0, _objectWithoutProperties2.default)(_useCard, ["children", "elevation", "isRounded"]);

  var elevationBorderRadius = isRounded ? _styles.ui.get('cardBorderRadius') : 0;
  var elevationClassName = (0, _element.useMemo)(function () {
    return (0, _styles.css)({
      borderRadius: elevationBorderRadius
    });
  }, [elevationBorderRadius]);
  return (0, _element.createElement)(_view.View, (0, _extends2.default)({}, otherProps, {
    ref: forwardedRef
  }), (0, _element.createElement)(_view.View, (0, _extends2.default)({}, _styles.ui.$('CardContent'), {
    css: styles.Content
  }), children), (0, _element.createElement)(_elevation.Elevation, (0, _extends2.default)({
    className: elevationClassName,
    isInteractive: false,
    value: elevation ? 1 : 0
  }, _styles.ui.$('CardElevation'))), (0, _element.createElement)(_elevation.Elevation, (0, _extends2.default)({
    className: elevationClassName,
    isInteractive: false,
    value: elevation
  }, _styles.ui.$('CardElevation'))));
}
/**
 * `Card` is a layout component, providing a flexible and extensible content container.
 *
 * `Card` provides convenient sub-components such as `CardBody`, `CardHeader`, and `CardFooter`.
 *
 * @example
 * ```jsx
 * import {
 *   Card,
 *   CardHeader,
 *   CardBody,
 *   CardFooter,
 *   Text,
 * } from `@wordpress/components/ui`;
 *
 * function Example() {
 *   return (
 *     <Card>
 *       <CardHeader>
 *         <Heading size={ 4 }>Card Title</Heading>
 *       </CardHeader>
 *       <CardBody>
 *         <Text>Card Content</Text>
 *       </CardBody>
 *       <CardFooter>
 *         <Text>Card Footer</Text>
 *       </CardFooter>
 *     </Card>
 *   );
 * }
 * ```
 */


var ConnectedCard = (0, _context.contextConnect)(Card, 'Card');
var _default = ConnectedCard;
exports.default = _default;
