// @wordpress/components/build/ui/card/styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.xSmall = exports.medium = exports.small = exports.rounded = exports.borderless = exports.borderRadius = exports.headerFooter = exports.InnerBody = exports.Body = exports.Content = exports.Header = exports.Card = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _styles = require("@wp-g2/styles");

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7, _templateObject8, _templateObject9, _templateObject10, _templateObject11, _templateObject12;

var Card = (0, _styles.css)(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2.default)(["\n\tbox-shadow: 0 0 0 1px ", ";\n\toutline: none;\n"])), _styles.ui.get('surfaceBorderColor'));
exports.Card = Card;
var Header = (0, _styles.css)(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2.default)(["\n\tborder-bottom: 1px solid;\n"])));
exports.Header = Header;
var Content = (0, _styles.css)(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2.default)(["\n\theight: 100%;\n"])));
exports.Content = Content;
var Body = (0, _styles.css)(_templateObject4 || (_templateObject4 = (0, _taggedTemplateLiteral2.default)(["\n\theight: auto;\n\tmax-height: 100%;\n\tpadding: ", ";\n"])), _styles.ui.get('cardPadding'));
exports.Body = Body;
var InnerBody = (0, _styles.css)(_templateObject5 || (_templateObject5 = (0, _taggedTemplateLiteral2.default)(["\n\tmargin-left: calc( ", " * -1 );\n\tmargin-right: calc( ", " * -1 );\n"])), _styles.ui.get('cardPadding'), _styles.ui.get('cardPadding'));
exports.InnerBody = InnerBody;
var headerFooter = (0, _styles.css)(_templateObject6 || (_templateObject6 = (0, _taggedTemplateLiteral2.default)(["\n\tborder-color: ", ";\n\tmin-height: ", ";\n\tpadding-bottom: ", ";\n\tpadding-left: ", ";\n\tpadding-right: ", ";\n\tpadding-top: ", ";\n"])), _styles.ui.get('colorDivider'), _styles.ui.get('cardHeaderHeight'), _styles.ui.get('cardHeaderFooterPaddingY'), _styles.ui.get('cardPaddingX'), _styles.ui.get('cardPaddingX'), _styles.ui.get('cardHeaderFooterPaddingY'));
exports.headerFooter = headerFooter;
var borderRadius = (0, _styles.css)(_templateObject7 || (_templateObject7 = (0, _taggedTemplateLiteral2.default)(["\n\t&:first-of-type {\n\t\tborder-top-left-radius: ", ";\n\t\tborder-top-right-radius: ", ";\n\t}\n\n\t&:last-of-type {\n\t\tborder-bottom-left-radius: ", ";\n\t\tborder-bottom-right-radius: ", ";\n\t}\n"])), _styles.ui.get('cardBorderRadius'), _styles.ui.get('cardBorderRadius'), _styles.ui.get('cardBorderRadius'), _styles.ui.get('cardBorderRadius'));
exports.borderRadius = borderRadius;
var borderless = (0, _styles.css)(_templateObject8 || (_templateObject8 = (0, _taggedTemplateLiteral2.default)(["\n\tbox-shadow: none;\n"])));
exports.borderless = borderless;
var rounded = (0, _styles.css)(_templateObject9 || (_templateObject9 = (0, _taggedTemplateLiteral2.default)(["\n\tborder-radius: ", ";\n"])), _styles.ui.get('cardBorderRadius'));
exports.rounded = rounded;
var small = (0, _styles.css)(_templateObject10 || (_templateObject10 = (0, _taggedTemplateLiteral2.default)(["\n\tmin-height: 30px;\n"])));
exports.small = small;
var medium = (0, _styles.css)(_templateObject11 || (_templateObject11 = (0, _taggedTemplateLiteral2.default)([""])));
exports.medium = medium;
var xSmall = (0, _styles.css)(_templateObject12 || (_templateObject12 = (0, _taggedTemplateLiteral2.default)(["\n\tmin-height: 24px;\n"])));
exports.xSmall = xSmall;
