// @wp-g2/components/dist/cjs/Spinner/Spinner.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _react = _interopRequireDefault(require("react"));

var _Spinner = require("./Spinner.styles");

var _Spinner2 = require("./Spinner.utils");

/**
 * @typedef Props
 * @property {import('react').CSSProperties['color']} [color] Color of `Spinner`.
 * @property {number} [size=16] Size of `Spinner`.
 */

/**
 *
 * @param {import('@wp-g2/create-styles').ViewOwnProps<Props, 'div'>} props
 * @param {import('react').Ref<any>} forwardedRef
 */
function Spinner(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'Spinner'),
      _useContextSystem$col = _useContextSystem.color,
      color = _useContextSystem$col === void 0 ? (0, _styles.get)('colorText') : _useContextSystem$col,
      _useContextSystem$siz = _useContextSystem.size,
      size = _useContextSystem$siz === void 0 ? _Spinner2.BASE_SIZE : _useContextSystem$siz,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["color", "size"]);

  return /*#__PURE__*/_react.default.createElement(_Spinner.ContainerView, Object.assign({}, otherProps, {
    "aria-busy": true,
    ref: forwardedRef,
    style: {
      height: size,
      width: size
    }
  }), /*#__PURE__*/_react.default.createElement(_Spinner.BarsWrapperView, {
    "aria-hidden": true,
    style: {
      transform: "scale(" + size / _Spinner2.BASE_SIZE * _Spinner2.BASE_SIZE / _Spinner2.WRAPPER_SIZE + ")"
    }
  }, /*#__PURE__*/_react.default.createElement(_Spinner.BarsView, {
    style: {
      color: color
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "InnerBar1"
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "InnerBar2"
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "InnerBar3"
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "InnerBar4"
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "InnerBar5"
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "InnerBar6"
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "InnerBar7"
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "InnerBar8"
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "InnerBar9"
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "InnerBar10"
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "InnerBar11"
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "InnerBar12"
  }))));
}

var _default = (0, _context.contextConnect)(Spinner, 'Spinner');

exports.default = _default;
