// @wp-g2/components/dist/cjs/RadioGroup/RadioGroup.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _i18n = require("@wordpress/i18n");

var _context = require("@wp-g2/context");

var _react = _interopRequireDefault(require("react"));

var _reakit = require("reakit");

var _View = require("../View");

var _RadioGroup = require("./RadioGroup.Context");

function RadioGroup(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'RadioGroup'),
      baseId = _useContextSystem.baseId,
      children = _useContextSystem.children,
      id = _useContextSystem.id,
      _useContextSystem$lab = _useContextSystem.label,
      label = _useContextSystem$lab === void 0 ? (0, _i18n.__)('RadioGroup') : _useContextSystem$lab,
      value = _useContextSystem.value,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["baseId", "children", "id", "label", "value"]);

  var radio = (0, _reakit.useRadioState)({
    baseId: baseId || id,
    state: value
  });

  var contextProps = _react.default.useMemo(function () {
    return {
      radio: radio
    };
  }, [radio]);

  return /*#__PURE__*/_react.default.createElement(_RadioGroup.RadioGroupContext.Provider, {
    value: contextProps
  }, /*#__PURE__*/_react.default.createElement(_reakit.RadioGroup, Object.assign({
    "aria-label": label,
    as: _View.View,
    ref: forwardedRef
  }, otherProps), children));
}

var _default = (0, _context.contextConnect)(RadioGroup, 'RadioGroup');

exports.default = _default;
