// @wp-g2/components/dist/cjs/select-dropdown/select-dropdown-styles.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.inline = exports.hidden = exports.DropdownMenu = exports.MenuWrapper = exports.popoverHidden = exports.Popover = exports.SelectDropdown = void 0;

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _styles = require("@wp-g2/styles");

var baseFieldStyles = _interopRequireWildcard(require("../BaseField/BaseField.styles"));

function _templateObject6() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tdisplay: none;\n"]);

  _templateObject6 = function () {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\twidth: 100%;\n"]);

  _templateObject5 = function () {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\toutline: none;\n"]);

  _templateObject4 = function () {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\topacity: 0;\n"]);

  _templateObject3 = function () {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\t", ";\n\topacity: 1;\n\toutline: none;\n\ttransition: 40ms opacity linear;\n\ttransition-delay: 20ms;\n"]);

  _templateObject2 = function () {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteralLoose2.default)(["\n\tposition: relative;\n"]);

  _templateObject = function () {
    return data;
  };

  return data;
}

var SelectDropdown = (0, _styles.css)(_templateObject());
exports.SelectDropdown = SelectDropdown;
var Popover = (0, _styles.css)(_templateObject2(), _styles.ui.zIndex('Dropdown'));
exports.Popover = Popover;
var popoverHidden = (0, _styles.css)(_templateObject3());
exports.popoverHidden = popoverHidden;
var MenuWrapper = (0, _styles.css)(_templateObject4());
exports.MenuWrapper = MenuWrapper;
var DropdownMenu = (0, _styles.css)(_templateObject5());
exports.DropdownMenu = DropdownMenu;
var hidden = (0, _styles.css)(_templateObject6());
exports.hidden = hidden;
var inline = baseFieldStyles.inline;
exports.inline = inline;
