// @wp-g2/components/dist/cjs/select-dropdown/select-dropdown-item.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _context = require("@wp-g2/context");

var _styles = require("@wp-g2/styles");

var _utils = require("@wp-g2/utils");

var _react = _interopRequireDefault(require("react"));

var _Menu = require("../Menu");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function SelectDropdownItem(props, forwardedRef) {
  var _useContextSystem = (0, _context.useContextSystem)(props, 'SelectDropdownItem'),
      className = _useContextSystem.className,
      index = _useContextSystem.index,
      isHighlighted = _useContextSystem.isHighlighted,
      isSelected = _useContextSystem.isSelected,
      label = _useContextSystem.label,
      name = _useContextSystem.name,
      renderItem = _useContextSystem.renderItem,
      otherProps = (0, _objectWithoutProperties2.default)(_useContextSystem, ["className", "index", "isHighlighted", "isSelected", "label", "name", "renderItem"]);

  var itemProps = _objectSpread({
    index: index,
    isHighlighted: isHighlighted,
    isSelected: isSelected,
    label: label,
    name: name
  }, otherProps);

  var content = (0, _utils.renderChildren)(renderItem || label || name, itemProps);
  return /*#__PURE__*/_react.default.createElement(_Menu.MenuItem, Object.assign({}, _styles.ui.$('SelectDropdownItem'), otherProps, {
    className: className,
    isSelected: isSelected,
    ref: forwardedRef
  }), content);
}

var _default = (0, _context.contextConnect)(SelectDropdownItem, 'SelectDropdownItem');

exports.default = _default;
