// @wordpress/block-editor/build-module/components/block-list/block.native.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import { createElement } from "@wordpress/element";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

/**
 * External dependencies
 */
import { View, Text, TouchableWithoutFeedback, Dimensions } from 'react-native';
import { pick } from 'lodash';
/**
 * WordPress dependencies
 */

import { Component, createRef, useMemo } from '@wordpress/element';
import { GlobalStylesContext, getMergedGlobalStyles, alignmentHelpers, useGlobalStyles } from '@wordpress/components';
import { withDispatch, withSelect } from '@wordpress/data';
import { compose, withPreferredColorScheme } from '@wordpress/compose';
import { getBlockType, __experimentalGetAccessibleBlockLabel as getAccessibleBlockLabel } from '@wordpress/blocks';
import { __experimentalUseEditorFeature as useEditorFeature } from '@wordpress/block-editor';
/**
 * Internal dependencies
 */

import styles from './block.scss';
import BlockEdit from '../block-edit';
import BlockInvalidWarning from './block-invalid-warning';
import BlockMobileToolbar from '../block-mobile-toolbar';
import { store as blockEditorStore } from '../../store';
var emptyArray = [];

function BlockForType(_ref) {
  var attributes = _ref.attributes,
      clientId = _ref.clientId,
      contentStyle = _ref.contentStyle,
      getBlockWidth = _ref.getBlockWidth,
      insertBlocksAfter = _ref.insertBlocksAfter,
      isSelected = _ref.isSelected,
      mergeBlocks = _ref.mergeBlocks,
      name = _ref.name,
      onBlockFocus = _ref.onBlockFocus,
      onChange = _ref.onChange,
      onDeleteBlock = _ref.onDeleteBlock,
      onReplace = _ref.onReplace,
      parentWidth = _ref.parentWidth,
      wrapperProps = _ref.wrapperProps,
      blockWidth = _ref.blockWidth;
  var defaultColors = useEditorFeature('color.palette') || emptyArray;
  var globalStyle = useGlobalStyles();
  var mergedStyle = useMemo(function () {
    return getMergedGlobalStyles(globalStyle, wrapperProps.style, attributes, defaultColors);
  }, [defaultColors, globalStyle, // I couldn't simply use attributes and wrapperProps.styles as a dependency because they are almost always a new reference.
  // Thanks to the JSON.stringify we check if the value is the same instead of reference.
  JSON.stringify(wrapperProps.style), JSON.stringify(pick(attributes, GlobalStylesContext.BLOCK_STYLE_ATTRIBUTES))]);
  return createElement(GlobalStylesContext.Provider, {
    value: mergedStyle
  }, createElement(BlockEdit, {
    name: name,
    isSelected: isSelected,
    attributes: attributes,
    setAttributes: onChange,
    onFocus: onBlockFocus,
    onReplace: onReplace,
    insertBlocksAfter: insertBlocksAfter,
    mergeBlocks: mergeBlocks // Block level styles
    ,
    wrapperProps: wrapperProps // inherited styles merged with block level styles
    ,
    mergedStyle: mergedStyle,
    clientId: clientId,
    parentWidth: parentWidth,
    contentStyle: contentStyle,
    onDeleteBlock: onDeleteBlock,
    blockWidth: blockWidth
  }), createElement(View, {
    onLayout: getBlockWidth
  }));
}

var BlockListBlock = /*#__PURE__*/function (_Component) {
  _inherits(BlockListBlock, _Component);

  var _super = _createSuper(BlockListBlock);

  function BlockListBlock() {
    var _this;

    _classCallCheck(this, BlockListBlock);

    _this = _super.apply(this, arguments);
    _this.insertBlocksAfter = _this.insertBlocksAfter.bind(_assertThisInitialized(_this));
    _this.onFocus = _this.onFocus.bind(_assertThisInitialized(_this));
    _this.getBlockWidth = _this.getBlockWidth.bind(_assertThisInitialized(_this));
    _this.state = {
      blockWidth: _this.props.blockWidth - 2 * _this.props.marginHorizontal
    };
    _this.anchorNodeRef = createRef();
    return _this;
  }

  _createClass(BlockListBlock, [{
    key: "onFocus",
    value: function onFocus() {
      var _this$props = this.props,
          firstToSelectId = _this$props.firstToSelectId,
          isSelected = _this$props.isSelected,
          onSelect = _this$props.onSelect;

      if (!isSelected) {
        onSelect(firstToSelectId);
      }
    }
  }, {
    key: "insertBlocksAfter",
    value: function insertBlocksAfter(blocks) {
      this.props.onInsertBlocks(blocks, this.props.order + 1);

      if (blocks[0]) {
        // focus on the first block inserted
        this.props.onSelect(blocks[0].clientId);
      }
    }
  }, {
    key: "getBlockWidth",
    value: function getBlockWidth(_ref2) {
      var nativeEvent = _ref2.nativeEvent;
      var layout = nativeEvent.layout;
      var blockWidth = this.state.blockWidth;
      var layoutWidth = Math.floor(layout.width);

      if (!blockWidth || !layoutWidth) {
        return;
      }

      if (blockWidth !== layoutWidth) {
        this.setState({
          blockWidth: layoutWidth
        });
      }
    }
  }, {
    key: "getBlockForType",
    value: function getBlockForType() {
      var blockWidth = this.state.blockWidth;
      return createElement(BlockForType, _extends({}, this.props, {
        onBlockFocus: this.onFocus,
        insertBlocksAfter: this.insertBlocksAfter,
        getBlockWidth: this.getBlockWidth,
        blockWidth: blockWidth
      }));
    }
  }, {
    key: "renderBlockTitle",
    value: function renderBlockTitle() {
      return createElement(View, {
        style: styles.blockTitle
      }, createElement(Text, null, "BlockType: ", this.props.name));
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          attributes = _this$props2.attributes,
          blockType = _this$props2.blockType,
          clientId = _this$props2.clientId,
          icon = _this$props2.icon,
          isSelected = _this$props2.isSelected,
          isValid = _this$props2.isValid,
          order = _this$props2.order,
          title = _this$props2.title,
          isDimmed = _this$props2.isDimmed,
          isTouchable = _this$props2.isTouchable,
          onDeleteBlock = _this$props2.onDeleteBlock,
          isStackedHorizontally = _this$props2.isStackedHorizontally,
          isParentSelected = _this$props2.isParentSelected,
          getStylesFromColorScheme = _this$props2.getStylesFromColorScheme,
          marginVertical = _this$props2.marginVertical,
          marginHorizontal = _this$props2.marginHorizontal,
          isInnerBlockSelected = _this$props2.isInnerBlockSelected,
          name = _this$props2.name;

      if (!attributes || !blockType) {
        return null;
      }

      var blockWidth = this.state.blockWidth;
      var align = attributes.align;
      var accessibilityLabel = getAccessibleBlockLabel(blockType, attributes, order + 1);
      var isFullWidth = alignmentHelpers.isFullWidth,
          isWider = alignmentHelpers.isWider,
          isContainerRelated = alignmentHelpers.isContainerRelated;
      var accessible = !(isSelected || isInnerBlockSelected);
      var screenWidth = Math.floor(Dimensions.get('window').width);
      var isScreenWidthEqual = blockWidth === screenWidth;
      var isScreenWidthWider = blockWidth < screenWidth;
      var isFullWidthToolbar = isFullWidth(align) || isScreenWidthEqual;
      return createElement(TouchableWithoutFeedback, {
        onPress: this.onFocus,
        accessible: accessible,
        accessibilityRole: 'button'
      }, createElement(View, {
        style: {
          flex: 1
        },
        accessibilityLabel: accessibilityLabel
      }, createElement(View, {
        pointerEvents: isTouchable ? 'auto' : 'box-only',
        accessibilityLabel: accessibilityLabel,
        style: [{
          marginVertical: marginVertical,
          marginHorizontal: marginHorizontal,
          flex: 1
        }, isDimmed && styles.dimmed]
      }, isSelected && createElement(View, {
        pointerEvents: "box-none",
        style: [styles.solidBorder, isFullWidth(align) && isScreenWidthWider && styles.borderFullWidth, isFullWidth(align) && isContainerRelated(name) && isScreenWidthWider && styles.containerBorderFullWidth, getStylesFromColorScheme(styles.solidBorderColor, styles.solidBorderColorDark)]
      }), isParentSelected && createElement(View, {
        style: [styles.dashedBorder, getStylesFromColorScheme(styles.dashedBorderColor, styles.dashedBorderColorDark)]
      }), isValid ? this.getBlockForType() : createElement(BlockInvalidWarning, {
        blockTitle: title,
        icon: icon
      }), createElement(View, {
        style: [styles.neutralToolbar, !isFullWidthToolbar && isContainerRelated(name) && isWider(screenWidth, 'mobile') && styles.containerToolbar],
        ref: this.anchorNodeRef
      }, isSelected && createElement(BlockMobileToolbar, {
        clientId: clientId,
        onDelete: onDeleteBlock,
        isStackedHorizontally: isStackedHorizontally,
        blockWidth: blockWidth,
        anchorNodeRef: this.anchorNodeRef.current,
        isFullWidth: isFullWidthToolbar
      })))));
    }
  }]);

  return BlockListBlock;
}(Component); // Helper function to memoize the wrapperProps since getEditWrapperProps always returns a new reference


var wrapperPropsCache = new WeakMap();
var emptyObj = {};

function getWrapperProps(value, getWrapperPropsFunction) {
  if (!getWrapperPropsFunction) {
    return emptyObj;
  }

  var cachedValue = wrapperPropsCache.get(value);

  if (!cachedValue) {
    var wrapperProps = getWrapperPropsFunction(value);
    wrapperPropsCache.set(value, wrapperProps);
    return wrapperProps;
  }

  return cachedValue;
}

export default compose([withSelect(function (select, _ref3) {
  var clientId = _ref3.clientId,
      rootClientId = _ref3.rootClientId;

  var _select = select(blockEditorStore),
      getBlockIndex = _select.getBlockIndex,
      isBlockSelected = _select.isBlockSelected,
      __unstableGetBlockWithoutInnerBlocks = _select.__unstableGetBlockWithoutInnerBlocks,
      getSelectedBlockClientId = _select.getSelectedBlockClientId,
      getLowestCommonAncestorWithSelectedBlock = _select.getLowestCommonAncestorWithSelectedBlock,
      getBlockParents = _select.getBlockParents,
      hasSelectedInnerBlock = _select.hasSelectedInnerBlock;

  var order = getBlockIndex(clientId, rootClientId);
  var isSelected = isBlockSelected(clientId);
  var isInnerBlockSelected = hasSelectedInnerBlock(clientId);

  var block = __unstableGetBlockWithoutInnerBlocks(clientId);

  var _ref4 = block || {},
      name = _ref4.name,
      attributes = _ref4.attributes,
      isValid = _ref4.isValid;

  var blockType = getBlockType(name || 'core/missing');
  var title = blockType.title;
  var icon = blockType.icon;
  var parents = getBlockParents(clientId, true);
  var parentId = parents[0] || '';
  var selectedBlockClientId = getSelectedBlockClientId();
  var commonAncestor = getLowestCommonAncestorWithSelectedBlock(clientId);
  var commonAncestorIndex = parents.indexOf(commonAncestor) - 1;
  var firstToSelectId = commonAncestor ? parents[commonAncestorIndex] : parents[parents.length - 1];
  var isParentSelected = // set false as a default value to prevent re-render when it's changed from null to false
  (selectedBlockClientId || false) && selectedBlockClientId === parentId;
  var selectedParents = selectedBlockClientId ? getBlockParents(selectedBlockClientId) : [];
  var isDescendantOfParentSelected = selectedParents.includes(parentId);
  var isTouchable = isSelected || isDescendantOfParentSelected || isParentSelected || parentId === '';
  return {
    icon: icon,
    name: name || 'core/missing',
    order: order,
    title: title,
    attributes: attributes,
    blockType: blockType,
    isSelected: isSelected,
    isInnerBlockSelected: isInnerBlockSelected,
    isValid: isValid,
    isParentSelected: isParentSelected,
    firstToSelectId: firstToSelectId,
    isTouchable: isTouchable,
    wrapperProps: getWrapperProps(attributes, blockType.getEditWrapperProps)
  };
}), withDispatch(function (dispatch, ownProps, _ref5) {
  var select = _ref5.select;

  var _dispatch = dispatch(blockEditorStore),
      insertBlocks = _dispatch.insertBlocks,
      _mergeBlocks = _dispatch.mergeBlocks,
      replaceBlocks = _dispatch.replaceBlocks,
      selectBlock = _dispatch.selectBlock,
      updateBlockAttributes = _dispatch.updateBlockAttributes;

  return {
    mergeBlocks: function mergeBlocks(forward) {
      var clientId = ownProps.clientId;

      var _select2 = select(blockEditorStore),
          getPreviousBlockClientId = _select2.getPreviousBlockClientId,
          getNextBlockClientId = _select2.getNextBlockClientId;

      if (forward) {
        var nextBlockClientId = getNextBlockClientId(clientId);

        if (nextBlockClientId) {
          _mergeBlocks(clientId, nextBlockClientId);
        }
      } else {
        var previousBlockClientId = getPreviousBlockClientId(clientId);

        if (previousBlockClientId) {
          _mergeBlocks(previousBlockClientId, clientId);
        }
      }
    },
    onInsertBlocks: function onInsertBlocks(blocks, index) {
      insertBlocks(blocks, index, ownProps.rootClientId);
    },
    onSelect: function onSelect() {
      var clientId = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : ownProps.clientId;
      var initialPosition = arguments.length > 1 ? arguments[1] : undefined;
      selectBlock(clientId, initialPosition);
    },
    onChange: function onChange(attributes) {
      updateBlockAttributes(ownProps.clientId, attributes);
    },
    onReplace: function onReplace(blocks, indexToSelect) {
      replaceBlocks([ownProps.clientId], blocks, indexToSelect);
    }
  };
}), withPreferredColorScheme])(BlockListBlock);
